# H2M
Author: Kexin Dong  
Date: Jan 4, 2024  

# h2m

[![PyPI - Version](https://img.shields.io/pypi/v/h2m.svg)](https://pypi.org/project/h2m)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/h2m.svg)](https://pypi.org/project/h2m)

-----

# Installation

```python
pip install -i https://test.pypi.org/simple/ h2m
```

H2M is a python package for precise modeling of human vairants in mice. See [Github](https://github.com/kexindon/H2M-py) for more information.


# Importing


```python
import h2m
import pandas as pd
```

# Loading data  
We should upload reference genome and GENCODE annotation data for both human and mouse, which could be directly downloaded from a public [dropbox](https://www.dropbox.com/scl/fo/1wtrnc9w6s9gemweuw2fv/h?rlkey=hli1z6tv096cjwit5oi6bwggg&dl=0).  
Both GRCh37 and GRCh38 human reference genomes are available. Upload the one that you are going to use.  


```python
path_h_ref, path_m_ref = 'h2m_data/Refgenome/GCF_000001405.25_GRCh37.p13_genomic.fna.gz', 'h2m_data/Refgenome/GCF_000001635.27_GRCm39_genomic.fna.gz'
# remember to replace the paths with yours
records_h, index_list_h = h2m.genome_loader(path_h_ref)
records_m, index_list_m  = h2m.genome_loader(path_m_ref)
```

It may take up to 3 minutes to load the ref genomes.


```python
path_h_anno, path_m_anno = 'h2m_data/Genecode/gencode_v19_GRCh37.db', 'h2m_data/Genecode/gencode_vm33_GRCm39.db'
# remember to replace the paths with yours
db_h, db_m = h2m.anno_loader(path_h_anno), h2m.anno_loader(path_m_anno)
```

A notebook file about how to generate the db file: *link of the ipynb file here in github*

# Query  
First of all, you can use H2M to query a human gene for the presence of mouse homologs.


```python
query_result = h2m.query('TP53')
```

    Query human gene: TP53;
    Mouse ortholog(s): Trp53;
    Homology type: one2one;
    Sequence Simalarity(%):77.3537.



```python
query_result
```




    [{'gene_name_m': 'Trp53',
      'gene_id_m': 'ENSMUSG00000059552',
      'homology_type': 'ortholog_one2one',
      'similarity': 77.3537}]



The output is a list of information for all the mouse ortholog(s) (if have; sometimes more than one).  
Each element is a dictionary of **mouse gene name**, **mouse gene id**, **homology type** (one to one/one to multiple/many to many), and **similarity of human and mouse gene in percentage**.


```python
h2m.query('U1')
```

    Query human gene: U1;
    Mouse ortholog(s): Gm24251,Gm24291,Gm24251;
    Homology type: one2many;
    Sequence Simalarity(%):50.0, 59.6591, 50.6944.





    [{'gene_name_m': 'Gm24251',
      'gene_id_m': 'ENSMUSG00000087949',
      'homology_type': 'ortholog_one2many',
      'similarity': 50.0},
     {'gene_name_m': 'Gm24291',
      'gene_id_m': 'ENSMUSG00000088258',
      'homology_type': 'ortholog_one2many',
      'similarity': 59.6591},
     {'gene_name_m': 'Gm24251',
      'gene_id_m': 'ENSMUSG00000087949',
      'homology_type': 'ortholog_one2many',
      'similarity': 50.6944}]




```python
h2m.query('TPT1P6')
```

    Error 7: The query human gene: TPT1P6 has no mouse ortholog.





    [None]



A print output is helpful in interactive tasks like a Jupyter Notebook. You can also turn off this.


```python
h2m.query('UBAC2', show = False)
```




    [{'gene_name_m': 'Ubac2',
      'gene_id_m': 'ENSMUSG00000041765',
      'homology_type': 'ortholog_one2one',
      'similarity': 89.2442}]



Except for gene names, both ENSEMBL gene id and transcript id are accepted to identify a human gene. You can use the **ty** parameter ('tx_id','gene_id' or 'name') to specify your input type, but this is totally optional.

Using gene id:


```python
query_result = h2m.query('ENSG00000141510')
```

    Query human gene: TP53;
    Mouse ortholog(s): Trp53;
    Homology type: one2one;
    Sequence Simalarity(%):77.3537.


Using transcript id. Should include a db annotation file with the same ref genome version.


```python
query_result = h2m.query('ENST00000269305.4', db=db_h, ty='tx_id')
```

    Query human gene: TP53;
    Mouse ortholog(s): Trp53;
    Homology type: one2one;
    Sequence Simalarity(%):77.3537.


# Get transcript ID

One gene may have different transcripts. For mutation modeling, it is important to specify one transcript. If you do not have this information in hand, you can use H2M to get it.

Again, both gene IDs and gene names are accepted as identificaitons for human and mouse genes.


```python
list_tx_id_h = h2m.get_tx_id('TP53', 'h', ver=37)
```

    Genome assembly: GRCh37;
    The canonical transcript is: ENST00000269305.4;
    You can choose from the 17 transcripts below for further analysis:
    (1)ENST00000269305.4 (2)ENST00000413465.2 (3)ENST00000359597.4 (4)ENST00000504290.1 (5)ENST00000510385.1 (6)ENST00000504937.1 (7)ENST00000455263.2 (8)ENST00000420246.2 (9)ENST00000445888.2 (10)ENST00000576024.1 (11)ENST00000509690.1 (12)ENST00000514944.1 (13)ENST00000574684.1 (14)ENST00000505014.1 (15)ENST00000508793.1 (16)ENST00000604348.1 (17)ENST00000503591.1
    



```python
list_tx_id_m = h2m.get_tx_id('ENSMUSG00000059552', 'm')
```

    Genome assembly: GRCm39;
    The canonical transcript is: ENSMUST00000108658.10;
    You can choose from the 6 transcripts below for further analysis:
    (1)ENSMUST00000108658.10 (2)ENSMUST00000171247.8 (3)ENSMUST00000005371.12 (4)ENSMUST00000147512.2 (5)ENSMUST00000108657.4 (6)ENSMUST00000130540.2
    


More information is offered except for a complete list of all transcripts.

- the chromosome


```python
list_tx_id_h[0]
```




    17



- the start and end location of the gene on the chromosome


```python
list_tx_id_h[1:3]
```




    [7565097, 7590856]



- the canonical transcript annotated by ENSEMBL database and used by major clinical datasets (e.g. AACR-GENIE)


```python
list_tx_id_h[3]
```




    'ENST00000269305.4'



- the complete list of transcripts starting with the canonical one


```python
list_tx_id_h[4]
```




    ['ENST00000269305.4',
     'ENST00000413465.2',
     'ENST00000359597.4',
     'ENST00000504290.1',
     'ENST00000510385.1',
     'ENST00000504937.1',
     'ENST00000455263.2',
     'ENST00000420246.2',
     'ENST00000445888.2',
     'ENST00000576024.1',
     'ENST00000509690.1',
     'ENST00000514944.1',
     'ENST00000574684.1',
     'ENST00000505014.1',
     'ENST00000508793.1',
     'ENST00000604348.1',
     'ENST00000503591.1']



The output is a list of information for all the mouse ortholog(s) (if have; sometimes more than one).  
Each element is a dictionary of **mouse gene name**, **mouse gene id**, **homology type** (one to one/one to multiple), and **similarity of human and mouse gene in percentage**.


# Modeling human variants in the mouse genome

## input

Now you can use H2M to model your human mutations of interest.  
You should have at least such information in hand:  
1. transcript id of the human gene
2. transcript id of the mouse gene
3. start location of human variants *on the chromosome*  
4. end location of human variants *on the chromosome*  
5. the alternative sequence *on the positive strand of the chromosome*  
6. mutation type (**str in ['SNP','DNP','ONP','INS','DEL'])
7. the version number of human ref genome

Usually, such information has been pretty well organized in processed sequencing data and is easy to get.  

## Usage

Taking *TP53* R273H (ENST00000269305.4:c.818G>A) as an example.


```python
tx_id_h, tx_id_m = list_tx_id_h[3], list_tx_id_m[3]
# use the canonical transcript
```


```python
model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7577120, 7577120, 'T', ty_h = 'SNP', ver = 37)
model_result
```

    /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
      warnings.warn(





    {'gene_name_h': 'TP53',
     'gene_id_h': 'ENSG00000141510.11',
     'tx_id_m': 'ENSMUST00000108658.10',
     'chr_h': 'chr17',
     'exon_num_h': 10,
     'strand_h': '-',
     'start_h': 7577120,
     'end_h': 7577120,
     'ref_seq_h': 'C',
     'alt_seq_h': 'T',
     'HGVSp_h': 'R273H',
     'classification_h': 'Missense',
     'classification_h_detail': None,
     'exon_h': 'E_7',
     'type_h': 'SNP',
     'status': True,
     'error': 0,
     'statement': 'This mutation can be originally modeled.',
     'gene_name_m': 'Trp53',
     'gene_id_m': 'ENSMUSG00000059552.14',
     'chr_m': 'chr11',
     'exon_num_m': 10,
     'strand_m': '+',
     'type_m': 'SNP',
     'classification_m': 'Missense',
     'exon_m': 'E_7',
     'start_m_ori': 69480434,
     'end_m_ori': 69480434,
     'ref_seq_m_ori': 'G',
     'alt_seq_m_ori': 'A',
     'HGVSp_m_ori': 'R270H',
     'start_m': 69480434,
     'end_m': 69480434,
     'ref_seq_m': 'G',
     'alt_seq_m': 'A',
     'HGVSp_m': 'R270H'}



We can see that this human mutaton can be originally modeled by introducing the same neucleotide alteration.

By setting **show_sequence = True**, we can output the sequences of the wild-type and mutated human gene, wild-type, originally-modeled, and alternatively-modeled (if exsist) mouse gene. (If it can be originally modeled, the new_seq_m_alt would be the same as new_seq_m_ori)


```python
model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7577120, 7577120, 'T', ty_h = 'SNP', ver = 37, show_sequence=True)
model_result
```

    /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
      warnings.warn(





    {'gene_name_h': 'TP53',
     'gene_id_h': 'ENSG00000141510.11',
     'tx_id_m': 'ENSMUST00000108658.10',
     'chr_h': 'chr17',
     'exon_num_h': 10,
     'strand_h': '-',
     'start_h': 7577120,
     'end_h': 7577120,
     'ref_seq_h': 'C',
     'alt_seq_h': 'T',
     'HGVSp_h': 'R273H',
     'classification_h': 'Missense',
     'classification_h_detail': None,
     'exon_h': 'E_7',
     'type_h': 'SNP',
     'status': True,
     'error': 0,
     'statement': 'This mutation can be originally modeled.',
     'gene_name_m': 'Trp53',
     'gene_id_m': 'ENSMUSG00000059552.14',
     'chr_m': 'chr11',
     'exon_num_m': 10,
     'strand_m': '+',
     'type_m': 'SNP',
     'classification_m': 'Missense',
     'exon_m': 'E_7',
     'start_m_ori': 69480434,
     'end_m_ori': 69480434,
     'ref_seq_m_ori': 'G',
     'alt_seq_m_ori': 'A',
     'HGVSp_m_ori': 'R270H',
     'start_m': 69480434,
     'end_m': 69480434,
     'ref_seq_m': 'G',
     'alt_seq_m': 'A',
     'HGVSp_m': 'R270H',
     'seq_h': Seq('ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACA...CTC'),
     'seq_m': Seq('ATGACTGCCATGGAGGAGTCACAGTCGGATATCAGCCTCGAGCTCCCTCTGAGC...TTA'),
     'new_seq_h': Seq('ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACA...CTC'),
     'new_seq_m_ori': Seq('ATGACTGCCATGGAGGAGTCACAGTCGGATATCAGCCTCGAGCTCCCTCTGAGC...TTA'),
     'new_seq_m_alt': Seq('ATGACTGCCATGGAGGAGTCACAGTCGGATATCAGCCTCGAGCTCCCTCTGAGC...TTA'),
     'human_idx': 817,
     'mouse_idx': 808}




```python
seq_h, seq_m, new_seq_h, new_seq_m = model_result['seq_h'], model_result['seq_m'], model_result['new_seq_h'],  model_result['new_seq_m_ori']
h2m.PrintAlign(seq_h, new_seq_h)
```

    ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCG-TGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGACATTCTCCACTTCTTGTTCCCCACTGACAGCCTCCCACCCCCATCTCTCCCTCCCCTGCCATTTTGGGTTTTGGGTCTTTGAACCCTTGCTTGCAATAGGTGTGCGTCAGAAGCACCCAGGACTTCCATTTGCTTTGTCCCGGGGCTCCACTGAACAAGTTGGCCTGCACTGGTGTTTTGTTGTGGGGAGGAGGATGGGGAGTAGGACATACCAGCTTAGATTTTAAGGTTTTTACTGTGAGGGATGTTTGGGAGATGTAAGAAATGTTCTTGCAGTTAAGGGTTAGTTTACAATCAGCCACATTCTAGGTAGGGGCCCACTTCACCGTACTAACCAGGGAAGCTGTCCCTCACTGTTGAATTTTCTCTAACTTCAAGGCCCATATCTGTGAAATGCTGGCATTTGCACCTACCTCACAGAGTGCATTGTGAGGGTTAATGAAATAATGTACATCTGGCCTTGAAACCACCTTTTATTACATGGGGTCTAGAACTTGACCCCCTTGAGGGTGCTTGTTCCCTCTCCCTGTTGGTCGGTGGGTTGGTAGTTTCTACAGTTGGGCAGCTGGTTAGGTAGAGGGAGTTGTCAAGTCTCTGCTGGCCCAGCCAAACCCTGTCTGACAACCTCTTGGTGAACCTTAGTACCTAAAAGGAAATCTCACCCCATCCCACACCCTGGAGGATTTCATCTCTTGTATATGATGATCTGGATCCACCAAGACTTGTTTTATGCTCAGGGTCAATTTCTTTTTTCTTTTTTTTTTTTTTTTTTCTTTTTCTTTGAGACTGGGTCTCGCTTTGTTGCCCAGGCTGGAGTGGAGTGGCGTGATCTTGGCTTACTGCAGCCTTTGCCTCCCCGGCTCGAGCAGTCCTGCCTCAGCCTCCGGAGTAGCTGGGACCACAGGTTCATGCCACCATGGCCAGCCAACTTTTGCATGTTTTGTAGAGATGGGGTCTCACAGTGTTGCCCAGGCTGGTCTCAAACTCCTGGGCTCAGGCGATCCACCTGTCTCAGCCTCCCAGAGTGCTGGGATTACAATTGTGAGCCACCACGTCCAGCTGGAAGGGTCAACATCTTTTACATTCTGCAAGCACATCTGCATTTTCACCCCACCCTTCCCCTCCTTCTCCCTTTTTATATCCCATTTTTATATCGATCTCTTATTTTACAATAAAACTTTGCTGCCACCTGTGTGTCTGAGGGGTGAACGCCAGTGCAGGCTACTGGGGTCAGCAGGTGCAGGGGTGAGTGAGGAGGTGCTGGGAAGCAGCCACCTGAGTCTGCAATGAGTGTGGGCTGGGGGGCCCAGTGCCCGGGTTCCGGGAGGGGAACAAAGGCTGGAGACTGGGTCAGTCTGCGGGCTGCATGACAACAAGGGAGGGGGTGGCTCCATTCATAACTCAGGAACCAACCGTCCCTCCTCCCCTCCGGCCACGGCTGGCACAAGGTTCTCTCCCTCCCCTGCTTCTAGGACTGGGCTGCTTCCCCCTCGGCAGCCTCTCACCAAGGATTACGGGATTTAAATGTCTGATTTAGCAAGGCTGAGCCTCCAGGGTGGCCATCTGCTCCATCAGAAAGTGGCAGGATACCTGGGTTCCCAAGGGGAACAGGGGTGGGTGCTACTGGATGGAGAGAGGCCAGTGGGAGGCCTGCTAGCCAGGGTCCCAGGAAAGTGGGGGCAGCTAAGGTAAGAGTAGGGGTGTGGGGCTAGGTCCTTCCCAGCATCCCCTCATCCTGGGCCTCATGCCAGGTAGCTGAATGAATTGAAGCTTTAAACTCTGCCAGGAAAACCTTTCAAAGGGCTTCTTGGGATAGGGAGGAGAGTCGGGTTGAGGAGCTCAGTACTGCCTGCCCATGCTCCTCAGGGCTGCTGGCTCCCAGGGAGGGGGGCTGGGAGCAGGCAGGCTCTTCCCCATCACCCACTGCTCTCTTGGAGCCAGTGCTTGAAGGGGCAGTCAGACATGGCTTGCCCTTCCTCCTCCCTGGTGGTGGAGATGGGTGTTAGGGTCCAGTGGGTGCTACTGTCCAGGGGGGCTTCTGGGGCCACCAGCCTGTCAGCTCATCAACCAGGCTGAAGGTGCAAGCAGGAGCCCCTTGCCTTGCCCCAAGGATCCCAGACAGCTATGAAGCCACCAGCCTTCCTGACCTCAAGACCACCTTTTTTTTTTCTCTTTCTTACTAGGGAATGCCAAACACTCTCCCCAGGAGATCCAGACCCGCCTCTTTCAGAGACTTTTAACTTAAACATCTGTCCCTACCCAGCAGGCAAACTAGAGCTCCTGAAGCTCAGTCCCTGTCCTTGCCTCTGTAGACAGGTCACCTTGATGAGCTTCCTTTTTTTTTTTTTAATTTTTTTTTATTTTAGGCTTTATTGGGGCATAATTGATCCCCCAAAATTGCATACATTCAAGGTATGCAGTGTGATGATTTGATATGGGGGTATATTGTGAAACCATTACCACAATCAAATTAATCAGCACGTCCATCATCACACACAGTTACCATTTGTGTGTGTGCACGTGTGTTCACCTACGACGAGGACACTTGGACCTACTCTGCAGATCTCAAGTAAACAGAAAATCTCCCTTTTTGACAACCATCCTCCACCCTTTCAATCCCAACCTTTTCCTAGATTATGTCCCTAGCTCTGTTTTTATTTCTGCTGTGCTGCTTCAGATCCATTCTGACTCTGCCAAACCCTTCTTTGTGAGCTGATAGATTGCTGGATTGAGAATTACAGCTGGGCGCGGTGGCTCACGCCTGTAATCCCAACACTGTGGGAGGCCAAGGCCGGCGGATCACTTGAGGTCAGGAGTTGGAGACCAGCCTGACCAACAAGATGAAACCCCATCTCTACTAAAAATACAAAATTAGCTGGGCATGGTGGTGCACGCCTGTAATCTCATCTTCTTGGGAGGCTGAGGCAGGAGAATTGCTTGAACCCGGGAGGTGGAGGTTGCAGTGAGCCAAGATCCTGCCATTGCACTCCAGCCTGGGCAACAACAGTGAAGCTCCATCTCAAAACACACAAAAAAAAGAAGTACAAAGTCTGAGACTTCAGGCCAGCTCTGCTACACTATATACTCTAACCTCTCTGGTCCTACTTGGTGACTTCTTTCCCTCTGGTCGTGTTCAAGTTCCCGTCCCATCCAGTCAAGCAGGTACTCATTGGTACCTTACCCTGTGCCAGGAGCTGTTCTAGGCCCTGGAAACCTATGGCAGACATGTTCCCTACCCTCCCACTCAAAGAGCCCAGGCCTTATCCTAATGAGATCTGAAATCAAATCTCCCAATTTCCTCATGGCTTCAGTCTAAACTTGTAATTCACAACCTTAAATCAATATGTTCTATTTTTTTATTTAGAAAACATTTCCGGCCAGGCACGGTGGATCACACCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCAGGAGGCAGAGGGTTGCAGTGAGCCGAGATTGCGCCATTGCACTCTAGCCTGGGCAACAGAGCAAGACTCCATCTCAAAAAAGAAAAAAAAATGGAAGAAAAAAAAATTTCCCCCTCATTTTAGGAACACGAGGTCTCCAAATCTAAAATTCGTACTCTGAGGAGATTGAATAGCCTTAAATGCTTTCATCATTAAAAAGAAAAGAAAGGAACCTGGTATGCATCCTAAAAATGAAAAATATACCTACCTGTAATCCCAGCACACAGCACATTGGGAGGCTAAAGCAGGAGGATAACTTGAGGCCAGGAGTTTCAGATCAGCCTGGGCAACATAGCAACACCCCATTTCTTTTTCTTTTCTTTTTTTTTTGGAGACACAGTCTCGCTCTGTTACTCAGGCTGGAGTGCAGTGGCTCAATCTCAGCTCACTGCAAGCTCTGCCTCCCAGGTTCATGCCATTCTCCTGCCTCAGCCTCCCGAGTAGCTGGGACTACAGGCGCCCGCCACCACGCCTGGCTAATTTTTTGTATTTTTAGTAGAGACAGGGTTTCACCGTGTTAGCCAGGATGGTCTCGATCTCCTGACCTCGTGATCCGCCAGCCTTGGCCTCCTAAAGTACTGGGATTACAGGCGTGAGCCACTGCGCCTGGCCACAACACCCCATTTCTATTTTAATAAAATAAAATACTGTGAAAAACATTTACAATTTTTAAATTTTAATTTTAAAATTAAACTTATATTTATTCATTTGTGTGTGTGGGTTTTTTTTTTTTTTTTTTTTTGCTTTTTTTTTGAGATGGAGTGTCACTCTGTCACCCAGGCTGGAGTGCAGTGGCGTGATCTCTGCCTCCCGGTTCAAGTGATTCTCCTGCCATAGCCTCCCAAGTAGCTGGGACTACAGGTACACGCCACCACGCCGGGTTAATTTTTGTATTTTTAGTAGAGACAGGATTTCACTGTGTCGCCAGGCTAGCCTCGAACTCCTGACCTCAGGTGATTCGCCCACCTTGGCCTCCCAAAGTGCTGTGATTACAAGCGTGAGCCACCGTGCCCAGCCCAAAGTTGGTTTTAATAGCAGAAAATCTATCAACATAATTCAATATATTAAATTTAGAAAGAAAAATTATCTATCATATCAACAGATACTGAAAGGAATTTGATTAAATTTCAGTAGCCATTTCCTTAAAAAAGAAAACACTTTAACACAGTAATAGACTGATAATGGAATACCAATTTTCCTAATAAGTTAAACATTAAGATAATTTCAATTAAGGTCAAGAGCTGGGCCAGGTGCAGTGGCTCACACCTGTAATCCCAACACTTTGGAGGCCAAGGTGGGTGGATCACCTGAGGTCAGGAGTGGAGACCAGCCTGGCTGACAATAGTGAAATCCTGCCTCTACTAAAAACACAAAAAATTAGCTGGGCATGGTGGTGGGCACCTATAATCCCAGCTACTGGGAAGGCTGAGACAGGAGAATTGCTTGAACCTGGGAGGCGGAGGTTGCAGTGAGCAAAGATCACACCATTGCACTCCAGCCTGGGCGACAGAGCCAGAGTCAGTCTCAAAAAAAAAAAGAGGTGGCCACACCTATAATCCAAACATTTTGTGAGGCCAAGGCAGGAGAATTGCTTCAGGCCAAGAGTTGAACACCTCGTCAACATAGCCAGACCTCTCTCTAGATAGATAGATAGATGATAGATAGAGAGATAGATAGATGATAGATAGAGAGATAGATAGATGATAGATAGATAGATAGATAGATAGATAGATAGATAGATAGATAGATAGATAGATAGATAATCTGGCCGGGTGTGGAGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCTGAGGCGGGCAGATCACGAGGACAAGAGATTGAAACCATCCTGGCTAACAAGGTGAAACCCCGTCTCTACTAAAAATACAAAAAATTAGGCGGGTGTGGTGGCACGCGCCTGTAGTCCTAGCTATTCAGGAGGCTGAGACAGGAGAATTGCTTGAATCCGAAAGGCGGAGGTTGCAGCGAGCCGAGATCGTGCCACTGCACTCCAGCCTGGGTGACAGAGCAAGACTCCATCTCAAAATAAATAAATAAATAATCAAGAACAGTATAAGGGGCTGTATGGTGGCTCATGCCTGTGATCCCAGCACTTTGGGAGGCCAAGGTGGGAGGATCCCTTGAGACCAGCCCAGGCAACAGAGAAAGACCCTGTCTCTATTTAAAAAAATTAAAAACTGGCCGGGCACGGTGGCTCACGCCTGTAATTCCAGCGCTTGGGAGGCCAAGGCAGGCACATCAGGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACGTGGTGAAACCCCGTCTCTACTAAAAATACAAAAAGTAGCTAGGCGTGGTGGCAGGCACCTGTAATCCCAGCTACTTGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCAGGAGGCGGAGGTTGCAGTGGGCAAAGATCGTGCCATTGCACTCAGCCTGGGTGACAGGGCAAGACTCCATCTCAAAATAAATAAACAAAGTAATTAATTAATTAAATTAAAAACTGTGGGGATATAGACTTACTCTGGTTTTATTTTTTCTTTTCTTTTCTTTTCTTTTTTCTGAGACGGAGTCTCGCTCTGTTGCCCAGGCTGGAGTACAGTGGCGTGGTTTCTGTTCTCTGCAACCTCCACCTCCCGGATTCAAGCGATTCTCTTGCCTCAGCCTCTTGAATACCTGGAATTACAGGTGCCTGCCACCACCCCCGGCTAATTTTTTGTATTTTTAGTAGAGACAGGGTTTCACCATGTTGGCCAAGCTGGTCTCGAACTCCTGACCTCATGATCCACCCGCCTCTGCCTCCCAAAGCACTGAGACTACAGGAGTGAGCCACTGTGCCCAGCCTACTCTGGTTTTAGTGCATTCAAGAGGAACAAAAAAGGAAGAAAATCACTAGTAAATATACCTCTTTCTGGTTAGAGTGGATGTTTGGAAATTATATATATATTATATTATATTATATATATTATATATATACACAAACACGTACATACATGCACACACATATATGCCTTTTTGATTATAGGATAGTATACCAAAACTCAGAAATATTATGGAATTAACAGAATTTAGTAAGGCAGATAAGTAGTAGGTAGAAAAATATTAATTTTATCTTCCAGCAGAAGCACTGTGAAAAATTAGACAACAAGAAAACATTCCATTCAAAATAATGACAATAAGGCCGGGCATGGTGGCTCACACCTGTAATCCCAGCACTTTGGGAGGCTGAGGCAGGAGGATCATCTGAGGTCAAGTTTGAGATCAGCCTGGCCAACATGGTGACACCCTGTCTCTACTGAAAATACAAAAATCAGCCAGCTATGGTAGTGTAAGCCTGTAATTCCAGCTACTCGGGAGGTCGAAGCAGAAGAATCACTTGAACCCAGGAGGCAGAGATTGCAGTGAGCCAAGATCCTGCCAGTGCTTTCCAGCCTGGGCAACAGTGTGAGGCTCCATCTCAAAAAAAAAAAAAAAAAAAAGACAATAGCAATAAACATTAAGAAATGTGTAATAGGAATGGCACACACAAAGAAGGAATGGCACAGAGCCTGTATGCAGAAGACCACAAACCCTTATTTAACGACGTAAGCCAAGATCCAAAGAAAATGATAGATTCTCAGATGGGAAAACTAAAAAAATAAGAAAAATCAATTATCTCGAGATAAATATAATATAATGCAATTTCAATTAGAATCCCAAATTTTCATTGTGTGTGTGTGTGAGTTGGGTAAATTTATCATAAATGTATAGGAACGAGTAAGTGTCACTAGTTGTTTAAATAAATACTGGATTTGGGCCAGGCATGGTGGCTCACGCCTCTAATCCCAGCACTTTGGGAGACCGAGGCGGGCAGATCATGAGGTCAGGAGATCGAGACCATCTGGCCAACATAGTGAAAACTCGTCTCTACTAAAGATACAAAAAATTAGCTGGGCATGGTGGCACGTGCCTGTAGTTCCAGCTACTCTGGAGGCTGAGGCAGGAGAGTTGCTTGAACCCGGGAGGTGGAGGTTGCAATGAGCCGAGATCCTGTCACTGCACTCCACCCTGGCGACAAAGTGAGACTCCGTCTCTCTCTCTCTCTTTAGGCCAAGGCAGGTGGATCACCTGAGGTCAGGAGTTCAAGACAGCCTGGCCAACATAGCGAAATCCCATCTCTACTAAAAATACAAAAATTAGCCTGGCAGTGGTGGCCCACGCCTGTAATCCCAGCTACTAAGGGGGCTGAGGCAGGAGGATCTCTTAACCAGGGAGGAGGAGGTTGCAGTGAGCAGAGATTGTGCCACTGCACTCCAGCCTGTGCAACAGAGTGAGACTCTGTCTC
    |||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||  ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
    ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGC-ATGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGACATTCTCCACTTCTTGTTCCCCACTGACAGCCTCCCACCCCCATCTCTCCCTCCCCTGCCATTTTGGGTTTTGGGTCTTTGAACCCTTGCTTGCAATAGGTGTGCGTCAGAAGCACCCAGGACTTCCATTTGCTTTGTCCCGGGGCTCCACTGAACAAGTTGGCCTGCACTGGTGTTTTGTTGTGGGGAGGAGGATGGGGAGTAGGACATACCAGCTTAGATTTTAAGGTTTTTACTGTGAGGGATGTTTGGGAGATGTAAGAAATGTTCTTGCAGTTAAGGGTTAGTTTACAATCAGCCACATTCTAGGTAGGGGCCCACTTCACCGTACTAACCAGGGAAGCTGTCCCTCACTGTTGAATTTTCTCTAACTTCAAGGCCCATATCTGTGAAATGCTGGCATTTGCACCTACCTCACAGAGTGCATTGTGAGGGTTAATGAAATAATGTACATCTGGCCTTGAAACCACCTTTTATTACATGGGGTCTAGAACTTGACCCCCTTGAGGGTGCTTGTTCCCTCTCCCTGTTGGTCGGTGGGTTGGTAGTTTCTACAGTTGGGCAGCTGGTTAGGTAGAGGGAGTTGTCAAGTCTCTGCTGGCCCAGCCAAACCCTGTCTGACAACCTCTTGGTGAACCTTAGTACCTAAAAGGAAATCTCACCCCATCCCACACCCTGGAGGATTTCATCTCTTGTATATGATGATCTGGATCCACCAAGACTTGTTTTATGCTCAGGGTCAATTTCTTTTTTCTTTTTTTTTTTTTTTTTTCTTTTTCTTTGAGACTGGGTCTCGCTTTGTTGCCCAGGCTGGAGTGGAGTGGCGTGATCTTGGCTTACTGCAGCCTTTGCCTCCCCGGCTCGAGCAGTCCTGCCTCAGCCTCCGGAGTAGCTGGGACCACAGGTTCATGCCACCATGGCCAGCCAACTTTTGCATGTTTTGTAGAGATGGGGTCTCACAGTGTTGCCCAGGCTGGTCTCAAACTCCTGGGCTCAGGCGATCCACCTGTCTCAGCCTCCCAGAGTGCTGGGATTACAATTGTGAGCCACCACGTCCAGCTGGAAGGGTCAACATCTTTTACATTCTGCAAGCACATCTGCATTTTCACCCCACCCTTCCCCTCCTTCTCCCTTTTTATATCCCATTTTTATATCGATCTCTTATTTTACAATAAAACTTTGCTGCCACCTGTGTGTCTGAGGGGTGAACGCCAGTGCAGGCTACTGGGGTCAGCAGGTGCAGGGGTGAGTGAGGAGGTGCTGGGAAGCAGCCACCTGAGTCTGCAATGAGTGTGGGCTGGGGGGCCCAGTGCCCGGGTTCCGGGAGGGGAACAAAGGCTGGAGACTGGGTCAGTCTGCGGGCTGCATGACAACAAGGGAGGGGGTGGCTCCATTCATAACTCAGGAACCAACCGTCCCTCCTCCCCTCCGGCCACGGCTGGCACAAGGTTCTCTCCCTCCCCTGCTTCTAGGACTGGGCTGCTTCCCCCTCGGCAGCCTCTCACCAAGGATTACGGGATTTAAATGTCTGATTTAGCAAGGCTGAGCCTCCAGGGTGGCCATCTGCTCCATCAGAAAGTGGCAGGATACCTGGGTTCCCAAGGGGAACAGGGGTGGGTGCTACTGGATGGAGAGAGGCCAGTGGGAGGCCTGCTAGCCAGGGTCCCAGGAAAGTGGGGGCAGCTAAGGTAAGAGTAGGGGTGTGGGGCTAGGTCCTTCCCAGCATCCCCTCATCCTGGGCCTCATGCCAGGTAGCTGAATGAATTGAAGCTTTAAACTCTGCCAGGAAAACCTTTCAAAGGGCTTCTTGGGATAGGGAGGAGAGTCGGGTTGAGGAGCTCAGTACTGCCTGCCCATGCTCCTCAGGGCTGCTGGCTCCCAGGGAGGGGGGCTGGGAGCAGGCAGGCTCTTCCCCATCACCCACTGCTCTCTTGGAGCCAGTGCTTGAAGGGGCAGTCAGACATGGCTTGCCCTTCCTCCTCCCTGGTGGTGGAGATGGGTGTTAGGGTCCAGTGGGTGCTACTGTCCAGGGGGGCTTCTGGGGCCACCAGCCTGTCAGCTCATCAACCAGGCTGAAGGTGCAAGCAGGAGCCCCTTGCCTTGCCCCAAGGATCCCAGACAGCTATGAAGCCACCAGCCTTCCTGACCTCAAGACCACCTTTTTTTTTTCTCTTTCTTACTAGGGAATGCCAAACACTCTCCCCAGGAGATCCAGACCCGCCTCTTTCAGAGACTTTTAACTTAAACATCTGTCCCTACCCAGCAGGCAAACTAGAGCTCCTGAAGCTCAGTCCCTGTCCTTGCCTCTGTAGACAGGTCACCTTGATGAGCTTCCTTTTTTTTTTTTTAATTTTTTTTTATTTTAGGCTTTATTGGGGCATAATTGATCCCCCAAAATTGCATACATTCAAGGTATGCAGTGTGATGATTTGATATGGGGGTATATTGTGAAACCATTACCACAATCAAATTAATCAGCACGTCCATCATCACACACAGTTACCATTTGTGTGTGTGCACGTGTGTTCACCTACGACGAGGACACTTGGACCTACTCTGCAGATCTCAAGTAAACAGAAAATCTCCCTTTTTGACAACCATCCTCCACCCTTTCAATCCCAACCTTTTCCTAGATTATGTCCCTAGCTCTGTTTTTATTTCTGCTGTGCTGCTTCAGATCCATTCTGACTCTGCCAAACCCTTCTTTGTGAGCTGATAGATTGCTGGATTGAGAATTACAGCTGGGCGCGGTGGCTCACGCCTGTAATCCCAACACTGTGGGAGGCCAAGGCCGGCGGATCACTTGAGGTCAGGAGTTGGAGACCAGCCTGACCAACAAGATGAAACCCCATCTCTACTAAAAATACAAAATTAGCTGGGCATGGTGGTGCACGCCTGTAATCTCATCTTCTTGGGAGGCTGAGGCAGGAGAATTGCTTGAACCCGGGAGGTGGAGGTTGCAGTGAGCCAAGATCCTGCCATTGCACTCCAGCCTGGGCAACAACAGTGAAGCTCCATCTCAAAACACACAAAAAAAAGAAGTACAAAGTCTGAGACTTCAGGCCAGCTCTGCTACACTATATACTCTAACCTCTCTGGTCCTACTTGGTGACTTCTTTCCCTCTGGTCGTGTTCAAGTTCCCGTCCCATCCAGTCAAGCAGGTACTCATTGGTACCTTACCCTGTGCCAGGAGCTGTTCTAGGCCCTGGAAACCTATGGCAGACATGTTCCCTACCCTCCCACTCAAAGAGCCCAGGCCTTATCCTAATGAGATCTGAAATCAAATCTCCCAATTTCCTCATGGCTTCAGTCTAAACTTGTAATTCACAACCTTAAATCAATATGTTCTATTTTTTTATTTAGAAAACATTTCCGGCCAGGCACGGTGGATCACACCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCAGGAGGCAGAGGGTTGCAGTGAGCCGAGATTGCGCCATTGCACTCTAGCCTGGGCAACAGAGCAAGACTCCATCTCAAAAAAGAAAAAAAAATGGAAGAAAAAAAAATTTCCCCCTCATTTTAGGAACACGAGGTCTCCAAATCTAAAATTCGTACTCTGAGGAGATTGAATAGCCTTAAATGCTTTCATCATTAAAAAGAAAAGAAAGGAACCTGGTATGCATCCTAAAAATGAAAAATATACCTACCTGTAATCCCAGCACACAGCACATTGGGAGGCTAAAGCAGGAGGATAACTTGAGGCCAGGAGTTTCAGATCAGCCTGGGCAACATAGCAACACCCCATTTCTTTTTCTTTTCTTTTTTTTTTGGAGACACAGTCTCGCTCTGTTACTCAGGCTGGAGTGCAGTGGCTCAATCTCAGCTCACTGCAAGCTCTGCCTCCCAGGTTCATGCCATTCTCCTGCCTCAGCCTCCCGAGTAGCTGGGACTACAGGCGCCCGCCACCACGCCTGGCTAATTTTTTGTATTTTTAGTAGAGACAGGGTTTCACCGTGTTAGCCAGGATGGTCTCGATCTCCTGACCTCGTGATCCGCCAGCCTTGGCCTCCTAAAGTACTGGGATTACAGGCGTGAGCCACTGCGCCTGGCCACAACACCCCATTTCTATTTTAATAAAATAAAATACTGTGAAAAACATTTACAATTTTTAAATTTTAATTTTAAAATTAAACTTATATTTATTCATTTGTGTGTGTGGGTTTTTTTTTTTTTTTTTTTTTGCTTTTTTTTTGAGATGGAGTGTCACTCTGTCACCCAGGCTGGAGTGCAGTGGCGTGATCTCTGCCTCCCGGTTCAAGTGATTCTCCTGCCATAGCCTCCCAAGTAGCTGGGACTACAGGTACACGCCACCACGCCGGGTTAATTTTTGTATTTTTAGTAGAGACAGGATTTCACTGTGTCGCCAGGCTAGCCTCGAACTCCTGACCTCAGGTGATTCGCCCACCTTGGCCTCCCAAAGTGCTGTGATTACAAGCGTGAGCCACCGTGCCCAGCCCAAAGTTGGTTTTAATAGCAGAAAATCTATCAACATAATTCAATATATTAAATTTAGAAAGAAAAATTATCTATCATATCAACAGATACTGAAAGGAATTTGATTAAATTTCAGTAGCCATTTCCTTAAAAAAGAAAACACTTTAACACAGTAATAGACTGATAATGGAATACCAATTTTCCTAATAAGTTAAACATTAAGATAATTTCAATTAAGGTCAAGAGCTGGGCCAGGTGCAGTGGCTCACACCTGTAATCCCAACACTTTGGAGGCCAAGGTGGGTGGATCACCTGAGGTCAGGAGTGGAGACCAGCCTGGCTGACAATAGTGAAATCCTGCCTCTACTAAAAACACAAAAAATTAGCTGGGCATGGTGGTGGGCACCTATAATCCCAGCTACTGGGAAGGCTGAGACAGGAGAATTGCTTGAACCTGGGAGGCGGAGGTTGCAGTGAGCAAAGATCACACCATTGCACTCCAGCCTGGGCGACAGAGCCAGAGTCAGTCTCAAAAAAAAAAAGAGGTGGCCACACCTATAATCCAAACATTTTGTGAGGCCAAGGCAGGAGAATTGCTTCAGGCCAAGAGTTGAACACCTCGTCAACATAGCCAGACCTCTCTCTAGATAGATAGATAGATGATAGATAGAGAGATAGATAGATGATAGATAGAGAGATAGATAGATGATAGATAGATAGATAGATAGATAGATAGATAGATAGATAGATAGATAGATAGATAGATAATCTGGCCGGGTGTGGAGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCTGAGGCGGGCAGATCACGAGGACAAGAGATTGAAACCATCCTGGCTAACAAGGTGAAACCCCGTCTCTACTAAAAATACAAAAAATTAGGCGGGTGTGGTGGCACGCGCCTGTAGTCCTAGCTATTCAGGAGGCTGAGACAGGAGAATTGCTTGAATCCGAAAGGCGGAGGTTGCAGCGAGCCGAGATCGTGCCACTGCACTCCAGCCTGGGTGACAGAGCAAGACTCCATCTCAAAATAAATAAATAAATAATCAAGAACAGTATAAGGGGCTGTATGGTGGCTCATGCCTGTGATCCCAGCACTTTGGGAGGCCAAGGTGGGAGGATCCCTTGAGACCAGCCCAGGCAACAGAGAAAGACCCTGTCTCTATTTAAAAAAATTAAAAACTGGCCGGGCACGGTGGCTCACGCCTGTAATTCCAGCGCTTGGGAGGCCAAGGCAGGCACATCAGGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACGTGGTGAAACCCCGTCTCTACTAAAAATACAAAAAGTAGCTAGGCGTGGTGGCAGGCACCTGTAATCCCAGCTACTTGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCAGGAGGCGGAGGTTGCAGTGGGCAAAGATCGTGCCATTGCACTCAGCCTGGGTGACAGGGCAAGACTCCATCTCAAAATAAATAAACAAAGTAATTAATTAATTAAATTAAAAACTGTGGGGATATAGACTTACTCTGGTTTTATTTTTTCTTTTCTTTTCTTTTCTTTTTTCTGAGACGGAGTCTCGCTCTGTTGCCCAGGCTGGAGTACAGTGGCGTGGTTTCTGTTCTCTGCAACCTCCACCTCCCGGATTCAAGCGATTCTCTTGCCTCAGCCTCTTGAATACCTGGAATTACAGGTGCCTGCCACCACCCCCGGCTAATTTTTTGTATTTTTAGTAGAGACAGGGTTTCACCATGTTGGCCAAGCTGGTCTCGAACTCCTGACCTCATGATCCACCCGCCTCTGCCTCCCAAAGCACTGAGACTACAGGAGTGAGCCACTGTGCCCAGCCTACTCTGGTTTTAGTGCATTCAAGAGGAACAAAAAAGGAAGAAAATCACTAGTAAATATACCTCTTTCTGGTTAGAGTGGATGTTTGGAAATTATATATATATTATATTATATTATATATATTATATATATACACAAACACGTACATACATGCACACACATATATGCCTTTTTGATTATAGGATAGTATACCAAAACTCAGAAATATTATGGAATTAACAGAATTTAGTAAGGCAGATAAGTAGTAGGTAGAAAAATATTAATTTTATCTTCCAGCAGAAGCACTGTGAAAAATTAGACAACAAGAAAACATTCCATTCAAAATAATGACAATAAGGCCGGGCATGGTGGCTCACACCTGTAATCCCAGCACTTTGGGAGGCTGAGGCAGGAGGATCATCTGAGGTCAAGTTTGAGATCAGCCTGGCCAACATGGTGACACCCTGTCTCTACTGAAAATACAAAAATCAGCCAGCTATGGTAGTGTAAGCCTGTAATTCCAGCTACTCGGGAGGTCGAAGCAGAAGAATCACTTGAACCCAGGAGGCAGAGATTGCAGTGAGCCAAGATCCTGCCAGTGCTTTCCAGCCTGGGCAACAGTGTGAGGCTCCATCTCAAAAAAAAAAAAAAAAAAAAGACAATAGCAATAAACATTAAGAAATGTGTAATAGGAATGGCACACACAAAGAAGGAATGGCACAGAGCCTGTATGCAGAAGACCACAAACCCTTATTTAACGACGTAAGCCAAGATCCAAAGAAAATGATAGATTCTCAGATGGGAAAACTAAAAAAATAAGAAAAATCAATTATCTCGAGATAAATATAATATAATGCAATTTCAATTAGAATCCCAAATTTTCATTGTGTGTGTGTGTGAGTTGGGTAAATTTATCATAAATGTATAGGAACGAGTAAGTGTCACTAGTTGTTTAAATAAATACTGGATTTGGGCCAGGCATGGTGGCTCACGCCTCTAATCCCAGCACTTTGGGAGACCGAGGCGGGCAGATCATGAGGTCAGGAGATCGAGACCATCTGGCCAACATAGTGAAAACTCGTCTCTACTAAAGATACAAAAAATTAGCTGGGCATGGTGGCACGTGCCTGTAGTTCCAGCTACTCTGGAGGCTGAGGCAGGAGAGTTGCTTGAACCCGGGAGGTGGAGGTTGCAATGAGCCGAGATCCTGTCACTGCACTCCACCCTGGCGACAAAGTGAGACTCCGTCTCTCTCTCTCTCTTTAGGCCAAGGCAGGTGGATCACCTGAGGTCAGGAGTTCAAGACAGCCTGGCCAACATAGCGAAATCCCATCTCTACTAAAAATACAAAAATTAGCCTGGCAGTGGTGGCCCACGCCTGTAATCCCAGCTACTAAGGGGGCTGAGGCAGGAGGATCTCTTAACCAGGGAGGAGGAGGTTGCAGTGAGCAGAGATTGTGCCACTGCACTCCAGCCTGTGCAACAGAGTGAGACTCTGTCTC
      Score=9011
    



```python
h2m.PrintAlign(seq_m, new_seq_m)
```

    ATGACTGCCATGGAGGAGTCACAGTCGGATATCAGCCTCGAGCTCCCTCTGAGCCAGGAGACATTTTCAGGCTTATGGAAACTACTTCCTCCAGAAGATATCCTGCCATCACCTCACTGCATGGACGATCTGTTGCTGCCCCAGGATGTTGAGGAGTTTTTTGAAGGCCCAAGTGAAGCCCTCCGAGTGTCAGGAGCTCCTGCAGCACAGGACCCTGTCACCGAGACCCCTGGGCCAGTGGCCCCTGCCCCAGCCACTCCATGGCCCCTGTCATCTTTTGTCCCTTCTCAAAAAACTTACCAGGGCAACTATGGCTTCCACCTGGGCTTCCTGCAGTCTGGGACAGCCAAGTCTGTTATGTGCACGTACTCTCCTCCCCTCAATAAGCTATTCTGCCAGCTGGCGAAGACGTGCCCTGTGCAGTTGTGGGTCAGCGCCACACCTCCAGCTGGGAGCCGTGTCCGCGCCATGGCCATCTACAAGAAGTCACAGCACATGACGGAGGTCGTGAGACGCTGCCCCCACCATGAGCGCTGCTCCGATGGTGATGGCCTGGCTCCTCCCCAGCATCTTATCCGGGTGGAAGGAAATTTGTATCCCGAGTATCTGGAAGACAGGCAGACTTTTCGCCACAGCGTGGTGGTACCTTATGAGCCACCCGAGGCCGGCTCTGAGTATACCACCATCCACTACAAGTACATGTGTAATAGCTCCTGCATGGGGGGCATGAACCGCCGACCTATCCTTACCATCATCACACTGGAAGACTCCAGTGGGAACCTTCTGGGACGGGACAGCTTTGAGGTTCG-TGTTTGTGCCTGCCCTGGGAGAGACCGCCGTACAGAAGAAGAAAATTTCCGCAAAAAGGAAGTCCTTTGCCCTGAACTGCCCCCAGGGAGCGCAAAGAGAGCGCTGCCCACCTGCACAAGCGCCTCTCCCCCGCAAAAGAAAAAACCACTTGATGGAGAGTATTTCACCCTCAAGATCCGCGGGCGTAAACGCTTCGAGATGTTCCGGGAGCTGAATGAGGCCTTAGAGTTAAAGGATGCCCATGCTACAGAGGAGTCTGGAGACAGCAGGGCTCACTCCAGCTACCTGAAGACCAAGAAGGGCCAGTCTACTTCCCGCCATAAAAAAACAATGGTCAAGAAAGTGGGGCCTGACTCAGACTGACTGCCTCTGCATCCCGTCCCCATCACCAGCCTCCCCCTCTCCTTGCTGTCTTATGACTTCAGGGCTGAGACACAATCCTCCCGGTCCCTTCTGCTGCCTTTTTTACCTTGTAGCTAGGGCTCAGCCCCCTCTCTGAGTAGTGGTTCCTGGCCCAAGTTGGGGAATAGGTTGATAGTTGTCAGGTCTCTGCTGGCCCAGCGAAATTCTATCCAGCCAGTTGTTGGACCCTGGCACCTACAATGAAATCTCACCCTACCCCACACCCTGTAAGATTCTATCTTGGGCCCTCATAGGGTCCATATCCTCCAGGGCCTACTTTCCTTCCATTCTGCAAAGCCTGTCTGCATTTATCCACCCCCCACCCTGTCTCCCTCTTTTTTTTTTTTTTACCCCTTTTTATATATCAATTTCCTATTTTACAATAAAATTTTGTTATCACTTA
    ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||  ||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||
    ATGACTGCCATGGAGGAGTCACAGTCGGATATCAGCCTCGAGCTCCCTCTGAGCCAGGAGACATTTTCAGGCTTATGGAAACTACTTCCTCCAGAAGATATCCTGCCATCACCTCACTGCATGGACGATCTGTTGCTGCCCCAGGATGTTGAGGAGTTTTTTGAAGGCCCAAGTGAAGCCCTCCGAGTGTCAGGAGCTCCTGCAGCACAGGACCCTGTCACCGAGACCCCTGGGCCAGTGGCCCCTGCCCCAGCCACTCCATGGCCCCTGTCATCTTTTGTCCCTTCTCAAAAAACTTACCAGGGCAACTATGGCTTCCACCTGGGCTTCCTGCAGTCTGGGACAGCCAAGTCTGTTATGTGCACGTACTCTCCTCCCCTCAATAAGCTATTCTGCCAGCTGGCGAAGACGTGCCCTGTGCAGTTGTGGGTCAGCGCCACACCTCCAGCTGGGAGCCGTGTCCGCGCCATGGCCATCTACAAGAAGTCACAGCACATGACGGAGGTCGTGAGACGCTGCCCCCACCATGAGCGCTGCTCCGATGGTGATGGCCTGGCTCCTCCCCAGCATCTTATCCGGGTGGAAGGAAATTTGTATCCCGAGTATCTGGAAGACAGGCAGACTTTTCGCCACAGCGTGGTGGTACCTTATGAGCCACCCGAGGCCGGCTCTGAGTATACCACCATCCACTACAAGTACATGTGTAATAGCTCCTGCATGGGGGGCATGAACCGCCGACCTATCCTTACCATCATCACACTGGAAGACTCCAGTGGGAACCTTCTGGGACGGGACAGCTTTGAGGTTC-ATGTTTGTGCCTGCCCTGGGAGAGACCGCCGTACAGAAGAAGAAAATTTCCGCAAAAAGGAAGTCCTTTGCCCTGAACTGCCCCCAGGGAGCGCAAAGAGAGCGCTGCCCACCTGCACAAGCGCCTCTCCCCCGCAAAAGAAAAAACCACTTGATGGAGAGTATTTCACCCTCAAGATCCGCGGGCGTAAACGCTTCGAGATGTTCCGGGAGCTGAATGAGGCCTTAGAGTTAAAGGATGCCCATGCTACAGAGGAGTCTGGAGACAGCAGGGCTCACTCCAGCTACCTGAAGACCAAGAAGGGCCAGTCTACTTCCCGCCATAAAAAAACAATGGTCAAGAAAGTGGGGCCTGACTCAGACTGACTGCCTCTGCATCCCGTCCCCATCACCAGCCTCCCCCTCTCCTTGCTGTCTTATGACTTCAGGGCTGAGACACAATCCTCCCGGTCCCTTCTGCTGCCTTTTTTACCTTGTAGCTAGGGCTCAGCCCCCTCTCTGAGTAGTGGTTCCTGGCCCAAGTTGGGGAATAGGTTGATAGTTGTCAGGTCTCTGCTGGCCCAGCGAAATTCTATCCAGCCAGTTGTTGGACCCTGGCACCTACAATGAAATCTCACCCTACCCCACACCCTGTAAGATTCTATCTTGGGCCCTCATAGGGTCCATATCCTCCAGGGCCTACTTTCCTTCCATTCTGCAAAGCCTGTCTGCATTTATCCACCCCCCACCCTGTCTCCCTCTTTTTTTTTTTTTTACCCCTTTTTATATATCAATTTCCTATTTTACAATAAAATTTTGTTATCACTTA
      Score=1614
    


Sometimes the mutation does not locate in the homologous region. Taking *TP53* D393Y as an example.


```python
model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7572932, 7572932, 'A', ty_h = 'SNP', ver = 37)
model_result
```

    /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
      warnings.warn(





    {'gene_name_h': 'TP53',
     'gene_id_h': 'ENSG00000141510.11',
     'tx_id_m': 'ENSMUST00000108658.10',
     'chr_h': 'chr17',
     'exon_num_h': 10,
     'strand_h': '-',
     'start_h': 7572932,
     'end_h': 7572932,
     'ref_seq_h': 'G',
     'alt_seq_h': 'T',
     'HGVSp_h': 'D393Y',
     'classification_h': 'Missense',
     'classification_h_detail': None,
     'exon_h': 'E_10',
     'type_h': 'SNP',
     'status': False,
     'error': 4,
     'statement': 'Error 4: Flanked segments are not identical.',
     'gene_name_m': 'Trp53',
     'gene_id_m': 'ENSMUSG00000059552.14',
     'chr_m': 'chr11',
     'exon_num_m': 10,
     'strand_m': '+',
     'type_m': None,
     'classification_m': None,
     'exon_m': None,
     'start_m_ori': None,
     'end_m_ori': None,
     'ref_seq_m_ori': None,
     'alt_seq_m_ori': None,
     'HGVSp_m_ori': None,
     'start_m': None,
     'end_m': None,
     'ref_seq_m': None,
     'alt_seq_m': None,
     'HGVSp_m': None}



Flanked size is a parameter that controls how many amino acids (or neucleotides, for non-coding mutations) are included on both sides of the mutated region. You can change that by parameter **flank_size**.


```python
model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7572932, 7572932, 'A', ty_h = 'SNP', ver = 37, flank_size=1)
model_result
```

    /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
      warnings.warn(





    {'gene_name_h': 'TP53',
     'gene_id_h': 'ENSG00000141510.11',
     'tx_id_m': 'ENSMUST00000108658.10',
     'chr_h': 'chr17',
     'exon_num_h': 10,
     'strand_h': '-',
     'start_h': 7572932,
     'end_h': 7572932,
     'ref_seq_h': 'C',
     'alt_seq_h': 'A',
     'HGVSp_h': 'D393Y',
     'classification_h': 'Missense',
     'classification_h_detail': None,
     'exon_h': 'E_10',
     'type_h': 'SNP',
     'status': True,
     'error': 0,
     'statement': 'This mutation can be originally modeled.',
     'gene_name_m': 'Trp53',
     'gene_id_m': 'ENSMUSG00000059552.14',
     'chr_m': 'chr11',
     'exon_num_m': 10,
     'strand_m': '+',
     'type_m': 'SNP',
     'classification_m': 'Missense',
     'exon_m': 'E_10',
     'start_m_ori': 69482252,
     'end_m_ori': 69482252,
     'ref_seq_m_ori': 'G',
     'alt_seq_m_ori': 'T',
     'HGVSp_m_ori': 'D390Y',
     'start_m': 69482252,
     'end_m': 69482252,
     'ref_seq_m': 'G',
     'alt_seq_m': 'T',
     'HGVSp_m': 'D390Y'}



With the flank size of 1 (which means 3 amino acids are taken into consideration in totall), this mutation can be originally modeled.

## Alternative modeling

Sometimes the human mutation cannot be originally modeled in the mouse genome by using the same neucleotide alteration. Under this circumsatance, some alternative modeling strategies may be found by searching the codon list of the target amino acids. Taking TP53 R249_T253delinsS as an example.


```python
model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7577523, 7577534, '-', ty_h = 'DEL', ver = 37)
```

The default maximum number of output alternatives is 5. You can definitly change that by the parameter **max_alternative**.


```python
model_result_long = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7577523, 7577534, '-', ty_h = 'DEL', ver = 37, max_alternative=10)
len(model_result), len(model_result_long)
```

    /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
      warnings.warn(





    (5, 6)



If you do not want to alternatively model variants, you can set **search_alternatve** to False.


```python
model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7577523, 7577534, '-', ty_h = 'DEL', ver = 37, search_alternative= False)
model_result['statement']
```

    /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
      warnings.warn(





    'Error 6: This mutation cannot be originally modeled and is awaiting alternative.'




```python
df_mutation['Variant_Classification'].value_counts()
```




    Missense_Mutation         946176
    Nonsense_Mutation         107587
    Frame_Shift_Del            94612
    Silent                     62774
    Intron                     43541
    Frame_Shift_Ins            39107
    Splice_Site                37777
    Splice_Region              37172
    In_Frame_Del               23885
    5'Flank                    17654
    In_Frame_Ins                8327
    3'UTR                       3225
    3'Flank                     3015
    5'UTR                       3007
    Translation_Start_Site      1671
    RNA                         1632
    Nonstop_Mutation             754
    In_Frame_DEL                  74
    Frame_Shift_DEL               62
    Name: Variant_Classification, dtype: int64



## Original modeling with uncertain effects

For frame-shifting mutations and mutations in the non-coding region, we cannot find such alternative modeling strategies with the same protein change effects. H2M will only offer the original modeling and its effect.

- Example 1: *TP53* C275Lfs*31


```python
model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7577115, 7577116, 'A', ty_h = 'INS', ver = 37)
model_result
```

    /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
      warnings.warn(





    {'gene_name_h': 'TP53',
     'gene_id_h': 'ENSG00000141510.11',
     'tx_id_m': 'ENSMUST00000108658.10',
     'chr_h': 'chr17',
     'exon_num_h': 10,
     'strand_h': '-',
     'start_h': 7577115,
     'end_h': 7577116,
     'ref_seq_h': '',
     'alt_seq_h': 'A',
     'HGVSp_h': 'C275Lfs*31',
     'classification_h': 'Frame_Shifting',
     'classification_h_detail': None,
     'exon_h': 'E_7',
     'type_h': 'INS',
     'status': True,
     'error': 2,
     'statement': 'Error 2: This mutation can be modeled, but the effect may not be consistent.',
     'gene_name_m': 'Trp53',
     'gene_id_m': 'ENSMUSG00000059552.14',
     'chr_m': 'chr11',
     'exon_num_m': 10,
     'strand_m': '+',
     'type_m': 'INS',
     'classification_m': 'Frame_Shifting',
     'exon_m': 'E_7',
     'start_m_ori': 69480438,
     'end_m_ori': 69480439,
     'ref_seq_m_ori': '',
     'alt_seq_m_ori': 'T',
     'HGVSp_m_ori': 'C272Lfs*24',
     'start_m': 69480438,
     'end_m': 69480439,
     'ref_seq_m': '',
     'alt_seq_m': 'T',
     'HGVSp_m': 'C272Lfs*24'}




```python
p_h, p_m = model_result['HGVSp_h'], model_result['HGVSp_m']
print(f'HGVSp_h: {p_h}, HGVSp_m: {p_m}')
```

    HGVSp_h: C275Lfs*31, HGVSp_m: C272Lfs*24


- Example 2: TP53 splice site mutation


```python
model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7578555, 7578555, 'T', ty_h = 'SNP', ver = 37)
model_result
```




    {'gene_name_h': 'TP53',
     'gene_id_h': 'ENSG00000141510.11',
     'tx_id_m': 'ENSMUST00000108658.10',
     'chr_h': 'chr17',
     'exon_num_h': 10,
     'strand_h': '-',
     'start_h': 7578555,
     'end_h': 7578555,
     'ref_seq_h': 'C',
     'alt_seq_h': 'T',
     'HGVSp_h': None,
     'classification_h': 'Non_Coding',
     'classification_h_detail': 'Intron',
     'exon_h': None,
     'type_h': 'SNP',
     'status': True,
     'error': 2,
     'statement': 'Error 2: This mutation can be modeled, but the effect may not be consistent.',
     'gene_name_m': 'Trp53',
     'gene_id_m': 'ENSMUSG00000059552.14',
     'chr_m': 'chr11',
     'exon_num_m': 10,
     'strand_m': '+',
     'type_m': 'SNP',
     'classification_m': 'Intron',
     'exon_m': None,
     'start_m_ori': None,
     'end_m_ori': None,
     'ref_seq_m_ori': None,
     'alt_seq_m_ori': None,
     'HGVSp_m_ori': None,
     'start_m': 69479189,
     'end_m': 69479189,
     'ref_seq_m': 'G',
     'alt_seq_m': 'A',
     'HGVSp_m': None}



## Additional function: modeling for base editing

When you set **param = 'BE'**, you will get modeling results that can be modeled by base editing (A->G, G->A, C->T, T->C, AA->GG, ...etc.). If one mutation can be originally modeled in the mouse genome but not in a BE style, alternative BE modeling strategies will be returned too.

Taking *KEAP1* F221L as an example.


```python
h2m.query('KEAP1')
```

    Query human gene: KEAP1;
    Mouse ortholog(s): Keap1;
    Homology type: one2one;
    Sequence Simalarity(%):94.0705.





    [{'gene_name_m': 'Keap1',
      'gene_id_m': 'ENSMUSG00000003308',
      'homology_type': 'ortholog_one2one',
      'similarity': 94.0705}]




```python
tx_id_h_2, tx_id_m_2 = h2m.get_tx_id('KEAP1','h',ver=37, show=False)[3], h2m.get_tx_id('Keap1','m', show=False)[3]
model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h_2, tx_id_m_2, 10602915, 10602915, 'C', ty_h = 'SNP', ver = 37, param='BE')
model_result
```

    /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
      warnings.warn(





    [{'gene_name_h': 'KEAP1',
      'gene_id_h': 'ENSG00000079999.9',
      'tx_id_m': 'ENSMUST00000164812.8',
      'chr_h': 'chr19',
      'exon_num_h': 5,
      'strand_h': '-',
      'start_h': 10602915,
      'end_h': 10602915,
      'ref_seq_h': 'G',
      'alt_seq_h': 'C',
      'HGVSp_h': 'F221L',
      'classification_h': 'Missense',
      'classification_h_detail': None,
      'exon_h': 'E_2',
      'type_h': 'SNP',
      'status': True,
      'error': 1,
      'statement': 'Error 1: This mutation can be alternatively modeled.',
      'gene_name_m': 'Keap1',
      'gene_id_m': 'ENSMUSG00000003308.16',
      'chr_m': 'chr9',
      'exon_num_m': 5,
      'strand_m': '-',
      'type_m': 'SNP',
      'classification_m': 'Missense',
      'exon_m': 'E_2',
      'start_m_ori': 21145346,
      'end_m_ori': 21145346,
      'ref_seq_m_ori': 'G',
      'alt_seq_m_ori': 'C',
      'HGVSp_m_ori': 'F221L',
      'start_m': 21145348,
      'end_m': 21145348,
      'ref_seq_m': 'A',
      'alt_seq_m': 'G',
      'HGVSp_m': 'F221L'}]



## Batch Processing

In this algorithm, alignment would take time. If you need to model human variants at scale, you could set **batch = True**, and input a prepared dictionary of alignment results by using function h2m.get_align. This would save time.

- load the sample data


```python
df = pd.read_csv('sample_data.csv')
df
```




<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Hugo_Symbol</th>
      <th>Entrez_Gene_Id</th>
      <th>Center</th>
      <th>NCBI_Build</th>
      <th>Chromosome</th>
      <th>Start_Position</th>
      <th>End_Position</th>
      <th>Strand</th>
      <th>Consequence</th>
      <th>Variant_Classification</th>
      <th>Variant_Type</th>
      <th>Reference_Allele</th>
      <th>Tumor_Seq_Allele1</th>
      <th>Tumor_Seq_Allele2</th>
      <th>dbSNP_RS</th>
      <th>dbSNP_Val_Status</th>
      <th>Tumor_Sample_Barcode</th>
      <th>Matched_Norm_Sample_Barcode</th>
      <th>Match_Norm_Seq_Allele1</th>
      <th>Match_Norm_Seq_Allele2</th>
      <th>Tumor_Validation_Allele1</th>
      <th>Tumor_Validation_Allele2</th>
      <th>Match_Norm_Validation_Allele1</th>
      <th>Match_Norm_Validation_Allele2</th>
      <th>Verification_Status</th>
      <th>...</th>
      <th>HGVSp_Short</th>
      <th>Transcript_ID</th>
      <th>RefSeq</th>
      <th>Protein_position</th>
      <th>Codons</th>
      <th>Exon_Number</th>
      <th>gnomAD_AF</th>
      <th>gnomAD_AFR_AF</th>
      <th>gnomAD_AMR_AF</th>
      <th>gnomAD_ASJ_AF</th>
      <th>gnomAD_EAS_AF</th>
      <th>gnomAD_FIN_AF</th>
      <th>gnomAD_NFE_AF</th>
      <th>gnomAD_OTH_AF</th>
      <th>gnomAD_SAS_AF</th>
      <th>FILTER</th>
      <th>Polyphen_Prediction</th>
      <th>Polyphen_Score</th>
      <th>SIFT_Prediction</th>
      <th>SIFT_Score</th>
      <th>SWISSPROT</th>
      <th>n_depth</th>
      <th>t_depth</th>
      <th>Annotation_Status</th>
      <th>mutationInCis_Flag</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>DNMT3A</td>
      <td>1788.0</td>
      <td>UCSF</td>
      <td>GRCh37</td>
      <td>2</td>
      <td>25468174</td>
      <td>25468174</td>
      <td>+</td>
      <td>missense_variant</td>
      <td>Missense_Mutation</td>
      <td>SNP</td>
      <td>T</td>
      <td>T</td>
      <td>C</td>
      <td>rs149738328</td>
      <td>NaN</td>
      <td>GENIE-UCSF-13266-8461T</td>
      <td>NORMAL</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.N501S</td>
      <td>ENST00000264709</td>
      <td>NM_175629.2</td>
      <td>501.0</td>
      <td>aAt/aGt</td>
      <td>13/23</td>
      <td>0.000320</td>
      <td>0.000124</td>
      <td>0.000261</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000185</td>
      <td>0.000434</td>
      <td>0.000330</td>
      <td>0.000459</td>
      <td>PASS</td>
      <td>benign</td>
      <td>0.046</td>
      <td>deleterious</td>
      <td>0.05</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>1007.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TP53</td>
      <td>7157.0</td>
      <td>DFCI</td>
      <td>GRCh37</td>
      <td>17</td>
      <td>7578212</td>
      <td>7578212</td>
      <td>+</td>
      <td>stop_gained</td>
      <td>Nonsense_Mutation</td>
      <td>SNP</td>
      <td>G</td>
      <td>G</td>
      <td>A</td>
      <td>rs397516436</td>
      <td>NaN</td>
      <td>GENIE-DFCI-094827-757551</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.R213*</td>
      <td>ENST00000269305</td>
      <td>NM_001126112.2</td>
      <td>213.0</td>
      <td>Cga/Tga</td>
      <td>6/11</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>PASS</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>382.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>2</th>
      <td>TP53</td>
      <td>7157.0</td>
      <td>DFCI</td>
      <td>GRCh37</td>
      <td>17</td>
      <td>7577090</td>
      <td>7577090</td>
      <td>+</td>
      <td>missense_variant</td>
      <td>Missense_Mutation</td>
      <td>SNP</td>
      <td>C</td>
      <td>C</td>
      <td>G</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>GENIE-DFCI-111478-664164</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.R283P</td>
      <td>ENST00000269305</td>
      <td>NM_001126112.2</td>
      <td>283.0</td>
      <td>cGc/cCc</td>
      <td>8/11</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>PASS</td>
      <td>probably_damaging</td>
      <td>0.914</td>
      <td>deleterious</td>
      <td>0.01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>528.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>3</th>
      <td>TP53</td>
      <td>7157.0</td>
      <td>WAKE</td>
      <td>GRCh37</td>
      <td>17</td>
      <td>7578406</td>
      <td>7578406</td>
      <td>+</td>
      <td>missense_variant</td>
      <td>Missense_Mutation</td>
      <td>SNP</td>
      <td>C</td>
      <td>NaN</td>
      <td>A</td>
      <td>rs28934578</td>
      <td>NaN</td>
      <td>GENIE-WAKE-F1296-01</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.R175L</td>
      <td>ENST00000269305</td>
      <td>NM_001126112.2</td>
      <td>175.0</td>
      <td>cGc/cTc</td>
      <td>5/11</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>PASS</td>
      <td>possibly_damaging</td>
      <td>0.902</td>
      <td>deleterious</td>
      <td>0.00</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>842.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>4</th>
      <td>DNMT3A</td>
      <td>1788.0</td>
      <td>UHN</td>
      <td>GRCh37</td>
      <td>2</td>
      <td>25463533</td>
      <td>25463534</td>
      <td>+</td>
      <td>frameshift_variant</td>
      <td>Frame_Shift_Ins</td>
      <td>INS</td>
      <td>TG</td>
      <td>TG</td>
      <td>CTCCTCT</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>GENIE-UHN-AGI943150-BM1</td>
      <td>NORMAL</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.N717Efs*64</td>
      <td>ENST00000264709</td>
      <td>NM_175629.2</td>
      <td>716.0</td>
      <td>gtCAac/gtAGAGGAGac</td>
      <td>18/23</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>PASS</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>3272.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>5</th>
      <td>KEAP1</td>
      <td>9817.0</td>
      <td>MSK</td>
      <td>GRCh37</td>
      <td>19</td>
      <td>10602256</td>
      <td>10602256</td>
      <td>+</td>
      <td>missense_variant</td>
      <td>Missense_Mutation</td>
      <td>SNP</td>
      <td>T</td>
      <td>T</td>
      <td>C</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>GENIE-MSK-P-0037236-T01-IM6</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.E441G</td>
      <td>ENST00000171111</td>
      <td>NM_203500.1</td>
      <td>441.0</td>
      <td>gAg/gGg</td>
      <td>3/6</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>PASS</td>
      <td>probably_damaging</td>
      <td>1.000</td>
      <td>deleterious</td>
      <td>0.00</td>
      <td>NaN</td>
      <td>451.0</td>
      <td>711.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>6</th>
      <td>TP53</td>
      <td>7157.0</td>
      <td>VICC</td>
      <td>GRCh37</td>
      <td>17</td>
      <td>7578403</td>
      <td>7578403</td>
      <td>+</td>
      <td>missense_variant</td>
      <td>Missense_Mutation</td>
      <td>SNP</td>
      <td>C</td>
      <td>C</td>
      <td>A</td>
      <td>rs786202962</td>
      <td>NaN</td>
      <td>GENIE-VICC-915932-unk-2</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>unknown</td>
      <td>...</td>
      <td>p.C176F</td>
      <td>ENST00000269305</td>
      <td>NM_001126112.2</td>
      <td>176.0</td>
      <td>tGc/tTc</td>
      <td>5/11</td>
      <td>0.000004</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000009</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>PASS</td>
      <td>probably_damaging</td>
      <td>0.998</td>
      <td>deleterious</td>
      <td>0.00</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>228.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>7</th>
      <td>TP53</td>
      <td>7157.0</td>
      <td>YALE</td>
      <td>GRCh37</td>
      <td>17</td>
      <td>7578406</td>
      <td>7578406</td>
      <td>+</td>
      <td>missense_variant</td>
      <td>Missense_Mutation</td>
      <td>SNP</td>
      <td>C</td>
      <td>C</td>
      <td>T</td>
      <td>rs28934578</td>
      <td>NaN</td>
      <td>GENIE-YALE-TPL817-1</td>
      <td>NORMAL</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.R175H</td>
      <td>ENST00000269305</td>
      <td>NM_001126112.2</td>
      <td>175.0</td>
      <td>cGc/cAc</td>
      <td>5/11</td>
      <td>0.000004</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000009</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>PASS</td>
      <td>benign</td>
      <td>0.319</td>
      <td>tolerated</td>
      <td>0.11</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>934.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>8</th>
      <td>TP53</td>
      <td>7157.0</td>
      <td>VICC</td>
      <td>GRCh37</td>
      <td>17</td>
      <td>7578517</td>
      <td>7578517</td>
      <td>+</td>
      <td>missense_variant</td>
      <td>Missense_Mutation</td>
      <td>SNP</td>
      <td>G</td>
      <td>G</td>
      <td>A</td>
      <td>rs750600586</td>
      <td>NaN</td>
      <td>GENIE-VICC-849730-unk-1</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>unknown</td>
      <td>...</td>
      <td>p.A138V</td>
      <td>ENST00000269305</td>
      <td>NM_001126112.2</td>
      <td>138.0</td>
      <td>gCc/gTc</td>
      <td>5/11</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>PASS</td>
      <td>possibly_damaging</td>
      <td>0.770</td>
      <td>deleterious</td>
      <td>0.03</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>518.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>9</th>
      <td>TP53</td>
      <td>7157.0</td>
      <td>DFCI</td>
      <td>GRCh37</td>
      <td>17</td>
      <td>7578413</td>
      <td>7578413</td>
      <td>+</td>
      <td>frameshift_variant</td>
      <td>Frame_Shift_Del</td>
      <td>DEL</td>
      <td>C</td>
      <td>C</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>GENIE-DFCI-007749-4752</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.V173*</td>
      <td>ENST00000269305</td>
      <td>NM_001126112.2</td>
      <td>173.0</td>
      <td>Gtg/tg</td>
      <td>5/11</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>PASS</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>196.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>10</th>
      <td>KEAP1</td>
      <td>9817.0</td>
      <td>DFCI</td>
      <td>GRCh37</td>
      <td>19</td>
      <td>10610391</td>
      <td>10610392</td>
      <td>+</td>
      <td>frameshift_variant</td>
      <td>Frame_Shift_Del</td>
      <td>DEL</td>
      <td>AG</td>
      <td>AG</td>
      <td>-</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>GENIE-DFCI-008260-5003</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.F107Qfs*19</td>
      <td>ENST00000171111</td>
      <td>NM_203500.1</td>
      <td>106.0</td>
      <td>gtCTtc/gttc</td>
      <td>2/6</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>PASS</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>74.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>11</th>
      <td>TP53</td>
      <td>7157.0</td>
      <td>MSK</td>
      <td>GRCh37</td>
      <td>17</td>
      <td>7578291</td>
      <td>7578291</td>
      <td>+</td>
      <td>splice_acceptor_variant</td>
      <td>Splice_Site</td>
      <td>SNP</td>
      <td>T</td>
      <td>T</td>
      <td>A</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>GENIE-MSK-P-0055396-T01-IM6</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.X187_splice</td>
      <td>ENST00000269305</td>
      <td>NM_001126112.2</td>
      <td>187.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>PASS</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>558.0</td>
      <td>788.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>12</th>
      <td>TP53</td>
      <td>7157.0</td>
      <td>DFCI</td>
      <td>GRCh37</td>
      <td>17</td>
      <td>7577127</td>
      <td>7577127</td>
      <td>+</td>
      <td>stop_gained</td>
      <td>Nonsense_Mutation</td>
      <td>SNP</td>
      <td>C</td>
      <td>C</td>
      <td>A</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>GENIE-DFCI-242287-3617171</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.E271*</td>
      <td>ENST00000269305</td>
      <td>NM_001126112.2</td>
      <td>271.0</td>
      <td>Gag/Tag</td>
      <td>8/11</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>PASS</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>243.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>13</th>
      <td>TP53</td>
      <td>7157.0</td>
      <td>MSK</td>
      <td>GRCh37</td>
      <td>17</td>
      <td>7578392</td>
      <td>7578392</td>
      <td>+</td>
      <td>stop_gained</td>
      <td>Nonsense_Mutation</td>
      <td>SNP</td>
      <td>C</td>
      <td>C</td>
      <td>A</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>GENIE-MSK-P-0033205-T01-IM6</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.E180*</td>
      <td>ENST00000269305</td>
      <td>NM_001126112.2</td>
      <td>180.0</td>
      <td>Gag/Tag</td>
      <td>5/11</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>PASS</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>487.0</td>
      <td>248.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>14</th>
      <td>TP53</td>
      <td>7157.0</td>
      <td>MSK</td>
      <td>GRCh37</td>
      <td>17</td>
      <td>7577121</td>
      <td>7577121</td>
      <td>+</td>
      <td>missense_variant</td>
      <td>Missense_Mutation</td>
      <td>SNP</td>
      <td>G</td>
      <td>G</td>
      <td>A</td>
      <td>rs121913343</td>
      <td>NaN</td>
      <td>GENIE-MSK-P-0027826-T02-IM6</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.R273C</td>
      <td>ENST00000269305</td>
      <td>NM_001126112.2</td>
      <td>273.0</td>
      <td>Cgt/Tgt</td>
      <td>8/11</td>
      <td>0.000012</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000054</td>
      <td>0.000046</td>
      <td>0.000009</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>PASS</td>
      <td>probably_damaging</td>
      <td>0.999</td>
      <td>deleterious</td>
      <td>0.00</td>
      <td>NaN</td>
      <td>548.0</td>
      <td>523.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>15</th>
      <td>TP53</td>
      <td>7157.0</td>
      <td>DFCI</td>
      <td>GRCh37</td>
      <td>17</td>
      <td>7577548</td>
      <td>7577548</td>
      <td>+</td>
      <td>missense_variant</td>
      <td>Missense_Mutation</td>
      <td>SNP</td>
      <td>C</td>
      <td>C</td>
      <td>A</td>
      <td>rs28934575</td>
      <td>NaN</td>
      <td>GENIE-DFCI-091107-2518440</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.G245C</td>
      <td>ENST00000269305</td>
      <td>NM_001126112.2</td>
      <td>245.0</td>
      <td>Ggc/Tgc</td>
      <td>7/11</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>PASS</td>
      <td>probably_damaging</td>
      <td>1.000</td>
      <td>deleterious</td>
      <td>0.00</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>195.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>16</th>
      <td>KEAP1</td>
      <td>9817.0</td>
      <td>MSK</td>
      <td>GRCh37</td>
      <td>19</td>
      <td>10602445</td>
      <td>10602445</td>
      <td>+</td>
      <td>missense_variant</td>
      <td>Missense_Mutation</td>
      <td>SNP</td>
      <td>C</td>
      <td>C</td>
      <td>T</td>
      <td>rs1244523465</td>
      <td>NaN</td>
      <td>GENIE-MSK-P-0059013-T02-IM7</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.G378D</td>
      <td>ENST00000171111</td>
      <td>NM_203500.1</td>
      <td>378.0</td>
      <td>gGc/gAc</td>
      <td>3/6</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>PASS</td>
      <td>probably_damaging</td>
      <td>1.000</td>
      <td>deleterious</td>
      <td>0.00</td>
      <td>NaN</td>
      <td>786.0</td>
      <td>994.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>17</th>
      <td>DNMT3A</td>
      <td>1788.0</td>
      <td>UCHI</td>
      <td>GRCh37</td>
      <td>2</td>
      <td>25457243</td>
      <td>25457243</td>
      <td>+</td>
      <td>missense_variant</td>
      <td>Missense_Mutation</td>
      <td>SNP</td>
      <td>G</td>
      <td>NaN</td>
      <td>A</td>
      <td>rs377577594</td>
      <td>NaN</td>
      <td>GENIE-UCHI-Patient347-T1</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.R882C</td>
      <td>ENST00000264709</td>
      <td>NM_175629.2</td>
      <td>882.0</td>
      <td>Cgc/Tgc</td>
      <td>23/23</td>
      <td>0.000123</td>
      <td>0.000185</td>
      <td>0.000029</td>
      <td>0.000199</td>
      <td>0.000218</td>
      <td>0.000139</td>
      <td>0.000123</td>
      <td>0.000326</td>
      <td>0.000065</td>
      <td>PASS</td>
      <td>probably_damaging</td>
      <td>0.958</td>
      <td>deleterious</td>
      <td>0.02</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>2936.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>18</th>
      <td>TP53</td>
      <td>7157.0</td>
      <td>NKI</td>
      <td>GRCh37</td>
      <td>17</td>
      <td>7578475</td>
      <td>7578475</td>
      <td>+</td>
      <td>missense_variant</td>
      <td>Missense_Mutation</td>
      <td>SNP</td>
      <td>G</td>
      <td>G</td>
      <td>A</td>
      <td>rs587782705</td>
      <td>NaN</td>
      <td>GENIE-NKI-HEQH-F0LA</td>
      <td>NORMAL</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.P152L</td>
      <td>ENST00000269305</td>
      <td>NM_001126112.2</td>
      <td>152.0</td>
      <td>cCg/cTg</td>
      <td>5/11</td>
      <td>0.000008</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000099</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>0.000009</td>
      <td>0.000000</td>
      <td>0.000000</td>
      <td>PASS</td>
      <td>possibly_damaging</td>
      <td>0.886</td>
      <td>tolerated</td>
      <td>0.08</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>2668.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
    <tr>
      <th>19</th>
      <td>TP53</td>
      <td>7157.0</td>
      <td>DFCI</td>
      <td>GRCh37</td>
      <td>17</td>
      <td>7579321</td>
      <td>7579322</td>
      <td>+</td>
      <td>frameshift_variant</td>
      <td>Frame_Shift_Del</td>
      <td>DEL</td>
      <td>CA</td>
      <td>CA</td>
      <td>-</td>
      <td>rs587780067</td>
      <td>NaN</td>
      <td>GENIE-DFCI-109522-950510</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>Unknown</td>
      <td>...</td>
      <td>p.V122Dfs*26</td>
      <td>ENST00000269305</td>
      <td>NM_001126112.2</td>
      <td>122.0</td>
      <td>gTG/g</td>
      <td>4/11</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>PASS</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>298.0</td>
      <td>SUCCESS</td>
      <td>False</td>
    </tr>
  </tbody>
</table>
<p>20 rows  64 columns</p>
</div>




```python
list_names = list(df['Hugo_Symbol'].unique())
align_dict = dict()
human_tx_id_dict = dict()
mouse_tx_id_dict = dict()
for gene_name_h in list_names:
    tx_id_h = h2m.get_tx_id(gene_name_h,'h',ver=37, show=False)[3]
    gene_name_m = h2m.query(gene_name_h, show=False)[0]['gene_name_m']
    tx_id_m = h2m.get_tx_id(gene_name_m,'m', show=False)[3]
    align_dict[tx_id_h] = h2m.get_align(records_h, index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, ver = 37)
    human_tx_id_dict[gene_name_h] = tx_id_h
    mouse_tx_id_dict[gene_name_h] = tx_id_m
```

    /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
      warnings.warn(



```python
human_tx_id_dict
```




    {'TP53': 'ENST00000269305.4',
     'DNMT3A': 'ENST00000264709.3',
     'KEAP1': 'ENST00000171111.5'}




```python
gene_name_h, start_h, end_h, ty_h, ref_seq_h, alt_seq_h
```




    (35207, 17, 7577574, 'Missense_Mutation', 'SNP', 'T')




```python
result_list = []
for i in range(len(df)):
    gene_name_h, start_h, end_h, ty_h, ref_seq_h, alt_seq_h = df.iloc[i,[0,5,6,10,11,13]].to_list()
    tx_id_h = human_tx_id_dict[gene_name_h]
    tx_id_m = mouse_tx_id_dict[gene_name_h]
    result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, start_h, end_h, alt_seq_h, ty_h,ver = 37, ref_seq_h=ref_seq_h)

    if type(result) == list: # multiple alternative modeling
        for k in result:
            k['index'] = i
            result_list.append(k)
    else:
        result['index'] = i
        result_list.append(result)

df_result = pd.DataFrame(result_list)
# df_result.to_csv('/result/test_data_result.csv',index=False)
df_result
```

    /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/Bio/Seq.py:2804: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.
      warnings.warn(





<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_name_h</th>
      <th>gene_id_h</th>
      <th>tx_id_m</th>
      <th>chr_h</th>
      <th>exon_num_h</th>
      <th>strand_h</th>
      <th>start_h</th>
      <th>end_h</th>
      <th>ref_seq_h</th>
      <th>alt_seq_h</th>
      <th>HGVSp_h</th>
      <th>classification_h</th>
      <th>classification_h_detail</th>
      <th>exon_h</th>
      <th>type_h</th>
      <th>status</th>
      <th>error</th>
      <th>statement</th>
      <th>gene_name_m</th>
      <th>gene_id_m</th>
      <th>chr_m</th>
      <th>exon_num_m</th>
      <th>strand_m</th>
      <th>type_m</th>
      <th>classification_m</th>
      <th>exon_m</th>
      <th>start_m_ori</th>
      <th>end_m_ori</th>
      <th>ref_seq_m_ori</th>
      <th>alt_seq_m_ori</th>
      <th>HGVSp_m_ori</th>
      <th>start_m</th>
      <th>end_m</th>
      <th>ref_seq_m</th>
      <th>alt_seq_m</th>
      <th>HGVSp_m</th>
      <th>index</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>DNMT3A</td>
      <td>ENSG00000119772.12</td>
      <td>ENSMUST00000020991.15</td>
      <td>chr2</td>
      <td>22</td>
      <td>-</td>
      <td>25468174</td>
      <td>25468174</td>
      <td>T</td>
      <td>C</td>
      <td>N501S</td>
      <td>Missense</td>
      <td>None</td>
      <td>E_12</td>
      <td>SNP</td>
      <td>True</td>
      <td>0</td>
      <td>This mutation can be originally modeled.</td>
      <td>Dnmt3a</td>
      <td>ENSMUSG00000020661.16</td>
      <td>chr12</td>
      <td>22</td>
      <td>+</td>
      <td>SNP</td>
      <td>Missense</td>
      <td>E_12</td>
      <td>3949220.0</td>
      <td>3949220.0</td>
      <td>A</td>
      <td>G</td>
      <td>N497S</td>
      <td>3949220.0</td>
      <td>3949220.0</td>
      <td>A</td>
      <td>G</td>
      <td>N497S</td>
      <td>0</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7578212</td>
      <td>7578212</td>
      <td>G</td>
      <td>A</td>
      <td>R213delins*</td>
      <td>Nonsense</td>
      <td>None</td>
      <td>E_5</td>
      <td>SNP</td>
      <td>True</td>
      <td>1</td>
      <td>Error 1: This mutation can be alternatively mo...</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>ONP</td>
      <td>Nonsense</td>
      <td>E_5</td>
      <td>69479529.0</td>
      <td>69479529.0</td>
      <td>C</td>
      <td>T</td>
      <td>R210C</td>
      <td>69479529.0</td>
      <td>69479531.0</td>
      <td>CGC</td>
      <td>TAA</td>
      <td>R210delins*</td>
      <td>1</td>
    </tr>
    <tr>
      <th>2</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7578212</td>
      <td>7578212</td>
      <td>C</td>
      <td>T</td>
      <td>R213delins*</td>
      <td>Nonsense</td>
      <td>None</td>
      <td>E_5</td>
      <td>SNP</td>
      <td>True</td>
      <td>1</td>
      <td>Error 1: This mutation can be alternatively mo...</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>ONP</td>
      <td>Nonsense</td>
      <td>E_5</td>
      <td>69479529.0</td>
      <td>69479529.0</td>
      <td>C</td>
      <td>T</td>
      <td>R210C</td>
      <td>69479529.0</td>
      <td>69479531.0</td>
      <td>CGC</td>
      <td>TAG</td>
      <td>R210delins*</td>
      <td>1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7578212</td>
      <td>7578212</td>
      <td>G</td>
      <td>A</td>
      <td>R213delins*</td>
      <td>Nonsense</td>
      <td>None</td>
      <td>E_5</td>
      <td>SNP</td>
      <td>True</td>
      <td>1</td>
      <td>Error 1: This mutation can be alternatively mo...</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>ONP</td>
      <td>Nonsense</td>
      <td>E_5</td>
      <td>69479529.0</td>
      <td>69479529.0</td>
      <td>C</td>
      <td>T</td>
      <td>R210C</td>
      <td>69479529.0</td>
      <td>69479531.0</td>
      <td>CGC</td>
      <td>TGA</td>
      <td>R210delins*</td>
      <td>1</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7577090</td>
      <td>7577090</td>
      <td>C</td>
      <td>G</td>
      <td>R283P</td>
      <td>Missense</td>
      <td>None</td>
      <td>E_7</td>
      <td>SNP</td>
      <td>True</td>
      <td>0</td>
      <td>This mutation can be originally modeled.</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>SNP</td>
      <td>Missense</td>
      <td>E_7</td>
      <td>69480464.0</td>
      <td>69480464.0</td>
      <td>G</td>
      <td>C</td>
      <td>R280P</td>
      <td>69480464.0</td>
      <td>69480464.0</td>
      <td>G</td>
      <td>C</td>
      <td>R280P</td>
      <td>2</td>
    </tr>
    <tr>
      <th>5</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7578406</td>
      <td>7578406</td>
      <td>C</td>
      <td>A</td>
      <td>R175L</td>
      <td>Missense</td>
      <td>None</td>
      <td>E_4</td>
      <td>SNP</td>
      <td>True</td>
      <td>0</td>
      <td>This mutation can be originally modeled.</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>SNP</td>
      <td>Missense</td>
      <td>E_4</td>
      <td>69479338.0</td>
      <td>69479338.0</td>
      <td>G</td>
      <td>T</td>
      <td>R172L</td>
      <td>69479338.0</td>
      <td>69479338.0</td>
      <td>G</td>
      <td>T</td>
      <td>R172L</td>
      <td>3</td>
    </tr>
    <tr>
      <th>6</th>
      <td>DNMT3A</td>
      <td>ENSG00000119772.12</td>
      <td>ENSMUST00000020991.15</td>
      <td>chr2</td>
      <td>22</td>
      <td>-</td>
      <td>25463532</td>
      <td>25463535</td>
      <td>TG</td>
      <td>CTCCTCT</td>
      <td>N717Efs*64</td>
      <td>Frame_Shifting</td>
      <td>None</td>
      <td>E_17</td>
      <td>INS</td>
      <td>True</td>
      <td>2</td>
      <td>Error 2: This mutation can be modeled, but the...</td>
      <td>Dnmt3a</td>
      <td>ENSMUSG00000020661.16</td>
      <td>chr12</td>
      <td>22</td>
      <td>+</td>
      <td>INS</td>
      <td>Frame_Shifting</td>
      <td>E_17</td>
      <td>3952445.0</td>
      <td>3952448.0</td>
      <td>CA</td>
      <td>AGAGGAG</td>
      <td>N713Efs*64</td>
      <td>3952445.0</td>
      <td>3952448.0</td>
      <td>CA</td>
      <td>AGAGGAG</td>
      <td>N713Efs*64</td>
      <td>4</td>
    </tr>
    <tr>
      <th>7</th>
      <td>KEAP1</td>
      <td>ENSG00000079999.9</td>
      <td>ENSMUST00000164812.8</td>
      <td>chr19</td>
      <td>5</td>
      <td>-</td>
      <td>10602256</td>
      <td>10602256</td>
      <td>T</td>
      <td>C</td>
      <td>E441G</td>
      <td>Missense</td>
      <td>None</td>
      <td>E_2</td>
      <td>SNP</td>
      <td>True</td>
      <td>0</td>
      <td>This mutation can be originally modeled.</td>
      <td>Keap1</td>
      <td>ENSMUSG00000003308.16</td>
      <td>chr9</td>
      <td>5</td>
      <td>-</td>
      <td>SNP</td>
      <td>Missense</td>
      <td>E_2</td>
      <td>21144687.0</td>
      <td>21144687.0</td>
      <td>T</td>
      <td>C</td>
      <td>E441G</td>
      <td>21144687.0</td>
      <td>21144687.0</td>
      <td>A</td>
      <td>G</td>
      <td>E441G</td>
      <td>5</td>
    </tr>
    <tr>
      <th>8</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7578403</td>
      <td>7578403</td>
      <td>C</td>
      <td>A</td>
      <td>C176F</td>
      <td>Missense</td>
      <td>None</td>
      <td>E_4</td>
      <td>SNP</td>
      <td>True</td>
      <td>0</td>
      <td>This mutation can be originally modeled.</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>SNP</td>
      <td>Missense</td>
      <td>E_4</td>
      <td>69479341.0</td>
      <td>69479341.0</td>
      <td>G</td>
      <td>T</td>
      <td>C173F</td>
      <td>69479341.0</td>
      <td>69479341.0</td>
      <td>G</td>
      <td>T</td>
      <td>C173F</td>
      <td>6</td>
    </tr>
    <tr>
      <th>9</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7578406</td>
      <td>7578406</td>
      <td>C</td>
      <td>T</td>
      <td>R175H</td>
      <td>Missense</td>
      <td>None</td>
      <td>E_4</td>
      <td>SNP</td>
      <td>True</td>
      <td>0</td>
      <td>This mutation can be originally modeled.</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>SNP</td>
      <td>Missense</td>
      <td>E_4</td>
      <td>69479338.0</td>
      <td>69479338.0</td>
      <td>G</td>
      <td>A</td>
      <td>R172H</td>
      <td>69479338.0</td>
      <td>69479338.0</td>
      <td>G</td>
      <td>A</td>
      <td>R172H</td>
      <td>7</td>
    </tr>
    <tr>
      <th>10</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7578517</td>
      <td>7578517</td>
      <td>G</td>
      <td>A</td>
      <td>A138V</td>
      <td>Missense</td>
      <td>None</td>
      <td>E_4</td>
      <td>SNP</td>
      <td>True</td>
      <td>0</td>
      <td>This mutation can be originally modeled.</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>SNP</td>
      <td>Missense</td>
      <td>E_4</td>
      <td>69479227.0</td>
      <td>69479227.0</td>
      <td>C</td>
      <td>T</td>
      <td>A135V</td>
      <td>69479227.0</td>
      <td>69479227.0</td>
      <td>C</td>
      <td>T</td>
      <td>A135V</td>
      <td>8</td>
    </tr>
    <tr>
      <th>11</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7578413</td>
      <td>7578413</td>
      <td>C</td>
      <td></td>
      <td>V173*</td>
      <td>Frame_Shifting</td>
      <td>None</td>
      <td>E_4</td>
      <td>DEL</td>
      <td>True</td>
      <td>2</td>
      <td>Error 2: This mutation can be modeled, but the...</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>DEL</td>
      <td>Frame_Shifting</td>
      <td>E_4</td>
      <td>69479331.0</td>
      <td>69479331.0</td>
      <td>G</td>
      <td></td>
      <td>V170*</td>
      <td>69479331.0</td>
      <td>69479331.0</td>
      <td>G</td>
      <td></td>
      <td>V170*</td>
      <td>9</td>
    </tr>
    <tr>
      <th>12</th>
      <td>KEAP1</td>
      <td>ENSG00000079999.9</td>
      <td>ENSMUST00000164812.8</td>
      <td>chr19</td>
      <td>5</td>
      <td>-</td>
      <td>10610391</td>
      <td>10610392</td>
      <td>AG</td>
      <td></td>
      <td>F107Qfs*19</td>
      <td>Frame_Shifting</td>
      <td>None</td>
      <td>E_1</td>
      <td>DEL</td>
      <td>True</td>
      <td>2</td>
      <td>Error 2: This mutation can be modeled, but the...</td>
      <td>Keap1</td>
      <td>ENSMUSG00000003308.16</td>
      <td>chr9</td>
      <td>5</td>
      <td>-</td>
      <td>DEL</td>
      <td>Frame_Shifting</td>
      <td>E_1</td>
      <td>21148687.0</td>
      <td>21148688.0</td>
      <td>AG</td>
      <td></td>
      <td>F107*</td>
      <td>21148687.0</td>
      <td>21148688.0</td>
      <td>CT</td>
      <td></td>
      <td>F107*</td>
      <td>10</td>
    </tr>
    <tr>
      <th>13</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7578291</td>
      <td>7578291</td>
      <td>T</td>
      <td>A</td>
      <td>None</td>
      <td>Non_Coding</td>
      <td>Intron</td>
      <td>None</td>
      <td>SNP</td>
      <td>True</td>
      <td>2</td>
      <td>Error 2: This mutation can be modeled, but the...</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>SNP</td>
      <td>Intron</td>
      <td>None</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>None</td>
      <td>None</td>
      <td>None</td>
      <td>69479450.0</td>
      <td>69479450.0</td>
      <td>A</td>
      <td>T</td>
      <td>None</td>
      <td>11</td>
    </tr>
    <tr>
      <th>14</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7577127</td>
      <td>7577127</td>
      <td>C</td>
      <td>A</td>
      <td>E271delins*</td>
      <td>Nonsense</td>
      <td>None</td>
      <td>E_7</td>
      <td>SNP</td>
      <td>True</td>
      <td>0</td>
      <td>This mutation can be originally modeled.</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>SNP</td>
      <td>Nonsense</td>
      <td>E_7</td>
      <td>69480427.0</td>
      <td>69480427.0</td>
      <td>G</td>
      <td>T</td>
      <td>E268delins*</td>
      <td>69480427.0</td>
      <td>69480427.0</td>
      <td>G</td>
      <td>T</td>
      <td>E268delins*</td>
      <td>12</td>
    </tr>
    <tr>
      <th>15</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7578392</td>
      <td>7578392</td>
      <td>C</td>
      <td>A</td>
      <td>E180delins*</td>
      <td>Nonsense</td>
      <td>None</td>
      <td>E_4</td>
      <td>SNP</td>
      <td>True</td>
      <td>0</td>
      <td>This mutation can be originally modeled.</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>SNP</td>
      <td>Nonsense</td>
      <td>E_4</td>
      <td>69479352.0</td>
      <td>69479352.0</td>
      <td>G</td>
      <td>T</td>
      <td>E177delins*</td>
      <td>69479352.0</td>
      <td>69479352.0</td>
      <td>G</td>
      <td>T</td>
      <td>E177delins*</td>
      <td>13</td>
    </tr>
    <tr>
      <th>16</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7577121</td>
      <td>7577121</td>
      <td>G</td>
      <td>A</td>
      <td>R273C</td>
      <td>Missense</td>
      <td>None</td>
      <td>E_7</td>
      <td>SNP</td>
      <td>True</td>
      <td>0</td>
      <td>This mutation can be originally modeled.</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>SNP</td>
      <td>Missense</td>
      <td>E_7</td>
      <td>69480433.0</td>
      <td>69480433.0</td>
      <td>C</td>
      <td>T</td>
      <td>R270C</td>
      <td>69480433.0</td>
      <td>69480433.0</td>
      <td>C</td>
      <td>T</td>
      <td>R270C</td>
      <td>14</td>
    </tr>
    <tr>
      <th>17</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7577548</td>
      <td>7577548</td>
      <td>C</td>
      <td>A</td>
      <td>G245C</td>
      <td>Missense</td>
      <td>None</td>
      <td>E_6</td>
      <td>SNP</td>
      <td>True</td>
      <td>0</td>
      <td>This mutation can be originally modeled.</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>SNP</td>
      <td>Missense</td>
      <td>E_6</td>
      <td>69480027.0</td>
      <td>69480027.0</td>
      <td>G</td>
      <td>T</td>
      <td>G242C</td>
      <td>69480027.0</td>
      <td>69480027.0</td>
      <td>G</td>
      <td>T</td>
      <td>G242C</td>
      <td>15</td>
    </tr>
    <tr>
      <th>18</th>
      <td>KEAP1</td>
      <td>ENSG00000079999.9</td>
      <td>ENSMUST00000164812.8</td>
      <td>chr19</td>
      <td>5</td>
      <td>-</td>
      <td>10602445</td>
      <td>10602445</td>
      <td>C</td>
      <td>T</td>
      <td>G378D</td>
      <td>Missense</td>
      <td>None</td>
      <td>E_2</td>
      <td>SNP</td>
      <td>True</td>
      <td>0</td>
      <td>This mutation can be originally modeled.</td>
      <td>Keap1</td>
      <td>ENSMUSG00000003308.16</td>
      <td>chr9</td>
      <td>5</td>
      <td>-</td>
      <td>SNP</td>
      <td>Missense</td>
      <td>E_2</td>
      <td>21144876.0</td>
      <td>21144876.0</td>
      <td>C</td>
      <td>T</td>
      <td>G378D</td>
      <td>21144876.0</td>
      <td>21144876.0</td>
      <td>G</td>
      <td>A</td>
      <td>G378D</td>
      <td>16</td>
    </tr>
    <tr>
      <th>19</th>
      <td>DNMT3A</td>
      <td>ENSG00000119772.12</td>
      <td>ENSMUST00000020991.15</td>
      <td>chr2</td>
      <td>22</td>
      <td>-</td>
      <td>25457243</td>
      <td>25457243</td>
      <td>G</td>
      <td>A</td>
      <td>R882C</td>
      <td>Missense</td>
      <td>None</td>
      <td>E_22</td>
      <td>SNP</td>
      <td>True</td>
      <td>0</td>
      <td>This mutation can be originally modeled.</td>
      <td>Dnmt3a</td>
      <td>ENSMUSG00000020661.16</td>
      <td>chr12</td>
      <td>22</td>
      <td>+</td>
      <td>SNP</td>
      <td>Missense</td>
      <td>E_22</td>
      <td>3957652.0</td>
      <td>3957652.0</td>
      <td>C</td>
      <td>T</td>
      <td>R878C</td>
      <td>3957652.0</td>
      <td>3957652.0</td>
      <td>C</td>
      <td>T</td>
      <td>R878C</td>
      <td>17</td>
    </tr>
    <tr>
      <th>20</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7578475</td>
      <td>7578475</td>
      <td>C</td>
      <td>T</td>
      <td>P152L</td>
      <td>Missense</td>
      <td>None</td>
      <td>E_4</td>
      <td>SNP</td>
      <td>False</td>
      <td>4</td>
      <td>Error 4: Flanked segments are not identical.</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>None</td>
      <td>None</td>
      <td>None</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>None</td>
      <td>None</td>
      <td>None</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>None</td>
      <td>None</td>
      <td>None</td>
      <td>18</td>
    </tr>
    <tr>
      <th>21</th>
      <td>TP53</td>
      <td>ENSG00000141510.11</td>
      <td>ENSMUST00000108658.10</td>
      <td>chr17</td>
      <td>10</td>
      <td>-</td>
      <td>7579321</td>
      <td>7579322</td>
      <td>TG</td>
      <td></td>
      <td>V122Dfs*26</td>
      <td>Frame_Shifting</td>
      <td>None</td>
      <td>E_3</td>
      <td>DEL</td>
      <td>False</td>
      <td>4</td>
      <td>Error 4: Flanked segments are not identical.</td>
      <td>Trp53</td>
      <td>ENSMUSG00000059552.14</td>
      <td>chr11</td>
      <td>10</td>
      <td>+</td>
      <td>None</td>
      <td>None</td>
      <td>None</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>None</td>
      <td>None</td>
      <td>None</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>None</td>
      <td>None</td>
      <td>None</td>
      <td>19</td>
    </tr>
  </tbody>
</table>
</div>



# Appendix

## 1. Database Sources

*Local* 
**Dictionaries**   
1. name_dict: Dictionary of human ensembl id to human gene name.  
2. homo_dict: Dictionary of to human gene name to all  (json-like format). (Prepared by using biomart)  
3. human_only_list: List of human-only gene names. (Prepared by using biomart)

**Databases**
1. GENECODE Databases  
- human version 37  
- human version 38  
- mouse  
2. Refgenome
- human version 37  
- human version 38  
- mouse  

*Online*
ENSEMBL API

## 2. Workflow

![workflow](images/workflow.png)

## 3. Functions

### 3.1 query(id, db,  ty, show)  
- Function  
    -  Query a human gene for the presence of mouse homologs.  
- Parameters   
    - id: string, identification of a human gene. Multiple input forms are accepted, including gene name, stable ensembl gene id with or without version number.  
    - ver (optional): int, specify the version of human, one of 37/38. Only necessary when using human gene ensembl id.  
    -  ty (optional): type of your input id. string, one of *name, gene_id*. Not necessary.  
 show (optional): bool, print summary of output or not.  
- output  
    -  A list of all mouse orthologs (if have; sometimes more than one.) Each element is a dictionary of mouse gene name, mouse gene id, homology type and similarity of human and mouse gene in percentage.  
- Example  


```python
h2m.query('TP53')
```

    Query human gene: TP53;
    Mouse ortholog(s): Trp53;
    Homology type: one2one;
    Sequence Simalarity(%):77.3537.





    [{'gene_name_m': 'Trp53',
      'gene_id_m': 'ENSMUSG00000059552',
      'homology_type': 'ortholog_one2one',
      'similarity': 77.3537}]




```python
h2m.query('ENSG00000141510')
```

    Query human gene: TP53;
    Mouse ortholog(s): Trp53;
    Homology type: one2one;
    Sequence Simalarity(%):77.3537.





    [{'gene_name_m': 'Trp53',
      'gene_id_m': 'ENSMUSG00000059552',
      'homology_type': 'ortholog_one2one',
      'similarity': 77.3537}]



### 3.2 get_tx(id, species, ver, ty, show)    
- Function  
    - Query a human or mouse gene for coordinate and information of all its transcripts.  

- Parameters   
    - id: string, identification of a human gene. Multiple input forms are accepted, including gene   
    - species: string, h for human or m for mouse.  
    - name, stable ensembl gene id with or without version number.  
    - ver: int, specify the version of human, one of 37/38. It is a necessary parameter.  
    - ty (optional): type of your input id. string, one of *name, gene_id*. Not necessary.  
    - show (optional): bool, print summary of output or not.  
- output  
    - A list [chromosome, start location(of gene), end location(of gene), canonical transcript id, list of all transcript id (the canonical one included and always at the first place), a list of additional information of each transcript]  
- example  


```python
h2m.get_tx_id('TP53','h',ver=37)
```

    Genome assembly: GRCh37;
    The canonical transcript is: ENST00000269305.4;
    You can choose from the 17 transcripts below for further analysis:
    (1)ENST00000269305.4 (2)ENST00000413465.2 (3)ENST00000359597.4 (4)ENST00000504290.1 (5)ENST00000510385.1 (6)ENST00000504937.1 (7)ENST00000455263.2 (8)ENST00000420246.2 (9)ENST00000445888.2 (10)ENST00000576024.1 (11)ENST00000509690.1 (12)ENST00000514944.1 (13)ENST00000574684.1 (14)ENST00000505014.1 (15)ENST00000508793.1 (16)ENST00000604348.1 (17)ENST00000503591.1
    





    [17,
     7565097,
     7590856,
     'ENST00000269305.4',
     ['ENST00000269305.4',
      'ENST00000413465.2',
      'ENST00000359597.4',
      'ENST00000504290.1',
      'ENST00000510385.1',
      'ENST00000504937.1',
      'ENST00000455263.2',
      'ENST00000420246.2',
      'ENST00000445888.2',
      'ENST00000576024.1',
      'ENST00000509690.1',
      'ENST00000514944.1',
      'ENST00000574684.1',
      'ENST00000505014.1',
      'ENST00000508793.1',
      'ENST00000604348.1',
      'ENST00000503591.1'],
     [{'object_type': 'Transcript',
       'source': 'havana',
       'display_name': 'TP53-018',
       'id': 'ENST00000413465',
       'start': 7565097,
       'end': 7579912,
       'db_type': 'core',
       'assembly_name': 'GRCh37',
       'biotype': 'protein_coding',
       'version': 2,
       'Exon': [{'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'id': 'ENSE00002337729',
         'start': 7579839,
         'end': 7579912,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7579700,
         'id': 'ENSE00002419584',
         'end': 7579721,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00003625790',
         'start': 7579312,
         'end': 7579590},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00003518480',
         'start': 7578371,
         'end': 7578554,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00003462942',
         'start': 7578177,
         'end': 7578289,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17',
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7577499,
         'id': 'ENSE00003504863',
         'end': 7577608},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7565097,
         'id': 'ENSE00001657961',
         'end': 7565332,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 2}],
       'Parent': 'ENSG00000141510',
       'species': 'homo_sapiens',
       'strand': -1,
       'Translation': {'length': 285,
        'species': 'homo_sapiens',
        'Parent': 'ENST00000413465',
        'object_type': 'Translation',
        'end': 7579912,
        'id': 'ENSP00000410739',
        'start': 7565257,
        'db_type': 'core',
        'version': 2},
       'is_canonical': 0,
       'logic_name': 'havana_homo_sapiens_37',
       'seq_region_name': '17'},
      {'assembly_name': 'GRCh37',
       'db_type': 'core',
       'Exon': [{'end': 7579912,
         'id': 'ENSE00002337729',
         'start': 7579839,
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1},
        {'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17',
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7579700,
         'id': 'ENSE00002419584',
         'end': 7579721},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7579590,
         'id': 'ENSE00003625790',
         'start': 7579312,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7578371,
         'id': 'ENSE00003518480',
         'end': 7578554},
        {'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17',
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7578177,
         'id': 'ENSE00003462942',
         'end': 7578289},
        {'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00003504863',
         'start': 7577499,
         'end': 7577608},
        {'version': 1,
         'seq_region_name': '17',
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'start': 7577019,
         'id': 'ENSE00003586720',
         'end': 7577155,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'end': 7576926,
         'start': 7576853,
         'id': 'ENSE00003636029',
         'species': 'homo_sapiens',
         'object_type': 'Exon'},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7569562,
         'start': 7569404,
         'id': 'ENSE00002204316',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'}],
       'version': 4,
       'biotype': 'protein_coding',
       'object_type': 'Transcript',
       'end': 7579912,
       'display_name': 'TP53-019',
       'start': 7569404,
       'id': 'ENST00000359597',
       'source': 'havana',
       'Translation': {'end': 7579912,
        'id': 'ENSP00000352610',
        'start': 7569524,
        'Parent': 'ENST00000359597',
        'species': 'homo_sapiens',
        'object_type': 'Translation',
        'length': 343,
        'version': 4,
        'db_type': 'core'},
       'strand': -1,
       'seq_region_name': '17',
       'is_canonical': 0,
       'logic_name': 'havana_homo_sapiens_37',
       'Parent': 'ENSG00000141510',
       'species': 'homo_sapiens'},
      {'db_type': 'core',
       'assembly_name': 'GRCh37',
       'version': 1,
       'Exon': [{'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'end': 7578811,
         'start': 7578371,
         'id': 'ENSE00002064269',
         'species': 'homo_sapiens',
         'object_type': 'Exon'},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00003465983',
         'start': 7578177,
         'end': 7578289,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7577608,
         'start': 7577499,
         'id': 'ENSE00003475276',
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'seq_region_name': '17',
         'version': 1},
        {'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7577155,
         'start': 7577019,
         'id': 'ENSE00003692058'},
        {'start': 7576853,
         'id': 'ENSE00003516179',
         'end': 7576926,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37'},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7576525,
         'id': 'ENSE00003576506',
         'end': 7576584,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1},
        {'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'id': 'ENSE00003634848',
         'start': 7573927,
         'end': 7574033,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7573008,
         'id': 'ENSE00003492844',
         'start': 7571720,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'seq_region_name': '17',
         'version': 1}],
       'biotype': 'retained_intron',
       'object_type': 'Transcript',
       'id': 'ENST00000504290',
       'display_name': 'TP53-006',
       'start': 7571720,
       'end': 7578811,
       'source': 'havana',
       'strand': -1,
       'is_canonical': 0,
       'logic_name': 'havana_homo_sapiens_37',
       'seq_region_name': '17',
       'Parent': 'ENSG00000141510',
       'species': 'homo_sapiens'},
      {'Parent': 'ENSG00000141510',
       'species': 'homo_sapiens',
       'seq_region_name': '17',
       'is_canonical': 0,
       'logic_name': 'havana_homo_sapiens_37',
       'strand': -1,
       'source': 'havana',
       'end': 7578811,
       'display_name': 'TP53-007',
       'id': 'ENST00000510385',
       'start': 7571720,
       'object_type': 'Transcript',
       'biotype': 'retained_intron',
       'Exon': [{'version': 1,
         'seq_region_name': '17',
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'id': 'ENSE00002064269',
         'start': 7578371,
         'end': 7578811,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7578289,
         'start': 7578177,
         'id': 'ENSE00003465983',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'end': 7577608,
         'start': 7577499,
         'id': 'ENSE00003475276',
         'species': 'homo_sapiens',
         'object_type': 'Exon'},
        {'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7577155,
         'id': 'ENSE00003692058',
         'start': 7577019},
        {'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00003516179',
         'start': 7576853,
         'end': 7576926},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7576525,
         'id': 'ENSE00003496026',
         'end': 7576657,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7573927,
         'id': 'ENSE00003634848',
         'end': 7574033,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'start': 7571720,
         'id': 'ENSE00003492844',
         'end': 7573008,
         'object_type': 'Exon',
         'species': 'homo_sapiens'}],
       'version': 1,
       'assembly_name': 'GRCh37',
       'db_type': 'core'},
      {'db_type': 'core',
       'assembly_name': 'GRCh37',
       'biotype': 'retained_intron',
       'version': 1,
       'Exon': [{'id': 'ENSE00002064269',
         'start': 7578371,
         'end': 7578811,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'version': 1,
         'seq_region_name': '17',
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1},
        {'end': 7578289,
         'id': 'ENSE00003465983',
         'start': 7578177,
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core'},
        {'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7577608,
         'start': 7577499,
         'id': 'ENSE00003475276'},
        {'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'end': 7577155,
         'id': 'ENSE00003692058',
         'start': 7577019,
         'species': 'homo_sapiens',
         'object_type': 'Exon'},
        {'end': 7576926,
         'start': 7576853,
         'id': 'ENSE00003516179',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7573927,
         'id': 'ENSE00003634848',
         'end': 7574033,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7571720,
         'id': 'ENSE00003492844',
         'end': 7573008,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'}],
       'object_type': 'Transcript',
       'source': 'havana',
       'start': 7571720,
       'display_name': 'TP53-008',
       'id': 'ENST00000504937',
       'end': 7578811,
       'strand': -1,
       'is_canonical': 0,
       'logic_name': 'havana_homo_sapiens_37',
       'seq_region_name': '17',
       'species': 'homo_sapiens',
       'Parent': 'ENSG00000141510'},
      {'Parent': 'ENSG00000141510',
       'species': 'homo_sapiens',
       'is_canonical': 1,
       'logic_name': 'ensembl_havana_transcript',
       'seq_region_name': '17',
       'strand': -1,
       'Translation': {'db_type': 'core',
        'version': 4,
        'object_type': 'Translation',
        'species': 'homo_sapiens',
        'Parent': 'ENST00000269305',
        'length': 393,
        'start': 7572927,
        'id': 'ENSP00000269305',
        'end': 7579912},
       'source': 'ensembl_havana',
       'display_name': 'TP53-001',
       'id': 'ENST00000269305',
       'start': 7571720,
       'end': 7590856,
       'object_type': 'Transcript',
       'biotype': 'protein_coding',
       'version': 4,
       'Exon': [{'version': 3,
         'seq_region_name': '17',
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'start': 7590695,
         'id': 'ENSE00001146308',
         'end': 7590856,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7579940,
         'id': 'ENSE00002667911',
         'start': 7579839,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'seq_region_name': '17',
         'version': 1},
        {'version': 1,
         'seq_region_name': '17',
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'id': 'ENSE00002419584',
         'start': 7579700,
         'end': 7579721,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'end': 7579590,
         'start': 7579312,
         'id': 'ENSE00003625790',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core'},
        {'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'seq_region_name': '17',
         'version': 1,
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7578554,
         'id': 'ENSE00003518480',
         'start': 7578371},
        {'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'start': 7578177,
         'id': 'ENSE00003462942',
         'end': 7578289,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'version': 1,
         'seq_region_name': '17',
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'start': 7577499,
         'id': 'ENSE00003504863',
         'end': 7577608,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'seq_region_name': '17',
         'version': 1,
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7577155,
         'start': 7577019,
         'id': 'ENSE00003586720'},
        {'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'end': 7576926,
         'id': 'ENSE00003636029',
         'start': 7576853,
         'species': 'homo_sapiens',
         'object_type': 'Exon'},
        {'version': 1,
         'seq_region_name': '17',
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'start': 7573927,
         'id': 'ENSE00003545950',
         'end': 7574033,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7571720,
         'id': 'ENSE00003605891',
         'end': 7573008,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'}],
       'db_type': 'core',
       'assembly_name': 'GRCh37'},
      {'strand': -1,
       'Translation': {'object_type': 'Translation',
        'species': 'homo_sapiens',
        'Parent': 'ENST00000455263',
        'length': 346,
        'start': 7576537,
        'id': 'ENSP00000398846',
        'end': 7579912,
        'db_type': 'core',
        'version': 2},
       'is_canonical': 0,
       'logic_name': 'ensembl_havana_transcript',
       'seq_region_name': '17',
       'species': 'homo_sapiens',
       'Parent': 'ENSG00000141510',
       'db_type': 'core',
       'assembly_name': 'GRCh37',
       'biotype': 'protein_coding',
       'version': 2,
       'Exon': [{'end': 7590799,
         'start': 7590695,
         'id': 'ENSE00002051192',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1},
        {'start': 7579839,
         'id': 'ENSE00002667911',
         'end': 7579940,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'version': 1,
         'seq_region_name': '17',
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1},
        {'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'end': 7579721,
         'start': 7579700,
         'id': 'ENSE00002419584',
         'species': 'homo_sapiens',
         'object_type': 'Exon'},
        {'version': 1,
         'seq_region_name': '17',
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'start': 7579312,
         'id': 'ENSE00003625790',
         'end': 7579590,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'version': 1,
         'seq_region_name': '17',
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'start': 7578371,
         'id': 'ENSE00003518480',
         'end': 7578554,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'end': 7578289,
         'id': 'ENSE00003462942',
         'start': 7578177,
         'species': 'homo_sapiens',
         'object_type': 'Exon'},
        {'end': 7577608,
         'start': 7577499,
         'id': 'ENSE00003504863',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1},
        {'end': 7577155,
         'start': 7577019,
         'id': 'ENSE00003586720',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core'},
        {'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7576853,
         'id': 'ENSE00003636029',
         'end': 7576926},
        {'end': 7576584,
         'start': 7576525,
         'id': 'ENSE00003511286',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1},
        {'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7574033,
         'id': 'ENSE00003634848',
         'start': 7573927},
        {'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17',
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7571722,
         'id': 'ENSE00002034209',
         'end': 7573008}],
       'object_type': 'Transcript',
       'source': 'ensembl_havana',
       'display_name': 'TP53-004',
       'start': 7571722,
       'id': 'ENST00000455263',
       'end': 7590799},
      {'object_type': 'Transcript',
       'source': 'ensembl_havana',
       'end': 7590799,
       'start': 7571722,
       'display_name': 'TP53-005',
       'id': 'ENST00000420246',
       'assembly_name': 'GRCh37',
       'db_type': 'core',
       'biotype': 'protein_coding',
       'Exon': [{'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17',
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00002051192',
         'start': 7590695,
         'end': 7590799},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7579839,
         'id': 'ENSE00002667911',
         'end': 7579940,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1},
        {'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17',
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7579700,
         'id': 'ENSE00002419584',
         'end': 7579721},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7579312,
         'id': 'ENSE00003625790',
         'end': 7579590,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7578554,
         'start': 7578371,
         'id': 'ENSE00003518480',
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'seq_region_name': '17',
         'version': 1},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7578289,
         'id': 'ENSE00003462942',
         'start': 7578177,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'seq_region_name': '17',
         'version': 1},
        {'end': 7577608,
         'id': 'ENSE00003504863',
         'start': 7577499,
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core'},
        {'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00003586720',
         'start': 7577019,
         'end': 7577155},
        {'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'end': 7576926,
         'start': 7576853,
         'id': 'ENSE00003636029',
         'species': 'homo_sapiens',
         'object_type': 'Exon'},
        {'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'id': 'ENSE00003552110',
         'start': 7576525,
         'end': 7576657,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00003634848',
         'start': 7573927,
         'end': 7574033},
        {'start': 7571722,
         'id': 'ENSE00002034209',
         'end': 7573008,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37'}],
       'version': 2,
       'species': 'homo_sapiens',
       'Parent': 'ENSG00000141510',
       'strand': -1,
       'Translation': {'length': 341,
        'object_type': 'Translation',
        'species': 'homo_sapiens',
        'Parent': 'ENST00000420246',
        'start': 7576625,
        'id': 'ENSP00000391127',
        'end': 7579912,
        'db_type': 'core',
        'version': 2},
       'seq_region_name': '17',
       'is_canonical': 0,
       'logic_name': 'ensembl_havana_transcript'},
      {'version': 2,
       'Exon': [{'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'end': 7590805,
         'start': 7590695,
         'id': 'ENSE00002030826',
         'species': 'homo_sapiens',
         'object_type': 'Exon'},
        {'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'end': 7579937,
         'start': 7579839,
         'id': 'ENSE00001596491',
         'species': 'homo_sapiens',
         'object_type': 'Exon'},
        {'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7579700,
         'id': 'ENSE00002419584',
         'end': 7579721},
        {'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'start': 7579312,
         'id': 'ENSE00003625790',
         'end': 7579590,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7578554,
         'start': 7578371,
         'id': 'ENSE00003518480',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7578177,
         'id': 'ENSE00003462942',
         'end': 7578289,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7577608,
         'start': 7577499,
         'id': 'ENSE00003504863',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00003586720',
         'start': 7577019,
         'end': 7577155},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00003636029',
         'start': 7576853,
         'end': 7576926,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7574033,
         'start': 7573927,
         'id': 'ENSE00003545950',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7573008,
         'start': 7571739,
         'id': 'ENSE00002037735',
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'seq_region_name': '17',
         'version': 1}],
       'biotype': 'protein_coding',
       'db_type': 'core',
       'assembly_name': 'GRCh37',
       'id': 'ENST00000445888',
       'display_name': 'TP53-002',
       'start': 7571739,
       'end': 7590805,
       'source': 'havana',
       'object_type': 'Transcript',
       'is_canonical': 0,
       'logic_name': 'havana_homo_sapiens_37',
       'seq_region_name': '17',
       'Translation': {'version': 2,
        'db_type': 'core',
        'end': 7579912,
        'id': 'ENSP00000391478',
        'start': 7572927,
        'Parent': 'ENST00000445888',
        'species': 'homo_sapiens',
        'object_type': 'Translation',
        'length': 393},
       'strand': -1,
       'Parent': 'ENSG00000141510',
       'species': 'homo_sapiens'},
      {'seq_region_name': '17',
       'is_canonical': 0,
       'logic_name': 'havana_homo_sapiens_37',
       'strand': -1,
       'Translation': {'Parent': 'ENST00000576024',
        'species': 'homo_sapiens',
        'object_type': 'Translation',
        'length': 31,
        'end': 7576905,
        'start': 7572967,
        'id': 'ENSP00000458393',
        'db_type': 'core',
        'version': 1},
       'Parent': 'ENSG00000141510',
       'species': 'homo_sapiens',
       'biotype': 'protein_coding',
       'Exon': [{'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7576905,
         'start': 7576853,
         'id': 'ENSE00002672443',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'seq_region_name': '17',
         'version': 2,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'end': 7573008,
         'start': 7572887,
         'id': 'ENSE00002226620',
         'species': 'homo_sapiens',
         'object_type': 'Exon'}],
       'version': 1,
       'assembly_name': 'GRCh37',
       'db_type': 'core',
       'source': 'havana',
       'end': 7576905,
       'display_name': 'TP53-017',
       'start': 7572887,
       'id': 'ENST00000576024',
       'object_type': 'Transcript'},
      {'species': 'homo_sapiens',
       'Parent': 'ENSG00000141510',
       'is_canonical': 0,
       'logic_name': 'havana_homo_sapiens_37',
       'seq_region_name': '17',
       'Translation': {'length': 199,
        'Parent': 'ENST00000509690',
        'species': 'homo_sapiens',
        'object_type': 'Translation',
        'end': 7578533,
        'start': 7576853,
        'id': 'ENSP00000425104',
        'db_type': 'core',
        'version': 1},
       'strand': -1,
       'start': 7576853,
       'display_name': 'TP53-015',
       'id': 'ENST00000509690',
       'end': 7590805,
       'source': 'havana',
       'object_type': 'Transcript',
       'version': 1,
       'Exon': [{'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7590805,
         'id': 'ENSE00002030826',
         'start': 7590695},
        {'version': 1,
         'seq_region_name': '17',
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'id': 'ENSE00003656695',
         'start': 7578371,
         'end': 7578554,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7578289,
         'id': 'ENSE00003462942',
         'start': 7578177},
        {'end': 7577608,
         'start': 7577499,
         'id': 'ENSE00003504863',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core'},
        {'end': 7577155,
         'start': 7577019,
         'id': 'ENSE00003586720',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core'},
        {'id': 'ENSE00003669712',
         'start': 7576853,
         'end': 7576926,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37'}],
       'biotype': 'protein_coding',
       'db_type': 'core',
       'assembly_name': 'GRCh37'},
      {'biotype': 'protein_coding',
       'version': 1,
       'Exon': [{'version': 1,
         'seq_region_name': '17',
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'start': 7590695,
         'id': 'ENSE00002051873',
         'end': 7590745,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'end': 7579940,
         'id': 'ENSE00002667911',
         'start': 7579839,
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1},
        {'end': 7579721,
         'start': 7579700,
         'id': 'ENSE00002419584',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00003518480',
         'start': 7578371,
         'end': 7578554,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7578289,
         'id': 'ENSE00003462942',
         'start': 7578177,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7577535,
         'id': 'ENSE00002048269',
         'end': 7577608,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1}],
       'db_type': 'core',
       'assembly_name': 'GRCh37',
       'source': 'havana',
       'id': 'ENST00000514944',
       'display_name': 'TP53-013',
       'start': 7577535,
       'end': 7590745,
       'object_type': 'Transcript',
       'is_canonical': 0,
       'logic_name': 'havana_homo_sapiens_37',
       'seq_region_name': '17',
       'strand': -1,
       'Translation': {'species': 'homo_sapiens',
        'Parent': 'ENST00000514944',
        'object_type': 'Translation',
        'length': 155,
        'end': 7579912,
        'id': 'ENSP00000423862',
        'start': 7577535,
        'db_type': 'core',
        'version': 1},
       'Parent': 'ENSG00000141510',
       'species': 'homo_sapiens'},
      {'strand': -1,
       'logic_name': 'havana_homo_sapiens_37',
       'is_canonical': 0,
       'seq_region_name': '17',
       'Parent': 'ENSG00000141510',
       'species': 'homo_sapiens',
       'db_type': 'core',
       'assembly_name': 'GRCh37',
       'biotype': 'processed_transcript',
       'version': 1,
       'Exon': [{'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7578437,
         'id': 'ENSE00002670535',
         'start': 7578371},
        {'start': 7577572,
         'id': 'ENSE00002677565',
         'end': 7577608,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'version': 1,
         'seq_region_name': '17',
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1}],
       'object_type': 'Transcript',
       'source': 'havana',
       'display_name': 'TP53-016',
       'id': 'ENST00000574684',
       'start': 7577572,
       'end': 7578437},
      {'Parent': 'ENSG00000141510',
       'species': 'homo_sapiens',
       'strand': -1,
       'is_canonical': 0,
       'logic_name': 'havana_homo_sapiens_37',
       'seq_region_name': '17',
       'object_type': 'Transcript',
       'source': 'havana',
       'id': 'ENST00000505014',
       'display_name': 'TP53-009',
       'start': 7577844,
       'end': 7590805,
       'db_type': 'core',
       'assembly_name': 'GRCh37',
       'biotype': 'retained_intron',
       'version': 1,
       'Exon': [{'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00002030826',
         'start': 7590695,
         'end': 7590805,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'},
        {'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00002084733',
         'start': 7579700,
         'end': 7579940},
        {'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17',
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00003532881',
         'start': 7579312,
         'end': 7579590},
        {'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'end': 7578554,
         'id': 'ENSE00003520683',
         'start': 7578371,
         'species': 'homo_sapiens',
         'object_type': 'Exon'},
        {'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'end': 7578289,
         'start': 7577844,
         'id': 'ENSE00002076714',
         'species': 'homo_sapiens',
         'object_type': 'Exon'}]},
      {'display_name': 'TP53-014',
       'start': 7578434,
       'id': 'ENST00000508793',
       'end': 7580752,
       'source': 'havana',
       'object_type': 'Transcript',
       'version': 1,
       'Exon': [{'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7580643,
         'id': 'ENSE00002062958',
         'end': 7580752},
        {'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'end': 7579940,
         'id': 'ENSE00002667911',
         'start': 7579839,
         'species': 'homo_sapiens',
         'object_type': 'Exon'},
        {'start': 7579700,
         'id': 'ENSE00002419584',
         'end': 7579721,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37'},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7579590,
         'start': 7579312,
         'id': 'ENSE00003625790',
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'seq_region_name': '17',
         'version': 1},
        {'end': 7578554,
         'start': 7578434,
         'id': 'ENSE00002073243',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1}],
       'biotype': 'protein_coding',
       'db_type': 'core',
       'assembly_name': 'GRCh37',
       'Parent': 'ENSG00000141510',
       'species': 'homo_sapiens',
       'is_canonical': 0,
       'logic_name': 'havana_homo_sapiens_37',
       'seq_region_name': '17',
       'Translation': {'length': 165,
        'object_type': 'Translation',
        'species': 'homo_sapiens',
        'Parent': 'ENST00000508793',
        'id': 'ENSP00000424104',
        'start': 7578434,
        'end': 7579912,
        'db_type': 'core',
        'version': 1},
       'strand': -1},
      {'is_canonical': 0,
       'logic_name': 'havana_homo_sapiens_37',
       'seq_region_name': '17',
       'Translation': {'start': 7578480,
        'id': 'ENSP00000473895',
        'end': 7579912,
        'length': 143,
        'object_type': 'Translation',
        'species': 'homo_sapiens',
        'Parent': 'ENST00000604348',
        'version': 1,
        'db_type': 'core'},
       'strand': -1,
       'Parent': 'ENSG00000141510',
       'species': 'homo_sapiens',
       'version': 1,
       'Exon': [{'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'end': 7590805,
         'id': 'ENSE00002030826',
         'start': 7590695,
         'species': 'homo_sapiens',
         'object_type': 'Exon'},
        {'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7579839,
         'id': 'ENSE00002667911',
         'end': 7579940},
        {'start': 7579700,
         'id': 'ENSE00002419584',
         'end': 7579721,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'seq_region_name': '17',
         'version': 1,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37'},
        {'object_type': 'Exon',
         'species': 'homo_sapiens',
         'id': 'ENSE00003625790',
         'start': 7579312,
         'end': 7579590,
         'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7578533,
         'start': 7578480,
         'id': 'ENSE00003670707',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17'}],
       'biotype': 'protein_coding',
       'db_type': 'core',
       'assembly_name': 'GRCh37',
       'start': 7578480,
       'display_name': 'TP53-020',
       'id': 'ENST00000604348',
       'end': 7590805,
       'source': 'havana',
       'object_type': 'Transcript'},
      {'Parent': 'ENSG00000141510',
       'species': 'homo_sapiens',
       'seq_region_name': '17',
       'logic_name': 'havana_homo_sapiens_37',
       'is_canonical': 0,
       'strand': -1,
       'Translation': {'version': 1,
        'db_type': 'core',
        'end': 7579912,
        'id': 'ENSP00000426252',
        'start': 7578547,
        'species': 'homo_sapiens',
        'Parent': 'ENST00000503591',
        'object_type': 'Translation',
        'length': 127},
       'source': 'havana',
       'end': 7590745,
       'start': 7578547,
       'display_name': 'TP53-003',
       'id': 'ENST00000503591',
       'object_type': 'Transcript',
       'biotype': 'protein_coding',
       'Exon': [{'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'seq_region_name': '17',
         'version': 1,
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7590745,
         'id': 'ENSE00002051873',
         'start': 7590695},
        {'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7580745,
         'id': 'ENSE00002065802',
         'start': 7580643,
         'strand': -1,
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'seq_region_name': '17',
         'version': 1},
        {'version': 1,
         'seq_region_name': '17',
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'strand': -1,
         'start': 7579839,
         'id': 'ENSE00002667911',
         'end': 7579940,
         'object_type': 'Exon',
         'species': 'homo_sapiens'},
        {'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'version': 1,
         'seq_region_name': '17',
         'species': 'homo_sapiens',
         'object_type': 'Exon',
         'end': 7579721,
         'start': 7579700,
         'id': 'ENSE00002419584'},
        {'strand': -1,
         'db_type': 'core',
         'assembly_name': 'GRCh37',
         'seq_region_name': '17',
         'version': 1,
         'object_type': 'Exon',
         'species': 'homo_sapiens',
         'start': 7579312,
         'id': 'ENSE00003625790',
         'end': 7579590},
        {'version': 1,
         'seq_region_name': '17',
         'assembly_name': 'GRCh37',
         'db_type': 'core',
         'strand': -1,
         'end': 7578554,
         'start': 7578547,
         'id': 'ENSE00002362269',
         'species': 'homo_sapiens',
         'object_type': 'Exon'}],
       'version': 1,
       'assembly_name': 'GRCh37',
       'db_type': 'core'}]]



### 3.3 model(records_h, index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, start_h, end_h, alt_seq_h, ty_h = None, ver = None, ref_seq_h = '', param = 'default', search_alternative = True, max_alternative = 5, nonstop_size = 300, flank_size = 2, batch = False, show_sequence = False, align_dict = None)  
- Function  
    - Model human variants in the mouse genome.  

- Parameters   
    - records_h, index_list_h, records_m, index_list_m: human and mouse reference genome.  
    - db_h, db_m: human and mouse GENCODE annotation.  
    - tx_id_h, tx_id_m: human and mouse transcript id (could get by **h2m.get_tx_id()**).  
    - start_h, end_h: int, start and end location of the mutation on the chromosome.  
    - alt_seq_h: str, human mutation alternative sequence.  
    - ty_h: str, human variantion type. One of ['SNP', 'DNP', 'ONP', 'INS', 'DEL'].  
    - ver: int, human ref genome number. 37 or 38.  
    - ref_seq_h (optional): str, human mutation reference sequence.    
    - param (optional): set param = 'BE' and will only output base editing modelable results.  
    - search_alternative (optional): set search_alternative = False and will only output original modeling results.  
    - max_alternative (optional): the maximum number of output alternatives of one human variants.  
    - nonstop_size (optional): the length of neucleotides that are included after the stop codon for alignment and translation in case of the nonstop mutations or frame shifting mutations.  
    - flank_size (optional): the number of amino acids or neucleotides (for non-coding mutations) that are included on each side of the mutation site.  
    - batch (optional): set batch = True and will use input align_dict to save time in batch processing.  
    - show_sequence (optional): set batch = True and will output the whole sequences.  
    - align_dict (optional): input a prepared dictionary of alignment indexes to save time in batch processing.  

- output  
    - a dictionary or a list of dictionaries that include comprehensive information for the human and mouse variants.  

- Other rules  
    1. If the mutation falls in the coding and non-coding regions at the same time, it would be considered and processed as a ORIGIAL-MODELING ONLY mutation.
    3. The alt_seq input should be in the positive strand and the start_h coordinate should be smaller than or equal the end_h coordinate.
    4. If the ref-seq or alt-see has no length, it could be input as  or -.

- example  


```python
tx_id_h, tx_id_m = list_tx_id_h[3], list_tx_id_m[3]
# use the canonical transcript
model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7577120, 7577120, 'T', ty_h = 'SNP', ver = 37)
model_result
```




    {'gene_name_h': 'TP53',
     'gene_id_h': 'ENSG00000141510.11',
     'tx_id_m': 'ENSMUST00000108658.10',
     'chr_h': 'chr17',
     'exon_num_h': 10,
     'strand_h': '-',
     'start_h': 7577120,
     'end_h': 7577120,
     'ref_seq_h': 'C',
     'alt_seq_h': 'T',
     'HGVSp_h': 'R273H',
     'classification_h': 'Missense',
     'classification_h_detail': None,
     'exon_h': 'E_7',
     'type_h': 'SNP',
     'status': True,
     'error': 0,
     'statement': 'This mutation can be originally modeled.',
     'gene_name_m': 'Trp53',
     'gene_id_m': 'ENSMUSG00000059552.14',
     'chr_m': 'chr11',
     'exon_num_m': 10,
     'strand_m': '+',
     'type_m': 'SNP',
     'classification_m': 'Missense',
     'exon_m': 'E_7',
     'start_m_ori': 69480434,
     'end_m_ori': 69480434,
     'ref_seq_m_ori': 'G',
     'alt_seq_m_ori': 'A',
     'HGVSp_m_ori': 'R270H',
     'start_m': 69480434,
     'end_m': 69480434,
     'ref_seq_m': 'G',
     'alt_seq_m': 'A',
     'HGVSp_m': 'R270H'}


## License

`h2m` is distributed under the terms of the [MIT](https://spdx.org/licenses/MIT.html) license.