{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# H2M Tutorial\n",
    "Author: Kexin Dong  \n",
    "Date: Jan 4, 2024  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "H2M is a python package for precise modeling of human vairants in mice."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Quick Start"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pip install h2m"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Importing packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "import h2m\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Loading data  \n",
    "We should upload reference genome and GENCODE annotation data for both human and mouse, which could be directly downloaded from a public [dropbox](https://www.dropbox.com/scl/fo/1wtrnc9w6s9gemweuw2fv/h?rlkey=hli1z6tv096cjwit5oi6bwggg&dl=0) (**Gencode** and **Refgenome** fold respectively).  \n",
    "Both GRCh37 and GRCh38 human reference genome assemblys are available. Upload the one that you are going to use.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_h_ref, path_m_ref = '/Users/gorkordkx/Documents/Database/RefGenome/ncbi-2023-09-12/GCF_000001405.25_GRCh37.p13_genomic.fna.gz', '/Users/gorkordkx/Documents/Database/RefGenome/mouse-2023-09-13/GCF_000001635.27_GRCm39_genomic.fna.gz'\n",
    "# remember to replace the paths with yours\n",
    "records_h, index_list_h = h2m.genome_loader(path_h_ref)\n",
    "records_m, index_list_m  = h2m.genome_loader(path_m_ref)\n",
    "\n",
    "path_h_anno, path_m_anno = '/Users/gorkordkx/Documents/Database/Genecode/gencode_v19_GRCh37.db', '/Users/gorkordkx/Documents/Database/Genecode/gencode_vm33_GRCm39.db'\n",
    "# remember to replace the paths with yours\n",
    "db_h, db_m = h2m.anno_loader(path_h_anno), h2m.anno_loader(path_m_anno)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It may take up to 3 minutes to load the ref genomes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A notebook file about how to generate the db file: [1_prepare_gencode_annotation_file.ipynb]('1_prepare_gencode_annotation_file.ipynb')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Query mouse orthologous genes\n",
    "First of all, you can use H2M to query a human gene for the presence of mouse homologs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query human gene: TP53;\n",
      "Mouse ortholog(s): Trp53;\n",
      "Homology type: one2one;\n",
      "Sequence Simalarity(%):77.3537.\n"
     ]
    }
   ],
   "source": [
    "query_result = h2m.query('TP53')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output is a list of information for all the mouse ortholog(s) (if have; sometimes more than one).  \n",
    "Each element is a dictionary of **mouse gene name**, **mouse gene id**, **homology type** (one to one/one to multiple/many to many), and **similarity of human and mouse gene in percentage**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query human gene: U1;\n",
      "Mouse ortholog(s): Gm24251,Gm24291,Gm24251;\n",
      "Homology type: one2many;\n",
      "Sequence Simalarity(%):50.0, 59.6591, 50.6944.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'gene_name_m': 'Gm24251',\n",
       "  'gene_id_m': 'ENSMUSG00000087949',\n",
       "  'homology_type': 'ortholog_one2many',\n",
       "  'similarity': 50.0},\n",
       " {'gene_name_m': 'Gm24291',\n",
       "  'gene_id_m': 'ENSMUSG00000088258',\n",
       "  'homology_type': 'ortholog_one2many',\n",
       "  'similarity': 59.6591},\n",
       " {'gene_name_m': 'Gm24251',\n",
       "  'gene_id_m': 'ENSMUSG00000087949',\n",
       "  'homology_type': 'ortholog_one2many',\n",
       "  'similarity': 50.6944}]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h2m.query('U1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Class 7: The query human gene: TPT1P6 has no mouse ortholog.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[None]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h2m.query('TPT1P6')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A print output is helpful in interactive tasks like a Jupyter Notebook. You can also turn off this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'gene_name_m': 'Ubac2',\n",
       "  'gene_id_m': 'ENSMUSG00000041765',\n",
       "  'homology_type': 'ortholog_one2one',\n",
       "  'similarity': 89.2442}]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h2m.query('UBAC2', show = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Except for gene names, both ENSEMBL gene id and transcript id are accepted to identify a human gene. You can use the **ty** parameter ('tx_id','gene_id' or 'name') to specify your input type, but this is totally optional."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using gene id:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query human gene: TP53;\n",
      "Mouse ortholog(s): Trp53;\n",
      "Homology type: one2one;\n",
      "Sequence Simalarity(%):77.3537.\n"
     ]
    }
   ],
   "source": [
    "query_result = h2m.query('ENSG00000141510')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using transcript id. Should include a db annotation file with the same ref genome version."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query human gene: TP53;\n",
      "Mouse ortholog(s): Trp53;\n",
      "Homology type: one2one;\n",
      "Sequence Simalarity(%):77.3537.\n"
     ]
    }
   ],
   "source": [
    "query_result = h2m.query('ENST00000269305.4', db=db_h, ty='tx_id')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The query result of all human genes, as well as corresponding transcript IDs, is also available as [a csv file]('https://www.dropbox.com/scl/fi/o6735wok25t5dstvpz9kd/Supp_Table_1_Homo_Genes.csv?rlkey=skvxlcfv4r8ksspiq5itxxjc9&dl=0')."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get transcript ID (Internet connection needed)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One gene may have different transcripts. For mutation modeling, it is important to specify one transcript. If you do not have this information in hand, you can use H2M to get it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Again, both gene IDs and gene names are accepted as identificaitons for human and mouse genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome assembly: GRCh37;\n",
      "The canonical transcript is: ENST00000269305.4;\n",
      "You can choose from the 17 transcripts below for further analysis:\n",
      "(1)ENST00000269305.4 (2)ENST00000413465.2 (3)ENST00000359597.4 (4)ENST00000504290.1 (5)ENST00000510385.1 (6)ENST00000504937.1 (7)ENST00000455263.2 (8)ENST00000420246.2 (9)ENST00000445888.2 (10)ENST00000576024.1 (11)ENST00000509690.1 (12)ENST00000514944.1 (13)ENST00000574684.1 (14)ENST00000505014.1 (15)ENST00000508793.1 (16)ENST00000604348.1 (17)ENST00000503591.1\n",
      "\n"
     ]
    }
   ],
   "source": [
    "list_tx_id_h = h2m.get_tx_id('TP53', 'h', ver=37)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome assembly: GRCm39;\n",
      "The canonical transcript is: ENSMUST00000108658.10;\n",
      "You can choose from the 6 transcripts below for further analysis:\n",
      "(1)ENSMUST00000108658.10 (2)ENSMUST00000171247.8 (3)ENSMUST00000005371.12 (4)ENSMUST00000147512.2 (5)ENSMUST00000108657.4 (6)ENSMUST00000130540.2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "list_tx_id_m = h2m.get_tx_id('ENSMUSG00000059552', 'm')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "More information is offered except for a complete list of all transcripts."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- the chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_tx_id_h[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- the start and end location of the gene on the chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[7565097, 7590856]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_tx_id_h[1:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- the canonical transcript annotated by ENSEMBL database and used by major clinical datasets (e.g. AACR-GENIE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ENST00000269305.4'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_tx_id_h[3]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- the complete list of transcripts starting with the canonical one"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ENST00000269305.4',\n",
       " 'ENST00000413465.2',\n",
       " 'ENST00000359597.4',\n",
       " 'ENST00000504290.1',\n",
       " 'ENST00000510385.1',\n",
       " 'ENST00000504937.1',\n",
       " 'ENST00000455263.2',\n",
       " 'ENST00000420246.2',\n",
       " 'ENST00000445888.2',\n",
       " 'ENST00000576024.1',\n",
       " 'ENST00000509690.1',\n",
       " 'ENST00000514944.1',\n",
       " 'ENST00000574684.1',\n",
       " 'ENST00000505014.1',\n",
       " 'ENST00000508793.1',\n",
       " 'ENST00000604348.1',\n",
       " 'ENST00000503591.1']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list_tx_id_h[4]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output is a list of information for all the mouse ortholog(s) (if have; sometimes more than one).  \n",
    "Each element is a dictionary of **mouse gene name**, **mouse gene id**, **homology type** (one to one/one to multiple), and **similarity of human and mouse gene in percentage**.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Modeling human variants in the mouse genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now you can use H2M to model your human mutations of interest.  \n",
    "You should have at least such information in hand:  \n",
    "1. transcript id of the human gene\n",
    "2. transcript id of the mouse gene\n",
    "3. start location of human variants *on the chromosome*  \n",
    "4. end location of human variants *on the chromosome*  \n",
    "5. the reference and alternative sequence *on the positive strand of the chromosome*  \n",
    "6. mutation type (**str in ['SNP','DNP','ONP','INS','DEL'])\n",
    "7. the version number of human ref genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Usually, such information has been pretty well organized in different databases and is easy to get.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Taking *TP53* R273H (ENST00000269305.4:c.818G>A) as an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "tx_id_h, tx_id_m = list_tx_id_h[3], list_tx_id_m[3]\n",
    "# use the canonical transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>gene_id_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>chr_h</th>\n",
       "      <th>exon_num_h</th>\n",
       "      <th>strand_h</th>\n",
       "      <th>match</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>HGVSc_h</th>\n",
       "      <th>HGVSp_h</th>\n",
       "      <th>classification_h</th>\n",
       "      <th>exon_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>status</th>\n",
       "      <th>class</th>\n",
       "      <th>statement</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>gene_id_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>exon_num_m</th>\n",
       "      <th>strand_m</th>\n",
       "      <th>type_m</th>\n",
       "      <th>classification_m</th>\n",
       "      <th>exon_m</th>\n",
       "      <th>start_m_ori</th>\n",
       "      <th>end_m_ori</th>\n",
       "      <th>ref_seq_m_ori</th>\n",
       "      <th>alt_seq_m_ori</th>\n",
       "      <th>HGVSc_m_ori</th>\n",
       "      <th>HGVSp_m_ori</th>\n",
       "      <th>start_m</th>\n",
       "      <th>end_m</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>alt_seq_m</th>\n",
       "      <th>HGVSc_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4:c.817G&gt;A</td>\n",
       "      <td>R273H</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_7</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Class 0: This mutation can be originally modeled.</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_7</td>\n",
       "      <td>69480434</td>\n",
       "      <td>69480434</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000108658.10:c.808G&gt;A</td>\n",
       "      <td>R270H</td>\n",
       "      <td>69480434</td>\n",
       "      <td>69480434</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000108658.10:c.808G&gt;A</td>\n",
       "      <td>R270H</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name_h           gene_id_h            tx_id_h  chr_h  exon_num_h  \\\n",
       "0        TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "\n",
       "  strand_h  match  start_h    end_h ref_seq_h alt_seq_h  \\\n",
       "0        -   True  7577120  7577120         C         T   \n",
       "\n",
       "                      HGVSc_h HGVSp_h classification_h exon_h type_h  status  \\\n",
       "0  ENST00000269305.4:c.817G>A   R273H         Missense    E_7    SNP    True   \n",
       "\n",
       "   class                                          statement gene_name_m  \\\n",
       "0      0  Class 0: This mutation can be originally modeled.       Trp53   \n",
       "\n",
       "               gene_id_m                tx_id_m  chr_m  exon_num_m strand_m  \\\n",
       "0  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11          10        +   \n",
       "\n",
       "  type_m classification_m exon_m  start_m_ori  end_m_ori ref_seq_m_ori  \\\n",
       "0    SNP         Missense    E_7     69480434   69480434             G   \n",
       "\n",
       "  alt_seq_m_ori                     HGVSc_m_ori HGVSp_m_ori   start_m  \\\n",
       "0             A  ENSMUST00000108658.10:c.808G>A       R270H  69480434   \n",
       "\n",
       "      end_m ref_seq_m alt_seq_m                         HGVSc_m HGVSp_m  \n",
       "0  69480434         G         A  ENSMUST00000108658.10:c.808G>A   R270H  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7577120, 7577120, 'C','T', ty_h = 'SNP', ver = 37)\n",
    "pd.DataFrame(model_result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Another non-coding example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>gene_id_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>chr_h</th>\n",
       "      <th>exon_num_h</th>\n",
       "      <th>strand_h</th>\n",
       "      <th>match</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>HGVSc_h</th>\n",
       "      <th>HGVSp_h</th>\n",
       "      <th>classification_h</th>\n",
       "      <th>exon_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>status</th>\n",
       "      <th>class</th>\n",
       "      <th>statement</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>gene_id_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>exon_num_m</th>\n",
       "      <th>strand_m</th>\n",
       "      <th>type_m</th>\n",
       "      <th>classification_m</th>\n",
       "      <th>exon_m</th>\n",
       "      <th>start_m_ori</th>\n",
       "      <th>end_m_ori</th>\n",
       "      <th>ref_seq_m_ori</th>\n",
       "      <th>alt_seq_m_ori</th>\n",
       "      <th>HGVSc_m_ori</th>\n",
       "      <th>HGVSp_m_ori</th>\n",
       "      <th>start_m</th>\n",
       "      <th>end_m</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>alt_seq_m</th>\n",
       "      <th>HGVSc_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7578291</td>\n",
       "      <td>7578291</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ENST00000269305.4:c.263A&gt;C</td>\n",
       "      <td>None</td>\n",
       "      <td>Non_Coding,I</td>\n",
       "      <td>None</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>Class 2: This mutation can be modeled, but the...</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Intron</td>\n",
       "      <td>None</td>\n",
       "      <td>69479450</td>\n",
       "      <td>69479450</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSMUST00000108658.10:c.260A&gt;C</td>\n",
       "      <td>None</td>\n",
       "      <td>69479450</td>\n",
       "      <td>69479450</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSMUST00000108658.10:c.260A&gt;C</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name_h           gene_id_h            tx_id_h  chr_h  exon_num_h  \\\n",
       "0        TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "\n",
       "  strand_h  match  start_h    end_h ref_seq_h alt_seq_h  \\\n",
       "0        -   True  7578291  7578291         T         G   \n",
       "\n",
       "                      HGVSc_h HGVSp_h classification_h exon_h type_h  status  \\\n",
       "0  ENST00000269305.4:c.263A>C    None     Non_Coding,I   None    SNP    True   \n",
       "\n",
       "   class                                          statement gene_name_m  \\\n",
       "0      2  Class 2: This mutation can be modeled, but the...       Trp53   \n",
       "\n",
       "               gene_id_m                tx_id_m  chr_m  exon_num_m strand_m  \\\n",
       "0  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11          10        +   \n",
       "\n",
       "  type_m classification_m exon_m  start_m_ori  end_m_ori ref_seq_m_ori  \\\n",
       "0    SNP           Intron   None     69479450   69479450             A   \n",
       "\n",
       "  alt_seq_m_ori                     HGVSc_m_ori HGVSp_m_ori   start_m  \\\n",
       "0             C  ENSMUST00000108658.10:c.260A>C        None  69479450   \n",
       "\n",
       "      end_m ref_seq_m alt_seq_m                         HGVSc_m HGVSp_m  \n",
       "0  69479450         A         C  ENSMUST00000108658.10:c.260A>C    None  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7578291,7578291, 'T','G', ty_h = 'SNP', ver = 37)\n",
    "pd.DataFrame(model_result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that this human mutaton can be originally modeled by introducing the same neucleotide alteration."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By setting **show_sequence = True**, we can output the sequences of the wild-type and mutated human gene, wild-type, originally-modeled, and alternatively-modeled (if exsist) mouse gene. (If it can be originally modeled, the new_seq_m_alt would be the same as new_seq_m_ori)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>gene_id_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>chr_h</th>\n",
       "      <th>exon_num_h</th>\n",
       "      <th>strand_h</th>\n",
       "      <th>match</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>HGVSc_h</th>\n",
       "      <th>HGVSp_h</th>\n",
       "      <th>classification_h</th>\n",
       "      <th>exon_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>status</th>\n",
       "      <th>class</th>\n",
       "      <th>statement</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>gene_id_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>exon_num_m</th>\n",
       "      <th>strand_m</th>\n",
       "      <th>type_m</th>\n",
       "      <th>classification_m</th>\n",
       "      <th>exon_m</th>\n",
       "      <th>start_m_ori</th>\n",
       "      <th>end_m_ori</th>\n",
       "      <th>ref_seq_m_ori</th>\n",
       "      <th>alt_seq_m_ori</th>\n",
       "      <th>HGVSc_m_ori</th>\n",
       "      <th>HGVSp_m_ori</th>\n",
       "      <th>start_m</th>\n",
       "      <th>end_m</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>alt_seq_m</th>\n",
       "      <th>HGVSc_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "      <th>seq_h</th>\n",
       "      <th>seq_m</th>\n",
       "      <th>new_seq_h</th>\n",
       "      <th>new_seq_m_ori</th>\n",
       "      <th>new_seq_m_alt</th>\n",
       "      <th>human_idx</th>\n",
       "      <th>mouse_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4:c.817G&gt;A</td>\n",
       "      <td>R273H</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_7</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Class 0: This mutation can be originally modeled.</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_7</td>\n",
       "      <td>69480434</td>\n",
       "      <td>69480434</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000108658.10:c.808G&gt;A</td>\n",
       "      <td>R270H</td>\n",
       "      <td>69480434</td>\n",
       "      <td>69480434</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000108658.10:c.808G&gt;A</td>\n",
       "      <td>R270H</td>\n",
       "      <td>(A, T, G, G, A, G, G, A, G, C, C, G, C, A, G, ...</td>\n",
       "      <td>(A, T, G, A, C, T, G, C, C, A, T, G, G, A, G, ...</td>\n",
       "      <td>(A, T, G, G, A, G, G, A, G, C, C, G, C, A, G, ...</td>\n",
       "      <td>(A, T, G, A, C, T, G, C, C, A, T, G, G, A, G, ...</td>\n",
       "      <td>(A, T, G, A, C, T, G, C, C, A, T, G, G, A, G, ...</td>\n",
       "      <td>817</td>\n",
       "      <td>808</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name_h           gene_id_h            tx_id_h  chr_h  exon_num_h  \\\n",
       "0        TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "\n",
       "  strand_h  match  start_h    end_h ref_seq_h alt_seq_h  \\\n",
       "0        -   True  7577120  7577120         C         T   \n",
       "\n",
       "                      HGVSc_h HGVSp_h classification_h exon_h type_h  status  \\\n",
       "0  ENST00000269305.4:c.817G>A   R273H         Missense    E_7    SNP    True   \n",
       "\n",
       "   class                                          statement gene_name_m  \\\n",
       "0      0  Class 0: This mutation can be originally modeled.       Trp53   \n",
       "\n",
       "               gene_id_m                tx_id_m  chr_m  exon_num_m strand_m  \\\n",
       "0  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11          10        +   \n",
       "\n",
       "  type_m classification_m exon_m  start_m_ori  end_m_ori ref_seq_m_ori  \\\n",
       "0    SNP         Missense    E_7     69480434   69480434             G   \n",
       "\n",
       "  alt_seq_m_ori                     HGVSc_m_ori HGVSp_m_ori   start_m  \\\n",
       "0             A  ENSMUST00000108658.10:c.808G>A       R270H  69480434   \n",
       "\n",
       "      end_m ref_seq_m alt_seq_m                         HGVSc_m HGVSp_m  \\\n",
       "0  69480434         G         A  ENSMUST00000108658.10:c.808G>A   R270H   \n",
       "\n",
       "                                               seq_h  \\\n",
       "0  (A, T, G, G, A, G, G, A, G, C, C, G, C, A, G, ...   \n",
       "\n",
       "                                               seq_m  \\\n",
       "0  (A, T, G, A, C, T, G, C, C, A, T, G, G, A, G, ...   \n",
       "\n",
       "                                           new_seq_h  \\\n",
       "0  (A, T, G, G, A, G, G, A, G, C, C, G, C, A, G, ...   \n",
       "\n",
       "                                       new_seq_m_ori  \\\n",
       "0  (A, T, G, A, C, T, G, C, C, A, T, G, G, A, G, ...   \n",
       "\n",
       "                                       new_seq_m_alt  human_idx  mouse_idx  \n",
       "0  (A, T, G, A, C, T, G, C, C, A, T, G, G, A, G, ...        817        808  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7577120, 7577120, 'C','T', ty_h = 'SNP', ver = 37, show_sequence=True)\n",
    "pd.DataFrame(model_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_h, seq_m, new_seq_h, new_seq_m = model_result[0]['seq_h'], model_result[0]['seq_m'], model_result[0]['new_seq_h'],  model_result[0]['new_seq_m_ori']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sometimes the mutation does not locate in the homologous region. Taking *TP53* D393Y as an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>gene_id_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>chr_h</th>\n",
       "      <th>exon_num_h</th>\n",
       "      <th>strand_h</th>\n",
       "      <th>match</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>HGVSc_h</th>\n",
       "      <th>HGVSp_h</th>\n",
       "      <th>classification_h</th>\n",
       "      <th>exon_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>status</th>\n",
       "      <th>class</th>\n",
       "      <th>statement</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>gene_id_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>exon_num_m</th>\n",
       "      <th>strand_m</th>\n",
       "      <th>type_m</th>\n",
       "      <th>classification_m</th>\n",
       "      <th>exon_m</th>\n",
       "      <th>start_m_ori</th>\n",
       "      <th>end_m_ori</th>\n",
       "      <th>ref_seq_m_ori</th>\n",
       "      <th>alt_seq_m_ori</th>\n",
       "      <th>HGVSc_m_ori</th>\n",
       "      <th>HGVSp_m_ori</th>\n",
       "      <th>start_m</th>\n",
       "      <th>end_m</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>alt_seq_m</th>\n",
       "      <th>HGVSc_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7572932</td>\n",
       "      <td>7572932</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.1176G&gt;T</td>\n",
       "      <td>D393Y</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_10</td>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "      <td>Class 4: Flanked segments are not identical.</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name_h           gene_id_h            tx_id_h  chr_h  exon_num_h  \\\n",
       "0        TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "\n",
       "  strand_h  match  start_h    end_h ref_seq_h alt_seq_h  \\\n",
       "0        -   True  7572932  7572932         C         A   \n",
       "\n",
       "                       HGVSc_h HGVSp_h classification_h exon_h type_h  status  \\\n",
       "0  ENST00000269305.4:c.1176G>T   D393Y         Missense   E_10    SNP   False   \n",
       "\n",
       "   class                                     statement gene_name_m  \\\n",
       "0      4  Class 4: Flanked segments are not identical.       Trp53   \n",
       "\n",
       "               gene_id_m                tx_id_m  chr_m  exon_num_m strand_m  \\\n",
       "0  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11          10        +   \n",
       "\n",
       "  type_m classification_m exon_m start_m_ori end_m_ori ref_seq_m_ori  \\\n",
       "0   None             None   None        None      None          None   \n",
       "\n",
       "  alt_seq_m_ori HGVSc_m_ori HGVSp_m_ori start_m end_m ref_seq_m alt_seq_m  \\\n",
       "0          None        None        None    None  None      None      None   \n",
       "\n",
       "  HGVSc_m HGVSp_m  \n",
       "0    None    None  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7572932, 7572932, 'C','A', ty_h = 'SNP', ver = 37)\n",
    "pd.DataFrame(model_result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Flanked size is a parameter that controls how many amino acids (or neucleotides, for non-coding mutations) are included on both sides of the mutated region. You can change that by parameter **flank_size**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>gene_id_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>chr_h</th>\n",
       "      <th>exon_num_h</th>\n",
       "      <th>strand_h</th>\n",
       "      <th>match</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>HGVSc_h</th>\n",
       "      <th>HGVSp_h</th>\n",
       "      <th>classification_h</th>\n",
       "      <th>exon_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>status</th>\n",
       "      <th>class</th>\n",
       "      <th>statement</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>gene_id_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>exon_num_m</th>\n",
       "      <th>strand_m</th>\n",
       "      <th>type_m</th>\n",
       "      <th>classification_m</th>\n",
       "      <th>exon_m</th>\n",
       "      <th>start_m_ori</th>\n",
       "      <th>end_m_ori</th>\n",
       "      <th>ref_seq_m_ori</th>\n",
       "      <th>alt_seq_m_ori</th>\n",
       "      <th>HGVSc_m_ori</th>\n",
       "      <th>HGVSp_m_ori</th>\n",
       "      <th>start_m</th>\n",
       "      <th>end_m</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>alt_seq_m</th>\n",
       "      <th>HGVSc_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7572932</td>\n",
       "      <td>7572932</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.1176G&gt;T</td>\n",
       "      <td>D393Y</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_10</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Class 0: This mutation can be originally modeled.</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_10</td>\n",
       "      <td>69482252</td>\n",
       "      <td>69482252</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000108658.10:c.1167G&gt;T</td>\n",
       "      <td>D390Y</td>\n",
       "      <td>69482252</td>\n",
       "      <td>69482252</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000108658.10:c.1167G&gt;T</td>\n",
       "      <td>D390Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name_h           gene_id_h            tx_id_h  chr_h  exon_num_h  \\\n",
       "0        TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "\n",
       "  strand_h  match  start_h    end_h ref_seq_h alt_seq_h  \\\n",
       "0        -   True  7572932  7572932         C         A   \n",
       "\n",
       "                       HGVSc_h HGVSp_h classification_h exon_h type_h  status  \\\n",
       "0  ENST00000269305.4:c.1176G>T   D393Y         Missense   E_10    SNP    True   \n",
       "\n",
       "   class                                          statement gene_name_m  \\\n",
       "0      0  Class 0: This mutation can be originally modeled.       Trp53   \n",
       "\n",
       "               gene_id_m                tx_id_m  chr_m  exon_num_m strand_m  \\\n",
       "0  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11          10        +   \n",
       "\n",
       "  type_m classification_m exon_m  start_m_ori  end_m_ori ref_seq_m_ori  \\\n",
       "0    SNP         Missense   E_10     69482252   69482252             G   \n",
       "\n",
       "  alt_seq_m_ori                      HGVSc_m_ori HGVSp_m_ori   start_m  \\\n",
       "0             T  ENSMUST00000108658.10:c.1167G>T       D390Y  69482252   \n",
       "\n",
       "      end_m ref_seq_m alt_seq_m                          HGVSc_m HGVSp_m  \n",
       "0  69482252         G         T  ENSMUST00000108658.10:c.1167G>T   D390Y  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7572932, 7572932, 'C','A', ty_h = 'SNP', ver = 37, flank_size=1)\n",
    "pd.DataFrame(model_result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With the flank size of 1 (which means 3 amino acids are taken into consideration in totall), this mutation can be originally modeled."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Alternative modeling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sometimes the human mutation cannot be originally modeled in the mouse genome by using the same neucleotide alteration. Under this circumsatance, some alternative modeling strategies may be found by searching the codon list of the target amino acids. Taking TP53 R249_T253delinsS as an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7577523, 7577534, 'GTGAGGATGGGC', '-', ty_h = 'DEL', ver = 37)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The default maximum number of output alternatives is 5. You can definitly change that by the parameter **max_alternative**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5, 6)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_result_long = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7577523, 7577534, 'GTGAGGATGGGC', '-', ty_h = 'DEL', ver = 37, max_alternative=10)\n",
    "len(model_result), len(model_result_long)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you do not want to alternatively model variants, you can set **search_alternatve** to False."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Class 6: This mutation cannot be originally modeled and is awaiting alternative.'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7577523, 7577534, 'GTGAGGATGGGC', '-', ty_h = 'DEL', ver = 37, search_alternative= False)\n",
    "model_result[0]['statement']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Original modeling with uncertain effects"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For frame-shifting mutations and mutations in the non-coding region, we cannot find such alternative modeling strategies with the same protein change effects. H2M will only offer the original modeling and its effect."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Example 1: *TP53* C275Lfs*31"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>gene_id_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>chr_h</th>\n",
       "      <th>exon_num_h</th>\n",
       "      <th>strand_h</th>\n",
       "      <th>match</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>HGVSc_h</th>\n",
       "      <th>HGVSp_h</th>\n",
       "      <th>classification_h</th>\n",
       "      <th>exon_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>status</th>\n",
       "      <th>class</th>\n",
       "      <th>statement</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>gene_id_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>exon_num_m</th>\n",
       "      <th>strand_m</th>\n",
       "      <th>type_m</th>\n",
       "      <th>classification_m</th>\n",
       "      <th>exon_m</th>\n",
       "      <th>start_m_ori</th>\n",
       "      <th>end_m_ori</th>\n",
       "      <th>ref_seq_m_ori</th>\n",
       "      <th>alt_seq_m_ori</th>\n",
       "      <th>HGVSc_m_ori</th>\n",
       "      <th>HGVSp_m_ori</th>\n",
       "      <th>start_m</th>\n",
       "      <th>end_m</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>alt_seq_m</th>\n",
       "      <th>HGVSc_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7577115</td>\n",
       "      <td>7577116</td>\n",
       "      <td></td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.821_822&gt;T</td>\n",
       "      <td>C275Lfs*31</td>\n",
       "      <td>Frame_Shifting</td>\n",
       "      <td>E_7</td>\n",
       "      <td>INS</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>Class 2: This mutation can be modeled, but the...</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>INS</td>\n",
       "      <td>Frame_Shifting</td>\n",
       "      <td>E_7</td>\n",
       "      <td>69480438</td>\n",
       "      <td>69480439</td>\n",
       "      <td></td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000108658.10:c.812_813&gt;T</td>\n",
       "      <td>C272Lfs*24</td>\n",
       "      <td>69480438</td>\n",
       "      <td>69480439</td>\n",
       "      <td></td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000108658.10:c.812_813&gt;T</td>\n",
       "      <td>C272Lfs*24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name_h           gene_id_h            tx_id_h  chr_h  exon_num_h  \\\n",
       "0        TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "\n",
       "  strand_h  match  start_h    end_h ref_seq_h alt_seq_h  \\\n",
       "0        -   True  7577115  7577116                   A   \n",
       "\n",
       "                         HGVSc_h     HGVSp_h classification_h exon_h type_h  \\\n",
       "0  ENST00000269305.4:c.821_822>T  C275Lfs*31   Frame_Shifting    E_7    INS   \n",
       "\n",
       "   status  class                                          statement  \\\n",
       "0    True      2  Class 2: This mutation can be modeled, but the...   \n",
       "\n",
       "  gene_name_m              gene_id_m                tx_id_m  chr_m  \\\n",
       "0       Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "\n",
       "   exon_num_m strand_m type_m classification_m exon_m  start_m_ori  end_m_ori  \\\n",
       "0          10        +    INS   Frame_Shifting    E_7     69480438   69480439   \n",
       "\n",
       "  ref_seq_m_ori alt_seq_m_ori                        HGVSc_m_ori HGVSp_m_ori  \\\n",
       "0                           T  ENSMUST00000108658.10:c.812_813>T  C272Lfs*24   \n",
       "\n",
       "    start_m     end_m ref_seq_m alt_seq_m                            HGVSc_m  \\\n",
       "0  69480438  69480439                   T  ENSMUST00000108658.10:c.812_813>T   \n",
       "\n",
       "      HGVSp_m  \n",
       "0  C272Lfs*24  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7577115, 7577116, '','A', ty_h = 'INS', ver = 37)\n",
    "pd.DataFrame(model_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HGVSp_h: C275Lfs*31, HGVSp_m: C272Lfs*24\n"
     ]
    }
   ],
   "source": [
    "p_h, p_m = model_result[0]['HGVSp_h'], model_result[0]['HGVSp_m']\n",
    "print(f'HGVSp_h: {p_h}, HGVSp_m: {p_m}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Example 2: TP53 splice site mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>gene_id_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>chr_h</th>\n",
       "      <th>exon_num_h</th>\n",
       "      <th>strand_h</th>\n",
       "      <th>match</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>HGVSc_h</th>\n",
       "      <th>HGVSp_h</th>\n",
       "      <th>classification_h</th>\n",
       "      <th>exon_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>status</th>\n",
       "      <th>class</th>\n",
       "      <th>statement</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>gene_id_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>exon_num_m</th>\n",
       "      <th>strand_m</th>\n",
       "      <th>type_m</th>\n",
       "      <th>classification_m</th>\n",
       "      <th>exon_m</th>\n",
       "      <th>start_m_ori</th>\n",
       "      <th>end_m_ori</th>\n",
       "      <th>ref_seq_m_ori</th>\n",
       "      <th>alt_seq_m_ori</th>\n",
       "      <th>HGVSc_m_ori</th>\n",
       "      <th>HGVSp_m_ori</th>\n",
       "      <th>start_m</th>\n",
       "      <th>end_m</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>alt_seq_m</th>\n",
       "      <th>HGVSc_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7578555</td>\n",
       "      <td>7578555</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4:c.30G&gt;A</td>\n",
       "      <td>None</td>\n",
       "      <td>Non_Coding,I</td>\n",
       "      <td>None</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>Class 2: This mutation can be modeled, but the...</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Intron</td>\n",
       "      <td>None</td>\n",
       "      <td>69479189</td>\n",
       "      <td>69479189</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000108658.10:c.30G&gt;A</td>\n",
       "      <td>None</td>\n",
       "      <td>69479189</td>\n",
       "      <td>69479189</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000108658.10:c.30G&gt;A</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name_h           gene_id_h            tx_id_h  chr_h  exon_num_h  \\\n",
       "0        TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "\n",
       "  strand_h  match  start_h    end_h ref_seq_h alt_seq_h  \\\n",
       "0        -   True  7578555  7578555         C         T   \n",
       "\n",
       "                     HGVSc_h HGVSp_h classification_h exon_h type_h  status  \\\n",
       "0  ENST00000269305.4:c.30G>A    None     Non_Coding,I   None    SNP    True   \n",
       "\n",
       "   class                                          statement gene_name_m  \\\n",
       "0      2  Class 2: This mutation can be modeled, but the...       Trp53   \n",
       "\n",
       "               gene_id_m                tx_id_m  chr_m  exon_num_m strand_m  \\\n",
       "0  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11          10        +   \n",
       "\n",
       "  type_m classification_m exon_m  start_m_ori  end_m_ori ref_seq_m_ori  \\\n",
       "0    SNP           Intron   None     69479189   69479189             G   \n",
       "\n",
       "  alt_seq_m_ori                    HGVSc_m_ori HGVSp_m_ori   start_m  \\\n",
       "0             A  ENSMUST00000108658.10:c.30G>A        None  69479189   \n",
       "\n",
       "      end_m ref_seq_m alt_seq_m                        HGVSc_m HGVSp_m  \n",
       "0  69479189         G         A  ENSMUST00000108658.10:c.30G>A    None  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7578555, 7578555, 'C', 'T', ty_h = 'SNP', ver = 37)\n",
    "pd.DataFrame(model_result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Additional function 1: modeling for base editing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When you set **param = 'BE'**, you will get modeling results that can be modeled by base editing (A->G, G->A, C->T, T->C, AA->GG, ...etc.). If one mutation can be originally modeled in the mouse genome but not in a BE style, alternative BE modeling strategies will be returned too."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Taking *KEAP1* F221L as an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query human gene: KEAP1;\n",
      "Mouse ortholog(s): Keap1;\n",
      "Homology type: one2one;\n",
      "Sequence Simalarity(%):94.0705.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'gene_name_m': 'Keap1',\n",
       "  'gene_id_m': 'ENSMUSG00000003308',\n",
       "  'homology_type': 'ortholog_one2one',\n",
       "  'similarity': 94.0705}]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h2m.query('KEAP1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>gene_id_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>chr_h</th>\n",
       "      <th>exon_num_h</th>\n",
       "      <th>strand_h</th>\n",
       "      <th>match</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>HGVSc_h</th>\n",
       "      <th>HGVSp_h</th>\n",
       "      <th>classification_h</th>\n",
       "      <th>exon_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>status</th>\n",
       "      <th>class</th>\n",
       "      <th>statement</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>gene_id_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>exon_num_m</th>\n",
       "      <th>strand_m</th>\n",
       "      <th>type_m</th>\n",
       "      <th>classification_m</th>\n",
       "      <th>exon_m</th>\n",
       "      <th>start_m_ori</th>\n",
       "      <th>end_m_ori</th>\n",
       "      <th>ref_seq_m_ori</th>\n",
       "      <th>alt_seq_m_ori</th>\n",
       "      <th>HGVSc_m_ori</th>\n",
       "      <th>HGVSp_m_ori</th>\n",
       "      <th>start_m</th>\n",
       "      <th>end_m</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>alt_seq_m</th>\n",
       "      <th>HGVSc_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KEAP1</td>\n",
       "      <td>ENSG00000079999.9</td>\n",
       "      <td>ENST00000171111.5</td>\n",
       "      <td>chr19</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>10602915</td>\n",
       "      <td>10602915</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000171111.5:c.662C&gt;A</td>\n",
       "      <td>F221L</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_2</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>Class 1: This mutation can be alternatively mo...</td>\n",
       "      <td>Keap1</td>\n",
       "      <td>ENSMUSG00000003308.16</td>\n",
       "      <td>ENSMUST00000164812.8</td>\n",
       "      <td>chr9</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_2</td>\n",
       "      <td>21145346</td>\n",
       "      <td>21145346</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000164812.8:c.662C&gt;A</td>\n",
       "      <td>F221L</td>\n",
       "      <td>21145348</td>\n",
       "      <td>21145348</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSMUST00000164812.8:c.660T&gt;C</td>\n",
       "      <td>F221L</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name_h          gene_id_h            tx_id_h  chr_h  exon_num_h  \\\n",
       "0       KEAP1  ENSG00000079999.9  ENST00000171111.5  chr19           5   \n",
       "\n",
       "  strand_h  match   start_h     end_h ref_seq_h alt_seq_h  \\\n",
       "0        -   True  10602915  10602915         G         T   \n",
       "\n",
       "                      HGVSc_h HGVSp_h classification_h exon_h type_h  status  \\\n",
       "0  ENST00000171111.5:c.662C>A   F221L         Missense    E_2    SNP    True   \n",
       "\n",
       "   class                                          statement gene_name_m  \\\n",
       "0      1  Class 1: This mutation can be alternatively mo...       Keap1   \n",
       "\n",
       "               gene_id_m               tx_id_m chr_m  exon_num_m strand_m  \\\n",
       "0  ENSMUSG00000003308.16  ENSMUST00000164812.8  chr9           5        -   \n",
       "\n",
       "  type_m classification_m exon_m  start_m_ori  end_m_ori ref_seq_m_ori  \\\n",
       "0    SNP         Missense    E_2     21145346   21145346             G   \n",
       "\n",
       "  alt_seq_m_ori                    HGVSc_m_ori HGVSp_m_ori   start_m  \\\n",
       "0             T  ENSMUST00000164812.8:c.662C>A       F221L  21145348   \n",
       "\n",
       "      end_m ref_seq_m alt_seq_m                        HGVSc_m HGVSp_m  \n",
       "0  21145348         A         G  ENSMUST00000164812.8:c.660T>C   F221L  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tx_id_h_2, tx_id_m_2 = h2m.get_tx_id('KEAP1','h',ver=37, show=False)[3], h2m.get_tx_id('Keap1','m', show=False)[3]\n",
    "model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h_2, tx_id_m_2, 10602915, 10602915, 'G','T', ty_h = 'SNP', ver = 37, param='BE')\n",
    "pd.DataFrame(model_result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Additional function 2: modeling by amino acid change input"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set **coor = 'aa'** and modeling human variants by amino acid change input. Use TP53 R175H as an example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>gene_id_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>chr_h</th>\n",
       "      <th>exon_num_h</th>\n",
       "      <th>strand_h</th>\n",
       "      <th>match</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>HGVSc_h</th>\n",
       "      <th>HGVSp_h</th>\n",
       "      <th>classification_h</th>\n",
       "      <th>exon_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>status</th>\n",
       "      <th>class</th>\n",
       "      <th>statement</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>gene_id_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>exon_num_m</th>\n",
       "      <th>strand_m</th>\n",
       "      <th>type_m</th>\n",
       "      <th>classification_m</th>\n",
       "      <th>exon_m</th>\n",
       "      <th>start_m_ori</th>\n",
       "      <th>end_m_ori</th>\n",
       "      <th>ref_seq_m_ori</th>\n",
       "      <th>alt_seq_m_ori</th>\n",
       "      <th>HGVSc_m_ori</th>\n",
       "      <th>HGVSp_m_ori</th>\n",
       "      <th>start_m</th>\n",
       "      <th>end_m</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>alt_seq_m</th>\n",
       "      <th>HGVSc_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7578405</td>\n",
       "      <td>7578407</td>\n",
       "      <td>GCG</td>\n",
       "      <td>GTG</td>\n",
       "      <td>ENST00000269305.4:c.522_524CGC&gt;CAC</td>\n",
       "      <td>R175H</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_4</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>Class 1: This mutation can be alternatively mo...</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_4</td>\n",
       "      <td>69479337</td>\n",
       "      <td>69479339</td>\n",
       "      <td>CGC</td>\n",
       "      <td>CAC</td>\n",
       "      <td>ENSMUST00000108658.10:c.513_515CGC&gt;CAC</td>\n",
       "      <td>R172H</td>\n",
       "      <td>69479338</td>\n",
       "      <td>69479338</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000108658.10:c.514G&gt;A</td>\n",
       "      <td>R172H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7578405</td>\n",
       "      <td>7578407</td>\n",
       "      <td>GCG</td>\n",
       "      <td>GTG</td>\n",
       "      <td>ENST00000269305.4:c.522_524CGC&gt;CAC</td>\n",
       "      <td>R175H</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_4</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>Class 1: This mutation can be alternatively mo...</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>DNP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_4</td>\n",
       "      <td>69479337</td>\n",
       "      <td>69479339</td>\n",
       "      <td>CGC</td>\n",
       "      <td>CAC</td>\n",
       "      <td>ENSMUST00000108658.10:c.513_515CGC&gt;CAC</td>\n",
       "      <td>R172H</td>\n",
       "      <td>69479338</td>\n",
       "      <td>69479339</td>\n",
       "      <td>GC</td>\n",
       "      <td>AT</td>\n",
       "      <td>ENSMUST00000108658.10:c.514_515GC&gt;AT</td>\n",
       "      <td>R172H</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_name_h           gene_id_h            tx_id_h  chr_h  exon_num_h  \\\n",
       "0        TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "1        TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "\n",
       "  strand_h  match  start_h    end_h ref_seq_h alt_seq_h  \\\n",
       "0        -   True  7578405  7578407       GCG       GTG   \n",
       "1        -   True  7578405  7578407       GCG       GTG   \n",
       "\n",
       "                              HGVSc_h HGVSp_h classification_h exon_h type_h  \\\n",
       "0  ENST00000269305.4:c.522_524CGC>CAC   R175H         Missense    E_4    SNP   \n",
       "1  ENST00000269305.4:c.522_524CGC>CAC   R175H         Missense    E_4    SNP   \n",
       "\n",
       "   status  class                                          statement  \\\n",
       "0    True      1  Class 1: This mutation can be alternatively mo...   \n",
       "1    True      1  Class 1: This mutation can be alternatively mo...   \n",
       "\n",
       "  gene_name_m              gene_id_m                tx_id_m  chr_m  \\\n",
       "0       Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "1       Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "\n",
       "   exon_num_m strand_m type_m classification_m exon_m  start_m_ori  end_m_ori  \\\n",
       "0          10        +    SNP         Missense    E_4     69479337   69479339   \n",
       "1          10        +    DNP         Missense    E_4     69479337   69479339   \n",
       "\n",
       "  ref_seq_m_ori alt_seq_m_ori                             HGVSc_m_ori  \\\n",
       "0           CGC           CAC  ENSMUST00000108658.10:c.513_515CGC>CAC   \n",
       "1           CGC           CAC  ENSMUST00000108658.10:c.513_515CGC>CAC   \n",
       "\n",
       "  HGVSp_m_ori   start_m     end_m ref_seq_m alt_seq_m  \\\n",
       "0       R172H  69479338  69479338         G         A   \n",
       "1       R172H  69479338  69479339        GC        AT   \n",
       "\n",
       "                                HGVSc_m HGVSp_m  \n",
       "0        ENSMUST00000108658.10:c.514G>A   R172H  \n",
       "1  ENSMUST00000108658.10:c.514_515GC>AT   R172H  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 175, 175, 'R', 'H', coor = 'aa', ty_h = 'SNP', ver = 37)\n",
    "pd.DataFrame(model_result)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Batch Processing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In batch processing, you need to build a pandas dataframe with columns as the following example:  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>class_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7579311</td>\n",
       "      <td>7579311</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577567</td>\n",
       "      <td>7577567</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>25398281</td>\n",
       "      <td>25398281</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7578392</td>\n",
       "      <td>7578392</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7578406</td>\n",
       "      <td>7578406</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577141</td>\n",
       "      <td>7577141</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579415</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577096</td>\n",
       "      <td>7577096</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577094</td>\n",
       "      <td>7577094</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>55249022</td>\n",
       "      <td>55249022</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7578235</td>\n",
       "      <td>7578235</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577539</td>\n",
       "      <td>7577539</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7578268</td>\n",
       "      <td>7578268</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7578536</td>\n",
       "      <td>7578536</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577068</td>\n",
       "      <td>7577068</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index gene_name_h            tx_id_h   start_h     end_h  \\\n",
       "0       0        TP53  ENST00000269305.4   7579311   7579311   \n",
       "1       1        KRAS  ENST00000256078.4  25398284  25398284   \n",
       "2       2        TP53  ENST00000269305.4   7577567   7577567   \n",
       "3       3        KRAS  ENST00000256078.4  25398284  25398284   \n",
       "4       4        KRAS  ENST00000256078.4  25398281  25398281   \n",
       "5       5        TP53  ENST00000269305.4   7578392   7578392   \n",
       "6       6        TP53  ENST00000269305.4   7578406   7578406   \n",
       "7       7        TP53  ENST00000269305.4   7577141   7577141   \n",
       "8       8        TP53  ENST00000269305.4   7579415   7579415   \n",
       "9       9        TP53  ENST00000269305.4   7577096   7577096   \n",
       "10     10        TP53  ENST00000269305.4   7577094   7577094   \n",
       "11     11        EGFR  ENST00000275493.2  55249022  55249022   \n",
       "12     12        TP53  ENST00000269305.4   7578235   7578235   \n",
       "13     13        TP53  ENST00000269305.4   7577539   7577539   \n",
       "14     14        TP53  ENST00000269305.4   7578268   7578268   \n",
       "15     15        TP53  ENST00000269305.4   7578536   7578536   \n",
       "16     16        TP53  ENST00000269305.4   7577068   7577068   \n",
       "\n",
       "              class_h type_h ref_seq_h alt_seq_h  \n",
       "0         Splice_Site    SNP         C         A  \n",
       "1   Missense_Mutation    SNP         C         A  \n",
       "2   Nonsense_Mutation    SNP         A         T  \n",
       "3   Missense_Mutation    SNP         C         T  \n",
       "4   Missense_Mutation    SNP         C         T  \n",
       "5   Missense_Mutation    SNP         C         T  \n",
       "6   Missense_Mutation    SNP         C         T  \n",
       "7   Missense_Mutation    SNP         C         A  \n",
       "8   Nonsense_Mutation    SNP         C         T  \n",
       "9   Missense_Mutation    SNP         T         C  \n",
       "10  Missense_Mutation    SNP         G         A  \n",
       "11  Missense_Mutation    SNP         G         A  \n",
       "12  Missense_Mutation    SNP         T         C  \n",
       "13  Missense_Mutation    SNP         G         A  \n",
       "14  Missense_Mutation    SNP         A         C  \n",
       "15  Missense_Mutation    SNP         T         G  \n",
       "16    Frame_Shift_Del    DEL         G         -  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sample = pd.read_csv('input/sample_data.csv')\n",
    "df_sample"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This can be achieved simply by using h2m built-in functions. The sample data used in this section can be found in the [dropbox link]('https://www.dropbox.com/scl/fo/369r665lxot7y66cdsft4/h?rlkey=ofy8hlr1dqq0t0qwp5qhenrsf&dl=0')."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Read from cBioPortal  \n",
    "In this format, users simply provide a set of variants with their genome coordinates and reference and alternate alleles, as well as the variant type. This format is compatible with all of the datasets in the cBioPortal, as well as AACR-GENIE. Download the txt mutation data file from such public dataset and then load it as follows:  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>25398285</td>\n",
       "      <td>25398285</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>ENST00000288602.6</td>\n",
       "      <td>140453136</td>\n",
       "      <td>140453136</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>55249071</td>\n",
       "      <td>55249071</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7577120</td>\n",
       "      <td>7577120</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>NRAS</td>\n",
       "      <td>ENST00000369535.4</td>\n",
       "      <td>115256529</td>\n",
       "      <td>115256529</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>728842</th>\n",
       "      <td>728842</td>\n",
       "      <td>BCORL1</td>\n",
       "      <td>ENST00000218147.7</td>\n",
       "      <td>129189861</td>\n",
       "      <td>129189861</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>728843</th>\n",
       "      <td>728843</td>\n",
       "      <td>BCORL1</td>\n",
       "      <td>ENST00000218147.7</td>\n",
       "      <td>129189992</td>\n",
       "      <td>129189992</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>728844</th>\n",
       "      <td>728844</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>ENST00000332070.3</td>\n",
       "      <td>133547943</td>\n",
       "      <td>133547943</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>728845</th>\n",
       "      <td>728845</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>ENST00000332070.3</td>\n",
       "      <td>133547964</td>\n",
       "      <td>133547964</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>728846</th>\n",
       "      <td>728846</td>\n",
       "      <td>PHF6</td>\n",
       "      <td>ENST00000332070.3</td>\n",
       "      <td>133551330</td>\n",
       "      <td>133551330</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>728847 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         index gene_name_h            tx_id_h    start_h      end_h type_h  \\\n",
       "0            0        KRAS  ENST00000256078.4   25398285   25398285    SNP   \n",
       "1            1        BRAF  ENST00000288602.6  140453136  140453136    SNP   \n",
       "2            2        EGFR  ENST00000275493.2   55249071   55249071    SNP   \n",
       "3            3        TP53  ENST00000269305.4    7577120    7577120    SNP   \n",
       "4            4        NRAS  ENST00000369535.4  115256529  115256529    SNP   \n",
       "...        ...         ...                ...        ...        ...    ...   \n",
       "728842  728842      BCORL1  ENST00000218147.7  129189861  129189861    SNP   \n",
       "728843  728843      BCORL1  ENST00000218147.7  129189992  129189992    SNP   \n",
       "728844  728844        PHF6  ENST00000332070.3  133547943  133547943    SNP   \n",
       "728845  728845        PHF6  ENST00000332070.3  133547964  133547964    SNP   \n",
       "728846  728846        PHF6  ENST00000332070.3  133551330  133551330    SNP   \n",
       "\n",
       "       ref_seq_h alt_seq_h  \n",
       "0              C         A  \n",
       "1              A         T  \n",
       "2              C         T  \n",
       "3              C         T  \n",
       "4              T         C  \n",
       "...          ...       ...  \n",
       "728842         T         C  \n",
       "728843         G         T  \n",
       "728844         G         A  \n",
       "728845         G         T  \n",
       "728846         C         T  \n",
       "\n",
       "[728847 rows x 8 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_aacr = '/Users/gorkordkx/Desktop/Flab - Drylab/Database/AACR-GENIE/v13.1/data_mutations_extended.txt'\n",
    "df = h2m.cbio_reader(path_aacr)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Read from ClinVar\n",
    "download a ClinVar vcf.gz file, and choose your desired Variation IDs that you wish to model. These vcf.gz files are available in the Reference Files Dropbox Link (under the clinvar folder) or the more up to date files can be accessed here: https://ftp.ncbi.nlm.nih.gov/pub/clinvar/   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig '1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '2' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '3' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '4' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '5' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '6' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '7' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '8' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '9' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '10' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '11' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '12' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '13' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '14' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '15' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '16' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '17' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '18' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '19' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '20' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '21' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '22' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'X' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'Y' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'MT' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_113889.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_167222.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NW_003315925.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NW_003315947.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NW_003315950.2' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>DVL1</td>\n",
       "      <td>1273425</td>\n",
       "      <td>1273426</td>\n",
       "      <td>Indel</td>\n",
       "      <td>AA</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>MSH6</td>\n",
       "      <td>48010028</td>\n",
       "      <td>48010028</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>RRM2B</td>\n",
       "      <td>103218209</td>\n",
       "      <td>103218209</td>\n",
       "      <td>INS</td>\n",
       "      <td>A</td>\n",
       "      <td>AT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>RRM2B</td>\n",
       "      <td>103231051</td>\n",
       "      <td>103231052</td>\n",
       "      <td>Indel</td>\n",
       "      <td>GC</td>\n",
       "      <td>AT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>25358662</td>\n",
       "      <td>25358663</td>\n",
       "      <td>DEL</td>\n",
       "      <td>CT</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7571192</td>\n",
       "      <td>7571192</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7571198</td>\n",
       "      <td>7571198</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7571206</td>\n",
       "      <td>7571206</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7571224</td>\n",
       "      <td>7571224</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7572147</td>\n",
       "      <td>7572148</td>\n",
       "      <td>DEL</td>\n",
       "      <td>AG</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index gene_name_h    start_h      end_h type_h ref_seq_h alt_seq_h\n",
       "0      0        DVL1    1273425    1273426  Indel        AA         G\n",
       "1      1        MSH6   48010028   48010028    SNP         G         T\n",
       "2      2       RRM2B  103218209  103218209    INS         A        AT\n",
       "3      3       RRM2B  103231051  103231052  Indel        GC        AT\n",
       "4      4        KRAS   25358662   25358663    DEL        CT         C\n",
       "5      5        TP53    7571192    7571192    SNP         G         C\n",
       "6      6        TP53    7571198    7571198    SNP         G         A\n",
       "7      7        TP53    7571206    7571206    SNP         G         A\n",
       "8      8        TP53    7571224    7571224    SNP         C         T\n",
       "9      9        TP53    7572147    7572148    DEL        AG         A"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filepath = '/Users/gorkordkx/Desktop/GRCh37_clinvar_20240206.vcf.gz'\n",
    "variation_ids = [925574, 925434, 926695, 925707, 325626, 1191613, 308061, 361149, 1205375, 208043]\n",
    "df = h2m.clinvar_reader(filepath, variation_ids)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Read from GenomAD  \n",
    "Search a specific gene in GenomAD browser, and download the conlusion csv.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>type_h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7661882</td>\n",
       "      <td>7661882</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7661888</td>\n",
       "      <td>7661888</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7661899</td>\n",
       "      <td>7661899</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7661904</td>\n",
       "      <td>7661904</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7661905</td>\n",
       "      <td>7661905</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2036</th>\n",
       "      <td>2036</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7676665</td>\n",
       "      <td>7676665</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2037</th>\n",
       "      <td>2037</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7676667</td>\n",
       "      <td>7676667</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2038</th>\n",
       "      <td>2038</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7676667</td>\n",
       "      <td>7676667</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2039</th>\n",
       "      <td>2039</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7676668</td>\n",
       "      <td>7676668</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2040</th>\n",
       "      <td>2040</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7676669</td>\n",
       "      <td>7676669</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2041 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      index gene_name_h  start_h    end_h ref_seq_h alt_seq_h type_h\n",
       "0         0        TP53  7661882  7661882         C         T    SNP\n",
       "1         1        TP53  7661888  7661888         T         A    SNP\n",
       "2         2        TP53  7661899  7661899         A         G    SNP\n",
       "3         3        TP53  7661904  7661904         C         T    SNP\n",
       "4         4        TP53  7661905  7661905         G         A    SNP\n",
       "...     ...         ...      ...      ...       ...       ...    ...\n",
       "2036   2036        TP53  7676665  7676665         C         T    SNP\n",
       "2037   2037        TP53  7676667  7676667         C         A    SNP\n",
       "2038   2038        TP53  7676667  7676667         C         T    SNP\n",
       "2039   2039        TP53  7676668  7676668         T         G    SNP\n",
       "2040   2040        TP53  7676669  7676669         G         T    SNP\n",
       "\n",
       "[2041 rows x 7 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = h2m.genomad_reader('/Users/gorkordkx/Downloads/gnomAD_v4.0.0_ENSG00000141510_2024_02_07_11_36_03.csv','TP53')\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get canonical transcript ID of human genes if needed, no internet connection needed"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There will be returning two dataframes for success and failures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>class_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7579311</td>\n",
       "      <td>7579311</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577567</td>\n",
       "      <td>7577567</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>25398281</td>\n",
       "      <td>25398281</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578392</td>\n",
       "      <td>7578392</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578406</td>\n",
       "      <td>7578406</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577141</td>\n",
       "      <td>7577141</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579415</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577096</td>\n",
       "      <td>7577096</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577094</td>\n",
       "      <td>7577094</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>55249022</td>\n",
       "      <td>55249022</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578235</td>\n",
       "      <td>7578235</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577539</td>\n",
       "      <td>7577539</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578268</td>\n",
       "      <td>7578268</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578536</td>\n",
       "      <td>7578536</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577068</td>\n",
       "      <td>7577068</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index gene_name_h   start_h     end_h            class_h type_h ref_seq_h  \\\n",
       "0       0        TP53   7579311   7579311        Splice_Site    SNP         C   \n",
       "1       1        KRAS  25398284  25398284  Missense_Mutation    SNP         C   \n",
       "2       2        TP53   7577567   7577567  Nonsense_Mutation    SNP         A   \n",
       "3       3        KRAS  25398284  25398284  Missense_Mutation    SNP         C   \n",
       "4       4        KRAS  25398281  25398281  Missense_Mutation    SNP         C   \n",
       "5       5        TP53   7578392   7578392  Missense_Mutation    SNP         C   \n",
       "6       6        TP53   7578406   7578406  Missense_Mutation    SNP         C   \n",
       "7       7        TP53   7577141   7577141  Missense_Mutation    SNP         C   \n",
       "8       8        TP53   7579415   7579415  Nonsense_Mutation    SNP         C   \n",
       "9       9        TP53   7577096   7577096  Missense_Mutation    SNP         T   \n",
       "10     10        TP53   7577094   7577094  Missense_Mutation    SNP         G   \n",
       "11     11        EGFR  55249022  55249022  Missense_Mutation    SNP         G   \n",
       "12     12        TP53   7578235   7578235  Missense_Mutation    SNP         T   \n",
       "13     13        TP53   7577539   7577539  Missense_Mutation    SNP         G   \n",
       "14     14        TP53   7578268   7578268  Missense_Mutation    SNP         A   \n",
       "15     15        TP53   7578536   7578536  Missense_Mutation    SNP         T   \n",
       "16     16        TP53   7577068   7577068    Frame_Shift_Del    DEL         G   \n",
       "\n",
       "   alt_seq_h  \n",
       "0          A  \n",
       "1          A  \n",
       "2          T  \n",
       "3          T  \n",
       "4          T  \n",
       "5          T  \n",
       "6          T  \n",
       "7          A  \n",
       "8          T  \n",
       "9          C  \n",
       "10         A  \n",
       "11         A  \n",
       "12         C  \n",
       "13         A  \n",
       "14         C  \n",
       "15         G  \n",
       "16         -  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df_sample.drop('tx_id_h', axis = 1)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No error occurs.\n"
     ]
    }
   ],
   "source": [
    "df = h2m.get_tx_batch(df,species='h',ver=37)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>class_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>ref_genome_h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7579311</td>\n",
       "      <td>7579311</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577567</td>\n",
       "      <td>7577567</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>25398281</td>\n",
       "      <td>25398281</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578392</td>\n",
       "      <td>7578392</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578406</td>\n",
       "      <td>7578406</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577141</td>\n",
       "      <td>7577141</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579415</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577096</td>\n",
       "      <td>7577096</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577094</td>\n",
       "      <td>7577094</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>55249022</td>\n",
       "      <td>55249022</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578235</td>\n",
       "      <td>7578235</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577539</td>\n",
       "      <td>7577539</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578268</td>\n",
       "      <td>7578268</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578536</td>\n",
       "      <td>7578536</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577068</td>\n",
       "      <td>7577068</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index gene_name_h   start_h     end_h            class_h type_h ref_seq_h  \\\n",
       "0       0        TP53   7579311   7579311        Splice_Site    SNP         C   \n",
       "1       1        KRAS  25398284  25398284  Missense_Mutation    SNP         C   \n",
       "2       2        TP53   7577567   7577567  Nonsense_Mutation    SNP         A   \n",
       "3       3        KRAS  25398284  25398284  Missense_Mutation    SNP         C   \n",
       "4       4        KRAS  25398281  25398281  Missense_Mutation    SNP         C   \n",
       "5       5        TP53   7578392   7578392  Missense_Mutation    SNP         C   \n",
       "6       6        TP53   7578406   7578406  Missense_Mutation    SNP         C   \n",
       "7       7        TP53   7577141   7577141  Missense_Mutation    SNP         C   \n",
       "8       8        TP53   7579415   7579415  Nonsense_Mutation    SNP         C   \n",
       "9       9        TP53   7577096   7577096  Missense_Mutation    SNP         T   \n",
       "10     10        TP53   7577094   7577094  Missense_Mutation    SNP         G   \n",
       "11     11        EGFR  55249022  55249022  Missense_Mutation    SNP         G   \n",
       "12     12        TP53   7578235   7578235  Missense_Mutation    SNP         T   \n",
       "13     13        TP53   7577539   7577539  Missense_Mutation    SNP         G   \n",
       "14     14        TP53   7578268   7578268  Missense_Mutation    SNP         A   \n",
       "15     15        TP53   7578536   7578536  Missense_Mutation    SNP         T   \n",
       "16     16        TP53   7577068   7577068    Frame_Shift_Del    DEL         G   \n",
       "\n",
       "   alt_seq_h            tx_id_h ref_genome_h  \n",
       "0          A  ENST00000269305.4       GRCh37  \n",
       "1          A  ENST00000256078.4       GRCh37  \n",
       "2          T  ENST00000269305.4       GRCh37  \n",
       "3          T  ENST00000256078.4       GRCh37  \n",
       "4          T  ENST00000256078.4       GRCh37  \n",
       "5          T  ENST00000269305.4       GRCh37  \n",
       "6          T  ENST00000269305.4       GRCh37  \n",
       "7          A  ENST00000269305.4       GRCh37  \n",
       "8          T  ENST00000269305.4       GRCh37  \n",
       "9          C  ENST00000269305.4       GRCh37  \n",
       "10         A  ENST00000269305.4       GRCh37  \n",
       "11         A  ENST00000275493.2       GRCh37  \n",
       "12         C  ENST00000269305.4       GRCh37  \n",
       "13         A  ENST00000269305.4       GRCh37  \n",
       "14         C  ENST00000269305.4       GRCh37  \n",
       "15         G  ENST00000269305.4       GRCh37  \n",
       "16         -  ENST00000269305.4       GRCh37  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Query mouse orthologous genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No error occurs.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>class_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>ref_genome_h</th>\n",
       "      <th>gene_name_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>55249022</td>\n",
       "      <td>55249022</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Egfr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Kras</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Kras</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>25398281</td>\n",
       "      <td>25398281</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Kras</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7579311</td>\n",
       "      <td>7579311</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577567</td>\n",
       "      <td>7577567</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578392</td>\n",
       "      <td>7578392</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>6</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578406</td>\n",
       "      <td>7578406</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577141</td>\n",
       "      <td>7577141</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>8</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579415</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577096</td>\n",
       "      <td>7577096</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>10</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577094</td>\n",
       "      <td>7577094</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578235</td>\n",
       "      <td>7578235</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577539</td>\n",
       "      <td>7577539</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578268</td>\n",
       "      <td>7578268</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578536</td>\n",
       "      <td>7578536</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577068</td>\n",
       "      <td>7577068</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index gene_name_h   start_h     end_h            class_h type_h ref_seq_h  \\\n",
       "0      11        EGFR  55249022  55249022  Missense_Mutation    SNP         G   \n",
       "1       1        KRAS  25398284  25398284  Missense_Mutation    SNP         C   \n",
       "2       3        KRAS  25398284  25398284  Missense_Mutation    SNP         C   \n",
       "3       4        KRAS  25398281  25398281  Missense_Mutation    SNP         C   \n",
       "4       0        TP53   7579311   7579311        Splice_Site    SNP         C   \n",
       "5       2        TP53   7577567   7577567  Nonsense_Mutation    SNP         A   \n",
       "6       5        TP53   7578392   7578392  Missense_Mutation    SNP         C   \n",
       "7       6        TP53   7578406   7578406  Missense_Mutation    SNP         C   \n",
       "8       7        TP53   7577141   7577141  Missense_Mutation    SNP         C   \n",
       "9       8        TP53   7579415   7579415  Nonsense_Mutation    SNP         C   \n",
       "10      9        TP53   7577096   7577096  Missense_Mutation    SNP         T   \n",
       "11     10        TP53   7577094   7577094  Missense_Mutation    SNP         G   \n",
       "12     12        TP53   7578235   7578235  Missense_Mutation    SNP         T   \n",
       "13     13        TP53   7577539   7577539  Missense_Mutation    SNP         G   \n",
       "14     14        TP53   7578268   7578268  Missense_Mutation    SNP         A   \n",
       "15     15        TP53   7578536   7578536  Missense_Mutation    SNP         T   \n",
       "16     16        TP53   7577068   7577068    Frame_Shift_Del    DEL         G   \n",
       "\n",
       "   alt_seq_h            tx_id_h ref_genome_h gene_name_m  \n",
       "0          A  ENST00000275493.2       GRCh37        Egfr  \n",
       "1          A  ENST00000256078.4       GRCh37        Kras  \n",
       "2          T  ENST00000256078.4       GRCh37        Kras  \n",
       "3          T  ENST00000256078.4       GRCh37        Kras  \n",
       "4          A  ENST00000269305.4       GRCh37       Trp53  \n",
       "5          T  ENST00000269305.4       GRCh37       Trp53  \n",
       "6          T  ENST00000269305.4       GRCh37       Trp53  \n",
       "7          T  ENST00000269305.4       GRCh37       Trp53  \n",
       "8          A  ENST00000269305.4       GRCh37       Trp53  \n",
       "9          T  ENST00000269305.4       GRCh37       Trp53  \n",
       "10         C  ENST00000269305.4       GRCh37       Trp53  \n",
       "11         A  ENST00000269305.4       GRCh37       Trp53  \n",
       "12         C  ENST00000269305.4       GRCh37       Trp53  \n",
       "13         A  ENST00000269305.4       GRCh37       Trp53  \n",
       "14         C  ENST00000269305.4       GRCh37       Trp53  \n",
       "15         G  ENST00000269305.4       GRCh37       Trp53  \n",
       "16         -  ENST00000269305.4       GRCh37       Trp53  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = h2m.query_batch(df)[0]\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get canonical transcript ID of mouse genes, no internet connection needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No error occurs.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>class_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>ref_genome_h</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>11</td>\n",
       "      <td>EGFR</td>\n",
       "      <td>55249022</td>\n",
       "      <td>55249022</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Egfr</td>\n",
       "      <td>ENSMUST00000020329.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>25398281</td>\n",
       "      <td>25398281</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7579311</td>\n",
       "      <td>7579311</td>\n",
       "      <td>Splice_Site</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577567</td>\n",
       "      <td>7577567</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>5</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578392</td>\n",
       "      <td>7578392</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>6</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578406</td>\n",
       "      <td>7578406</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577141</td>\n",
       "      <td>7577141</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>8</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579415</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577096</td>\n",
       "      <td>7577096</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>10</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577094</td>\n",
       "      <td>7577094</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578235</td>\n",
       "      <td>7578235</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577539</td>\n",
       "      <td>7577539</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578268</td>\n",
       "      <td>7578268</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7578536</td>\n",
       "      <td>7578536</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>TP53</td>\n",
       "      <td>7577068</td>\n",
       "      <td>7577068</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>DEL</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index gene_name_h   start_h     end_h            class_h type_h ref_seq_h  \\\n",
       "0      11        EGFR  55249022  55249022  Missense_Mutation    SNP         G   \n",
       "1       1        KRAS  25398284  25398284  Missense_Mutation    SNP         C   \n",
       "2       3        KRAS  25398284  25398284  Missense_Mutation    SNP         C   \n",
       "3       4        KRAS  25398281  25398281  Missense_Mutation    SNP         C   \n",
       "4       0        TP53   7579311   7579311        Splice_Site    SNP         C   \n",
       "5       2        TP53   7577567   7577567  Nonsense_Mutation    SNP         A   \n",
       "6       5        TP53   7578392   7578392  Missense_Mutation    SNP         C   \n",
       "7       6        TP53   7578406   7578406  Missense_Mutation    SNP         C   \n",
       "8       7        TP53   7577141   7577141  Missense_Mutation    SNP         C   \n",
       "9       8        TP53   7579415   7579415  Nonsense_Mutation    SNP         C   \n",
       "10      9        TP53   7577096   7577096  Missense_Mutation    SNP         T   \n",
       "11     10        TP53   7577094   7577094  Missense_Mutation    SNP         G   \n",
       "12     12        TP53   7578235   7578235  Missense_Mutation    SNP         T   \n",
       "13     13        TP53   7577539   7577539  Missense_Mutation    SNP         G   \n",
       "14     14        TP53   7578268   7578268  Missense_Mutation    SNP         A   \n",
       "15     15        TP53   7578536   7578536  Missense_Mutation    SNP         T   \n",
       "16     16        TP53   7577068   7577068    Frame_Shift_Del    DEL         G   \n",
       "\n",
       "   alt_seq_h            tx_id_h ref_genome_h gene_name_m  \\\n",
       "0          A  ENST00000275493.2       GRCh37        Egfr   \n",
       "1          A  ENST00000256078.4       GRCh37        Kras   \n",
       "2          T  ENST00000256078.4       GRCh37        Kras   \n",
       "3          T  ENST00000256078.4       GRCh37        Kras   \n",
       "4          A  ENST00000269305.4       GRCh37       Trp53   \n",
       "5          T  ENST00000269305.4       GRCh37       Trp53   \n",
       "6          T  ENST00000269305.4       GRCh37       Trp53   \n",
       "7          T  ENST00000269305.4       GRCh37       Trp53   \n",
       "8          A  ENST00000269305.4       GRCh37       Trp53   \n",
       "9          T  ENST00000269305.4       GRCh37       Trp53   \n",
       "10         C  ENST00000269305.4       GRCh37       Trp53   \n",
       "11         A  ENST00000269305.4       GRCh37       Trp53   \n",
       "12         C  ENST00000269305.4       GRCh37       Trp53   \n",
       "13         A  ENST00000269305.4       GRCh37       Trp53   \n",
       "14         C  ENST00000269305.4       GRCh37       Trp53   \n",
       "15         G  ENST00000269305.4       GRCh37       Trp53   \n",
       "16         -  ENST00000269305.4       GRCh37       Trp53   \n",
       "\n",
       "                  tx_id_m  \n",
       "0   ENSMUST00000020329.13  \n",
       "1    ENSMUST00000111710.8  \n",
       "2    ENSMUST00000111710.8  \n",
       "3    ENSMUST00000111710.8  \n",
       "4   ENSMUST00000108658.10  \n",
       "5   ENSMUST00000108658.10  \n",
       "6   ENSMUST00000108658.10  \n",
       "7   ENSMUST00000108658.10  \n",
       "8   ENSMUST00000108658.10  \n",
       "9   ENSMUST00000108658.10  \n",
       "10  ENSMUST00000108658.10  \n",
       "11  ENSMUST00000108658.10  \n",
       "12  ENSMUST00000108658.10  \n",
       "13  ENSMUST00000108658.10  \n",
       "14  ENSMUST00000108658.10  \n",
       "15  ENSMUST00000108658.10  \n",
       "16  ENSMUST00000108658.10  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = h2m.get_tx_batch(df, species='m')[0]\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Batch modeling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No error occurs.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>gene_id_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>chr_h</th>\n",
       "      <th>exon_num_h</th>\n",
       "      <th>strand_h</th>\n",
       "      <th>match</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>HGVSc_h</th>\n",
       "      <th>HGVSp_h</th>\n",
       "      <th>classification_h</th>\n",
       "      <th>exon_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>status</th>\n",
       "      <th>class</th>\n",
       "      <th>statement</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>gene_id_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>exon_num_m</th>\n",
       "      <th>strand_m</th>\n",
       "      <th>type_m</th>\n",
       "      <th>classification_m</th>\n",
       "      <th>exon_m</th>\n",
       "      <th>start_m_ori</th>\n",
       "      <th>end_m_ori</th>\n",
       "      <th>ref_seq_m_ori</th>\n",
       "      <th>alt_seq_m_ori</th>\n",
       "      <th>HGVSc_m_ori</th>\n",
       "      <th>HGVSp_m_ori</th>\n",
       "      <th>start_m</th>\n",
       "      <th>end_m</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>alt_seq_m</th>\n",
       "      <th>HGVSc_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "      <th>index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7579311</td>\n",
       "      <td>7579311</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.279G&gt;T</td>\n",
       "      <td>None</td>\n",
       "      <td>Non_Coding,I</td>\n",
       "      <td>None</td>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "      <td>Class 4: Flanked segments are not identical.</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>ENSG00000133703.7</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>chr12</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000256078.4:c.34G&gt;T</td>\n",
       "      <td>G12V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Class 0: This mutation can be originally modeled.</td>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUSG00000030265.15</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "      <td>chr6</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_1</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000111710.8:c.34G&gt;T</td>\n",
       "      <td>G12V</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000111710.8:c.34G&gt;T</td>\n",
       "      <td>G12V</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7577567</td>\n",
       "      <td>7577567</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4:c.713T&gt;A</td>\n",
       "      <td>C238delins*</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>E_6</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Class 0: This mutation can be originally modeled.</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>E_6</td>\n",
       "      <td>69480008.0</td>\n",
       "      <td>69480008.0</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000108658.10:c.704T&gt;A</td>\n",
       "      <td>C235delins*</td>\n",
       "      <td>69480008.0</td>\n",
       "      <td>69480008.0</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000108658.10:c.704T&gt;A</td>\n",
       "      <td>C235delins*</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>ENSG00000133703.7</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>chr12</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000256078.4:c.34G&gt;A</td>\n",
       "      <td>G12D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Class 0: This mutation can be originally modeled.</td>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUSG00000030265.15</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "      <td>chr6</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_1</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000111710.8:c.34G&gt;A</td>\n",
       "      <td>G12D</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000111710.8:c.34G&gt;A</td>\n",
       "      <td>G12D</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>ENSG00000133703.7</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>chr12</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>25398281</td>\n",
       "      <td>25398281</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000256078.4:c.37G&gt;A</td>\n",
       "      <td>G13D</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_1</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Class 0: This mutation can be originally modeled.</td>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUSG00000030265.15</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "      <td>chr6</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_1</td>\n",
       "      <td>145192494.0</td>\n",
       "      <td>145192494.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000111710.8:c.37G&gt;A</td>\n",
       "      <td>G13D</td>\n",
       "      <td>145192494.0</td>\n",
       "      <td>145192494.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000111710.8:c.37G&gt;A</td>\n",
       "      <td>G13D</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7578392</td>\n",
       "      <td>7578392</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4:c.537G&gt;A</td>\n",
       "      <td>E180K</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_4</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Class 0: This mutation can be originally modeled.</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_4</td>\n",
       "      <td>69479352.0</td>\n",
       "      <td>69479352.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000108658.10:c.528G&gt;A</td>\n",
       "      <td>E177K</td>\n",
       "      <td>69479352.0</td>\n",
       "      <td>69479352.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000108658.10:c.528G&gt;A</td>\n",
       "      <td>E177K</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7578406</td>\n",
       "      <td>7578406</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4:c.523G&gt;A</td>\n",
       "      <td>R175H</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_4</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Class 0: This mutation can be originally modeled.</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_4</td>\n",
       "      <td>69479338.0</td>\n",
       "      <td>69479338.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000108658.10:c.514G&gt;A</td>\n",
       "      <td>R172H</td>\n",
       "      <td>69479338.0</td>\n",
       "      <td>69479338.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000108658.10:c.514G&gt;A</td>\n",
       "      <td>R172H</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7577141</td>\n",
       "      <td>7577141</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.796G&gt;T</td>\n",
       "      <td>G266V</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_7</td>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "      <td>Class 4: Flanked segments are not identical.</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7579415</td>\n",
       "      <td>7579415</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENST00000269305.4:c.271G&gt;A</td>\n",
       "      <td>W91delins*</td>\n",
       "      <td>Nonsense</td>\n",
       "      <td>E_3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "      <td>Class 4: Flanked segments are not identical.</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7577096</td>\n",
       "      <td>7577096</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENST00000269305.4:c.841A&gt;G</td>\n",
       "      <td>D281G</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_7</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Class 0: This mutation can be originally modeled.</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_7</td>\n",
       "      <td>69480458.0</td>\n",
       "      <td>69480458.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSMUST00000108658.10:c.832A&gt;G</td>\n",
       "      <td>D278G</td>\n",
       "      <td>69480458.0</td>\n",
       "      <td>69480458.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSMUST00000108658.10:c.832A&gt;G</td>\n",
       "      <td>D278G</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7577094</td>\n",
       "      <td>7577094</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.843C&gt;T</td>\n",
       "      <td>R282W</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_7</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>Class 1: This mutation can be alternatively mo...</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>ONP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_7</td>\n",
       "      <td>69480460.0</td>\n",
       "      <td>69480460.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000108658.10:c.834C&gt;T</td>\n",
       "      <td>R279C</td>\n",
       "      <td>69480460.0</td>\n",
       "      <td>69480462.0</td>\n",
       "      <td>CGC</td>\n",
       "      <td>TGG</td>\n",
       "      <td>ENSMUST00000108658.10:c.834_836CGC&gt;TGG</td>\n",
       "      <td>R279W</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENSG00000146648.11</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>28</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55249022</td>\n",
       "      <td>55249022</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000275493.2:c.2319G&gt;A</td>\n",
       "      <td>V774M</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_20</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>Class 1: This mutation can be alternatively mo...</td>\n",
       "      <td>Egfr</td>\n",
       "      <td>ENSMUSG00000020122.17</td>\n",
       "      <td>ENSMUST00000020329.13</td>\n",
       "      <td>chr11</td>\n",
       "      <td>28</td>\n",
       "      <td>+</td>\n",
       "      <td>ONP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_20</td>\n",
       "      <td>16846930.0</td>\n",
       "      <td>16846930.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000020329.13:c.2325G&gt;A</td>\n",
       "      <td>V776I</td>\n",
       "      <td>16846930.0</td>\n",
       "      <td>16846932.0</td>\n",
       "      <td>GTA</td>\n",
       "      <td>ATG</td>\n",
       "      <td>ENSMUST00000020329.13:c.2325_2327GTA&gt;ATG</td>\n",
       "      <td>V776M</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7578235</td>\n",
       "      <td>7578235</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENST00000269305.4:c.613A&gt;G</td>\n",
       "      <td>Y205C</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_5</td>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "      <td>Class 4: Flanked segments are not identical.</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7577539</td>\n",
       "      <td>7577539</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENST00000269305.4:c.741C&gt;T</td>\n",
       "      <td>R248W</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_6</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>Class 1: This mutation can be alternatively mo...</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>ONP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_6</td>\n",
       "      <td>69480036.0</td>\n",
       "      <td>69480036.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000108658.10:c.732C&gt;T</td>\n",
       "      <td>R245C</td>\n",
       "      <td>69480036.0</td>\n",
       "      <td>69480038.0</td>\n",
       "      <td>CGC</td>\n",
       "      <td>TGG</td>\n",
       "      <td>ENSMUST00000108658.10:c.732_734CGC&gt;TGG</td>\n",
       "      <td>R245W</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7578268</td>\n",
       "      <td>7578268</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ENST00000269305.4:c.580T&gt;G</td>\n",
       "      <td>L194R</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_5</td>\n",
       "      <td>SNP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>Class 0: This mutation can be originally modeled.</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>SNP</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_5</td>\n",
       "      <td>69479473.0</td>\n",
       "      <td>69479473.0</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSMUST00000108658.10:c.571T&gt;G</td>\n",
       "      <td>L191R</td>\n",
       "      <td>69479473.0</td>\n",
       "      <td>69479473.0</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSMUST00000108658.10:c.571T&gt;G</td>\n",
       "      <td>L191R</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7578536</td>\n",
       "      <td>7578536</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ENST00000269305.4:c.393A&gt;C</td>\n",
       "      <td>K132Q</td>\n",
       "      <td>Missense</td>\n",
       "      <td>E_4</td>\n",
       "      <td>SNP</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "      <td>Class 4: Flanked segments are not identical.</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENSG00000141510.11</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>chr17</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>7577068</td>\n",
       "      <td>7577068</td>\n",
       "      <td>G</td>\n",
       "      <td></td>\n",
       "      <td>ENST00000269305.4:c.869C&gt;</td>\n",
       "      <td>K291Rfs*54</td>\n",
       "      <td>Frame_Shifting</td>\n",
       "      <td>E_7</td>\n",
       "      <td>DEL</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "      <td>Class 4: Flanked segments are not identical.</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene_name_h           gene_id_h            tx_id_h  chr_h  exon_num_h  \\\n",
       "0         TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "1         KRAS   ENSG00000133703.7  ENST00000256078.4  chr12           4   \n",
       "2         TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "3         KRAS   ENSG00000133703.7  ENST00000256078.4  chr12           4   \n",
       "4         KRAS   ENSG00000133703.7  ENST00000256078.4  chr12           4   \n",
       "5         TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "6         TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "7         TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "8         TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "9         TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "10        TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "11        EGFR  ENSG00000146648.11  ENST00000275493.2   chr7          28   \n",
       "12        TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "13        TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "14        TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "15        TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "16        TP53  ENSG00000141510.11  ENST00000269305.4  chr17          10   \n",
       "\n",
       "   strand_h  match   start_h     end_h ref_seq_h alt_seq_h  \\\n",
       "0         -   True   7579311   7579311         C         A   \n",
       "1         -   True  25398284  25398284         C         A   \n",
       "2         -   True   7577567   7577567         A         T   \n",
       "3         -   True  25398284  25398284         C         T   \n",
       "4         -   True  25398281  25398281         C         T   \n",
       "5         -   True   7578392   7578392         C         T   \n",
       "6         -   True   7578406   7578406         C         T   \n",
       "7         -   True   7577141   7577141         C         A   \n",
       "8         -   True   7579415   7579415         C         T   \n",
       "9         -   True   7577096   7577096         T         C   \n",
       "10        -   True   7577094   7577094         G         A   \n",
       "11        +   True  55249022  55249022         G         A   \n",
       "12        -   True   7578235   7578235         T         C   \n",
       "13        -   True   7577539   7577539         G         A   \n",
       "14        -   True   7578268   7578268         A         C   \n",
       "15        -   True   7578536   7578536         T         G   \n",
       "16        -   True   7577068   7577068         G             \n",
       "\n",
       "                        HGVSc_h      HGVSp_h classification_h exon_h type_h  \\\n",
       "0    ENST00000269305.4:c.279G>T         None     Non_Coding,I   None    SNP   \n",
       "1     ENST00000256078.4:c.34G>T         G12V         Missense    E_1    SNP   \n",
       "2    ENST00000269305.4:c.713T>A  C238delins*         Nonsense    E_6    SNP   \n",
       "3     ENST00000256078.4:c.34G>A         G12D         Missense    E_1    SNP   \n",
       "4     ENST00000256078.4:c.37G>A         G13D         Missense    E_1    SNP   \n",
       "5    ENST00000269305.4:c.537G>A        E180K         Missense    E_4    SNP   \n",
       "6    ENST00000269305.4:c.523G>A        R175H         Missense    E_4    SNP   \n",
       "7    ENST00000269305.4:c.796G>T        G266V         Missense    E_7    SNP   \n",
       "8    ENST00000269305.4:c.271G>A   W91delins*         Nonsense    E_3    SNP   \n",
       "9    ENST00000269305.4:c.841A>G        D281G         Missense    E_7    SNP   \n",
       "10   ENST00000269305.4:c.843C>T        R282W         Missense    E_7    SNP   \n",
       "11  ENST00000275493.2:c.2319G>A        V774M         Missense   E_20    SNP   \n",
       "12   ENST00000269305.4:c.613A>G        Y205C         Missense    E_5    SNP   \n",
       "13   ENST00000269305.4:c.741C>T        R248W         Missense    E_6    SNP   \n",
       "14   ENST00000269305.4:c.580T>G        L194R         Missense    E_5    SNP   \n",
       "15   ENST00000269305.4:c.393A>C        K132Q         Missense    E_4    SNP   \n",
       "16    ENST00000269305.4:c.869C>   K291Rfs*54   Frame_Shifting    E_7    DEL   \n",
       "\n",
       "    status  class                                          statement  \\\n",
       "0    False      4       Class 4: Flanked segments are not identical.   \n",
       "1     True      0  Class 0: This mutation can be originally modeled.   \n",
       "2     True      0  Class 0: This mutation can be originally modeled.   \n",
       "3     True      0  Class 0: This mutation can be originally modeled.   \n",
       "4     True      0  Class 0: This mutation can be originally modeled.   \n",
       "5     True      0  Class 0: This mutation can be originally modeled.   \n",
       "6     True      0  Class 0: This mutation can be originally modeled.   \n",
       "7    False      4       Class 4: Flanked segments are not identical.   \n",
       "8    False      4       Class 4: Flanked segments are not identical.   \n",
       "9     True      0  Class 0: This mutation can be originally modeled.   \n",
       "10    True      1  Class 1: This mutation can be alternatively mo...   \n",
       "11    True      1  Class 1: This mutation can be alternatively mo...   \n",
       "12   False      4       Class 4: Flanked segments are not identical.   \n",
       "13    True      1  Class 1: This mutation can be alternatively mo...   \n",
       "14    True      0  Class 0: This mutation can be originally modeled.   \n",
       "15   False      4       Class 4: Flanked segments are not identical.   \n",
       "16   False      4       Class 4: Flanked segments are not identical.   \n",
       "\n",
       "   gene_name_m              gene_id_m                tx_id_m  chr_m  \\\n",
       "0        Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "1         Kras  ENSMUSG00000030265.15   ENSMUST00000111710.8   chr6   \n",
       "2        Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "3         Kras  ENSMUSG00000030265.15   ENSMUST00000111710.8   chr6   \n",
       "4         Kras  ENSMUSG00000030265.15   ENSMUST00000111710.8   chr6   \n",
       "5        Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "6        Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "7        Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "8        Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "9        Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "10       Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "11        Egfr  ENSMUSG00000020122.17  ENSMUST00000020329.13  chr11   \n",
       "12       Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "13       Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "14       Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "15       Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "16       Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "\n",
       "    exon_num_m strand_m type_m classification_m exon_m  start_m_ori  \\\n",
       "0           10        +   None             None   None          NaN   \n",
       "1            4        -    SNP         Missense    E_1  145192497.0   \n",
       "2           10        +    SNP         Nonsense    E_6   69480008.0   \n",
       "3            4        -    SNP         Missense    E_1  145192497.0   \n",
       "4            4        -    SNP         Missense    E_1  145192494.0   \n",
       "5           10        +    SNP         Missense    E_4   69479352.0   \n",
       "6           10        +    SNP         Missense    E_4   69479338.0   \n",
       "7           10        +   None             None   None          NaN   \n",
       "8           10        +   None             None   None          NaN   \n",
       "9           10        +    SNP         Missense    E_7   69480458.0   \n",
       "10          10        +    ONP         Missense    E_7   69480460.0   \n",
       "11          28        +    ONP         Missense   E_20   16846930.0   \n",
       "12          10        +   None             None   None          NaN   \n",
       "13          10        +    ONP         Missense    E_6   69480036.0   \n",
       "14          10        +    SNP         Missense    E_5   69479473.0   \n",
       "15          10        +   None             None   None          NaN   \n",
       "16          10        +   None             None   None          NaN   \n",
       "\n",
       "      end_m_ori ref_seq_m_ori alt_seq_m_ori                      HGVSc_m_ori  \\\n",
       "0           NaN          None          None                             None   \n",
       "1   145192497.0             C             A     ENSMUST00000111710.8:c.34G>T   \n",
       "2    69480008.0             T             A   ENSMUST00000108658.10:c.704T>A   \n",
       "3   145192497.0             C             T     ENSMUST00000111710.8:c.34G>A   \n",
       "4   145192494.0             C             T     ENSMUST00000111710.8:c.37G>A   \n",
       "5    69479352.0             G             A   ENSMUST00000108658.10:c.528G>A   \n",
       "6    69479338.0             G             A   ENSMUST00000108658.10:c.514G>A   \n",
       "7           NaN          None          None                             None   \n",
       "8           NaN          None          None                             None   \n",
       "9    69480458.0             A             G   ENSMUST00000108658.10:c.832A>G   \n",
       "10   69480460.0             C             T   ENSMUST00000108658.10:c.834C>T   \n",
       "11   16846930.0             G             A  ENSMUST00000020329.13:c.2325G>A   \n",
       "12          NaN          None          None                             None   \n",
       "13   69480036.0             C             T   ENSMUST00000108658.10:c.732C>T   \n",
       "14   69479473.0             T             G   ENSMUST00000108658.10:c.571T>G   \n",
       "15          NaN          None          None                             None   \n",
       "16          NaN          None          None                             None   \n",
       "\n",
       "    HGVSp_m_ori      start_m        end_m ref_seq_m alt_seq_m  \\\n",
       "0          None          NaN          NaN      None      None   \n",
       "1          G12V  145192497.0  145192497.0         C         A   \n",
       "2   C235delins*   69480008.0   69480008.0         T         A   \n",
       "3          G12D  145192497.0  145192497.0         C         T   \n",
       "4          G13D  145192494.0  145192494.0         C         T   \n",
       "5         E177K   69479352.0   69479352.0         G         A   \n",
       "6         R172H   69479338.0   69479338.0         G         A   \n",
       "7          None          NaN          NaN      None      None   \n",
       "8          None          NaN          NaN      None      None   \n",
       "9         D278G   69480458.0   69480458.0         A         G   \n",
       "10        R279C   69480460.0   69480462.0       CGC       TGG   \n",
       "11        V776I   16846930.0   16846932.0       GTA       ATG   \n",
       "12         None          NaN          NaN      None      None   \n",
       "13        R245C   69480036.0   69480038.0       CGC       TGG   \n",
       "14        L191R   69479473.0   69479473.0         T         G   \n",
       "15         None          NaN          NaN      None      None   \n",
       "16         None          NaN          NaN      None      None   \n",
       "\n",
       "                                     HGVSc_m      HGVSp_m  index  \n",
       "0                                       None         None      0  \n",
       "1               ENSMUST00000111710.8:c.34G>T         G12V      1  \n",
       "2             ENSMUST00000108658.10:c.704T>A  C235delins*      2  \n",
       "3               ENSMUST00000111710.8:c.34G>A         G12D      3  \n",
       "4               ENSMUST00000111710.8:c.37G>A         G13D      4  \n",
       "5             ENSMUST00000108658.10:c.528G>A        E177K      5  \n",
       "6             ENSMUST00000108658.10:c.514G>A        R172H      6  \n",
       "7                                       None         None      7  \n",
       "8                                       None         None      8  \n",
       "9             ENSMUST00000108658.10:c.832A>G        D278G      9  \n",
       "10    ENSMUST00000108658.10:c.834_836CGC>TGG        R279W     10  \n",
       "11  ENSMUST00000020329.13:c.2325_2327GTA>ATG        V776M     11  \n",
       "12                                      None         None     12  \n",
       "13    ENSMUST00000108658.10:c.732_734CGC>TGG        R245W     13  \n",
       "14            ENSMUST00000108658.10:c.571T>G        L191R     14  \n",
       "15                                      None         None     15  \n",
       "16                                      None         None     16  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "result = h2m.model_batch(df, records_h, index_list_h, records_m, index_list_m, db_h, db_m, 37)\n",
    "result[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Appendix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Update History"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1.1.0 Jan 16, 2023\n",
    "- how to get strand_h  \n",
    "- compatibility of null exonlist  \n",
    "- compatibility of non-coding mutations  \n",
    "- add save common canonical tx dicitonaries\n",
    "\n",
    "1.1.1 Feb 7, 2023\n",
    "- amino acid function\n",
    "- batch processing function"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Database Sources"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*Local* \n",
    "**Dictionaries**   \n",
    "1. name_dict: Dictionary of human ensembl id to human gene name.  \n",
    "2. homo_dict: Dictionary of to human gene name to all  (json-like format). (Prepared by using biomart)  \n",
    "3. human_only_list: List of human-only gene names. (Prepared by using biomart)\n",
    "\n",
    "**Databases**\n",
    "1. GENECODE Databases  \n",
    "- human version 37  \n",
    "- human version 38  \n",
    "- mouse  \n",
    "2. Refgenome\n",
    "- human version 37  \n",
    "- human version 38  \n",
    "- mouse  \n",
    "\n",
    "*Online*\n",
    "ENSEMBL API"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![workflow](images/workflow.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### cbio_reader( ):   \n",
    "- Function\n",
    "    - Generate h2m input from cbioportal data.  \n",
    "\n",
    "- Parameter:\n",
    "    - path (str): the path of mutation data in txt format.\n",
    "    - keep (bool): True: keep all the original columns in the dataframe/ False: keep the necesssary columns for h2m only. Default to False.  \n",
    "\n",
    "- Output: \n",
    "    - An input dataframe for h2m modeling.\n",
    "\n",
    "- Example:  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "h2m.cbio_reader('.../data_mutations.txt', keep=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### clinvar_reader( ):\n",
    "\n",
    "- Generate h2m input from ClinVar data.  \n",
    "\n",
    "- Parameter:\n",
    "    - path (str): the path of clinvar renference vcf.gz data.\n",
    "    - list_of_ids (list): the list of variation ids.\n",
    "    - keep (bool): True: keep all the original columns in the dataframe/ False: keep the necesssary columns for h2m only. Default to False.  \n",
    "\n",
    "- Output: \n",
    "    - An input dataframe for h2m modeling.\n",
    "\n",
    "- Example:  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath = '.../GrCh37_clinvar_20230923.vcf.gz'  \n",
    "variation_ids = [925574, 925434, 926695, 925707, 325626, 1191613, 308061, 361149, 1205375, 208043]   \n",
    "df = h2m.clinvar_reader(filepath, variation_ids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### genomad_reader( ):\n",
    "\n",
    "- Generate h2m input from GenomAD data.  \n",
    "\n",
    "- Parameter:\n",
    "    - path (str): the path of GenomAD csv data.\n",
    "    - gene_name (str): the gene name.\n",
    "    - keep (bool): True: keep all the original columns in the dataframe/ False: keep the necesssary columns for h2m only. Default to False.  \n",
    "\n",
    "- Output: \n",
    "    - An input dataframe for h2m modeling.\n",
    "\n",
    "- Example: "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = h2m.genomad_reader('/Users/gorkordkx/Downloads/gnomAD_v4.0.0_ENSG00000141510_2024_02_07_11_36_03.csv','TP53')\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### query( )  \n",
    "- Function  \n",
    "    -  Query a human gene for the presence of mouse homologs.  \n",
    "- Parameters   \n",
    "    - id: string, identification of a human gene. Multiple input forms are accepted, including gene name, stable ensembl gene id with or without version number.  \n",
    "    - ver (optional): int, specify the version of human, one of 37/38. Only necessary when using human gene ensembl id.  \n",
    "    -  ty (optional): type of your input id. string, one of *name, gene_id*. Not necessary.  \n",
    " show (optional): bool, print summary of output or not.  \n",
    "- output  \n",
    "    -  A list of all mouse orthologs (if have; sometimes more than one.) Each element is a dictionary of mouse gene name, mouse gene id, homology type and similarity of human and mouse gene in percentage.  \n",
    "- Example  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 462,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query human gene: TP53;\n",
      "Mouse ortholog(s): Trp53;\n",
      "Homology type: one2one;\n",
      "Sequence Simalarity(%):77.3537.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'gene_name_m': 'Trp53',\n",
       "  'gene_id_m': 'ENSMUSG00000059552',\n",
       "  'homology_type': 'ortholog_one2one',\n",
       "  'similarity': 77.3537}]"
      ]
     },
     "execution_count": 462,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h2m.query('TP53')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 465,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query human gene: TP53;\n",
      "Mouse ortholog(s): Trp53;\n",
      "Homology type: one2one;\n",
      "Sequence Simalarity(%):77.3537.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'gene_name_m': 'Trp53',\n",
       "  'gene_id_m': 'ENSMUSG00000059552',\n",
       "  'homology_type': 'ortholog_one2one',\n",
       "  'similarity': 77.3537}]"
      ]
     },
     "execution_count": 465,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h2m.query('ENSG00000141510')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.4 query_batch( )    \n",
    "- Function  \n",
    "    - Query a human or mouse gene for coordinate and information of all its transcripts.  \n",
    "\n",
    "- Parameters   \n",
    "    - id: string, identification of a human gene. Multiple input forms are accepted, including gene   \n",
    "    - species: string, ‘h’ for human or ‘m’ for mouse.  \n",
    "    - name, stable ensembl gene id with or without version number.  \n",
    "    - ver: int, specify the version of human, one of 37/38. It is a necessary parameter.  \n",
    "    - ty (optional): type of your input id. string, one of *name, gene_id*.\n",
    "    - show (optional): bool, print summary of output or not.  \n",
    "- output  \n",
    "    - A list [chromosome, start location(of gene), end location(of gene), canonical transcript id, list of all transcript id (the canonical one included and always at the first place), a list of additional information of each transcript]  \n",
    "- example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "h2m.query_batch(df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get_tx( )    \n",
    "- Function  \n",
    "    - Query a human or mouse gene for coordinate and information of all its transcripts.  \n",
    "\n",
    "- Parameters   \n",
    "    - id: string, identification of a human gene. Multiple input forms are accepted, including gene   \n",
    "    - species: string, ‘h’ for human or ‘m’ for mouse.  \n",
    "    - name, stable ensembl gene id with or without version number.  \n",
    "    - ver: int, specify the version of human, one of 37/38. It is a necessary parameter.  \n",
    "    - ty (optional): type of your input id. string, one of *name, gene_id*.\n",
    "    - show (optional): bool, print summary of output or not.  \n",
    "- output  \n",
    "    - A list [chromosome, start location(of gene), end location(of gene), canonical transcript id, list of all transcript id (the canonical one included and always at the first place), a list of additional information of each transcript]  \n",
    "- example  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome assembly: GRCh37;\n",
      "The canonical transcript is: ENST00000269305.4;\n",
      "You can choose from the 17 transcripts below for further analysis:\n",
      "(1)ENST00000269305.4 (2)ENST00000413465.2 (3)ENST00000359597.4 (4)ENST00000504290.1 (5)ENST00000510385.1 (6)ENST00000504937.1 (7)ENST00000455263.2 (8)ENST00000420246.2 (9)ENST00000445888.2 (10)ENST00000576024.1 (11)ENST00000509690.1 (12)ENST00000514944.1 (13)ENST00000574684.1 (14)ENST00000505014.1 (15)ENST00000508793.1 (16)ENST00000604348.1 (17)ENST00000503591.1\n",
      "\n"
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the the current cell or a previous cell. Please review the code in the cell(s) to identify a possible cause of the failure. Click <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. View Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "x = h2m.get_tx_id('TP53','h',ver=37)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### get_tx_batch( )\n",
    "- Function:\n",
    "    - Batch query of canonical transcript IDs of human or mouse genes.\n",
    "\n",
    "- Parameters:  \n",
    "    - df (Pandas DataFrame): Must include a column of gene names named 'gene_name_h'/'gene_name_m', depending on the species. An index column is recommended.\n",
    "    - species (str): 'h' for human or 'm' for mouse.  \n",
    "    - ver (int): specify the version of human, one of 37/38.\n",
    "\n",
    "- Return:  \n",
    "    - Two dataframes. The first dataframe is the processed original dataframe with canonical transcirpt id attached in the column named 'tx_id_h'/'tx_id_m'. Thesecond dataframe contains all rows that are not successfully processed.\n",
    "\n",
    "- Example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "h2m.get_tx_batch(df,'h',ver=37)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### model( )  \n",
    "- Function  \n",
    "    - Model human variants in the mouse genome.  \n",
    "\n",
    "- Workflow  \n",
    "\n",
    "![model_workflow](images/model_workflow.png)  \n",
    "\n",
    "- Original modeling with uncertain effects:  \n",
    "![original_uncertain](images/original_uncertain.png)  \n",
    "\n",
    "- Alternative modeling with uncertain effects:  \n",
    "![alternative](images/alternative.png)  \n",
    "\n",
    "- All possible status:  \n",
    "![status](images/status.png)  \n",
    "\n",
    "- Parameters   \n",
    "    - records_h, index_list_h, records_m, index_list_m: human and mouse reference genome.  \n",
    "    - db_h, db_m: human and mouse GENCODE annotation.  \n",
    "    - tx_id_h, tx_id_m: human and mouse transcript id (could get by **h2m.get_tx_id()**).  \n",
    "    - start_h, end_h: int, start and end location of the mutation on the chromosome.  \n",
    "    - alt_seq_h: str, human mutation alternative sequence.  \n",
    "    - ty_h: str, human variantion type. One of ['SNP', 'DNP', 'ONP', 'INS', 'DEL'].  \n",
    "    - ver: int, human ref genome number. 37 or 38.  \n",
    "    - ref_seq_h (optional): str, human mutation reference sequence.    \n",
    "    - param (optional): set param = 'BE' and will only output base editing modelable results.  \n",
    "    - search_alternative (optional): set search_alternative = False and will only output original modeling results.  \n",
    "    - max_alternative (optional): the maximum number of output alternatives of one human variants.  \n",
    "    - nonstop_size (optional): the length of neucleotides that are included after the stop codon for alignment and translation in case of the nonstop mutations or frame shifting mutations.  \n",
    "    - flank_size (optional): the number of amino acids or neucleotides (for non-coding mutations) that are included on each side of the mutation site.  \n",
    "    - splicing_size (optional): \n",
    "    - batch (optional): set batch = True and will use input align_dict to save time in batch processing.  \n",
    "    - show_sequence (optional): set batch = True and will output the whole sequences.  \n",
    "    - align_dict (optional): input a prepared dictionary of alignment indexes to save time in batch processing.  \n",
    "    - memory_protect (optional): default True. Break long alignments that may lead to death of the kernel.    \n",
    "    - memory_size (optional): maxlength of aligned sequence when memory_protect == True.  \n",
    "\n",
    "- output  \n",
    "    - a dictionary or a list of dictionaries that include comprehensive information for the human and mouse variants.  \n",
    "\n",
    "- Other rules  \n",
    "    1. If the mutation falls in the coding and non-coding regions at the same time, it would be considered and processed as a ORIGIAL-MODELING ONLY mutation.\n",
    "    3. The alt_seq input should be in the positive strand and the start_h coordinate should be smaller than or equal the end_h coordinate.\n",
    "    4. If the ref-seq or alt-see has no length, it could be input as ‘’ or ‘-’.\n",
    "\n",
    "- example  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gene_name_h': 'TP53',\n",
       " 'gene_id_h': 'ENSG00000141510.11',\n",
       " 'tx_id_m': 'ENSMUST00000108658.10',\n",
       " 'chr_h': 'chr17',\n",
       " 'exon_num_h': 10,\n",
       " 'strand_h': '-',\n",
       " 'start_h': 7577120,\n",
       " 'end_h': 7577120,\n",
       " 'ref_seq_h': 'C',\n",
       " 'alt_seq_h': 'T',\n",
       " 'HGVSp_h': 'R273H',\n",
       " 'classification_h': 'Missense',\n",
       " 'classification_h_detail': None,\n",
       " 'exon_h': 'E_7',\n",
       " 'type_h': 'SNP',\n",
       " 'status': True,\n",
       " 'error': 0,\n",
       " 'statement': 'This mutation can be originally modeled.',\n",
       " 'gene_name_m': 'Trp53',\n",
       " 'gene_id_m': 'ENSMUSG00000059552.14',\n",
       " 'chr_m': 'chr11',\n",
       " 'exon_num_m': 10,\n",
       " 'strand_m': '+',\n",
       " 'type_m': 'SNP',\n",
       " 'classification_m': 'Missense',\n",
       " 'exon_m': 'E_7',\n",
       " 'start_m_ori': 69480434,\n",
       " 'end_m_ori': 69480434,\n",
       " 'ref_seq_m_ori': 'G',\n",
       " 'alt_seq_m_ori': 'A',\n",
       " 'HGVSp_m_ori': 'R270H',\n",
       " 'start_m': 69480434,\n",
       " 'end_m': 69480434,\n",
       " 'ref_seq_m': 'G',\n",
       " 'alt_seq_m': 'A',\n",
       " 'HGVSp_m': 'R270H'}"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tx_id_h, tx_id_m = list_tx_id_h[3], list_tx_id_m[3]\n",
    "# use the canonical transcript\n",
    "model_result = h2m.model(records_h,index_list_h, records_m, index_list_m, db_h, db_m, tx_id_h, tx_id_m, 7577120, 7577120, 'T', ty_h = 'SNP', ver = 37)\n",
    "model_result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### model_batch( )\n",
    "- Function:\n",
    "    - Batch modeling of human variants in the mouse genome.  \n",
    "\n",
    "- Parameters:  \n",
    "    - df (Pandas DataFrame): Must include columns {'start_h','end_h','type_h','ref_seq_h','alt_seq_h','tx_id_h','tx_id_m','index'}.\n",
    "    - ecords_h, index_list_h, records_m, index_list_m: reference genome files\n",
    "    - db_h, db_m: genomic annotation files\n",
    "    - ver (int): specify the version of human, one of 37/38.\n",
    "    - param (optional): set param = 'BE' and will only output base editing modelable results.  \n",
    "    - coor (optional): default = 'nc'. set input = 'aa' and will be compatable with input of amino acid variants.\n",
    "    - search_alternative (optional): set search_alternative = False and will only output original modeling results.  \n",
    "    - max_alternative (optional): the maximum number of output alternatives of one human variants.  \n",
    "    - nonstop_size (optional): the length of neucleotides that are included after the stop codon for alignment and translation in case of the nonstop mutations orframe shifting mutations.  \n",
    "    - flank_size (optional): the number of amino acids or neucleotides (for non-coding mutations) that are included on each side of the mutation site.  \n",
    "    - splicing_size (optional): \n",
    "    - batch (optional): set batch = True and will use input align_dict to save time in batch processing.  \n",
    "    - show_sequence (optional): set batch = True and will output the whole sequences.  \n",
    "    - align_dict (optional): input a prepared dictionary of alignment indexes to save time in batch processing.  \n",
    "    - memory_protect (optional): default True. Break long alignments that may lead to death of the kernel.    \n",
    "    - memory_size (optional): maxlength of aligned sequence when memory_protect == True.  \n",
    "\n",
    "- Return:  \n",
    "    - Two dataframes. The first dataframe is the processed original dataframe. The second dataframe contains all rows that are not successfully processed.\n",
    "\n",
    "- Example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "h2m.model_batch(df, records_h, index_list_h, records_m, index_list_m, db_h, db_m, ver = 37, param = 'BE')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
