{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Retrieve the human-mouse orthologous genes and canonical transcripts  \n",
    "This script generates H2M Database. \n",
    "\n",
    "**Author**: Kexin Dong  \n",
    "**Date**: Apr 9, 2024    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 0. Import packages  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/lib/python3.12/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import bioh2m as h2m\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "plt.rcParams['font.family'] = 'sans-serif'\n",
    "plt.rcParams['font.sans-serif'] = ['Arial']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def label_numbers(numbers):\n",
    "    last_number = None\n",
    "    count = 0\n",
    "    labeled_numbers = []\n",
    "\n",
    "    for num in numbers:\n",
    "        if num == last_number:\n",
    "            count += 1\n",
    "        else:\n",
    "            count = 1\n",
    "        last_number = num\n",
    "        labeled_numbers.append(count)\n",
    "\n",
    "    return labeled_numbers\n",
    "\n",
    "\n",
    "def get_label(h,x):\n",
    "    return f\"{h}-M{x:02}\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Generate input data of H2M from downloaded raw data  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (1) AACR-GENIE  \n",
    "Downloaded from Synapse (https://www.synapse.org/#!Synapse:syn53210170)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "851088"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path_aacr = '/Users/kexindong/Documents/GitHub/Database/PublicDatabase/AACR-GENIE/v15.0/data_mutations_extended.txt'\n",
    "df = h2m.cbio_reader(path_aacr, keep=True)\n",
    "df = pd.DataFrame(df[['gene_name_h','tx_id_h','start_h','end_h','ref_seq_h','alt_seq_h','type_h']].value_counts()).reset_index(drop=False)\n",
    "# count frequencies\n",
    "len(df) #851088"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['ID'] = [f\"AACR-H{(x+1):07}\" for x in list(range(len(df)))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/aacr_input_original.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/aacr_input_original.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Manually check gene symbols and add-ons  \n",
    "\n",
    "Histone proteins, G protein-coupled receptors, \n",
    "NO MOUSE ORTHOLOG: GPR32, RPL21P4 (pseudogene), hcg4p8 (pseudogene), CLTCL1, IFIT1P1 (pseudogene), Prkacg, segments of the T-cell receptor alpha joining (TRAJ) gene TRAJ59.  TCRs. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_catalog = pd.read_csv('/Users/kexindong/Desktop/Supp_Table_1_Homo_Genes_manually_updated.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_of_aacr_symbol = {\n",
    "    'H3F3A':'H3-3A',\n",
    "    'H3F3B':'H3-3B',\n",
    "    'WHSC1':'NSD1',\n",
    "    'GPR124':'ADGRA2',\n",
    "    'MKL1':'MRTFA',\n",
    "    'PARK2':'PRKN',\n",
    "    'MRE11A':'MRE11',\n",
    "    'MLLT4':'AFDN',\n",
    "    'CASC5': 'KNL1',\n",
    "    'MEF2BNB-MEF2B':'BORCS8-MEF2B',\n",
    "    'PAK7':'PAK5',\n",
    "    'HIST1H1C':'H1-2',\n",
    "    'HIST1H3B': 'H3C1',\n",
    "    'HIST1H1E': 'H1-4',\n",
    "    'HIST1H2BD': 'H2BC5',\n",
    "    'HIST3H3': 'H3-4',\n",
    "    'HIST1H3D': 'H3C2',\n",
    "    'HIST1H3E': 'H3C3',\n",
    "    'HIST1H3J': 'H3C12',\n",
    "    'HIST1H2BK': 'H2BC8',\n",
    "    'HIST1H3F': 'H3C4',\n",
    "    'HIST1H1B': 'H1-5',\n",
    "    'HIST1H1D': 'H1-3',\n",
    "    'HIST1H2AC': 'H2AC1',\n",
    "    'HIST1H2BJ': 'H2BC13',\n",
    "    'HIST1H4E': 'H4C5',\n",
    "    'FAM46C':'TENT5C',\n",
    "    'WHSC1L1':'NSD3',\n",
    "    'SETD8':'KMT5A',\n",
    "    'LPHN3':'ADGRL3',\n",
    "    'BAI3':'ADGRB3',\n",
    "    'SEPT9':'NAPB',\n",
    "    'BRE':'BABAM2',\n",
    "    'RFWD2':'COP1',\n",
    "    'TCEB1':'ELOC',\n",
    "    'GBA':'GBA1',\n",
    "    'PVRL4':'NECTIN4',\n",
    "    'ICK':'GCKR',\n",
    "    'GNB2L1':'RACK1',\n",
    "    'MGEA5':'OGA',\n",
    "    'DIRC2':'SLC49A4',\n",
    "    'LARGE':'LARGE1',\n",
    "    'TMEM173B':'STING1',\n",
    "    'SEPT5':'SEPTIN5'\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_catalog['gene_name_h'] =  [dict_of_aacr_symbol.get(x,x) for x in df_catalog['gene_name_h']]\n",
    "df_catalog['gene_id_h_37'] = [h2m.dict_tx_h[x][0] for x in df_catalog['gene_name_h']] \n",
    "df_catalog['gene_id_h_37'] = [np.nan if x is None else x for x in df_catalog['gene_id_h_37']]\n",
    "df_catalog.to_csv('/Users/kexindong/Desktop/Supp_Table_1_Homo_Genes_manually_updated.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['gene_name_h'] = [dict_of_aacr_symbol.get(x,x) for x in df['gene_name_h']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There were rows that could not be processed.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>count</th>\n",
       "      <th>ID</th>\n",
       "      <th>gene_name_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENST00000265724.3</td>\n",
       "      <td>87225124</td>\n",
       "      <td>87225124</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>2</td>\n",
       "      <td>AACR-H0155944</td>\n",
       "      <td>Abcb1a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENST00000265724.3</td>\n",
       "      <td>87225124</td>\n",
       "      <td>87225124</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>2</td>\n",
       "      <td>AACR-H0155944</td>\n",
       "      <td>Abcb1b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENST00000265724.3</td>\n",
       "      <td>87174220</td>\n",
       "      <td>87174221</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>2</td>\n",
       "      <td>AACR-H0156000</td>\n",
       "      <td>Abcb1a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENST00000265724.3</td>\n",
       "      <td>87174220</td>\n",
       "      <td>87174221</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>2</td>\n",
       "      <td>AACR-H0156000</td>\n",
       "      <td>Abcb1b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENST00000265724.3</td>\n",
       "      <td>87133531</td>\n",
       "      <td>87133531</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0260329</td>\n",
       "      <td>Abcb1a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926619</th>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>ENST00000254323.2</td>\n",
       "      <td>13923972</td>\n",
       "      <td>13923972</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0384705</td>\n",
       "      <td>Zswim4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926620</th>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>ENST00000254323.2</td>\n",
       "      <td>13928736</td>\n",
       "      <td>13928736</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0384706</td>\n",
       "      <td>Zswim4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926621</th>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>ENST00000254323.2</td>\n",
       "      <td>13928772</td>\n",
       "      <td>13928772</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0384707</td>\n",
       "      <td>Zswim4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926622</th>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>ENST00000254323.2</td>\n",
       "      <td>13930268</td>\n",
       "      <td>13930268</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0384708</td>\n",
       "      <td>Zswim4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926623</th>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>ENST00000254323.2</td>\n",
       "      <td>13941861</td>\n",
       "      <td>13941861</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0851088</td>\n",
       "      <td>Zswim4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>926624 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_name_h            tx_id_h   start_h     end_h ref_seq_h alt_seq_h  \\\n",
       "0            ABCB1  ENST00000265724.3  87225124  87225124         T         -   \n",
       "1            ABCB1  ENST00000265724.3  87225124  87225124         T         -   \n",
       "2            ABCB1  ENST00000265724.3  87174220  87174221         -         C   \n",
       "3            ABCB1  ENST00000265724.3  87174220  87174221         -         C   \n",
       "4            ABCB1  ENST00000265724.3  87133531  87133531         A         T   \n",
       "...            ...                ...       ...       ...       ...       ...   \n",
       "926619      ZSWIM4  ENST00000254323.2  13923972  13923972         G         A   \n",
       "926620      ZSWIM4  ENST00000254323.2  13928736  13928736         T         C   \n",
       "926621      ZSWIM4  ENST00000254323.2  13928772  13928772         G         T   \n",
       "926622      ZSWIM4  ENST00000254323.2  13930268  13930268         C         T   \n",
       "926623      ZSWIM4  ENST00000254323.2  13941861  13941861         T         C   \n",
       "\n",
       "       type_h  count             ID gene_name_m  \n",
       "0         DEL      2  AACR-H0155944      Abcb1a  \n",
       "1         DEL      2  AACR-H0155944      Abcb1b  \n",
       "2         INS      2  AACR-H0156000      Abcb1a  \n",
       "3         INS      2  AACR-H0156000      Abcb1b  \n",
       "4         SNP      1  AACR-H0260329      Abcb1a  \n",
       "...       ...    ...            ...         ...  \n",
       "926619    SNP      1  AACR-H0384705      Zswim4  \n",
       "926620    SNP      1  AACR-H0384706      Zswim4  \n",
       "926621    SNP      1  AACR-H0384707      Zswim4  \n",
       "926622    SNP      1  AACR-H0384708      Zswim4  \n",
       "926623    SNP      1  AACR-H0851088      Zswim4  \n",
       "\n",
       "[926624 rows x 10 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df, df_wrong = h2m.query_batch(df)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AACR-GENIE # genes mapped with mouse orthologs: 1477'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_1 = len(df['gene_name_h'].unique())\n",
    "f'AACR-GENIE # genes mapped with mouse orthologs: {l_1}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AACR-GENIE # genes mapped with mouse orthologs: 154'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_2 = len(df_wrong['gene_name_h'].unique())\n",
    "f'AACR-GENIE # genes mapped with mouse orthologs: {l_2}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(set_unmapped)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_mapped = set(df['gene_name_h'].unique())\n",
    "set_unmapped = set(df_wrong['gene_name_h'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_catalog['gene_name_h'] =  [dict_of_aacr_symbol.get(x,x) for x in df_catalog['gene_name_h']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_wrong.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/aacr_without_homo_gene.csv',index=False)\n",
    "# df_wrong = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/aacr_without_homo_gene.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No error occurs.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>count</th>\n",
       "      <th>ID</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENST00000265724.3</td>\n",
       "      <td>87225124</td>\n",
       "      <td>87225124</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>2</td>\n",
       "      <td>AACR-H0155944</td>\n",
       "      <td>Abcb1a</td>\n",
       "      <td>ENSMUST00000047753.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENST00000265724.3</td>\n",
       "      <td>87225124</td>\n",
       "      <td>87225124</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>2</td>\n",
       "      <td>AACR-H0155944</td>\n",
       "      <td>Abcb1b</td>\n",
       "      <td>ENSMUST00000009058.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENST00000265724.3</td>\n",
       "      <td>87174220</td>\n",
       "      <td>87174221</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>2</td>\n",
       "      <td>AACR-H0156000</td>\n",
       "      <td>Abcb1a</td>\n",
       "      <td>ENSMUST00000047753.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENST00000265724.3</td>\n",
       "      <td>87174220</td>\n",
       "      <td>87174221</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>2</td>\n",
       "      <td>AACR-H0156000</td>\n",
       "      <td>Abcb1b</td>\n",
       "      <td>ENSMUST00000009058.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENST00000265724.3</td>\n",
       "      <td>87133531</td>\n",
       "      <td>87133531</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0260329</td>\n",
       "      <td>Abcb1a</td>\n",
       "      <td>ENSMUST00000047753.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926619</th>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>ENST00000254323.2</td>\n",
       "      <td>13923972</td>\n",
       "      <td>13923972</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0384705</td>\n",
       "      <td>Zswim4</td>\n",
       "      <td>ENSMUST00000039480.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926620</th>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>ENST00000254323.2</td>\n",
       "      <td>13928736</td>\n",
       "      <td>13928736</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0384706</td>\n",
       "      <td>Zswim4</td>\n",
       "      <td>ENSMUST00000039480.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926621</th>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>ENST00000254323.2</td>\n",
       "      <td>13928772</td>\n",
       "      <td>13928772</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0384707</td>\n",
       "      <td>Zswim4</td>\n",
       "      <td>ENSMUST00000039480.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926622</th>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>ENST00000254323.2</td>\n",
       "      <td>13930268</td>\n",
       "      <td>13930268</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0384708</td>\n",
       "      <td>Zswim4</td>\n",
       "      <td>ENSMUST00000039480.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926623</th>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>ENST00000254323.2</td>\n",
       "      <td>13941861</td>\n",
       "      <td>13941861</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0851088</td>\n",
       "      <td>Zswim4</td>\n",
       "      <td>ENSMUST00000039480.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>926624 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_name_h            tx_id_h   start_h     end_h ref_seq_h alt_seq_h  \\\n",
       "0            ABCB1  ENST00000265724.3  87225124  87225124         T         -   \n",
       "1            ABCB1  ENST00000265724.3  87225124  87225124         T         -   \n",
       "2            ABCB1  ENST00000265724.3  87174220  87174221         -         C   \n",
       "3            ABCB1  ENST00000265724.3  87174220  87174221         -         C   \n",
       "4            ABCB1  ENST00000265724.3  87133531  87133531         A         T   \n",
       "...            ...                ...       ...       ...       ...       ...   \n",
       "926619      ZSWIM4  ENST00000254323.2  13923972  13923972         G         A   \n",
       "926620      ZSWIM4  ENST00000254323.2  13928736  13928736         T         C   \n",
       "926621      ZSWIM4  ENST00000254323.2  13928772  13928772         G         T   \n",
       "926622      ZSWIM4  ENST00000254323.2  13930268  13930268         C         T   \n",
       "926623      ZSWIM4  ENST00000254323.2  13941861  13941861         T         C   \n",
       "\n",
       "       type_h  count             ID gene_name_m                tx_id_m  \n",
       "0         DEL      2  AACR-H0155944      Abcb1a   ENSMUST00000047753.5  \n",
       "1         DEL      2  AACR-H0155944      Abcb1b  ENSMUST00000009058.10  \n",
       "2         INS      2  AACR-H0156000      Abcb1a   ENSMUST00000047753.5  \n",
       "3         INS      2  AACR-H0156000      Abcb1b  ENSMUST00000009058.10  \n",
       "4         SNP      1  AACR-H0260329      Abcb1a   ENSMUST00000047753.5  \n",
       "...       ...    ...            ...         ...                    ...  \n",
       "926619    SNP      1  AACR-H0384705      Zswim4   ENSMUST00000039480.7  \n",
       "926620    SNP      1  AACR-H0384706      Zswim4   ENSMUST00000039480.7  \n",
       "926621    SNP      1  AACR-H0384707      Zswim4   ENSMUST00000039480.7  \n",
       "926622    SNP      1  AACR-H0384708      Zswim4   ENSMUST00000039480.7  \n",
       "926623    SNP      1  AACR-H0851088      Zswim4   ENSMUST00000039480.7  \n",
       "\n",
       "[926624 rows x 11 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df,df_wrong = h2m.get_tx_batch(df, species='m')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "numbers = label_numbers(df['ID'])\n",
    "df['ID'] = [get_label(h,x) for h,x in zip(df['ID'],numbers)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>count</th>\n",
       "      <th>ID</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENST00000265724.3</td>\n",
       "      <td>87225124</td>\n",
       "      <td>87225124</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>2</td>\n",
       "      <td>AACR-H0155944-M01</td>\n",
       "      <td>Abcb1a</td>\n",
       "      <td>ENSMUST00000047753.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENST00000265724.3</td>\n",
       "      <td>87225124</td>\n",
       "      <td>87225124</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>2</td>\n",
       "      <td>AACR-H0155944-M02</td>\n",
       "      <td>Abcb1b</td>\n",
       "      <td>ENSMUST00000009058.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENST00000265724.3</td>\n",
       "      <td>87174220</td>\n",
       "      <td>87174221</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>2</td>\n",
       "      <td>AACR-H0156000-M01</td>\n",
       "      <td>Abcb1a</td>\n",
       "      <td>ENSMUST00000047753.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENST00000265724.3</td>\n",
       "      <td>87174220</td>\n",
       "      <td>87174221</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>2</td>\n",
       "      <td>AACR-H0156000-M02</td>\n",
       "      <td>Abcb1b</td>\n",
       "      <td>ENSMUST00000009058.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCB1</td>\n",
       "      <td>ENST00000265724.3</td>\n",
       "      <td>87133531</td>\n",
       "      <td>87133531</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0260329-M01</td>\n",
       "      <td>Abcb1a</td>\n",
       "      <td>ENSMUST00000047753.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926619</th>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>ENST00000254323.2</td>\n",
       "      <td>13923972</td>\n",
       "      <td>13923972</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0384705-M01</td>\n",
       "      <td>Zswim4</td>\n",
       "      <td>ENSMUST00000039480.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926620</th>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>ENST00000254323.2</td>\n",
       "      <td>13928736</td>\n",
       "      <td>13928736</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0384706-M01</td>\n",
       "      <td>Zswim4</td>\n",
       "      <td>ENSMUST00000039480.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926621</th>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>ENST00000254323.2</td>\n",
       "      <td>13928772</td>\n",
       "      <td>13928772</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0384707-M01</td>\n",
       "      <td>Zswim4</td>\n",
       "      <td>ENSMUST00000039480.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926622</th>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>ENST00000254323.2</td>\n",
       "      <td>13930268</td>\n",
       "      <td>13930268</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0384708-M01</td>\n",
       "      <td>Zswim4</td>\n",
       "      <td>ENSMUST00000039480.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>926623</th>\n",
       "      <td>ZSWIM4</td>\n",
       "      <td>ENST00000254323.2</td>\n",
       "      <td>13941861</td>\n",
       "      <td>13941861</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "      <td>AACR-H0851088-M01</td>\n",
       "      <td>Zswim4</td>\n",
       "      <td>ENSMUST00000039480.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>926624 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_name_h            tx_id_h   start_h     end_h ref_seq_h alt_seq_h  \\\n",
       "0            ABCB1  ENST00000265724.3  87225124  87225124         T         -   \n",
       "1            ABCB1  ENST00000265724.3  87225124  87225124         T         -   \n",
       "2            ABCB1  ENST00000265724.3  87174220  87174221         -         C   \n",
       "3            ABCB1  ENST00000265724.3  87174220  87174221         -         C   \n",
       "4            ABCB1  ENST00000265724.3  87133531  87133531         A         T   \n",
       "...            ...                ...       ...       ...       ...       ...   \n",
       "926619      ZSWIM4  ENST00000254323.2  13923972  13923972         G         A   \n",
       "926620      ZSWIM4  ENST00000254323.2  13928736  13928736         T         C   \n",
       "926621      ZSWIM4  ENST00000254323.2  13928772  13928772         G         T   \n",
       "926622      ZSWIM4  ENST00000254323.2  13930268  13930268         C         T   \n",
       "926623      ZSWIM4  ENST00000254323.2  13941861  13941861         T         C   \n",
       "\n",
       "       type_h  count                 ID gene_name_m                tx_id_m  \n",
       "0         DEL      2  AACR-H0155944-M01      Abcb1a   ENSMUST00000047753.5  \n",
       "1         DEL      2  AACR-H0155944-M02      Abcb1b  ENSMUST00000009058.10  \n",
       "2         INS      2  AACR-H0156000-M01      Abcb1a   ENSMUST00000047753.5  \n",
       "3         INS      2  AACR-H0156000-M02      Abcb1b  ENSMUST00000009058.10  \n",
       "4         SNP      1  AACR-H0260329-M01      Abcb1a   ENSMUST00000047753.5  \n",
       "...       ...    ...                ...         ...                    ...  \n",
       "926619    SNP      1  AACR-H0384705-M01      Zswim4   ENSMUST00000039480.7  \n",
       "926620    SNP      1  AACR-H0384706-M01      Zswim4   ENSMUST00000039480.7  \n",
       "926621    SNP      1  AACR-H0384707-M01      Zswim4   ENSMUST00000039480.7  \n",
       "926622    SNP      1  AACR-H0384708-M01      Zswim4   ENSMUST00000039480.7  \n",
       "926623    SNP      1  AACR-H0851088-M01      Zswim4   ENSMUST00000039480.7  \n",
       "\n",
       "[926624 rows x 11 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "926624"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df.rename(columns={'ID':'index'})\n",
    "len(df['index'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/aacr_input.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_of_addon = set(df_wrong['ID']).intersection(set(df['ID']))\n",
    "df_addon_aacr = df[df['ID'].isin(set_of_addon)].reset_index(drop=True)\n",
    "df,df_wrong = h2m.get_tx_batch(df_addon_aacr, species='m')\n",
    "df\n",
    "numbers = label_numbers(df['ID'])\n",
    "df['ID'] = [get_label(h,x) for h,x in zip(df['ID'],numbers)]\n",
    "df = df.rename(columns={'ID':'index'})\n",
    "len(df['index'].unique())\n",
    "df.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/aacr_input_addon.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (2) COSMIC  \n",
    "Downloaded from COSMIC website (https://cancer.sanger.ac.uk/cosmic/download/cosmic/v99/mutantcensus)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cosmic = pd.read_csv('/Users/kexindong/Documents/GitHub/Database/PublicDatabase/COSMIC/Cosmic_MutantCensus_Tsv_v99_GRCh37/Cosmic_MutantCensus_v99_GRCh37.tsv', sep='\\t')\n",
    "df_cosmic = df_cosmic.drop_duplicates()\n",
    "df_cosmic['GENOMIC_WT_ALLELE'] = df_cosmic['GENOMIC_WT_ALLELE'].fillna('')\n",
    "df_cosmic['GENOMIC_MUT_ALLELE'] = df_cosmic['GENOMIC_MUT_ALLELE'].fillna('')\n",
    "df_cosmic = df_cosmic[['GENE_SYMBOL','TRANSCRIPT_ACCESSION','GENOME_START','GENOME_STOP','GENOMIC_WT_ALLELE','GENOMIC_MUT_ALLELE','MUTATION_ID']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cosmic.columns = ['gene_name_h','tx_id_h','start_h','end_h','ref_seq_h','alt_seq_h','cosmic_ID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>cosmic_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>ENST00000250003.3</td>\n",
       "      <td>17741555.0</td>\n",
       "      <td>17741556.0</td>\n",
       "      <td></td>\n",
       "      <td>CCTCA</td>\n",
       "      <td>20866010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>ENST00000250003.3</td>\n",
       "      <td>17742468.0</td>\n",
       "      <td>17742468.0</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td>20866012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>ENST00000250003.3</td>\n",
       "      <td>17741690.0</td>\n",
       "      <td>17741690.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>20866014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>ENST00000250003.3</td>\n",
       "      <td>17741428.0</td>\n",
       "      <td>17741428.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>20865978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>ENST00000250003.3</td>\n",
       "      <td>17742960.0</td>\n",
       "      <td>17742960.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>20866016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1897039</th>\n",
       "      <td>NFKBIE</td>\n",
       "      <td>ENST00000275015.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>184763869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1897040</th>\n",
       "      <td>ASXL1</td>\n",
       "      <td>ENST00000375687.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>184754753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1897041</th>\n",
       "      <td>ASXL1</td>\n",
       "      <td>ENST00000375687.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>184754753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1897042</th>\n",
       "      <td>ASXL1</td>\n",
       "      <td>ENST00000375687.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>184754753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1897043</th>\n",
       "      <td>ASXL1</td>\n",
       "      <td>ENST00000375687.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>184763905</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1897044 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name_h            tx_id_h     start_h       end_h ref_seq_h  \\\n",
       "0             MYOD1  ENST00000250003.3  17741555.0  17741556.0             \n",
       "1             MYOD1  ENST00000250003.3  17742468.0  17742468.0         C   \n",
       "2             MYOD1  ENST00000250003.3  17741690.0  17741690.0         C   \n",
       "3             MYOD1  ENST00000250003.3  17741428.0  17741428.0         G   \n",
       "4             MYOD1  ENST00000250003.3  17742960.0  17742960.0         G   \n",
       "...             ...                ...         ...         ...       ...   \n",
       "1897039      NFKBIE  ENST00000275015.5         NaN         NaN             \n",
       "1897040       ASXL1  ENST00000375687.4         NaN         NaN             \n",
       "1897041       ASXL1  ENST00000375687.4         NaN         NaN             \n",
       "1897042       ASXL1  ENST00000375687.4         NaN         NaN             \n",
       "1897043       ASXL1  ENST00000375687.4         NaN         NaN             \n",
       "\n",
       "        alt_seq_h  cosmic_ID  \n",
       "0           CCTCA   20866010  \n",
       "1                   20866012  \n",
       "2               T   20866014  \n",
       "3               A   20865978  \n",
       "4               A   20866016  \n",
       "...           ...        ...  \n",
       "1897039            184763869  \n",
       "1897040            184754753  \n",
       "1897041            184754753  \n",
       "1897042            184754753  \n",
       "1897043            184763905  \n",
       "\n",
       "[1897044 rows x 7 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cosmic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cosmic['type_h'] = [h2m.get_type(x,y) for x,y in zip(df_cosmic['ref_seq_h'], df_cosmic['alt_seq_h'])]\n",
    "df_cosmic = df_cosmic[['gene_name_h','tx_id_h','start_h','end_h','ref_seq_h','alt_seq_h','type_h','cosmic_ID']]\n",
    "df_cosmic = df_cosmic.dropna(subset=['start_h','end_h'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>cosmic_ID</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>ENST00000250003.3</td>\n",
       "      <td>17741555.0</td>\n",
       "      <td>17741556.0</td>\n",
       "      <td></td>\n",
       "      <td>CCTCA</td>\n",
       "      <td>INS</td>\n",
       "      <td>20866010</td>\n",
       "      <td>COSMIC-H0000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>ENST00000250003.3</td>\n",
       "      <td>17742468.0</td>\n",
       "      <td>17742468.0</td>\n",
       "      <td>C</td>\n",
       "      <td></td>\n",
       "      <td>DEL</td>\n",
       "      <td>20866012</td>\n",
       "      <td>COSMIC-H0000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>ENST00000250003.3</td>\n",
       "      <td>17741690.0</td>\n",
       "      <td>17741690.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>20866014</td>\n",
       "      <td>COSMIC-H0000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>ENST00000250003.3</td>\n",
       "      <td>17741428.0</td>\n",
       "      <td>17741428.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>20865978</td>\n",
       "      <td>COSMIC-H0000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MYOD1</td>\n",
       "      <td>ENST00000250003.3</td>\n",
       "      <td>17742960.0</td>\n",
       "      <td>17742960.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>20866016</td>\n",
       "      <td>COSMIC-H0000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1895470</th>\n",
       "      <td>LYN</td>\n",
       "      <td>ENST00000519728.1</td>\n",
       "      <td>56863058.0</td>\n",
       "      <td>56863058.0</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>76559148</td>\n",
       "      <td>COSMIC-H1771597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1895471</th>\n",
       "      <td>LYN</td>\n",
       "      <td>ENST00000519728.1</td>\n",
       "      <td>56916981.0</td>\n",
       "      <td>56916981.0</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>76555762</td>\n",
       "      <td>COSMIC-H1771598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1895472</th>\n",
       "      <td>LYN</td>\n",
       "      <td>ENST00000519728.1</td>\n",
       "      <td>56922493.0</td>\n",
       "      <td>56922493.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>76559152</td>\n",
       "      <td>COSMIC-H1771599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1895473</th>\n",
       "      <td>LYN</td>\n",
       "      <td>ENST00000519728.1</td>\n",
       "      <td>56912024.0</td>\n",
       "      <td>56912024.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>76559154</td>\n",
       "      <td>COSMIC-H1771600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1895474</th>\n",
       "      <td>LYN</td>\n",
       "      <td>ENST00000519728.1</td>\n",
       "      <td>56844593.0</td>\n",
       "      <td>56844593.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>76559156</td>\n",
       "      <td>COSMIC-H1771601</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1771601 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name_h            tx_id_h     start_h       end_h ref_seq_h  \\\n",
       "0             MYOD1  ENST00000250003.3  17741555.0  17741556.0             \n",
       "1             MYOD1  ENST00000250003.3  17742468.0  17742468.0         C   \n",
       "2             MYOD1  ENST00000250003.3  17741690.0  17741690.0         C   \n",
       "3             MYOD1  ENST00000250003.3  17741428.0  17741428.0         G   \n",
       "4             MYOD1  ENST00000250003.3  17742960.0  17742960.0         G   \n",
       "...             ...                ...         ...         ...       ...   \n",
       "1895470         LYN  ENST00000519728.1  56863058.0  56863058.0         A   \n",
       "1895471         LYN  ENST00000519728.1  56916981.0  56916981.0         T   \n",
       "1895472         LYN  ENST00000519728.1  56922493.0  56922493.0         G   \n",
       "1895473         LYN  ENST00000519728.1  56912024.0  56912024.0         G   \n",
       "1895474         LYN  ENST00000519728.1  56844593.0  56844593.0         A   \n",
       "\n",
       "        alt_seq_h type_h  cosmic_ID               ID  \n",
       "0           CCTCA    INS   20866010  COSMIC-H0000001  \n",
       "1                    DEL   20866012  COSMIC-H0000002  \n",
       "2               T    SNP   20866014  COSMIC-H0000003  \n",
       "3               A    SNP   20865978  COSMIC-H0000004  \n",
       "4               A    SNP   20866016  COSMIC-H0000005  \n",
       "...           ...    ...        ...              ...  \n",
       "1895470         C    SNP   76559148  COSMIC-H1771597  \n",
       "1895471         C    SNP   76555762  COSMIC-H1771598  \n",
       "1895472         A    SNP   76559152  COSMIC-H1771599  \n",
       "1895473         A    SNP   76559154  COSMIC-H1771600  \n",
       "1895474         G    SNP   76559156  COSMIC-H1771601  \n",
       "\n",
       "[1771601 rows x 9 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cosmic['ID'] = [f\"COSMIC-H{(x+1):07}\" for x in list(range(len(df_cosmic)))]\n",
    "df_cosmic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_cosmic.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/cosmic_input_original.csv',index=False)\n",
    "df_cosmic = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/cosmic_input_original.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1771601"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_cosmic['ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cosmic['gene_name_h'] = [dict_of_aacr_symbol.get(x,x) for x in df_cosmic['gene_name_h']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There were rows that could not be processed.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>cosmic_ID</th>\n",
       "      <th>ID</th>\n",
       "      <th>gene_name_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995.3</td>\n",
       "      <td>52570740.0</td>\n",
       "      <td>52570740.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>38273025</td>\n",
       "      <td>COSMIC-H0056496</td>\n",
       "      <td>A1cf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995.3</td>\n",
       "      <td>52566444.0</td>\n",
       "      <td>52566444.0</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DEL</td>\n",
       "      <td>38273035</td>\n",
       "      <td>COSMIC-H0056497</td>\n",
       "      <td>A1cf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995.3</td>\n",
       "      <td>52595946.0</td>\n",
       "      <td>52595946.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>38270819</td>\n",
       "      <td>COSMIC-H0056498</td>\n",
       "      <td>A1cf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995.3</td>\n",
       "      <td>52601695.0</td>\n",
       "      <td>52601695.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>38273050</td>\n",
       "      <td>COSMIC-H0056499</td>\n",
       "      <td>A1cf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995.3</td>\n",
       "      <td>52572755.0</td>\n",
       "      <td>52572755.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>38273056</td>\n",
       "      <td>COSMIC-H0056500</td>\n",
       "      <td>A1cf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941206</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15833958.0</td>\n",
       "      <td>15833958.0</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>25670786</td>\n",
       "      <td>COSMIC-H0925849</td>\n",
       "      <td>Zrsr2-ps1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941207</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15838385.0</td>\n",
       "      <td>15838385.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>25670792</td>\n",
       "      <td>COSMIC-H0925850</td>\n",
       "      <td>Zrsr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941208</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15838385.0</td>\n",
       "      <td>15838385.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>25670792</td>\n",
       "      <td>COSMIC-H0925850</td>\n",
       "      <td>Zrsr2-ps1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941209</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15841057.0</td>\n",
       "      <td>15841058.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>INS</td>\n",
       "      <td>25670796</td>\n",
       "      <td>COSMIC-H0925851</td>\n",
       "      <td>Zrsr2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941210</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15841057.0</td>\n",
       "      <td>15841058.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>INS</td>\n",
       "      <td>25670796</td>\n",
       "      <td>COSMIC-H0925851</td>\n",
       "      <td>Zrsr2-ps1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1941211 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name_h            tx_id_h     start_h       end_h ref_seq_h  \\\n",
       "0              A1CF  ENST00000373995.3  52570740.0  52570740.0         C   \n",
       "1              A1CF  ENST00000373995.3  52566444.0  52566444.0         T   \n",
       "2              A1CF  ENST00000373995.3  52595946.0  52595946.0         G   \n",
       "3              A1CF  ENST00000373995.3  52601695.0  52601695.0         A   \n",
       "4              A1CF  ENST00000373995.3  52572755.0  52572755.0         C   \n",
       "...             ...                ...         ...         ...       ...   \n",
       "1941206       ZRSR2  ENST00000307771.7  15833958.0  15833958.0         T   \n",
       "1941207       ZRSR2  ENST00000307771.7  15838385.0  15838385.0         C   \n",
       "1941208       ZRSR2  ENST00000307771.7  15838385.0  15838385.0         C   \n",
       "1941209       ZRSR2  ENST00000307771.7  15841057.0  15841058.0       NaN   \n",
       "1941210       ZRSR2  ENST00000307771.7  15841057.0  15841058.0       NaN   \n",
       "\n",
       "        alt_seq_h type_h  cosmic_ID               ID gene_name_m  \n",
       "0               T    SNP   38273025  COSMIC-H0056496        A1cf  \n",
       "1             NaN    DEL   38273035  COSMIC-H0056497        A1cf  \n",
       "2               A    SNP   38270819  COSMIC-H0056498        A1cf  \n",
       "3               T    SNP   38273050  COSMIC-H0056499        A1cf  \n",
       "4               A    SNP   38273056  COSMIC-H0056500        A1cf  \n",
       "...           ...    ...        ...              ...         ...  \n",
       "1941206         A    SNP   25670786  COSMIC-H0925849   Zrsr2-ps1  \n",
       "1941207         T    SNP   25670792  COSMIC-H0925850       Zrsr2  \n",
       "1941208         T    SNP   25670792  COSMIC-H0925850   Zrsr2-ps1  \n",
       "1941209        AA    INS   25670796  COSMIC-H0925851       Zrsr2  \n",
       "1941210        AA    INS   25670796  COSMIC-H0925851   Zrsr2-ps1  \n",
       "\n",
       "[1941211 rows x 10 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cosmic, df_wrong = h2m.query_batch(df_cosmic)\n",
    "df_cosmic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'COSMIC # genes mapped with mouse orthologs: 724'"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_1 = len(df_cosmic['gene_name_h'].unique())\n",
    "f'COSMIC # genes mapped with mouse orthologs: {l_1}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'COSMIC # genes not mapped with mouse orthologs: 11'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_1 = len(df_wrong['gene_name_h'].unique())\n",
    "f'COSMIC # genes not mapped with mouse orthologs: {l_1}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_mapped = set_mapped | set(df_cosmic['gene_name_h'].unique())\n",
    "set_unmapped = set_unmapped | set(df_wrong['gene_name_h'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_wrong.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/cosmic_without_homo_gene.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ZNF479', 'CARS', 'C15orf65', 'CLTCL1', 'FGFR1OP', 'ZNF331',\n",
       "       'HIST1H4I', 'MDS2', 'SEPT6', 'P2RY8', 'FCRL4'], dtype=object)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_wrong['gene_name_h'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No error occurs.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>cosmic_ID</th>\n",
       "      <th>ID</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995.3</td>\n",
       "      <td>52570740.0</td>\n",
       "      <td>52570740.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>38273025</td>\n",
       "      <td>COSMIC-H0056496</td>\n",
       "      <td>A1cf</td>\n",
       "      <td>ENSMUST00000075838.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995.3</td>\n",
       "      <td>52566444.0</td>\n",
       "      <td>52566444.0</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DEL</td>\n",
       "      <td>38273035</td>\n",
       "      <td>COSMIC-H0056497</td>\n",
       "      <td>A1cf</td>\n",
       "      <td>ENSMUST00000075838.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995.3</td>\n",
       "      <td>52595946.0</td>\n",
       "      <td>52595946.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>38270819</td>\n",
       "      <td>COSMIC-H0056498</td>\n",
       "      <td>A1cf</td>\n",
       "      <td>ENSMUST00000075838.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995.3</td>\n",
       "      <td>52601695.0</td>\n",
       "      <td>52601695.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>38273050</td>\n",
       "      <td>COSMIC-H0056499</td>\n",
       "      <td>A1cf</td>\n",
       "      <td>ENSMUST00000075838.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995.3</td>\n",
       "      <td>52572755.0</td>\n",
       "      <td>52572755.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>38273056</td>\n",
       "      <td>COSMIC-H0056500</td>\n",
       "      <td>A1cf</td>\n",
       "      <td>ENSMUST00000075838.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941206</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15833958.0</td>\n",
       "      <td>15833958.0</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>25670786</td>\n",
       "      <td>COSMIC-H0925849</td>\n",
       "      <td>Zrsr2-ps1</td>\n",
       "      <td>ENSMUST00000229075.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941207</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15838385.0</td>\n",
       "      <td>15838385.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>25670792</td>\n",
       "      <td>COSMIC-H0925850</td>\n",
       "      <td>Zrsr2</td>\n",
       "      <td>ENSMUST00000112289.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941208</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15838385.0</td>\n",
       "      <td>15838385.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>25670792</td>\n",
       "      <td>COSMIC-H0925850</td>\n",
       "      <td>Zrsr2-ps1</td>\n",
       "      <td>ENSMUST00000229075.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941209</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15841057.0</td>\n",
       "      <td>15841058.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>INS</td>\n",
       "      <td>25670796</td>\n",
       "      <td>COSMIC-H0925851</td>\n",
       "      <td>Zrsr2</td>\n",
       "      <td>ENSMUST00000112289.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941210</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15841057.0</td>\n",
       "      <td>15841058.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>INS</td>\n",
       "      <td>25670796</td>\n",
       "      <td>COSMIC-H0925851</td>\n",
       "      <td>Zrsr2-ps1</td>\n",
       "      <td>ENSMUST00000229075.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1941211 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name_h            tx_id_h     start_h       end_h ref_seq_h  \\\n",
       "0              A1CF  ENST00000373995.3  52570740.0  52570740.0         C   \n",
       "1              A1CF  ENST00000373995.3  52566444.0  52566444.0         T   \n",
       "2              A1CF  ENST00000373995.3  52595946.0  52595946.0         G   \n",
       "3              A1CF  ENST00000373995.3  52601695.0  52601695.0         A   \n",
       "4              A1CF  ENST00000373995.3  52572755.0  52572755.0         C   \n",
       "...             ...                ...         ...         ...       ...   \n",
       "1941206       ZRSR2  ENST00000307771.7  15833958.0  15833958.0         T   \n",
       "1941207       ZRSR2  ENST00000307771.7  15838385.0  15838385.0         C   \n",
       "1941208       ZRSR2  ENST00000307771.7  15838385.0  15838385.0         C   \n",
       "1941209       ZRSR2  ENST00000307771.7  15841057.0  15841058.0       NaN   \n",
       "1941210       ZRSR2  ENST00000307771.7  15841057.0  15841058.0       NaN   \n",
       "\n",
       "        alt_seq_h type_h  cosmic_ID               ID gene_name_m  \\\n",
       "0               T    SNP   38273025  COSMIC-H0056496        A1cf   \n",
       "1             NaN    DEL   38273035  COSMIC-H0056497        A1cf   \n",
       "2               A    SNP   38270819  COSMIC-H0056498        A1cf   \n",
       "3               T    SNP   38273050  COSMIC-H0056499        A1cf   \n",
       "4               A    SNP   38273056  COSMIC-H0056500        A1cf   \n",
       "...           ...    ...        ...              ...         ...   \n",
       "1941206         A    SNP   25670786  COSMIC-H0925849   Zrsr2-ps1   \n",
       "1941207         T    SNP   25670792  COSMIC-H0925850       Zrsr2   \n",
       "1941208         T    SNP   25670792  COSMIC-H0925850   Zrsr2-ps1   \n",
       "1941209        AA    INS   25670796  COSMIC-H0925851       Zrsr2   \n",
       "1941210        AA    INS   25670796  COSMIC-H0925851   Zrsr2-ps1   \n",
       "\n",
       "                      tx_id_m  \n",
       "0        ENSMUST00000075838.8  \n",
       "1        ENSMUST00000075838.8  \n",
       "2        ENSMUST00000075838.8  \n",
       "3        ENSMUST00000075838.8  \n",
       "4        ENSMUST00000075838.8  \n",
       "...                       ...  \n",
       "1941206  ENSMUST00000229075.2  \n",
       "1941207  ENSMUST00000112289.9  \n",
       "1941208  ENSMUST00000229075.2  \n",
       "1941209  ENSMUST00000112289.9  \n",
       "1941210  ENSMUST00000229075.2  \n",
       "\n",
       "[1941211 rows x 11 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cosmic,df_wrong = h2m.get_tx_batch(df_cosmic, species='m')\n",
    "df_cosmic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "numbers = label_numbers(df_cosmic['ID'])\n",
    "df_cosmic['ID'] = [get_label(h,x) for h,x in zip(df_cosmic['ID'],numbers)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>cosmic_ID</th>\n",
       "      <th>ID</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995.3</td>\n",
       "      <td>52570740.0</td>\n",
       "      <td>52570740.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>38273025</td>\n",
       "      <td>COSMIC-H0056496-M01</td>\n",
       "      <td>A1cf</td>\n",
       "      <td>ENSMUST00000075838.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995.3</td>\n",
       "      <td>52566444.0</td>\n",
       "      <td>52566444.0</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DEL</td>\n",
       "      <td>38273035</td>\n",
       "      <td>COSMIC-H0056497-M01</td>\n",
       "      <td>A1cf</td>\n",
       "      <td>ENSMUST00000075838.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995.3</td>\n",
       "      <td>52595946.0</td>\n",
       "      <td>52595946.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>38270819</td>\n",
       "      <td>COSMIC-H0056498-M01</td>\n",
       "      <td>A1cf</td>\n",
       "      <td>ENSMUST00000075838.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995.3</td>\n",
       "      <td>52601695.0</td>\n",
       "      <td>52601695.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>38273050</td>\n",
       "      <td>COSMIC-H0056499-M01</td>\n",
       "      <td>A1cf</td>\n",
       "      <td>ENSMUST00000075838.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>ENST00000373995.3</td>\n",
       "      <td>52572755.0</td>\n",
       "      <td>52572755.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>38273056</td>\n",
       "      <td>COSMIC-H0056500-M01</td>\n",
       "      <td>A1cf</td>\n",
       "      <td>ENSMUST00000075838.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941206</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15833958.0</td>\n",
       "      <td>15833958.0</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>25670786</td>\n",
       "      <td>COSMIC-H0925849-M02</td>\n",
       "      <td>Zrsr2-ps1</td>\n",
       "      <td>ENSMUST00000229075.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941207</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15838385.0</td>\n",
       "      <td>15838385.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>25670792</td>\n",
       "      <td>COSMIC-H0925850-M01</td>\n",
       "      <td>Zrsr2</td>\n",
       "      <td>ENSMUST00000112289.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941208</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15838385.0</td>\n",
       "      <td>15838385.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>25670792</td>\n",
       "      <td>COSMIC-H0925850-M02</td>\n",
       "      <td>Zrsr2-ps1</td>\n",
       "      <td>ENSMUST00000229075.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941209</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15841057.0</td>\n",
       "      <td>15841058.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>INS</td>\n",
       "      <td>25670796</td>\n",
       "      <td>COSMIC-H0925851-M01</td>\n",
       "      <td>Zrsr2</td>\n",
       "      <td>ENSMUST00000112289.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1941210</th>\n",
       "      <td>ZRSR2</td>\n",
       "      <td>ENST00000307771.7</td>\n",
       "      <td>15841057.0</td>\n",
       "      <td>15841058.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AA</td>\n",
       "      <td>INS</td>\n",
       "      <td>25670796</td>\n",
       "      <td>COSMIC-H0925851-M02</td>\n",
       "      <td>Zrsr2-ps1</td>\n",
       "      <td>ENSMUST00000229075.2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1941211 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name_h            tx_id_h     start_h       end_h ref_seq_h  \\\n",
       "0              A1CF  ENST00000373995.3  52570740.0  52570740.0         C   \n",
       "1              A1CF  ENST00000373995.3  52566444.0  52566444.0         T   \n",
       "2              A1CF  ENST00000373995.3  52595946.0  52595946.0         G   \n",
       "3              A1CF  ENST00000373995.3  52601695.0  52601695.0         A   \n",
       "4              A1CF  ENST00000373995.3  52572755.0  52572755.0         C   \n",
       "...             ...                ...         ...         ...       ...   \n",
       "1941206       ZRSR2  ENST00000307771.7  15833958.0  15833958.0         T   \n",
       "1941207       ZRSR2  ENST00000307771.7  15838385.0  15838385.0         C   \n",
       "1941208       ZRSR2  ENST00000307771.7  15838385.0  15838385.0         C   \n",
       "1941209       ZRSR2  ENST00000307771.7  15841057.0  15841058.0       NaN   \n",
       "1941210       ZRSR2  ENST00000307771.7  15841057.0  15841058.0       NaN   \n",
       "\n",
       "        alt_seq_h type_h  cosmic_ID                   ID gene_name_m  \\\n",
       "0               T    SNP   38273025  COSMIC-H0056496-M01        A1cf   \n",
       "1             NaN    DEL   38273035  COSMIC-H0056497-M01        A1cf   \n",
       "2               A    SNP   38270819  COSMIC-H0056498-M01        A1cf   \n",
       "3               T    SNP   38273050  COSMIC-H0056499-M01        A1cf   \n",
       "4               A    SNP   38273056  COSMIC-H0056500-M01        A1cf   \n",
       "...           ...    ...        ...                  ...         ...   \n",
       "1941206         A    SNP   25670786  COSMIC-H0925849-M02   Zrsr2-ps1   \n",
       "1941207         T    SNP   25670792  COSMIC-H0925850-M01       Zrsr2   \n",
       "1941208         T    SNP   25670792  COSMIC-H0925850-M02   Zrsr2-ps1   \n",
       "1941209        AA    INS   25670796  COSMIC-H0925851-M01       Zrsr2   \n",
       "1941210        AA    INS   25670796  COSMIC-H0925851-M02   Zrsr2-ps1   \n",
       "\n",
       "                      tx_id_m  \n",
       "0        ENSMUST00000075838.8  \n",
       "1        ENSMUST00000075838.8  \n",
       "2        ENSMUST00000075838.8  \n",
       "3        ENSMUST00000075838.8  \n",
       "4        ENSMUST00000075838.8  \n",
       "...                       ...  \n",
       "1941206  ENSMUST00000229075.2  \n",
       "1941207  ENSMUST00000112289.9  \n",
       "1941208  ENSMUST00000229075.2  \n",
       "1941209  ENSMUST00000112289.9  \n",
       "1941210  ENSMUST00000229075.2  \n",
       "\n",
       "[1941211 rows x 11 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_cosmic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cosmic = df_cosmic.rename(columns={'ID':'index'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_cosmic.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/cosmic_input.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_wrong = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/v1_input/cosmic_without_homo_gene.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No error occurs.\n"
     ]
    }
   ],
   "source": [
    "set_of_addon = set(df_wrong['ID']).intersection(set(df_cosmic['ID']))\n",
    "df_addon_aacr = df_cosmic[df_cosmic['ID'].isin(set_of_addon)].reset_index(drop=True)\n",
    "df,df_wrong = h2m.get_tx_batch(df_addon_aacr, species='m')\n",
    "df\n",
    "numbers = label_numbers(df['ID'])\n",
    "df['ID'] = [get_label(h,x) for h,x in zip(df['ID'],numbers)]\n",
    "df = df.rename(columns={'ID':'index'})\n",
    "len(df['index'].unique())\n",
    "df.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/cosmic_input_addon.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (3) ClinVar  \n",
    "Downloaded from ClinVar FTP Site (https://ftp.ncbi.nlm.nih.gov/pub/clinvar/)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read from vcf file  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::idx_find_and_load] Could not retrieve index file for '/Users/kexindong/Documents/GitHub/Database/PublicDatabase/ClinVar/GRCh37_clinvar_20240206.vcf.gz'\n",
      "[W::vcf_parse] Contig '1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '2' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '3' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '4' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '5' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '6' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '7' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '8' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '9' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '10' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '11' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '12' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '13' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '14' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '15' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '16' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '17' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '18' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '19' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '20' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '21' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '22' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'X' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'Y' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'MT' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_113889.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NT_167222.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NW_003315925.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NW_003315947.1' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'NW_003315950.2' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    }
   ],
   "source": [
    "path = '/Users/kexindong/Documents/GitHub/Database/PublicDatabase/ClinVar/GRCh37_clinvar_20240206.vcf.gz'\n",
    "df_clinvar = h2m.clinvar_reader(path,keep=True)\n",
    "df_clinvar = h2m.clinvar_to_maf(df_clinvar)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>Variation_ID</th>\n",
       "      <th>Allele_ID</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>CLNHGVS</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>clinvar_id</th>\n",
       "      <th>format</th>\n",
       "      <th>ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>1</td>\n",
       "      <td>69134</td>\n",
       "      <td>69134</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2205837</td>\n",
       "      <td>2193183</td>\n",
       "      <td>(Likely_benign,)</td>\n",
       "      <td>(NC_000001.10:g.69134A&gt;G,)</td>\n",
       "      <td>(Inborn_genetic_diseases,)</td>\n",
       "      <td>2205837</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>1</td>\n",
       "      <td>69581</td>\n",
       "      <td>69581</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2252161</td>\n",
       "      <td>2238986</td>\n",
       "      <td>(Uncertain_significance,)</td>\n",
       "      <td>(NC_000001.10:g.69581C&gt;G,)</td>\n",
       "      <td>(Inborn_genetic_diseases,)</td>\n",
       "      <td>2252161</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>1</td>\n",
       "      <td>69682</td>\n",
       "      <td>69682</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2396347</td>\n",
       "      <td>2386655</td>\n",
       "      <td>(Uncertain_significance,)</td>\n",
       "      <td>(NC_000001.10:g.69682G&gt;A,)</td>\n",
       "      <td>(Inborn_genetic_diseases,)</td>\n",
       "      <td>2396347</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0000003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>1</td>\n",
       "      <td>69769</td>\n",
       "      <td>69769</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2288999</td>\n",
       "      <td>2278803</td>\n",
       "      <td>(Uncertain_significance,)</td>\n",
       "      <td>(NC_000001.10:g.69769T&gt;C,)</td>\n",
       "      <td>(Inborn_genetic_diseases,)</td>\n",
       "      <td>2288999</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0000004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>1</td>\n",
       "      <td>69995</td>\n",
       "      <td>69995</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2351346</td>\n",
       "      <td>2333177</td>\n",
       "      <td>(Uncertain_significance,)</td>\n",
       "      <td>(NC_000001.10:g.69995G&gt;C,)</td>\n",
       "      <td>(Inborn_genetic_diseases,)</td>\n",
       "      <td>2351346</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0000005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2353133</th>\n",
       "      <td>2353133</td>\n",
       "      <td>ABO</td>\n",
       "      <td>NW_003315925.1</td>\n",
       "      <td>83615</td>\n",
       "      <td>83615</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>17735</td>\n",
       "      <td>32774</td>\n",
       "      <td>(Benign,)</td>\n",
       "      <td>None</td>\n",
       "      <td>(not_provided|ABO_blood_group_system,)</td>\n",
       "      <td>17735</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2353134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2353134</th>\n",
       "      <td>2353134</td>\n",
       "      <td>PECAM1</td>\n",
       "      <td>NW_003315947.1</td>\n",
       "      <td>181683</td>\n",
       "      <td>181683</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>156304</td>\n",
       "      <td>166084</td>\n",
       "      <td>(Benign,)</td>\n",
       "      <td>(NW_003315947.1:g.181683C&gt;G,)</td>\n",
       "      <td>(Three_Vessel_Coronary_Disease|not_provided|PL...</td>\n",
       "      <td>156304</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2353135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2353135</th>\n",
       "      <td>2353135</td>\n",
       "      <td>KCNJ18</td>\n",
       "      <td>NW_003315950.2</td>\n",
       "      <td>355765</td>\n",
       "      <td>355765</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>206</td>\n",
       "      <td>15245</td>\n",
       "      <td>(risk_factor,)</td>\n",
       "      <td>None</td>\n",
       "      <td>(Thyrotoxic_periodic_paralysis, _susceptibilit...</td>\n",
       "      <td>206</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2353136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2353136</th>\n",
       "      <td>2353136</td>\n",
       "      <td>KCNJ18</td>\n",
       "      <td>NW_003315950.2</td>\n",
       "      <td>356212</td>\n",
       "      <td>356212</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>205</td>\n",
       "      <td>15244</td>\n",
       "      <td>(risk_factor,)</td>\n",
       "      <td>None</td>\n",
       "      <td>(Thyrotoxic_periodic_paralysis, _susceptibilit...</td>\n",
       "      <td>205</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2353137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2353137</th>\n",
       "      <td>2353137</td>\n",
       "      <td>KCNJ18</td>\n",
       "      <td>NW_003315950.2</td>\n",
       "      <td>356248</td>\n",
       "      <td>356248</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>207</td>\n",
       "      <td>15246</td>\n",
       "      <td>(risk_factor,)</td>\n",
       "      <td>None</td>\n",
       "      <td>(Thyrotoxic_periodic_paralysis, _susceptibilit...</td>\n",
       "      <td>207</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2353138</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2353138 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           index gene_name_h      Chromosome  start_h   end_h ref_seq_h  \\\n",
       "0              0       OR4F5               1    69134   69134         A   \n",
       "1              1       OR4F5               1    69581   69581         C   \n",
       "2              2       OR4F5               1    69682   69682         G   \n",
       "3              3       OR4F5               1    69769   69769         T   \n",
       "4              4       OR4F5               1    69995   69995         G   \n",
       "...          ...         ...             ...      ...     ...       ...   \n",
       "2353133  2353133         ABO  NW_003315925.1    83615   83615         C   \n",
       "2353134  2353134      PECAM1  NW_003315947.1   181683  181683         C   \n",
       "2353135  2353135      KCNJ18  NW_003315950.2   355765  355765         G   \n",
       "2353136  2353136      KCNJ18  NW_003315950.2   356212  356212         C   \n",
       "2353137  2353137      KCNJ18  NW_003315950.2   356248  356248         A   \n",
       "\n",
       "        alt_seq_h type_h  Variation_ID  Allele_ID                     CLNSIG  \\\n",
       "0               G    SNP       2205837    2193183           (Likely_benign,)   \n",
       "1               G    SNP       2252161    2238986  (Uncertain_significance,)   \n",
       "2               A    SNP       2396347    2386655  (Uncertain_significance,)   \n",
       "3               C    SNP       2288999    2278803  (Uncertain_significance,)   \n",
       "4               C    SNP       2351346    2333177  (Uncertain_significance,)   \n",
       "...           ...    ...           ...        ...                        ...   \n",
       "2353133         -    DEL         17735      32774                  (Benign,)   \n",
       "2353134         G    SNP        156304     166084                  (Benign,)   \n",
       "2353135         A    SNP           206      15245             (risk_factor,)   \n",
       "2353136         T    SNP           205      15244             (risk_factor,)   \n",
       "2353137         G    SNP           207      15246             (risk_factor,)   \n",
       "\n",
       "                               CLNHGVS  \\\n",
       "0           (NC_000001.10:g.69134A>G,)   \n",
       "1           (NC_000001.10:g.69581C>G,)   \n",
       "2           (NC_000001.10:g.69682G>A,)   \n",
       "3           (NC_000001.10:g.69769T>C,)   \n",
       "4           (NC_000001.10:g.69995G>C,)   \n",
       "...                                ...   \n",
       "2353133                           None   \n",
       "2353134  (NW_003315947.1:g.181683C>G,)   \n",
       "2353135                           None   \n",
       "2353136                           None   \n",
       "2353137                           None   \n",
       "\n",
       "                                                     CLNDN clinvar_id format  \\\n",
       "0                               (Inborn_genetic_diseases,)    2205837    MAF   \n",
       "1                               (Inborn_genetic_diseases,)    2252161    MAF   \n",
       "2                               (Inborn_genetic_diseases,)    2396347    MAF   \n",
       "3                               (Inborn_genetic_diseases,)    2288999    MAF   \n",
       "4                               (Inborn_genetic_diseases,)    2351346    MAF   \n",
       "...                                                    ...        ...    ...   \n",
       "2353133             (not_provided|ABO_blood_group_system,)      17735    MAF   \n",
       "2353134  (Three_Vessel_Coronary_Disease|not_provided|PL...     156304    MAF   \n",
       "2353135  (Thyrotoxic_periodic_paralysis, _susceptibilit...        206    MAF   \n",
       "2353136  (Thyrotoxic_periodic_paralysis, _susceptibilit...        205    MAF   \n",
       "2353137  (Thyrotoxic_periodic_paralysis, _susceptibilit...        207    MAF   \n",
       "\n",
       "                       ID  \n",
       "0        CLINVAR-H0000001  \n",
       "1        CLINVAR-H0000002  \n",
       "2        CLINVAR-H0000003  \n",
       "3        CLINVAR-H0000004  \n",
       "4        CLINVAR-H0000005  \n",
       "...                   ...  \n",
       "2353133  CLINVAR-H2353134  \n",
       "2353134  CLINVAR-H2353135  \n",
       "2353135  CLINVAR-H2353136  \n",
       "2353136  CLINVAR-H2353137  \n",
       "2353137  CLINVAR-H2353138  \n",
       "\n",
       "[2353138 rows x 16 columns]"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clinvar = df_clinvar.rename(columns={'ID':'clinvar_id'})\n",
    "df_clinvar['ID'] = [f\"CLINVAR-H{(x+1):07}\" for x in list(range(len(df_clinvar)))]\n",
    "df_clinvar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_clinvar.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/clinvar_input_original.csv',index=False)\n",
    "df_clinvar = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/clinvar_input_original.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2353138"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_clinvar['ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clinvar['gene_name_h'] = [dict_of_aacr_symbol.get(x,x) for x in df_clinvar['gene_name_h']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There were rows that could not be processed.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>Variation_ID</th>\n",
       "      <th>Allele_ID</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>CLNHGVS</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>clinvar_id</th>\n",
       "      <th>format</th>\n",
       "      <th>ID</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>ref_genome_h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>1</td>\n",
       "      <td>69134</td>\n",
       "      <td>69134</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2205837</td>\n",
       "      <td>2193183</td>\n",
       "      <td>('Likely_benign',)</td>\n",
       "      <td>('NC_000001.10:g.69134A&gt;G',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2205837</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0000001</td>\n",
       "      <td>ENST00000335137.3</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>1</td>\n",
       "      <td>69581</td>\n",
       "      <td>69581</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2252161</td>\n",
       "      <td>2238986</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000001.10:g.69581C&gt;G',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2252161</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0000002</td>\n",
       "      <td>ENST00000335137.3</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>1</td>\n",
       "      <td>69682</td>\n",
       "      <td>69682</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2396347</td>\n",
       "      <td>2386655</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000001.10:g.69682G&gt;A',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2396347</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0000003</td>\n",
       "      <td>ENST00000335137.3</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>1</td>\n",
       "      <td>69769</td>\n",
       "      <td>69769</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2288999</td>\n",
       "      <td>2278803</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000001.10:g.69769T&gt;C',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2288999</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0000004</td>\n",
       "      <td>ENST00000335137.3</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>1</td>\n",
       "      <td>69995</td>\n",
       "      <td>69995</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2351346</td>\n",
       "      <td>2333177</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000001.10:g.69995G&gt;C',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2351346</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0000005</td>\n",
       "      <td>ENST00000335137.3</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2348062</th>\n",
       "      <td>2353133</td>\n",
       "      <td>ABO</td>\n",
       "      <td>NW_003315925.1</td>\n",
       "      <td>83615</td>\n",
       "      <td>83615</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>DEL</td>\n",
       "      <td>17735</td>\n",
       "      <td>32774</td>\n",
       "      <td>('Benign',)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>('not_provided|ABO_blood_group_system',)</td>\n",
       "      <td>17735</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2353134</td>\n",
       "      <td>ENST00000453660.2</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2348063</th>\n",
       "      <td>2353134</td>\n",
       "      <td>PECAM1</td>\n",
       "      <td>NW_003315947.1</td>\n",
       "      <td>181683</td>\n",
       "      <td>181683</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>156304</td>\n",
       "      <td>166084</td>\n",
       "      <td>('Benign',)</td>\n",
       "      <td>('NW_003315947.1:g.181683C&gt;G',)</td>\n",
       "      <td>('Three_Vessel_Coronary_Disease|not_provided|P...</td>\n",
       "      <td>156304</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2353135</td>\n",
       "      <td>None</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2348064</th>\n",
       "      <td>2353135</td>\n",
       "      <td>KCNJ18</td>\n",
       "      <td>NW_003315950.2</td>\n",
       "      <td>355765</td>\n",
       "      <td>355765</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>206</td>\n",
       "      <td>15245</td>\n",
       "      <td>('risk_factor',)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>('Thyrotoxic_periodic_paralysis', '_susceptibi...</td>\n",
       "      <td>206</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2353136</td>\n",
       "      <td>ENST00000583088.1</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2348065</th>\n",
       "      <td>2353136</td>\n",
       "      <td>KCNJ18</td>\n",
       "      <td>NW_003315950.2</td>\n",
       "      <td>356212</td>\n",
       "      <td>356212</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>205</td>\n",
       "      <td>15244</td>\n",
       "      <td>('risk_factor',)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>('Thyrotoxic_periodic_paralysis', '_susceptibi...</td>\n",
       "      <td>205</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2353137</td>\n",
       "      <td>ENST00000583088.1</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2348066</th>\n",
       "      <td>2353137</td>\n",
       "      <td>KCNJ18</td>\n",
       "      <td>NW_003315950.2</td>\n",
       "      <td>356248</td>\n",
       "      <td>356248</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>207</td>\n",
       "      <td>15246</td>\n",
       "      <td>('risk_factor',)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>('Thyrotoxic_periodic_paralysis', '_susceptibi...</td>\n",
       "      <td>207</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2353138</td>\n",
       "      <td>ENST00000583088.1</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2348067 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           index gene_name_h      Chromosome  start_h   end_h ref_seq_h  \\\n",
       "0              0       OR4F5               1    69134   69134         A   \n",
       "1              1       OR4F5               1    69581   69581         C   \n",
       "2              2       OR4F5               1    69682   69682         G   \n",
       "3              3       OR4F5               1    69769   69769         T   \n",
       "4              4       OR4F5               1    69995   69995         G   \n",
       "...          ...         ...             ...      ...     ...       ...   \n",
       "2348062  2353133         ABO  NW_003315925.1    83615   83615         C   \n",
       "2348063  2353134      PECAM1  NW_003315947.1   181683  181683         C   \n",
       "2348064  2353135      KCNJ18  NW_003315950.2   355765  355765         G   \n",
       "2348065  2353136      KCNJ18  NW_003315950.2   356212  356212         C   \n",
       "2348066  2353137      KCNJ18  NW_003315950.2   356248  356248         A   \n",
       "\n",
       "        alt_seq_h type_h  Variation_ID  Allele_ID  \\\n",
       "0               G    SNP       2205837    2193183   \n",
       "1               G    SNP       2252161    2238986   \n",
       "2               A    SNP       2396347    2386655   \n",
       "3               C    SNP       2288999    2278803   \n",
       "4               C    SNP       2351346    2333177   \n",
       "...           ...    ...           ...        ...   \n",
       "2348062         -    DEL         17735      32774   \n",
       "2348063         G    SNP        156304     166084   \n",
       "2348064         A    SNP           206      15245   \n",
       "2348065         T    SNP           205      15244   \n",
       "2348066         G    SNP           207      15246   \n",
       "\n",
       "                              CLNSIG                          CLNHGVS  \\\n",
       "0                 ('Likely_benign',)     ('NC_000001.10:g.69134A>G',)   \n",
       "1        ('Uncertain_significance',)     ('NC_000001.10:g.69581C>G',)   \n",
       "2        ('Uncertain_significance',)     ('NC_000001.10:g.69682G>A',)   \n",
       "3        ('Uncertain_significance',)     ('NC_000001.10:g.69769T>C',)   \n",
       "4        ('Uncertain_significance',)     ('NC_000001.10:g.69995G>C',)   \n",
       "...                              ...                              ...   \n",
       "2348062                  ('Benign',)                              NaN   \n",
       "2348063                  ('Benign',)  ('NW_003315947.1:g.181683C>G',)   \n",
       "2348064             ('risk_factor',)                              NaN   \n",
       "2348065             ('risk_factor',)                              NaN   \n",
       "2348066             ('risk_factor',)                              NaN   \n",
       "\n",
       "                                                     CLNDN  clinvar_id format  \\\n",
       "0                             ('Inborn_genetic_diseases',)     2205837    MAF   \n",
       "1                             ('Inborn_genetic_diseases',)     2252161    MAF   \n",
       "2                             ('Inborn_genetic_diseases',)     2396347    MAF   \n",
       "3                             ('Inborn_genetic_diseases',)     2288999    MAF   \n",
       "4                             ('Inborn_genetic_diseases',)     2351346    MAF   \n",
       "...                                                    ...         ...    ...   \n",
       "2348062           ('not_provided|ABO_blood_group_system',)       17735    MAF   \n",
       "2348063  ('Three_Vessel_Coronary_Disease|not_provided|P...      156304    MAF   \n",
       "2348064  ('Thyrotoxic_periodic_paralysis', '_susceptibi...         206    MAF   \n",
       "2348065  ('Thyrotoxic_periodic_paralysis', '_susceptibi...         205    MAF   \n",
       "2348066  ('Thyrotoxic_periodic_paralysis', '_susceptibi...         207    MAF   \n",
       "\n",
       "                       ID            tx_id_h ref_genome_h  \n",
       "0        CLINVAR-H0000001  ENST00000335137.3       GRCh37  \n",
       "1        CLINVAR-H0000002  ENST00000335137.3       GRCh37  \n",
       "2        CLINVAR-H0000003  ENST00000335137.3       GRCh37  \n",
       "3        CLINVAR-H0000004  ENST00000335137.3       GRCh37  \n",
       "4        CLINVAR-H0000005  ENST00000335137.3       GRCh37  \n",
       "...                   ...                ...          ...  \n",
       "2348062  CLINVAR-H2353134  ENST00000453660.2       GRCh37  \n",
       "2348063  CLINVAR-H2353135               None       GRCh37  \n",
       "2348064  CLINVAR-H2353136  ENST00000583088.1       GRCh37  \n",
       "2348065  CLINVAR-H2353137  ENST00000583088.1       GRCh37  \n",
       "2348066  CLINVAR-H2353138  ENST00000583088.1       GRCh37  \n",
       "\n",
       "[2348067 rows x 18 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clinvar, df_wrong = h2m.get_tx_batch(df_clinvar, 'h',ver=37)\n",
    "df_clinvar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_wrong.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/clinvar_without_tx_id.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There were rows that could not be processed.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>Variation_ID</th>\n",
       "      <th>Allele_ID</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>CLNHGVS</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>clinvar_id</th>\n",
       "      <th>format</th>\n",
       "      <th>ID</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>ref_genome_h</th>\n",
       "      <th>gene_name_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2135678</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858782</td>\n",
       "      <td>58858782</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2336128</td>\n",
       "      <td>2328468</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000019.9:g.58858782G&gt;C',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2336128</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2135679</td>\n",
       "      <td>ENST00000263100.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>A1bg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2135679</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858882</td>\n",
       "      <td>58858882</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2521649</td>\n",
       "      <td>2696794</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000019.9:g.58858882C&gt;G',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2521649</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2135680</td>\n",
       "      <td>ENST00000263100.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>A1bg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2135680</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858888</td>\n",
       "      <td>58858888</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2552144</td>\n",
       "      <td>2714907</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000019.9:g.58858888C&gt;G',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2552144</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2135681</td>\n",
       "      <td>ENST00000263100.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>A1bg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2135681</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858923</td>\n",
       "      <td>58858923</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2484196</td>\n",
       "      <td>2455282</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000019.9:g.58858923G&gt;A',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2484196</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2135682</td>\n",
       "      <td>ENST00000263100.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>A1bg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2135682</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58861805</td>\n",
       "      <td>58861805</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2623749</td>\n",
       "      <td>2787053</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000019.9:g.58861805C&gt;T',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2623749</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2135683</td>\n",
       "      <td>ENST00000263100.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>A1bg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2513281</th>\n",
       "      <td>83147</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1</td>\n",
       "      <td>78098759</td>\n",
       "      <td>78098759</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2331197</td>\n",
       "      <td>2327429</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000001.10:g.78098759T&gt;C',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2331197</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0083148</td>\n",
       "      <td>ENST00000370801.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Zzz3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2513282</th>\n",
       "      <td>83148</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1</td>\n",
       "      <td>78098762</td>\n",
       "      <td>78098762</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2618943</td>\n",
       "      <td>2782688</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000001.10:g.78098762T&gt;C',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2618943</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0083149</td>\n",
       "      <td>ENST00000370801.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Zzz3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2513283</th>\n",
       "      <td>83149</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1</td>\n",
       "      <td>78098790</td>\n",
       "      <td>78098790</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2456830</td>\n",
       "      <td>2431257</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000001.10:g.78098790T&gt;C',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2456830</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0083150</td>\n",
       "      <td>ENST00000370801.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Zzz3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2513284</th>\n",
       "      <td>83150</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1</td>\n",
       "      <td>78098992</td>\n",
       "      <td>78098992</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2408495</td>\n",
       "      <td>2391953</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000001.10:g.78098992T&gt;A',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2408495</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0083151</td>\n",
       "      <td>ENST00000370801.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Zzz3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2513285</th>\n",
       "      <td>83151</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1</td>\n",
       "      <td>78099017</td>\n",
       "      <td>78099017</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2271322</td>\n",
       "      <td>2257695</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000001.10:g.78099017C&gt;T',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2271322</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0083152</td>\n",
       "      <td>ENST00000370801.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Zzz3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2513286 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           index gene_name_h Chromosome   start_h     end_h ref_seq_h  \\\n",
       "0        2135678        A1BG         19  58858782  58858782         G   \n",
       "1        2135679        A1BG         19  58858882  58858882         C   \n",
       "2        2135680        A1BG         19  58858888  58858888         C   \n",
       "3        2135681        A1BG         19  58858923  58858923         G   \n",
       "4        2135682        A1BG         19  58861805  58861805         C   \n",
       "...          ...         ...        ...       ...       ...       ...   \n",
       "2513281    83147        ZZZ3          1  78098759  78098759         T   \n",
       "2513282    83148        ZZZ3          1  78098762  78098762         T   \n",
       "2513283    83149        ZZZ3          1  78098790  78098790         T   \n",
       "2513284    83150        ZZZ3          1  78098992  78098992         T   \n",
       "2513285    83151        ZZZ3          1  78099017  78099017         C   \n",
       "\n",
       "        alt_seq_h type_h  Variation_ID  Allele_ID  \\\n",
       "0               C    SNP       2336128    2328468   \n",
       "1               G    SNP       2521649    2696794   \n",
       "2               G    SNP       2552144    2714907   \n",
       "3               A    SNP       2484196    2455282   \n",
       "4               T    SNP       2623749    2787053   \n",
       "...           ...    ...           ...        ...   \n",
       "2513281         C    SNP       2331197    2327429   \n",
       "2513282         C    SNP       2618943    2782688   \n",
       "2513283         C    SNP       2456830    2431257   \n",
       "2513284         A    SNP       2408495    2391953   \n",
       "2513285         T    SNP       2271322    2257695   \n",
       "\n",
       "                              CLNSIG                          CLNHGVS  \\\n",
       "0        ('Uncertain_significance',)   ('NC_000019.9:g.58858782G>C',)   \n",
       "1        ('Uncertain_significance',)   ('NC_000019.9:g.58858882C>G',)   \n",
       "2        ('Uncertain_significance',)   ('NC_000019.9:g.58858888C>G',)   \n",
       "3        ('Uncertain_significance',)   ('NC_000019.9:g.58858923G>A',)   \n",
       "4        ('Uncertain_significance',)   ('NC_000019.9:g.58861805C>T',)   \n",
       "...                              ...                              ...   \n",
       "2513281  ('Uncertain_significance',)  ('NC_000001.10:g.78098759T>C',)   \n",
       "2513282  ('Uncertain_significance',)  ('NC_000001.10:g.78098762T>C',)   \n",
       "2513283  ('Uncertain_significance',)  ('NC_000001.10:g.78098790T>C',)   \n",
       "2513284  ('Uncertain_significance',)  ('NC_000001.10:g.78098992T>A',)   \n",
       "2513285  ('Uncertain_significance',)  ('NC_000001.10:g.78099017C>T',)   \n",
       "\n",
       "                                CLNDN  clinvar_id format                ID  \\\n",
       "0        ('Inborn_genetic_diseases',)     2336128    MAF  CLINVAR-H2135679   \n",
       "1        ('Inborn_genetic_diseases',)     2521649    MAF  CLINVAR-H2135680   \n",
       "2        ('Inborn_genetic_diseases',)     2552144    MAF  CLINVAR-H2135681   \n",
       "3        ('Inborn_genetic_diseases',)     2484196    MAF  CLINVAR-H2135682   \n",
       "4        ('Inborn_genetic_diseases',)     2623749    MAF  CLINVAR-H2135683   \n",
       "...                               ...         ...    ...               ...   \n",
       "2513281  ('Inborn_genetic_diseases',)     2331197    MAF  CLINVAR-H0083148   \n",
       "2513282  ('Inborn_genetic_diseases',)     2618943    MAF  CLINVAR-H0083149   \n",
       "2513283  ('Inborn_genetic_diseases',)     2456830    MAF  CLINVAR-H0083150   \n",
       "2513284  ('Inborn_genetic_diseases',)     2408495    MAF  CLINVAR-H0083151   \n",
       "2513285  ('Inborn_genetic_diseases',)     2271322    MAF  CLINVAR-H0083152   \n",
       "\n",
       "                   tx_id_h ref_genome_h gene_name_m  \n",
       "0        ENST00000263100.3       GRCh37        A1bg  \n",
       "1        ENST00000263100.3       GRCh37        A1bg  \n",
       "2        ENST00000263100.3       GRCh37        A1bg  \n",
       "3        ENST00000263100.3       GRCh37        A1bg  \n",
       "4        ENST00000263100.3       GRCh37        A1bg  \n",
       "...                    ...          ...         ...  \n",
       "2513281  ENST00000370801.3       GRCh37        Zzz3  \n",
       "2513282  ENST00000370801.3       GRCh37        Zzz3  \n",
       "2513283  ENST00000370801.3       GRCh37        Zzz3  \n",
       "2513284  ENST00000370801.3       GRCh37        Zzz3  \n",
       "2513285  ENST00000370801.3       GRCh37        Zzz3  \n",
       "\n",
       "[2513286 rows x 19 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clinvar, df_wrong = h2m.query_batch(df_clinvar)\n",
    "df_clinvar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CLINVAR # genes mapped with mouse orthologs: 17396'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_1 = len(df_clinvar['gene_name_h'].unique())\n",
    "f'CLINVAR # genes mapped with mouse orthologs: {l_1}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_mapped = set_mapped | set(df_cosmic['gene_name_h'].unique())\n",
    "set_unmapped = set_unmapped | set(df_wrong['gene_name_h'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>Variation_ID</th>\n",
       "      <th>Allele_ID</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>CLNHGVS</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>clinvar_id</th>\n",
       "      <th>format</th>\n",
       "      <th>ID</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>ref_genome_h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1072</td>\n",
       "      <td>HES4</td>\n",
       "      <td>1</td>\n",
       "      <td>934467</td>\n",
       "      <td>934467</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2261120</td>\n",
       "      <td>2255281</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000001.10:g.934467C&gt;G',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2261120</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0001073</td>\n",
       "      <td>ENST00000428771.2</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1073</td>\n",
       "      <td>HES4</td>\n",
       "      <td>1</td>\n",
       "      <td>934470</td>\n",
       "      <td>934470</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2360191</td>\n",
       "      <td>2352994</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000001.10:g.934470G&gt;C',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2360191</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0001074</td>\n",
       "      <td>ENST00000428771.2</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1074</td>\n",
       "      <td>HES4</td>\n",
       "      <td>1</td>\n",
       "      <td>934476</td>\n",
       "      <td>934476</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2363714</td>\n",
       "      <td>2353874</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000001.10:g.934476T&gt;A',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2363714</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0001075</td>\n",
       "      <td>ENST00000428771.2</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1075</td>\n",
       "      <td>HES4</td>\n",
       "      <td>1</td>\n",
       "      <td>934485</td>\n",
       "      <td>934485</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2360190</td>\n",
       "      <td>2352993</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000001.10:g.934485G&gt;C',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2360190</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0001076</td>\n",
       "      <td>ENST00000428771.2</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1076</td>\n",
       "      <td>HES4</td>\n",
       "      <td>1</td>\n",
       "      <td>934491</td>\n",
       "      <td>934491</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2389046</td>\n",
       "      <td>2374091</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000001.10:g.934491G&gt;A',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2389046</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0001077</td>\n",
       "      <td>ENST00000428771.2</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21499</th>\n",
       "      <td>2350089</td>\n",
       "      <td>CLIC2</td>\n",
       "      <td>X</td>\n",
       "      <td>154528454</td>\n",
       "      <td>154528454</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>807878</td>\n",
       "      <td>798237</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000023.10:g.154528454C&gt;A',)</td>\n",
       "      <td>('not_provided',)</td>\n",
       "      <td>807878</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2350090</td>\n",
       "      <td>ENST00000369449.2</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21500</th>\n",
       "      <td>2350090</td>\n",
       "      <td>CLIC2</td>\n",
       "      <td>X</td>\n",
       "      <td>154563670</td>\n",
       "      <td>154563670</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>446044</td>\n",
       "      <td>439316</td>\n",
       "      <td>('Benign/Likely_benign',)</td>\n",
       "      <td>('NC_000023.10:g.154563670G&gt;A',)</td>\n",
       "      <td>('not_provided',)</td>\n",
       "      <td>446044</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2350091</td>\n",
       "      <td>ENST00000369449.2</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21501</th>\n",
       "      <td>2350091</td>\n",
       "      <td>CLIC2</td>\n",
       "      <td>X</td>\n",
       "      <td>154563718</td>\n",
       "      <td>154563718</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1675174</td>\n",
       "      <td>1666531</td>\n",
       "      <td>('Conflicting_classifications_of_pathogenicity',)</td>\n",
       "      <td>('NC_000023.10:g.154563718C&gt;T',)</td>\n",
       "      <td>('Inborn_genetic_diseases|X-linked_intellectua...</td>\n",
       "      <td>1675174</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2350092</td>\n",
       "      <td>ENST00000369449.2</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21502</th>\n",
       "      <td>2350092</td>\n",
       "      <td>CLIC2</td>\n",
       "      <td>X</td>\n",
       "      <td>154563723</td>\n",
       "      <td>154563723</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2623060</td>\n",
       "      <td>2789876</td>\n",
       "      <td>('Uncertain_significance',)</td>\n",
       "      <td>('NC_000023.10:g.154563723C&gt;T',)</td>\n",
       "      <td>('Inborn_genetic_diseases',)</td>\n",
       "      <td>2623060</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2350093</td>\n",
       "      <td>ENST00000369449.2</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21503</th>\n",
       "      <td>2350197</td>\n",
       "      <td>TSPY1</td>\n",
       "      <td>Y</td>\n",
       "      <td>9304636</td>\n",
       "      <td>9304636</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>769492</td>\n",
       "      <td>706350</td>\n",
       "      <td>('Likely_benign',)</td>\n",
       "      <td>('NC_000024.9:g.9304636C&gt;T',)</td>\n",
       "      <td>('not_provided',)</td>\n",
       "      <td>769492</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2350198</td>\n",
       "      <td>ENST00000451548.1</td>\n",
       "      <td>GRCh37</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21504 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         index gene_name_h Chromosome    start_h      end_h ref_seq_h  \\\n",
       "0         1072        HES4          1     934467     934467         C   \n",
       "1         1073        HES4          1     934470     934470         G   \n",
       "2         1074        HES4          1     934476     934476         T   \n",
       "3         1075        HES4          1     934485     934485         G   \n",
       "4         1076        HES4          1     934491     934491         G   \n",
       "...        ...         ...        ...        ...        ...       ...   \n",
       "21499  2350089       CLIC2          X  154528454  154528454         C   \n",
       "21500  2350090       CLIC2          X  154563670  154563670         G   \n",
       "21501  2350091       CLIC2          X  154563718  154563718         C   \n",
       "21502  2350092       CLIC2          X  154563723  154563723         C   \n",
       "21503  2350197       TSPY1          Y    9304636    9304636         C   \n",
       "\n",
       "      alt_seq_h type_h  Variation_ID  Allele_ID  \\\n",
       "0             G    SNP       2261120    2255281   \n",
       "1             C    SNP       2360191    2352994   \n",
       "2             A    SNP       2363714    2353874   \n",
       "3             C    SNP       2360190    2352993   \n",
       "4             A    SNP       2389046    2374091   \n",
       "...         ...    ...           ...        ...   \n",
       "21499         A    SNP        807878     798237   \n",
       "21500         A    SNP        446044     439316   \n",
       "21501         T    SNP       1675174    1666531   \n",
       "21502         T    SNP       2623060    2789876   \n",
       "21503         T    SNP        769492     706350   \n",
       "\n",
       "                                                  CLNSIG  \\\n",
       "0                            ('Uncertain_significance',)   \n",
       "1                            ('Uncertain_significance',)   \n",
       "2                            ('Uncertain_significance',)   \n",
       "3                            ('Uncertain_significance',)   \n",
       "4                            ('Uncertain_significance',)   \n",
       "...                                                  ...   \n",
       "21499                        ('Uncertain_significance',)   \n",
       "21500                          ('Benign/Likely_benign',)   \n",
       "21501  ('Conflicting_classifications_of_pathogenicity',)   \n",
       "21502                        ('Uncertain_significance',)   \n",
       "21503                                 ('Likely_benign',)   \n",
       "\n",
       "                                CLNHGVS  \\\n",
       "0         ('NC_000001.10:g.934467C>G',)   \n",
       "1         ('NC_000001.10:g.934470G>C',)   \n",
       "2         ('NC_000001.10:g.934476T>A',)   \n",
       "3         ('NC_000001.10:g.934485G>C',)   \n",
       "4         ('NC_000001.10:g.934491G>A',)   \n",
       "...                                 ...   \n",
       "21499  ('NC_000023.10:g.154528454C>A',)   \n",
       "21500  ('NC_000023.10:g.154563670G>A',)   \n",
       "21501  ('NC_000023.10:g.154563718C>T',)   \n",
       "21502  ('NC_000023.10:g.154563723C>T',)   \n",
       "21503     ('NC_000024.9:g.9304636C>T',)   \n",
       "\n",
       "                                                   CLNDN  clinvar_id format  \\\n",
       "0                           ('Inborn_genetic_diseases',)     2261120    MAF   \n",
       "1                           ('Inborn_genetic_diseases',)     2360191    MAF   \n",
       "2                           ('Inborn_genetic_diseases',)     2363714    MAF   \n",
       "3                           ('Inborn_genetic_diseases',)     2360190    MAF   \n",
       "4                           ('Inborn_genetic_diseases',)     2389046    MAF   \n",
       "...                                                  ...         ...    ...   \n",
       "21499                                  ('not_provided',)      807878    MAF   \n",
       "21500                                  ('not_provided',)      446044    MAF   \n",
       "21501  ('Inborn_genetic_diseases|X-linked_intellectua...     1675174    MAF   \n",
       "21502                       ('Inborn_genetic_diseases',)     2623060    MAF   \n",
       "21503                                  ('not_provided',)      769492    MAF   \n",
       "\n",
       "                     ID            tx_id_h ref_genome_h  \n",
       "0      CLINVAR-H0001073  ENST00000428771.2       GRCh37  \n",
       "1      CLINVAR-H0001074  ENST00000428771.2       GRCh37  \n",
       "2      CLINVAR-H0001075  ENST00000428771.2       GRCh37  \n",
       "3      CLINVAR-H0001076  ENST00000428771.2       GRCh37  \n",
       "4      CLINVAR-H0001077  ENST00000428771.2       GRCh37  \n",
       "...                 ...                ...          ...  \n",
       "21499  CLINVAR-H2350090  ENST00000369449.2       GRCh37  \n",
       "21500  CLINVAR-H2350091  ENST00000369449.2       GRCh37  \n",
       "21501  CLINVAR-H2350092  ENST00000369449.2       GRCh37  \n",
       "21502  CLINVAR-H2350093  ENST00000369449.2       GRCh37  \n",
       "21503  CLINVAR-H2350198  ENST00000451548.1       GRCh37  \n",
       "\n",
       "[21504 rows x 18 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_wrong"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CLINVAR # genes not mapped with mouse orthologs: 650'"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_1 = len(df_wrong['gene_name_h'].unique())\n",
    "f'CLINVAR # genes not mapped with mouse orthologs: {l_1}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_mapped = set_mapped | set(df_clinvar['gene_name_h'].unique())\n",
    "set_unmapped = set_unmapped | set(df_wrong['gene_name_h'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "with open('/Users/kexindong/Documents/GitHub/Output/h2m_database/gene_mapped.pkl', 'wb') as file:\n",
    "    pickle.dump(set_mapped, file)\n",
    "\n",
    "with open('/Users/kexindong/Documents/GitHub/Output/h2m_database/gene_unmapped.pkl', 'wb') as file:\n",
    "    pickle.dump(set_unmapped, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17428"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set_mapped)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "793"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set_unmapped)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_wrong.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/clinvar_without_homo_gene.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['HES4', 'SCNN1D', 'CALML6', 'NBPF1', 'ACTL8', 'NBPF3', 'FCN3',\n",
       "       'IFI6', 'GJA9', 'ZFP69B', 'ZNF684', 'BEST4', 'CYP4Z1', 'DIRAS3',\n",
       "       'LINC01140', 'NBPF4', 'NBPF6', 'CHI3L2', 'CD58', 'NBPF9', 'NBPF8',\n",
       "       'NBPF10', 'NBPF11', 'NBPF12', 'NBPF14', 'PPIAL4G', 'PPIAL4D',\n",
       "       'NBPF20', 'NBPF15', 'PPIAL4E', 'FLG-AS1', 'LCE3A', 'LCE2B',\n",
       "       'LCE4A', 'LCE1D', 'SPRR2G', 'S100A12', 'AQP10', 'TDRD10', 'FCRL4',\n",
       "       'CD1A', 'CD1B', 'CD1E', 'HSPA6', 'SLC9C2', 'CFHR4', 'CFHR5',\n",
       "       'MIR181A1HG', 'C4BPB', 'FAM177B', 'ZP4', 'OR2C3', 'GCSAML',\n",
       "       'OR2G2', 'OR2G3', 'OR2M4', 'OR2T3', 'OR2T34', 'OR2T10', 'OR2T11',\n",
       "       'OR2T27', 'OR14I1', 'PGBD2', 'TP53I3', 'TRMT61B', 'CAPN14',\n",
       "       'OXER1', 'TTC31', 'TRABD2A', 'GNLY', 'PLGLB1', 'PLGLB2', 'ZNF514',\n",
       "       'ANKRD36C', 'ANKRD36', 'ANKRD36B', 'NT5DC4', 'IL37', 'CYP27C1',\n",
       "       'RAB6C', 'GPR148', 'FTCDNL1', 'CASP10', 'CPO', 'DIRC3', 'ZNF385D',\n",
       "       'ZNF619', 'ZNF620', 'ZNF621', 'ZNF662', 'ZNF660', 'ZNF197',\n",
       "       'ZNF501', 'ZNF589', 'IQCF2', 'ADAMTS9-AS2', 'EBLN2', 'HHLA2',\n",
       "       'GUCA1C', 'ZNF80', 'ARGFX', 'PARP15', 'GRK7', 'TM4SF18', 'TERC',\n",
       "       'MCF2L2', 'HTR3D', 'HTR3C', 'HTR3E', 'MAGEF1', 'ZNF141', 'MYL5',\n",
       "       'S100P', 'CCDC96', 'GPR78', 'FGFBP2', 'DCAF16', 'GBA3', 'TLR10',\n",
       "       'STATH', 'HTN3', 'HTN1', 'FDCSP', 'MUC7', 'THAP6', 'SCD5', 'THAP9',\n",
       "       'HERC5', 'RNF175', 'DDX60L', 'PLEKHG4B', 'SETD9', 'CCNI2', 'HMHB1',\n",
       "       'TIGD6', 'RPL26L1', 'FAM153A', 'BTNL8', 'BTNL3', 'C6orf52',\n",
       "       'BTN3A2', 'BTN3A3', 'ZNF391', 'ZNF165', 'ZSCAN16', 'ZSCAN9',\n",
       "       'ZSCAN31', 'ZSCAN23', 'ZNF311', 'OR2B3', 'MAS1L', 'HCG4', 'MUC22',\n",
       "       'PSORS1C1', 'MCCD1', 'HLA-DPA1', 'HLA-DPB1', 'MLN', 'CLPSL1',\n",
       "       'ETV7', 'KCNK17', 'NCR2', 'DEFB114', 'DEFB112', 'GSTA4', 'COL21A1',\n",
       "       'KIAA1586', 'EYS', 'HTR1E', 'GJB7', 'CASC6', 'MCHR2', 'KPNA5',\n",
       "       'TMEM244', 'VNN2', 'KIF25', 'FRMD1', 'RBAKDN', 'STK17A', 'COA1',\n",
       "       'ZNF479', 'ZNF680', 'ZNF138', 'ZNF273', 'ERV3-1', 'ZNF92', 'SAMD9',\n",
       "       'ZNF789', 'FAM200A', 'OR2AE1', 'TRIM4', 'NAT16', 'MOGAT3', 'LAMB4',\n",
       "       'THAP5', 'FEZF1-AS1', 'SMKR1', 'WEE2-AS1', 'TAS2R5', 'OR6V1',\n",
       "       'EPHA1-AS1', 'ARHGEF35', 'ZNF425', 'ZNF783', 'ZNF862', 'GIMAP2',\n",
       "       'ZNF596', 'FAM66B', 'PXDNL', 'PSKH2', 'FER1L6-AS1', 'FER1L6-AS2',\n",
       "       'ZNF572', 'PRNCR1', 'CASC19', 'CASC8', 'CASC11', 'CCDC26',\n",
       "       'TSNARE1', 'GLI4', 'ZNF16', 'GLIS3-AS1', 'INSL4', 'IFNW1',\n",
       "       'CDKN2B-AS1', 'ANKRD18B', 'ANKRD18A', 'ANKRD20A1', 'PRKACG',\n",
       "       'ZNF484', 'FAM120AOS', 'ZNF782', 'OR13C4', 'OR1L1', 'OR1L3',\n",
       "       'OR1K1', 'ZNF79', 'PAEP', 'LCNL1', 'EXD3', 'TUBB8', 'IDI2-AS1',\n",
       "       'PROSER2-AS1', 'EBLN1', 'PRTFDC1', 'ANKRD30A', 'ZNF33A', 'ZNF37A',\n",
       "       'ZNF33B', 'ZNF485', 'TMEM72-AS1', 'AGAP4', 'AGAP6', 'JMJD1C-AS1',\n",
       "       'ADIRF', 'KLLN', 'LIPJ', 'IFIT5', 'TLX1NB', 'PNLIPRP3', 'ARMS2',\n",
       "       'VENTX', 'KRTAP5-AS1', 'H19', 'C11orf21', 'KCNQ1DN', 'SLC22A18AS',\n",
       "       'MRGPRG-AS1', 'C11orf40', 'MMP26', 'OR51B5', 'OR56A1', 'ZNF215',\n",
       "       'ZNF214', 'NCR3LG1', 'IGSF22', 'DCDC1', 'FAM180B', 'OR4S1',\n",
       "       'TRIM48', 'OR5D13', 'OR5F1', 'OR6Q1', 'GLYATL2', 'GLYATL1',\n",
       "       'FAM111B', 'TCN1', 'SCGB1D1', 'SCGB1D2', 'SCGB1D4', 'SLC22A11',\n",
       "       'SPDYC', 'MYEOV', 'PAAF1', 'KDM4E', 'ANGPTL5', 'CARD18', 'NXPE1',\n",
       "       'OR6M1', 'OR6T1', 'TP53AIP1', 'GALNT8', 'RBP5', 'CD163L1',\n",
       "       'NANOGNB', 'A2ML1', 'KLRF1', 'CLEC2B', 'TAS2R8', 'TAS2R9', 'PRH1',\n",
       "       'OVCH1', 'KRT3', 'LACRT', 'DCD', 'MUCL1', 'OR10A7', 'OR6C4',\n",
       "       'IL26', 'TSPAN19', 'PLEKHG7', 'DEPDC4', 'ZNF140', 'ZNF10', 'ANHX',\n",
       "       'CSNK1A1L', 'MLNR', 'DLEU2', 'DHRS12', 'PRR20B', 'ATXN8OS', 'DAOA',\n",
       "       'OR11H12', 'OR11H2', 'OR4K14', 'OR4K13', 'OR4K17', 'RBM23',\n",
       "       'CHMP4A', 'SERPINA4', 'GOLGA6L6', 'GOLGA6L1', 'GOLGA6L2', 'NPAP1',\n",
       "       'SNORD108', 'SNHG14', 'PGBD4', 'NOX5', 'ST20', 'GOLGA6L10',\n",
       "       'GOLGA6L4', 'ZNF774', 'HBM', 'C1QTNF8', 'IL32', 'ZNF75A',\n",
       "       'ZSCAN32', 'ZNF500', 'NPIPA1', 'NPIPA5', 'CLEC19A', 'NPIPB3',\n",
       "       'NPIPB4', 'NPIPB5', 'C16orf82', 'NPIPB7', 'NPIPB8', 'NPIPB11',\n",
       "       'ZNF785', 'PYDC1', 'ZNF843', 'ABCC11', 'CETP', 'ZNF19',\n",
       "       'SNAI3-AS1', 'SLC22A31', 'OR1G1', 'ZNF232', 'ZNF594', 'KRBA2',\n",
       "       'CDRT15', 'TBC1D26', 'CCDC144A', 'TBC1D28', 'CDRT15L2', 'CCL16',\n",
       "       'CCL14', 'TBC1D3B', 'TBC1D3H', 'TBC1D3F', 'TBC1D3', 'GSDMB',\n",
       "       'KRT37', 'KRT38', 'DBF4B', 'PRAC1', 'PRAC2', 'NACA2', 'ABCA10',\n",
       "       'ZACN', 'TMEM105', 'CLUL1', 'ENOSF1', 'ANKRD30B', 'ZNF396',\n",
       "       'CCDC102B', 'SMIM21', 'AZU1', 'ZNF57', 'RAX2', 'TMIGD2',\n",
       "       'HSD11B1L', 'FUT6', 'FUT3', 'FUT5', 'CAPS', 'SH2D3A', 'FBN3',\n",
       "       'OR7D2', 'ZNF699', 'ZNF559', 'ZNF177', 'P2RY11', 'ICAM3',\n",
       "       'TSPAN16', 'ZNF564', 'STX10', 'CLEC17A', 'ZNF333', 'ZNF101',\n",
       "       'ZNF737', 'ZNF626', 'ZNF430', 'ZNF714', 'ZNF431', 'ZNF100',\n",
       "       'ZNF257', 'ZNF98', 'ZNF492', 'ZNF730', 'ZNF675', 'RPSA2',\n",
       "       'SCGB2B2', 'ZNF302', 'ZNF181', 'ZNF599', 'ZNF30', 'HPN-AS1',\n",
       "       'THAP8', 'ZNF529', 'ZNF461', 'ZNF829', 'ZNF568', 'ZNF527',\n",
       "       'ZNF571', 'ZNF540', 'ZNF607', 'ZNF573', 'CLC', 'LEUTX', 'ZNF576',\n",
       "       'ZNF283', 'ZNF404', 'ZNF155', 'ZNF230', 'ZNF222', 'ZNF223',\n",
       "       'ZNF224', 'ZNF225', 'ZNF227', 'ZNF285', 'ZNF229', 'IGFL1', 'TPRX1',\n",
       "       'ELSPBP1', 'CARD8', 'ZNF114', 'ADM5', 'GPR32', 'ZNF577', 'ZNF649',\n",
       "       'ZNF350', 'ZNF614', 'ZNF841', 'ZNF836', 'ZNF766', 'ZNF880',\n",
       "       'ZNF528', 'ZNF534', 'ZNF578', 'ZNF808', 'ZNF701', 'ZNF83',\n",
       "       'ZNF600', 'ZNF468', 'ZNF320', 'ZNF888', 'ZNF816-ZNF321P', 'ZNF816',\n",
       "       'ERVV-1', 'ERVV-2', 'ZNF415', 'ZNF677', 'ZNF765', 'ZNF331', 'DPRX',\n",
       "       'VSTM1', 'FCAR', 'ZNF581', 'NLRP11', 'NLRP13', 'NLRP8', 'ZNF470',\n",
       "       'ZNF71', 'ZNF835', 'ZIM2', 'ZIM3', 'DUXA', 'ZNF264', 'ZNF805',\n",
       "       'ZNF460', 'ZNF543', 'ZNF547', 'VN1R1', 'ZNF530', 'ZNF211',\n",
       "       'ZNF671', 'ZNF776', 'ZNF587', 'ZNF814', 'ZNF417', 'ZSCAN1',\n",
       "       'ZNF135', 'ZNF497', 'ZNF132', 'SIRPB2', 'ZNF343', 'GNRH2',\n",
       "       'ZNF133', 'CST4', 'CST1', 'CST2', 'C20orf203', 'PI3', 'WFDC6',\n",
       "       'BCAS4', 'APCDD1L', 'TNFRSF6B', 'SLC2A4RG', 'NPBWR2', 'KRTAP25-1',\n",
       "       'KRTAP23-1', 'KRTAP19-2', 'KRTAP19-7', 'KRTAP22-2', 'KRTAP6-2',\n",
       "       'KRTAP22-1', 'KRTAP6-1', 'KRTAP20-3', 'KRTAP21-3', 'KRTAP21-2',\n",
       "       'KRTAP21-1', 'TSPEAR-AS1', 'OR11H1', 'XKR3', 'GAB4', 'CLTCL1',\n",
       "       'ZNF74', 'RGL4', 'ZNF70', 'SLC2A11', 'PIWIL3', 'C22orf31',\n",
       "       'SEC14L6', 'C22orf42', 'RFPL3S', 'APOL5', 'APOBEC3A', 'IL17REL',\n",
       "       'SHOX', 'SLC25A6', 'ASMTL', 'P2RY8', 'ZBED1', 'XG', 'GYG2', 'ARSD',\n",
       "       'ARSH', 'ARSF', 'MXRA5', 'VCX3A', 'VCX', 'PNPLA4', 'VCX2', 'VCX3B',\n",
       "       'WWC3', 'ATXN3L', 'SCML1', 'DDX53', 'SUPT20HL2', 'SUPT20HL1',\n",
       "       'MAGEB10', 'MAGEB3', 'MPC1L', 'ZNF41', 'ZNF81', 'GAGE10',\n",
       "       'GAGE12J', 'GAGE12C', 'GAGE12D', 'GAGE2E', 'GAGE12G', 'GAGE12H',\n",
       "       'GAGE2A', 'PAGE1', 'PAGE4', 'SPANXN5', 'XAGE5', 'XAGE3', 'PAGE2B',\n",
       "       'PAGE2', 'PAGE5', 'PAGE3', 'SPIN3', 'FAAH2', 'MTMR8', 'RAB41',\n",
       "       'PABPC1L2B', 'FAM133A', 'TCEAL2', 'BEX5', 'LINC00630', 'TCEAL4',\n",
       "       'TMEM31', 'LUZP4', 'CT83', 'ZNF75D', 'SAGE1', 'SPANXC', 'SPANXD',\n",
       "       'SPANXN4', 'SPANXN3', 'SPANXN2', 'SPANXN1', 'MIR510', 'CSAG1',\n",
       "       'TMEM187', 'CTAG2', 'CLIC2', 'TSPY1'], dtype=object)"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_wrong['gene_name_h'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There were rows that could not be processed.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>type_h</th>\n",
       "      <th>Variation_ID</th>\n",
       "      <th>Allele_ID</th>\n",
       "      <th>CLNSIG</th>\n",
       "      <th>CLNHGVS</th>\n",
       "      <th>CLNDN</th>\n",
       "      <th>clinvar_id</th>\n",
       "      <th>format</th>\n",
       "      <th>ID</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>ref_genome_h</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2135678</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858782</td>\n",
       "      <td>58858782</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2336128</td>\n",
       "      <td>2328468</td>\n",
       "      <td>(Uncertain_significance,)</td>\n",
       "      <td>(NC_000019.9:g.58858782G&gt;C,)</td>\n",
       "      <td>(Inborn_genetic_diseases,)</td>\n",
       "      <td>2336128</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2135679</td>\n",
       "      <td>ENST00000263100.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>A1bg</td>\n",
       "      <td>ENSMUST00000096418.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2135679</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858882</td>\n",
       "      <td>58858882</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2521649</td>\n",
       "      <td>2696794</td>\n",
       "      <td>(Uncertain_significance,)</td>\n",
       "      <td>(NC_000019.9:g.58858882C&gt;G,)</td>\n",
       "      <td>(Inborn_genetic_diseases,)</td>\n",
       "      <td>2521649</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2135680</td>\n",
       "      <td>ENST00000263100.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>A1bg</td>\n",
       "      <td>ENSMUST00000096418.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2135680</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858888</td>\n",
       "      <td>58858888</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2552144</td>\n",
       "      <td>2714907</td>\n",
       "      <td>(Uncertain_significance,)</td>\n",
       "      <td>(NC_000019.9:g.58858888C&gt;G,)</td>\n",
       "      <td>(Inborn_genetic_diseases,)</td>\n",
       "      <td>2552144</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2135681</td>\n",
       "      <td>ENST00000263100.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>A1bg</td>\n",
       "      <td>ENSMUST00000096418.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2135681</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58858923</td>\n",
       "      <td>58858923</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2484196</td>\n",
       "      <td>2455282</td>\n",
       "      <td>(Uncertain_significance,)</td>\n",
       "      <td>(NC_000019.9:g.58858923G&gt;A,)</td>\n",
       "      <td>(Inborn_genetic_diseases,)</td>\n",
       "      <td>2484196</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2135682</td>\n",
       "      <td>ENST00000263100.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>A1bg</td>\n",
       "      <td>ENSMUST00000096418.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2135682</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>58861805</td>\n",
       "      <td>58861805</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2623749</td>\n",
       "      <td>2787053</td>\n",
       "      <td>(Uncertain_significance,)</td>\n",
       "      <td>(NC_000019.9:g.58861805C&gt;T,)</td>\n",
       "      <td>(Inborn_genetic_diseases,)</td>\n",
       "      <td>2623749</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H2135683</td>\n",
       "      <td>ENST00000263100.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>A1bg</td>\n",
       "      <td>ENSMUST00000096418.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2513279</th>\n",
       "      <td>83147</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1</td>\n",
       "      <td>78098759</td>\n",
       "      <td>78098759</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2331197</td>\n",
       "      <td>2327429</td>\n",
       "      <td>(Uncertain_significance,)</td>\n",
       "      <td>(NC_000001.10:g.78098759T&gt;C,)</td>\n",
       "      <td>(Inborn_genetic_diseases,)</td>\n",
       "      <td>2331197</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0083148</td>\n",
       "      <td>ENST00000370801.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Zzz3</td>\n",
       "      <td>ENSMUST00000106100.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2513280</th>\n",
       "      <td>83148</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1</td>\n",
       "      <td>78098762</td>\n",
       "      <td>78098762</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2618943</td>\n",
       "      <td>2782688</td>\n",
       "      <td>(Uncertain_significance,)</td>\n",
       "      <td>(NC_000001.10:g.78098762T&gt;C,)</td>\n",
       "      <td>(Inborn_genetic_diseases,)</td>\n",
       "      <td>2618943</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0083149</td>\n",
       "      <td>ENST00000370801.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Zzz3</td>\n",
       "      <td>ENSMUST00000106100.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2513281</th>\n",
       "      <td>83149</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1</td>\n",
       "      <td>78098790</td>\n",
       "      <td>78098790</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2456830</td>\n",
       "      <td>2431257</td>\n",
       "      <td>(Uncertain_significance,)</td>\n",
       "      <td>(NC_000001.10:g.78098790T&gt;C,)</td>\n",
       "      <td>(Inborn_genetic_diseases,)</td>\n",
       "      <td>2456830</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0083150</td>\n",
       "      <td>ENST00000370801.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Zzz3</td>\n",
       "      <td>ENSMUST00000106100.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2513282</th>\n",
       "      <td>83150</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1</td>\n",
       "      <td>78098992</td>\n",
       "      <td>78098992</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2408495</td>\n",
       "      <td>2391953</td>\n",
       "      <td>(Uncertain_significance,)</td>\n",
       "      <td>(NC_000001.10:g.78098992T&gt;A,)</td>\n",
       "      <td>(Inborn_genetic_diseases,)</td>\n",
       "      <td>2408495</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0083151</td>\n",
       "      <td>ENST00000370801.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Zzz3</td>\n",
       "      <td>ENSMUST00000106100.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2513283</th>\n",
       "      <td>83151</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1</td>\n",
       "      <td>78099017</td>\n",
       "      <td>78099017</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2271322</td>\n",
       "      <td>2257695</td>\n",
       "      <td>(Uncertain_significance,)</td>\n",
       "      <td>(NC_000001.10:g.78099017C&gt;T,)</td>\n",
       "      <td>(Inborn_genetic_diseases,)</td>\n",
       "      <td>2271322</td>\n",
       "      <td>MAF</td>\n",
       "      <td>CLINVAR-H0083152</td>\n",
       "      <td>ENST00000370801.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Zzz3</td>\n",
       "      <td>ENSMUST00000106100.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2513284 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           index gene_name_h Chromosome   start_h     end_h ref_seq_h  \\\n",
       "0        2135678        A1BG         19  58858782  58858782         G   \n",
       "1        2135679        A1BG         19  58858882  58858882         C   \n",
       "2        2135680        A1BG         19  58858888  58858888         C   \n",
       "3        2135681        A1BG         19  58858923  58858923         G   \n",
       "4        2135682        A1BG         19  58861805  58861805         C   \n",
       "...          ...         ...        ...       ...       ...       ...   \n",
       "2513279    83147        ZZZ3          1  78098759  78098759         T   \n",
       "2513280    83148        ZZZ3          1  78098762  78098762         T   \n",
       "2513281    83149        ZZZ3          1  78098790  78098790         T   \n",
       "2513282    83150        ZZZ3          1  78098992  78098992         T   \n",
       "2513283    83151        ZZZ3          1  78099017  78099017         C   \n",
       "\n",
       "        alt_seq_h type_h  Variation_ID  Allele_ID                     CLNSIG  \\\n",
       "0               C    SNP       2336128    2328468  (Uncertain_significance,)   \n",
       "1               G    SNP       2521649    2696794  (Uncertain_significance,)   \n",
       "2               G    SNP       2552144    2714907  (Uncertain_significance,)   \n",
       "3               A    SNP       2484196    2455282  (Uncertain_significance,)   \n",
       "4               T    SNP       2623749    2787053  (Uncertain_significance,)   \n",
       "...           ...    ...           ...        ...                        ...   \n",
       "2513279         C    SNP       2331197    2327429  (Uncertain_significance,)   \n",
       "2513280         C    SNP       2618943    2782688  (Uncertain_significance,)   \n",
       "2513281         C    SNP       2456830    2431257  (Uncertain_significance,)   \n",
       "2513282         A    SNP       2408495    2391953  (Uncertain_significance,)   \n",
       "2513283         T    SNP       2271322    2257695  (Uncertain_significance,)   \n",
       "\n",
       "                               CLNHGVS                       CLNDN clinvar_id  \\\n",
       "0         (NC_000019.9:g.58858782G>C,)  (Inborn_genetic_diseases,)    2336128   \n",
       "1         (NC_000019.9:g.58858882C>G,)  (Inborn_genetic_diseases,)    2521649   \n",
       "2         (NC_000019.9:g.58858888C>G,)  (Inborn_genetic_diseases,)    2552144   \n",
       "3         (NC_000019.9:g.58858923G>A,)  (Inborn_genetic_diseases,)    2484196   \n",
       "4         (NC_000019.9:g.58861805C>T,)  (Inborn_genetic_diseases,)    2623749   \n",
       "...                                ...                         ...        ...   \n",
       "2513279  (NC_000001.10:g.78098759T>C,)  (Inborn_genetic_diseases,)    2331197   \n",
       "2513280  (NC_000001.10:g.78098762T>C,)  (Inborn_genetic_diseases,)    2618943   \n",
       "2513281  (NC_000001.10:g.78098790T>C,)  (Inborn_genetic_diseases,)    2456830   \n",
       "2513282  (NC_000001.10:g.78098992T>A,)  (Inborn_genetic_diseases,)    2408495   \n",
       "2513283  (NC_000001.10:g.78099017C>T,)  (Inborn_genetic_diseases,)    2271322   \n",
       "\n",
       "        format                ID            tx_id_h ref_genome_h gene_name_m  \\\n",
       "0          MAF  CLINVAR-H2135679  ENST00000263100.3       GRCh37        A1bg   \n",
       "1          MAF  CLINVAR-H2135680  ENST00000263100.3       GRCh37        A1bg   \n",
       "2          MAF  CLINVAR-H2135681  ENST00000263100.3       GRCh37        A1bg   \n",
       "3          MAF  CLINVAR-H2135682  ENST00000263100.3       GRCh37        A1bg   \n",
       "4          MAF  CLINVAR-H2135683  ENST00000263100.3       GRCh37        A1bg   \n",
       "...        ...               ...                ...          ...         ...   \n",
       "2513279    MAF  CLINVAR-H0083148  ENST00000370801.3       GRCh37        Zzz3   \n",
       "2513280    MAF  CLINVAR-H0083149  ENST00000370801.3       GRCh37        Zzz3   \n",
       "2513281    MAF  CLINVAR-H0083150  ENST00000370801.3       GRCh37        Zzz3   \n",
       "2513282    MAF  CLINVAR-H0083151  ENST00000370801.3       GRCh37        Zzz3   \n",
       "2513283    MAF  CLINVAR-H0083152  ENST00000370801.3       GRCh37        Zzz3   \n",
       "\n",
       "                      tx_id_m  \n",
       "0        ENSMUST00000096418.5  \n",
       "1        ENSMUST00000096418.5  \n",
       "2        ENSMUST00000096418.5  \n",
       "3        ENSMUST00000096418.5  \n",
       "4        ENSMUST00000096418.5  \n",
       "...                       ...  \n",
       "2513279  ENSMUST00000106100.9  \n",
       "2513280  ENSMUST00000106100.9  \n",
       "2513281  ENSMUST00000106100.9  \n",
       "2513282  ENSMUST00000106100.9  \n",
       "2513283  ENSMUST00000106100.9  \n",
       "\n",
       "[2513284 rows x 20 columns]"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clinvar,df_wrong = h2m.get_tx_batch(df_clinvar, species='m')\n",
    "df_clinvar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [],
   "source": [
    "numbers = label_numbers(df_clinvar['ID'])\n",
    "df_clinvar['ID'] = [get_label(h,x) for h,x in zip(df_clinvar['ID'],numbers)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clinvar = df_clinvar.rename(columns={'ID':'index'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clinvar.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/clinvar_input.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clinvar = df_clinvar.iloc[:,1:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (4) Put all together and split into 16 files  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = pd.concat([df[['gene_name_h','tx_id_h','tx_id_m','start_h','end_h','ref_seq_h','alt_seq_h','type_h','index']].rename(columns={'index':'id'}),\n",
    "         df_cosmic[['gene_name_h','tx_id_h','tx_id_m','start_h','end_h','ref_seq_h','alt_seq_h','type_h','index']].rename(columns={'index':'id'}),\n",
    "         df_clinvar[['gene_name_h','tx_id_h','tx_id_m','start_h','end_h','ref_seq_h','alt_seq_h','type_h','index']].rename(columns={'index':'id'})],ignore_index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = df_all.rename(columns={'id':'index'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = df_all.sort_values(by = ['tx_id_h','tx_id_m']).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = df_all.dropna(subset = ['tx_id_h','tx_id_m']).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Run H2M"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "See folder **run_h2m**  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Process result data  \n",
    "## (1) Combine left over inputs  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "left_over_path = '/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/left_over_round_2/'\n",
    "list_of_left_over = []\n",
    "for name in range(1,17):\n",
    "    path = f'{left_over_path}/h2m_db_left_over_{name}_r2.csv'\n",
    "    list_of_left_over.append(pd.read_csv(path))\n",
    "df_left_over = pd.concat(list_of_left_over).sort_values(by='index').reset_index(drop=True)\n",
    "df_left_over = df_left_over.rename(columns={'index':'ID'})\n",
    "df_left_over['database'] = [x.split('-')[0] for x in df_left_over['ID']]\n",
    "df_left_over.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/left_over.csv',index=False)\n",
    "del df_left_over"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (2) Combine results  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "def label_numbers(numbers):\n",
    "    last_number = None\n",
    "    count = 0\n",
    "    labeled_numbers = []\n",
    "\n",
    "    for num in numbers:\n",
    "        if num == last_number:\n",
    "            count += 1\n",
    "        else:\n",
    "            count = 1\n",
    "        last_number = num\n",
    "        labeled_numbers.append(count)\n",
    "\n",
    "    return labeled_numbers\n",
    "\n",
    "def get_label_variant(h,x):\n",
    "    return f\"{h}-V{x:02}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# bind result 8\n",
    "path_8 = '/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/result_round_1/h2m_db_result_8_'\n",
    "list_8 = []\n",
    "for name in [1,3,4,5,6,7,8,9,10,11,13]:\n",
    "    path = f'{path_8}{name}.csv'\n",
    "    list_8.append(pd.read_csv(path))\n",
    "df_result_8 = pd.concat(list_8).sort_values(by='index').reset_index(drop=True)\n",
    "df_result_8.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/result_round_1/h2m_db_result_8.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/result_round_1/h2m_db_result_'\n",
    "# bind result 1-16 round 1 \n",
    "path = '/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/result_round_1/h2m_db_result_'\n",
    "list_of_result = []\n",
    "for name in range(1,17):\n",
    "    path_of_result = f'{path}{name}.csv'\n",
    "    list_of_result.append(pd.read_csv(path_of_result))\n",
    "df_result = pd.concat(list_of_result).sort_values(by='index').reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# bind result 1-16 round 2\n",
    "path = '/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/result_round_2/h2m_db_result_'\n",
    "list_of_result = []\n",
    "for name in range(1,17):\n",
    "    path_of_result = f'{path}{name}_r2.csv'\n",
    "    list_of_result.append(pd.read_csv(path_of_result))\n",
    "df_result_r2 = pd.concat(list_of_result).sort_values(by='index').reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_result_all = pd.concat([df_result, df_result_r2]).sort_values(by='index').reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_result_all.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/df_result_raw.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_result_all = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/df_result_raw.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>gene_id_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>chr_h</th>\n",
       "      <th>exon_num_h</th>\n",
       "      <th>strand_h</th>\n",
       "      <th>match</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>...</th>\n",
       "      <th>alt_seq_m_ori</th>\n",
       "      <th>HGVSc_m_ori</th>\n",
       "      <th>HGVSp_m_ori</th>\n",
       "      <th>start_m</th>\n",
       "      <th>end_m</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>alt_seq_m</th>\n",
       "      <th>HGVSc_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "      <th>index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>ENSG00000133703.7</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>chr12</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000111710.8:c.35G&gt;A</td>\n",
       "      <td>G12D</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000111710.8:c.35G&gt;A</td>\n",
       "      <td>G12D</td>\n",
       "      <td>AACR-H0000001-M01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>ENSG00000133703.7</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>chr12</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000111710.8:c.35G&gt;T</td>\n",
       "      <td>G12V</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000111710.8:c.35G&gt;T</td>\n",
       "      <td>G12V</td>\n",
       "      <td>AACR-H0000002-M01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>ENSG00000157764.8</td>\n",
       "      <td>ENST00000288602.6</td>\n",
       "      <td>chr7</td>\n",
       "      <td>18</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>140453136</td>\n",
       "      <td>140453136</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000002487.15:c.1910T&gt;A</td>\n",
       "      <td>V637E</td>\n",
       "      <td>39604717.0</td>\n",
       "      <td>39604717.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000002487.15:c.1910T&gt;A</td>\n",
       "      <td>V637E</td>\n",
       "      <td>AACR-H0000003-M01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>ENSG00000121879.3</td>\n",
       "      <td>ENST00000263967.3</td>\n",
       "      <td>chr3</td>\n",
       "      <td>20</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>178936091</td>\n",
       "      <td>178936091</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000029201.14:c.1633G&gt;A</td>\n",
       "      <td>E545K</td>\n",
       "      <td>32502197.0</td>\n",
       "      <td>32502197.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000029201.14:c.1633G&gt;A</td>\n",
       "      <td>E545K</td>\n",
       "      <td>AACR-H0000004-M01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>ENSG00000121879.3</td>\n",
       "      <td>ENST00000263967.3</td>\n",
       "      <td>chr3</td>\n",
       "      <td>20</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>178952085</td>\n",
       "      <td>178952085</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSMUST00000029201.14:c.3140A&gt;G</td>\n",
       "      <td>H1047R</td>\n",
       "      <td>32516913.0</td>\n",
       "      <td>32516913.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSMUST00000029201.14:c.3140A&gt;G</td>\n",
       "      <td>H1047R</td>\n",
       "      <td>AACR-H0000005-M01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4564993</th>\n",
       "      <td>LYN</td>\n",
       "      <td>ENSG00000254087.3</td>\n",
       "      <td>ENST00000519728.1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>56912024</td>\n",
       "      <td>56912024</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000041377.13:c.1252G&gt;A</td>\n",
       "      <td>G418S</td>\n",
       "      <td>3783020.0</td>\n",
       "      <td>3783020.0</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSMUST00000041377.13:c.1252G&gt;C</td>\n",
       "      <td>G418R</td>\n",
       "      <td>COSMIC-H1771600-M01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4564994</th>\n",
       "      <td>LYN</td>\n",
       "      <td>ENSG00000254087.3</td>\n",
       "      <td>ENST00000519728.1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>56912024</td>\n",
       "      <td>56912024</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000041377.13:c.1252G&gt;A</td>\n",
       "      <td>G418S</td>\n",
       "      <td>3783020.0</td>\n",
       "      <td>3783022.0</td>\n",
       "      <td>GGC</td>\n",
       "      <td>AGG</td>\n",
       "      <td>ENSMUST00000041377.13:c.1252_1254GGC&gt;AGG</td>\n",
       "      <td>G418R</td>\n",
       "      <td>COSMIC-H1771600-M01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4564995</th>\n",
       "      <td>LYN</td>\n",
       "      <td>ENSG00000254087.3</td>\n",
       "      <td>ENST00000519728.1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>56912024</td>\n",
       "      <td>56912024</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000041377.13:c.1252G&gt;A</td>\n",
       "      <td>G418S</td>\n",
       "      <td>3783020.0</td>\n",
       "      <td>3783022.0</td>\n",
       "      <td>GGC</td>\n",
       "      <td>CGG</td>\n",
       "      <td>ENSMUST00000041377.13:c.1252_1254GGC&gt;CGG</td>\n",
       "      <td>G418R</td>\n",
       "      <td>COSMIC-H1771600-M01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4564996</th>\n",
       "      <td>LYN</td>\n",
       "      <td>ENSG00000254087.3</td>\n",
       "      <td>ENST00000519728.1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>56912024</td>\n",
       "      <td>56912024</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000041377.13:c.1252G&gt;A</td>\n",
       "      <td>G418S</td>\n",
       "      <td>3783020.0</td>\n",
       "      <td>3783022.0</td>\n",
       "      <td>GGC</td>\n",
       "      <td>CGT</td>\n",
       "      <td>ENSMUST00000041377.13:c.1252_1254GGC&gt;CGT</td>\n",
       "      <td>G418R</td>\n",
       "      <td>COSMIC-H1771600-M01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4564997</th>\n",
       "      <td>LYN</td>\n",
       "      <td>ENSG00000254087.3</td>\n",
       "      <td>ENST00000519728.1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>56912024</td>\n",
       "      <td>56912024</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000041377.13:c.1252G&gt;A</td>\n",
       "      <td>G418S</td>\n",
       "      <td>3783020.0</td>\n",
       "      <td>3783022.0</td>\n",
       "      <td>GGC</td>\n",
       "      <td>AGA</td>\n",
       "      <td>ENSMUST00000041377.13:c.1252_1254GGC&gt;AGA</td>\n",
       "      <td>G418R</td>\n",
       "      <td>COSMIC-H1771600-M01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4564998 rows × 43 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name_h          gene_id_h            tx_id_h  chr_h  exon_num_h  \\\n",
       "0              KRAS  ENSG00000133703.7  ENST00000256078.4  chr12           4   \n",
       "1              KRAS  ENSG00000133703.7  ENST00000256078.4  chr12           4   \n",
       "2              BRAF  ENSG00000157764.8  ENST00000288602.6   chr7          18   \n",
       "3            PIK3CA  ENSG00000121879.3  ENST00000263967.3   chr3          20   \n",
       "4            PIK3CA  ENSG00000121879.3  ENST00000263967.3   chr3          20   \n",
       "...             ...                ...                ...    ...         ...   \n",
       "4564993         LYN  ENSG00000254087.3  ENST00000519728.1   chr8          12   \n",
       "4564994         LYN  ENSG00000254087.3  ENST00000519728.1   chr8          12   \n",
       "4564995         LYN  ENSG00000254087.3  ENST00000519728.1   chr8          12   \n",
       "4564996         LYN  ENSG00000254087.3  ENST00000519728.1   chr8          12   \n",
       "4564997         LYN  ENSG00000254087.3  ENST00000519728.1   chr8          12   \n",
       "\n",
       "        strand_h  match    start_h      end_h ref_seq_h  ... alt_seq_m_ori  \\\n",
       "0              -   True   25398284   25398284         C  ...             T   \n",
       "1              -   True   25398284   25398284         C  ...             A   \n",
       "2              -   True  140453136  140453136         A  ...             T   \n",
       "3              +   True  178936091  178936091         G  ...             A   \n",
       "4              +   True  178952085  178952085         A  ...             G   \n",
       "...          ...    ...        ...        ...       ...  ...           ...   \n",
       "4564993        +   True   56912024   56912024         G  ...             A   \n",
       "4564994        +   True   56912024   56912024         G  ...             A   \n",
       "4564995        +   True   56912024   56912024         G  ...             A   \n",
       "4564996        +   True   56912024   56912024         G  ...             A   \n",
       "4564997        +   True   56912024   56912024         G  ...             A   \n",
       "\n",
       "                             HGVSc_m_ori HGVSp_m_ori      start_m  \\\n",
       "0           ENSMUST00000111710.8:c.35G>A        G12D  145192497.0   \n",
       "1           ENSMUST00000111710.8:c.35G>T        G12V  145192497.0   \n",
       "2        ENSMUST00000002487.15:c.1910T>A       V637E   39604717.0   \n",
       "3        ENSMUST00000029201.14:c.1633G>A       E545K   32502197.0   \n",
       "4        ENSMUST00000029201.14:c.3140A>G      H1047R   32516913.0   \n",
       "...                                  ...         ...          ...   \n",
       "4564993  ENSMUST00000041377.13:c.1252G>A       G418S    3783020.0   \n",
       "4564994  ENSMUST00000041377.13:c.1252G>A       G418S    3783020.0   \n",
       "4564995  ENSMUST00000041377.13:c.1252G>A       G418S    3783020.0   \n",
       "4564996  ENSMUST00000041377.13:c.1252G>A       G418S    3783020.0   \n",
       "4564997  ENSMUST00000041377.13:c.1252G>A       G418S    3783020.0   \n",
       "\n",
       "               end_m ref_seq_m  alt_seq_m  \\\n",
       "0        145192497.0         C          T   \n",
       "1        145192497.0         C          A   \n",
       "2         39604717.0         A          T   \n",
       "3         32502197.0         G          A   \n",
       "4         32516913.0         A          G   \n",
       "...              ...       ...        ...   \n",
       "4564993    3783020.0         G          C   \n",
       "4564994    3783022.0       GGC        AGG   \n",
       "4564995    3783022.0       GGC        CGG   \n",
       "4564996    3783022.0       GGC        CGT   \n",
       "4564997    3783022.0       GGC        AGA   \n",
       "\n",
       "                                          HGVSc_m HGVSp_m                index  \n",
       "0                    ENSMUST00000111710.8:c.35G>A    G12D    AACR-H0000001-M01  \n",
       "1                    ENSMUST00000111710.8:c.35G>T    G12V    AACR-H0000002-M01  \n",
       "2                 ENSMUST00000002487.15:c.1910T>A   V637E    AACR-H0000003-M01  \n",
       "3                 ENSMUST00000029201.14:c.1633G>A   E545K    AACR-H0000004-M01  \n",
       "4                 ENSMUST00000029201.14:c.3140A>G  H1047R    AACR-H0000005-M01  \n",
       "...                                           ...     ...                  ...  \n",
       "4564993           ENSMUST00000041377.13:c.1252G>C   G418R  COSMIC-H1771600-M01  \n",
       "4564994  ENSMUST00000041377.13:c.1252_1254GGC>AGG   G418R  COSMIC-H1771600-M01  \n",
       "4564995  ENSMUST00000041377.13:c.1252_1254GGC>CGG   G418R  COSMIC-H1771600-M01  \n",
       "4564996  ENSMUST00000041377.13:c.1252_1254GGC>CGT   G418R  COSMIC-H1771600-M01  \n",
       "4564997  ENSMUST00000041377.13:c.1252_1254GGC>AGA   G418R  COSMIC-H1771600-M01  \n",
       "\n",
       "[4564998 rows x 43 columns]"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_result_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_addon = pd.read_csv(f'/Users/kexindong/Documents/GitHub/Output/h2m_database/output_addon.csv')\n",
    "numbers = label_numbers(df_addon['index'])\n",
    "df_addon['ID_mouse'] = [get_label_variant(h,x) for h,x in zip(df_addon['index'], numbers)]\n",
    "df_addon = df_addon.rename(columns={'index':'ID_human'})\n",
    "df_addon = df_addon.sort_values(by='ID_mouse').reset_index(drop=True)\n",
    "df_addon['ref_seq_h'] = df_addon['ref_seq_h'].fillna('-')\n",
    "df_addon['alt_seq_h'] = df_addon['alt_seq_h'].fillna('-')\n",
    "df_addon['Database'] = [x.split('-')[0] for x in df_addon['ID_mouse']]\n",
    "# remove non-modelable mouse IDs\n",
    "import numpy as np\n",
    "df_addon.loc[df_addon['status'] == False, 'ID_mouse'] = np.nan\n",
    "# remove M01 in the human ID column \n",
    "df_addon['ID_human'] = [x[:-4] for x in df_addon['ID_human']]\n",
    "df_addon['ID_human']\n",
    "df_addon['ID_db'] =[ dict_clinvar.get(x,np.nan) for x in df_addon['ID_human']]\n",
    "df_addon = df_addon[df_addon['class']!=5].reset_index(drop=True)\n",
    "df_binded = pd.concat([df_addon, df_binded])\n",
    "df_binded = df_binded.sort_values(['gene_name_h','type_h','gene_name_m','class','Database','ID_human'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded = df_binded.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Result data cleaning  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "numbers = label_numbers(df_result_all['index'])\n",
    "df_result_all['ID_mouse'] = [get_label_variant(h,x) for h,x in zip(df_result_all['index'], numbers)]\n",
    "df_binded = df_result_all.rename(columns={'index':'ID_human'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded = df_binded.sort_values(by='ID_mouse').reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded['ref_seq_h'] = df_binded['ref_seq_h'].fillna('-')\n",
    "df_binded['alt_seq_h'] = df_binded['alt_seq_h'].fillna('-')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded['Database'] = [x.split('-')[0] for x in df_binded['ID_mouse']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['gene_name_h', 'gene_id_h', 'tx_id_h', 'chr_h', 'exon_num_h',\n",
       "       'strand_h', 'match', 'start_h', 'end_h', 'ref_seq_h', 'alt_seq_h',\n",
       "       'HGVSc_h', 'HGVSp_h', 'classification_h', 'exon_h', 'type_h', 'status',\n",
       "       'class', 'statement', 'flank_size_left', 'flank_size_right',\n",
       "       'gene_name_m', 'gene_id_m', 'tx_id_m', 'chr_m', 'exon_num_m',\n",
       "       'strand_m', 'type_m', 'classification_m', 'exon_m', 'start_m_ori',\n",
       "       'end_m_ori', 'ref_seq_m_ori', 'alt_seq_m_ori', 'HGVSc_m_ori',\n",
       "       'HGVSp_m_ori', 'start_m', 'end_m', 'ref_seq_m', 'alt_seq_m', 'HGVSc_m',\n",
       "       'HGVSp_m', 'ID_human', 'ID_mouse', 'Database'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_binded.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove non-modelable mouse IDs\n",
    "import numpy as np\n",
    "df_binded.loc[df_binded['status'] == False, 'ID_mouse'] = np.nan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove M01 in the human ID column \n",
    "df_binded['ID_human'] = [x[:-4] for x in df_binded['ID_human']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Add original database ID  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "path = '/Users/kexindong/Documents/GitHub/Output/h2m_database/'\n",
    "# later\n",
    "with open(f'{path}dict_ori_id.json', 'r') as file:\n",
    "    dict_clinvar = json.load(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### DONOT NEED TO RUN\n",
    "path = '/Users/kexindong/Documents/GitHub/Output/h2m_database/'\n",
    "name_1 = 'clinvar_input.csv'\n",
    "name_2 = 'cosmic_input.csv'\n",
    "clinvar = pd.read_csv(f'{path}{name_1}')\n",
    "clinvar['index'] = [x[:-4] for x in clinvar['index']]\n",
    "dict_clinvar = dict(zip(clinvar['index'],clinvar['clinvar_id']))\n",
    "clinvar[clinvar['Variation_ID']==337847]\n",
    "del clinvar\n",
    "cosmic = pd.read_csv(f'{path}{name_2}')\n",
    "cosmic['index'] = [x[:-4] for x in cosmic['index'] ]\n",
    "dict_clinvar.update(dict(zip(cosmic['index'],cosmic['cosmic_ID'])))\n",
    "del cosmic\n",
    "with open(f'{path}dict_ori_id.json', 'w') as file:\n",
    "    json.dump(dict_clinvar, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded['ID_db'] =[ dict_clinvar.get(x,np.nan) for x in df_binded['ID_human']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>gene_id_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>chr_h</th>\n",
       "      <th>exon_num_h</th>\n",
       "      <th>strand_h</th>\n",
       "      <th>match</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>...</th>\n",
       "      <th>start_m</th>\n",
       "      <th>end_m</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>alt_seq_m</th>\n",
       "      <th>HGVSc_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "      <th>ID_human</th>\n",
       "      <th>ID_mouse</th>\n",
       "      <th>Database</th>\n",
       "      <th>ID_db</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>ENSG00000133703.7</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>chr12</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000111710.8:c.35G&gt;A</td>\n",
       "      <td>G12D</td>\n",
       "      <td>AACR-H0000001</td>\n",
       "      <td>AACR-H0000001-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>ENSG00000133703.7</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>chr12</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000111710.8:c.35G&gt;T</td>\n",
       "      <td>G12V</td>\n",
       "      <td>AACR-H0000002</td>\n",
       "      <td>AACR-H0000002-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>ENSG00000157764.8</td>\n",
       "      <td>ENST00000288602.6</td>\n",
       "      <td>chr7</td>\n",
       "      <td>18</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>140453136</td>\n",
       "      <td>140453136</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>39604717.0</td>\n",
       "      <td>39604717.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000002487.15:c.1910T&gt;A</td>\n",
       "      <td>V637E</td>\n",
       "      <td>AACR-H0000003</td>\n",
       "      <td>AACR-H0000003-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>ENSG00000121879.3</td>\n",
       "      <td>ENST00000263967.3</td>\n",
       "      <td>chr3</td>\n",
       "      <td>20</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>178936091</td>\n",
       "      <td>178936091</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>32502197.0</td>\n",
       "      <td>32502197.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000029201.14:c.1633G&gt;A</td>\n",
       "      <td>E545K</td>\n",
       "      <td>AACR-H0000004</td>\n",
       "      <td>AACR-H0000004-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>ENSG00000121879.3</td>\n",
       "      <td>ENST00000263967.3</td>\n",
       "      <td>chr3</td>\n",
       "      <td>20</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>178952085</td>\n",
       "      <td>178952085</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>32516913.0</td>\n",
       "      <td>32516913.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSMUST00000029201.14:c.3140A&gt;G</td>\n",
       "      <td>H1047R</td>\n",
       "      <td>AACR-H0000005</td>\n",
       "      <td>AACR-H0000005-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4564993</th>\n",
       "      <td>LYN</td>\n",
       "      <td>ENSG00000254087.3</td>\n",
       "      <td>ENST00000519728.1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>56912024</td>\n",
       "      <td>56912024</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>3783020.0</td>\n",
       "      <td>3783020.0</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSMUST00000041377.13:c.1252G&gt;C</td>\n",
       "      <td>G418R</td>\n",
       "      <td>COSMIC-H1771600</td>\n",
       "      <td>COSMIC-H1771600-M01-V01</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>76559154.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4564994</th>\n",
       "      <td>LYN</td>\n",
       "      <td>ENSG00000254087.3</td>\n",
       "      <td>ENST00000519728.1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>56912024</td>\n",
       "      <td>56912024</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>3783020.0</td>\n",
       "      <td>3783022.0</td>\n",
       "      <td>GGC</td>\n",
       "      <td>AGG</td>\n",
       "      <td>ENSMUST00000041377.13:c.1252_1254GGC&gt;AGG</td>\n",
       "      <td>G418R</td>\n",
       "      <td>COSMIC-H1771600</td>\n",
       "      <td>COSMIC-H1771600-M01-V02</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>76559154.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4564995</th>\n",
       "      <td>LYN</td>\n",
       "      <td>ENSG00000254087.3</td>\n",
       "      <td>ENST00000519728.1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>56912024</td>\n",
       "      <td>56912024</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>3783020.0</td>\n",
       "      <td>3783022.0</td>\n",
       "      <td>GGC</td>\n",
       "      <td>CGG</td>\n",
       "      <td>ENSMUST00000041377.13:c.1252_1254GGC&gt;CGG</td>\n",
       "      <td>G418R</td>\n",
       "      <td>COSMIC-H1771600</td>\n",
       "      <td>COSMIC-H1771600-M01-V03</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>76559154.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4564996</th>\n",
       "      <td>LYN</td>\n",
       "      <td>ENSG00000254087.3</td>\n",
       "      <td>ENST00000519728.1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>56912024</td>\n",
       "      <td>56912024</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>3783020.0</td>\n",
       "      <td>3783022.0</td>\n",
       "      <td>GGC</td>\n",
       "      <td>CGT</td>\n",
       "      <td>ENSMUST00000041377.13:c.1252_1254GGC&gt;CGT</td>\n",
       "      <td>G418R</td>\n",
       "      <td>COSMIC-H1771600</td>\n",
       "      <td>COSMIC-H1771600-M01-V04</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>76559154.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4564997</th>\n",
       "      <td>LYN</td>\n",
       "      <td>ENSG00000254087.3</td>\n",
       "      <td>ENST00000519728.1</td>\n",
       "      <td>chr8</td>\n",
       "      <td>12</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>56912024</td>\n",
       "      <td>56912024</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>3783020.0</td>\n",
       "      <td>3783022.0</td>\n",
       "      <td>GGC</td>\n",
       "      <td>AGA</td>\n",
       "      <td>ENSMUST00000041377.13:c.1252_1254GGC&gt;AGA</td>\n",
       "      <td>G418R</td>\n",
       "      <td>COSMIC-H1771600</td>\n",
       "      <td>COSMIC-H1771600-M01-V05</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>76559154.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4564998 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name_h          gene_id_h            tx_id_h  chr_h  exon_num_h  \\\n",
       "0              KRAS  ENSG00000133703.7  ENST00000256078.4  chr12           4   \n",
       "1              KRAS  ENSG00000133703.7  ENST00000256078.4  chr12           4   \n",
       "2              BRAF  ENSG00000157764.8  ENST00000288602.6   chr7          18   \n",
       "3            PIK3CA  ENSG00000121879.3  ENST00000263967.3   chr3          20   \n",
       "4            PIK3CA  ENSG00000121879.3  ENST00000263967.3   chr3          20   \n",
       "...             ...                ...                ...    ...         ...   \n",
       "4564993         LYN  ENSG00000254087.3  ENST00000519728.1   chr8          12   \n",
       "4564994         LYN  ENSG00000254087.3  ENST00000519728.1   chr8          12   \n",
       "4564995         LYN  ENSG00000254087.3  ENST00000519728.1   chr8          12   \n",
       "4564996         LYN  ENSG00000254087.3  ENST00000519728.1   chr8          12   \n",
       "4564997         LYN  ENSG00000254087.3  ENST00000519728.1   chr8          12   \n",
       "\n",
       "        strand_h  match    start_h      end_h ref_seq_h  ...      start_m  \\\n",
       "0              -   True   25398284   25398284         C  ...  145192497.0   \n",
       "1              -   True   25398284   25398284         C  ...  145192497.0   \n",
       "2              -   True  140453136  140453136         A  ...   39604717.0   \n",
       "3              +   True  178936091  178936091         G  ...   32502197.0   \n",
       "4              +   True  178952085  178952085         A  ...   32516913.0   \n",
       "...          ...    ...        ...        ...       ...  ...          ...   \n",
       "4564993        +   True   56912024   56912024         G  ...    3783020.0   \n",
       "4564994        +   True   56912024   56912024         G  ...    3783020.0   \n",
       "4564995        +   True   56912024   56912024         G  ...    3783020.0   \n",
       "4564996        +   True   56912024   56912024         G  ...    3783020.0   \n",
       "4564997        +   True   56912024   56912024         G  ...    3783020.0   \n",
       "\n",
       "               end_m ref_seq_m alt_seq_m  \\\n",
       "0        145192497.0         C         T   \n",
       "1        145192497.0         C         A   \n",
       "2         39604717.0         A         T   \n",
       "3         32502197.0         G         A   \n",
       "4         32516913.0         A         G   \n",
       "...              ...       ...       ...   \n",
       "4564993    3783020.0         G         C   \n",
       "4564994    3783022.0       GGC       AGG   \n",
       "4564995    3783022.0       GGC       CGG   \n",
       "4564996    3783022.0       GGC       CGT   \n",
       "4564997    3783022.0       GGC       AGA   \n",
       "\n",
       "                                          HGVSc_m HGVSp_m         ID_human  \\\n",
       "0                    ENSMUST00000111710.8:c.35G>A    G12D    AACR-H0000001   \n",
       "1                    ENSMUST00000111710.8:c.35G>T    G12V    AACR-H0000002   \n",
       "2                 ENSMUST00000002487.15:c.1910T>A   V637E    AACR-H0000003   \n",
       "3                 ENSMUST00000029201.14:c.1633G>A   E545K    AACR-H0000004   \n",
       "4                 ENSMUST00000029201.14:c.3140A>G  H1047R    AACR-H0000005   \n",
       "...                                           ...     ...              ...   \n",
       "4564993           ENSMUST00000041377.13:c.1252G>C   G418R  COSMIC-H1771600   \n",
       "4564994  ENSMUST00000041377.13:c.1252_1254GGC>AGG   G418R  COSMIC-H1771600   \n",
       "4564995  ENSMUST00000041377.13:c.1252_1254GGC>CGG   G418R  COSMIC-H1771600   \n",
       "4564996  ENSMUST00000041377.13:c.1252_1254GGC>CGT   G418R  COSMIC-H1771600   \n",
       "4564997  ENSMUST00000041377.13:c.1252_1254GGC>AGA   G418R  COSMIC-H1771600   \n",
       "\n",
       "                        ID_mouse Database       ID_db  \n",
       "0          AACR-H0000001-M01-V01     AACR         NaN  \n",
       "1          AACR-H0000002-M01-V01     AACR         NaN  \n",
       "2          AACR-H0000003-M01-V01     AACR         NaN  \n",
       "3          AACR-H0000004-M01-V01     AACR         NaN  \n",
       "4          AACR-H0000005-M01-V01     AACR         NaN  \n",
       "...                          ...      ...         ...  \n",
       "4564993  COSMIC-H1771600-M01-V01   COSMIC  76559154.0  \n",
       "4564994  COSMIC-H1771600-M01-V02   COSMIC  76559154.0  \n",
       "4564995  COSMIC-H1771600-M01-V03   COSMIC  76559154.0  \n",
       "4564996  COSMIC-H1771600-M01-V04   COSMIC  76559154.0  \n",
       "4564997  COSMIC-H1771600-M01-V05   COSMIC  76559154.0  \n",
       "\n",
       "[4564998 rows x 46 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_binded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded = df_binded[df_binded['class']!=5].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/df_result_cleaned_v2.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/df_result_cleaned_v2.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Add columns indicating BE/PE pre-computed or not  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_be = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/BEGG/be_result_all.csv')\n",
    "df_pe = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_result_all.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pe['ID'] = [x[:-4] for x in list(df_pe.iloc[:25885,:]['ID'])] + list(df_pe.iloc[25885:,:]['ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_pe = set(df_pe['ID'])\n",
    "set_be = set(df_be['ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded['pe_human'] = df_binded['ID_human'].isin(set_pe)\n",
    "df_binded['pe_mouse'] = df_binded['ID_mouse'].isin(set_pe)\n",
    "df_binded['be_human'] = df_binded['ID_human'].isin(set_be)\n",
    "df_binded['be_mouse'] = df_binded['ID_mouse'].isin(set_be)\n",
    "df_binded['expanded'] = df_binded['ID_human'].isin(set_pe) | df_binded['ID_mouse'].isin(set_pe) | df_binded['ID_human'].isin(set_be) |  df_binded['ID_mouse'].isin(set_be)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded = df_binded.sort_values(['gene_name_h','type_h','gene_name_m','class','expanded','Database','ID_human'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### - Drop duplicates  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded = df_binded.drop_duplicates(subset=['gene_name_h',\n",
    " 'gene_id_h',\n",
    " 'tx_id_h',\n",
    " 'chr_h',\n",
    " 'exon_num_h',\n",
    " 'strand_h',\n",
    " 'match',\n",
    " 'start_h',\n",
    " 'end_h',\n",
    " 'ref_seq_h',\n",
    " 'alt_seq_h',\n",
    " 'HGVSc_h',\n",
    " 'HGVSp_h',\n",
    " 'classification_h',\n",
    " 'exon_h',\n",
    " 'type_h',\n",
    " 'status',\n",
    " 'class',\n",
    " 'statement',\n",
    " 'flank_size_left',\n",
    " 'flank_size_right',\n",
    " 'gene_name_m',\n",
    " 'gene_id_m',\n",
    " 'tx_id_m',\n",
    " 'chr_m',\n",
    " 'exon_num_m',\n",
    " 'strand_m',\n",
    " 'type_m',\n",
    " 'classification_m',\n",
    " 'exon_m',\n",
    " 'start_m_ori',\n",
    " 'end_m_ori',\n",
    " 'ref_seq_m_ori',\n",
    " 'alt_seq_m_ori',\n",
    " 'HGVSc_m_ori',\n",
    " 'HGVSp_m_ori',\n",
    " 'start_m',\n",
    " 'end_m',\n",
    " 'ref_seq_m',\n",
    " 'alt_seq_m',\n",
    " 'HGVSc_m',\n",
    " 'HGVSp_m',\n",
    " 'Database',\n",
    " 'ID_db',\n",
    " 'pe_human',\n",
    " 'pe_mouse',\n",
    " 'be_human',\n",
    " 'be_mouse',\n",
    " 'expanded',\n",
    " 'flank_size'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded = df_binded.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/df_result_cleaned_v2.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Generate genome editing library design input  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- human prime editing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_h_anno, path_m_anno = '/Users/kexindong/Documents/GitHub/Database/Genecode/gencode_v19_GRCh37.db', '/Users/kexindong/Documents/GitHub/Database/Genecode/gencode_vm33_GRCm39.db'\n",
    "# remember to replace the paths with yours\n",
    "db_h = h2m.anno_loader(path_h_anno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/rl/65f1zcd575n092yn3xlqlqfw0000gn/T/ipykernel_17689/2865124907.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  df_pe_h['Chromosome'] = [x[3:] for x in df_pe_h['Chromosome']].copy()\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/aacr_input.csv')\n",
    "list_of_id = list(df[df['count']>20]['index'].unique())\n",
    "list_of_id = [x[:-4] for x in list_of_id]\n",
    "df_pe = df_binded[df_binded['ID_human'].isin(list_of_id)].drop_duplicates().reset_index(drop=True)\n",
    "df_pe.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/subset_human.csv')\n",
    "df_pe_h = df_pe[['gene_name_h','tx_id_h','start_h','end_h','ref_seq_h','alt_seq_h','type_h','ID_human','chr_h']]\n",
    "df_pe_h.columns = ['Hugo_Symbol','tx_id_h','Start_Position','End_Position', 'Reference_Allele','Tumor_Seq_Allele2','Variant_Type','ID','Chromosome']\n",
    "df_pe_h\n",
    "df_pe_h['Chromosome'] = [x[3:] for x in df_pe_h['Chromosome']].copy()\n",
    "df_pe_h.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_human.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- human base editing  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "def be_amenable(df):\n",
    "    set_be = {('C','T'),\n",
    "              ('T','C'),\n",
    "              ('A','G'),\n",
    "              ('G','A'),}\n",
    "    df = df[df['Variant_Type']=='SNP'].reset_index(drop=True)\n",
    "    df = df[[(x,y) in set(set_be) for x,y in zip(df['Reference_Allele'],df['Tumor_Seq_Allele2'])]].reset_index(drop=True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "be_amenable(df_pe_h).to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/BEGG/be_human.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- mouse prime editing  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000001-M01-V01</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000002-M01-V01</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Braf</td>\n",
       "      <td>ENSMUST00000002487.15</td>\n",
       "      <td>39604717.0</td>\n",
       "      <td>39604717.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000003-M01-V01</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pik3ca</td>\n",
       "      <td>ENSMUST00000029201.14</td>\n",
       "      <td>32502197.0</td>\n",
       "      <td>32502197.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000004-M01-V01</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Pik3ca</td>\n",
       "      <td>ENSMUST00000029201.14</td>\n",
       "      <td>32516913.0</td>\n",
       "      <td>32516913.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000005-M01-V01</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4704</th>\n",
       "      <td>Ntrk2</td>\n",
       "      <td>ENSMUST00000079828.7</td>\n",
       "      <td>59022231.0</td>\n",
       "      <td>59022231.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005026-M01-V01</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4705</th>\n",
       "      <td>Egfr</td>\n",
       "      <td>ENSMUST00000020329.13</td>\n",
       "      <td>16810181.0</td>\n",
       "      <td>16810181.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005027-M01-V01</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4706</th>\n",
       "      <td>Xpo1</td>\n",
       "      <td>ENSMUST00000102869.8</td>\n",
       "      <td>23235114.0</td>\n",
       "      <td>23235114.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005028-M01-V01</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4707</th>\n",
       "      <td>Ep300</td>\n",
       "      <td>ENSMUST00000068387.11</td>\n",
       "      <td>81525575.0</td>\n",
       "      <td>81525576.0</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>INS</td>\n",
       "      <td>AACR-H0005029-M01-V01</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4708</th>\n",
       "      <td>Bcl9</td>\n",
       "      <td>ENSMUST00000046521.14</td>\n",
       "      <td>97117238.0</td>\n",
       "      <td>97117238.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005030-M01-V01</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4709 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Hugo_Symbol                tx_id_m  Start_Position  End_Position  \\\n",
       "0           Kras   ENSMUST00000111710.8     145192497.0   145192497.0   \n",
       "1           Kras   ENSMUST00000111710.8     145192497.0   145192497.0   \n",
       "2           Braf  ENSMUST00000002487.15      39604717.0    39604717.0   \n",
       "3         Pik3ca  ENSMUST00000029201.14      32502197.0    32502197.0   \n",
       "4         Pik3ca  ENSMUST00000029201.14      32516913.0    32516913.0   \n",
       "...          ...                    ...             ...           ...   \n",
       "4704       Ntrk2   ENSMUST00000079828.7      59022231.0    59022231.0   \n",
       "4705        Egfr  ENSMUST00000020329.13      16810181.0    16810181.0   \n",
       "4706        Xpo1   ENSMUST00000102869.8      23235114.0    23235114.0   \n",
       "4707       Ep300  ENSMUST00000068387.11      81525575.0    81525576.0   \n",
       "4708        Bcl9  ENSMUST00000046521.14      97117238.0    97117238.0   \n",
       "\n",
       "     Reference_Allele Tumor_Seq_Allele2 Variant_Type                     ID  \\\n",
       "0                   C                 T          SNP  AACR-H0000001-M01-V01   \n",
       "1                   C                 A          SNP  AACR-H0000002-M01-V01   \n",
       "2                   A                 T          SNP  AACR-H0000003-M01-V01   \n",
       "3                   G                 A          SNP  AACR-H0000004-M01-V01   \n",
       "4                   A                 G          SNP  AACR-H0000005-M01-V01   \n",
       "...               ...               ...          ...                    ...   \n",
       "4704                C                 T          SNP  AACR-H0005026-M01-V01   \n",
       "4705                C                 T          SNP  AACR-H0005027-M01-V01   \n",
       "4706                A                 G          SNP  AACR-H0005028-M01-V01   \n",
       "4707                -                 A          INS  AACR-H0005029-M01-V01   \n",
       "4708                C                 T          SNP  AACR-H0005030-M01-V01   \n",
       "\n",
       "     Chromosome  \n",
       "0             6  \n",
       "1             6  \n",
       "2             6  \n",
       "3             3  \n",
       "4             3  \n",
       "...         ...  \n",
       "4704         13  \n",
       "4705         11  \n",
       "4706         11  \n",
       "4707         15  \n",
       "4708          3  \n",
       "\n",
       "[4709 rows x 9 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pe_m = df_binded[df_binded['ID_human'].isin(list_of_id)].reset_index(drop=True)\n",
    "df_pe_m = df_pe_m[df_pe_m['status']==True].reset_index(drop=True)\n",
    "df_pe_m = df_pe_m[['gene_name_m','tx_id_m','start_m','end_m','ref_seq_m','alt_seq_m','type_m','ID_mouse','chr_m']]\n",
    "df_pe_m = df_pe_m.rename(columns={'ID_mouse':'ID'})\n",
    "df_pe_m['chr_m'] = [x[3:] for x in df_pe_m['chr_m']]\n",
    "df_pe_m.columns =['Hugo_Symbol','tx_id_m','Start_Position','End_Position','Reference_Allele','Tumor_Seq_Allele2','Variant_Type','ID','Chromosome']\n",
    "df_pe_m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pe_m.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_mouse.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- mouse base editing  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "be_amenable(df_pe_m).to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/BEGG/be_mouse.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## (3) Run PEGG and BEGG  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kexindong/Documents/GitHub/h2m-public/.conda/lib/python3.10/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from pegg import prime\n",
    "from pegg import base\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_of_gene = list(df['tx_id_h'].unique())\n",
    "list_chr = []\n",
    "for x in list_of_gene:\n",
    "    try:\n",
    "        chr = str(db_h[x].chrom)[3:]\n",
    "    except:\n",
    "        chr = None\n",
    "    list_chr.append(chr)\n",
    "dict_of_chr = dict(zip(list_of_gene, list_chr))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '/Users/kexindong/Documents/GitHub/Output/h2m_database/sample_data/'\n",
    "df = pd.read_csv(f'{path}db_1_sample_1000.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_h_ref, path_m_ref = '/Users/kexindong/Documents/GitHub/Database/RefGenome/ncbi-2023-09-12/GCF_000001405.25_GRCh37.p13_genomic.fna.gz', '/Users/kexindong/Documents/GitHub/Database/RefGenome/mouse-2023-09-13/GCF_000001635.27_GRCm39_genomic.fna.gz'\n",
    "chrom_dict_h, i = prime.genome_loader(path_h_ref)\n",
    "chrom_dict_m, i = prime.genome_loader(path_m_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['gene_name_h', 'gene_id_h', 'tx_id_h', 'chr_h', 'exon_num_h',\n",
       "       'strand_h', 'match', 'start_h', 'end_h', 'ref_seq_h', 'alt_seq_h',\n",
       "       'HGVSc_h', 'HGVSp_h', 'classification_h', 'exon_h', 'type_h', 'status',\n",
       "       'class', 'statement', 'flank_size_left', 'flank_size_right',\n",
       "       'gene_name_m', 'gene_id_m', 'tx_id_m', 'chr_m', 'exon_num_m',\n",
       "       'strand_m', 'type_m', 'classification_m', 'exon_m', 'start_m_ori',\n",
       "       'end_m_ori', 'ref_seq_m_ori', 'alt_seq_m_ori', 'HGVSc_m_ori',\n",
       "       'HGVSp_m_ori', 'start_m', 'end_m', 'ref_seq_m', 'alt_seq_m', 'HGVSc_m',\n",
       "       'HGVSp_m', 'ID_human', 'ID_mouse', 'Database'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['chr_h'] = [x[3:] for x in df['chr_h']]\n",
    "df['chr_m'] = [x[3:] for x in df['chr_m']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_h = df[[\n",
    "    'gene_name_h','tx_id_h','start_h', 'end_h', 'ref_seq_h','alt_seq_h','type_h','ID_human',\n",
    "       'chr_h'\n",
    "]]\n",
    "\n",
    "df_m = df[[\n",
    "    'gene_name_m','tx_id_m','start_m', 'end_m', 'ref_seq_m','alt_seq_m','type_m','ID_mouse',\n",
    "       'chr_m'\n",
    "]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_h.columns = ['Hugo_Symbol','tx_id_h','Start_Position','End_Position', 'Reference_Allele','Tumor_Seq_Allele2','Variant_Type','ID','Chromosome']\n",
    "df_m.columns = ['Hugo_Symbol','tx_id_h','Start_Position','End_Position', 'Reference_Allele','Tumor_Seq_Allele2','Variant_Type','ID','Chromosome']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [],
   "source": [
    "def be_amenable(df):\n",
    "    set_be = {('C','T'),\n",
    "              ('T','C'),\n",
    "              ('A','G'),\n",
    "              ('G','A'),}\n",
    "    df = df[df['Variant_Type']=='SNP'].reset_index(drop=True)\n",
    "    df = df[[(x,y) in set(set_be) for x,y in zip(df['Reference_Allele'],df['Tumor_Seq_Allele2'])]].reset_index(drop=True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_m = df_m.dropna(subset='ID').reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Bcl6</td>\n",
       "      <td>ENSMUST00000023151.6</td>\n",
       "      <td>23791861.0</td>\n",
       "      <td>23791861.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000996-M01-V01</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Esr1</td>\n",
       "      <td>ENSMUST00000105590.8</td>\n",
       "      <td>4662747.0</td>\n",
       "      <td>4662747.0</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002687-M01-V01</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Keap1</td>\n",
       "      <td>ENSMUST00000164812.8</td>\n",
       "      <td>21145012.0</td>\n",
       "      <td>21145012.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002754-M01-V01</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Mlh1</td>\n",
       "      <td>ENSMUST00000035079.10</td>\n",
       "      <td>111057546.0</td>\n",
       "      <td>111057546.0</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0014742-M01-V01</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Tsc2</td>\n",
       "      <td>ENSMUST00000227745.2</td>\n",
       "      <td>24842559.0</td>\n",
       "      <td>24842559.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0017727-M01-V01</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>773</th>\n",
       "      <td>Atic</td>\n",
       "      <td>ENSMUST00000027384.6</td>\n",
       "      <td>71615930.0</td>\n",
       "      <td>71615930.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>COSMIC-H1601022-M01-V01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>774</th>\n",
       "      <td>Atic</td>\n",
       "      <td>ENSMUST00000027384.6</td>\n",
       "      <td>71608130.0</td>\n",
       "      <td>71608130.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>COSMIC-H1601059-M01-V01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>775</th>\n",
       "      <td>Atic</td>\n",
       "      <td>ENSMUST00000027384.6</td>\n",
       "      <td>71601743.0</td>\n",
       "      <td>71601743.0</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>COSMIC-H1603731-M01-V01</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>776</th>\n",
       "      <td>Prdm2</td>\n",
       "      <td>ENSMUST00000105778.8</td>\n",
       "      <td>142859478.0</td>\n",
       "      <td>142859478.0</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>COSMIC-H1677359-M01-V01</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>777</th>\n",
       "      <td>Prdm2</td>\n",
       "      <td>ENSMUST00000105778.8</td>\n",
       "      <td>142858921.0</td>\n",
       "      <td>142858922.0</td>\n",
       "      <td>CG</td>\n",
       "      <td>AT</td>\n",
       "      <td>DNP</td>\n",
       "      <td>COSMIC-H1683868-M01-V02</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>778 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Hugo_Symbol                tx_id_h  Start_Position  End_Position  \\\n",
       "0          Bcl6   ENSMUST00000023151.6      23791861.0    23791861.0   \n",
       "1          Esr1   ENSMUST00000105590.8       4662747.0     4662747.0   \n",
       "2         Keap1   ENSMUST00000164812.8      21145012.0    21145012.0   \n",
       "3          Mlh1  ENSMUST00000035079.10     111057546.0   111057546.0   \n",
       "4          Tsc2   ENSMUST00000227745.2      24842559.0    24842559.0   \n",
       "..          ...                    ...             ...           ...   \n",
       "773        Atic   ENSMUST00000027384.6      71615930.0    71615930.0   \n",
       "774        Atic   ENSMUST00000027384.6      71608130.0    71608130.0   \n",
       "775        Atic   ENSMUST00000027384.6      71601743.0    71601743.0   \n",
       "776       Prdm2   ENSMUST00000105778.8     142859478.0   142859478.0   \n",
       "777       Prdm2   ENSMUST00000105778.8     142858921.0   142858922.0   \n",
       "\n",
       "    Reference_Allele Tumor_Seq_Allele2 Variant_Type                       ID  \\\n",
       "0                  C                 A          SNP    AACR-H0000996-M01-V01   \n",
       "1                  T                 C          SNP    AACR-H0002687-M01-V01   \n",
       "2                  C                 A          SNP    AACR-H0002754-M01-V01   \n",
       "3                  T                 C          SNP    AACR-H0014742-M01-V01   \n",
       "4                  G                 A          SNP    AACR-H0017727-M01-V01   \n",
       "..               ...               ...          ...                      ...   \n",
       "773                A                 G          SNP  COSMIC-H1601022-M01-V01   \n",
       "774                G                 A          SNP  COSMIC-H1601059-M01-V01   \n",
       "775                A                 C          SNP  COSMIC-H1603731-M01-V01   \n",
       "776                T                 C          SNP  COSMIC-H1677359-M01-V01   \n",
       "777               CG                AT          DNP  COSMIC-H1683868-M01-V02   \n",
       "\n",
       "    Chromosome  \n",
       "0           16  \n",
       "1           10  \n",
       "2            9  \n",
       "3            9  \n",
       "4           17  \n",
       "..         ...  \n",
       "773          1  \n",
       "774          1  \n",
       "775          1  \n",
       "776          4  \n",
       "777          4  \n",
       "\n",
       "[778 rows x 9 columns]"
      ]
     },
     "execution_count": 194,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_h = be_amenable(df_h)\n",
    "df_m = be_amenable(df_m)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 198,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_m['Reference_Allele'] = [str(x) for x in df_m['Reference_Allele']]\n",
    "df_m['Tumor_Seq_Allele2'] = [str(x) for x in df_m['Tumor_Seq_Allele2']]\n",
    "df_m['Start_Position'] = [int(x) for x in df_m['Start_Position']]\n",
    "df_m['End_Position'] = [int(x) for x in df_m['End_Position']]\n",
    "df_h['Reference_Allele'] = [str(x) for x in df_h['Reference_Allele']]\n",
    "df_h['Tumor_Seq_Allele2'] = [str(x) for x in df_h['Tumor_Seq_Allele2']]\n",
    "df_h['Start_Position'] = [int(x) for x in df_h['Start_Position']]\n",
    "df_h['End_Position'] = [int(x) for x in df_h['End_Position']]\n",
    "df_be_h = base.run_base(df_h, 'cBioPortal', chrom_dict_h, PAM='NGN',filtration='ABE+CBE', ideal_edit_window=[4, 8], auto_SNP_filter=True, proto_size=19,context_size=120, RE_sites=None, polyT_threshold=4, before_proto_context=5,sensor_length=40, sensor_orientation='reverse-complement', sensor=True)\n",
    "df_be_m = base.run_base(df_m, 'cBioPortal', chrom_dict_m, PAM='NGN',filtration='ABE+CBE', ideal_edit_window=[4, 8], auto_SNP_filter=True, proto_size=19,context_size=120, RE_sites=None, polyT_threshold=4, before_proto_context=5,sensor_length=40, sensor_orientation='reverse-complement', sensor=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_be = pd.concat([df_be_h, df_be_m])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_be.to_csv(f'{path}db_2_be.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "PEGG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pe_input = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_human.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pegg = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_human_result.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_of_pe_inluced_id = list(df_pe_input['ID'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pegg_sub = df_pegg[df_pegg['ID']=='AACR-H0002687-M01'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>...</th>\n",
       "      <th>indel_size</th>\n",
       "      <th>PEGG2_Score</th>\n",
       "      <th>RF_Score</th>\n",
       "      <th>pegRNA_rank</th>\n",
       "      <th>sensor_wt</th>\n",
       "      <th>sensor_alt</th>\n",
       "      <th>sensor_orientation</th>\n",
       "      <th>sensor_error</th>\n",
       "      <th>contains_polyT_terminator</th>\n",
       "      <th>pegRNA_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>41</td>\n",
       "      <td>ESR1</td>\n",
       "      <td>ENSMUST00000105590.8</td>\n",
       "      <td>152129399</td>\n",
       "      <td>152129399</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002687-M01</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>23.348941</td>\n",
       "      <td>37.452766</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGACAGCTGCGGC...</td>\n",
       "      <td>GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGGCAGCTGCGGC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>41</td>\n",
       "      <td>ESR1</td>\n",
       "      <td>ENSMUST00000105590.8</td>\n",
       "      <td>152129399</td>\n",
       "      <td>152129399</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002687-M01</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>23.112691</td>\n",
       "      <td>36.314398</td>\n",
       "      <td>2.0</td>\n",
       "      <td>GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGACAGCTGCGGC...</td>\n",
       "      <td>GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGGCAGCTGCGGC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>41</td>\n",
       "      <td>ESR1</td>\n",
       "      <td>ENSMUST00000105590.8</td>\n",
       "      <td>152129399</td>\n",
       "      <td>152129399</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002687-M01</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>22.681445</td>\n",
       "      <td>36.314398</td>\n",
       "      <td>3.0</td>\n",
       "      <td>GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGACAGCTGCGGC...</td>\n",
       "      <td>GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGGCAGCTGCGGC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>41</td>\n",
       "      <td>ESR1</td>\n",
       "      <td>ENSMUST00000105590.8</td>\n",
       "      <td>152129399</td>\n",
       "      <td>152129399</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002687-M01</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>23.632858</td>\n",
       "      <td>36.145389</td>\n",
       "      <td>4.0</td>\n",
       "      <td>GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGACAGCTGCGGC...</td>\n",
       "      <td>GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGGCAGCTGCGGC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>41</td>\n",
       "      <td>ESR1</td>\n",
       "      <td>ENSMUST00000105590.8</td>\n",
       "      <td>152129399</td>\n",
       "      <td>152129399</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002687-M01</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>22.780257</td>\n",
       "      <td>33.087854</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGACAGCTGCGGC...</td>\n",
       "      <td>GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGGCAGCTGCGGC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_209</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   mutation_idx Hugo_Symbol               tx_id_h  Start_Position  \\\n",
       "0            41        ESR1  ENSMUST00000105590.8       152129399   \n",
       "1            41        ESR1  ENSMUST00000105590.8       152129399   \n",
       "2            41        ESR1  ENSMUST00000105590.8       152129399   \n",
       "3            41        ESR1  ENSMUST00000105590.8       152129399   \n",
       "4            41        ESR1  ENSMUST00000105590.8       152129399   \n",
       "\n",
       "   End_Position Reference_Allele Tumor_Seq_Allele2 Variant_Type  \\\n",
       "0     152129399                T                 C          SNP   \n",
       "1     152129399                T                 C          SNP   \n",
       "2     152129399                T                 C          SNP   \n",
       "3     152129399                T                 C          SNP   \n",
       "4     152129399                T                 C          SNP   \n",
       "\n",
       "                  ID Chromosome  ...  indel_size  PEGG2_Score   RF_Score  \\\n",
       "0  AACR-H0002687-M01          6  ...           0    23.348941  37.452766   \n",
       "1  AACR-H0002687-M01          6  ...           0    23.112691  36.314398   \n",
       "2  AACR-H0002687-M01          6  ...           0    22.681445  36.314398   \n",
       "3  AACR-H0002687-M01          6  ...           0    23.632858  36.145389   \n",
       "4  AACR-H0002687-M01          6  ...           0    22.780257  33.087854   \n",
       "\n",
       "  pegRNA_rank                                          sensor_wt  \\\n",
       "0         1.0  GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGACAGCTGCGGC...   \n",
       "1         2.0  GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGACAGCTGCGGC...   \n",
       "2         3.0  GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGACAGCTGCGGC...   \n",
       "3         4.0  GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGACAGCTGCGGC...   \n",
       "4         5.0  GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGACAGCTGCGGC...   \n",
       "\n",
       "                                          sensor_alt  sensor_orientation  \\\n",
       "0  GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGGCAGCTGCGGC...  reverse-complement   \n",
       "1  GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGGCAGCTGCGGC...  reverse-complement   \n",
       "2  GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGGCAGCTGCGGC...  reverse-complement   \n",
       "3  GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGGCAGCTGCGGC...  reverse-complement   \n",
       "4  GGGCACCTGCTGGCCGTGGGGCTGCAGGAAAGGCGGCAGCTGCGGC...  reverse-complement   \n",
       "\n",
       "  sensor_error  contains_polyT_terminator   pegRNA_id  \n",
       "0     No Error                      False  pegRNA_205  \n",
       "1     No Error                      False  pegRNA_206  \n",
       "2     No Error                      False  pegRNA_207  \n",
       "3     No Error                      False  pegRNA_208  \n",
       "4     No Error                      False  pegRNA_209  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pegg_sub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_m = df_m[[x[:-4] in list_of_pe_inluced_id for x in df_m['ID']]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Efficiency predictions on non-NGG PAMs are un-tested\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>...</th>\n",
       "      <th>indel_size</th>\n",
       "      <th>PEGG2_Score</th>\n",
       "      <th>RF_Score</th>\n",
       "      <th>pegRNA_rank</th>\n",
       "      <th>sensor_wt</th>\n",
       "      <th>sensor_alt</th>\n",
       "      <th>sensor_orientation</th>\n",
       "      <th>sensor_error</th>\n",
       "      <th>contains_polyT_terminator</th>\n",
       "      <th>pegRNA_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Esr1</td>\n",
       "      <td>ENSMUST00000105590.8</td>\n",
       "      <td>4662747</td>\n",
       "      <td>4662747</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002687-M01-V01</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>18.882586</td>\n",
       "      <td>27.552490</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CGCTGATGCTGCTGCACCCGCCGCCGCAGCTGTCTCCTTTCCTGCA...</td>\n",
       "      <td>CGCTGATGCTGCTGCACCCGCCGCCGCAGCTGCCTCCTTTCCTGCA...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>Esr1</td>\n",
       "      <td>ENSMUST00000105590.8</td>\n",
       "      <td>4662747</td>\n",
       "      <td>4662747</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002687-M01-V01</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>22.159396</td>\n",
       "      <td>27.281115</td>\n",
       "      <td>2.0</td>\n",
       "      <td>AGCCCGCTGATGCTGCTGCACCCGCCGCCGCAGCTGTCTCCTTTCC...</td>\n",
       "      <td>AGCCCGCTGATGCTGCTGCACCCGCCGCCGCAGCTGCCTCCTTTCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>Esr1</td>\n",
       "      <td>ENSMUST00000105590.8</td>\n",
       "      <td>4662747</td>\n",
       "      <td>4662747</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002687-M01-V01</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>22.371025</td>\n",
       "      <td>27.281115</td>\n",
       "      <td>3.0</td>\n",
       "      <td>AGCCCGCTGATGCTGCTGCACCCGCCGCCGCAGCTGTCTCCTTTCC...</td>\n",
       "      <td>AGCCCGCTGATGCTGCTGCACCCGCCGCCGCAGCTGCCTCCTTTCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>Esr1</td>\n",
       "      <td>ENSMUST00000105590.8</td>\n",
       "      <td>4662747</td>\n",
       "      <td>4662747</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002687-M01-V01</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>18.820443</td>\n",
       "      <td>27.081155</td>\n",
       "      <td>4.0</td>\n",
       "      <td>CGCTGATGCTGCTGCACCCGCCGCCGCAGCTGTCTCCTTTCCTGCA...</td>\n",
       "      <td>CGCTGATGCTGCTGCACCCGCCGCCGCAGCTGCCTCCTTTCCTGCA...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>Esr1</td>\n",
       "      <td>ENSMUST00000105590.8</td>\n",
       "      <td>4662747</td>\n",
       "      <td>4662747</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002687-M01-V01</td>\n",
       "      <td>10</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>23.026340</td>\n",
       "      <td>23.273683</td>\n",
       "      <td>5.0</td>\n",
       "      <td>AGCCCGCTGATGCTGCTGCACCCGCCGCCGCAGCTGTCTCCTTTCC...</td>\n",
       "      <td>AGCCCGCTGATGCTGCTGCACCCGCCGCCGCAGCTGCCTCCTTTCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   mutation_idx Hugo_Symbol               tx_id_h  Start_Position  \\\n",
       "0             0        Esr1  ENSMUST00000105590.8         4662747   \n",
       "1             0        Esr1  ENSMUST00000105590.8         4662747   \n",
       "2             0        Esr1  ENSMUST00000105590.8         4662747   \n",
       "3             0        Esr1  ENSMUST00000105590.8         4662747   \n",
       "4             0        Esr1  ENSMUST00000105590.8         4662747   \n",
       "\n",
       "   End_Position Reference_Allele Tumor_Seq_Allele2 Variant_Type  \\\n",
       "0       4662747                T                 C          SNP   \n",
       "1       4662747                T                 C          SNP   \n",
       "2       4662747                T                 C          SNP   \n",
       "3       4662747                T                 C          SNP   \n",
       "4       4662747                T                 C          SNP   \n",
       "\n",
       "                      ID Chromosome  ...  indel_size  PEGG2_Score   RF_Score  \\\n",
       "0  AACR-H0002687-M01-V01         10  ...           0    18.882586  27.552490   \n",
       "1  AACR-H0002687-M01-V01         10  ...           0    22.159396  27.281115   \n",
       "2  AACR-H0002687-M01-V01         10  ...           0    22.371025  27.281115   \n",
       "3  AACR-H0002687-M01-V01         10  ...           0    18.820443  27.081155   \n",
       "4  AACR-H0002687-M01-V01         10  ...           0    23.026340  23.273683   \n",
       "\n",
       "  pegRNA_rank                                          sensor_wt  \\\n",
       "0         1.0  CGCTGATGCTGCTGCACCCGCCGCCGCAGCTGTCTCCTTTCCTGCA...   \n",
       "1         2.0  AGCCCGCTGATGCTGCTGCACCCGCCGCCGCAGCTGTCTCCTTTCC...   \n",
       "2         3.0  AGCCCGCTGATGCTGCTGCACCCGCCGCCGCAGCTGTCTCCTTTCC...   \n",
       "3         4.0  CGCTGATGCTGCTGCACCCGCCGCCGCAGCTGTCTCCTTTCCTGCA...   \n",
       "4         5.0  AGCCCGCTGATGCTGCTGCACCCGCCGCCGCAGCTGTCTCCTTTCC...   \n",
       "\n",
       "                                          sensor_alt  sensor_orientation  \\\n",
       "0  CGCTGATGCTGCTGCACCCGCCGCCGCAGCTGCCTCCTTTCCTGCA...  reverse-complement   \n",
       "1  AGCCCGCTGATGCTGCTGCACCCGCCGCCGCAGCTGCCTCCTTTCC...  reverse-complement   \n",
       "2  AGCCCGCTGATGCTGCTGCACCCGCCGCCGCAGCTGCCTCCTTTCC...  reverse-complement   \n",
       "3  CGCTGATGCTGCTGCACCCGCCGCCGCAGCTGCCTCCTTTCCTGCA...  reverse-complement   \n",
       "4  AGCCCGCTGATGCTGCTGCACCCGCCGCCGCAGCTGCCTCCTTTCC...  reverse-complement   \n",
       "\n",
       "  sensor_error  contains_polyT_terminator  pegRNA_id  \n",
       "0     No Error                      False   pegRNA_0  \n",
       "1     No Error                      False   pegRNA_1  \n",
       "2     No Error                      False   pegRNA_2  \n",
       "3     No Error                      False   pegRNA_3  \n",
       "4     No Error                      False   pegRNA_4  \n",
       "\n",
       "[5 rows x 46 columns]"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pe_m = prime.run(df_m, 'cBioPortal', chrom_dict=chrom_dict_m, PAM='NGN',rankby='RF_Score', pegRNAs_per_mut=5, RTT_lengths=[5, 10, 15, 25, 30],PBS_lengths=[8, 10, 13, 15], min_RHA_size=1, RE_sites=None, polyT_threshold=4,proto_size=19, context_size=120, before_proto_context=5, sensor_length=60,sensor_orientation='reverse-complement', sensor=True)\n",
    "df_pe_m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.concat([df_pegg_sub, df_pe_m])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Result Statistics  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (1) Model-able Statistics  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_aacr = pd.read_csv('/Users/gorkordkx/Documents/Output/h2m/1_murine_mutation_database_output/aacr_output/aacr_result_f0.csv')\n",
    "df_cosmic = pd.read_csv('/Users/gorkordkx/Documents/Output/h2m/1_murine_mutation_database_output/cosmic_output/cosmic_result_f0.csv')\n",
    "df_clinvar = pd.read_csv('/Users/gorkordkx/Documents/Output/h2m/1_murine_mutation_database_output/clinvar_output/clinvar_result_f0.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "aacr: 683645 out of 799660 (85.49%) mutations can be modeled.\n",
      "\n",
      "cosmic: 473858 out of 598809 (79.13%) mutations can be modeled.\n",
      "\n",
      "clinvar: 1585810 out of 2021074 (78.46%) mutations can be modeled.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "name_list = ['aacr','cosmic','clinvar']\n",
    "df_list = [df_aacr, df_cosmic, df_clinvar]\n",
    "\n",
    "for x in [0,1,2]:\n",
    "    l_1 = len(df_list[x]['index'].unique())\n",
    "    df_sub = df_list[x][df_list[x]['status']==True]\n",
    "    l_2 = len(df_sub['index'].unique())\n",
    "    print(f'{name_list[x]}: {l_2} out of {l_1} ({l_2/l_1* 100:.2f}%) mutations can be modeled.\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.8340348780487805"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "799660+598809+2021074"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>gene_id_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>chr_h</th>\n",
       "      <th>exon_num_h</th>\n",
       "      <th>strand_h</th>\n",
       "      <th>match</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>...</th>\n",
       "      <th>start_m</th>\n",
       "      <th>end_m</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>alt_seq_m</th>\n",
       "      <th>HGVSc_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "      <th>ID_human</th>\n",
       "      <th>ID_mouse</th>\n",
       "      <th>Database</th>\n",
       "      <th>ID_db</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1127</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENSG00000146648.11</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>28</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55241707</td>\n",
       "      <td>55241707</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000020329.13:c.2161G&gt;A</td>\n",
       "      <td>G721S</td>\n",
       "      <td>AACR-H0001067</td>\n",
       "      <td>AACR-H0001067-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137860</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENSG00000146648.11</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>28</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55241707</td>\n",
       "      <td>55241708</td>\n",
       "      <td>GG</td>\n",
       "      <td>...</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>16841272.0</td>\n",
       "      <td>GG</td>\n",
       "      <td>TC</td>\n",
       "      <td>ENSMUST00000020329.13:c.2161_2162GG&gt;TC</td>\n",
       "      <td>G721S</td>\n",
       "      <td>AACR-H0125830</td>\n",
       "      <td>AACR-H0125830-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>619908</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENSG00000146648.11</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>28</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55241706</td>\n",
       "      <td>55241707</td>\n",
       "      <td>GG</td>\n",
       "      <td>...</td>\n",
       "      <td>16841270.0</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>GG</td>\n",
       "      <td>AA</td>\n",
       "      <td>ENSMUST00000020329.13:c.2160_2161GG&gt;AA</td>\n",
       "      <td>G721S</td>\n",
       "      <td>AACR-H0574620</td>\n",
       "      <td>AACR-H0574620-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1762934</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENSG00000146648.11</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>28</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55241707</td>\n",
       "      <td>55241707</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000020329.13:c.2161G&gt;A</td>\n",
       "      <td>G721S</td>\n",
       "      <td>CLINVAR-H0893912</td>\n",
       "      <td>CLINVAR-H0893912-M01-V01</td>\n",
       "      <td>CLINVAR</td>\n",
       "      <td>16612.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3847616</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENSG00000146648.11</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>28</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55241707</td>\n",
       "      <td>55241707</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000020329.13:c.2161G&gt;A</td>\n",
       "      <td>G721S</td>\n",
       "      <td>COSMIC-H0900805</td>\n",
       "      <td>COSMIC-H0900805-M01-V01</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>24005626.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3863169</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENSG00000146648.11</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>28</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55241707</td>\n",
       "      <td>55241707</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000020329.13:c.2161G&gt;A</td>\n",
       "      <td>G721S</td>\n",
       "      <td>COSMIC-H0928414</td>\n",
       "      <td>COSMIC-H0928414-M01-V01</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>24005626.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3863424</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENSG00000146648.11</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>28</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55241707</td>\n",
       "      <td>55241707</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000020329.13:c.2161G&gt;A</td>\n",
       "      <td>G721S</td>\n",
       "      <td>COSMIC-H0928775</td>\n",
       "      <td>COSMIC-H0928775-M01-V01</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>24005626.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3863812</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENSG00000146648.11</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>28</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55241707</td>\n",
       "      <td>55241707</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000020329.13:c.2161G&gt;A</td>\n",
       "      <td>G721S</td>\n",
       "      <td>COSMIC-H0929331</td>\n",
       "      <td>COSMIC-H0929331-M01-V01</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>24005626.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3863819</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENSG00000146648.11</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>28</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55241707</td>\n",
       "      <td>55241707</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000020329.13:c.2161G&gt;A</td>\n",
       "      <td>G721S</td>\n",
       "      <td>COSMIC-H0929342</td>\n",
       "      <td>COSMIC-H0929342-M01-V01</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>24005626.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3863875</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENSG00000146648.11</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>chr7</td>\n",
       "      <td>28</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55241707</td>\n",
       "      <td>55241707</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>16841271.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000020329.13:c.2161G&gt;A</td>\n",
       "      <td>G721S</td>\n",
       "      <td>COSMIC-H0929424</td>\n",
       "      <td>COSMIC-H0929424-M01-V01</td>\n",
       "      <td>COSMIC</td>\n",
       "      <td>24005626.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene_name_h           gene_id_h            tx_id_h chr_h  exon_num_h  \\\n",
       "1127           EGFR  ENSG00000146648.11  ENST00000275493.2  chr7          28   \n",
       "137860         EGFR  ENSG00000146648.11  ENST00000275493.2  chr7          28   \n",
       "619908         EGFR  ENSG00000146648.11  ENST00000275493.2  chr7          28   \n",
       "1762934        EGFR  ENSG00000146648.11  ENST00000275493.2  chr7          28   \n",
       "3847616        EGFR  ENSG00000146648.11  ENST00000275493.2  chr7          28   \n",
       "...             ...                 ...                ...   ...         ...   \n",
       "3863169        EGFR  ENSG00000146648.11  ENST00000275493.2  chr7          28   \n",
       "3863424        EGFR  ENSG00000146648.11  ENST00000275493.2  chr7          28   \n",
       "3863812        EGFR  ENSG00000146648.11  ENST00000275493.2  chr7          28   \n",
       "3863819        EGFR  ENSG00000146648.11  ENST00000275493.2  chr7          28   \n",
       "3863875        EGFR  ENSG00000146648.11  ENST00000275493.2  chr7          28   \n",
       "\n",
       "        strand_h  match   start_h     end_h ref_seq_h  ...     start_m  \\\n",
       "1127           +   True  55241707  55241707         G  ...  16841271.0   \n",
       "137860         +   True  55241707  55241708        GG  ...  16841271.0   \n",
       "619908         +   True  55241706  55241707        GG  ...  16841270.0   \n",
       "1762934        +   True  55241707  55241707         G  ...  16841271.0   \n",
       "3847616        +   True  55241707  55241707         G  ...  16841271.0   \n",
       "...          ...    ...       ...       ...       ...  ...         ...   \n",
       "3863169        +   True  55241707  55241707         G  ...  16841271.0   \n",
       "3863424        +   True  55241707  55241707         G  ...  16841271.0   \n",
       "3863812        +   True  55241707  55241707         G  ...  16841271.0   \n",
       "3863819        +   True  55241707  55241707         G  ...  16841271.0   \n",
       "3863875        +   True  55241707  55241707         G  ...  16841271.0   \n",
       "\n",
       "              end_m ref_seq_m alt_seq_m  \\\n",
       "1127     16841271.0         G         A   \n",
       "137860   16841272.0        GG        TC   \n",
       "619908   16841271.0        GG        AA   \n",
       "1762934  16841271.0         G         A   \n",
       "3847616  16841271.0         G         A   \n",
       "...             ...       ...       ...   \n",
       "3863169  16841271.0         G         A   \n",
       "3863424  16841271.0         G         A   \n",
       "3863812  16841271.0         G         A   \n",
       "3863819  16841271.0         G         A   \n",
       "3863875  16841271.0         G         A   \n",
       "\n",
       "                                        HGVSc_m HGVSp_m          ID_human  \\\n",
       "1127            ENSMUST00000020329.13:c.2161G>A   G721S     AACR-H0001067   \n",
       "137860   ENSMUST00000020329.13:c.2161_2162GG>TC   G721S     AACR-H0125830   \n",
       "619908   ENSMUST00000020329.13:c.2160_2161GG>AA   G721S     AACR-H0574620   \n",
       "1762934         ENSMUST00000020329.13:c.2161G>A   G721S  CLINVAR-H0893912   \n",
       "3847616         ENSMUST00000020329.13:c.2161G>A   G721S   COSMIC-H0900805   \n",
       "...                                         ...     ...               ...   \n",
       "3863169         ENSMUST00000020329.13:c.2161G>A   G721S   COSMIC-H0928414   \n",
       "3863424         ENSMUST00000020329.13:c.2161G>A   G721S   COSMIC-H0928775   \n",
       "3863812         ENSMUST00000020329.13:c.2161G>A   G721S   COSMIC-H0929331   \n",
       "3863819         ENSMUST00000020329.13:c.2161G>A   G721S   COSMIC-H0929342   \n",
       "3863875         ENSMUST00000020329.13:c.2161G>A   G721S   COSMIC-H0929424   \n",
       "\n",
       "                         ID_mouse Database       ID_db  \n",
       "1127        AACR-H0001067-M01-V01     AACR         NaN  \n",
       "137860      AACR-H0125830-M01-V01     AACR         NaN  \n",
       "619908      AACR-H0574620-M01-V01     AACR         NaN  \n",
       "1762934  CLINVAR-H0893912-M01-V01  CLINVAR     16612.0  \n",
       "3847616   COSMIC-H0900805-M01-V01   COSMIC  24005626.0  \n",
       "...                           ...      ...         ...  \n",
       "3863169   COSMIC-H0928414-M01-V01   COSMIC  24005626.0  \n",
       "3863424   COSMIC-H0928775-M01-V01   COSMIC  24005626.0  \n",
       "3863812   COSMIC-H0929331-M01-V01   COSMIC  24005626.0  \n",
       "3863819   COSMIC-H0929342-M01-V01   COSMIC  24005626.0  \n",
       "3863875   COSMIC-H0929424-M01-V01   COSMIC  24005626.0  \n",
       "\n",
       "[74 rows x 46 columns]"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_binded[(df_binded['gene_name_h']=='EGFR') & (df_binded['HGVSp_h']=='G719S')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2743313"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "683645+473858+1585810"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Class 0: This mutation can be originally modeled.                                   566996\n",
       "Class 4: Flanked segments are not identical.                                        117841\n",
       "Class 2: This mutation can be modeled, but the effect may not be consistent.         97055\n",
       "Class 1: This mutation can be alternatively modeled.                                 93736\n",
       "Class 3: This mutation cannot be originally modeled and no alternative is found.       745\n",
       "Class 5: Coordinate error. This mutation is not in the query gene.                       2\n",
       "Name: statement, dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_aacr['statement'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_aacr = pd.read_csv('.../aacr_output/aacr_result_f0.csv')\n",
    "df_cosmic = pd.read_csv('.../cosmic_output/cosmic_result_f0.csv')\n",
    "df_clinvar = pd.read_csv('.../clinvar_output/clinvar_result_f0.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/df_result_cleaned.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3682634"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_binded['ID_mouse'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_name_h\n",
       "PRPF8    1227\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_binded[df_binded['flank_size']==1959]['gene_name_h'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded['flank_size_left'] = df_binded['flank_size_left'].fillna('0aa')\n",
    "df_binded['flank_size_right'] = df_binded['flank_size_right'].fillna('0aa')\n",
    "df_binded['flank_size'] = [int(x[:-2])+int(y[:-2]) for x,y in zip(df_binded['flank_size_left'], df_binded['flank_size_right'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3345179"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_binded[df_binded['flank_size']>0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1785931"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_binded[df_binded['flank_size']>20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.5338820433824318"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "1785931/3345179"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAE8EAAAZHCAYAAADqb7SpAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8fJSN1AAAACXBIWXMAAC4jAAAuIwF4pT92AAEAAElEQVR4nOzcbZDV9X338e8ubDSsoGK4iagBNBBrK3EXNY1CIAneNOIdyNnd1gmjY5I2kzTpxCZN00Qt6SSdyWBNTWIbZzQzyS6HFVIimETUVEMLlJiFGKBWRAFBIUpwkXvY68F1XV4Xiefs2d1zfv+z7Os1w4z4/57z/8DAA568a7q6uroCAAAAAAAAAAAAAAAAAAAAAAAAABKozXoAAAAAAAAAAAAAAAAAAAAAAAAAAAOHCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyQzYCN63vvWtqKmp+b0f06ZNy3oaAAAAAAAAAAAAAAAAAAAAAAAAwAlrQEbwdu/eHV/+8peznlEV1qxZ85YxwCx/DB48uE+/prlz5/b63ffff3+Zfmcr68iRIzFixIge//ruuOOOrKcDAAAAAAAAAAAAAAAAAAAAAAAwwA3ICN7f/M3fxKuvvpr1jKrQ0dGR9YSqks/ns55QkkcffTR+85vfZD0DAAAAAAAAAAAAAAAAAAAAAAAAemzARfAWL14c9913X9YzqoYI3vEef/zxfhFIbG1tzXoCAAAAAAAAAAAAAAAAAAAAAAAA9MqAiuBt2bIlbr311qxnVBURvOMdOXIkFi1alPWMog4cOBA//OEPs54BAAAAAAAAAAAAAAAAAAAAAAAAvTJgInh79uyJa665Jnbv3p31lKrR1dUV69aty3pG1cnn81lPKGrp0qXR2dmZ9QwAAAAAAAAAAAAAAAAAAAAAAADolQERwTt06FDccMMN8atf/SrrKVVl06ZNYmpv4Yknnohdu3ZlPaOg1tbWrCcAAAAAAAAAAAAAAAAAAAAAAABAr53wEbwjR47En/7pn8YTTzyR9ZSq09HRkfWEqnT06NF46KGHsp7xljo7O2Pp0qVZzwAAAAAAAAAAAAAAAAAAAAAAAIBeO6EjeAcPHoxZs2ZFe3t71lOqkgheYfl8PusJb2nx4sVx4MCBrGcAAAAAAAAAAAAAAAAAAAAAAABAr52wEbz9+/fHtddeG0uWLMl6StVau3Zt1hOq1pNPPhmvvPJK1jN+T1tbW9YTAAAAAAAAAAAAAAAAAAAAAAAAoE9OyAjejh07Ytq0afHTn/406ylVraOjo9ubF198Mbq6upL+OHLkSOV/8d04evRoPPTQQ1nPOM5vfvObePTRR7OeAQAAAAAAAAAAAAAAAAAAAAAAAH1ywkXw1qxZE5MnT47Vq1dnPaWqvfrqq7Ft27aiN+94xzvinHPOSbSo+uTz+awnHKe9vb0qAoEAAAAAAAAAAAAAAAAAAAAAAADQFydUBO/BBx+MqVOnxvbt27OeUvU6Ojq6vWlsbKz8kIyceeaZcfrppxe9eeqpp2LHjh2JFnWvra0t6wkAAAAAAAAAAAAAAAAAAAAAAADQZydEBG/37t2Ry+Vi7ty5sX///qzn9AulRPAaGhoqPyQjdXV1ccMNNxS9OXbsWLS3tydaVNxLL70UTz31VMHnEydOTLgGAAAAAAAAAAAAAAAAAAAAAAAAeq/fR/CWL18eF154YeTz+ayn9CulRPAaGxsrPyRDTU1N3d5Uy5+rBQsWxLFjxwo+b25uTrgGAAAAAAAAAAAAAAAAAAAAAAAAeq/fRvB27NgRLS0tMWPGjNi2bVvWc/odEbyID37wgzFixIiiNytWrIjt27cnWlRYa2tr0ectLS2JlgAAAAAAAAAAAAAAAAAAAAAAAEDf9LsI3pEjR+Luu++OiRMndhsG+11jxoyp0Kr+5eDBg7Fx48aiN8OHD4+xY8emGZSRQYMGxezZs4vedHV1xcKFCxMtemubNm2KNWvWFHze2NgY7373uxMuAgAAAAAAAAAAAAAAAAAAAAAAgN7rNxG8o0ePxoMPPhgTJ06Mz372s9HZ2dmjz3/mM5+Jv//7v6/Quv7lmWeeiSNHjhS9aWhoSLQmW01NTd3eLFiwIMGSwrqLPba0tCRaAgAAAAAAAAAAAAAAAAAAAAAAAH3XLyJ4y5YtiwsuuCDmzp0bzz//fI8+e9JJJ8V9990X8+fPj5qamgot7F/Wrl3b7U1jY2OCJdm7/PLL48wzzyx6s3Llyti6dWuiRb+vWASvtrY2crlcwjUAAAAAAAAAAAAAAAAAAAAAAADQN/0igvcv//Iv8d///d89/tzYsWPj5z//eXzsYx+rwKr+q6Ojo9ubgRLBq62tjTlz5hS96erqioULFyZadLx169bF+vXrCz6fOnVqjBkzJuEiAAAAAAAAAAAAAAAAAAAAAAAA6Jt+EcHrjVtuuSXWrVsXkydPznpK1RHBO14ul+v2Jp/PJ1jy+9ra2oo+b25uTrQEAAAAAAAAAAAAAAAAAAAAAAAAyuOEi+C9613viqVLl8b9998fQ4cOzXpO1enq6op169YVvTnttNNi/PjxiRZl733ve1+MHTu26M2qVavixRdfTDPo/1MsgldXVxezZ89OuAYAAAAAAAAAAAAAAAAAAAAAAAD67oSJ4NXV1cXtt98e69evjz/5kz/Jek7V2rx5c+zZs6foTUNDQ6I11SOXy3V7k8/nEyz5f1auXBmbN28u+PzKK6+M4cOHJ1wEAAAAAAAAAAAAAAAAAAAAAAAAfXdCRPBmzpwZzzzzTPzjP/5jDBkyJOs5Va2jo6Pbm8bGxsoPqTLVGMFra2sr+rylpSXREgAAAAAAAAAAAAAAAAAAAAAAACiffh3Bu+SSS2L58uWxZMmSmDBhQtZz+oVSIngNDQ2VH1JlLrroopg4cWLRmzVr1sTmzZuT7Dl27FjR6F59fX1cd911SbYAAAAAAAAAAAAAAAAAAAAAAABAOfXLCF5DQ0P88Ic/jFWrVsWHPvShrOf0K6VE8BobGys/pArlcrlub4qF6crpZz/7WezYsaPg82uvvTaGDBmSZAsAAAAAAAAAAAAAAAAAAAAAAACU0+CsB5SqpqYmZsyYEZ/73OdixowZWc/pt9auXVv0+bBhw+K88857y2d79+6Nf//3f49Vq1bF2rVrY/PmzbF9+/bYu3dvHDlyJIYMGRJDhw6Ns88+O84999x473vfG1OmTInGxsaoq6urxC+nrHK5XNx1111Fb/L5fHz+85+v+JbW1taiz1taWiq+AQAAAAAAAAAAAAAAAAAAAAAAACqhX0Twbr755vjGN74R5557btZT+rXdu3fHli1bit40NDRETU3Nmz8/cOBAtLe3xw9+8INYvnx5HD58uOBnOzs7o7OzM7Zv3x6rVq2KH/zgBxERMXz48Ljxxhtj7ty5cdlll5XnF1MBf/AHfxB/9Ed/FL/61a8K3jz99NPx3HPPFQwFlsPhw4dj0aJFBZ8PHz48rrzyyoq9HwAAAAAAAAAAAAAAAAAAAAAAACqpNusBpZg1a5YAXhl0dHR0e9PQ0BAREXv27Ik77rgjzjnnnLj55pvjkUceKRrAK+a1116L7373u3H55ZfH5ZdfHo8++mivvieFXC7X7U0+n6/ohp/85Cfx2muvFXw+e/bsqKurq+gGAAAAAAAAAAAAAAAAAAAAAAAAqJR+EcGjPEqJ4E2aNCnmz58f48aNizvvvDN27dpV1g0rVqyIK664Im688cbYtm1bWb+7HJqamrq9qXQEr7W1tejz5ubmir4fAAAAAAAAAAAAAAAAAAAAAAAAKkkEbwApJYJ3++23x1/91V/F7t27K7pl8eLFMWnSpPjRj35U0ff01LnnnhuTJ08uerN27dp49tlnK/L+ffv2xZIlSwo+HzNmTEydOrUi7wYAAAAAAAAAAAAAAAAAAAAAAIAURPAGkFIieDt37qz8kP/jtddei2uvvTa++tWvJntnKXK5XLc3+Xy+Iu9++OGHY+/evQWf53K5qK311xYAAAAAAAAAAAAAAAAAAAAAAID+S01rgDh06FBs2LAh6xlv6Utf+lJ86lOfynrGm3K5XNTU1BS9WbBgQUXe3draWvR5S0tLRd4LAAAAAAAAAAAAAAAAAAAAAAAAqQzOegBp/PrXv47Dhw/36Ttqa2vjvPPOiwsvvDBGjhwZp512Whw7diw6Oztj69atsXHjxnj22Wd79d3//M//HPX19fG1r32tTxvL4eyzz473v//9sWLFioI3zzzzTGzYsCHOP//8sr13z5498cgjjxR8PmHChGhsbCzb+wAAAAAAAAAAAAAAAAAAAAAAACALIngDREdHR68/+6EPfSj+7M/+LK677ro4/fTTi96++uqr8fDDD8f3vve9ePzxx3v0nq9//evxrne9K/78z/+811vLJZfLFY3gRUTk8/n4yle+UrZ3Ll68OA4ePFjweXNzc9neBQAAAAAAAAAAAAAAAAAAAAAAAFmpzXoAaaxdu7bHn5k+fXqsXr06li9fHnPnzu02gBcRccYZZ8RHP/rReOyxx2LNmjUxZcqUHr3zs5/9bPzyl7/s8dZyu+mmm6K2tvhfj3w+X9Z3tra2Fn0uggcAAAAAAAAAAAAAAAAAAAAAAMCJYHDWA0ijo6Oj5Nv6+vq455574pZbbunTOxsbG+PJJ5+Mu+++O/76r/86Dh8+3O1nDh48GC0tLbFu3bqoq6vr0/v7YvTo0fGBD3wgnnjiiYI369evj1//+tdxwQUX9Pl9O3fujMcee6zg84aGhpg4cWKf31NpO3fujF27dvXqsyNGjIiRI0eWeREAAAAAAAAAAAAAAAAAAAAAAADVRgRvgFi7dm1Jd+ecc04sWbIkJk2aVLZ3f+Yzn4k//MM/jBtuuCH27t3b7f3GjRvjn/7pn+Jzn/tc2Tb0RlNTU9EIXkTEggUL4q677urzuxYuXBhHjx4t+Ly5ubnP70jhW9/6Vtx55529+uxXvvKVuOOOO8o7CAAAAAAAAAAAAAAAAAAAAAAAgKojgjcAvPDCC/Hb3/6227sxY8bEz372sxg3blzZN3z4wx+Of/u3f4urrroqDh8+3O39vHnz4mMf+1gMGzas7FtKNWvWrPjkJz8ZR44cKXizcOHCskTw2traCj6rqamJpqamPr+j2t15551vBvTu+d7CGDn6nRkvAgAAAAAAAAAAAAAAAAAAAAAAqkVuxmVZT6CMarMeQOW98cYbcc0118QFF1wQ9fX1b3lz8sknx7JlyyoSwPu/PvjBD8b8+fNLut2zZ0/cd999FdtSijPOOCM+/OEPF73ZuHFjrFu3rk/v2bp1a6xYsaLg8ylTpsRZZ53Vp3cAAAAAAAAAAAAAAAAAAAAAAABAtRic9QAq74ILLogf/ehHb/78lVdeic2bNx/34wMf+EBceOGFFd/yyU9+MhYvXhyPPfZYt7f33ntv3H777RXfVExTU1P8+Mc/LnqTz+f79HvX1tYWXV1dBZ+3tLT0+rtT+4u/+Iu46aabSrp96aWX4sorr6zwIgAAAAAAAAAAAAAAAAAAAAAAAKqNCN4ANGrUqBg1alS8733vy+T93/jGN6KhoSGOHTtW9O7FF1+MVatWxaWXXppo2e+7/vrr46STToqDBw8WvMnn8zFv3rxev6O1tbXgs7q6upg9e3avvzu1kSNHxsiRI0u6ra+vL/hs5pTJMXbs2DKtAgAAIAuvv/56PPHEE2/+fPr06TFs2LAMFwEAADDQ+bcqAAAA1ci/VwEAAKg2/q0KAABANfrdf69yYqjNegADz6RJk+Kqq64q6XbRokUVXlPcqaee2u3W//mf/4lf/vKXvfr+Z599tuhnr7jiijjjjDN69d0AAAAAAAAAAAAAAAAAAAAAAABQjUTwyMRtt91W0t2TTz5Z4SXda2pq6vYmn8/36rtbW1uLPm9ubu7V9wIAAAAAAAAAAAAAAAAAAAAAAEC1EsEjE1dccUWcfPLJ3d49/fTTcfDgwQSLCps5c2YMGTKk6M3ChQt79d1tbW0Fnw0ZMiSuv/76Xn0vAAAAAAAAAAAAAAAAAAAAAAAAVCsRPDIxZMiQmDx5crd3hw4dik2bNiVYVFh9fX1cc801RW82bdoUv/jFL3r0vR0dHbFx48aCz2fOnBn19fU9+k4AAAAAAAAAAAAAAAAAAAAAAACodiJ4ZKaxsbGku82bN1d4Sfeampq6vcnn8z36ztbW1qLPW1paevR9AAAAAAAAAAAAAAAAAAAAAAAA0B+I4JGZ8ePHl3T30ksvVXhJ966++uoYNmxY0ZueRPC6urqira2t4PPTTz89rrrqqpK/DwAAAAAAAAAAAAAAAAAAAAAAAPoLETwyM2bMmJLu3njjjQov6d7JJ58c1113XdGbF154IVavXl3S9/3Hf/xHbNmypeDzWbNmxdve9rYebQQAAAAAAAAAAAAAAAAAAAAAAID+QASPzNTX15d0Vw0RvIiIXC7X7U0+ny/pu9ra2oo+b25uLul7AAAAAAAAAAAAAAAAAAAAAAAAoL8RwSMzb3vb20q6GzRoUIWXlOaKK66I4cOHF71pb2/v9nuOHj0aCxcuLPj8ne98Z0ybNq2n8wAAAAAAAAAAAAAAAAAAAAAAAKBfEMEjM/v37y/pbsiQIRVeUpq6urq48cYbi968+OKLsXLlyqI3jz/+eLzyyisFn+dyuait9VcTAAAAAAAAAAAAAAAAAAAAAACAE5PSFpnp7Ows6e7UU0+t8JLS5XK5bm8WLFhQ9HlbW1vR5y0tLT3aBAAAAAAAAAAAAAAAAAAAAAAAAP2JCB6Z2bZtW0l3Z599doWXlG769OkxatSoojft7e3R1dX1ls8OHToUixYtKvjZ8847Ly6++OI+bQQAAAAAAAAAAAAAAAAAAAAAAIBqNjjrAaRz6NCh2LlzZ+zatSt27tz55n+fccYZ8dGPfjT5nk2bNpV0N3bs2MoO6YFBgwbF7Nmz49577y14s23btli5cmX88R//8e89+8lPfhK//e1vC362ubm5HDMBAAAAAAAAAAAAAAAAAAAAAACgaongneBmzpwZGzdujF27dsWePXve8uacc87JJIL3i1/8otuboUOHxvjx4xOsKV0ulysawYuIWLhw4VtG8Nra2op+TgQPAAAAAAAAAAAAAAAAAAAAAACAE11t1gOorJ07d8Zzzz1XMIAXEbFly5Z4+eWXE66K6OzsjI6Ojm7vGhoaoqampvKDeuDyyy+Ps846q+hNe3t7dHV1Hff/9u/fH0uWLCn4mfe+971x/vnnl2UjAAAAAAAAAAAAAAAAAAAAAAAAVCsRvBPchAkTSrr7z//8zwovOd5Pf/rTOHz4cLd306dPT7CmZ2pqauKmm24qerN169ZYvXr1cf9v2bJlsXfv3oKfaW5uLss+AAAAAAAAAAAAAAAAAAAAAAAAqGYieCe4iy++uKS79vb2Ci853gMPPFDS3Uc+8pHKDumlpqambm9+9/c0n88XvK2pqRHBAwAAAAAAAAAAAAAAAAAAAAAAYEAQwTvBvf/97y/pbvHixfH6669XeM3/9txzz8UjjzzS7d24ceOisbExwaKeu+SSS2LcuHFFbx566KE3/3vfvn3x8MMPF7y97LLL4uyzzy7bPgAAAAAAAAAAAAAAAAAAAAAAAKhWIngnuIaGhhg9enS3d/v374/vf//7CRZFfPnLX46jR492e3frrbdGTU1NgkW9k8vlij7fvHlzPP300xERsWzZsti3b1/B25aWlrJuAwAAAAAAAAAAAAAAAAAAAAAAgGolgneCq62tjRtuuKGk27vuuis6OzsruufRRx+N1tbWbu/e/va3x6233lrRLX3V1NTU7c2iRYsiImLhwoUFbwYPHhw33XRT2XYBAAAAAAAAAAAAAAAAAAAAAABANRPBGwBuu+22ku5efvnluOuuuyq245VXXolbbrmlpNuPf/zjMXr06IptKYdJkybFe97znqI3ixYtigMHDsSyZcsK3syYMSPe8Y53lHseAAAAAAAAAAAAAAAAAAAAAAAAVCURvAHgoosuiqlTp5Z0O3/+/Fi6dGnZN7zxxhtx/fXXx7Zt27q9PfXUU+MLX/hC2TdUQi6XK/p8w4YNcc8998TevXsL3jQ3N5d7FgAAAAAAAAAAAAAAAAAAAAAAAFQtEbwBYt68eSXdHT16NObMmROrVq0q27tff/31uPrqq2PlypUl3f/DP/xDjBo1qmzvr6SmpqZub+64446Cz97+9rfH9ddfX75BAAAAAAAAAAAAAAAAAAAAAAAAUOVE8DI2d+7cqKmp6fbHtGnT+vSeKVOmxKxZs0q63bdvX8yYMSPa2tr69M6IiA0bNsSll14aTz31VEn3U6dOjU984hN9fm8q73nPe+LCCy8serN///6Cz6655poYOnRouWcBAAAAAAAAAAAAAAAAAAAAAABA1RLBG0DuvffeGDFiREm3nZ2d0dzcHLfddlu8/PLLPX7Xvn37Yt68eXHRRRfFxo0bS/rMiBEjorW1NWpr+9cfy6ampl5/tqWlpYxLAAAAAAAAAAAAAAAAAAAAAAAAoPr1r9oYfTJq1Kh48MEHY9CgQSV/5rvf/W6MGzcuPv3pT8fq1avj2LFjRe83bNgQX/rSl2LcuHHxd3/3d3Hw4MGS3nPSSSdFe3t7nHnmmSVvqxa5XK5XnzvttNPi6quvLvMaAAAAAAAAAAAAAAAAAAAAAAAAqG6Dsx5AWldffXXMnz8/Pv3pT5f8mQMHDsQ3v/nN+OY3vxmnn356XHrppTF69OgYOXJkDB48OF5//fV4/vnno6OjI7Zv397jTbW1tfH9738/pk6d2uPPVoPx48fHxRdfHP/1X//Vo8/deOONcdJJJ1VoFQAAAAAAAAAAAAAAAAAAAAAAAFQnEbwB6FOf+lTs27cvvvCFL/T4s7t3744f//jHZdtSV1cX999/f8yaNats35mFpqamHkfwmpubK7QGAAAAAAAAAAAAAAAAAAAAAAAAqldt1gPIxuc///n4zne+E3V1dZltGDp0aCxdujRuvvnmzDaUy5w5c6Kmpqbk+9GjR8f06dMruAgAAAAAAAAAAAAAAAAAAAAAAACqkwjeAPbxj388HnvssRgzZkzyd19yySWxZs2amDFjRvJ3V8JZZ50Vl112Wcn3c+bMiUGDBlVwEQAAAAAAAAAAAAAAAAAAAAAAAFQnEbwBbsqUKbF+/fr4xCc+EbW1lf/jcMopp8S8efNixYoVMWHChIq/L6WmpqaSb1taWiq4BAAAAAAAAAAAAAAAAAAAAAAAAKqXCB4xbNiw+Pa3vx3PPPNMzJkzJwYNGlT2d9TX18df/uVfxvPPPx9/+7d/G4MHDy77O7I2e/bskn7vxo8fH5deemmCRQAAAAAAAAAAAAAAAAAAAAAAAFB9RPB40/nnnx8LFiyILVu2xFe/+tVoaGiImpqaXn9fXV1dTJ06Nf71X/81duzYEXfffXeMGDGijIury6hRo2LatGnd3jU3N1d+DAAAAAAAAAAAAAAAAAAAAAAAAFSpwVkPSGnu3Lkxd+7crGcc54EHHogHHngg6xnHOfPMM+OLX/xifPGLX4yXX345fv7zn8fTTz8d69evj61bt8b27dtj7969sX///qitrY36+vo45ZRTYvTo0fHud787JkyYEJdccklMnTo1TjnllOT7s/w9Xb58eSbvjYjo6urK7N0AAAAAAAAAAAAAAAAAAAAAAABQqgEVwaPnRo8eHbNnz47Zs2dnPQUAAAAAAAAAAAAAAAAAAAAAAAA4AdRmPQAAAAAAAAAAAAAAAAAAAAAAAACAgUMEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkBmc9AOB3Xf71x2PwqaOyntGtF772kawnAAAAAAAAAAAAAAAAAAAAAAAA9Du1WQ8AAAAAAAAAAAAAAAAAAAAAAAAAYOAQwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAADgf7Fzv7Fa1/Ufx9/nBBggpWTO0IJNxSbLFVEuf4qagURygI5wXedqbqy2btTW6kZbba3W6lZ36pabbbW8dXEuOPyVUMSBpjZMy2VzlWVpIoEGItoBOXC685uZeH3POde5rvf3Qh+PjSl8Ptf3+1J3blzeeAJAGhE8AAAAAAAAAAAAAAAAAAAAAAAAANKI4AEAAAAAAAAAAAAAAAAAAAAAAACQRgQPAAAAAAAAAAAAAAAAAAAAAAAAgDQieAAAAAAAAAAAAAAAAAAAAAAAAACkEcEDAAAAAAAAAAAAAAAAAAAAAAAAII0IHgAAAAAAAAAAAAAAAAAAAAAAAABpRPAAAAAAAAAAAAAAAAAAAAAAAAAASCOCBwAAAAAAAAAAAAAAAAAAAAAAAEAaETwAAAAAAAAAAAAAAAAAAAAAAAAA0ojgAQAAAAAAAAAAAAAAAAAAAAAAAJBGBA8AAAAAAAAAAAAAAAAAAAAAAACANCJ4AAAAAAAAAAAAAAAAAAAAAAAAAKQRwQMAAAAAAAAAAAAAAAAAAAAAAAAgjQgeAAAAAAAAAAAAAAAAAAAAAAAAAGlE8AAAAAAAAAAAAAAAAAAAAAAAAABII4IHAAAAAAAAAAAAAAAAAAAAAAAAQBoRPAAAAAAAAAAAAAAAAAAAAAAAAADSiOABAAAAAAAAAAAAAAAAAAAAAAAAkEYEDwAAAAAAAAAAAAAAAAAAAAAAAIA0IngAAAAAAAAAAAAAAAAAAAAAAAAApBHBAwAAAAAAAAAAAAAAAAAAAAAAACCNCB4AAAAAAAAAAAAAAAAAAAAAAAAAaUTwAAAAAAAAAAAAAAAAAAAAAAAAAEgjggcAAAAAAAAAAAAAAAAAAAAAAABAGhE8AAAAAAAAAAAAAAAAAAAAAAAAANKI4AEAAAAAAAAAAAAAAAAAAAAAAACQRgQPAAAAAAAAAAAAAAAAAAAAAAAAgDQieAAAAAAAAAAAAAAAAAAAAAAAAACkEcEDAAAAAAAAAAAAAAAAAAAAAAAAII0IHgAAAAAAAAAAAAAAAAAAAAAAAABpRPAAAAAAAAAAAAAAAAAAAAAAAAAASCOCBwAAAAAAAAAAAAAAAAAAAAAAAEAaETwAAAAAAAAAAAAAAAAAAAAAAAAA0ojgAQAAAAAAAAAAAAAAAAAAAAAAAJBGBA8AAAAAAAAAAAAAAAAAAAAAAACANCJ4AAAAAAAAAAAAAAAAAAAAAAAAAKQRwQMAAAAAAAAAAAAAAAAAAAAAAAAgjQgeAAAAAAAAAAAAAAAAAAAAAAAAAGlE8AAAAAAAAAAAAAAAAAAAAAAAAABII4IHAAAAAAAAAAAAAAAAAAAAAAAAQBoRPAAAAAAAAAAAAAAAAAAAAAAAAADSiOABAAAAAAAAAAAAAAAAAAAAAAAAkEYEDwAAAAAAAAAAAAAAAAAAAAAAAIA0IngAAAAAAAAAAAAAAAAAAAAAAAAApBHBAwAAAAAAAAAAAAAAAAAAAAAAACCNCB4AAAAAAAAAAAAAAAAAAAAAAAAAaUTwAAAAAAAAAAAAAAAAAAAAAAAAAEgjggcAAAAAAAAAAAAAAAAAAAAAAABAGhE8AAAAAAAAAAAAAAAAAAAAAAAAANKI4AEAAAAAAAAAAAAAAAAAAAAAAACQRgQPAAAAAAAAAAAAAAAAAAAAAAAAgDQieAAAAAAAAAAAAAAAAAAAAAAAAACkEcEDAAAAAAAAAAAAAAAAAAAAAAAAII0IHgAAAAAAAAAAAAAAAAAAAAAAAABpRPAAAAAAAAAAAAAAAAAAAAAAAAAASCOCBwAAAAAAAAAAAAAAAAAAAAAAAEAaETwAAAAAAAAAAAAAAAAAAAAAAAAA0ojgAQAAAAAAAAAAAAAAAAAAAAAAAJBGBA8AAAAAAAAAAAAAAAAAAAAAAACANCJ4AAAAAAAAAAAAAAAAAAAAAAAAAKQRwQMAAAAAAAAAAAAAAAAAAAAAAAAgjQgeAAAAAAAAAAAAAAAAAAAAAAAAAGlE8AAAAAAAAAAAAAAAAAAAAAAAAABII4IHAAAAAAAAAAAAAAAAAAAAAAAAQBoRPAAAAAAAAAAAAAAAAAAAAAAAAADSiOABAAAAAAAAAAAAAAAAAAAAAAAAkEYEDwAAAAAAAAAAAAAAAAAAAAAAAIA0IngAAAAAAAAAAAAAAAAAAAAAAAAApBHBAwAAAAAAAAAAAAAAAAAAAAAAACCNCB4AAAAAAAAAAAAAAAAAAAAAAAAAaUTwAAAAAAAAAAAAAAAAAAAAAAAAAEgjggcAAAAAAAAAAAAAAAAAAAAAAABAGhE8AAAAAAAAAAAAAAAAAAAAAAAAANKI4AEAAAAAAAAAAAAAAAAAAAAAAACQRgQPAAAAAAAAAAAAAAAAAAAAAAAAgDQieAAAAAAAAAAAAAAAAAAAAAAAAACkEcEDAAAAAAAAAAAAAAAAAAAAAAAAII0IHgAAAAAAAAAAAAAAAAAAAAAAAABpRPAAAAAAAAAAAAAAAAAAAAAAAAAASCOCBwAAAAAAAAAAAAAAAAAAAAAAAEAaETwAAAAAAAAAAAAAAAAAAAAAAAAA0ojgAQAAAAAAAAAAAAAAAAAAAAAAAJBGBA8AAAAAAAAAAAAAAAAAAAAAAACANCJ4AAAAAAAAAAAAAAAAAAAAAAAAAKQRwQMAAAAAAAAAAAAAAAAAAAAAAAAgjQgeAAAAAAAAAAAAAAAAAAAAAAAAAGlE8AAAAAAAAAAAAAAAAAAAAAAAAABII4IHAAAAAAAAAAAAAAAAAAAAAAAAQBoRPAAAAAAAAAAAAAAAAAAAAAAAAADSiOABAAAAAAAAAAAAAAAAAAAAAAAAkEYEDwAAAAAAAAAAAAAAAAAAAAAAAIA0IngAAAAAAAAAAAAAAAAAAAAAAAAApBHBAwAAAAAAAAAAAAAAAAAAAAAAACCNCB4AAAAAAAAAAAAAAAAAAAAAAAAAaUTwAAAAAAAAAAAAAAAAAAAAAAAAAEgjggcAAAAAAAAAAAAAAAAAAAAAAABAGhE8AAAAAAAAAAAAAAAAAAAAAAAAANKI4AEAAAAAAAAAAAAAAAAAAAAAAACQRgQPAAAAAAAAAAAAAAAAAAAAAAAAgDQieAAAAAAAAAAAAAAAAAAAAAAAAACkEcEDAAAAAAAAAAAAAAAAAAAAAAAAII0IHgAAAAAAAAAAAAAAAAAAAAAAAABpRPAAAAAAAAAAAAAAAAAAAAAAAAAASCOCBwAAAAAAAAAAAAAAAAAAAAAAAEAaETwAAAAAAAAAAAAAAAAAAAAAAAAA0ojgAQAAAAAAAAAAAAAAAAAAAAAAAJBGBA8AAAAAAAAAAAAAAAAAAAAAAACANCJ4AAAAAAAAAAAAAAAAAAAAAAAAAKQRwQMAAAAAAAAAAAAAAAAAAAAAAAAgjQgeAAAAAAAAAAAAAAAAAAAAAAAAAGlE8AAAAAAAAAAAAAAAAAAAAAAAAABII4IHAAAAAAAAAAAAAAAAAAAAAAAAQBoRPAAAAAAAAAAAAAAAAAAAAAAAAADSiOABAAAAAAAAAAAAAAAAAAAAAAAAkEYEDwAAAAAAAAAAAAAAAAAAAAAAAIA0IngAAAAAAAAAAAAAAAAAAAAAAAAApBHBAwAAAAAAAAAAAAAAAAAAAAAAACCNCB4AAAAAAAAAAAAAAAAAAAAAAAAAaUTwAAAAAAAAAAAAAAAAAAAAAAAAAEgjggcAAAAAAAAAAAAAAAAAAAAAAABAGhE8AAAAAAAAAAAAAAAAAAAAAAAAANKI4AEAAAAAAAAAAAAAAAAAAAAAAACQRgQPAAAAAAAAAAAAAAAAAAAAAAAAgDQieAAAAAAAAAAAAAAAAAAAAAAAAACkEcEDAAAAAAAAAAAAAAAAAAAAAAAAII0IHgAAAAAAAAAAAAAAAAAAAAAAAABpRPAAAAAAAAAAAAAAAAAAAAAAAAAASCOCR9vdfvvt0dPTc8avG264oexpAAAAAAAAAAAAAAAAAAAAAAAAQMlE8GirI0eOxHe/+91SN/ziF794ywjfeH7ddtttpW6fiFWrVk34n0+IEAAAAAAAAAAAAAAAAAAAAAAAgLKJ4NFW3/72t+Nf//pX2TNatm3btjhx4kTZM8b00ksvxc6dO8ueAQAAAAAAAAAAAAAAAAAAAAAAABMmgkfbbN68Oe64446yZ0zKyy+/HHfffXfZM8a0adOmeO2118qeAQAAAAAAAAAAAAAAAAAAAAAAABMmgkdbPPvss/GlL32p7Blt0Wg0yp4wpnq9XvYEAAAAAAAAAAAAAAAAAAAAAAAAaIkIHpN29OjRuOWWW+LIkSNlT2mL7du3x/Hjx8ue0dTBgwdjz549Zc8AAAAAAAAAAAAAAAAAAAAAAACAlojgMSmvvfZarF69Op544omyp7TNsWPHYufOnWXPaGrDhg1x6tSpsmcAAAAAAAAAAAAAAAAAAAAAAABAS0TwaNnIyEh84QtfiD179pQ9pe0GBwfLntBUvV4vewIAAAAAAAAAAAAAAAAAAAAAAAC0TASPlpw4cSL6+/tj48aNZU/piLvuuiuGh4fLnnGGZ555Jn7961+XPQMAAAAAAAAAAAAAAAAAAAAAAABaJoLHhA0PD0dfX19s27at7Ckd8+qrr8aOHTvKnnGGwcHBGB0dLXsGAAAAAAAAAAAAAAAAAAAAAAAAtEwEjwk5cOBA3HDDDbFr166yp3Rco9Eoe8IZ6vV62RMAAAAAAAAAAAAAAAAAAAAAAABgUkTwGLdHH300Fi1aFI888kjZU1Ls2LEjXn311bJnvO6Pf/xjPP7442XPAAAAAAAAAAAAAAAAAAAAAAAAgEkRwWNc7rzzzli8eHE8//zzZU9pmwULFhSe//vf/4677rorac3Y1q9fX/YEAAAAAAAAAAAAAAAAAAAAAAAAmDQRPAodOXIkKpVKrFu3LoaHh8ue01aVSmXMO41GI2HJ+NTr9aZnvb29cfnllyeuAQAAAAAAAAAAAAAAAAAAAAAAgNaI4NHU7t2746qrruqqEFw7rVq1Kt797ncX3tm5c2e88sorSYua++1vfxt//vOfm54vXrw45syZk7gIAAAAAAAAAAAAAAAAAAAAAAAAWiOCxxkOHDgQtVotlixZEs8991zZczpm1qxZ8dnPfrbwzvDwcGzfvj1pUXP1er3wvFarJS0BAAAAAAAAAAAAAAAAAAAAAACAyRHB43UjIyPxk5/8JK644ooxo2tvdvHFF3doVWdVq9Ux7zQajYQlzY2Ojsbg4GDT86lTp0Z/f3/iIgAAAAAAAAAAAAAAAAAAAAAAAGidCB5x6tSpuPPOO+OKK66Ib3zjG3Hs2LEJff7rX/96/OAHP+jQus665ZZbYubMmYV37r777gn/O2mnhx56KP7xj380PV+2bFnMnj07cREAAAAAAAAAAAAAAAAAAAAAAAC0TgTvHe6Xv/xlLFiwINatWxdPP/30hD57zjnnxB133BE//vGPo6enp0MLO2vGjBmxYsWKwjvHjx+PrVu3Ji06U71eLzwfGBhIWgIAAAAAAAAAAAAAAAAAAAAAAACTJ4L3DvfTn/40/vSnP034c/PmzYsHH3wwvvzlL3dgVa5qtTrmnUajkbDkTKdOnYoNGzY0PZ85c2asXLkycREAAAAAAAAAAAAAAAAAAAAAAABMjggeE/bFL34xfv/738eiRYvKntIWy5Yti/e+972Fd3bt2hVHjx5NWvRf9913X7zwwgtNz/v6+mLGjBmJiwAAAAAAAAAAAAAAAAAAAAAAAGByRPAYt7lz58aOHTviZz/7WcyaNavsOW1zzjnnxKpVqwrvnDhxIrZu3Zoz6A3q9Xrhea1WS1oCAAAAAAAAAAAAAAAAAAAAAAAA7SGCx5imTp0a3/zmN+PJJ5+M5cuXlz2nI6rV6ph3BgcHE5b814kTJ2Lz5s1Nz2fPnh0333xz4iIAAAAAAAAAAAAAAAAAAAAAAACYPBE8Cq1YsSL+8Ic/xI9+9KOYMWNG2XM65jOf+Uy8733vK7xz7733xksvvZQzKCJ27twZR48ebXp+6623xtSpU9P2AAAAAAAAAAAAAAAAAAAAAAAAQDuI4PGWPvnJT8bu3btj27ZtMX/+/LLndNyUKVOiv7+/8M7Jkydj8+bNSYsi6vV64XmtVktaAgAAAAAAAAAAAAAAAAAAAAAAAO0jgsf/WLhwYWzZsiX27dsXN910U9lzUlWr1THvNBqNhCURr7zySmzfvr3p+cUXXxzXXXddyhYAAAAAAAAAAAAAAAAAAAAAAABoJxE8oqenJ5YuXRq7du2Kxx57LFauXFn2pFJcf/31cdFFFxXeue++++Lw4cMd37J169YYHh5uel6tVqO3148vAAAAAAAAAAAAAAAAAAAAAAAAZx8VrXe42267LZ566qm45557YsmSJWXPKVVvb2+sWbOm8M7Jkydj06ZNHd+yfv36wvOBgYGObwAAAAAAAAAAAAAAAAAAAAAAAIBOEMF7h+vv749LL7207Bldo1KpjHmn0Wh0dMPhw4fjnnvuaXo+f/78+PjHP97RDQAAAAAAAAAAAAAAAAAAAAAAANApInjwBtdcc0186EMfKryzZ8+eePHFFzu2YWhoKE6ePNn0fGBgoGPvBgAAAAAAAAAAAAAAAAAAAAAAgE4TwYM36OnpibVr1xbeGRkZiU2bNnVsQ71eLzyv1WodezcAAAAAAAAAAAAAAAAAAAAAAAB0mggevEmlUhnzTqPR6Mi7Dxw4EPfff3/T84ULF8b8+fM78m4AAAAAAAAAAAAAAAAAAAAAAADIIIIHb7Jo0aK47LLLCu/s3bs3Dh061PZ3NxqNOH36dNPzWq3W9ncCAAAAAAAAAAAAAAAAAAAAAABAJhE8eAuVSqXw/NSpUzE0NNT299br9aZnvb29Y+4CAAAAAAAAAAAAAAAAAAAAAACAbieCB29hPLG5RqPR1nf+7W9/i3379jU9v+666+KSSy5p6zsBAAAAAAAAAAAAAAAAAAAAAAAgmwgevIWPfOQjceWVVxbeeeCBB+LgwYNte+f69esLzwcGBtr2LgAAAAAAAAAAAAAAAAAAAAAAACjLlLIHQLeqVCrxve99r+n56dOnY+PGjfHVr361Le+r1+tNz6ZOnRpr1qxpy3s66dChQ/HCCy+M6+7+/fs7vKbzBu99qOwJAAAA3W3GBa//7c59T5Q4BAAAAP6f76oAAAB0I99XAQAA6Da+qwIAANCN3vB9lbcHETxoolqtFkbwIiIGBwfbEsF78skn44knmv9PwKVLl8bs2bMn/Z5Ou/322+P73/9+2TMAAAAAAAAAAAAAAAAAAAAAAADoYr1lD4BuNX/+/PjYxz5WeOehhx6K559/ftLvqtfrhee1Wm3S7wAAAAAAAAAAAAAAAAAAAAAAAIBuIIIHBSqVSuH56dOnY+PGjZN+z/r165uezZgxI1auXDnpdwAAAAAAAAAAAAAAAAAAAAAAAEA3mFL2AOhmlUolvvWtbxXeaTQa8bWvfa3ld/zmN7+Jv/zlL03P+/r6YubMmS0/P9NXvvKVWLNmzbju7t+/P26++eYOLwIAAAAAAAAAAAAAAAAAAAAAAKDbiOBBgXnz5sXVV18d+/bta3rn4Ycfjueeey4uueSSlt5Rr9cLzwcGBlp6bhkuvPDCuPDCC8d192wJ+xWpLPm/sicAAAB0rZdffjn27Nnz+u9vvPHGeM973lPiIgAAAN7pfFcFAACgG/m+CgAAQLfxXRUAAIBu9Obvq7w99JY9ALpdtVotPB8dHY0NGza09OzTp09Ho9Foen7++efHsmXLWno2AAAAAAAAAAAAAAAAAAAAAAAAdCMRPBjD2rVro7e3+EelKGRX5Fe/+lXs37+/6Xl/f39MmzatpWcDAAAAAAAAAAAAAAAAAAAAAABANxLBgzHMmTMnrr322sI7+/bti2effXbCz67X64XntVptws8EAAAAAAAAAAAAAAAAAAAAAACAbiaCB+NQrVYLz0dHR2PDhg0TeubIyEgMDQ01PZ8zZ05cf/31E3omAAAAAAAAAAAAAAAAAAAAAAAAdDsRPBiHW2+9Nd71rncV3mk0GhN65r333hsvvvhi0/NKpRK9vX5EAQAAAAAAAAAAAAAAAAAAAAAAeHtR2IJxeP/73x+f/vSnC+888sgj8fe//33cz6zX64XnAwMD434WAAAAAAAAAAAAAAAAAAAAAAAAnC1E8GCcqtXqmHcajca4nnX8+PHYsmVL0/PLLrssPvGJT4x3GgAAAAAAAAAAAAAAAAAAAAAAAJw1RPBgnFavXh3Tpk0rvDPeCN6OHTvi2LFjTc8HBgYmtA0AAAAAAAAAAAAAAAAAAAAAAADOFiJ4ME7nn39+LF26tPDOY489Fk8//fSYz6rX64XntVptQtsAAAAAAAAAAAAAAAAAAAAAAADgbCGCBxNQrVbHvNNoNArPjx07Fjt27Gh6/tGPfjQ+/OEPT3gbAAAAAAAAAAAAAAAAAAAAAAAAnA1E8GAC+vr6Yvr06YV3BgcHC8+3bNkSx48fb3peq9Va2gYAAAAAAAAAAAAAAAAAAAAAAABnAxE8mIBZs2bF8uXLC+88/vjj8dRTTzU9r9frTc96enqiWq22vA8AAAAAAAAAAAAAAAAAAAAAAAC6nQgeTNB4InUbN258yz8/fPhw7N69u+nnrr322vjgBz/Y8jYAAAAAAAAAAAAAAAAAAAAAAADodiJ4MEGf+9zn4txzzy28s2HDhrf886GhoTh58mTTzw0MDExqGwAAAAAAAAAAAAAAAAAAAAAAAHQ7ETyYoOnTp0dfX1/hnd/97nfx17/+9Yw/HxwcbPqZKVOmxJo1aya9DwAAAAAAAAAAAAAAAAAAAAAAALqZCB60oFKpjHln48aN//P7Q4cOxd69e5veX7JkSVxwwQWTnQYAAAAAAAAAAAAAAAAAAAAAAABdTQQPWrBs2bI477zzCu+8OYI3NDQUp06danq/Vqu1YxoAAAAAAAAAAAAAAAAAAAAAAAB0NRE8aMG0adNi9erVhXceffTReOaZZ17/faPRaHp3+vTpsWrVqnbNAwAAAAAAAAAAAAAAAAAAAAAAgK4lggctqlQqY94ZGhqKiIiDBw/GAw880PTeihUr4txzz23bNgAAAAAAAAAAAAAAAAAAAAAAAOhWInjQoptuuikuuOCCwjubNm16/a+nT59uem9gYKCt2wAAAAAAAAAAAAAAAAAAAAAAAKBbieBBi6ZMmRL9/f2Fdx5++OH45z//GUNDQ03vnHfeebF8+fJ2zwMAAAAAAAAAAAAAAAAAAAAAAICuJIIHk1CtVgvPR0dH4+c//3ncf//9Te98/vOfj2nTprV7GgAAAAAAAAAAAAAAAAAAAAAAAHQlETyYhMWLF8cHPvCBwjs//OEPY2RkpOl5rVZr9ywAAAAAAAAAAAAAAAAAAAAAAADoWiJ4MAm9vb2xZs2awjvDw8NNzy666KK48cYb2z0LAAAAAAAAAAAAAAAAAAAAAAAAupYIHkxStVpt+bOVSiV6e/0YAgAAAAAAAAAAAAAAAAAAAAAA8M6hvgWT9KlPfSrmzp3b0mcHBgbavAYAAAAAAAAAAAAAAAAAAAAAAAC6mwgetMHatWsn/JlLL700rr766g6sAQAAAAAAAAAAAAAAAAAAAAAAgO4lggdtUK1WUz4DAAAAAAAAAAAAAAAAAAAAAAAAZzsRPGiDhQsXxuWXXz6hz9RqtQ6tAQAAAAAAAAAAAAAAAAAAAAAAgO4lggdtUqlUxn33qquuiiuvvLKDawAAAAAAAAAAAAAAAAAAAAAAAKA7ieBBm1Sr1XHfrdVqHVwCAAAAAAAAAAAAAAAAAAAAAAAA3UsED9pkwYIFsWDBgjHv9fT0TCiYBwAAAAAAAAAAAAAAAAAAAAAAAG8nInjQRuOJ211zzTUxd+7chDUAAAAAAAAAAAAAAAAAAAAAAADQfUTwaIt169bF6Oho4a+9e/d2zZbR0dGYN29e29/9ne98Z8z3Pvjgg21/b0TE3r17u+a/AQAAAAAAAAAAAAAAAAAAAAAAADQjggcAAAAAAAAAAAAAAAAAAAAAAABAGhE8AAAAAAAAAAAAAAAAAAAAAAAAANKI4AEAAAAAAAAAAAAAAAAAAAAAAACQRgQPAAAAAAAAAAAAAAAAAAAAAAAAgDQieAAAAAAAAAAAAAAAAAAAAAAAAACkEcEDAAAAAAAAAAAAAAAAAAAAAAAAII0IHgAAAAAAAAAAAAAAAAAAAAAAAABpRPAAAAAAAAAAAAAAAAAAAAAAAAAASCOCBwAAAAAAAAAAAAAAAAAAAAAAAEAaETwAAAAAAAAAAAAAAAAAAAAAAAAA0ojgAQAAAAAAAAAAAAAAAAAAAAAAAJBGBA8AAAAAAAAAAAAAAAAAAAAAAACANCJ4AAAAAAAAAAAAAAAAAAAAAAAAAKQRwQMAAAAAAAAAAAAAAAAAAAAAAAAgjQgeAAAAAAAAAAAAAAAAAAAAAAAAAGlE8AAAAAAAAAAAAAAAAAAAAAAAAABII4IHAAAAAAAAAAAAAAAAAAAAAAAAQBoRPAAAAAAAAAAAAAAAAAAAAAAAAADSiOABAAAAAAAAAAAAAAAAAAAAAAAAkEYEDwAAAAAAAAAAAAAAAAAAAAAAAIA0IngAAAAAAAAAAAAAAAAAAAAAAAAApBHBAwAAAAAAAAAAAAAAAAAAAAAAACCNCB4AAAAAAAAAAAAAAAAAAAAAAAAAaUTwAAAAAAAAAAAAAAAAAAAAAAAAAEgjggcAAAAAAAAAAAAAAAAAAAAAAABAGhE8AAAAAAAAAAAAAAAAAAAAAAAAANKI4AEAAAAAAAAAAAAAAAAAAAAAAACQRgQPAAAAAAAAAAAAAAAAAAAAAAAAgDQieAAAAAAAAAAAAAAAAAAAAAAAAACkEcEDAAAAAAAAAAAAAAAAAAAAAAAAII0IHgAAAAAAAAAAAAAAAAAAAAAAAABpRPAAAAAAAAAAAAAAAAAAAAAAAAAASCOCBwAAAAAAAAAAAAAAAAAAAAAAAEAaETwAAAAAAAAAAAAAAAAAAAAAAAAA0ojgAQAAAAAAAAAAAAAAAAAAAAAAAJBGBA8AAAAAAAAAAAAAAAAAAAAAAACANCJ4AAAAAAAAAAAAAAAAAAAAAAAAAKQRwQMAAAAAAAAAAAAAAAAAAAAAAAAgjQgeAAAAAAAAAAAAAAAAAAAAAAAAAGlE8AAAAAAAAAAAAAAAAAAAAAAAAABII4IHAAAAAAAAAAAAAAAAAAAAAAAAQBoRPAAAAAAAAAAAAAAAAAAAAAAAAADSiOABAAAAAAAAAAAAAAAAAAAAAAAAkEYEDwAAAAAAAAAAAAAAAAAAAAAAAIA0IngAAAAAAAAAAAAAAAAAAAAAAAAApBHBAwAAAAAAAAAAAAAAAAAAAAAAACCNCB4AAAAAAAAAAAAAAAAAAAAAAAAAaUTwAAAAAAAAAAAAAAAAAAAAAAAAAEgjggcAAAAAAAAAAAAAAAAAAAAAAABAGhE8AAAAAAAAAAAAAAAAAAAAAAAAANKI4AEAAAAAAAAAAAAAAAAAAAAAAACQRgQPAAAAAAAAAAAAAAAAAAAAAAAAgDQieAAAAAAAAAAAAAAAAAAAAAAAAACkEcEDAAAAAAAAAAAAAAAAAAAAAAAAII0IHgAAAAAAAAAAAAAAAAAAAAAAAABpRPAAAAAAAAAAAAAAAAAAAAAAAAAASCOCBwAAAAAAAAAAAAAAAAAAAAAAAEAaETwAAAAAAAAAAAAAAAAAAAAAAAAA0ojgAQAAAAAAAAAAAAAAAAAAAAAAAJBGBA8AAAAAAAAAAAAAAAAAAAAAAACANCJ4AAAAAAAAAAAAAAAAAAAAAAAAAKQRwQMAAAAAAAAAAAAAAAAAAAAAAAAgjQgeAAAAAAAAAAAAAAAAAAAAAAAAAGlE8AAAAAAAAAAAAAAAAAAAAAAAAABII4IHAAAAAAAAAAAAAAAAAAAAAAAAQBoRPAAAAAAAAAAAAAAAAAAAAAAAAADSiOABAAAAAAAAAAAAAAAAAAAAAAAAkEYEDwAAAAAAAAAAAAAAAAAAAAAAAIA0IngAAAAAAAAAAAAAAAAAAAAAAAAApBHBAwAAAAAAAAAAAAAAAAAAAAAAACCNCB4AAAAAAAAAAAAAAAAAAAAAAAAAaUTwAAAAAAAAAAAAAAAAAAAAAAAAAEgjggcAAAAAAAAAAAAAAAAAAAAAAABAGhE8AAAAAAAAAAAAAAAAAAAAAAAAANKI4AEAAAAAAAAAAAAAAAAAAAAAAACQRgQPAAAAAAAAAAAAAAAAAAAAAAAAgDQieAAAAAAAAAAAAAAAAAAAAAAAAACkEcEDAAAAAAAAAAAAAAAAAAAAAAAAII0IHgAAAAAAAAAAAAAAAAAAAAAAAABpRPAAAAAAAAAAAAAAAAAAAAAAAAAASCOCBwAAAAAAAAAAAAAAAAAAAAAAAEAaETwAAAAAAAAAAAAAAAAAAAAAAAAA0ojgAQAAAAAAAAAAAAAAAAAAAAAAAJBGBA8AAAAAAAAAAAAAAAAAAAAAAACANCJ4AAAAAAAAAAAAAAAAAAAAAAAAAKQRwQMAAAAAAAAAAAAAAAAAAAAAAAAgjQgeAAAAAAAAAAAAAAAAAAAAAAAAAGlE8AAAAAAAAAAAAAAAAAAAAAAAAABII4IHAAAAAAAAAAAAAAAAAAAAAAAAQBoRPAAAAAAAAAAAAAAAAAAAAAAAAADSiOABAAAAAAAAAAAAAAAAAAAAAAAAkEYEDwAAAAAAAAAAAAAAAAAAAAAAAIA0IngAAAAAAAAAAAAAAAAAAAAAAAAApBHBAwAAAAAAAAAAAAAAAAAAAAAAACCNCB4AAAAAAAAAAAAAAAAAAAAAAAAAaUTwAAAAAAAAAAAAAAAAAAAAAAAAAEgjggcAAAAAAAAAAAAAAAAAAAAAAABAGhE8AAAAAAAAAAAAAAAAAAAAAAAAANKI4AEAAAAAAAAAAAAAAAAAAAAAAACQRgQPAAAAAAAAAAAAAAAAAAAAAAAAgDQieAAAAAAAAAAAAAAAAAAAAAAAAACkEcEDAAAAAAAAAAAAAAAAAAAAAAAAII0IHgAAAAAAAAAAAAAAAAAAAAAAAABpRPAAAAAAAAAAAAAAAAAAAAAAAAAASCOCBwAAAAAAAAAAAAAAAAAAAAAAAEAaETwAAAAAAAAAAAAAAAAAAAAAAAAA0ojgAQAAAAAAAAAAAAAAAAAAAAAAAJBGBA8AAAAAAAAAAAAAAAAAAAAAAACANCJ4AAAAAAAAAAAAAAAAAAAAAAAAAKQRwQMAAAAAAAAAAAAAAAAAAAAAAAAgjQgeAAAAAAAAAAAAAAAAAAAAAAAAAGlE8AAAAAAAAAAAAAAAAAAAAAAAAABII4IHAAAAAAAAAAAAAAAAAAAAAAAAQBoRPAAAAAAAAAAAAAAAAAAAAAAAAADSiOABAAAAAAAAAAAAAAAAAAAAAAAAkEYEDwAAAAAAAAAAAAAAAAAAAAAAAIA0IngAAAAAAAAAAAAAAAAAAAAAAAAApBHBAwAAAAAAAAAAAAAAAAAAAAAAACCNCB4AAAAAAAAAAAAAAAAAAAAAAAAAaUTwAAAAAAAAAAAAAAAAAAAAAAAAAEgjggcAAAAAAAAAAAAAAAAAAAAAAABAGhE8AAAAAAAAAAAAAAAAAAAAAAAAANKI4AEAAAAAAAAAAAAAAAAAAAAAAACQRgQPAAAAAAAAAAAAAAAAAAAAAAAAgDQieAAAAAAAAAAAAAAAAAAAAAAAAACkEcEDAAAAAAAAAAAAAAAAAAAAAAAAII0IHgAAAAAAAAAAAAAAAAAAAAAAAABpRPAAAAAAAAAAAAAAAAAAAAAAAAAASCOCBwAAAAAAAAAAAAAAAAAAAAAAAEAaETwAAAAAAAAAAAAAAAAAAAAAAAAA0ojgAQAAAAAAAAAAAAAAAAAAAAAAAJBGBA8AAAAAAAAAAAAAAAAAAAAAAACANCJ4AAAAAAAAAAAAAAAAAAAAAAAAAKQRwQMAAAAAAAAAAAAAAAAAAAAAAAAgjQgeAAAAAAAAAAAAAAAAAAAAAAAAAGlE8AAAAAAAAAAAAAAAAAAAAAAAAABII4IHAAAAAAAAAAAAAAAAAAAAAAAAQBoRPAAAAAAAAAAAAAAAAAAAAAAAAADSiOABAAAAAAAAAAAAAAAAAAAAAAAAkEYEDwAAAAAAAAAAAAAAAAAAAAAAAIA0IngAAAAAAAAAAAAAAAAAAAAAAAAApBHBAwAAAAAAAAAAAAAAAAAAAAAAACCNCB4AAAAAAAAAAAAAAAAAAAAAAAAAaUTwAAAAAAAAAAAAAAAAAAAAAAAAAEgjggcAAAAAAAAAAAAAAAAAAAAAAABAGhE8AAAAAAAAAAAAAAAAAAAAAAAAANKI4AEAAAAAAAAAAAAAAAAAAAAAAACQRgQPAAAAAAAAAAAAAAAAAAAAAAAAgDQieAAAAAAAAAAAAAAAAAAAAAAAAACkEcEDAAAAAAAAAAAAAAAAAAAAAAAAII0IHgAAAAAAAAAAAAAAAAAAAAAAAABpRPAAAAAAAAAAAAAAAAAAAAAAAAAASCOCBwAAAAAAAAAAAAAAAAAAAAAAAEAaETwAAAAAAAAAAAAAAAAAAAAAAAAA0ojgAQAAAAAAAAAAAAAAAAAAAAAAAJBGBA8AAAAAAAAAAAAAAAAAAAAAAACANCJ4AAAAAAAAAAAAAAAAAAAAAAAAAKQRwQMAAAAAAAAAAAAAAAAAAAAAAAAgjQgeAAAAAAAAAAAAAAAAAAAAAAAAAGlE8AAAAAAAAAAAAAAAAAAAAAAAAABII4IHAAAAAAAAAAAAAAAAAAAAAAAAQBoRPAAAAAAAAAAAAAAAAAAAAAAAAADSiOABAAAAAAAAAAAAAAAAAAAAAAAAkEYEDwAAAAAAAAAAAAAAAAAAAAAAAIA0IngAAAAAAAAAAAAAAAAAAAAAAAAApBHBAwAAAAAAAAAAAAAAAAAAAAAAACCNCB4AAAAAAAAAAAAAAAAAAAAAAAAAaUTwAAAAAAAAAAAAAAAAAAAAAAAAAEgjggcAAAAAAAAAAAAAAAAAAAAAAABAGhE8AAAAAAAAAAAAAAAAAAAAAAAAANKI4AEAAAAAAAAAAAAAAAAAAAAAAACQRgQPAAAAAAAAAAAAAAAAAAAAAAAAgDQieAAAAAAAAAAAAAAAAAAAAAAAAACkEcEDAAAAAAAAAAAAAAAAAAAAAAAAII0IHgAAAAAAAAAAAAAAAAAAAAAAAABpRPAAAAAAAAAAAAAAAAAAAAAAAAAASCOCBwAAAAAAAAAAAAAAAAAAAAAAAEAaETwAAAAAAAAAAAAAAAAAAAAAAAAA0ojgAQAAAAAAAAAAAAAAAAAAAAAAAJBGBA8AAAAAAAAAAAAAAAAAAAAAAACANCJ4AAAAAAAAAAAAAAAAAAAAAAAAAKQRwQMAAAAAAAAAAAAAAAAAAAAAAAAgjQgeAAAAAAAAAAAAAAAAAAAAAAAAAGlE8AAAAAAAAAAAAAAAAAAAAAAAAABII4IHAAAAAAAAAMB/2Ln7IKvruv/j74WVG5dbR/BeCRRzYBRYSZmyvCtRkhsJ9uzmiE3jzVjDZDVNf2TR0ExZk12WUzpD400xu5wFQQU10hwUzRzJxRuSCDCGxRFR7lnkbn9//K68Li/dc84uez7n6Hk8ZpjC7/uc7wtG/9h/ngAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgVbht27ZFVVVV2f3avHlzl/9Mc+bM6fJ7b7vttm782y2uMWPGdPrPd/3115d6NgAAAAAAAAAAAAAAAAAAAAAAABVOBK/CtbS0lHpCWWlubi71hIL84x//iNWrV5d6BgAAAAAAAAAAAAAAAAAAAAAAAHSaCF6FE8H7oLVr134s4nKNjY2lngAAAAAAAAAAAAAAAAAAAAAAAABdIoJX4UTwPiybzZZ6Ql5NTU2lngAAAAAAAAAAAAAAAAAAAAAAAABdIoJX4UTwPqy5ubnUE3JatWpVrFu3rtQzAAAAAAAAAAAAAAAAAAAAAAAAoEtE8CrY/v37Y+3ataWeUXbWrVsXL730UqlndKixsbHUEwAAAAAAAAAAAAAAAAAAAAAAAKDLRPAq2KuvvhqHDh0q9YyytGDBglJP+Ejt7e1luw0AAAAAAAAAAAAAAAAAAAAAAAAKIYJXwVpaWko9oWw1NzeXesJHeuaZZ2Lz5s2lngEAAAAAAAAAAAAAAAAAAAAAAABdJoJXwVavXl3qCWVrw4YN8eKLL5Z6xoc0NTWVegIAAAAAAAAAAAAAAAAAAAAAAAAcFRG8CtbS0pL3ZsWKFdHe3p7816mnnlr8v4A8stlsqSd8wKFDh6K5ubnUMwAAAAAAAAAAAAAAAAAAAAAAAOCoiOBVqPb29nj55Zdz3lRVVcXYsWMTLSo/5Race+KJJ2Lbtm2lngEAAAAAAAAAAAAAAAAAAAAAAABHRQSvQm3YsCF27dqV82bkyJHRv3//RIvSGzVqVM7nb7zxRvztb39LtCa/pqamUk8AAAAAAAAAAAAAAAAAAAAAAACAoyaCV6FaWlry3owbN674Q0qorq4u7002m02wJL/9+/fH4sWLO3w+cODAOPHEExMuAgAAAAAAAAAAAAAAAAAAAAAAgK4RwatQhUTwamtriz+khDKZTN6bhQsXRnt7e4I1uT366KOxa9euDp9PmzYtevfunXARAAAAAAAAAAAAAAAAAAAAAAAAdI0IXoUSwYs466yzYuzYsTlvNm3aFM8//3yiRR1rbGzM+byhoSHREgAAAAAAAAAAAAAAAAAAAAAAADg6IngVKl8Er6qqKsaNG5dmTAllMpm8N9lsNsGSju3ZsyeWLVvW4fMTTjghLr300oSLAAAAAAAAAAAAAAAAAAAAAAAAoOtE8CrQO++8E5s3b855M2LEiBgwYECiRaVTV1eX96a5uTna29sTrPloS5Ysiba2tg6fz5w5M3r27JlwEQAAAAAAAAAAAAAAAAAAAAAAAHSdCF4FWr16dd6b2traBEtK74wzzogLL7ww501ra2s8++yziRZ9WGNjY87n9fX1iZYAAAAAAAAAAAAAAAAAAAAAAADA0RPBq0AtLS15byolghcRkclk8t5ks9kESz7snXfeiT//+c8dPv/Upz4VEyZMSLgIAAAAAAAAAAAAAAAAAAAAAAAAjo4IXgUSwfugGTNmRI8euf9TWLRoURw5ciTRog++9+DBgx0+LyTgBwAAAAAAAAAAAAAAAAAAAAAAAOVEBK8CFRLBGzduXPGHlImTTz45Lrroopw3W7ZsiZUrVyZa9D8aGxtzPm9oaEi0BAAAAAAAAAAAAAAAAAAAAAAAALqHCF6Fee+99+L111/PeTN8+PAYNGhQmkFlIpPJ5L1ZsGBBgiX/Y8uWLfH00093+Hz06NExevTohIsAAAAAAAAAAAAAAAAAAAAAAADg6IngVZjXXnstDh48mPOmtrY20ZryMX369Kiurs55s2jRojhy5EiiRRHZbDbn+xoaGpJtAQAAAAAAAAAAAAAAAAAAAAAAgO4igldhWlpa8t6MGzeu+EPKzJAhQ+LSSy/NefPWW2/FihUrEi2KaGxszPm8vr4+0RIAAAAAAAAAAAAAAAAAAAAAAADoPiJ4FaaQCF5tbW3xh5ShTCaT9yabzSZYErF+/fp44YUXOnw+YcKEGDZsWJItAAAAAAAAAAAAAAAAAAAAAAAA0J2qSz2AtFavXp33pqMI3nvvvRfPPvtsPPfcc9HS0hIbNmyI1tbW2L17dxw4cCD69u0bNTU1ceqpp8aIESNi9OjRcdFFF8UFF1wQffv27e4/SrebNm1a3HzzzXHgwIEObx588MG46667omfPnkXd0tTUlPN5Q0NDUd8PAAAAAAAAAAAAAAAAAAAAAAAAxSKCV2HyRfCGDRsWxx133Pu/P3z4cCxdujT++Mc/xrJly6Ktra3Dz+7Zsyf27NkTb731VqxatSqy2WxERNTU1MTkyZPjuuuui4kTJ3bPH6QIBg0aFFdccUU88sgjHd5s3bo1nnrqqbj88suLuiVXBK9nz54xc+bMor4fAAAAAAAAAAAAAAAAAAAAAAAAiqVHqQeQzsaNG2Pnzp05b8aNGxcREfv3749f/epXMXz48Jg6dWosXLgwZwAvl71790ZjY2NceeWVce65574fxytHdXV1eW+Kvf/VV1+NV199tcPnl112WQwdOrSoGwAAAAAAAAAAAAAAAAAAAAAAAKBYRPAqSEtLS96bcePGxf333x8jRoyIb3/727Fp06Zu3fDKK69EXV1dXHzxxbFmzZpu/e7uMGXKlOjbt2/Om8WLF8ehQ4eKtqGxsTHn8/r6+qK9GwAAAAAAAAAAAAAAAAAAAAAAAIpNBK+CFBLB+/Wvfx3XX399bNmypahbVqxYEeeff37MmzevqO/prH79+sWkSZNy3mzbti3+8pe/FG1DU1NTh8/69OkT11xzTdHeDQAAAAAAAAAAAAAAAAAAAAAAAMUmgldBCongbd26tfhD/ltbW1vccMMNcfPNN8eRI0eSvTefurq6vDfZbLYo737hhRdiw4YNHT6/6qqrYsCAAUV5NwAAAAAAAAAAAAAAAAAAAAAAAKQggldBConglcI999wTX/nKV+LgwYOlnhIREZMmTYr+/fvnvFm8eHFR9jY2NuZ83tDQ0O3vBAAAAAAAAAAAAAAAAAAAAAAAgJSqSz2ANLZv3x6bNm06qu+oqqqKYcOGxbnnnhsnnXRSDBo0KKqqqmLPnj3R2toa//znP2PNmjVx5MiRTn/34sWLY9asWTF//vyoqqo6qp1Hq2/fvjF58uSYP39+hzfvvvtuPPHEE3HllVd223uPHDkS2Wy2w+cDBgyISZMmddv7AAAAAAAAAAAAAAAAAAAAAAAAoBRE8CpES0tLlz87fvz4uP7662PatGlx0kkn5bzdvXt3PP744zF//vx45JFHOhXEa2xsjNNOOy1uv/32Lm/tLnV1dTkjeBER2Wy2WyN4Tz/9dGzZsqXD59OmTYs+ffp02/sAAAAAAAAAAAAAAAAAAAAAAACgFHqUegBprF69utOfGTt2bCxfvjxeeOGFuOWWW/IG8CIi+vfvHzNmzIglS5bE66+/HlOmTOnUO3/xi1/EsmXLOr21u11xxRUxaNCgnDdLliyJgwcPdts7Gxsbcz6vr6/vtncBAAAAAAAAAAAAAAAAAAAAAABAqVSXegBptLS0FHxbXV0dc+fOje9+97tRXd31f0XOOuusWLJkSTQ1NcWNN94Yu3fvzvuZ9vb2+NrXvhZr166NwYMHd/ndR6tXr14xbdq0uPfeezu82bFjRyxfvjwmTZp01O87ePBgLFy4sMPnQ4cOjcsvv/yo31NsW7dujbfffrug29bW1iKvKb4Ff3621BMAAADK27HHv/9/H/vbKyUcAgAAAP/Nz6oAAACUIz+vAgAAUG78rAoAAEA5+l8/r/LJIIJXIQqN4A0ePDgWLlwYl156abe9O5PJxKc//emYOHFivPXWW3nv33777bjtttvirrvu6rYNXZHJZHJG8CIiFixY0C0RvOXLl8e7777b4fMZM2ZEz549j/o9xfbb3/42fvzjH5d6BgAAAAAAAAAAAAAAAAAAAAAAAGWsR6kHUHwHDhyINWvW5L0bMGBALF++vFsDeP8xZsyYePLJJ2PgwIEF3d99992xfv36bt/RGZdddlkMGTIk583DDz8c77333lG/q6mpKefzhoaGo34HAAAAAAAAAAAAAAAAAAAAAAAAlAMRvAqwffv2mDRpUpx33nkxYMCAj7ypqqqKbDYb559/ftF2jBo1KubPn1/Q7eHDh+OOO+4o2pZC9OzZM6ZPn57zZufOnfGnP/3pqN7T1tYWDz30UIfPhw0bFhMmTDiqdwAAAAAAAAAAAAAAAAAAAAAAAEC5qC71AIrvhBNOiMWLF7//+3fffTc2btwYGzZsiI0bN8bGjRvjrLPOiiuuuKLoWyZNmhRf//rX4/e//33e2/vuuy9uv/326NevX9F3dSSTycTdd9+d8yabzcbkyZO7/I6lS5fG7t27c26oqqrq8vendMstt8SMGTMKum1tbU3y7xwAAAAAAAAAAAAAAAAAAAAAAADlRQSvAh133HFx3HHHRW1tbUne/5Of/CSamppi7969Oe/27dsXy5Yti7q6ukTLPuyiiy6Kk08+ObZs2dLhzcMPPxz79++PPn36dOkdjY2NOZ/X19d36XtLYejQoTF06NCCbmtqaoq8pvjqvvjZUk8AAAAoW7t27Yqnnnrq/d9fcsklMWDAgBIuAgAAoNL5WRUAAIBy5OdVAAAAyo2fVQEAAChH//fnVT4ZepR6AJXnxBNPjFmzZhV0++CDDxZ5TW49evSIGTNm5LzZvXt3PPbYY136/l27duX87KhRo+Lcc8/t0ncDAAAAAAAAAAAAAAAAAAAAAABAORLBoyRuuOGGgu6eeeaZIi/JL5PJ5L3JZrNd+u7FixfH/v37O3xeX1/fpe8FAAAAAAAAAAAAAAAAAAAAAACAciWCR0mMGTMmTj/99Lx3b775ZrzxxhvFH5TDhRdeGGeccUbOm6VLl0ZbW1unv7uxsTHncxE8AAAAAAAAAAAAAAAAAAAAAAAAPmlE8CiZz3/+8wXdrVmzpshL8qurq8v5fM+ePfHoo4926ju3bdsWTz75ZIfPL7jgghg+fHinvhMAAAAAAAAAAAAAAAAAAAAAAADKnQgeJVNbW1vQ3caNG4u8JL9MJpP3JpvNduo7m5ub49ChQx0+b2ho6NT3AQAAAAAAAAAAAAAAAAAAAAAAwMeBCB4lM3z48ILuWltbi7wkv7Fjx8bIkSNz3ixdujT27dtX8Hc2NjZ2+Kxnz54xc+bMgr8LAAAAAAAAAAAAAAAAAAAAAAAAPi5E8CiZU045paC7vXv3FnlJYerq6nI+37dvXyxdurSg79q8eXOsXLmyw+eXXHJJnHjiiZ3aBwAAAAAAAAAAAAAAAAAAAAAAAB8HIniUTE1NTUF3H5cIXkRENpst6LsWLFgQ7e3tHT6vr68veBcAAAAAAAAAAAAAAAAAAAAAAAB8nIjgUTK9evUq6K5nz55FXlKYUaNGxejRo3PePProowVF+xobGzt81rt375g+fXqn9wEAAAAAAAAAAAAAAAAAAAAAAMDHgQgeJdPW1lbQ3bHHHlvkJYXLZDI5n7e1tcUjjzyS82bdunWxatWqDp9feeWVMXDgwC7tAwAAAAAAAAAAAAAAAAAAAAAAgHIngkfJ7N69u6C7cgrC1dXV5b1ZsGBBzudNTU05nzc0NHRqEwAAAAAAAAAAAAAAAAAAAAAAAHyciOBRMps3by7o7rTTTivyksKdeeaZUVtbm/Pm8ccfzxn4yxXB69+/f3z5y1/u8j4AAAAAAAAAAAAAAAAAAAAAAAAod9WlHkA6hw4dirfffju2bt36/v9u3bo1evToEbNnz06+Z/369QXdDRs2rLhDOimTycSqVas6fL5///5YunRp1NfXf+jZK6+8EmvWrOnws1OnTo2+fft2y04AAAAAAAAAAAAAAAAAAAAAAAAoRyJ4n3A33XRTrFixIrZu3Ro7duyI9vb2D9306tUrbrrppujdu3fSbblCcv/beeedV+QlnTNz5sz43ve+95F/l//R3Nz8kRG8pqamnN/9UZ8BAAAAAAAAAAAAAAAAAAAAAACAT5IepR5Ace3YsSPWrl0b27dv7zDaduDAgXjppZeS7mpvb4+VK1fmvTvjjDPi+OOPT7CocKeffnpMmDAh581jjz0We/bs+dA/z2azHX7m+OOPjy9+8YtHvQ8AAAAAAAAAAAAAAAAAAAAAAADKmQjeJ9zIkSMLuvvrX/9a5CUf9Pe//z3efPPNvHeXXHJJgjWdV1dXl/P5/v37Y9myZR/4Z6tWrYp//etfHX5mxowZUV1d3S37AAAAAAAAAAAAAAAAAAAAAAAAoFyJ4H3CjR8/vqC7hQsXFnnJB913330F3U2aNKm4Q7po5syZ0aNH7v98/u/faTabzXnf0NBw1LsAAAAAAAAAAAAAAAAAAAAAAACg3IngfcJNmDAhqqqq8t4999xzsW7dugSLInbs2BF/+MMf8t7169cvJk6cmGBR55144onxhS98IefNY489Fvv27Xv/983NzR3enn766fHZz3622/YBAAAAAAAAAAAAAAAAAAAAAABAuRLB+4QbMmRIjB8/vqDbefPmFXnN//fzn/88du7cmfcuk8lEv379Eizqmrq6upzP9+7dG48//nhERLz44ouxcePGDm8zmUxBsUIAAAAAAAAAAAAAAAAAAAAAAAD4uBPBqwDTp08v6O43v/lNbNq0qahbXnvttbjjjjvy3lVVVcU3v/nNom45WtOnT4/q6uqcNw8++GBERDQ3N+e8q6+v77ZdAAAAAAAAAAAAAAAAAAAAAAAAUM5E8CrArFmzolevXnnv2tra4jvf+U7Rduzbty+uvfbaeO+99/LeTpkyJc4777yibekOxx9/fFx22WU5b5YuXRoHDhyIRYsWdXhzzjnnxJgxY7p5HQAAAAAAAAAAAAAAAAAAAAAAAJQnEbwKcMIJJ0QmkynoduHChXHPPfd0+4ZDhw7FtddeGy0tLXlvq6urY+7cud2+oRjy/b3u3Lkz/uu//ivWr1/f4U19fX13zwIAAAAAAAAAAAAAAAAAAAAAAICyJYJXIX74wx/GMcccU9DtN77xjXjooYe67d0HDhyIhoaGWLx4cUH3t956a4wePbrb3l9M06ZNi169euW8mTNnTs7nIngAAAAAAAAAAAAAAAAAAAAAAABUEhG8MjBnzpyoqqrK+2vYsGFdfseIESNi9uzZBd0ePnw4ZsyYEXfeeWeX3/cfra2tcckll0Rzc3NB92effXb86Ec/Our3pjJw4MCYOHFizpu2trYOn40fPz7OPPPM7p4FAAAAAAAAAAAAAAAAAAAAAAAAZUsEr4LMnTs3zj777IJuDx48GN/61rdi6tSpsX79+k6/6+DBg/G73/0uRo0aFc8991xBn+nTp09ks9moqanp9PtKKZPJdPmzDQ0N3bgEAAAAAAAAAAAAAAAAAAAAAAAAyp8IXgXp27dvNDU1xbHHHlvwZx566KE4++yzY9asWbFixYo4cOBAzvs33ngjfvazn8XIkSPjlltuiZ07dxb0nqqqqrj33nvj3HPPLXhbuZg8eXKn/k7/o0ePHlFXV1eERQAAAAAAAAAAAAAAAAAAAAAAAFC+qks9gLTGjBkT8+fPj+nTp8eRI0cK+szhw4fjgQceiAceeCBqamriwgsvjFNOOSWGDBkSffr0id27d8e///3vePnll2Pjxo1d2vXLX/4yMplMlz5bajU1NTFp0qRobm7u1OcuvvjiOOmkk4q0CgAAAAAAAAAAAAAAAAAAAAAAAMqTCF4Fmjp1ajzwwAMxa9asOHz4cKc+u3fv3njyySe7dc9Pf/rTuPXWW7v1O1PLZDKdjuDV19cXaQ0AAAAAAAAAAAAAAAAAAAAAAACUrx6lHkBpfPWrX41FixZFv379SrbhmGOOifvvvz++//3vl2xDd7nqqquif//+Bd/36tUrpk+fXsRFAAAAAAAAAAAAAAAAAAAAAAAAUJ5E8CrYlClT4vnnn49zzjkn+btHjhwZK1eujOuuuy75u4uhT58+MWXKlILvJ06cGIMHDy7iIgAAAAAAAAAAAAAAAAAAAAAAAChPIngVbtSoUfHSSy/FD37wg+jdu3fR39erV6+49dZb46WXXorPfOYzRX9fSplMpuDbhoaGIi4BAAAAAAAAAAAAAAAAAAAAAACA8iWCR/Tu3Tvmzp0b69atixtvvDH69OnT7e/o1atXzJo1K9auXRt33HFHHHvssd3+jlL70pe+FIMHD857169fv7j66qsTLAIAAAAAAAAAAAAAAAAAAAAAAIDyI4LH+0477bS45557YsuWLXHnnXfG5z73uejZs2eXv69Hjx5RW1sbd9xxR7S2tsZ9990Xw4YN677BZeaYY46Ja665Ju/dlClTPpERQAAAAAAAAAAAAAAAAAAAAAAAAChEdakHEDFnzpyYM2dOqWe8b/DgwTF79uyYPXt2bN++PVauXBmrVq2K1157LTZt2hStra2xa9euaGtri4iImpqaqKmpiaFDh8aZZ54ZI0eOjPPPPz8uvvjiGDx4cPL9pfz7nDdvXsybN68k737jjTdK8l4AAAAAAAAAAAAAAAAAAAAAAADoDBE8cho8eHBcffXVcfXVV5d6CgAAAAAAAAAAAAAAAAAAAAAAAPAJ0KPUAwAAAAAAAAAAAAAAAAAAAAAAAACoHCJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAPw/9u4+SKuybuD4j3UlgSxUQCY1Vk0tdVQQRUzxjYARKN9GdMzGBANxxkonXLN8QWWXRsdU1FAWE5UK30Jby8gXVMxBmZXRwkBjRQFfFhWRBXFlnz+ansenlL13ufc6e+/5fP7Dvc65fvfMzq7nPuf+LgAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkU571AAClqqKyNusRcqW+emTWIwAAAAAAAAAAAAAAAAAAAAAAAEVQlvUAAAAAAAAAAAAAAAAAAAAAAAAAAOSHCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJBMedYDAEAhKiprsx4hV+qrR2Y9AgAAAAAAAAAAAAAAAAAAAAAAnVRZ1gMAAAAAAAAAAAAAAAAAAAAAAAAAkB8ieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkI4IHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkU571AABAx1NRWZv1CLlSXz0y6xEAAAAAAAAAAAAAAAAAAAAAAJIpy3oAAAAAAAAAAAAAAAAAAAAAAAAAAPJDBA8AAAAAAAAAAAAAAAAAAAAAAACAZETwAAAAAAAAAAAAAAAAAAAAAAAAAEhGBA8AAAAAAAAAAAAAAAAAAAAAAACAZETwAAAAAAAAAAAAAAAAAAAAAAAAAEhGBA8AAAAAAAAAAAAAAAAAAAAAAACAZETwAAAAAAAAAAAAAAAAAAAAAAAAAEhGBA8AAAAAAAAAAAAAAAAAAAAAAACAZETwAAAAAAAAAAAAAAAAAAAAAAAAAEhGBA8AAAAAAAAAAAAAAAAAAAAAAACAZETwAAAAAAAAAAAAAAAAAAAAAAAAAEhGBA8AAAAAAAAAAAAAAAAAAAAAAACAZETwAAAAAAAAAAAAAAAAAAAAAAAAAEhGBA8AAAAAAAAAAAAAAAAAAAAAAACAZETwAAAAAAAAAAAAAAAAAAAAAAAAAEhGBA8AAAAAAAAAAAAAAAAAAAAAAACAZMqzHgAAAIDiqKiszXqEXKmvHpn1CAAAAAAAAAAAAAAAAAAAAFCSyrIeAAAAAAAAAAAAAAAAAAAAAAAAAID8EMEDAAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAACSEcEDAAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAACSKc96AACAvKuorM16hFyprx6Z9QgAAAAAAAAAAAAAAAAAAAAAkGtlWQ8AAAAAAAAAAAAAAAAAAAAAAAAAQH6I4AEAAAAAAAAAAAAAAAAAAAAAAACQjAgeAAAAAAAAAAAAAAAAAAAAAAAAAMmI4AEAAAAAAAAAAAAAAAAAAAAAAACQjAgeAAAAAAAAAAAAAAAAAAAAAAAAAMmI4AEAAAAAAAAAAAAAAAAAAAAAAACQjAgeAAAAAAAAAAAAAAAAAAAAAAAAAMmI4AEAAAAAAAAAAAAAAAAAAAAAAACQjAgeAAAAAAAAAAAAAAAAAAAAAAAAAMmI4AEAAAAAAAAAAAAAAAAAAAAAAACQjAgeAAAAAAAAAAAAAAAAAAAAAAAAAMmI4AEAAAAAAAAAAAAAAAAAAAAAAACQjAgeAAAAAAAAAAAAAAAAAAAAAAAAAMmI4AEAAAAAAAAAAAAAAAAAAAAAAACQjAgeAAAAAAAAAAAAAAAAAAAAAAAAAMmI4AEAAAAAAAAAAAAAAAAAAAAAAACQjAgeAAAAAAAAAAAAAAAAAAAAAAAAAMmUZz0AAACkVFFZm/UIuVJfPTLrEQAAAAAAAAAAAAAAAAAAAADoYMqyHgAAAAAAAAAAAAAAAAAAAAAAAACA/BDBAwAAAAAAAAAAAAAAAAAAAAAAACAZETwAAAAAAAAAAAAAAAAAAAAAAAAAkinPegAAAKDzqqiszXoEAAAAAAAAAAAAAAAAAAAAADqYsqwHAAAAAAAAAAAAAAAAAAAAAAAAACA/RPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASEYEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASKY86wEAAAAAWlJRWZv1CLlSXz0y6xEAAAAAAAAAAAAAAAAAAIBOrCzrAQAAAAAAAAAAAAAAAAAAAAAAAADIDxE8AAAAAAAAAAAAAAAAAAAAAAAAAJIpz3oAAAAAKEUVlbVZjwAAAAAAAAAAAAAAAAAAAAAlqSzrAQAAAAAAAAAAAAAAAAAAAAAAAADIDxE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIpz3oAAAAAADqWisraIp3pU289/fWpIp2z86mvHpn1CAC0QfF+X1IIvy8BAAAAAAAAAAAAAACgcynLegAAAAAAAAAAAAAAAAAAAAAAAAAA8kMEDwAAAAAAAAAAAAAAAAAAAAAAAIBkRPAAAAAAAAAAAAAAAAAAAAAAAAAASKY86wEAAAAAIM8qKmuzHiFX6qtHZj0CAAAAAAAAAAAAAAAAAEDulWU9AAAAAAAAAAAAAAAAAAAAAAAAAAD5IYIHAAAAAAAAAAAAAAAAAAAAAAAAQDIieAAAAAAAAAAAAAAAAAAAAAAAAAAkU571AAAAAAAAqVRU1mY9Qq7UV4/MegQAAAAAAAAAAAAAAAAAoAMSwQMAAAAAgE5A5BEoFj9P0hKNBQAAAAAAAAAAAAAA8qgs6wEAAAAAAAAAAAAAAAAAAAAAAAAAyA8RPAAAAAAAAAAAAAAAAAAAAAAAAACSKc96AAAAAAAAOqeKytqsRwAAAAAAAAAAAAAAAAAAOqCyrAcAAAAAAAAAAAAAAAAAAAAAAAAAID9E8AAAAAAAAAAAAAAAAAAAAAAAAABIRgQPAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAABIRgQPAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAABIRgQPAAAAAAAAAAAAAAAAAAAAAAAAgGTKsx4AAAAAAAAAtqSisjbrEQAAAAAAAAAAAAAAAIAiKst6AAAAAAAAAAAAAAAAAAAAAAAAAADyQwQPAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAABIRgQPAAAAAAAAAAAAAAAAAAAAAAAAgGTKsx4AAAAAAAAAIK8qKmuzHiFX6qtHZj0CAAAAAAAAAAAAAAAQEWVZDwAAAAAAAAAAAAAAAAAAAAAAAABAfpRnPQAAAAAAAAAApFBRWZv1CLlSXz0y6xEAAAAAAAAAAAAAAOigRPAAAAAAAAAAgKLrmNHBTz0m8denshujHYgOAgAAAAAAAAAAAAClpCzrAQAAAAAAAAAAAAAAAAAAAAAAAADIj/KWlwAAAAAAAAAA0JFVVNZmPUKu1FePzHoEAAAAAAAAAAAAAChpZVkPAAAAAAAAAAAAAAAAAAAAAAAAAEB+lGc9AAAAAAAAAAAAlJKKytqsR8iV+uqRWY8AAAAAAAAAAAAAQJGVZT0AAAAAAAAAAAAAAAAAAAAAAAAAAPkhggcAAAAAAAAAAAAAAAAAAAAAAABAMuVZDwAAAAAAAAAAAPB5Kiprsx4hV+qrR2Y9Qq74/k7L9zcAAAAAAAAAAEDHIYIHAAAAAAAAAAAAACVMVDMtUU0AAAAAAAAAgK0nggcAAAAAAAAAAAB0eiJhaYmE0Zn5eZKWnycAAAAAAAAA0DmJ4AEAAAAAAAAAAABQVCJhAAAAAAAAAADAlpRlPQAAAAAAAAAAAAAAAAAAAAAAAAAA+VGe9QAAAAAAAAAAAAAAAED2Kiprsx4hV+qrR2Y9AgAAAAAAAEBmRPAAAAAAAAAAAAAAAAAAoACisWmJxqbl+zst398AAAAA5J0IHgAAAAAAAAAAAAAAHVI+Ihyfeqz/r09lNwYAAAAAAAAAJCSCBwAAAAAAAAAAAAAAQKeWj6gmeVVfPTLrEQAAAOgEvH+Slut5AAARPAAAAAAAAAAAAAAAgOR8qBgoFj9P6Mx8fwMAAAAAdF4ieAAAAAAAAAAAAAAAAAAAABSFiGla9dUjsx4BAAAA2kQEDwAAAAAAAAAAgIjwoTQAAAAAgFS8H5uWSBgAAABAxyOCBwAAAAAAAAAAAAAAAAAAdFqig631qY+g//Wp7MYAgE7M/5/QmYlQA1AoETwAAAAAAAAAAAAAAAAAAACAFohWAQAAFI8IHgAAAAAAAAAAAAAAAAAAAJQgUTYAAEjH/39nrTy+852sZ6CYRPAAAAAAAAAAAAAAAAAAAAAAAABKjCgbUMrKsh4AAAAAAAAAAAAAAAAAAAAAAAAAgPwQwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJIRwQMAAAAAAAAAAAAAAAAAAAAAAAAgmfKsBwAAAAAAAAAAAAAAAAAAAAAAAEpfRWVt1iMAUCLKsh4AAAAAAAAAAAAAAAAAAAAAAAAAgPwQwQMAAAAAAAAAAAAAAAAAAAAAAAAgGRE8AAAAAAAAAAAAAAAAAAAAAAAAAJLp0tzc3Jz1EEB+vf3223HzzTf/778nTpwYffr0yXAiAAAAisH1HgAAAB2Na1UAAAA6IterAAAAdDSuVQEAAOiIXK92TiJ4QKb+9re/xf777/+//37ppZdiv/32y3AiAAAAisH1HgAAAB2Na1UAAAA6IterAAAAdDSuVQEAAOiIXK92TmVZDwAAAAAAAAAAAAAAAAAAAAAAAABAfojgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyYjgAQAAAAAAAAAAAAAAAAAAAAAAAJCMCB4AAAAAAAAAAAAAAAAAAAAAAAAAyZRnPQCQb717947LLrvs//0bAACA0ud6DwAAgI7GtSoAAAAdketVAAAAOhrXqgAAAHRErlc7py7Nzc3NWQ8BAAAAAAAAAAAAAAAAAAAAAAAAQD6UZT0AAAAAAAAAAAAAAAAAAAAAAAAAAPkhggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACQjggcAAAAAAAAAAAAAAAAAAAAAAABAMiJ4AAAAAAAAAAAAAAAAAAAAAAAAACRTnvUAAP/p5ptvjvPOO++//vtRRx0VTzzxRPqBAAAASsy6deti2bJlsXLlyli9enWsX78+NmzYEM3NzdGtW7fo1q1b9O3bN3bdddfYc889Y8cdd8x65IJt2rQpFixYEM8880y88MIL8c9//jNWrVoV69ati40bN8Z2220X22+/ffTs2TP22muv2HfffWPAgAExbNiw6NmzZ9bjAwAA0Ek1NjbG/Pnz49lnn43FixdHfX19rFq1KtavXx8bN26M7t27x5e+9KXYYYcdYp999ol99903Dj300DjuuOOie/fuWY8PAACQaw0NDfH444/H888/H8uWLYtly5bFmjVr4sMPP4zGxsbYdttto0ePHtG3b9/YbbfdYv/994+BAwfGcccdF7169cp6/C1yvQoAAFA6Vq5cGU8//XQ899xzsWTJknj99dfjzTffjMbGxtiwYUN07do1unfvHjvttFP069cvvva1r8UhhxwSgwcPjm984xtZj18QzwEDAACkU6r9nrzc48zL6yxEl+bm5uashwD4t/feey/22muvWLNmzX99raP/EgUAAMjKypUr45FHHol58+bFwoULY/ny5dGat3wqKirikEMOiREjRsTxxx8fffv2bcdp2+app56K2267Le6///5Yv359q48vLy+PI444IsaNGxennnpqbLvttu0wJQAAAMWydu3aOOigg6K+vv4zv96vX7/P/Voqzc3N8fDDD0dNTU08/PDD8dFHH7X6HNttt10ce+yxMWHChBg1alR06dKlHSYFAADgPzU0NMRdd90Vd955Z9TV1bXq/uq/lZWVxeDBg+Pss8+O0047rcN80MD1KgAAQOlYs2ZNzJw5M37729+2+fo0ImLPPfeMk08+OcaPHx977LFHkafcep4DBgAASKvU+j15uceZl9fZWiJ4QIcyYcKEmD59+md+rSP+EgUAAMjKJ598Evfdd1/cdttt8dhjj8XmzZuLct6ysrIYPnx4nHvuuR3iDbAFCxbEpEmT4plnninaOXfZZZe45JJLYvz48VFWVla08wIAAFA8Z5xxRsyePftzv551BO+hhx6Kn/70p/HSSy8V7Zx77713TJ48OcaMGVO0cwIAAPD/rVq1Kq688sqYOXNmbNq0qWjn7dOnT1x88cUxceLE6Nq1a9HO21quVwEAAErDO++8E1dddVXceuutsXHjxqKdt6ysLE444YSYMmVK7LPPPkU7b1t5DhgAACAbpdTvycs9zry8zrYQwQM6jAceeCBOOumkz/16R/slCgAAkJXf/OY3cfnll8fSpUvbdZ/+/ftHdXV1DBs2rF33+SyNjY3xox/9KGbMmNHmv2rZkkMOOSRmzJgRBxxwQLucHwAAgLaZPXt2nHHGGVtck1UEr6GhIcaNGxdz585ttz2GDx8e06dPj379+rXbHgAAAHnzySefRFVVVUyZMiU2bNjQbvvsv//+cfvtt8fAgQPbbY/P4noVAACgdMycOTN+/OMfxwcffNBue5SXl8eFF14YkydPziTW7jlgAACA7JRKvycv9zjz8jq3hgge0CGsWLEiDjrooHjvvfc+d01H+SUKAACQleXLl8c555wTjz76aNJ9zzzzzLjhhhuiZ8+eSfZbsWJFjBgxIpYsWdLue3Xv3j1qamritNNOa/e9AAAAaNlrr70WBx54YKxdu3aL67KI4C1evDhGjhwZK1eubPe9dtppp7jnnnvimGOOafe9AAAAOrtly5bFd7/73Vi4cGGS/bp27Ro33nhj/OAHP0iyn+tVAACA0tDY2Bjf+9734r777ku250EHHRRz586Nr371q8n29BwwAABAdkql35OXe5x5eZ1bqyzrAQDWrl0bo0aN2uIvUAAAgLyrra2Ngw8+OHkALyLizjvvjAEDBsTf//73dt/rlVdeiSOPPDLJgy8R/3qg6PTTT48pU6Yk2Q8AAIDPt3nz5jjzzDNbDOBlYeHChXHMMcckeQglImLNmjUxbNiwmDVrVpL9AAAAOqunn346DjvssGQBvIiITZs2xfjx4+Pyyy9v971crwIAAJSGhoaGGDJkSNIAXkTECy+8EIMGDYq6urok+3kOGAAAIDul0u/Jyz3OvLzOYhDBAzK1adOmOPHEE+PFF1/MehQAAIAOq6amJr797W9n+ubj8uXL4/DDD4+nnnqq3fb4xz/+EUOGDIkVK1a02x6f55JLLonrr78++b4AAAD8n6qqqna97myrBQsWxNChQ5Nflzc1NcXZZ58d999/f9J9AQAAOosHH3wwhg4dGu+++24m+19xxRXt+iF816sAAACl4f33349hw4bFokWLMtn/zTffjG9961vx0ksvtes+ngMGAADITqn0e/JyjzMvr7NYujQ3NzdnPQSQT01NTXH66afHvffeW9D6o446Kp544on2HQoAAKCDuf3222Ps2LHRUd7C2X777WPevHkxaNCgop533bp1MXDgwFi6dGmrjuvXr1/0798/KioqokePHrF+/fpYuXJl1NXVxSuvvNKqc3Xp0iXuu+++OPHEE1t1HAAAAFvvueeei8MPPzyampoKWt+vX7+or69v36EiYuXKldG/f/945513WnXcPvvsEwceeGDssssu0b1791i3bl2sWLEiFi1aFK+//nqrztW1a9d48skni34tDgAA0JnNnz8/RowYERs3bmzVcT169IhDDz009thjj9hxxx1j8+bNsXr16njjjTfi2WefjU2bNrXqfF26dIl77rknTj755FYd1xLXqwAAAKVh8+bNMWLEiJg3b16rj+3Tp08MGjQodt555+jVq1c0NjbGW2+9FcuWLYu6urpWP1v8la98JRYtWhR9+/Zt9Swt8RwwAABAdkql35OXe5x5eZ3FJIIHZOKjjz6KU089NR588MGCjxHBAwAA8ubJJ5+MoUOHxscff9yq477+9a/H0KFD49hjj42Kioro1atX9O7dO5qamqKhoSEaGhqirq4uHnvssXj00Udb/WZa7969Y9GiRbHbbru16rgtGTNmTMyZM6egtd26dYuxY8fGhAkTYr/99vvcdS+//HLcdtttMX369Fi/fn1B595hhx2irq4u+vXrV9B6AAAAtt769eujf//+sWzZsoKPSRHBa2pqiqOPPjoWLFhQ0PqePXvGxIkTY9y4cbH77rt/7rpFixbFLbfcErNmzSr4mn/33XePurq6+PKXv1zQegAAgDxbsmRJDB48ONauXVvQ+rKyshg9enSce+65cdxxx0V5eflnrlu3bl08/PDDccMNN8QzzzxT8Dw9e/aMF198MXbdddeCj9kS16sAAACl47LLLovJkycXvL5r164xduzYGDt2bAwYMCC6dOnymevefvvtmDt3bkyZMqVV902POOKIeOKJJ2KbbbYp+JhCeA4YAAAgG6XS78nLPc68vM5iE8EDktuwYUOccMIJ8ec//7lVx4ngAQAAefL+++/HvvvuG6tXry74mNGjR8ell14aAwcOLPiYjz/+OGbNmhVVVVXx6quvFnzcoYceGgsWLPjcD4C0xq9+9as499xzC1p77LHHxowZM7b4ht5/Wr16dUyYMKHgN3KHDBkS8+fPL/j8AAAAbJ1x48ZFTU1Nq45JEcGrrKyMqVOnFrT21FNPjWnTpkXv3r0LPv/SpUvj7LPPLvhBlzPPPDNmzZpV8PkBAADy6KOPPopBgwbF4sWLC1p/wAEHxPTp0+Owww5r1T533313XHjhhfHWW28VtH706NGt+uDJlrheBQAAKA0vvvhiHHzwwQV/OH348OExffr0VgXcNm3aFDfeeGNcfPHFBe9zzTXXxIUXXljwHi3xHDAAAEA2Sqnfk5d7nHl5ncUmggcktXr16jjhhBNi4cKFrT5WBA8AAMiT73//+/HrX/+6oLW77rprzJkzJwYPHtzm/T755JO49NJLo6qqKgp9u2jq1KkxadKkNu8ZEfHuu+/GnnvuGe+//36La88777y4/vrr2/zXJ6+++ur4+c9/XtDrmz17dpx++ult2gcAAIDCPfDAA3HSSSe1+rj2juAtXbo09ttvv2hqampxbVVVVVRWVrZpn6ampvjhD38YN998c0HrFyxYEIcffnib9gIAAMiDCy64IK677rqC1p5xxhlRU1MTX/jCF9q012uvvRYjRoyIl19+uaD1jz/+eBx99NFt2uvfXK8CAACUjuHDhxccIpg0aVJUVVVFWVlZm/aaP39+nHLKKdHQ0NDi2h49esSrr74aO++8c5v2+jTPAQMAAGSjlPo9ebnHmZfX2R7a9m4IQBs8//zzMXDgwDb9AgUAAMiTF154Ie64446C1g4ZMiQWLVq0VQG8iIhtttkmrr766pg7d2588YtfLOiYK664It56662t2nfy5MkFPfgyfvz4mDZtWpsffImIuOSSS+KXv/xlQWsnTZoUjY2Nbd4LAACAlq1atSrOOeecrMf4TJMmTSroIZQpU6a0+SGUiIjy8vK46aab4oILLiho/fnnn19wvB4AACBvnnvuuYLvB/7sZz+Lu+66q80BvIh/BdqffvrpOPDAAwtaf9VVV7V5r39zvQoAAFAaFi5cWHAA7yc/+UlMnTq1zQG8iH8FDB555JHo3r17i2vXr18fv/jFL9q816d5DhgAACC9Uuv35OUeZ15eZ3sQwQOSuOOOO2LIkCGxatWqrEcBAADo8C677LKC3lTq379//OlPf4o+ffoUbe/Ro0fH7373u4IeMmlsbIypU6e2ea9XXnmloL828c1vfjOmTZvW5n0+7fzzz49Jkya1uO6NN96ImpqaouwJAADAf2tubo6zzjor1qxZk/Uo/2X+/Pkxd+7cFteNGTMmLr744qLsee2118aYMWNaXLdo0aL4wx/+UJQ9AQAAOpuLLrqooPusEydOjCuvvLIoe+60004xZ86cgv7Q2KOPPhpLlixp816uVwEAAErHrbfeWtC6UaNGbdWzuJ82YMCAgv8I94wZM2LDhg1btZ/ngAEAANIrtX5PXu5x5uV1thcRPKBdvffeezFmzJg466yztvpNUQAAgDxYvnx5PPTQQy2u69WrVzzwwAPRrVu3os9w/PHHx/XXX1/Q2q15COaaa66Jjz/+eItrunbtGjNnzozy8vI27fFZpkyZEkceeWSL66699tqC/vIGAAAArXfdddfFvHnzsh7jM1VVVbW4Zscdd4ybbrqpqPvW1NTE3nvv3eK6Yn0IBgAAoDP54x//GI8//niL60aMGBE33HBDUffee++946KLLipo7d13393mfVyvAgAAlIampqa49957W1y33XbbxbRp06JLly5F2/uUU06JUaNGtbjugw8+KOjD+VviOWAAAIB0SrXfk5d7nHl5ne1FBA9oN3/5y1/igAMOiDlz5mQ9CgAAQMmoqamJ5ubmFtddd9110a9fv3abY+LEiTFo0KAW161bty5+//vft/r8H374YcyePbvFdeeff35Bb8K1xjbbbBO33HJLiw/UvPbaawU9hAQAAEDrLF68OP6nvXsP13JM+wd+rpXVfk8KUS8lJKSNkEQqu8HYjZGSzcvYDNlvw8SYYSZ72U8lDMMcNhNmyCDlTRQiShQSjUyifWqt3x/zzs94se571fPcaz09n89x+GOs73Nd5+WfOc77vp7zufjii6u7jB80Z86ceOaZZxJzw4YNiw033DCnezdo0CDVIIaJEyfGpEmTcro3AABAoRs2bFhipnnz5jF69OioVatWzvc/88wzU/WJa/v+Ub8KAABQOF599dX46quvEnOnnnpqXu4C/+Y3v0k1WO/pp59e6z3cAwYAAMhOoc7vKZZ3nMVyznwyBA/Iuc8++yyOPvro6Nu3b3zyySfVXQ4AAEBBSXPZol27dvHzn/88r3WUlJTE9ddfnyo7duzYKq//wAMPxOLFiyvN1K5dO84555wqr51Gx44d49hjj03MjRw5Mi/7AwAAFKsVK1bEgAEDYuXKlT+aadSoUYYVfdddd92VOJy+RYsWceKJJ+Zl//79+0efPn0Sc/pVAACAb73xxhupLuxff/31sfHGG+elhkaNGsUxxxyTmJs5c2bMmzevyuvrVwEAAApH2i+VH3300XnZf/vtt48uXbok5saPH7/We7gHDAAAkH+FPr+nWN5xFss588kQPCBnVq9eHTfccEN06NAh/vjHP1bps5tttlmeqgIAACgcH3zwQcycOTMxd/7550etWrXyXs+uu+4anTp1Ssy9+OKLVV77zjvvTMwcccQR0apVqyqvndYvf/nLxMy4cePW6ksoAAAA/LDzzjsvpk+f/qN/r1evXlxzzTUZVvSt1atXp7rgcdJJJ0WdOnXyVkeafvVPf/pTrFixIm81AAAAFJLbb789MbP77rvHoEGD8lrHwIEDU+VeeumlKq2rXwUAACgsae4Cb7LJJqkG1a2tAw44IDHz0UcfxbJly9ZqffeAAQAA8md9mN9TLO84i+Wc+WYIHrDO1qxZE6NHj44OHTrEWWedlfgLHv/XkCFD4sorr8xTdQAAAIXjhRdeSJU78MAD81vIfzjooIMSM/PmzYuFCxemXnP27NkxZcqUxFy+v4Sy4447Rrdu3SrNlJeXx0MPPZTXOgAAAIrF008/HbfcckulmWuvvTa23XbbjCr6rvHjx8f8+fMTc/nuVw888MDEL4MsWrQonnrqqbzWAQAAUAiWLVsW999/f2Ju6NChea+lS5cu0bp168Tcm2++WaV19asAAACF5ZNPPknMtGvXLq81pBmwV1FREbNnz67y2u4BAwAA5Mf6NL+nWN5xFss5880QPGCdPPXUU9GxY8cYPHhwlR941qlTJ+644464/vrro6SkJE8VAgAAFI5XX301MbPNNtvEJptskkE1/7LXXnulys2aNSv1mn/5y18SM02aNEm997pIM+Rv7Nixea8DAABgfbdgwYI47rjjKs30798/TjvttIwq+r40/eq2224bW2+9dV7rqFWrVhxwwAGJOf0qAABAxLhx42LJkiWVZrp06RL9+/fPpJ4+ffpU+vdGjRrFl19+WaU19asAAACFJalPjYjEL6avq5YtW6bKffXVV1Ve2z1gAACA3Fvf5vcUyzvOYjlnvhmCB6yTO++8M2bOnFnlz7Vt2zYmTJgQJ510Uh6qAgAAKEyvv/56YmaPPfbIoJJvpR249/nnn6de89lnn03M7LvvvlFWVpZ6zbV14IEHJmYmTJgQX3/9dd5rAQAAWJ8df/zx8Y9//ONH/77hhhvGyJEjq/XyTZp+Nc2XKHIhTb/69NNPR0VFRQbVAAAA1FxpLukPGTIk/4X8rxNPPDEuvfTSuPHGG+OBBx6IZ555Jl5//fX45JNPYuXKlfH111/H7bffXqU19asAAACFpby8PDGT776padOmqXLLly+v8truAQMAAOTe+ja/p1jecRbLOfPNEDwgc8cff3xMmzYtunbtWt2lAAAA1ChpfqFj0003zaCSb6UdgpfmVysj/nWxZ8KECYm5Xr16pVpvXe24447RpEmTSjPffPNNjBs3LpN6AAAA1ke33XZb4lCCO+64I3UPmg9ffPFFvPPOO4m5rPrVNEPw58+fn2qgPgAAwPqqoqIinnzyyUozDRo0iEMPPTSjiiJ69uwZV155ZZxxxhnx85//PPr27Rs77bRTbLbZZlG7du0qr6dfBQAAKDxpBtAtWLAgrzUsXLgwVa5x48ZVWtc9YAAAgJqjps7vKZZ3nMVyziwYggdkpk2bNvHkk0/GPffcE40aNarucgAAAGqU5cuXx+rVqxNzG264YQbV5M8777wTX331VWJu1113zaCaiJKSklQPedNc2AEAAOD7ZsyYEeecc06lmWOPPTYOO+ywjCr6YZMmTUr8ZcOSkpLo0aNHJvVsuOGGsdVWWyXm9KsAAEAxe/vtt+PTTz+tNNO3b9+oX79+RhXlnn4VAACg8DRv3jwxM3fu3LzWMH/+/FS5Zs2aVWld94ABAACqX02f31Ms7ziL5ZxZ2KC6CwDWf2VlZTFkyJC44oorCvoiEQAAQD7Vq1cvvvzyy1i+fHnMmzcvPv300/jss8/i008//c4/aR5C5dLnn3+eKpf2lyCnTp2amKlXr17ssMMOqdbLhe7du8dzzz1XaeaVV17JqBoAAID1x6pVq+Loo4+O5cuX/2imbdu2cdNNN2VY1Q9L06+2a9cu1RdWcqV79+7xwQcfVJrRrwIAAMVs0qRJiZl99903g0ryR78KAABQeDp27JiYmT17dnz88cexxRZb5KWGNF9oLysri80337xK67oHDAAAUH0KZX5PsbzjLJZzZsEQPCCvfvKTn8Tvf//72Hrrrau7FAAAgIJQr169aNeuXbRr1666S4mIiHfffTdVrkWLFqlyb7zxRmJm6623jlq1aqVaLxe22WabxMy0adOivLw8SktLM6gIAABg/XDppZfG66+//qN/Ly0tjXvvvTf1YPV8StOvbrfddvkv5D+k6VfT1A0AALC+mjx5cmJm9913z6CS/NGvAgAAFJ5ddtklVe7pp5+Ok08+OS81PPnkk4mZnXbaKerWrVuldd0DBgAAqB6FNL+nWN5xFss5s6BbB/Kie/fuMW7cuHjiiScK4v9AAQAA+GEvvfRSqlyHDh1S5WbMmJGYSfOgLZfat2+fmFmyZEl8/PHHGVQDAACwfnj++edj+PDhlWbOP//82GOPPTKqqHKF2q++9957sXr16gyqAQAAqHmSfqG+YcOGmX+pINf0qwAAAIWne/fu0aRJk8TcjTfeGOXl5Tnf/29/+1uqfrJ3795VXrtQ+1T3gAEAgEJViPN7CrV3rOo7zmI5ZxYMwQNyauedd47HHnssXnnllejTp091lwMAAMA6evzxxxMzrVu3TnVZJ+JfD8iS1MQHexERs2bNynMlAAAA64cvv/wyBg0aVOkXNjp37hzDhg3LsKofV15eHh988EFirib2q6tXr445c+ZkUA0AAEDNsnLlynjnnXcqzXTs2DFKSwv3urx+FQAAoDDVrVs3jj766MTcu+++G2PGjMnp3uXl5XHhhRemyg4ePLjK67sHDAAAkI1Cnd9TLO84i+WcWSnct/pAjVFSUhL9+vWLZ555JqZMmRIHH3xwdZcEAABADkyYMCHVg7i99tor1Xrl5eXx0UcfJebatGmTar1c2WijjaJevXqJuZr2YA8AAKCmOvnkk+OTTz750b/XrVs37rvvvigrK8uwqh83b968WLVqVWIu63518803T5XTrwIAAMXogw8+iDVr1lSa2XbbbTOqJj/0qwAAAIXr1FNPTTWYfciQITFz5syc7XvRRRfFG2+8kZjr1atXbLfddlVa2z1gAACA/Fof5vcUyzvOYjlnVjao7gKAwjZw4MAYPnx4bLXVVtVdCgAAADl2/fXXp8r169cvVW7+/PmxevXqxNymm26aar1catmyZXz44YeVZiob4AAAAMC/jBo1Kh5++OFKM9dcc02Vv1CRT3Pnzk2Vy7pfbdGiRdSqVStxqIN+FQAAKEZpfszrv/7rv1Kvt2LFinjhhRfi5ZdfjrfffjtmzJgRCxcujK+++irWrFkTDRo0iObNm8dWW20V22yzTeyxxx7Ru3fvaNGixboco1L6VQAAgMK1/fbbx8knnxy33XZbpblFixbFQQcdFOPHj4+WLVuu05533313XHvttYm50tLSGD58eJXXdw8YAAAgf9aX+T3F8o6zWM6ZFUPwgHVy2GGHVXcJAAAA5MHrr78ejz76aGKufv36qX9RZN68ealy1XH5pVWrVomXXz799NNsigEAAChQs2fPjjPOOKPSTL9+/eKXv/xlRhWlU1P71dLS0mjRokXMnz+/0px+FQAAKEbvv/9+YqZNmzaV/r2ioiL++te/xp133hnPPPNMLFu27EezixYtikWLFsXs2bPj2WefjZtvvjlKS0ujT58+MXjw4DjyyCNjgw1yezVfvwoAAFDYrrrqqnjyySfj448/rjT33nvvRdeuXePRRx+Nrl27Vnmf8vLyuPjii+Oaa65Jlf/FL36xVvvU1D41wj1gAACg8K0v83tqau+Y63ecxXLOrJRWdwEAAAAA1DxDhgyJioqKxNxhhx0WjRo1SrXm559/nirXqlWrVLlcSrPnF198kUElAAAAhWnNmjUxYMCAWLx48Y9mmjdvHiNHjoySkpIMK0uWpl+tX79+6v43l/SrAAAAP+yDDz5IzLRs2fIH/31FRUXce++90a5du9h///3jscceq3QA3o8pLy+PZ599NgYMGBAdOnSI0aNHp3rHmpZ+FQAAoLA1b948xo4dm6pv++STT6Jnz54xZMiQxC+q/6dx48bFbrvtlnoAXq9eveL6669Pvf5/cg8YAACAJMXyjrNYzpkVQ/AAAAAA+I4//OEPMX78+MRcSUlJnHvuuanX/ec//5kq17Rp09Rr5krjxo0TMzXtwR4AAEBNMmzYsJg0aVKlmTvuuCPzXzRMI02/Wh29aoR+FQAA4Mek+WX6Fi1afO/fTZkyJbp37x7HHntszJ49O2f1zJ49OwYPHhy9e/eOWbNm5WRN/SoAAEDh69SpU4wdOzaaN2+emF25cmXceOONseWWW8YhhxwSI0aMiKlTp8a8efNi5cqVsXjx4pg9e3Y8//zzcemll8bOO+8cffv2jVdeeSVVLV27do1HH300ateuvVZncQ8YAACAJMXyjrNYzpmVDaq7AAAAAABqjg8//DCGDBmSKnvwwQfHDjvskHrthQsXJmbq1q0bZWVlqdfMlYYNGyZmvv766wwqAQAAKDwvv/xy/PrXv640M2jQoDj88MMzqqhq0vSraS6E5IN+FQAA4IeluZTfrFmz7/zvW2+9Nc4+++xYtWpVvsqK8ePHR5cuXWLUqFFx6KGHrtNa+lUAAID1Q69evWLy5Mlx8MEHx/Tp0xPzy5cvj8cffzwef/zxnNVw2GGHxZgxY6JevXprvYZ7wAAAACQplnecxXLOrJRWdwEAAAAA1AwrV66MI444IhYvXpyYrV27dlxzzTVVWj/NutX1YK9Ro0aJmZr2YA8AAKAmWLx4cRxzzDGxZs2aH820adMmbr755gyrqhr9KgAAQOFJMwSvQYMGERFRXl4e//3f/x2nn356Xgfg/dvixYvjsMMOi2uvvXad10miXwUAACgMW221VUyZMiWuuuqqqF+/fmb7tmjRIkaOHBmPPPLIOg3Ai9CnAgAAkKxYesdiOWdWDMEDAAAAICIiTj755HjttddSZS+44ILYeuutq7T+0qVLEzM1+dctli1blkElAAAAheW0006LOXPm/OjfS0tL49577622fi8N/SoAAEDhSTMEr2HDhlFeXh4nnHBC3H333RlU9V0XXHBBXHbZZWv9ef0qAADA+qVOnTpxySWXxIwZM+Kss86KJk2a5G2vpk2bxgUXXBAzZ86MwYMH52RNfSoAAABJiqV3LJZzZmWD6i4AAAAAgOp3xRVXxOjRo1NlO3XqFBdffHGV90jzYKysrKzK6+ZCmn1XrFiRQSUAAACF46GHHooxY8ZUmjnvvPOiV69eGVW0dvSrAAAAhWfhwoWJmbKysjj33HNj1KhRqdZs1KhRdOnSJdq3bx/NmjWLsrKyWLJkSXz88cfx1ltvxfvvv1/lOq+88spo2bJlnHbaaVX+rH4VAABg/bT55pvH1VdfHZ07d44zzjgjFi1alNP1u3XrFo8//nhssskmOV1XnwoAAECSYukdi+WcWTEEDwAAAKDIjRgxIn71q1+lytavXz8eeuihqFu3bpX3WbVqVWJmgw2q53FVmn3T1A8AAFAs5s6dG7/4xS8qzey0004xbNiwjCpae/pVAACAwrJq1apYs2ZNYu7hhx+O66+/vtJMWVlZDBgwII499tjYY489olatWj+a/eSTT+Khhx6KW2+9NebMmZO63jPOOCPat28f/fr1S/2ZCP0qAADA+ujNN9+M3/3ud/HEE0/E4sWL87LHq6++Gu3bt4+jjjoqTjnllOjSpUtO1tWnAgAAkKRYesdiOWdWSqu7AAAAAACqz+jRo+P0009Pnb/11ltj2223Xau90nwRpSY/2EtTPwAAQDEoLy+PgQMHxqJFi340U7du3bjvvvuidu3a2RW2lvSrAAAAhSXthfyTTjqp0r8feuih8f7778fIkSOjd+/elQ7Ai4ho3bp1nHPOOTFz5sy44YYbomHDhqnqKC8vj2OPPTa++OKLVPl/068CAACsP95777046qijonPnznH//ffnbQDevy1dujTuueee6Nq1awwYMCDmzp27zmvqUwEAAEhSLL1jsZwzK4bgAQAAABSp0aNHx/HHHx8VFRWp8ueee24MHjx4rferyQ/2ysrKEjM17cEeAABAdbn22mvjxRdfrDTz29/+Njp27JhRRetGvwoAAFBY0g7BW7Zs2Q/++3r16sXo0aPjz3/+c2yxxRZV3r+srCzOPPPMmDp1auofEJs/f3788pe/rNI++lUAAID1w1133RU77LBDPPTQQ6nv7ObSAw88EB06dIjf//7367SOPhUAAIAkxdI7Fss5s1I9/6UAAAAAqFb33HNPnHTSSVFeXp4qf9BBB8U111yzTnuWlJTkJAMAAED1mTJlSlx22WWVZvr27RtnnHFGRhWtO/0qAABAYUk7BO+HNGnSJJ588snYfffd17mO9u3bx4QJE6Jfv34xZcqUxPyDDz4YZ555ZvTo0SPV+vpVAACAwrZy5co4+eSTY/To0VX+bJ06daJz587Rrl27aNmyZdSrVy9WrlwZCxcujDlz5sTUqVNj0aJFqddbvnx5nHfeeTFlypT4wx/+EPXq1atyTfpUAAAAkhRL71gs58yKIXgAAAAARea6666Lc889N/WvSfbs2TMefPDBKC0tXad90/yCxOrVq9dpj7X1zTffJGaq65c3AAAAaoply5bFgAEDKu2hmjdvHqNGjSqoixv6VQAAgMKSplf6IXXq1ImxY8fmZADevzVv3jyefvrp6N69e3z44YeJ+UsvvTTGjRuXam39KgAAQOEqLy+PAQMGxJ///OfUn9lggw3i8MMPj4EDB0afPn2iTp06P5qtqKiIKVOmxMMPPxx33313LFy4MNUeDz74YLz//vvx3HPPRePGjVPXFqFPBQAAIFmx9I7Fcs6srNs3lwEAAAAoKEOHDo1zzjkn9QC8bt26xdixY9fqFx//r9q1aydmquvBXpp9a9WqlUElAAAANddZZ50VM2fOrDRz++23x6abbppRRbmhXwUAACgON910U/Ts2TPn67Zo0SIefvjhVP3Zc889F2+//XaqdfWrAAAAhev000+v0gC8Qw45JGbOnBl//OMfY//99690AF5ERElJSXTt2jWuueaamDt3bgwbNizq16+faq/XXnstDj/88Cr3lPpUAAAAkhRL71gs58yKIXgAAAAARWD16tVx4oknxlVXXZX6M926dYtnnnkmmjRpkpMaCv3BXpr6AQAA1ldPPPFE3HnnnZVmBg4cGEcccURGFeWOfhUAAKCwrE0f1L9//zjppJPyUM2/dO3aNc4888xU2dtvvz1VTr8KAABQmO6777647bbbUmXr1asXI0eOjEcffTS23HLLtdqvfv36MXTo0Jg6dWrsuOOOqT7z7LPPxqmnnlqlffSpAAAAJCmW3rFYzpkVQ/AAAAAA1nNLly6Ngw8+OO65557Un9ltt91i3Lhx0bRp05zVUVZWlpj55ptvcrZfVaTZt0GDBhlUAgAAUPPMnz8/TjjhhEozbdq0iVtuuSWjinJLvwoAAFBYqnohv6SkJIYPH56nar510UUXRcOGDRNzjz76aFRUVCTm9KsAAACF58svv4xzzjknVbZJkybx7LPPxuDBg3Oyd4cOHWLixImx7777psrfdddd8eyzz6ZeX58KAABAkmLpHYvlnFkxBA8AAABgPfbZZ59Fr1694qmnnkr9mb59+8YzzzwTjRs3zmktab7wsWTJkpzumVaafWvagz0AAIAsVFRUxODBg+OLL7740UxpaWnce++9Oe8js6JfBQAAKCxVHYK3//77R8eOHfNUzbc22mijOProoxNzn376abz22muJOf0qAABA4bnkkkvi888/T8zVqVMnnnjiidh9991zun+DBg3isccei7322itV/vTTT49Vq1alyupTAQAASFIsvWOxnDMrhuABAAAArKemTZsWu+yyS0ydOjX1Zw4//PAYO3ZsXh5ibbjhhomZxYsX53zfNNLsm+bBJAAAwPrmpptuir/97W+VZs4999zo1atXRhXlnn4VAACgsNStW7dK+eOPPz5PlXzfoEGDUuVefvnlxIx+FQAAoLAsWrQoRo0alSp7/fXX5+0da506deKRRx6JLbbYIjH73nvvxR133JFqXX0qAAAASYqldyyWc2bFEDwAAACA9dDYsWOjZ8+eMXfu3NSfOeWUU+Khhx6K2rVr56WmNA/2li5dGhUVFXnZvzJpHuylqR8AAGB98tZbb8WFF15YaWbHHXeMK6+8MqOK8qPQL6LoVwEAgGJTq1ataNy4caps7dq1o3///nmu6Fs9evSIpk2bJuZeeeWVxIx+FQAAoLCMGTMmli9fnpjbe++945RTTslrLc2bN4+RI0emyuZyCJ57wAAAAMWtWN5xFss5s2IIHgAAAMB65tprr42DDz64Sg/JrrzyyhgxYkSUlubvcVGaB2MVFRXV8nDv66+/TszUtAd7AAAA+bRy5coYMGBArFix4kczdevWjfvvvz9vw9SzkqbfS9M35oN+FQAA4IdttNFGqXI77bRTNGjQIM/VfKtWrVrRrVu3xNyMGTMSM/pVAACAwjJq1KhUueHDh+e3kP+19957xyGHHJKYmz59erz88suJOfeAAQAASFIs7ziL5ZxZMQQPAAAAYD2xcuXKGDRoUFxwwQVRXl6e6jNlZWUxatSouPTSS/NcXfoHY//4xz/yXMn3zZ8/PzGz8cYbZ1AJAABAzXDBBRfEW2+9VWnmN7/5TXTs2DGjivInTb+6aNGiWLVqVQbVfJd+FQAA4IelHYK3ww475LmS79txxx0TMx999FFiRr8KAABQOBYvXhxvvPFGYm7vvfeOnXbaKe/1/NtFF12UKvfII48kZtwDBgAAIEmxvOMslnNmxRA8AAAAgPXAggULok+fPjFmzJjUn2natGn89a9/jWOPPTaPlX1r8803T5VL85At19Lsudlmm2VQCQAAQPV75pln4qabbqo006dPnzjzzDMzqii/0vSrFRUVmX9ZY82aNbFgwYLEnH4VAAAoRi1atEiVa9OmTZ4r+b40febChQtj5cqV67yOfhUAAKBmmDx5cqofsD7mmGMyqOZb3bt3j2222SYxN2nSpMSMe8AAAAAkKZZ3nMVyzqwYggcAAABQ4KZPnx677LJLTJw4MfVn2rZtGxMnToy99947j5V9V/PmzaNJkyaJuer4Bcg0e7Zu3TqDSgAAAKrfAw88EBUVFZVmnnvuuSgtLY2SkpKc/7PXXnsl1vjRRx+lWqtt27aJa2255Zap/rtk3a8uWLAg1Rdl9KsAAEAx2mKLLVLlmjdvnudKvq9ly5apckuXLq307/pVAACAwvHqq6+myvXr1y/PlXzfAQcckJh5/fXX45tvvqk04x4wAAAASYrlHWexnDMrhuABAAAAFLC//e1vsdtuu8WcOXNSf2a33XaLV155Jbbbbrs8VvbD0jzc+/jjjzOo5Fuff/55rFixIjGX9sEkAAAAhaVp06bRrFmzxFzW/Wra/fSrAABAMWrXrl2qXN26dfNcyffVq1cvVW758uWV/l2/CgAAUDjmzZuXmNl0001js802y6Ca79p1110TMytWrIj3338/MeceMAAAAJUplnecxXLOrBiCBwAAAFCgxowZEwceeGB8/fXXqT9z9NFHx9///vfYeOON81jZj0vzcGzWrFkZVFL1/bbeeus8VwIAAEB1KdR+tWHDhrHppptmUA0AAEDNknYIXtKguXyoVatWqlxZWVliRr8KAABQGL788svEzFZbbZVBJd+X9v7rP//5z8RMofapEe4BAwAAZKVQe8eqvuMslnNmwRA8AAAAgAL0u9/9Lo499thYvXp1qnxJSUkMGzYs7r///qhTp06eq/tx2223XWLmvffey6CSb6X55coNN9wwNtpoowyqAQAAoDoUar/aoUOHDCoBAACoedq3b58qV5UfFMuVtIP3GjdunJjRrwIAABSGRYsWJWaaN2+e/0J+QKtWrVLl0gzyK9Q+1T1gAACA7BRq71jVd5zFcs4sGIIHAAAAUGDOOeecOP/886OioiJVvl69evGnP/0phg4dmufKknXt2jUx8+6772ZQybfeeeedxEznzp0zqAQAAIDqol8FAAAoLFtvvXXUr18/MTd37twMqvmuNIMPateuHXXr1k3M6VcBAAAKw4oVKxIzZWVlGVTyfWn654h0g+T1qQAAACQplt6xWM6ZBUPwAAAAAApERUVFnH766XHdddel/kyrVq3ixRdfjMMPPzyPlaWX5sHeZ599Fp988kkG1fzL5MmTEzM18cEeAAAAuZOmX3399ddj9erVGVTzL/pVAACAH1erVq1UvVyaS/65Nm/evMTMxhtvnGot/SoAAEBhaNCgQWJm8eLFGVTyfStXrkyVq1OnTmLGPWAAAACSFMs7zmI5ZxYMwQMAAAAoABUVFXHKKafErbfemvozO+ywQ0yePDm6deuWx8qqZtNNN41NNtkkMffqq69mUE1EeXl5TJkyJTG3++67Z1ANAAAA1aVz585Rq1atSjMrVqyIt956K5N6FixYEB9++GFiTr8KAAAUsx49eiRm3nzzzVizZk0G1XxrxowZiZntt98+1Vr6VQAAgMLQqFGjxMz8+fMzqOT7vvzyy1S5hg0bJmbcAwYAACBJsbzjLJZzZsEQPAAAAIACcNppp8Udd9yROt+/f/+YMGFCbL755nmsau306tUrMTN+/PgMKvnXL2kk/bJmSUlJ7LHHHpnUAwAAQPWoV69eql9kzKpfTbNP06ZNo1OnThlUAwAAUDP17NkzMbNo0aJUv3afS2+88UZiJu0QPP0qAABAYWjTpk1iZsaMGbF69eoMqvmu2bNnp8q1bt06Vc49YAAAACpTLO84i+WcWTAEDwAAAKCGu/zyy+O2225LnT/hhBNi7NixqX5Vsjrst99+iZknn3wyg0oi/vKXvyRmdt5552jevHkG1QAAANQMo0aNioqKimr75/nnn0+ssU2bNqnWSvOLhv+Wpl8dO3Zs6vXWRZp+tU+fPlFa6toHAABQvPr06RP16tVLzGX17jEi4tNPP40ZM2Yk5nbaaafUa+pXAQAAar527dolZlauXBlTp07NoJrvevXVVxMzJSUlsdVWW6Vazz1gAAAAkhTLO85iOWe+1byKAAAAAPj/RowYEcOGDUudv/zyy+Puu++ODTbYII9VrZt99903SkpKKs3MmjUr3n333bzX8sQTTyRmfvKTn+S9DgAAAKrf/vvvn5gZP358fPXVV3mtY/Xq1fH0008n5vSrAABAsatfv37ss88+ibnRo0fHmjVrMqgo3RcYSktLo1+/fqnX1K8CAADUfF26dEmVS3NvNdfGjRuXmGnfvn2qQfMR7gEDAACQrFjecRbLOfPNEDwAAACAGurvf/97nHHGGamyJSUlceutt8YVV1yR36JyoGXLltG1a9fE3F133ZXXOqZMmRKvv/56Yu6II47Iax0AAADUDF27do2WLVtWmlm1alWMHj06r3U8/vjj8fnnn1eaqVu3bo29iAIAAJClQw45JDHzySefpPrV+1wYOXJkYmaXXXaJFi1apF5TvwoAAFDzderUKZo1a5aYu//++zMb1B4RMW/evHjxxRcTc3vuuWfqNd0DBgAAIEmxvOMslnPmmyF4AAAAADXQ3Llz46ijjkp10aW0tDRGjRoVp556agaV5cagQYMSM6NGjYqlS5fmrYabb745MbP77rvHdtttl7caAAAAqDlKSkrimGOOSczddtttUV5enrc60vSrhx12WDRv3jxvNQAAABSKI444Iho1apSYu+KKK/Lay0VEvPTSSzFp0qTE3EEHHVSldfWrAAAANV9JSUnst99+ibkPP/wwHnvssfwX9L9uuOGGVL1i//79q7Sue8AAAABUpljecRbLOfPNEDwAAACAGmbNmjXxs5/9LBYsWJAqf8stt6S6TFKTHHPMMVG3bt1KM19++WVcffXVedn/zTffjPvuuy8xd9JJJ+VlfwAAAGqmE088MTEzY8aMuOeee/Ky/9ixY+PFF19MzOlXAQAA/qVRo0ap3pW++eabce+99+atjoqKirjooosSc7Vr147jjjuuyuvrVwEAAGq+AQMGpMpdcsklsWrVqjxXEzFnzpy45ZZbEnNNmzaNAw88sEpruwcMAABAkmJ5x1ks58wnQ/AAAAAAapjhw4fH//zP/6TKnn/++XHKKafkuaLca9q0aRx++OGJueHDh8f06dNzuvfq1avjlFNOiTVr1lSaa9asWRxxxBE53RsAAICabZtttomePXsm5i6++OKYP39+TvdevHhxDBkyJDG3zTbbRK9evXK6NwAAQCE7/fTTo7Q0+Vr8kCFDYs6cOXmpYcSIETFx4sTE3FFHHRUtW7as8vr6VQAAgJqvX79+0bZt28TczJkz47e//W1eaykvL48TTjghVqxYkZgdOHBg1KlTp0rruwcMAABAkmJ5x1ks58wnQ/AAAAAAapAZM2bEZZddlir705/+NO+XYPLp/PPPj5KSkkozK1eujMMOOyy+/vrrnO17wQUXpBoyePbZZ0e9evVyti8AAACF4cILL0zMfPHFF3HkkUfG6tWrc7JnRUVFHHfccfHBBx8kZi+55JKc7AkAALC+2GabbWLgwIGJua+++iqOPPLIWLJkSU73nzRpUpx99tmpsmm+gPBj9KsAAAA12wYbbBDnnXdequyvfvWr+Mtf/pK3Ws4777x4/vnnE3O1a9dOXfP/5R4wAAAASYrlHWexnDNfDMEDAAAAqEHOPffcWLlyZWKudevWcc899yReHqnJOnXqFEcddVRibubMmXHQQQfF8uXL13nPG264Ia677rrE3GabbRbnnHPOOu8HAABA4TnggANit912S8y99NJLMXDgwCgvL1/nPc8666z485//nJjr0qVLDBgwYJ33AwAAWN/8+te/jvr16yfmXnvttdh///1j6dKlOdl32rRpccABB8SqVasSs8ccc0x07tx5rffSrwIAANR8J5xwQrRr1y4xV15eHkcddVSMHTs25zVceumlqe7KRkSceuqpsfnmm6/VPu4BAwAAkKRY3nEWyznzxRA8AAAAgBri+eefjyeffDIxV1JSEqNGjYpmzZplUFV+DRs2LMrKyhJzL774YvTv3z/++c9/rvVeV111VZx99tmps379EQAAoHhdffXVqXIPPvhg/OxnP1vrL2ysXr06TjnllLjxxhtT5X//+98X9EB8AACAfNlss83i8ssvT5V96aWXYrfddosZM2as057PPfdc7LnnnrFw4cLEbKNGjeLaa69dp/0i9KsAAAA1XZ06dWLEiBGpssuWLYtDDjkkrrnmmlizZs06771kyZI45phj4te//nWqfOvWrWPYsGHrtKd7wAAAACQplnecxXLOfDAEDwAAAKCGuOyyy1LlKioqYp999omSkpIa888VV1yxVmdu165dlb6M0qVLlxg3blyV9pg/f34ccsghMXTo0KioqEjM9+zZMwYNGlSlPQAAAFi/7LnnnnHiiSemyj7yyCPRo0ePmDp1apX2mDVrVvTu3Ttuv/32VPmf//zn0bt37yrtAQAAUEzOO++86N+/f6rstGnTokuXLnH11VfHsmXLqrTP0qVL44ILLoh+/frFokWLUn3myiuvjE022aRK+/wQ/SoAAEDN17dv3zj99NNTZdesWRMXXnhh7LLLLvH888+v9Z6PPvpo7LDDDnH//fenyteqVStGjhwZjRo1Wus9I9wDBgAAIFmxvOMslnPmgyF4AAAAADXA1KlTY8KECdVdRrW48MILo0ePHqmyH330UfTt2zcOOOCAGDduXKW/fPn+++/HhRdeGO3bt4/HH3881fotWrSIBx98MEpLPTYDAAAodtdff31sueWWqbLTpk2Lbt26xdFHHx2TJk2q9MsX06ZNi1/84hex/fbbx8SJE1Otv/XWW8cdd9yRKgsAAFCsSkpKYsyYMdGmTZtU+WXLlsUll1wS7dq1i6FDh8b7779faX7evHlx9dVXx1ZbbRXXXnttlJeXp9rnsMMOizPPPDNVNg39KgAAQM03fPjw2GWXXVLnp0yZEnvvvXd07949RowYEZ999lniZ+bNmxe33HJLdOrUKQ499NCYM2dO6v2uvPLK2GeffVLnK+MeMAAAAEmK5R1nsZwz10oq0oy9B8izUaNGxXHHHVdpZs8994wXXnghm4IAAAAydtxxx8WoUaOqu4y1dvnll8cVV1yx1p+fPXt29OjRIxYsWFClz2244YbRrVu3aN++fTRu3DhWrFgRn332WUydOjVmzJhRpbVKS0vj6aefjn79+lXpcwAAAOTWCy+8EHvttVelmTZt2sSHH36Y91omT54cvXv3juXLl1fpc5tsskl079492rZtG40aNYqlS5fG3Llz47XXXqty3fXq1YtXXnklOnXqVKXPAQAAFKtZs2ZFr169Yv78+VX+7Oabbx7du3ePTTbZJJo2bRorVqyIefPmxbRp02L69OlVXm/bbbeNyZMnR8OGDav82croVwEAAGq+L774Inr16hXvvvtulT9bUlISbdu2jZ122ik23XTTaNKkSVRUVMTixYvj448/jrfffjtmz569VnWdcsopMWLEiLX67I9xDxgAAKBmqMnze4rlHWexnDOXDMEDaoSa/H+iAAAA+bZixYpo0aJFLFmypLpLWWvrOgQv4l8P9/baa69YtmxZboqqgpKSkrj55pvjtNNOy3xvAAAAvqsmDcGLiHjiiSfi0EMPjTVr1mSy33+qXbt2PPjgg/HTn/40870BAAAK2dtvvx19+/Zdq0F4udKhQ4cYN25ctG7dOi/r61cBAABqvk8//TT22WeftRqElw8nnnhi3HnnnVFSUpLztd0DBgAAqH41fX5PsbzjLJZz5kppdRcAAAAAUOyef/75gh6Alyvdu3ePp556Kpo0aZLpvqWlpXHnnXe6+AIAAMAPOuigg+KBBx6IOnXqZLpv3bp149FHHy2oSygAAAA1xfbbb1+tv2y/4447xvjx4/M2AC9CvwoAAFAINt1005g4cWL06tWrukuJYcOGxV133ZWXAXgR7gEDAACQrFjecRbLOXPFEDwAAACAajZx4sTqLqHG2HPPPWPChAnRtm3bTParW7dujBkzJk488cRM9gMAAKAwHXnkkfHXv/41Ntpoo0z2a9q0aYwdOzb233//TPYDAABYH22xxRYxceLEGDhwYKb7HnPMMTF+/PjYeOON876XfhUAAKDma9asWYwbNy4uvPDCKC3N/mvdrVq1irFjx8bQoUPzvpd7wAAAACQplnecxXLOXDAEDwAAAKCaTZ8+vbpLqFG23377ePPNN2PQoEF53adjx44xadKkOProo/O6DwAAAOuH3r17x1tvvRX77bdfXvfp2bNnvP7669GnT5+87gMAAFAMGjVqFPfee2889thj0bp167zuteGGG8ZDDz0UY8aMicaNG+d1r/+kXwUAAKj5ysrK4je/+U1MmDAhunbtmsmepaWlcfzxx8dbb70VBxxwQCZ7RrgHDAAAQLJiecdZLOdcV4bgAQAAAFSzjz/+uLpLqHEaN24co0ePjhdeeCF23XXXnK7dokWLGD58eLzxxhux44475nRtAAAA1m+tWrWKp556Kh599NHo2LFjTtdu06ZNjBw5MsaPHx9t27bN6doAAADF7uCDD45Zs2bFddddFxtvvHFO127atGkMGzYsZs+eHUceeWRO105LvwoAAFAYdt1115g8eXL88Y9/jJ133jkve5SVlcXPfvazeOONN+Kee+6JjTbaKC/7VMY9YAAAAJIUyzvOYjnnujAEDwAAAKCaLV68uLpLqLH23HPPePnll2P8+PExaNCgaNCgwVqtU1JSEj169Ijbb789Pvzwwzj77LNjgw02yHG1AAAAFItDDjkkpk2bFk8++WQceuihUbt27bVap1atWrHPPvvE/fffH7NmzYrBgwdHSUlJjqsFAAAgIqJu3bpx1llnxdy5c+NPf/pT7Lfffmvdz5WWlsaee+4ZI0aMiA8//DCGDh0ajRs3znHFVadfBQAAqPlKSkriqKOOiilTpsTLL78cp512Wmy++ebrvGaPHj3it7/9bXz00Ufx4IMPRqdOnXJU8dpzDxgAAIAkxfKOs1jOuTZKKioqKqq7CAAAAABIY9WqVTFp0qR48cUXY/r06TFr1qz49NNPY8mSJbFs2bLYYIMNon79+rHxxhvHFltsEdtvv3107do1+vTpE61ataru8gEAAFhPLVu2LCZMmBAvvfRSvPvuuzFr1qyYP39+LF26NJYvXx5lZWXRoEGDaNWqVWyxxRbRqVOn6N69e/Tp0yeaNWtW3eUDAAAUreXLl8eECRNi4sSJMWPGjJg1a1Z89tlnsXTp0li6dGnUqVMnmjRpEk2bNo02bdpE586do3PnzrHHHnsUxPtH/SoAAEDhePfdd2Py5MkxZcqU+OCDD+Ljjz+Ozz///P/3cLVq1Yr69etHw4YNY7PNNos2bdrEtttuG926dYtddtklWrRoUd1HSOQeMAAAAEmK5R1nsZwzDUPwAAAAAAAAAAAAAAAAAAAAAAAAAMhMaXUXAAAAAAAAAAAAAAAAAAAAAAAAAEDxMAQPAAAAAAAAAAAAAAAAAAAAAAAAgMwYggcAAAAAAAAAAAAAAAAAAAAAAABAZgzBAwAAAAAAAAAAAAAAAAAAAAAAACAzhuABAAAAAAAAAAAAAAAAAAAAAAAAkBlD8AAAAAAAAAAAAAAAAAAAAAAAAADIjCF4AAAAAAAAAAAAAAAAAAAAAAAAAGTGEDwAAAAAAAAAAAAAAAAAAAAAAAAAMmMIHgAAAAAAAAAAAAAAAAAAAAAAAACZMQQPAAAAAAAAAAAAAAAAAAAAAAAAgMwYggcAAAAAAAAAAAAAAAAAAAAAAABAZgzBAwAAAAAAAAAAAAAAAAAAAAAAACAzhuABAAAAAAAAAAAAAAAAAAAAAAAAkBlD8AAAAAAAAAAAAAAAAAAAAAAAAADIjCF4AAAAAAAAAAAAAAAAAAAAAAAAAGTGEDwAAAAAAAAAAAAAAAAAAAAAAAAAMmMIHgAAAAAAAAAAAAAAAAAAAAAAAACZMQQPAAAAAAAAAAAAAAAAAAAAAAAAgMwYggcAAAAAAAAAAAAAAAAAAAAAAABAZgzBAwAAAAAAAAAAAAAAAAAAAAAAACAzhuABAAAAAAAAAAAAAAAAAAAAAAAAkBlD8AAAAAAAAAAAAAAAAAAAAAAAAADIjCF4AAAAAAAAAAAAAAAAAAAAAAAAAGTGEDwAAAAAAAAAAAAAAAAAAAAAAAAAMmMIHgAAAAAAAAAAAAAAAAAAAAAAAACZMQQPAAAAAAAAAAAAAAAAAAAAAAAAgMwYggcAAAAAAAAAAAAAAAAAAAAAAABAZgzBAwAAAAAAAAAAAAAAAAAAAAAAACAzhuABAAAAAAAAAAAAAAAAAAAAAAAAkBlD8AAAAAAAAAAAAAAAAAAAAAAAAADIjCF4AAAAAAAAAAAAAAAAAAAAAAAAAGTGEDwAAAAAAAAAAAAAAAAAAAAAAAAAMmMIHgAAAAAAAAAAAAAAAAAAAAAAAACZMQQPAAAAAAAAAAAAAAAAAAAAAAAAgMwYggcAAAAAAAAAAAAAAAAAAAAAAABAZgzBAwAAAAAAAAAAAAAAAAAAAAAAACAzhuABAAAAAAAAAAAAAAAAAAAAAAAAkBlD8AAAAAAAAAAAAAAAAAAAAAAAAADIjCF4AAAAAAAAAAAAAAAAAAAAAAAAAGTGEDwAAAAAAAAAAAAAAAAAAAAAAAAAMmMIHgAAAAAAAAAAAAAAAAAAAAAAAACZMQQPAAAAAAAAAAAAAAAAAAAAAAAAgMwYggcAAAAAAAAAAAAAAAAAAAAAAABAZgzBAwAAAAAAAAAAAAAAAAAAAAAAACAzhuABAAAAAAAAAAAAAAAAAAAAAAAAkBlD8AAAAAAAAAAAAAAAAAAAAAAAAADIjCF4AAAAAAAAAAAAAAAAAAAAAAAAAGTGEDwAAAAAAAAAAAAAAAAAAAAAAAAAMmMIHgAAAAAAAAAAAAAAAAAAAAAAAACZ+X/dJiYRNQAAAANJREFUxs/GUo0IVAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 6000x1800 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20,6), dpi = 300)\n",
    "# Assuming df_binded is your DataFrame\n",
    "df_binded['flank_size'].hist(bins=1000)\n",
    "# Set the x-axis limit to show only 0 to 750\n",
    "plt.xlim(1, 100)\n",
    "ax.set_ylim((0,1))\n",
    "ax.yaxis.set_ticks([x*10000 for x in [50,100,150]])\n",
    "ax.yaxis.set_ticklabels(['.5M', '1M', '1.5M'], size=25, fontfamily=\"Arial\", fontweight=100)\n",
    "ax.xaxis.set_ticks([1,20,40,60,80,100])\n",
    "ax.xaxis.set_ticklabels([1,20,40,60,80,100], size=25, fontfamily=\"Arial\", fontweight=100)\n",
    "# ax.yaxis.set_tick_params(labelbottom=False, labeltop=True, length=0)\n",
    "ax.set_axisbelow(True)\n",
    "ax.grid(axis = \"y\", color=\"#A8BAC4\", lw=1.2)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"bottom\"].set_visible(False)\n",
    "ax.spines[\"left\"].set_lw(1.5)\n",
    "ax.spines[\"left\"].set_capstyle(\"butt\")\n",
    "plt.yticks(fontsize = 25)\n",
    "# Display the plot\n",
    "plt.savefig(f'/Users/kexindong/Documents/GitHub/Output/h2m_database/figures/database/hist_flank_size.svg', bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "name_list = ['AACR','COSMIC','CLINVAR']\n",
    "list_of_n = []\n",
    "list_of_fs = []\n",
    "list_of_input = []\n",
    "list_of_output = []\n",
    "list_of_class_0 = []\n",
    "list_of_class_1 = []\n",
    "list_of_class_2 = []\n",
    "list_of_class_3 = []\n",
    "\n",
    "for x in [0,1,2]:\n",
    "    name = name_list[x]\n",
    "    for fs in [1,5,11,21]:\n",
    "        df = df_binded[df_binded['Database']==name].reset_index(drop=True)\n",
    "        df = df[(df['flank_size'] >= fs) or (df['flank_size'] >= 0)] \n",
    "        list_of_n.append(name)\n",
    "        list_of_fs.append(fs)\n",
    "        l_1 = len(df['index'].unique())\n",
    "        df_sub = df[df['status']==True]\n",
    "        l_2 = len(df_sub['index'].unique())\n",
    "        print(f'{name} with {fs}: {l_2} out of {l_1} ({l_2/l_1* 100:.2f}%) mutations can be modeled.\\n')\n",
    "        list_of_output.append(l_2)\n",
    "        list_of_input.append(l_1)\n",
    "        list_of_class_0.append(len(df[df['class']==0]['index'].unique()))\n",
    "        list_of_class_1.append(len(df[df['class']==1]['index'].unique()))\n",
    "        list_of_class_2.append(len(df[df['class']==2]['index'].unique()))\n",
    "        list_of_class_3.append(len(df[df['class']==3]['index'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AACR with 1: 656168 out of 803974 (81.62%) mutations can be modeled.\n",
      "\n",
      "AACR with 5: 569309 out of 803974 (70.81%) mutations can be modeled.\n",
      "\n",
      "AACR with 11: 491026 out of 803974 (61.07%) mutations can be modeled.\n",
      "\n",
      "AACR with 21: 403953 out of 803974 (50.24%) mutations can be modeled.\n",
      "\n",
      "COSMIC with 1: 834682 out of 1044671 (79.90%) mutations can be modeled.\n",
      "\n",
      "COSMIC with 5: 707170 out of 1044671 (67.69%) mutations can be modeled.\n",
      "\n",
      "COSMIC with 11: 639560 out of 1044671 (61.22%) mutations can be modeled.\n",
      "\n",
      "COSMIC with 21: 507215 out of 1044671 (48.55%) mutations can be modeled.\n",
      "\n",
      "CLINVAR with 1: 1506299 out of 2038473 (73.89%) mutations can be modeled.\n",
      "\n",
      "CLINVAR with 5: 1198748 out of 2038473 (58.81%) mutations can be modeled.\n",
      "\n",
      "CLINVAR with 11: 973006 out of 2038473 (47.73%) mutations can be modeled.\n",
      "\n",
      "CLINVAR with 21: 742799 out of 2038473 (36.44%) mutations can be modeled.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "name_list = ['AACR','COSMIC','CLINVAR']\n",
    "list_of_n = []\n",
    "list_of_fs = []\n",
    "list_of_input = []\n",
    "list_of_output = []\n",
    "list_of_class_0 = []\n",
    "list_of_class_1 = []\n",
    "list_of_class_2 = []\n",
    "list_of_class_3 = []\n",
    "\n",
    "for x in [0,1,2]:\n",
    "    name = name_list[x]\n",
    "    df = df_binded[df_binded['Database']==name].reset_index(drop=True)\n",
    "    l_1 = len(df['ID_human'].unique())\n",
    "    for fs in [0,2,4,10,20,30,40,50,100,200,500,1000]:\n",
    "        df = df[df['flank_size'] >= fs] \n",
    "        list_of_n.append(name)\n",
    "        list_of_fs.append(fs)\n",
    "        df_sub = df[df['status']==True]\n",
    "        l_2 = len(df_sub['ID_human'].unique())\n",
    "        print(f'{name} with {fs}: {l_2} out of {l_1} ({l_2/l_1* 100:.2f}%) mutations can be modeled.\\n')\n",
    "        list_of_output.append(l_2)\n",
    "        list_of_input.append(l_1)\n",
    "        list_of_class_0.append(len(df[df['class']==0]['ID_human'].unique()))\n",
    "        list_of_class_1.append(len(df[df['class']==1]['ID_human'].unique()))\n",
    "        list_of_class_2.append(len(df[df['class']==2]['ID_human'].unique()))\n",
    "        list_of_class_3.append(len(df[df['class']==3]['ID_human'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(zip(list_of_n, list_of_fs,\n",
    "list_of_input,\n",
    "list_of_output,\n",
    "list_of_class_0,\n",
    "list_of_class_1,\n",
    "list_of_class_2,\n",
    "list_of_class_3))\n",
    "df.columns = ['Database','Flank_Size','Num_Input','Num_Output','Class_0','Class_1','Class_2','Class_3']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Database</th>\n",
       "      <th>Flank_Size</th>\n",
       "      <th>Num_Input</th>\n",
       "      <th>Num_Output</th>\n",
       "      <th>Class_0</th>\n",
       "      <th>Class_1</th>\n",
       "      <th>Class_2</th>\n",
       "      <th>Class_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AACR</td>\n",
       "      <td>1</td>\n",
       "      <td>803974</td>\n",
       "      <td>656168</td>\n",
       "      <td>537651</td>\n",
       "      <td>31913</td>\n",
       "      <td>86915</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AACR</td>\n",
       "      <td>5</td>\n",
       "      <td>803974</td>\n",
       "      <td>569309</td>\n",
       "      <td>478084</td>\n",
       "      <td>27402</td>\n",
       "      <td>63953</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AACR</td>\n",
       "      <td>11</td>\n",
       "      <td>803974</td>\n",
       "      <td>491026</td>\n",
       "      <td>415992</td>\n",
       "      <td>23561</td>\n",
       "      <td>51537</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AACR</td>\n",
       "      <td>21</td>\n",
       "      <td>803974</td>\n",
       "      <td>403953</td>\n",
       "      <td>344597</td>\n",
       "      <td>19107</td>\n",
       "      <td>40282</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>COSMIC</td>\n",
       "      <td>1</td>\n",
       "      <td>1044671</td>\n",
       "      <td>834682</td>\n",
       "      <td>692349</td>\n",
       "      <td>52114</td>\n",
       "      <td>90869</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>COSMIC</td>\n",
       "      <td>5</td>\n",
       "      <td>1044671</td>\n",
       "      <td>707170</td>\n",
       "      <td>639845</td>\n",
       "      <td>46937</td>\n",
       "      <td>20680</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>COSMIC</td>\n",
       "      <td>11</td>\n",
       "      <td>1044671</td>\n",
       "      <td>639560</td>\n",
       "      <td>587041</td>\n",
       "      <td>43216</td>\n",
       "      <td>9494</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>COSMIC</td>\n",
       "      <td>21</td>\n",
       "      <td>1044671</td>\n",
       "      <td>507215</td>\n",
       "      <td>466917</td>\n",
       "      <td>35535</td>\n",
       "      <td>4851</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CLINVAR</td>\n",
       "      <td>1</td>\n",
       "      <td>2038473</td>\n",
       "      <td>1506299</td>\n",
       "      <td>1209054</td>\n",
       "      <td>71314</td>\n",
       "      <td>226858</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CLINVAR</td>\n",
       "      <td>5</td>\n",
       "      <td>2038473</td>\n",
       "      <td>1198748</td>\n",
       "      <td>1034049</td>\n",
       "      <td>58723</td>\n",
       "      <td>106341</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CLINVAR</td>\n",
       "      <td>11</td>\n",
       "      <td>2038473</td>\n",
       "      <td>973006</td>\n",
       "      <td>854486</td>\n",
       "      <td>47854</td>\n",
       "      <td>70807</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CLINVAR</td>\n",
       "      <td>21</td>\n",
       "      <td>2038473</td>\n",
       "      <td>742799</td>\n",
       "      <td>658849</td>\n",
       "      <td>35996</td>\n",
       "      <td>48003</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Database  Flank_Size  Num_Input  Num_Output  Class_0  Class_1  Class_2  \\\n",
       "0      AACR           1     803974      656168   537651    31913    86915   \n",
       "1      AACR           5     803974      569309   478084    27402    63953   \n",
       "2      AACR          11     803974      491026   415992    23561    51537   \n",
       "3      AACR          21     803974      403953   344597    19107    40282   \n",
       "4    COSMIC           1    1044671      834682   692349    52114    90869   \n",
       "5    COSMIC           5    1044671      707170   639845    46937    20680   \n",
       "6    COSMIC          11    1044671      639560   587041    43216     9494   \n",
       "7    COSMIC          21    1044671      507215   466917    35535     4851   \n",
       "8   CLINVAR           1    2038473     1506299  1209054    71314   226858   \n",
       "9   CLINVAR           5    2038473     1198748  1034049    58723   106341   \n",
       "10  CLINVAR          11    2038473      973006   854486    47854    70807   \n",
       "11  CLINVAR          21    2038473      742799   658849    35996    48003   \n",
       "\n",
       "    Class_3  \n",
       "0         0  \n",
       "1         0  \n",
       "2         0  \n",
       "3         0  \n",
       "4         0  \n",
       "5         0  \n",
       "6         0  \n",
       "7         0  \n",
       "8         0  \n",
       "9         0  \n",
       "10        0  \n",
       "11        0  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/fs_stat',index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
