{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Retrieve the human-mouse orthologous genes and canonical transcripts  \n",
    "This script:\n",
    "- Retrieved human mutation-derived immunogenic peptides and mapped them to the mouse genome;\n",
    "- Retrieved mouse tumor-mutation derived peptides and mapped them to the mouse genome;\n",
    "\n",
    "\n",
    "**Author**: Kexin Dong  \n",
    "**Date**: May 19, 2024  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## H2M  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kexindong/Documents/GitHub/h2m-public/.conda/lib/python3.10/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import bioh2m as h2m\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "import requests\n",
    "import time\n",
    "import ast\n",
    "from adjustText import adjust_text\n",
    "import matplotlib.patheffects as PathEffects\n",
    "import seaborn as sns\n",
    "plt.rcParams['font.family'] = 'sans-serif'\n",
    "plt.rcParams['font.sans-serif'] = ['Arial']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load and pre-preocess data (N = 1,835)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_h_validated = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/neoantigen/validated_tsnadb2_download.txt', header=0, sep='\\t')\n",
    "df_h_validated = df_h_validated[['Mutation Type','Gene','Mutation','Mutant Peptide','HLA']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Mutation</th>\n",
       "      <th>Mutant Peptide</th>\n",
       "      <th>HLA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNV</td>\n",
       "      <td>SEC22C</td>\n",
       "      <td>SEC22C_H248Y</td>\n",
       "      <td>AEHSLQVAY</td>\n",
       "      <td>B44:03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TKT</td>\n",
       "      <td>TKT_R438W</td>\n",
       "      <td>AMFWSVPTV</td>\n",
       "      <td>A02:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNV</td>\n",
       "      <td>KIF6B</td>\n",
       "      <td>KIF6B_L1020P</td>\n",
       "      <td>APARLERRHSAL</td>\n",
       "      <td>B07:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNV</td>\n",
       "      <td>ACSS3</td>\n",
       "      <td>ACSS3_H194Y</td>\n",
       "      <td>ARIGAIYSL</td>\n",
       "      <td>C07:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SNV</td>\n",
       "      <td>AP2S1</td>\n",
       "      <td>AP2S1_N86K</td>\n",
       "      <td>AYLEAIHKF</td>\n",
       "      <td>A24:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1851</th>\n",
       "      <td>INDEL</td>\n",
       "      <td>C2CD3</td>\n",
       "      <td>C2CD3_R1186del</td>\n",
       "      <td>RSPTAEGVL</td>\n",
       "      <td>B07:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1852</th>\n",
       "      <td>INDEL</td>\n",
       "      <td>SPEN</td>\n",
       "      <td>KLLE1864_1867del</td>\n",
       "      <td>SPRGEAQLK</td>\n",
       "      <td>B07:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1853</th>\n",
       "      <td>INDEL</td>\n",
       "      <td>DDX3X</td>\n",
       "      <td>TR213_214del</td>\n",
       "      <td>TRYPTPVQK</td>\n",
       "      <td>A03:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1854</th>\n",
       "      <td>INDEL</td>\n",
       "      <td>PDE12</td>\n",
       "      <td>PDE12_S50del</td>\n",
       "      <td>VPSEPKLLS</td>\n",
       "      <td>B07:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1855</th>\n",
       "      <td>INDEL</td>\n",
       "      <td>ZMAT3</td>\n",
       "      <td>ZMAT3_V240fs</td>\n",
       "      <td>WPVLLLNV</td>\n",
       "      <td>B51:01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1835 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Mutation Type    Gene          Mutation Mutant Peptide     HLA\n",
       "0              SNV  SEC22C      SEC22C_H248Y      AEHSLQVAY  B44:03\n",
       "1              SNV     TKT         TKT_R438W      AMFWSVPTV  A02:01\n",
       "2              SNV   KIF6B      KIF6B_L1020P   APARLERRHSAL  B07:02\n",
       "3              SNV   ACSS3       ACSS3_H194Y      ARIGAIYSL  C07:02\n",
       "4              SNV   AP2S1        AP2S1_N86K      AYLEAIHKF  A24:02\n",
       "...            ...     ...               ...            ...     ...\n",
       "1851         INDEL   C2CD3    C2CD3_R1186del      RSPTAEGVL  B07:02\n",
       "1852         INDEL    SPEN  KLLE1864_1867del      SPRGEAQLK  B07:02\n",
       "1853         INDEL   DDX3X      TR213_214del      TRYPTPVQK  A03:01\n",
       "1854         INDEL   PDE12      PDE12_S50del      VPSEPKLLS  B07:02\n",
       "1855         INDEL   ZMAT3      ZMAT3_V240fs       WPVLLLNV  B51:01\n",
       "\n",
       "[1835 rows x 5 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_h_validated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SNV-derived (N = 1,627)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Mutation</th>\n",
       "      <th>Mutant Peptide</th>\n",
       "      <th>HLA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNV</td>\n",
       "      <td>SEC22C</td>\n",
       "      <td>H248Y</td>\n",
       "      <td>AEHSLQVAY</td>\n",
       "      <td>B44:03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TKT</td>\n",
       "      <td>R438W</td>\n",
       "      <td>AMFWSVPTV</td>\n",
       "      <td>A02:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNV</td>\n",
       "      <td>KIF6B</td>\n",
       "      <td>L1020P</td>\n",
       "      <td>APARLERRHSAL</td>\n",
       "      <td>B07:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNV</td>\n",
       "      <td>ACSS3</td>\n",
       "      <td>H194Y</td>\n",
       "      <td>ARIGAIYSL</td>\n",
       "      <td>C07:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SNV</td>\n",
       "      <td>AP2S1</td>\n",
       "      <td>N86K</td>\n",
       "      <td>AYLEAIHKF</td>\n",
       "      <td>A24:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1622</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TNFAIP1</td>\n",
       "      <td>R48L</td>\n",
       "      <td>YTTVRALTL</td>\n",
       "      <td>B08:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1623</th>\n",
       "      <td>SNV</td>\n",
       "      <td>CEL</td>\n",
       "      <td>T471L</td>\n",
       "      <td>YVFGKPFATPLGY</td>\n",
       "      <td>A29:02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1624</th>\n",
       "      <td>SNV</td>\n",
       "      <td>LRPPRC</td>\n",
       "      <td>T1335I</td>\n",
       "      <td>YVSEKDVISAK</td>\n",
       "      <td>A68:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1625</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TSR1</td>\n",
       "      <td>H561Y</td>\n",
       "      <td>YVSEVPVSV</td>\n",
       "      <td>C17:01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1626</th>\n",
       "      <td>SNV</td>\n",
       "      <td>HEPH</td>\n",
       "      <td>I622T</td>\n",
       "      <td>YYSAVDPTK</td>\n",
       "      <td>A23:01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1627 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Mutation Type     Gene Mutation Mutant Peptide     HLA\n",
       "0              SNV   SEC22C    H248Y      AEHSLQVAY  B44:03\n",
       "1              SNV      TKT    R438W      AMFWSVPTV  A02:01\n",
       "2              SNV    KIF6B   L1020P   APARLERRHSAL  B07:02\n",
       "3              SNV    ACSS3    H194Y      ARIGAIYSL  C07:02\n",
       "4              SNV    AP2S1     N86K      AYLEAIHKF  A24:02\n",
       "...            ...      ...      ...            ...     ...\n",
       "1622           SNV  TNFAIP1     R48L      YTTVRALTL  B08:01\n",
       "1623           SNV      CEL    T471L  YVFGKPFATPLGY  A29:02\n",
       "1624           SNV   LRPPRC   T1335I    YVSEKDVISAK  A68:01\n",
       "1625           SNV     TSR1    H561Y      YVSEVPVSV  C17:01\n",
       "1626           SNV     HEPH    I622T      YYSAVDPTK  A23:01\n",
       "\n",
       "[1627 rows x 5 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_h_validated = df_h_validated.dropna(subset ='Mutation')\n",
    "df_h_validated = df_h_validated[df_h_validated['Mutation Type']!='Fusion']\n",
    "df_h_validated['Mutation'] = [str(x).split('_')[1] for x in df_h_validated['Mutation']]\n",
    "df_h_validated = df_h_validated[df_h_validated['Mutation Type']=='SNV'].reset_index(drop=True)\n",
    "df_h_validated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select MHC-I binding (N = 1,211)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Mutation</th>\n",
       "      <th>Mutant Peptide</th>\n",
       "      <th>HLA</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>end_h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNV</td>\n",
       "      <td>SEC22C</td>\n",
       "      <td>H248Y</td>\n",
       "      <td>AEHSLQVAY</td>\n",
       "      <td>B44:03</td>\n",
       "      <td>H</td>\n",
       "      <td>248</td>\n",
       "      <td>Y</td>\n",
       "      <td>248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TKT</td>\n",
       "      <td>R438W</td>\n",
       "      <td>AMFWSVPTV</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>R</td>\n",
       "      <td>438</td>\n",
       "      <td>W</td>\n",
       "      <td>438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNV</td>\n",
       "      <td>KIF6B</td>\n",
       "      <td>L1020P</td>\n",
       "      <td>APARLERRHSAL</td>\n",
       "      <td>B07:02</td>\n",
       "      <td>L</td>\n",
       "      <td>1020</td>\n",
       "      <td>P</td>\n",
       "      <td>1020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNV</td>\n",
       "      <td>ACSS3</td>\n",
       "      <td>H194Y</td>\n",
       "      <td>ARIGAIYSL</td>\n",
       "      <td>C07:02</td>\n",
       "      <td>H</td>\n",
       "      <td>194</td>\n",
       "      <td>Y</td>\n",
       "      <td>194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SNV</td>\n",
       "      <td>AP2S1</td>\n",
       "      <td>N86K</td>\n",
       "      <td>AYLEAIHKF</td>\n",
       "      <td>A24:02</td>\n",
       "      <td>N</td>\n",
       "      <td>86</td>\n",
       "      <td>K</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1206</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TNFAIP1</td>\n",
       "      <td>R48L</td>\n",
       "      <td>YTTVRALTL</td>\n",
       "      <td>B08:01</td>\n",
       "      <td>R</td>\n",
       "      <td>48</td>\n",
       "      <td>L</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1207</th>\n",
       "      <td>SNV</td>\n",
       "      <td>CEL</td>\n",
       "      <td>T471L</td>\n",
       "      <td>YVFGKPFATPLGY</td>\n",
       "      <td>A29:02</td>\n",
       "      <td>T</td>\n",
       "      <td>471</td>\n",
       "      <td>L</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1208</th>\n",
       "      <td>SNV</td>\n",
       "      <td>LRPPRC</td>\n",
       "      <td>T1335I</td>\n",
       "      <td>YVSEKDVISAK</td>\n",
       "      <td>A68:01</td>\n",
       "      <td>T</td>\n",
       "      <td>1335</td>\n",
       "      <td>I</td>\n",
       "      <td>1335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1209</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TSR1</td>\n",
       "      <td>H561Y</td>\n",
       "      <td>YVSEVPVSV</td>\n",
       "      <td>C17:01</td>\n",
       "      <td>H</td>\n",
       "      <td>561</td>\n",
       "      <td>Y</td>\n",
       "      <td>561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1210</th>\n",
       "      <td>SNV</td>\n",
       "      <td>HEPH</td>\n",
       "      <td>I622T</td>\n",
       "      <td>YYSAVDPTK</td>\n",
       "      <td>A23:01</td>\n",
       "      <td>I</td>\n",
       "      <td>622</td>\n",
       "      <td>T</td>\n",
       "      <td>622</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1211 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Mutation Type     Gene Mutation Mutant Peptide     HLA ref_seq_h start_h  \\\n",
       "0              SNV   SEC22C    H248Y      AEHSLQVAY  B44:03         H     248   \n",
       "1              SNV      TKT    R438W      AMFWSVPTV  A02:01         R     438   \n",
       "2              SNV    KIF6B   L1020P   APARLERRHSAL  B07:02         L    1020   \n",
       "3              SNV    ACSS3    H194Y      ARIGAIYSL  C07:02         H     194   \n",
       "4              SNV    AP2S1     N86K      AYLEAIHKF  A24:02         N      86   \n",
       "...            ...      ...      ...            ...     ...       ...     ...   \n",
       "1206           SNV  TNFAIP1     R48L      YTTVRALTL  B08:01         R      48   \n",
       "1207           SNV      CEL    T471L  YVFGKPFATPLGY  A29:02         T     471   \n",
       "1208           SNV   LRPPRC   T1335I    YVSEKDVISAK  A68:01         T    1335   \n",
       "1209           SNV     TSR1    H561Y      YVSEVPVSV  C17:01         H     561   \n",
       "1210           SNV     HEPH    I622T      YYSAVDPTK  A23:01         I     622   \n",
       "\n",
       "     alt_seq_h end_h  \n",
       "0            Y   248  \n",
       "1            W   438  \n",
       "2            P  1020  \n",
       "3            Y   194  \n",
       "4            K    86  \n",
       "...        ...   ...  \n",
       "1206         L    48  \n",
       "1207         L   471  \n",
       "1208         I  1335  \n",
       "1209         Y   561  \n",
       "1210         T   622  \n",
       "\n",
       "[1211 rows x 9 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def split_mutation(mutation):\n",
    "    parts = re.findall(r'([A-Z])(\\d+)([A-Z])', mutation)\n",
    "    if parts:\n",
    "        return parts[0] \n",
    "    return ('', '', '') \n",
    "\n",
    "df_h_validated[['ref_seq_h', 'start_h', 'alt_seq_h']] = df_h_validated['Mutation'].apply(lambda x: pd.Series(split_mutation(x)))\n",
    "df_h_validated['end_h'] = df_h_validated['start_h']\n",
    "df_h_validated = df_h_validated.dropna(subset = 'HLA').reset_index(drop=True)\n",
    "df_h_validated = df_h_validated[[x[0] in ['A','B','C'] for x in df_h_validated['HLA']]]\n",
    "df_h_validated = df_h_validated.reset_index(drop=True)\n",
    "df_h_validated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Select 9-aa-long peptides (N = 1,211)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Mutation</th>\n",
       "      <th>Mutant Peptide</th>\n",
       "      <th>HLA</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>peptide_index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNV</td>\n",
       "      <td>SEC22C</td>\n",
       "      <td>H248Y</td>\n",
       "      <td>AEHSLQVAY</td>\n",
       "      <td>B44:03</td>\n",
       "      <td>H</td>\n",
       "      <td>248</td>\n",
       "      <td>Y</td>\n",
       "      <td>248</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TKT</td>\n",
       "      <td>R438W</td>\n",
       "      <td>AMFWSVPTV</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>R</td>\n",
       "      <td>438</td>\n",
       "      <td>W</td>\n",
       "      <td>438</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNV</td>\n",
       "      <td>ACSS3</td>\n",
       "      <td>H194Y</td>\n",
       "      <td>ARIGAIYSL</td>\n",
       "      <td>C07:02</td>\n",
       "      <td>H</td>\n",
       "      <td>194</td>\n",
       "      <td>Y</td>\n",
       "      <td>194</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNV</td>\n",
       "      <td>AP2S1</td>\n",
       "      <td>N86K</td>\n",
       "      <td>AYLEAIHKF</td>\n",
       "      <td>A24:02</td>\n",
       "      <td>N</td>\n",
       "      <td>86</td>\n",
       "      <td>K</td>\n",
       "      <td>86</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TMEM48</td>\n",
       "      <td>F169L</td>\n",
       "      <td>CLNEYHLFL</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>F</td>\n",
       "      <td>169</td>\n",
       "      <td>L</td>\n",
       "      <td>169</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>637</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TOX2</td>\n",
       "      <td>S382Y</td>\n",
       "      <td>YPAPQPPVL</td>\n",
       "      <td>B08:01</td>\n",
       "      <td>S</td>\n",
       "      <td>382</td>\n",
       "      <td>Y</td>\n",
       "      <td>382</td>\n",
       "      <td>637</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>638</th>\n",
       "      <td>SNV</td>\n",
       "      <td>NR1D1</td>\n",
       "      <td>G39D</td>\n",
       "      <td>YSDNSNDSF</td>\n",
       "      <td>A01:01</td>\n",
       "      <td>G</td>\n",
       "      <td>39</td>\n",
       "      <td>D</td>\n",
       "      <td>39</td>\n",
       "      <td>638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>639</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TNFAIP1</td>\n",
       "      <td>R48L</td>\n",
       "      <td>YTTVRALTL</td>\n",
       "      <td>B08:01</td>\n",
       "      <td>R</td>\n",
       "      <td>48</td>\n",
       "      <td>L</td>\n",
       "      <td>48</td>\n",
       "      <td>639</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>640</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TSR1</td>\n",
       "      <td>H561Y</td>\n",
       "      <td>YVSEVPVSV</td>\n",
       "      <td>C17:01</td>\n",
       "      <td>H</td>\n",
       "      <td>561</td>\n",
       "      <td>Y</td>\n",
       "      <td>561</td>\n",
       "      <td>640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>641</th>\n",
       "      <td>SNV</td>\n",
       "      <td>HEPH</td>\n",
       "      <td>I622T</td>\n",
       "      <td>YYSAVDPTK</td>\n",
       "      <td>A23:01</td>\n",
       "      <td>I</td>\n",
       "      <td>622</td>\n",
       "      <td>T</td>\n",
       "      <td>622</td>\n",
       "      <td>641</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>642 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Mutation Type     Gene Mutation Mutant Peptide     HLA ref_seq_h start_h  \\\n",
       "0             SNV   SEC22C    H248Y      AEHSLQVAY  B44:03         H     248   \n",
       "1             SNV      TKT    R438W      AMFWSVPTV  A02:01         R     438   \n",
       "2             SNV    ACSS3    H194Y      ARIGAIYSL  C07:02         H     194   \n",
       "3             SNV    AP2S1     N86K      AYLEAIHKF  A24:02         N      86   \n",
       "4             SNV   TMEM48    F169L      CLNEYHLFL  A02:01         F     169   \n",
       "..            ...      ...      ...            ...     ...       ...     ...   \n",
       "637           SNV     TOX2    S382Y      YPAPQPPVL  B08:01         S     382   \n",
       "638           SNV    NR1D1     G39D      YSDNSNDSF  A01:01         G      39   \n",
       "639           SNV  TNFAIP1     R48L      YTTVRALTL  B08:01         R      48   \n",
       "640           SNV     TSR1    H561Y      YVSEVPVSV  C17:01         H     561   \n",
       "641           SNV     HEPH    I622T      YYSAVDPTK  A23:01         I     622   \n",
       "\n",
       "    alt_seq_h end_h  peptide_index  \n",
       "0           Y   248              0  \n",
       "1           W   438              1  \n",
       "2           Y   194              2  \n",
       "3           K    86              3  \n",
       "4           L   169              4  \n",
       "..        ...   ...            ...  \n",
       "637         Y   382            637  \n",
       "638         D    39            638  \n",
       "639         L    48            639  \n",
       "640         Y   561            640  \n",
       "641         T   622            641  \n",
       "\n",
       "[642 rows x 10 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_h_validated = df_h_validated[[len(x)==9 for x in df_h_validated['Mutant Peptide']]].reset_index(drop=True)\n",
    "df_h_validated['peptide_index'] = range(len(df_h_validated))\n",
    "df_h_validated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check and rename genes  (N = 586)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_h_validated['gene_name_h'] = df_h_validated['Gene']\n",
    "dict_of_aacr_symbol = {\n",
    "    'H3F3A':'H3-3A',\n",
    "    'H3F3B':'H3-3B',\n",
    "    'WHSC1':'NSD1',\n",
    "    'GPR124':'ADGRA2',\n",
    "    'MKL1':'MRTFA',\n",
    "    'PARK2':'PRKN',\n",
    "    'MRE11A':'MRE11',\n",
    "    'MLLT4':'AFDN',\n",
    "    'CASC5': 'KNL1',\n",
    "    'MEF2BNB-MEF2B':'BORCS8-MEF2B',\n",
    "    'PAK7':'PAK5',\n",
    "    'HIST1H1C':'H1-2',\n",
    "    'HIST1H3B': 'H3C1',\n",
    "    'HIST1H1E': 'H1-4',\n",
    "    'HIST1H2BD': 'H2BC5',\n",
    "    'HIST3H3': 'H3-4',\n",
    "    'HIST1H3D': 'H3C2',\n",
    "    'HIST1H3E': 'H3C3',\n",
    "    'HIST1H3J': 'H3C12',\n",
    "    'HIST1H2BK': 'H2BC8',\n",
    "    'HIST1H3F': 'H3C4',\n",
    "    'HIST1H1B': 'H1-5',\n",
    "    'HIST1H1D': 'H1-3',\n",
    "    'HIST1H2AC': 'H2AC1',\n",
    "    'HIST1H2BJ': 'H2BC13',\n",
    "    'HIST1H4E': 'H4C5',\n",
    "    'FAM46C':'TENT5C',\n",
    "    'WHSC1L1':'NSD3',\n",
    "    'SETD8':'KMT5A',\n",
    "    'LPHN3':'ADGRL3',\n",
    "    'BAI3':'ADGRB3',\n",
    "    'SEPT9':'NAPB',\n",
    "    'BRE':'BABAM2',\n",
    "    'RFWD2':'COP1',\n",
    "    'TCEB1':'ELOC',\n",
    "    'GBA':'GBA1',\n",
    "    'PVRL4':'NECTIN4',\n",
    "    'ICK':'GCKR',\n",
    "    'GNB2L1':'RACK1',\n",
    "    'MGEA5':'OGA',\n",
    "    'DIRC2':'SLC49A4',\n",
    "    'LARGE':'LARGE1',\n",
    "    'TMEM173B':'STING1',\n",
    "    'SEPT5':'SEPTIN5'\n",
    "    }\n",
    "df_h_validated['gene_name_h'] = [dict_of_aacr_symbol.get(x,x) for x in df_h_validated['gene_name_h']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### H2M modeling  (N = 402/489 for mouse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There were rows that could not be processed.\n",
      "There were rows that could not be processed.\n",
      "No error occurs.\n"
     ]
    }
   ],
   "source": [
    "df_h_validated = h2m.get_tx_batch(df_h_validated, 'h', ver=37)[0]\n",
    "df_h_validated = h2m.query_batch(df_h_validated)[0]\n",
    "df_h_validated = h2m.get_tx_batch(df_h_validated, 'm', ver=37)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Mutation</th>\n",
       "      <th>Mutant Peptide</th>\n",
       "      <th>HLA</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>peptide_index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>ref_genome_h</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>type_h</th>\n",
       "      <th>index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNV</td>\n",
       "      <td>SEC22C</td>\n",
       "      <td>H248Y</td>\n",
       "      <td>AEHSLQVAY</td>\n",
       "      <td>B44:03</td>\n",
       "      <td>H</td>\n",
       "      <td>248</td>\n",
       "      <td>Y</td>\n",
       "      <td>248</td>\n",
       "      <td>0</td>\n",
       "      <td>SEC22C</td>\n",
       "      <td>ENST00000264454.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Sec22c</td>\n",
       "      <td>ENSMUST00000078547.12</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TKT</td>\n",
       "      <td>R438W</td>\n",
       "      <td>AMFWSVPTV</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>R</td>\n",
       "      <td>438</td>\n",
       "      <td>W</td>\n",
       "      <td>438</td>\n",
       "      <td>1</td>\n",
       "      <td>TKT</td>\n",
       "      <td>ENST00000423516.1</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Tkt</td>\n",
       "      <td>ENSMUST00000022529.8</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNV</td>\n",
       "      <td>ACSS3</td>\n",
       "      <td>H194Y</td>\n",
       "      <td>ARIGAIYSL</td>\n",
       "      <td>C07:02</td>\n",
       "      <td>H</td>\n",
       "      <td>194</td>\n",
       "      <td>Y</td>\n",
       "      <td>194</td>\n",
       "      <td>2</td>\n",
       "      <td>ACSS3</td>\n",
       "      <td>ENST00000548058.1</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Acss3</td>\n",
       "      <td>ENSMUST00000165067.9</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNV</td>\n",
       "      <td>AP2S1</td>\n",
       "      <td>N86K</td>\n",
       "      <td>AYLEAIHKF</td>\n",
       "      <td>A24:02</td>\n",
       "      <td>N</td>\n",
       "      <td>86</td>\n",
       "      <td>K</td>\n",
       "      <td>86</td>\n",
       "      <td>3</td>\n",
       "      <td>AP2S1</td>\n",
       "      <td>ENST00000263270.6</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Ap2s1</td>\n",
       "      <td>ENSMUST00000086112.8</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SNV</td>\n",
       "      <td>MED15</td>\n",
       "      <td>P677S</td>\n",
       "      <td>DANSFLQSV</td>\n",
       "      <td>B51:01</td>\n",
       "      <td>P</td>\n",
       "      <td>677</td>\n",
       "      <td>S</td>\n",
       "      <td>677</td>\n",
       "      <td>5</td>\n",
       "      <td>MED15</td>\n",
       "      <td>ENST00000263205.7</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Med15</td>\n",
       "      <td>ENSMUST00000012259.9</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>882</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TOX2</td>\n",
       "      <td>S382Y</td>\n",
       "      <td>YPAPQPPVL</td>\n",
       "      <td>B08:01</td>\n",
       "      <td>S</td>\n",
       "      <td>382</td>\n",
       "      <td>Y</td>\n",
       "      <td>382</td>\n",
       "      <td>637</td>\n",
       "      <td>TOX2</td>\n",
       "      <td>ENST00000341197.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Tox2</td>\n",
       "      <td>ENSMUST00000099110.10</td>\n",
       "      <td>SNP</td>\n",
       "      <td>882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>883</th>\n",
       "      <td>SNV</td>\n",
       "      <td>NR1D1</td>\n",
       "      <td>G39D</td>\n",
       "      <td>YSDNSNDSF</td>\n",
       "      <td>A01:01</td>\n",
       "      <td>G</td>\n",
       "      <td>39</td>\n",
       "      <td>D</td>\n",
       "      <td>39</td>\n",
       "      <td>638</td>\n",
       "      <td>NR1D1</td>\n",
       "      <td>ENST00000246672.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Nr1d1</td>\n",
       "      <td>ENSMUST00000064941.7</td>\n",
       "      <td>SNP</td>\n",
       "      <td>883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>884</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TNFAIP1</td>\n",
       "      <td>R48L</td>\n",
       "      <td>YTTVRALTL</td>\n",
       "      <td>B08:01</td>\n",
       "      <td>R</td>\n",
       "      <td>48</td>\n",
       "      <td>L</td>\n",
       "      <td>48</td>\n",
       "      <td>639</td>\n",
       "      <td>TNFAIP1</td>\n",
       "      <td>ENST00000226225.2</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Tnfaip1</td>\n",
       "      <td>ENSMUST00000108277.3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>885</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TSR1</td>\n",
       "      <td>H561Y</td>\n",
       "      <td>YVSEVPVSV</td>\n",
       "      <td>C17:01</td>\n",
       "      <td>H</td>\n",
       "      <td>561</td>\n",
       "      <td>Y</td>\n",
       "      <td>561</td>\n",
       "      <td>640</td>\n",
       "      <td>TSR1</td>\n",
       "      <td>ENST00000301364.5</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Tsr1</td>\n",
       "      <td>ENSMUST00000045807.14</td>\n",
       "      <td>SNP</td>\n",
       "      <td>885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>886</th>\n",
       "      <td>SNV</td>\n",
       "      <td>HEPH</td>\n",
       "      <td>I622T</td>\n",
       "      <td>YYSAVDPTK</td>\n",
       "      <td>A23:01</td>\n",
       "      <td>I</td>\n",
       "      <td>622</td>\n",
       "      <td>T</td>\n",
       "      <td>622</td>\n",
       "      <td>641</td>\n",
       "      <td>HEPH</td>\n",
       "      <td>ENST00000519389.1</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Heph</td>\n",
       "      <td>ENSMUST00000033553.14</td>\n",
       "      <td>SNP</td>\n",
       "      <td>886</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>586 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Mutation Type     Gene Mutation Mutant Peptide     HLA ref_seq_h start_h  \\\n",
       "0             SNV   SEC22C    H248Y      AEHSLQVAY  B44:03         H     248   \n",
       "1             SNV      TKT    R438W      AMFWSVPTV  A02:01         R     438   \n",
       "2             SNV    ACSS3    H194Y      ARIGAIYSL  C07:02         H     194   \n",
       "3             SNV    AP2S1     N86K      AYLEAIHKF  A24:02         N      86   \n",
       "4             SNV    MED15    P677S      DANSFLQSV  B51:01         P     677   \n",
       "..            ...      ...      ...            ...     ...       ...     ...   \n",
       "882           SNV     TOX2    S382Y      YPAPQPPVL  B08:01         S     382   \n",
       "883           SNV    NR1D1     G39D      YSDNSNDSF  A01:01         G      39   \n",
       "884           SNV  TNFAIP1     R48L      YTTVRALTL  B08:01         R      48   \n",
       "885           SNV     TSR1    H561Y      YVSEVPVSV  C17:01         H     561   \n",
       "886           SNV     HEPH    I622T      YYSAVDPTK  A23:01         I     622   \n",
       "\n",
       "    alt_seq_h end_h  peptide_index gene_name_h            tx_id_h  \\\n",
       "0           Y   248              0      SEC22C  ENST00000264454.3   \n",
       "1           W   438              1         TKT  ENST00000423516.1   \n",
       "2           Y   194              2       ACSS3  ENST00000548058.1   \n",
       "3           K    86              3       AP2S1  ENST00000263270.6   \n",
       "4           S   677              5       MED15  ENST00000263205.7   \n",
       "..        ...   ...            ...         ...                ...   \n",
       "882         Y   382            637        TOX2  ENST00000341197.4   \n",
       "883         D    39            638       NR1D1  ENST00000246672.3   \n",
       "884         L    48            639     TNFAIP1  ENST00000226225.2   \n",
       "885         Y   561            640        TSR1  ENST00000301364.5   \n",
       "886         T   622            641        HEPH  ENST00000519389.1   \n",
       "\n",
       "    ref_genome_h gene_name_m                tx_id_m type_h  index  \n",
       "0         GRCh37      Sec22c  ENSMUST00000078547.12    SNP      0  \n",
       "1         GRCh37         Tkt   ENSMUST00000022529.8    SNP      1  \n",
       "2         GRCh37       Acss3   ENSMUST00000165067.9    SNP      2  \n",
       "3         GRCh37       Ap2s1   ENSMUST00000086112.8    SNP      3  \n",
       "4         GRCh37       Med15   ENSMUST00000012259.9    SNP      4  \n",
       "..           ...         ...                    ...    ...    ...  \n",
       "882       GRCh37        Tox2  ENSMUST00000099110.10    SNP    882  \n",
       "883       GRCh37       Nr1d1   ENSMUST00000064941.7    SNP    883  \n",
       "884       GRCh37     Tnfaip1   ENSMUST00000108277.3    SNP    884  \n",
       "885       GRCh37        Tsr1  ENSMUST00000045807.14    SNP    885  \n",
       "886       GRCh37        Heph  ENSMUST00000033553.14    SNP    886  \n",
       "\n",
       "[586 rows x 17 columns]"
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_h_validated.drop_duplicates(subset='peptide_index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_h_ref, path_m_ref = '/Users/kexindong/Documents/GitHub/Database/RefGenome/ncbi-2023-09-12/GCF_000001405.25_GRCh37.p13_genomic.fna.gz', '/Users/kexindong/Documents/GitHub/Database/RefGenome/mouse-2023-09-13/GCF_000001635.27_GRCm39_genomic.fna.gz'\n",
    "# remember to replace the paths with yours\n",
    "records_h, index_list_h = h2m.genome_loader(path_h_ref)\n",
    "records_m, index_list_m = h2m.genome_loader(path_m_ref)\n",
    "path_h_anno, path_m_anno = '/Users/kexindong/Documents/GitHub/Database/Genecode/gencode_v19_GRCh37.db', '/Users/kexindong/Documents/GitHub/Database/Genecode/gencode_vm33_GRCm39.db'\n",
    "db_h, db_m = h2m.anno_loader(path_h_anno), h2m.anno_loader(path_m_anno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There were rows that could not be processed.\n"
     ]
    }
   ],
   "source": [
    "df_h_validated['type_h'] = 'SNP'\n",
    "df_h_validated['index'] = range(len(df_h_validated))\n",
    "df_result_h2m, df_fail = h2m.model_batch(df_h_validated, records_h, index_list_h, records_m, index_list_m, db_h, db_m, 37, coor='aa', show_sequence=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_result_h2m.to_csv('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/h2m_neoantigen_result_v3.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    402\n",
       "4     57\n",
       "Name: class, dtype: int64"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_result_h2m.drop_duplicates(subset=['HGVSp_h','gene_name_h'])['class'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    489\n",
       "4    164\n",
       "Name: class, dtype: int64"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_result_h2m.drop_duplicates(subset=['HGVSp_h','gene_name_h','tx_id_m'])['class'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutation Type</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Mutation</th>\n",
       "      <th>Mutant Peptide</th>\n",
       "      <th>HLA</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>start_h</th>\n",
       "      <th>alt_seq_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>peptide_index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>ref_genome_h</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>type_h</th>\n",
       "      <th>index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SNV</td>\n",
       "      <td>SEC22C</td>\n",
       "      <td>H248Y</td>\n",
       "      <td>AEHSLQVAY</td>\n",
       "      <td>B44:03</td>\n",
       "      <td>H</td>\n",
       "      <td>248</td>\n",
       "      <td>Y</td>\n",
       "      <td>248</td>\n",
       "      <td>0</td>\n",
       "      <td>SEC22C</td>\n",
       "      <td>ENST00000264454.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Sec22c</td>\n",
       "      <td>ENSMUST00000078547.12</td>\n",
       "      <td>SNP</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TKT</td>\n",
       "      <td>R438W</td>\n",
       "      <td>AMFWSVPTV</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>R</td>\n",
       "      <td>438</td>\n",
       "      <td>W</td>\n",
       "      <td>438</td>\n",
       "      <td>1</td>\n",
       "      <td>TKT</td>\n",
       "      <td>ENST00000423516.1</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Tkt</td>\n",
       "      <td>ENSMUST00000022529.8</td>\n",
       "      <td>SNP</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SNV</td>\n",
       "      <td>ACSS3</td>\n",
       "      <td>H194Y</td>\n",
       "      <td>ARIGAIYSL</td>\n",
       "      <td>C07:02</td>\n",
       "      <td>H</td>\n",
       "      <td>194</td>\n",
       "      <td>Y</td>\n",
       "      <td>194</td>\n",
       "      <td>2</td>\n",
       "      <td>ACSS3</td>\n",
       "      <td>ENST00000548058.1</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Acss3</td>\n",
       "      <td>ENSMUST00000165067.9</td>\n",
       "      <td>SNP</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SNV</td>\n",
       "      <td>AP2S1</td>\n",
       "      <td>N86K</td>\n",
       "      <td>AYLEAIHKF</td>\n",
       "      <td>A24:02</td>\n",
       "      <td>N</td>\n",
       "      <td>86</td>\n",
       "      <td>K</td>\n",
       "      <td>86</td>\n",
       "      <td>3</td>\n",
       "      <td>AP2S1</td>\n",
       "      <td>ENST00000263270.6</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Ap2s1</td>\n",
       "      <td>ENSMUST00000086112.8</td>\n",
       "      <td>SNP</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SNV</td>\n",
       "      <td>MED15</td>\n",
       "      <td>P677S</td>\n",
       "      <td>DANSFLQSV</td>\n",
       "      <td>B51:01</td>\n",
       "      <td>P</td>\n",
       "      <td>677</td>\n",
       "      <td>S</td>\n",
       "      <td>677</td>\n",
       "      <td>5</td>\n",
       "      <td>MED15</td>\n",
       "      <td>ENST00000263205.7</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Med15</td>\n",
       "      <td>ENSMUST00000012259.9</td>\n",
       "      <td>SNP</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>882</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TOX2</td>\n",
       "      <td>S382Y</td>\n",
       "      <td>YPAPQPPVL</td>\n",
       "      <td>B08:01</td>\n",
       "      <td>S</td>\n",
       "      <td>382</td>\n",
       "      <td>Y</td>\n",
       "      <td>382</td>\n",
       "      <td>637</td>\n",
       "      <td>TOX2</td>\n",
       "      <td>ENST00000341197.4</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Tox2</td>\n",
       "      <td>ENSMUST00000099110.10</td>\n",
       "      <td>SNP</td>\n",
       "      <td>882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>883</th>\n",
       "      <td>SNV</td>\n",
       "      <td>NR1D1</td>\n",
       "      <td>G39D</td>\n",
       "      <td>YSDNSNDSF</td>\n",
       "      <td>A01:01</td>\n",
       "      <td>G</td>\n",
       "      <td>39</td>\n",
       "      <td>D</td>\n",
       "      <td>39</td>\n",
       "      <td>638</td>\n",
       "      <td>NR1D1</td>\n",
       "      <td>ENST00000246672.3</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Nr1d1</td>\n",
       "      <td>ENSMUST00000064941.7</td>\n",
       "      <td>SNP</td>\n",
       "      <td>883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>884</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TNFAIP1</td>\n",
       "      <td>R48L</td>\n",
       "      <td>YTTVRALTL</td>\n",
       "      <td>B08:01</td>\n",
       "      <td>R</td>\n",
       "      <td>48</td>\n",
       "      <td>L</td>\n",
       "      <td>48</td>\n",
       "      <td>639</td>\n",
       "      <td>TNFAIP1</td>\n",
       "      <td>ENST00000226225.2</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Tnfaip1</td>\n",
       "      <td>ENSMUST00000108277.3</td>\n",
       "      <td>SNP</td>\n",
       "      <td>884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>885</th>\n",
       "      <td>SNV</td>\n",
       "      <td>TSR1</td>\n",
       "      <td>H561Y</td>\n",
       "      <td>YVSEVPVSV</td>\n",
       "      <td>C17:01</td>\n",
       "      <td>H</td>\n",
       "      <td>561</td>\n",
       "      <td>Y</td>\n",
       "      <td>561</td>\n",
       "      <td>640</td>\n",
       "      <td>TSR1</td>\n",
       "      <td>ENST00000301364.5</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Tsr1</td>\n",
       "      <td>ENSMUST00000045807.14</td>\n",
       "      <td>SNP</td>\n",
       "      <td>885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>886</th>\n",
       "      <td>SNV</td>\n",
       "      <td>HEPH</td>\n",
       "      <td>I622T</td>\n",
       "      <td>YYSAVDPTK</td>\n",
       "      <td>A23:01</td>\n",
       "      <td>I</td>\n",
       "      <td>622</td>\n",
       "      <td>T</td>\n",
       "      <td>622</td>\n",
       "      <td>641</td>\n",
       "      <td>HEPH</td>\n",
       "      <td>ENST00000519389.1</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>Heph</td>\n",
       "      <td>ENSMUST00000033553.14</td>\n",
       "      <td>SNP</td>\n",
       "      <td>886</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>887 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Mutation Type     Gene Mutation Mutant Peptide     HLA ref_seq_h start_h  \\\n",
       "0             SNV   SEC22C    H248Y      AEHSLQVAY  B44:03         H     248   \n",
       "1             SNV      TKT    R438W      AMFWSVPTV  A02:01         R     438   \n",
       "2             SNV    ACSS3    H194Y      ARIGAIYSL  C07:02         H     194   \n",
       "3             SNV    AP2S1     N86K      AYLEAIHKF  A24:02         N      86   \n",
       "4             SNV    MED15    P677S      DANSFLQSV  B51:01         P     677   \n",
       "..            ...      ...      ...            ...     ...       ...     ...   \n",
       "882           SNV     TOX2    S382Y      YPAPQPPVL  B08:01         S     382   \n",
       "883           SNV    NR1D1     G39D      YSDNSNDSF  A01:01         G      39   \n",
       "884           SNV  TNFAIP1     R48L      YTTVRALTL  B08:01         R      48   \n",
       "885           SNV     TSR1    H561Y      YVSEVPVSV  C17:01         H     561   \n",
       "886           SNV     HEPH    I622T      YYSAVDPTK  A23:01         I     622   \n",
       "\n",
       "    alt_seq_h end_h  peptide_index gene_name_h            tx_id_h  \\\n",
       "0           Y   248              0      SEC22C  ENST00000264454.3   \n",
       "1           W   438              1         TKT  ENST00000423516.1   \n",
       "2           Y   194              2       ACSS3  ENST00000548058.1   \n",
       "3           K    86              3       AP2S1  ENST00000263270.6   \n",
       "4           S   677              5       MED15  ENST00000263205.7   \n",
       "..        ...   ...            ...         ...                ...   \n",
       "882         Y   382            637        TOX2  ENST00000341197.4   \n",
       "883         D    39            638       NR1D1  ENST00000246672.3   \n",
       "884         L    48            639     TNFAIP1  ENST00000226225.2   \n",
       "885         Y   561            640        TSR1  ENST00000301364.5   \n",
       "886         T   622            641        HEPH  ENST00000519389.1   \n",
       "\n",
       "    ref_genome_h gene_name_m                tx_id_m type_h  index  \n",
       "0         GRCh37      Sec22c  ENSMUST00000078547.12    SNP      0  \n",
       "1         GRCh37         Tkt   ENSMUST00000022529.8    SNP      1  \n",
       "2         GRCh37       Acss3   ENSMUST00000165067.9    SNP      2  \n",
       "3         GRCh37       Ap2s1   ENSMUST00000086112.8    SNP      3  \n",
       "4         GRCh37       Med15   ENSMUST00000012259.9    SNP      4  \n",
       "..           ...         ...                    ...    ...    ...  \n",
       "882       GRCh37        Tox2  ENSMUST00000099110.10    SNP    882  \n",
       "883       GRCh37       Nr1d1   ENSMUST00000064941.7    SNP    883  \n",
       "884       GRCh37     Tnfaip1   ENSMUST00000108277.3    SNP    884  \n",
       "885       GRCh37        Tsr1  ENSMUST00000045807.14    SNP    885  \n",
       "886       GRCh37        Heph  ENSMUST00000033553.14    SNP    886  \n",
       "\n",
       "[887 rows x 17 columns]"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_h_validated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_process = df_result_h2m.drop_duplicates(subset=['HGVSp_h','gene_name_h','tx_id_m']).reset_index(drop=True)\n",
    "df_process = pd.merge(df_h_validated[['Mutant Peptide',  'HLA',\n",
    "        'index']], df_process, on = 'index', how = 'left')\n",
    "df_final_success = df_process[df_process['status']==True].reset_index(drop=True)\n",
    "df_final_success['flank_size_left'] = df_final_success['flank_size_left'].fillna('0aa')\n",
    "df_final_success['flank_size'] = [int(x[:-2])+int(y[:-2]) for x,y in zip(df_final_success['flank_size_left'], df_final_success['flank_size_right'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Mutant Peptide</th>\n",
       "      <th>HLA</th>\n",
       "      <th>index</th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>gene_id_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>chr_h</th>\n",
       "      <th>exon_num_h</th>\n",
       "      <th>strand_h</th>\n",
       "      <th>match</th>\n",
       "      <th>...</th>\n",
       "      <th>mouse_tx_idx_ori</th>\n",
       "      <th>mouse_p_idx_ori</th>\n",
       "      <th>mouse_new_p_idx_ori</th>\n",
       "      <th>new_seq_m</th>\n",
       "      <th>mouse_tx_idx</th>\n",
       "      <th>mouse_p_idx</th>\n",
       "      <th>mouse_new_p_idx</th>\n",
       "      <th>dist_h</th>\n",
       "      <th>dist_m</th>\n",
       "      <th>flank_size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AEHSLQVAY</td>\n",
       "      <td>B44:03</td>\n",
       "      <td>0</td>\n",
       "      <td>SEC22C</td>\n",
       "      <td>ENSG00000093183.9</td>\n",
       "      <td>ENST00000264454.3</td>\n",
       "      <td>chr3</td>\n",
       "      <td>6.0</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>[741, 742, 743]</td>\n",
       "      <td>[247]</td>\n",
       "      <td>[247]</td>\n",
       "      <td>ATGTCCATGATCCTTTTTGCCAGCATCGTACGGGTGAGGGATGGAC...</td>\n",
       "      <td>[741, 742, 743]</td>\n",
       "      <td>[247, 247, 247]</td>\n",
       "      <td>[247]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AMFWSVPTV</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>1</td>\n",
       "      <td>TKT</td>\n",
       "      <td>ENSG00000163931.11</td>\n",
       "      <td>ENST00000423516.1</td>\n",
       "      <td>chr3</td>\n",
       "      <td>15.0</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>[1287, 1288, 1289]</td>\n",
       "      <td>[429]</td>\n",
       "      <td>[429]</td>\n",
       "      <td>ATGGAAGGTTACCATAAGCCAGATCAGCAGAAGCTCCAGGCCCTGA...</td>\n",
       "      <td>[1287, 1288, 1289]</td>\n",
       "      <td>[429, 429, 429]</td>\n",
       "      <td>[429]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ARIGAIYSL</td>\n",
       "      <td>C07:02</td>\n",
       "      <td>2</td>\n",
       "      <td>ACSS3</td>\n",
       "      <td>ENSG00000111058.3</td>\n",
       "      <td>ENST00000548058.1</td>\n",
       "      <td>chr12</td>\n",
       "      <td>16.0</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>[564, 565, 566]</td>\n",
       "      <td>[188]</td>\n",
       "      <td>[188]</td>\n",
       "      <td>ATGAAGCCATCCTGGTTGCAATGTCGCAAAGTAACAGGCGCCGGGA...</td>\n",
       "      <td>[564, 565]</td>\n",
       "      <td>[188, 188]</td>\n",
       "      <td>[188]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AYLEAIHKF</td>\n",
       "      <td>A24:02</td>\n",
       "      <td>3</td>\n",
       "      <td>AP2S1</td>\n",
       "      <td>ENSG00000042753.7</td>\n",
       "      <td>ENST00000263270.6</td>\n",
       "      <td>chr19</td>\n",
       "      <td>5.0</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>[255, 256, 257]</td>\n",
       "      <td>[85]</td>\n",
       "      <td>[85]</td>\n",
       "      <td>ATGATCCGATTCATCCTTATCCAGAACCGGGCAGGCAAGACGCGCC...</td>\n",
       "      <td>[257]</td>\n",
       "      <td>[85]</td>\n",
       "      <td>[85]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DANSFLQSV</td>\n",
       "      <td>B51:01</td>\n",
       "      <td>4</td>\n",
       "      <td>MED15</td>\n",
       "      <td>ENSG00000099917.13</td>\n",
       "      <td>ENST00000263205.7</td>\n",
       "      <td>chr22</td>\n",
       "      <td>18.0</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>[2031, 2032, 2033]</td>\n",
       "      <td>[677]</td>\n",
       "      <td>[677]</td>\n",
       "      <td>ATGGACGTTTCGGGGCAGGAGACCGACTGGCGTAGCGCCGCCTTTC...</td>\n",
       "      <td>[2031, 2032, 2033]</td>\n",
       "      <td>[677, 677, 677]</td>\n",
       "      <td>[677]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>484</th>\n",
       "      <td>YPAPQPPVL</td>\n",
       "      <td>B08:01</td>\n",
       "      <td>882</td>\n",
       "      <td>TOX2</td>\n",
       "      <td>ENSG00000124191.13</td>\n",
       "      <td>ENST00000341197.4</td>\n",
       "      <td>chr20</td>\n",
       "      <td>9.0</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>[1269, 1270, 1271]</td>\n",
       "      <td>[423]</td>\n",
       "      <td>[423]</td>\n",
       "      <td>ATGGGGGGCGGGGGTGCCGGGCCTCCAGCCAATAGAGGACCAAGAG...</td>\n",
       "      <td>[1269, 1270, 1271]</td>\n",
       "      <td>[423, 423, 423]</td>\n",
       "      <td>[423]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>485</th>\n",
       "      <td>YSDNSNDSF</td>\n",
       "      <td>A01:01</td>\n",
       "      <td>883</td>\n",
       "      <td>NR1D1</td>\n",
       "      <td>ENSG00000126368.5</td>\n",
       "      <td>ENST00000246672.3</td>\n",
       "      <td>chr17</td>\n",
       "      <td>8.0</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>[114, 115, 116]</td>\n",
       "      <td>[38]</td>\n",
       "      <td>[38]</td>\n",
       "      <td>ATGACGACCCTGGACTCCAATAACAACACAGGTGGTGTTATCACCT...</td>\n",
       "      <td>[115]</td>\n",
       "      <td>[38]</td>\n",
       "      <td>[38]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>486</th>\n",
       "      <td>YTTVRALTL</td>\n",
       "      <td>B08:01</td>\n",
       "      <td>884</td>\n",
       "      <td>TNFAIP1</td>\n",
       "      <td>ENSG00000109079.5</td>\n",
       "      <td>ENST00000226225.2</td>\n",
       "      <td>chr17</td>\n",
       "      <td>6.0</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>[141, 142, 143]</td>\n",
       "      <td>[47]</td>\n",
       "      <td>[47]</td>\n",
       "      <td>ATGTCAGGGGACACCTGTCTGTGTCCAGCCTCGGGGGCCAAGCCCA...</td>\n",
       "      <td>[142]</td>\n",
       "      <td>[47]</td>\n",
       "      <td>[47]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>487</th>\n",
       "      <td>YVSEVPVSV</td>\n",
       "      <td>C17:01</td>\n",
       "      <td>885</td>\n",
       "      <td>TSR1</td>\n",
       "      <td>ENSG00000167721.6</td>\n",
       "      <td>ENST00000301364.5</td>\n",
       "      <td>chr17</td>\n",
       "      <td>15.0</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>[1677, 1678, 1679]</td>\n",
       "      <td>[559]</td>\n",
       "      <td>[559]</td>\n",
       "      <td>ATGGCGGCTCACCGCTCCGGGCCGCTGAAGCAGCAGAATAAAGCTC...</td>\n",
       "      <td>[1677, 1678, 1679]</td>\n",
       "      <td>[559, 559, 559]</td>\n",
       "      <td>[559]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>488</th>\n",
       "      <td>YYSAVDPTK</td>\n",
       "      <td>A23:01</td>\n",
       "      <td>886</td>\n",
       "      <td>HEPH</td>\n",
       "      <td>ENSG00000089472.12</td>\n",
       "      <td>ENST00000519389.1</td>\n",
       "      <td>chrX</td>\n",
       "      <td>21.0</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>[1698, 1699, 1700]</td>\n",
       "      <td>[566]</td>\n",
       "      <td>[566]</td>\n",
       "      <td>ATGAAGGCAGGCCATCTTCTCTGGGCTTTACTGTTGATGCACTCCT...</td>\n",
       "      <td>[1698, 1699, 1700]</td>\n",
       "      <td>[566, 566, 566]</td>\n",
       "      <td>[566]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>489 rows × 62 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Mutant Peptide     HLA  index gene_name_h           gene_id_h  \\\n",
       "0        AEHSLQVAY  B44:03      0      SEC22C   ENSG00000093183.9   \n",
       "1        AMFWSVPTV  A02:01      1         TKT  ENSG00000163931.11   \n",
       "2        ARIGAIYSL  C07:02      2       ACSS3   ENSG00000111058.3   \n",
       "3        AYLEAIHKF  A24:02      3       AP2S1   ENSG00000042753.7   \n",
       "4        DANSFLQSV  B51:01      4       MED15  ENSG00000099917.13   \n",
       "..             ...     ...    ...         ...                 ...   \n",
       "484      YPAPQPPVL  B08:01    882        TOX2  ENSG00000124191.13   \n",
       "485      YSDNSNDSF  A01:01    883       NR1D1   ENSG00000126368.5   \n",
       "486      YTTVRALTL  B08:01    884     TNFAIP1   ENSG00000109079.5   \n",
       "487      YVSEVPVSV  C17:01    885        TSR1   ENSG00000167721.6   \n",
       "488      YYSAVDPTK  A23:01    886        HEPH  ENSG00000089472.12   \n",
       "\n",
       "               tx_id_h  chr_h  exon_num_h strand_h  match  ...  \\\n",
       "0    ENST00000264454.3   chr3         6.0        -  False  ...   \n",
       "1    ENST00000423516.1   chr3        15.0        -  False  ...   \n",
       "2    ENST00000548058.1  chr12        16.0        +  False  ...   \n",
       "3    ENST00000263270.6  chr19         5.0        -   True  ...   \n",
       "4    ENST00000263205.7  chr22        18.0        +  False  ...   \n",
       "..                 ...    ...         ...      ...    ...  ...   \n",
       "484  ENST00000341197.4  chr20         9.0        +  False  ...   \n",
       "485  ENST00000246672.3  chr17         8.0        -   True  ...   \n",
       "486  ENST00000226225.2  chr17         6.0        +   True  ...   \n",
       "487  ENST00000301364.5  chr17        15.0        -   True  ...   \n",
       "488  ENST00000519389.1   chrX        21.0        +  False  ...   \n",
       "\n",
       "       mouse_tx_idx_ori  mouse_p_idx_ori mouse_new_p_idx_ori  \\\n",
       "0       [741, 742, 743]            [247]               [247]   \n",
       "1    [1287, 1288, 1289]            [429]               [429]   \n",
       "2       [564, 565, 566]            [188]               [188]   \n",
       "3       [255, 256, 257]             [85]                [85]   \n",
       "4    [2031, 2032, 2033]            [677]               [677]   \n",
       "..                  ...              ...                 ...   \n",
       "484  [1269, 1270, 1271]            [423]               [423]   \n",
       "485     [114, 115, 116]             [38]                [38]   \n",
       "486     [141, 142, 143]             [47]                [47]   \n",
       "487  [1677, 1678, 1679]            [559]               [559]   \n",
       "488  [1698, 1699, 1700]            [566]               [566]   \n",
       "\n",
       "                                             new_seq_m        mouse_tx_idx  \\\n",
       "0    ATGTCCATGATCCTTTTTGCCAGCATCGTACGGGTGAGGGATGGAC...     [741, 742, 743]   \n",
       "1    ATGGAAGGTTACCATAAGCCAGATCAGCAGAAGCTCCAGGCCCTGA...  [1287, 1288, 1289]   \n",
       "2    ATGAAGCCATCCTGGTTGCAATGTCGCAAAGTAACAGGCGCCGGGA...          [564, 565]   \n",
       "3    ATGATCCGATTCATCCTTATCCAGAACCGGGCAGGCAAGACGCGCC...               [257]   \n",
       "4    ATGGACGTTTCGGGGCAGGAGACCGACTGGCGTAGCGCCGCCTTTC...  [2031, 2032, 2033]   \n",
       "..                                                 ...                 ...   \n",
       "484  ATGGGGGGCGGGGGTGCCGGGCCTCCAGCCAATAGAGGACCAAGAG...  [1269, 1270, 1271]   \n",
       "485  ATGACGACCCTGGACTCCAATAACAACACAGGTGGTGTTATCACCT...               [115]   \n",
       "486  ATGTCAGGGGACACCTGTCTGTGTCCAGCCTCGGGGGCCAAGCCCA...               [142]   \n",
       "487  ATGGCGGCTCACCGCTCCGGGCCGCTGAAGCAGCAGAATAAAGCTC...  [1677, 1678, 1679]   \n",
       "488  ATGAAGGCAGGCCATCTTCTCTGGGCTTTACTGTTGATGCACTCCT...  [1698, 1699, 1700]   \n",
       "\n",
       "         mouse_p_idx mouse_new_p_idx dist_h dist_m flank_size  \n",
       "0    [247, 247, 247]           [247]    NaN    NaN         12  \n",
       "1    [429, 429, 429]           [429]    NaN    NaN         60  \n",
       "2         [188, 188]           [188]    NaN    NaN         24  \n",
       "3               [85]            [85]    NaN    NaN        142  \n",
       "4    [677, 677, 677]           [677]    NaN    NaN         57  \n",
       "..               ...             ...    ...    ...        ...  \n",
       "484  [423, 423, 423]           [423]    NaN    NaN          5  \n",
       "485             [38]            [38]    NaN    NaN         32  \n",
       "486             [47]            [47]    NaN    NaN         78  \n",
       "487  [559, 559, 559]           [559]    NaN    NaN         13  \n",
       "488  [566, 566, 566]           [566]    NaN    NaN         22  \n",
       "\n",
       "[489 rows x 62 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_final_success = df_final_success.reset_index(drop=True)\n",
    "df_final_success"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 9-mer splicing (N = 300/383 for mouse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_9mers_containing_position(sequence, position):\n",
    "    \"\"\"\n",
    "    Generate all 9-mer slices from a given DNA sequence that include a specific position.\n",
    "\n",
    "    Args:\n",
    "    sequence (str): The DNA sequence.\n",
    "    position (int): The 1-based position within the sequence.\n",
    "\n",
    "    Returns:\n",
    "    list: A list of all 9-mer slices including the specified position.\n",
    "    \"\"\"\n",
    "    # Adjust position to 0-based for easier slicing\n",
    "    sequence = str(h2m.Translate(sequence))\n",
    "    position = position[0]\n",
    "\n",
    "    k_mers = [\"\"] * 9 \n",
    "    start_index = position - 8\n",
    "\n",
    "    for i in range(9):\n",
    "        if 0 <= start_index + i <= len(sequence) - 9:  \n",
    "            k_mer = sequence[start_index + i: start_index + i + 9]\n",
    "            k_mers[i] = k_mer\n",
    "    \n",
    "    # Initialize list to collect 9-mers\n",
    "    \n",
    "    return k_mers\n",
    "\n",
    "def expand_row(row):\n",
    "    return pd.DataFrame({\n",
    "        'human_9mers': row['human_mouse_9mers']['human_9mers'],\n",
    "        'mouse_9mers': row['human_mouse_9mers']['mouse_9mers'],\n",
    "        'index':row['index'],\n",
    "        'HLA':row['HLA'],\n",
    "        'Mutant Peptide':row['Mutant Peptide']\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_final_success['human_mouse_9mers'] = df_final_success.apply(\n",
    "    lambda row: {\n",
    "        'human_9mers': find_9mers_containing_position(row['new_seq_h'], row['human_p_idx']),\n",
    "        'mouse_9mers': find_9mers_containing_position(row['new_seq_m'], row['mouse_p_idx'])\n",
    "    }, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "expanded_rows = pd.concat([expand_row(row) for _, row in df_final_success.iterrows()]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "o = expanded_rows[expanded_rows['human_9mers']==expanded_rows['Mutant Peptide']].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1    489\n",
       "4    164\n",
       "Name: class, dtype: int64"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_result_h2m.drop_duplicates(subset=['HGVSp_h','gene_name_h','tx_id_m'])['class'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human_9mers</th>\n",
       "      <th>mouse_9mers</th>\n",
       "      <th>index</th>\n",
       "      <th>HLA</th>\n",
       "      <th>Mutant Peptide</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AYLEAIHKF</td>\n",
       "      <td>AYLEAIHKF</td>\n",
       "      <td>3</td>\n",
       "      <td>A24:02</td>\n",
       "      <td>AYLEAIHKF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FLYNLLTRV</td>\n",
       "      <td>FMYNLLTRV</td>\n",
       "      <td>8</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>FLYNLLTRV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GLFGDIYLA</td>\n",
       "      <td>GLFGDIYLA</td>\n",
       "      <td>9</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>GLFGDIYLA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRIAFFLKY</td>\n",
       "      <td>GKIAFFLKF</td>\n",
       "      <td>10</td>\n",
       "      <td>B27:05</td>\n",
       "      <td>GRIAFFLKY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HMTEVVRHC</td>\n",
       "      <td>HMTEVVRHC</td>\n",
       "      <td>12</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>HMTEVVRHC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>YLSELLQTV</td>\n",
       "      <td>YLSELLQTV</td>\n",
       "      <td>880</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>YLSELLQTV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>YNTDDIEFY</td>\n",
       "      <td>YNTDDIEFY</td>\n",
       "      <td>881</td>\n",
       "      <td>A29:02</td>\n",
       "      <td>YNTDDIEFY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>YSDNSNDSF</td>\n",
       "      <td>YSDSSNDSF</td>\n",
       "      <td>883</td>\n",
       "      <td>A01:01</td>\n",
       "      <td>YSDNSNDSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>YTTVRALTL</td>\n",
       "      <td>YTTVRALTL</td>\n",
       "      <td>884</td>\n",
       "      <td>B08:01</td>\n",
       "      <td>YTTVRALTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>YVSEVPVSV</td>\n",
       "      <td>YVSDVPVSV</td>\n",
       "      <td>885</td>\n",
       "      <td>C17:01</td>\n",
       "      <td>YVSEVPVSV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>383 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    human_9mers mouse_9mers  index     HLA Mutant Peptide\n",
       "0     AYLEAIHKF   AYLEAIHKF      3  A24:02      AYLEAIHKF\n",
       "1     FLYNLLTRV   FMYNLLTRV      8  A02:01      FLYNLLTRV\n",
       "2     GLFGDIYLA   GLFGDIYLA      9  A02:01      GLFGDIYLA\n",
       "3     GRIAFFLKY   GKIAFFLKF     10  B27:05      GRIAFFLKY\n",
       "4     HMTEVVRHC   HMTEVVRHC     12  A02:01      HMTEVVRHC\n",
       "..          ...         ...    ...     ...            ...\n",
       "378   YLSELLQTV   YLSELLQTV    880  A02:01      YLSELLQTV\n",
       "379   YNTDDIEFY   YNTDDIEFY    881  A29:02      YNTDDIEFY\n",
       "380   YSDNSNDSF   YSDSSNDSF    883  A01:01      YSDNSNDSF\n",
       "381   YTTVRALTL   YTTVRALTL    884  B08:01      YTTVRALTL\n",
       "382   YVSEVPVSV   YVSDVPVSV    885  C17:01      YVSEVPVSV\n",
       "\n",
       "[383 rows x 5 columns]"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "o"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "300"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(o['human_9mers'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>HGVSp_h</th>\n",
       "      <th>status</th>\n",
       "      <th>flank_size</th>\n",
       "      <th>human_9mers</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "      <th>mouse_9mers</th>\n",
       "      <th>index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCB10</td>\n",
       "      <td>S405Y</td>\n",
       "      <td>True</td>\n",
       "      <td>39</td>\n",
       "      <td>GLYGNLIVL</td>\n",
       "      <td>Abcb10</td>\n",
       "      <td>S370Y</td>\n",
       "      <td>GLYGNLIVL</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCC6</td>\n",
       "      <td>Q715E</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>GEELDPPWL</td>\n",
       "      <td>Abcc6</td>\n",
       "      <td>Q713E</td>\n",
       "      <td>REELDLPWL</td>\n",
       "      <td>739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACE</td>\n",
       "      <td>S167F</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>FLDPDLTNI</td>\n",
       "      <td>Ace</td>\n",
       "      <td>S172F</td>\n",
       "      <td>FLDPELTNI</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACE2</td>\n",
       "      <td>D615Y</td>\n",
       "      <td>True</td>\n",
       "      <td>19</td>\n",
       "      <td>AYQSIKVRI</td>\n",
       "      <td>Ace2</td>\n",
       "      <td>D615Y</td>\n",
       "      <td>AYQSIKVRI</td>\n",
       "      <td>379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACTG1</td>\n",
       "      <td>G366S</td>\n",
       "      <td>True</td>\n",
       "      <td>375</td>\n",
       "      <td>ESSPSIVHR</td>\n",
       "      <td>Actg1</td>\n",
       "      <td>G366S</td>\n",
       "      <td>ESSPSIVHR</td>\n",
       "      <td>711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>ZNF611</td>\n",
       "      <td>D404G</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>KVCGTAFTW</td>\n",
       "      <td>Zfp966</td>\n",
       "      <td>D268G</td>\n",
       "      <td>KQCGKAFVR</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>ZNF611</td>\n",
       "      <td>D404G</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>KVCGTAFTW</td>\n",
       "      <td>Zfp975</td>\n",
       "      <td>D305G</td>\n",
       "      <td>NQCGKGFAQ</td>\n",
       "      <td>228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>ZNF611</td>\n",
       "      <td>D404G</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>KVCGTAFTW</td>\n",
       "      <td>Zfp976</td>\n",
       "      <td>D333G</td>\n",
       "      <td>NQCGKGFAQ</td>\n",
       "      <td>229</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>ZNF611</td>\n",
       "      <td>D404G</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>KVCGTAFTW</td>\n",
       "      <td>5430403G16Rik</td>\n",
       "      <td>D330G</td>\n",
       "      <td>YKCGQCGKA</td>\n",
       "      <td>195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>ZNF611</td>\n",
       "      <td>D404G</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>KVCGTAFTW</td>\n",
       "      <td>Zfp967</td>\n",
       "      <td>D268G</td>\n",
       "      <td>KQCGKAFVR</td>\n",
       "      <td>225</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>383 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name_h HGVSp_h status  flank_size human_9mers    gene_name_m HGVSp_m  \\\n",
       "0        ABCB10   S405Y   True          39   GLYGNLIVL         Abcb10   S370Y   \n",
       "1         ABCC6   Q715E   True           3   GEELDPPWL          Abcc6   Q713E   \n",
       "2           ACE   S167F   True          11   FLDPDLTNI            Ace   S172F   \n",
       "3          ACE2   D615Y   True          19   AYQSIKVRI           Ace2   D615Y   \n",
       "4         ACTG1   G366S   True         375   ESSPSIVHR          Actg1   G366S   \n",
       "..          ...     ...    ...         ...         ...            ...     ...   \n",
       "378      ZNF611   D404G   True           1   KVCGTAFTW         Zfp966   D268G   \n",
       "379      ZNF611   D404G   True           0   KVCGTAFTW         Zfp975   D305G   \n",
       "380      ZNF611   D404G   True           0   KVCGTAFTW         Zfp976   D333G   \n",
       "381      ZNF611   D404G   True           1   KVCGTAFTW  5430403G16Rik   D330G   \n",
       "382      ZNF611   D404G   True           1   KVCGTAFTW         Zfp967   D268G   \n",
       "\n",
       "    mouse_9mers  index  \n",
       "0     GLYGNLIVL    119  \n",
       "1     REELDLPWL    739  \n",
       "2     FLDPELTNI     88  \n",
       "3     AYQSIKVRI    379  \n",
       "4     ESSPSIVHR    711  \n",
       "..          ...    ...  \n",
       "378   KQCGKAFVR    224  \n",
       "379   NQCGKGFAQ    228  \n",
       "380   NQCGKGFAQ    229  \n",
       "381   YKCGQCGKA    195  \n",
       "382   KQCGKAFVR    225  \n",
       "\n",
       "[383 rows x 9 columns]"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_output = df_final_success.merge(o)[['gene_name_h','HGVSp_h','status','flank_size','human_9mers','gene_name_m','HGVSp_m','mouse_9mers','index']].sort_values(by='gene_name_h').reset_index(drop=True)\n",
    "df_output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate NetMHCpan inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, x in o.iterrows():\n",
    "    k = x['index']\n",
    "    l = x['human_9mers']\n",
    "    with open('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/input_h_3.txt', 'a') as file:\n",
    "        file.writelines(f'>{k}\\n{l}\\n')\n",
    "    l = x['mouse_9mers']\n",
    "    with open('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/input_m_3.txt', 'a') as file:\n",
    "        file.writelines(f'>{k}\\n{l}\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human_9mers</th>\n",
       "      <th>mouse_9mers</th>\n",
       "      <th>index</th>\n",
       "      <th>HLA</th>\n",
       "      <th>Mutant Peptide</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AYLEAIHKF</td>\n",
       "      <td>AYLEAIHKF</td>\n",
       "      <td>3</td>\n",
       "      <td>A24:02</td>\n",
       "      <td>AYLEAIHKF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FLYNLLTRV</td>\n",
       "      <td>FMYNLLTRV</td>\n",
       "      <td>8</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>FLYNLLTRV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GLFGDIYLA</td>\n",
       "      <td>GLFGDIYLA</td>\n",
       "      <td>9</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>GLFGDIYLA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GRIAFFLKY</td>\n",
       "      <td>GKIAFFLKF</td>\n",
       "      <td>10</td>\n",
       "      <td>B27:05</td>\n",
       "      <td>GRIAFFLKY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HMTEVVRHC</td>\n",
       "      <td>HMTEVVRHC</td>\n",
       "      <td>12</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>HMTEVVRHC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>YLSELLQTV</td>\n",
       "      <td>YLSELLQTV</td>\n",
       "      <td>880</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>YLSELLQTV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>YNTDDIEFY</td>\n",
       "      <td>YNTDDIEFY</td>\n",
       "      <td>881</td>\n",
       "      <td>A29:02</td>\n",
       "      <td>YNTDDIEFY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>YSDNSNDSF</td>\n",
       "      <td>YSDSSNDSF</td>\n",
       "      <td>883</td>\n",
       "      <td>A01:01</td>\n",
       "      <td>YSDNSNDSF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>YTTVRALTL</td>\n",
       "      <td>YTTVRALTL</td>\n",
       "      <td>884</td>\n",
       "      <td>B08:01</td>\n",
       "      <td>YTTVRALTL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>YVSEVPVSV</td>\n",
       "      <td>YVSDVPVSV</td>\n",
       "      <td>885</td>\n",
       "      <td>C17:01</td>\n",
       "      <td>YVSEVPVSV</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>383 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    human_9mers mouse_9mers  index     HLA Mutant Peptide\n",
       "0     AYLEAIHKF   AYLEAIHKF      3  A24:02      AYLEAIHKF\n",
       "1     FLYNLLTRV   FMYNLLTRV      8  A02:01      FLYNLLTRV\n",
       "2     GLFGDIYLA   GLFGDIYLA      9  A02:01      GLFGDIYLA\n",
       "3     GRIAFFLKY   GKIAFFLKF     10  B27:05      GRIAFFLKY\n",
       "4     HMTEVVRHC   HMTEVVRHC     12  A02:01      HMTEVVRHC\n",
       "..          ...         ...    ...     ...            ...\n",
       "378   YLSELLQTV   YLSELLQTV    880  A02:01      YLSELLQTV\n",
       "379   YNTDDIEFY   YNTDDIEFY    881  A29:02      YNTDDIEFY\n",
       "380   YSDNSNDSF   YSDSSNDSF    883  A01:01      YSDNSNDSF\n",
       "381   YTTVRALTL   YTTVRALTL    884  B08:01      YTTVRALTL\n",
       "382   YVSEVPVSV   YVSDVPVSV    885  C17:01      YVSEVPVSV\n",
       "\n",
       "[383 rows x 5 columns]"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "o"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Human results (top score is selected for each peptide among different MHC alleles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_h_i_list = []\n",
    "for x in ['A','B1','B2','C1','C2']:\n",
    "    result_h_i_list.append(pd.read_csv(f'/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/output_h_mhci_3_{x}.txt', header=0, sep='\\t', comment=\"#\"))\n",
    "result_h_i = pd.concat(result_h_i_list).reset_index(drop=True)\n",
    "del result_h_i_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_h_i['allele'] = [x[4:] for x in result_h_i['allele']]\n",
    "result_h_i['allele'] = [''.join(str(x).split('*')) for x in result_h_i['allele']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'A01:01, A02:01, A02:06, A03:01, A11:01, A23:01, A24:02, A25:01, A26:01, A29:02, A30:01, A30:02, A31:01, A32:01, A33:03, B07:02, B08:01, B13:01, B13:02, B14:02, B15:01, B15:02, B15:25, B18:01, B27:02, B27:05, B35:01, B35:03, B37:01, B38:01, B39:01, B40:01, B40:02, B44:02, B44:03, B46:01, B48:01, B49:01, B50:01, B51:01, B52:01, B53:01, B55:01, B56:01, B57:01, B58:01, B58:02, C01:02, C02:02, C02:09, C03:02, C03:03, C03:04, C04:01, C05:01, C06:02, C07:01, C07:02, C07:04, C08:01, C08:02, C12:02, C12:03, C14:02, C15:02, C16:01, C17:01, E01:01, E01:03, G01:01, G01:02, G01:03, G01:04, G01:06'"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "', '.join(result_h_i['allele'].sort_values().unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_h_i_unique = result_h_i.sort_values(by='score', ascending=False).reset_index(drop=True).drop_duplicates(subset='peptide')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "k = pd.merge(df_output,result_h_i_unique[['allele','peptide','score','percentile_rank']].rename(columns={'peptide':'human_9mers',\n",
    "                                                                                                           'allele':'HLA_allele',\n",
    "                                                                                                           'score':'score_h',\n",
    "                                                                                                           'percentile_rank':'percentile_rank_h'}),how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>HGVSp_h</th>\n",
       "      <th>status</th>\n",
       "      <th>flank_size</th>\n",
       "      <th>human_9mers</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "      <th>mouse_9mers</th>\n",
       "      <th>index</th>\n",
       "      <th>HLA_allele</th>\n",
       "      <th>score_h</th>\n",
       "      <th>percentile_rank_h</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCB10</td>\n",
       "      <td>S405Y</td>\n",
       "      <td>True</td>\n",
       "      <td>39</td>\n",
       "      <td>GLYGNLIVL</td>\n",
       "      <td>Abcb10</td>\n",
       "      <td>S370Y</td>\n",
       "      <td>GLYGNLIVL</td>\n",
       "      <td>119</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>0.876096</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCC6</td>\n",
       "      <td>Q715E</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>GEELDPPWL</td>\n",
       "      <td>Abcc6</td>\n",
       "      <td>Q713E</td>\n",
       "      <td>REELDLPWL</td>\n",
       "      <td>739</td>\n",
       "      <td>B40:01</td>\n",
       "      <td>0.750505</td>\n",
       "      <td>0.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACE</td>\n",
       "      <td>S167F</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>FLDPDLTNI</td>\n",
       "      <td>Ace</td>\n",
       "      <td>S172F</td>\n",
       "      <td>FLDPELTNI</td>\n",
       "      <td>88</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>0.987693</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACE2</td>\n",
       "      <td>D615Y</td>\n",
       "      <td>True</td>\n",
       "      <td>19</td>\n",
       "      <td>AYQSIKVRI</td>\n",
       "      <td>Ace2</td>\n",
       "      <td>D615Y</td>\n",
       "      <td>AYQSIKVRI</td>\n",
       "      <td>379</td>\n",
       "      <td>A24:02</td>\n",
       "      <td>0.837559</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACTG1</td>\n",
       "      <td>G366S</td>\n",
       "      <td>True</td>\n",
       "      <td>375</td>\n",
       "      <td>ESSPSIVHR</td>\n",
       "      <td>Actg1</td>\n",
       "      <td>G366S</td>\n",
       "      <td>ESSPSIVHR</td>\n",
       "      <td>711</td>\n",
       "      <td>A33:03</td>\n",
       "      <td>0.909769</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>ZNF611</td>\n",
       "      <td>D404G</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>KVCGTAFTW</td>\n",
       "      <td>Zfp966</td>\n",
       "      <td>D268G</td>\n",
       "      <td>KQCGKAFVR</td>\n",
       "      <td>224</td>\n",
       "      <td>B57:01</td>\n",
       "      <td>0.988189</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>ZNF611</td>\n",
       "      <td>D404G</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>KVCGTAFTW</td>\n",
       "      <td>Zfp975</td>\n",
       "      <td>D305G</td>\n",
       "      <td>NQCGKGFAQ</td>\n",
       "      <td>228</td>\n",
       "      <td>B57:01</td>\n",
       "      <td>0.988189</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>ZNF611</td>\n",
       "      <td>D404G</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>KVCGTAFTW</td>\n",
       "      <td>Zfp976</td>\n",
       "      <td>D333G</td>\n",
       "      <td>NQCGKGFAQ</td>\n",
       "      <td>229</td>\n",
       "      <td>B57:01</td>\n",
       "      <td>0.988189</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>ZNF611</td>\n",
       "      <td>D404G</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>KVCGTAFTW</td>\n",
       "      <td>5430403G16Rik</td>\n",
       "      <td>D330G</td>\n",
       "      <td>YKCGQCGKA</td>\n",
       "      <td>195</td>\n",
       "      <td>B57:01</td>\n",
       "      <td>0.988189</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>ZNF611</td>\n",
       "      <td>D404G</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>KVCGTAFTW</td>\n",
       "      <td>Zfp967</td>\n",
       "      <td>D268G</td>\n",
       "      <td>KQCGKAFVR</td>\n",
       "      <td>225</td>\n",
       "      <td>B57:01</td>\n",
       "      <td>0.988189</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>383 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name_h HGVSp_h status  flank_size human_9mers    gene_name_m HGVSp_m  \\\n",
       "0        ABCB10   S405Y   True          39   GLYGNLIVL         Abcb10   S370Y   \n",
       "1         ABCC6   Q715E   True           3   GEELDPPWL          Abcc6   Q713E   \n",
       "2           ACE   S167F   True          11   FLDPDLTNI            Ace   S172F   \n",
       "3          ACE2   D615Y   True          19   AYQSIKVRI           Ace2   D615Y   \n",
       "4         ACTG1   G366S   True         375   ESSPSIVHR          Actg1   G366S   \n",
       "..          ...     ...    ...         ...         ...            ...     ...   \n",
       "378      ZNF611   D404G   True           1   KVCGTAFTW         Zfp966   D268G   \n",
       "379      ZNF611   D404G   True           0   KVCGTAFTW         Zfp975   D305G   \n",
       "380      ZNF611   D404G   True           0   KVCGTAFTW         Zfp976   D333G   \n",
       "381      ZNF611   D404G   True           1   KVCGTAFTW  5430403G16Rik   D330G   \n",
       "382      ZNF611   D404G   True           1   KVCGTAFTW         Zfp967   D268G   \n",
       "\n",
       "    mouse_9mers  index HLA_allele   score_h  percentile_rank_h  \n",
       "0     GLYGNLIVL    119     A02:01  0.876096               0.04  \n",
       "1     REELDLPWL    739     B40:01  0.750505               0.13  \n",
       "2     FLDPELTNI     88     A02:01  0.987693               0.01  \n",
       "3     AYQSIKVRI    379     A24:02  0.837559               0.04  \n",
       "4     ESSPSIVHR    711     A33:03  0.909769               0.01  \n",
       "..          ...    ...        ...       ...                ...  \n",
       "378   KQCGKAFVR    224     B57:01  0.988189               0.01  \n",
       "379   NQCGKGFAQ    228     B57:01  0.988189               0.01  \n",
       "380   NQCGKGFAQ    229     B57:01  0.988189               0.01  \n",
       "381   YKCGQCGKA    195     B57:01  0.988189               0.01  \n",
       "382   KQCGKAFVR    225     B57:01  0.988189               0.01  \n",
       "\n",
       "[383 rows x 12 columns]"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_m_i = pd.read_csv(f'/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/output_m_mhci_3.txt', header=0, sep='\\t', comment=\"#\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_m_i['allele'] = [x[4:] for x in result_m_i['allele']]\n",
    "result_m_i['allele'] = [''.join(str(x).split('*')) for x in result_m_i['allele']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mouse results (top score is selected for each peptide among different MHC alleles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_m_i_unique = result_m_i.sort_values(by='score', ascending=False).reset_index(drop=True).drop_duplicates(subset='peptide')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [],
   "source": [
    "l_both_best = pd.merge(k,result_m_i_unique[['allele','peptide','score','percentile_rank']].rename(columns={'peptide':'mouse_9mers',\n",
    "                                                                                                           'allele':'H2_allele',\n",
    "                                                                                                           'percentile_rank':'percentile_rank_m',\n",
    "                                                                                                           'score':'score_m'}),how='left')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mouse Kb results (most common allele: Kb/Db)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_m_i_kb = result_m_i[result_m_i['allele'].isin(['Kb','Db'])].reset_index(drop=True).sort_values(by='score', ascending=False).reset_index(drop=True).drop_duplicates(subset='peptide')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [],
   "source": [
    "l_final = pd.merge(l_both_best,result_m_i_kb[['peptide','score','percentile_rank']].rename(columns={'peptide':'mouse_9mers',\n",
    "                                                                                                           'percentile_rank':'percentile_rank_kbdb',\n",
    "                                                                                                           'score':'score_kbdb'}),how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>HGVSp_h</th>\n",
       "      <th>status</th>\n",
       "      <th>flank_size</th>\n",
       "      <th>human_9mers</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "      <th>mouse_9mers</th>\n",
       "      <th>index</th>\n",
       "      <th>HLA_allele</th>\n",
       "      <th>score_h</th>\n",
       "      <th>percentile_rank_h</th>\n",
       "      <th>H2_allele</th>\n",
       "      <th>score_m</th>\n",
       "      <th>percentile_rank_m</th>\n",
       "      <th>score_kbdb</th>\n",
       "      <th>percentile_rank_kbdb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCB10</td>\n",
       "      <td>S405Y</td>\n",
       "      <td>True</td>\n",
       "      <td>39</td>\n",
       "      <td>GLYGNLIVL</td>\n",
       "      <td>Abcb10</td>\n",
       "      <td>S370Y</td>\n",
       "      <td>GLYGNLIVL</td>\n",
       "      <td>119</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>0.876096</td>\n",
       "      <td>0.04</td>\n",
       "      <td>Qa1</td>\n",
       "      <td>0.346745</td>\n",
       "      <td>0.79</td>\n",
       "      <td>0.253319</td>\n",
       "      <td>0.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCC6</td>\n",
       "      <td>Q715E</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>GEELDPPWL</td>\n",
       "      <td>Abcc6</td>\n",
       "      <td>Q713E</td>\n",
       "      <td>REELDLPWL</td>\n",
       "      <td>739</td>\n",
       "      <td>B40:01</td>\n",
       "      <td>0.750505</td>\n",
       "      <td>0.13</td>\n",
       "      <td>Qa2</td>\n",
       "      <td>0.192596</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.005281</td>\n",
       "      <td>7.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACE</td>\n",
       "      <td>S167F</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>FLDPDLTNI</td>\n",
       "      <td>Ace</td>\n",
       "      <td>S172F</td>\n",
       "      <td>FLDPELTNI</td>\n",
       "      <td>88</td>\n",
       "      <td>A02:01</td>\n",
       "      <td>0.987693</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Qa1</td>\n",
       "      <td>0.464993</td>\n",
       "      <td>0.34</td>\n",
       "      <td>0.021126</td>\n",
       "      <td>3.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACE2</td>\n",
       "      <td>D615Y</td>\n",
       "      <td>True</td>\n",
       "      <td>19</td>\n",
       "      <td>AYQSIKVRI</td>\n",
       "      <td>Ace2</td>\n",
       "      <td>D615Y</td>\n",
       "      <td>AYQSIKVRI</td>\n",
       "      <td>379</td>\n",
       "      <td>A24:02</td>\n",
       "      <td>0.837559</td>\n",
       "      <td>0.04</td>\n",
       "      <td>Kd</td>\n",
       "      <td>0.917959</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.038557</td>\n",
       "      <td>2.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACTG1</td>\n",
       "      <td>G366S</td>\n",
       "      <td>True</td>\n",
       "      <td>375</td>\n",
       "      <td>ESSPSIVHR</td>\n",
       "      <td>Actg1</td>\n",
       "      <td>G366S</td>\n",
       "      <td>ESSPSIVHR</td>\n",
       "      <td>711</td>\n",
       "      <td>A33:03</td>\n",
       "      <td>0.909769</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Qa1</td>\n",
       "      <td>0.089522</td>\n",
       "      <td>5.80</td>\n",
       "      <td>0.001162</td>\n",
       "      <td>21.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>ZNF611</td>\n",
       "      <td>D404G</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>KVCGTAFTW</td>\n",
       "      <td>Zfp966</td>\n",
       "      <td>D268G</td>\n",
       "      <td>KQCGKAFVR</td>\n",
       "      <td>224</td>\n",
       "      <td>B57:01</td>\n",
       "      <td>0.988189</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Qa1</td>\n",
       "      <td>0.014362</td>\n",
       "      <td>20.00</td>\n",
       "      <td>0.000303</td>\n",
       "      <td>38.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>ZNF611</td>\n",
       "      <td>D404G</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>KVCGTAFTW</td>\n",
       "      <td>Zfp975</td>\n",
       "      <td>D305G</td>\n",
       "      <td>NQCGKGFAQ</td>\n",
       "      <td>228</td>\n",
       "      <td>B57:01</td>\n",
       "      <td>0.988189</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Qa2</td>\n",
       "      <td>0.001335</td>\n",
       "      <td>33.00</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>47.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>380</th>\n",
       "      <td>ZNF611</td>\n",
       "      <td>D404G</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>KVCGTAFTW</td>\n",
       "      <td>Zfp976</td>\n",
       "      <td>D333G</td>\n",
       "      <td>NQCGKGFAQ</td>\n",
       "      <td>229</td>\n",
       "      <td>B57:01</td>\n",
       "      <td>0.988189</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Qa2</td>\n",
       "      <td>0.001335</td>\n",
       "      <td>33.00</td>\n",
       "      <td>0.000112</td>\n",
       "      <td>47.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>381</th>\n",
       "      <td>ZNF611</td>\n",
       "      <td>D404G</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>KVCGTAFTW</td>\n",
       "      <td>5430403G16Rik</td>\n",
       "      <td>D330G</td>\n",
       "      <td>YKCGQCGKA</td>\n",
       "      <td>195</td>\n",
       "      <td>B57:01</td>\n",
       "      <td>0.988189</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Kk</td>\n",
       "      <td>0.000284</td>\n",
       "      <td>50.00</td>\n",
       "      <td>0.000108</td>\n",
       "      <td>55.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>382</th>\n",
       "      <td>ZNF611</td>\n",
       "      <td>D404G</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "      <td>KVCGTAFTW</td>\n",
       "      <td>Zfp967</td>\n",
       "      <td>D268G</td>\n",
       "      <td>KQCGKAFVR</td>\n",
       "      <td>225</td>\n",
       "      <td>B57:01</td>\n",
       "      <td>0.988189</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Qa1</td>\n",
       "      <td>0.014362</td>\n",
       "      <td>20.00</td>\n",
       "      <td>0.000303</td>\n",
       "      <td>38.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>383 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name_h HGVSp_h status  flank_size human_9mers    gene_name_m HGVSp_m  \\\n",
       "0        ABCB10   S405Y   True          39   GLYGNLIVL         Abcb10   S370Y   \n",
       "1         ABCC6   Q715E   True           3   GEELDPPWL          Abcc6   Q713E   \n",
       "2           ACE   S167F   True          11   FLDPDLTNI            Ace   S172F   \n",
       "3          ACE2   D615Y   True          19   AYQSIKVRI           Ace2   D615Y   \n",
       "4         ACTG1   G366S   True         375   ESSPSIVHR          Actg1   G366S   \n",
       "..          ...     ...    ...         ...         ...            ...     ...   \n",
       "378      ZNF611   D404G   True           1   KVCGTAFTW         Zfp966   D268G   \n",
       "379      ZNF611   D404G   True           0   KVCGTAFTW         Zfp975   D305G   \n",
       "380      ZNF611   D404G   True           0   KVCGTAFTW         Zfp976   D333G   \n",
       "381      ZNF611   D404G   True           1   KVCGTAFTW  5430403G16Rik   D330G   \n",
       "382      ZNF611   D404G   True           1   KVCGTAFTW         Zfp967   D268G   \n",
       "\n",
       "    mouse_9mers  index HLA_allele   score_h  percentile_rank_h H2_allele  \\\n",
       "0     GLYGNLIVL    119     A02:01  0.876096               0.04       Qa1   \n",
       "1     REELDLPWL    739     B40:01  0.750505               0.13       Qa2   \n",
       "2     FLDPELTNI     88     A02:01  0.987693               0.01       Qa1   \n",
       "3     AYQSIKVRI    379     A24:02  0.837559               0.04        Kd   \n",
       "4     ESSPSIVHR    711     A33:03  0.909769               0.01       Qa1   \n",
       "..          ...    ...        ...       ...                ...       ...   \n",
       "378   KQCGKAFVR    224     B57:01  0.988189               0.01       Qa1   \n",
       "379   NQCGKGFAQ    228     B57:01  0.988189               0.01       Qa2   \n",
       "380   NQCGKGFAQ    229     B57:01  0.988189               0.01       Qa2   \n",
       "381   YKCGQCGKA    195     B57:01  0.988189               0.01        Kk   \n",
       "382   KQCGKAFVR    225     B57:01  0.988189               0.01       Qa1   \n",
       "\n",
       "      score_m  percentile_rank_m  score_kbdb  percentile_rank_kbdb  \n",
       "0    0.346745               0.79    0.253319                  0.24  \n",
       "1    0.192596               0.38    0.005281                  7.80  \n",
       "2    0.464993               0.34    0.021126                  3.00  \n",
       "3    0.917959               0.01    0.038557                  2.60  \n",
       "4    0.089522               5.80    0.001162                 21.00  \n",
       "..        ...                ...         ...                   ...  \n",
       "378  0.014362              20.00    0.000303                 38.00  \n",
       "379  0.001335              33.00    0.000112                 47.00  \n",
       "380  0.001335              33.00    0.000112                 47.00  \n",
       "381  0.000284              50.00    0.000108                 55.00  \n",
       "382  0.014362              20.00    0.000303                 38.00  \n",
       "\n",
       "[383 rows x 17 columns]"
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 194,
   "metadata": {},
   "outputs": [],
   "source": [
    "l_final.to_csv('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/h2m_neoantigen_result_v3.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## M2H  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load and pre-preocess data from 2 studies (N = 246)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dms_1 = pd.read_excel('/Users/kexindong/Documents/GitHub/Output/neoantigen/mouse_dms/li_snv.xlsx', sheet_name='Sup Data File 3 SNV Oncogenes', header=1)\n",
    "dms_2 = pd.read_excel('/Users/kexindong/Documents/GitHub/Output/neoantigen/mouse_dms/li_indels.xlsx', sheet_name='Sup Data File 4 INDEL Oncogenes', header=1)\n",
    "dms_3 = pd.read_excel('/Users/kexindong/Documents/GitHub/Output/neoantigen/mouse_dms/niknafs.xlsx',sheet_name='Table S3', header=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dms_1['source'] = 'Li 2020'\n",
    "dms_2['source'] = 'Li 2020'\n",
    "dms_3['source'] = 'Niknafs 2019'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dms_input_1 = dms_1[['Gene symbol','Position','REF','Alt','Sample ID']]\n",
    "dms_input_1.columns = ['gene_name_m','start_m','ref_seq_m','alt_seq_m','index']\n",
    "dms_input_1['end_m'] =dms_input_1['start_m']\n",
    "dms_input_1['index'] = [f'1_{(x+1):04}' for x in range(len(dms_input_1))]\n",
    "dms_input_1['type_m'] = 'SNP'\n",
    "dms_coor_1 = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/dms_1_bed.txt', sep='\\t',header=0)\n",
    "dms_1['Position'] = dms_coor_1['start']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species = 'mus_musculus'\n",
    "input_assembly = 'GRCm38'\n",
    "output_assembly = 'GRCm39'\n",
    "list_start, list_end = [],[]\n",
    "for i, x in dms_2.iterrows():\n",
    "    print(f'{i}/{len(dms_2)}')\n",
    "    region = str(x['Chr'])\n",
    "    start = x['Start']\n",
    "    end = x['End']\n",
    "    tu = convert_coordinates(species, input_assembly, output_assembly, region, start, end)\n",
    "    list_start.append(tu[0])\n",
    "    list_end.append(tu[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dms_2['Start'] = list_start\n",
    "dms_2['End'] = list_end\n",
    "dms_input_2 = dms_2[['Gene symbol','Start','End','Reference','Alt','Sample ID','Sequence variant type']]\n",
    "dms_input_2.columns = ['gene_name_h','start_h','end_h','ref_seq_h','alt_seq_h','index', 'type_h']\n",
    "dms_input_2['type_h'] = [h2m.get_type(x,y) for x,y in zip(dms_input_2['ref_seq_h'], dms_input_2['alt_seq_h'])]\n",
    "dms_input_2.columns =  ['gene_name_m','start_m','end_m','ref_seq_m','alt_seq_m','index' ,'type_m']\n",
    "dms_input_2['index'] = [f'2_{(x+1):04}' for x in range(len(dms_input_2))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dms_input_3 = dms_3[['Gene Symbol','Position (mm10)','Reference Base','Alternate Base']]\n",
    "# dms_coor_3 = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/neoantigen/P1xyCbSc9wkWYBBr.bed', sep='\\t')\n",
    "# dms_coor_3.columns = ['chr','start','end']\n",
    "dms_input_3.columns = ['gene_name_m','start_m','ref_seq_m','alt_seq_m']\n",
    "dms_input_3['start_m'] = list_start\n",
    "dms_input_3['end_m'] = list_end\n",
    "dms_input_3['index'] = [f'3_{(x+1):04}' for x in range(len(dms_input_3))]\n",
    "dms_input_3['type_m'] = 'SNP'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dms_input_3['index'] = [f'3_{(x+1):05}' for x in range(len(dms_input_3))]\n",
    "dms_input_2['index'] = [f'2_{(x+1):05}' for x in range(len(dms_input_2))]\n",
    "dms_input_1['index'] = [f'1_{(x+1):05}' for x in range(len(dms_input_1))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dms_input = pd.concat([dms_input_1, dms_input_2, dms_input_3]).reset_index(drop=True)\n",
    "# dms_input.to_csv('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/m2h_input.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "dms_input = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/m2h_input.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "dms_input = dms_input[[x[0]!='4' for x in dms_input['index']]]\n",
    "dms_input.to_csv('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/m2h_input_v2.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### M2H modeling (N = 137)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There were rows that could not be processed.\n"
     ]
    }
   ],
   "source": [
    "dms = h2m.get_tx_batch(dms_input,'m')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There were rows that could not be processed.\n"
     ]
    }
   ],
   "source": [
    "dms = h2m.query_batch(dms,'m2h')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No error occurs.\n"
     ]
    }
   ],
   "source": [
    "dms = h2m.get_tx_batch(dms,'h',37)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There were rows that could not be processed.\n"
     ]
    }
   ],
   "source": [
    "df_result = h2m.model_batch(dms, records_h, index_list_h, records_m, index_list_m, db_h, db_m, 37, direction='m2h', show_sequence=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "o = df_result[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [],
   "source": [
    "o = o[[x[0]!='4' for x in o['index']]].reset_index(drop=True)\n",
    "o.to_csv('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/m2h_output_v2.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {},
   "outputs": [],
   "source": [
    "o = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/m2h_output_v2.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "137"
      ]
     },
     "execution_count": 337,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(o[o['status']==True]['index'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_process = o.drop_duplicates(subset=['HGVSp_m','gene_name_m','tx_id_h']).reset_index(drop=True)\n",
    "df_final_success = df_process[df_process['status']==True].reset_index(drop=True)\n",
    "df_final_success = df_final_success.dropna(subset = ['flank_size_left','flank_size_right']).reset_index(drop=True)\n",
    "df_final_success['flank_size_left'] = df_final_success['flank_size_left'].fillna('0aa')\n",
    "df_final_success['flank_size'] = [int(x[:-2])+int(y[:-2]) for x,y in zip(df_final_success['flank_size_left'], df_final_success['flank_size_right'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_9mers_containing_position(sequence, position):\n",
    "    \"\"\"\n",
    "    Generate all 9-mer slices from a given DNA sequence that include a specific position.\n",
    "\n",
    "    Args:\n",
    "    sequence (str): The DNA sequence.\n",
    "    position (int): The 1-based position within the sequence.\n",
    "\n",
    "    Returns:\n",
    "    list: A list of all 9-mer slices including the specified position.\n",
    "    \"\"\"\n",
    "    def string_to_list(string):\n",
    "        try:\n",
    "            result = ast.literal_eval(string)\n",
    "            if isinstance(result, list):\n",
    "                return result\n",
    "            else:\n",
    "                raise ValueError(\"The provided string does not evaluate to a list\")\n",
    "        except:\n",
    "            raise ValueError(\"Invalid string format for list conversion\")\n",
    "    sequence = str(h2m.Translate(sequence))\n",
    "    position = string_to_list(position)[0]\n",
    "\n",
    "    k_mers = [\"\"] * 9 \n",
    "    start_index = position - 8\n",
    "\n",
    "    for i in range(9):\n",
    "        if 0 <= start_index + i <= len(sequence) - 9:\n",
    "            k_mer = sequence[start_index + i: start_index + i + 9]\n",
    "            k_mers[i] = k_mer\n",
    "    return k_mers\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_final_success = df_final_success.dropna(subset=['human_new_p_idx','mouse_new_p_idx']).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_final_success['human_mouse_9mers'] = df_final_success.apply(\n",
    "    lambda row: {\n",
    "        'human_9mers': find_9mers_containing_position(row['new_seq_h'], row['human_new_p_idx']),\n",
    "        'mouse_9mers': find_9mers_containing_position(row['new_seq_m'], row['mouse_new_p_idx'])\n",
    "    }, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [],
   "source": [
    "def expand_row(row):\n",
    "    return pd.DataFrame({\n",
    "        'human_9mers': row['human_mouse_9mers']['human_9mers'],\n",
    "        'mouse_9mers': row['human_mouse_9mers']['mouse_9mers'],\n",
    "        'index':row['index']\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [],
   "source": [
    "expanded_rows = pd.concat([expand_row(row) for index, row in df_final_success.iterrows()]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [],
   "source": [
    "expanded_rows['identical'] = [x==y for x,y in zip(expanded_rows['human_9mers'], expanded_rows['mouse_9mers'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [],
   "source": [
    "expanded_rows = expanded_rows[['*' not in x for x in expanded_rows['human_9mers']]].reset_index(drop=True)\n",
    "expanded_rows = expanded_rows[['*' not in x for x in expanded_rows['mouse_9mers']]].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [],
   "source": [
    "expanded_rows = expanded_rows[[len(x)>0 for x in expanded_rows['human_9mers']]].reset_index(drop=True)\n",
    "expanded_rows = expanded_rows[[len(x)>0 for x in expanded_rows['mouse_9mers']]].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>gene_id_m</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>chr_m</th>\n",
       "      <th>exon_num_m</th>\n",
       "      <th>strand_m</th>\n",
       "      <th>matcm</th>\n",
       "      <th>start_m</th>\n",
       "      <th>end_m</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>...</th>\n",
       "      <th>human_tx_idx_ori</th>\n",
       "      <th>human_p_idx_ori</th>\n",
       "      <th>human_new_p_idx_ori</th>\n",
       "      <th>dist_h</th>\n",
       "      <th>new_seq_h</th>\n",
       "      <th>human_tx_idx</th>\n",
       "      <th>human_p_idx</th>\n",
       "      <th>human_new_p_idx</th>\n",
       "      <th>index</th>\n",
       "      <th>human_mouse_9mers</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>69480028</td>\n",
       "      <td>69480028</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>[733]</td>\n",
       "      <td>[244]</td>\n",
       "      <td>[244]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTC...</td>\n",
       "      <td>[733]</td>\n",
       "      <td>[244]</td>\n",
       "      <td>[244]</td>\n",
       "      <td>1_00001</td>\n",
       "      <td>{'human_9mers': ['MCNSSCMGA', 'CNSSCMGAM', 'NS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552.14</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>chr11</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>69480040</td>\n",
       "      <td>69480040</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>[745]</td>\n",
       "      <td>[248]</td>\n",
       "      <td>[248]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTC...</td>\n",
       "      <td>[744, 745]</td>\n",
       "      <td>[248, 248]</td>\n",
       "      <td>[248]</td>\n",
       "      <td>1_00002</td>\n",
       "      <td>{'human_9mers': ['SCMGGMNRL', 'CMGGMNRLP', 'MG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Notch1</td>\n",
       "      <td>ENSMUSG00000026923.16</td>\n",
       "      <td>ENSMUST00000028288.5</td>\n",
       "      <td>chr2</td>\n",
       "      <td>34</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>26356613</td>\n",
       "      <td>26356613</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>[5032]</td>\n",
       "      <td>[1677]</td>\n",
       "      <td>[1677]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATGCCGCCGCTCCTGGCGCCCCTGCTCTGCCTGGCGCTGCTGCCCG...</td>\n",
       "      <td>[5032]</td>\n",
       "      <td>[1677]</td>\n",
       "      <td>[1677]</td>\n",
       "      <td>1_00004</td>\n",
       "      <td>{'human_9mers': ['DVRGSIVYP', 'VRGSIVYPE', 'RG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Notch1</td>\n",
       "      <td>ENSMUSG00000026923.16</td>\n",
       "      <td>ENSMUST00000028288.5</td>\n",
       "      <td>chr2</td>\n",
       "      <td>34</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>26371174</td>\n",
       "      <td>26371174</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>[910]</td>\n",
       "      <td>[303]</td>\n",
       "      <td>[303]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATGCCGCCGCTCCTGGCGCCCCTGCTCTGCCTGGCGCTGCTGCCCG...</td>\n",
       "      <td>[910]</td>\n",
       "      <td>[303]</td>\n",
       "      <td>[303]</td>\n",
       "      <td>1_00005</td>\n",
       "      <td>{'human_9mers': ['VDECQLMPS', 'DECQLMPSA', 'EC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Notch1</td>\n",
       "      <td>ENSMUSG00000026923.16</td>\n",
       "      <td>ENSMUST00000028288.5</td>\n",
       "      <td>chr2</td>\n",
       "      <td>34</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>26371195</td>\n",
       "      <td>26371195</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>[889]</td>\n",
       "      <td>[296]</td>\n",
       "      <td>[296]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATGCCGCCGCTCCTGGCGCCCCTGCTCTGCCTGGCGCTGCTGCCCG...</td>\n",
       "      <td>[889]</td>\n",
       "      <td>[296]</td>\n",
       "      <td>[296]</td>\n",
       "      <td>1_00006</td>\n",
       "      <td>{'human_9mers': ['GQYCTEDVG', 'QYCTEDVGE', 'YC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>Bcl9</td>\n",
       "      <td>ENSMUSG00000038256.16</td>\n",
       "      <td>ENSMUST00000046521.14</td>\n",
       "      <td>chr3</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>97115926</td>\n",
       "      <td>97115926</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>[2769]</td>\n",
       "      <td>[923]</td>\n",
       "      <td>[923]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATGCATTCCAGTAACCCTAAAGTGAGGAGCTCTCCATCAGGAAACA...</td>\n",
       "      <td>[2769]</td>\n",
       "      <td>[923]</td>\n",
       "      <td>[923]</td>\n",
       "      <td>3_00153</td>\n",
       "      <td>{'human_9mers': ['ASPVHLKSS', 'SPVHLKSSS', 'PV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>145</th>\n",
       "      <td>Sbno2</td>\n",
       "      <td>ENSMUSG00000035673.11</td>\n",
       "      <td>ENSMUST00000219260.2</td>\n",
       "      <td>chr10</td>\n",
       "      <td>31</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>79896637</td>\n",
       "      <td>79896637</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>[2613]</td>\n",
       "      <td>[871]</td>\n",
       "      <td>[871]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATGCTTGCAGTGGGGCCCGCCATGGACAGGGATTACCCGCAGCATG...</td>\n",
       "      <td>[2613, 2614, 2615]</td>\n",
       "      <td>[871, 871, 871]</td>\n",
       "      <td>[871]</td>\n",
       "      <td>3_00154</td>\n",
       "      <td>{'human_9mers': ['RFASIVAKW', 'FASIVAKWL', 'AS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>Mamdc4</td>\n",
       "      <td>ENSMUSG00000026941.17</td>\n",
       "      <td>ENSMUST00000095117.10</td>\n",
       "      <td>chr2</td>\n",
       "      <td>29</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>25460022</td>\n",
       "      <td>25460022</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>[224]</td>\n",
       "      <td>[74]</td>\n",
       "      <td>[74]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATGCCTCTGTCCAGCCACCTGCTGCCCGCCTTGGTCCTGTTCCTGG...</td>\n",
       "      <td>[224]</td>\n",
       "      <td>[74]</td>\n",
       "      <td>[74]</td>\n",
       "      <td>3_00155</td>\n",
       "      <td>{'human_9mers': ['DFEQDPCGC', 'FEQDPCGCR', 'EQ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147</th>\n",
       "      <td>Srxn1</td>\n",
       "      <td>ENSMUSG00000032802.9</td>\n",
       "      <td>ENSMUST00000041500.8</td>\n",
       "      <td>chr2</td>\n",
       "      <td>2</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>151947715</td>\n",
       "      <td>151947715</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>[54]</td>\n",
       "      <td>[18]</td>\n",
       "      <td>[18]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATGGGGCTGCGTGCAGGAGGAACGCTGGGCAGGGCCGGCGCGGGTC...</td>\n",
       "      <td>[54]</td>\n",
       "      <td>[18]</td>\n",
       "      <td>[18]</td>\n",
       "      <td>3_00156</td>\n",
       "      <td>{'human_9mers': ['RAGAGRGAA', 'AGAGRGAAE', 'GA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>1110004F10Rik</td>\n",
       "      <td>ENSMUSG00000030663.13</td>\n",
       "      <td>ENSMUST00000032899.12</td>\n",
       "      <td>chr7</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>115703549</td>\n",
       "      <td>115703549</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>[426]</td>\n",
       "      <td>[142]</td>\n",
       "      <td>[142]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ATGAGTGCTGCCAGAGAGTCTCACCCGCATGGGGTGAAGCGTTCAG...</td>\n",
       "      <td>[426]</td>\n",
       "      <td>[142]</td>\n",
       "      <td>[142]</td>\n",
       "      <td>3_00157</td>\n",
       "      <td>{'human_9mers': ['DSESDSESQ', 'SESDSESQK', 'ES...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>149 rows × 60 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene_name_m              gene_id_m                tx_id_m  chr_m  \\\n",
       "0            Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "1            Trp53  ENSMUSG00000059552.14  ENSMUST00000108658.10  chr11   \n",
       "2           Notch1  ENSMUSG00000026923.16   ENSMUST00000028288.5   chr2   \n",
       "3           Notch1  ENSMUSG00000026923.16   ENSMUST00000028288.5   chr2   \n",
       "4           Notch1  ENSMUSG00000026923.16   ENSMUST00000028288.5   chr2   \n",
       "..             ...                    ...                    ...    ...   \n",
       "144           Bcl9  ENSMUSG00000038256.16  ENSMUST00000046521.14   chr3   \n",
       "145          Sbno2  ENSMUSG00000035673.11   ENSMUST00000219260.2  chr10   \n",
       "146         Mamdc4  ENSMUSG00000026941.17  ENSMUST00000095117.10   chr2   \n",
       "147          Srxn1   ENSMUSG00000032802.9   ENSMUST00000041500.8   chr2   \n",
       "148  1110004F10Rik  ENSMUSG00000030663.13  ENSMUST00000032899.12   chr7   \n",
       "\n",
       "     exon_num_m strand_m  matcm    start_m      end_m ref_seq_m  ...  \\\n",
       "0            10        +   True   69480028   69480028         G  ...   \n",
       "1            10        +   True   69480040   69480040         G  ...   \n",
       "2            34        -   True   26356613   26356613         A  ...   \n",
       "3            34        -   True   26371174   26371174         T  ...   \n",
       "4            34        -   True   26371195   26371195         T  ...   \n",
       "..          ...      ...    ...        ...        ...       ...  ...   \n",
       "144           7        -   True   97115926   97115926         G  ...   \n",
       "145          31        -   True   79896637   79896637         G  ...   \n",
       "146          29        -   True   25460022   25460022         C  ...   \n",
       "147           2        +   True  151947715  151947715         C  ...   \n",
       "148           5        +   True  115703549  115703549         G  ...   \n",
       "\n",
       "    human_tx_idx_ori human_p_idx_ori human_new_p_idx_ori dist_h  \\\n",
       "0              [733]           [244]               [244]    NaN   \n",
       "1              [745]           [248]               [248]    NaN   \n",
       "2             [5032]          [1677]              [1677]    NaN   \n",
       "3              [910]           [303]               [303]    NaN   \n",
       "4              [889]           [296]               [296]    NaN   \n",
       "..               ...             ...                 ...    ...   \n",
       "144           [2769]           [923]               [923]    NaN   \n",
       "145           [2613]           [871]               [871]    NaN   \n",
       "146            [224]            [74]                [74]    NaN   \n",
       "147             [54]            [18]                [18]    NaN   \n",
       "148            [426]           [142]               [142]    NaN   \n",
       "\n",
       "                                             new_seq_h        human_tx_idx  \\\n",
       "0    ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTC...               [733]   \n",
       "1    ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTC...          [744, 745]   \n",
       "2    ATGCCGCCGCTCCTGGCGCCCCTGCTCTGCCTGGCGCTGCTGCCCG...              [5032]   \n",
       "3    ATGCCGCCGCTCCTGGCGCCCCTGCTCTGCCTGGCGCTGCTGCCCG...               [910]   \n",
       "4    ATGCCGCCGCTCCTGGCGCCCCTGCTCTGCCTGGCGCTGCTGCCCG...               [889]   \n",
       "..                                                 ...                 ...   \n",
       "144  ATGCATTCCAGTAACCCTAAAGTGAGGAGCTCTCCATCAGGAAACA...              [2769]   \n",
       "145  ATGCTTGCAGTGGGGCCCGCCATGGACAGGGATTACCCGCAGCATG...  [2613, 2614, 2615]   \n",
       "146  ATGCCTCTGTCCAGCCACCTGCTGCCCGCCTTGGTCCTGTTCCTGG...               [224]   \n",
       "147  ATGGGGCTGCGTGCAGGAGGAACGCTGGGCAGGGCCGGCGCGGGTC...                [54]   \n",
       "148  ATGAGTGCTGCCAGAGAGTCTCACCCGCATGGGGTGAAGCGTTCAG...               [426]   \n",
       "\n",
       "         human_p_idx  human_new_p_idx    index  \\\n",
       "0              [244]            [244]  1_00001   \n",
       "1         [248, 248]            [248]  1_00002   \n",
       "2             [1677]           [1677]  1_00004   \n",
       "3              [303]            [303]  1_00005   \n",
       "4              [296]            [296]  1_00006   \n",
       "..               ...              ...      ...   \n",
       "144            [923]            [923]  3_00153   \n",
       "145  [871, 871, 871]            [871]  3_00154   \n",
       "146             [74]             [74]  3_00155   \n",
       "147             [18]             [18]  3_00156   \n",
       "148            [142]            [142]  3_00157   \n",
       "\n",
       "                                     human_mouse_9mers  \n",
       "0    {'human_9mers': ['MCNSSCMGA', 'CNSSCMGAM', 'NS...  \n",
       "1    {'human_9mers': ['SCMGGMNRL', 'CMGGMNRLP', 'MG...  \n",
       "2    {'human_9mers': ['DVRGSIVYP', 'VRGSIVYPE', 'RG...  \n",
       "3    {'human_9mers': ['VDECQLMPS', 'DECQLMPSA', 'EC...  \n",
       "4    {'human_9mers': ['GQYCTEDVG', 'QYCTEDVGE', 'YC...  \n",
       "..                                                 ...  \n",
       "144  {'human_9mers': ['ASPVHLKSS', 'SPVHLKSSS', 'PV...  \n",
       "145  {'human_9mers': ['RFASIVAKW', 'FASIVAKWL', 'AS...  \n",
       "146  {'human_9mers': ['DFEQDPCGC', 'FEQDPCGCR', 'EQ...  \n",
       "147  {'human_9mers': ['RAGAGRGAA', 'AGAGRGAAE', 'GA...  \n",
       "148  {'human_9mers': ['DSESDSESQ', 'SESDSESQK', 'ES...  \n",
       "\n",
       "[149 rows x 60 columns]"
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_final_success"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human_9mers</th>\n",
       "      <th>mouse_9mers</th>\n",
       "      <th>index</th>\n",
       "      <th>identical</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MCNSSCMGA</td>\n",
       "      <td>MCNSSCMGA</td>\n",
       "      <td>1_00001</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CNSSCMGAM</td>\n",
       "      <td>CNSSCMGAM</td>\n",
       "      <td>1_00001</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NSSCMGAMN</td>\n",
       "      <td>NSSCMGAMN</td>\n",
       "      <td>1_00001</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SSCMGAMNR</td>\n",
       "      <td>SSCMGAMNR</td>\n",
       "      <td>1_00001</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SCMGAMNRR</td>\n",
       "      <td>SCMGAMNRR</td>\n",
       "      <td>1_00001</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1280</th>\n",
       "      <td>DSESQKEES</td>\n",
       "      <td>DSESQKEES</td>\n",
       "      <td>3_00157</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1281</th>\n",
       "      <td>SESQKEESA</td>\n",
       "      <td>SESQKEESA</td>\n",
       "      <td>3_00157</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1282</th>\n",
       "      <td>ESQKEESAE</td>\n",
       "      <td>ESQKEESAE</td>\n",
       "      <td>3_00157</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1283</th>\n",
       "      <td>SQKEESAEE</td>\n",
       "      <td>SQKEESAEE</td>\n",
       "      <td>3_00157</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1284</th>\n",
       "      <td>QKEESAEEL</td>\n",
       "      <td>QKEESAEEL</td>\n",
       "      <td>3_00157</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1231 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     human_9mers mouse_9mers    index  identical\n",
       "0      MCNSSCMGA   MCNSSCMGA  1_00001       True\n",
       "1      CNSSCMGAM   CNSSCMGAM  1_00001       True\n",
       "2      NSSCMGAMN   NSSCMGAMN  1_00001       True\n",
       "3      SSCMGAMNR   SSCMGAMNR  1_00001       True\n",
       "4      SCMGAMNRR   SCMGAMNRR  1_00001       True\n",
       "...          ...         ...      ...        ...\n",
       "1280   DSESQKEES   DSESQKEES  3_00157       True\n",
       "1281   SESQKEESA   SESQKEESA  3_00157       True\n",
       "1282   ESQKEESAE   ESQKEESAE  3_00157       True\n",
       "1283   SQKEESAEE   SQKEESAEE  3_00157       True\n",
       "1284   QKEESAEEL   QKEESAEEL  3_00157       True\n",
       "\n",
       "[1231 rows x 4 columns]"
      ]
     },
     "execution_count": 248,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expanded_rows.drop_duplicates('human_9mers')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate NetMHCpan inputs (N = 1231/1089)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1231"
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(expanded_rows['human_9mers'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1089"
      ]
     },
     "execution_count": 330,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(expanded_rows['mouse_9mers'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, x in expanded_rows.iterrows():\n",
    "    k = x['index']\n",
    "    l = x['human_9mers']\n",
    "    with open('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/m2h_input_h_sub_3.txt', 'a') as file:\n",
    "        file.writelines(f'>{k}\\n{l}\\n')\n",
    "    l = x['mouse_9mers']\n",
    "    with open('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/m2h_input_m_sub_3.txt', 'a') as file:\n",
    "        file.writelines(f'>{k}\\n{l}\\n')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Human results (top score is selected for each peptide among different MHC alleles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_h_i_list = []\n",
    "for x in ['A1','A2','B1','B2','B3','B4','C1','C2','EG']:\n",
    "    result_h_i_list.append(pd.read_csv(f'/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/m2h_output_v3/m2h_output_h_{x}.txt', header=0, sep='\\t', comment=\"#\"))\n",
    "result_h_i = pd.concat(result_h_i_list).reset_index(drop=True)\n",
    "del result_h_i_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_h_i.to_csv('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/m2h_output_v3/m2h_output_h_v3.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 316,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_h_i = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/m2h_output_v3/m2h_output_h_v3.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_h_i['allele'] = [x[4:] for x in result_h_i['allele']]\n",
    "result_h_i['allele'] = [''.join(str(x).split('*')) for x in result_h_i['allele']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_h_i_unique = result_h_i.sort_values(by='score', ascending=False).reset_index(drop=True).drop_duplicates(subset='peptide')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = pd.merge(df_output,result_h_i_unique[['allele','peptide','score','percentile_rank']].rename(columns={'peptide':'human_9mers',\n",
    "                                                                                                           'allele':'HLA_alleles',\n",
    "                                                                                                           'score':'score_h',\n",
    "                                                                                                           'percentile_rank':'percentile_rank_h'\n",
    "                                                                                                           }),how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = g.drop_duplicates().reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mouse results (top score is selected for each peptide among different MHC alleles)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_m_i_list = []\n",
    "for x in ['1','2']:\n",
    "    result_m_i_list.append(pd.read_csv(f'/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/m2h_output_v3/m2h_output_m_{x}.txt', header=0, sep='\\t', comment=\"#\"))\n",
    "result_m_i = pd.concat(result_m_i_list).reset_index(drop=True)\n",
    "del result_m_i_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_m_i.to_csv('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/m2h_output_v3/m2h_output_m_v3.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_m_i = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/m2h_output_v3/m2h_output_m_v3.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_m_i['allele'] = [x[4:] for x in result_m_i['allele']]\n",
    "result_m_i['allele'] = [''.join(str(x).split('*')) for x in result_m_i['allele']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_m_i_unique = result_m_i.sort_values(by='score', ascending=False).reset_index(drop=True).drop_duplicates(subset='peptide')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {},
   "outputs": [],
   "source": [
    "l_both_best = pd.merge(g,result_m_i_unique[['allele','peptide','score','percentile_rank']].rename(columns={'peptide':'mouse_9mers',\n",
    "                                                                                                           'allele':'H2_allele',\n",
    "                                                                                                           'percentile_rank':'percentile_rank_m',\n",
    "                                                                                                           'score':'score_m'}),how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>HGVSp_h</th>\n",
       "      <th>status</th>\n",
       "      <th>flank_size</th>\n",
       "      <th>human_9mers</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "      <th>mouse_9mers</th>\n",
       "      <th>index</th>\n",
       "      <th>HLA_alleles</th>\n",
       "      <th>score_h</th>\n",
       "      <th>percentile_rank_h</th>\n",
       "      <th>H2_allele</th>\n",
       "      <th>score_m</th>\n",
       "      <th>percentile_rank_m</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCB9</td>\n",
       "      <td>Q520delinsQ</td>\n",
       "      <td>True</td>\n",
       "      <td>115</td>\n",
       "      <td>RPHTQVLQN</td>\n",
       "      <td>Abcb9</td>\n",
       "      <td>Q516delinsQ</td>\n",
       "      <td>RPHTQVLQN</td>\n",
       "      <td>3_00022</td>\n",
       "      <td>B55:01</td>\n",
       "      <td>0.058610</td>\n",
       "      <td>1.70</td>\n",
       "      <td>Dq</td>\n",
       "      <td>0.008150</td>\n",
       "      <td>14.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCB9</td>\n",
       "      <td>Q520delinsQ</td>\n",
       "      <td>True</td>\n",
       "      <td>115</td>\n",
       "      <td>RTRPHTQVL</td>\n",
       "      <td>Abcb9</td>\n",
       "      <td>Q516delinsQ</td>\n",
       "      <td>RTRPHTQVL</td>\n",
       "      <td>3_00022</td>\n",
       "      <td>A30:01</td>\n",
       "      <td>0.925992</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Qa1</td>\n",
       "      <td>0.804364</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCB9</td>\n",
       "      <td>Q520delinsQ</td>\n",
       "      <td>True</td>\n",
       "      <td>115</td>\n",
       "      <td>TRPHTQVLQ</td>\n",
       "      <td>Abcb9</td>\n",
       "      <td>Q516delinsQ</td>\n",
       "      <td>TRPHTQVLQ</td>\n",
       "      <td>3_00022</td>\n",
       "      <td>B27:05</td>\n",
       "      <td>0.009969</td>\n",
       "      <td>4.40</td>\n",
       "      <td>Qa1</td>\n",
       "      <td>0.010245</td>\n",
       "      <td>24.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCB9</td>\n",
       "      <td>Q520delinsQ</td>\n",
       "      <td>True</td>\n",
       "      <td>115</td>\n",
       "      <td>PHTQVLQNV</td>\n",
       "      <td>Abcb9</td>\n",
       "      <td>Q516delinsQ</td>\n",
       "      <td>PHTQVLQNV</td>\n",
       "      <td>3_00022</td>\n",
       "      <td>B38:01</td>\n",
       "      <td>0.212903</td>\n",
       "      <td>0.48</td>\n",
       "      <td>Kd</td>\n",
       "      <td>0.025846</td>\n",
       "      <td>2.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCB9</td>\n",
       "      <td>Q520delinsQ</td>\n",
       "      <td>True</td>\n",
       "      <td>115</td>\n",
       "      <td>HTQVLQNVS</td>\n",
       "      <td>Abcb9</td>\n",
       "      <td>Q516delinsQ</td>\n",
       "      <td>HTQVLQNVS</td>\n",
       "      <td>3_00022</td>\n",
       "      <td>A30:01</td>\n",
       "      <td>0.007771</td>\n",
       "      <td>9.70</td>\n",
       "      <td>Qa1</td>\n",
       "      <td>0.003275</td>\n",
       "      <td>36.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3986</th>\n",
       "      <td>ZNF878</td>\n",
       "      <td>P23delinsP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>LDPSQKNLC</td>\n",
       "      <td>Zfp455</td>\n",
       "      <td>P24delinsP</td>\n",
       "      <td>LEPAQWDLY</td>\n",
       "      <td>3_00036</td>\n",
       "      <td>B37:01</td>\n",
       "      <td>0.005016</td>\n",
       "      <td>6.90</td>\n",
       "      <td>Qa1</td>\n",
       "      <td>0.059734</td>\n",
       "      <td>8.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3987</th>\n",
       "      <td>ZNF878</td>\n",
       "      <td>P23delinsP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>PVQRNLYQD</td>\n",
       "      <td>Zfp455</td>\n",
       "      <td>P24delinsP</td>\n",
       "      <td>PAQWDLYRD</td>\n",
       "      <td>3_00036</td>\n",
       "      <td>A30:01</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>77.00</td>\n",
       "      <td>Kq</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>83.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3988</th>\n",
       "      <td>ZNF878</td>\n",
       "      <td>P23delinsP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>LDPVQRNLY</td>\n",
       "      <td>Zfp455</td>\n",
       "      <td>P24delinsP</td>\n",
       "      <td>LEPAQWDLY</td>\n",
       "      <td>3_00036</td>\n",
       "      <td>A01:01</td>\n",
       "      <td>0.079631</td>\n",
       "      <td>1.20</td>\n",
       "      <td>Qa1</td>\n",
       "      <td>0.059734</td>\n",
       "      <td>8.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3989</th>\n",
       "      <td>ZNF878</td>\n",
       "      <td>P23delinsP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>REEWALLGP</td>\n",
       "      <td>Zfp455</td>\n",
       "      <td>P24delinsP</td>\n",
       "      <td>PEEWECLEP</td>\n",
       "      <td>3_00036</td>\n",
       "      <td>B50:01</td>\n",
       "      <td>0.295977</td>\n",
       "      <td>0.32</td>\n",
       "      <td>Qa2</td>\n",
       "      <td>0.002958</td>\n",
       "      <td>24.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3990</th>\n",
       "      <td>ZNF878</td>\n",
       "      <td>P23delinsP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>DPVQRNLYQ</td>\n",
       "      <td>Zfp455</td>\n",
       "      <td>P24delinsP</td>\n",
       "      <td>EPAQWDLYR</td>\n",
       "      <td>3_00036</td>\n",
       "      <td>B35:01</td>\n",
       "      <td>0.012099</td>\n",
       "      <td>3.40</td>\n",
       "      <td>Kq</td>\n",
       "      <td>0.006219</td>\n",
       "      <td>14.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3991 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name_h      HGVSp_h  status  flank_size human_9mers gene_name_m  \\\n",
       "0          ABCB9  Q520delinsQ    True         115   RPHTQVLQN       Abcb9   \n",
       "1          ABCB9  Q520delinsQ    True         115   RTRPHTQVL       Abcb9   \n",
       "2          ABCB9  Q520delinsQ    True         115   TRPHTQVLQ       Abcb9   \n",
       "3          ABCB9  Q520delinsQ    True         115   PHTQVLQNV       Abcb9   \n",
       "4          ABCB9  Q520delinsQ    True         115   HTQVLQNVS       Abcb9   \n",
       "...          ...          ...     ...         ...         ...         ...   \n",
       "3986      ZNF878   P23delinsP    True           0   LDPSQKNLC      Zfp455   \n",
       "3987      ZNF878   P23delinsP    True           0   PVQRNLYQD      Zfp455   \n",
       "3988      ZNF878   P23delinsP    True           0   LDPVQRNLY      Zfp455   \n",
       "3989      ZNF878   P23delinsP    True           0   REEWALLGP      Zfp455   \n",
       "3990      ZNF878   P23delinsP    True           0   DPVQRNLYQ      Zfp455   \n",
       "\n",
       "          HGVSp_m mouse_9mers    index HLA_alleles   score_h  \\\n",
       "0     Q516delinsQ   RPHTQVLQN  3_00022      B55:01  0.058610   \n",
       "1     Q516delinsQ   RTRPHTQVL  3_00022      A30:01  0.925992   \n",
       "2     Q516delinsQ   TRPHTQVLQ  3_00022      B27:05  0.009969   \n",
       "3     Q516delinsQ   PHTQVLQNV  3_00022      B38:01  0.212903   \n",
       "4     Q516delinsQ   HTQVLQNVS  3_00022      A30:01  0.007771   \n",
       "...           ...         ...      ...         ...       ...   \n",
       "3986   P24delinsP   LEPAQWDLY  3_00036      B37:01  0.005016   \n",
       "3987   P24delinsP   PAQWDLYRD  3_00036      A30:01  0.000029   \n",
       "3988   P24delinsP   LEPAQWDLY  3_00036      A01:01  0.079631   \n",
       "3989   P24delinsP   PEEWECLEP  3_00036      B50:01  0.295977   \n",
       "3990   P24delinsP   EPAQWDLYR  3_00036      B35:01  0.012099   \n",
       "\n",
       "      percentile_rank_h H2_allele   score_m  percentile_rank_m  \n",
       "0                  1.70        Dq  0.008150              14.00  \n",
       "1                  0.01       Qa1  0.804364               0.01  \n",
       "2                  4.40       Qa1  0.010245              24.00  \n",
       "3                  0.48        Kd  0.025846               2.40  \n",
       "4                  9.70       Qa1  0.003275              36.00  \n",
       "...                 ...       ...       ...                ...  \n",
       "3986               6.90       Qa1  0.059734               8.40  \n",
       "3987              77.00        Kq  0.000031              83.00  \n",
       "3988               1.20       Qa1  0.059734               8.40  \n",
       "3989               0.32       Qa2  0.002958              24.00  \n",
       "3990               3.40        Kq  0.006219              14.00  \n",
       "\n",
       "[3991 rows x 15 columns]"
      ]
     },
     "execution_count": 356,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_both_best"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mouse Kb/Db results (most common allele: Kb/Db)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_m_i_kb = result_m_i[result_m_i['allele'].isin(['Kb','Db'])].reset_index(drop=True).sort_values(by='score', ascending=False).reset_index(drop=True).drop_duplicates(subset='peptide')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "metadata": {},
   "outputs": [],
   "source": [
    "l_final = pd.merge(l_both_best,result_m_i_kb[['peptide','score','percentile_rank']].rename(columns={'peptide':'mouse_9mers',\n",
    "                                                                                                           'percentile_rank':'percentile_rank_kbdb',\n",
    "                                                                                                           'score':'score_kbdb'}),how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 359,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>HGVSp_h</th>\n",
       "      <th>status</th>\n",
       "      <th>flank_size</th>\n",
       "      <th>human_9mers</th>\n",
       "      <th>gene_name_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "      <th>mouse_9mers</th>\n",
       "      <th>index</th>\n",
       "      <th>HLA_alleles</th>\n",
       "      <th>score_h</th>\n",
       "      <th>percentile_rank_h</th>\n",
       "      <th>H2_allele</th>\n",
       "      <th>score_m</th>\n",
       "      <th>percentile_rank_m</th>\n",
       "      <th>score_kbdb</th>\n",
       "      <th>percentile_rank_kbdb</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABCB9</td>\n",
       "      <td>Q520delinsQ</td>\n",
       "      <td>True</td>\n",
       "      <td>115</td>\n",
       "      <td>RPHTQVLQN</td>\n",
       "      <td>Abcb9</td>\n",
       "      <td>Q516delinsQ</td>\n",
       "      <td>RPHTQVLQN</td>\n",
       "      <td>3_00022</td>\n",
       "      <td>B55:01</td>\n",
       "      <td>0.058610</td>\n",
       "      <td>1.70</td>\n",
       "      <td>Dq</td>\n",
       "      <td>0.008150</td>\n",
       "      <td>14.00</td>\n",
       "      <td>0.000145</td>\n",
       "      <td>50.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABCB9</td>\n",
       "      <td>Q520delinsQ</td>\n",
       "      <td>True</td>\n",
       "      <td>115</td>\n",
       "      <td>RTRPHTQVL</td>\n",
       "      <td>Abcb9</td>\n",
       "      <td>Q516delinsQ</td>\n",
       "      <td>RTRPHTQVL</td>\n",
       "      <td>3_00022</td>\n",
       "      <td>A30:01</td>\n",
       "      <td>0.925992</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Qa1</td>\n",
       "      <td>0.804364</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.176445</td>\n",
       "      <td>0.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABCB9</td>\n",
       "      <td>Q520delinsQ</td>\n",
       "      <td>True</td>\n",
       "      <td>115</td>\n",
       "      <td>TRPHTQVLQ</td>\n",
       "      <td>Abcb9</td>\n",
       "      <td>Q516delinsQ</td>\n",
       "      <td>TRPHTQVLQ</td>\n",
       "      <td>3_00022</td>\n",
       "      <td>B27:05</td>\n",
       "      <td>0.009969</td>\n",
       "      <td>4.40</td>\n",
       "      <td>Qa1</td>\n",
       "      <td>0.010245</td>\n",
       "      <td>24.00</td>\n",
       "      <td>0.000825</td>\n",
       "      <td>24.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABCB9</td>\n",
       "      <td>Q520delinsQ</td>\n",
       "      <td>True</td>\n",
       "      <td>115</td>\n",
       "      <td>PHTQVLQNV</td>\n",
       "      <td>Abcb9</td>\n",
       "      <td>Q516delinsQ</td>\n",
       "      <td>PHTQVLQNV</td>\n",
       "      <td>3_00022</td>\n",
       "      <td>B38:01</td>\n",
       "      <td>0.212903</td>\n",
       "      <td>0.48</td>\n",
       "      <td>Kd</td>\n",
       "      <td>0.025846</td>\n",
       "      <td>2.40</td>\n",
       "      <td>0.001880</td>\n",
       "      <td>16.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABCB9</td>\n",
       "      <td>Q520delinsQ</td>\n",
       "      <td>True</td>\n",
       "      <td>115</td>\n",
       "      <td>HTQVLQNVS</td>\n",
       "      <td>Abcb9</td>\n",
       "      <td>Q516delinsQ</td>\n",
       "      <td>HTQVLQNVS</td>\n",
       "      <td>3_00022</td>\n",
       "      <td>A30:01</td>\n",
       "      <td>0.007771</td>\n",
       "      <td>9.70</td>\n",
       "      <td>Qa1</td>\n",
       "      <td>0.003275</td>\n",
       "      <td>36.00</td>\n",
       "      <td>0.000603</td>\n",
       "      <td>25.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3986</th>\n",
       "      <td>ZNF878</td>\n",
       "      <td>P23delinsP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>LDPSQKNLC</td>\n",
       "      <td>Zfp455</td>\n",
       "      <td>P24delinsP</td>\n",
       "      <td>LEPAQWDLY</td>\n",
       "      <td>3_00036</td>\n",
       "      <td>B37:01</td>\n",
       "      <td>0.005016</td>\n",
       "      <td>6.90</td>\n",
       "      <td>Qa1</td>\n",
       "      <td>0.059734</td>\n",
       "      <td>8.40</td>\n",
       "      <td>0.005304</td>\n",
       "      <td>7.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3987</th>\n",
       "      <td>ZNF878</td>\n",
       "      <td>P23delinsP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>PVQRNLYQD</td>\n",
       "      <td>Zfp455</td>\n",
       "      <td>P24delinsP</td>\n",
       "      <td>PAQWDLYRD</td>\n",
       "      <td>3_00036</td>\n",
       "      <td>A30:01</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>77.00</td>\n",
       "      <td>Kq</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>83.00</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>98.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3988</th>\n",
       "      <td>ZNF878</td>\n",
       "      <td>P23delinsP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>LDPVQRNLY</td>\n",
       "      <td>Zfp455</td>\n",
       "      <td>P24delinsP</td>\n",
       "      <td>LEPAQWDLY</td>\n",
       "      <td>3_00036</td>\n",
       "      <td>A01:01</td>\n",
       "      <td>0.079631</td>\n",
       "      <td>1.20</td>\n",
       "      <td>Qa1</td>\n",
       "      <td>0.059734</td>\n",
       "      <td>8.40</td>\n",
       "      <td>0.005304</td>\n",
       "      <td>7.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3989</th>\n",
       "      <td>ZNF878</td>\n",
       "      <td>P23delinsP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>REEWALLGP</td>\n",
       "      <td>Zfp455</td>\n",
       "      <td>P24delinsP</td>\n",
       "      <td>PEEWECLEP</td>\n",
       "      <td>3_00036</td>\n",
       "      <td>B50:01</td>\n",
       "      <td>0.295977</td>\n",
       "      <td>0.32</td>\n",
       "      <td>Qa2</td>\n",
       "      <td>0.002958</td>\n",
       "      <td>24.00</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>95.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3990</th>\n",
       "      <td>ZNF878</td>\n",
       "      <td>P23delinsP</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>DPVQRNLYQ</td>\n",
       "      <td>Zfp455</td>\n",
       "      <td>P24delinsP</td>\n",
       "      <td>EPAQWDLYR</td>\n",
       "      <td>3_00036</td>\n",
       "      <td>B35:01</td>\n",
       "      <td>0.012099</td>\n",
       "      <td>3.40</td>\n",
       "      <td>Kq</td>\n",
       "      <td>0.006219</td>\n",
       "      <td>14.00</td>\n",
       "      <td>0.000065</td>\n",
       "      <td>56.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3991 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_name_h      HGVSp_h  status  flank_size human_9mers gene_name_m  \\\n",
       "0          ABCB9  Q520delinsQ    True         115   RPHTQVLQN       Abcb9   \n",
       "1          ABCB9  Q520delinsQ    True         115   RTRPHTQVL       Abcb9   \n",
       "2          ABCB9  Q520delinsQ    True         115   TRPHTQVLQ       Abcb9   \n",
       "3          ABCB9  Q520delinsQ    True         115   PHTQVLQNV       Abcb9   \n",
       "4          ABCB9  Q520delinsQ    True         115   HTQVLQNVS       Abcb9   \n",
       "...          ...          ...     ...         ...         ...         ...   \n",
       "3986      ZNF878   P23delinsP    True           0   LDPSQKNLC      Zfp455   \n",
       "3987      ZNF878   P23delinsP    True           0   PVQRNLYQD      Zfp455   \n",
       "3988      ZNF878   P23delinsP    True           0   LDPVQRNLY      Zfp455   \n",
       "3989      ZNF878   P23delinsP    True           0   REEWALLGP      Zfp455   \n",
       "3990      ZNF878   P23delinsP    True           0   DPVQRNLYQ      Zfp455   \n",
       "\n",
       "          HGVSp_m mouse_9mers    index HLA_alleles   score_h  \\\n",
       "0     Q516delinsQ   RPHTQVLQN  3_00022      B55:01  0.058610   \n",
       "1     Q516delinsQ   RTRPHTQVL  3_00022      A30:01  0.925992   \n",
       "2     Q516delinsQ   TRPHTQVLQ  3_00022      B27:05  0.009969   \n",
       "3     Q516delinsQ   PHTQVLQNV  3_00022      B38:01  0.212903   \n",
       "4     Q516delinsQ   HTQVLQNVS  3_00022      A30:01  0.007771   \n",
       "...           ...         ...      ...         ...       ...   \n",
       "3986   P24delinsP   LEPAQWDLY  3_00036      B37:01  0.005016   \n",
       "3987   P24delinsP   PAQWDLYRD  3_00036      A30:01  0.000029   \n",
       "3988   P24delinsP   LEPAQWDLY  3_00036      A01:01  0.079631   \n",
       "3989   P24delinsP   PEEWECLEP  3_00036      B50:01  0.295977   \n",
       "3990   P24delinsP   EPAQWDLYR  3_00036      B35:01  0.012099   \n",
       "\n",
       "      percentile_rank_h H2_allele   score_m  percentile_rank_m  score_kbdb  \\\n",
       "0                  1.70        Dq  0.008150              14.00    0.000145   \n",
       "1                  0.01       Qa1  0.804364               0.01    0.176445   \n",
       "2                  4.40       Qa1  0.010245              24.00    0.000825   \n",
       "3                  0.48        Kd  0.025846               2.40    0.001880   \n",
       "4                  9.70       Qa1  0.003275              36.00    0.000603   \n",
       "...                 ...       ...       ...                ...         ...   \n",
       "3986               6.90       Qa1  0.059734               8.40    0.005304   \n",
       "3987              77.00        Kq  0.000031              83.00    0.000004   \n",
       "3988               1.20       Qa1  0.059734               8.40    0.005304   \n",
       "3989               0.32       Qa2  0.002958              24.00    0.000003   \n",
       "3990               3.40        Kq  0.006219              14.00    0.000065   \n",
       "\n",
       "      percentile_rank_kbdb  \n",
       "0                    50.00  \n",
       "1                     0.59  \n",
       "2                    24.00  \n",
       "3                    16.00  \n",
       "4                    25.00  \n",
       "...                    ...  \n",
       "3986                  7.70  \n",
       "3987                 98.00  \n",
       "3988                  7.70  \n",
       "3989                 95.00  \n",
       "3990                 56.00  \n",
       "\n",
       "[3991 rows x 17 columns]"
      ]
     },
     "execution_count": 359,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 360,
   "metadata": {},
   "outputs": [],
   "source": [
    "l_final.to_csv('/Users/kexindong/Documents/GitHub/Output/neoantigen/final_result/m2h_neoantigen_result_v3_.csv',index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
