{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This script runs prime editing guide RNA design for H2M database.  \n",
    "\n",
    "Kexin Dong  \n",
    "\n",
    "May 1, 2024  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kexindong/Documents/GitHub/h2m-public/.conda/lib/python3.10/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import h2m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kexindong/Documents/GitHub/h2m_private/.conda/lib/python3.9/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from pegg import prime\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_human.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_h_ref, path_m_ref = '/Users/kexindong/Documents/GitHub/Database/RefGenome/ncbi-2023-09-12/GCF_000001405.25_GRCh37.p13_genomic.fna.gz', '/Users/kexindong/Documents/GitHub/Database/RefGenome/mouse-2023-09-13/GCF_000001635.27_GRCm39_genomic.fna.gz'\n",
    "chrom_dict_h, i = prime.genome_loader(path_h_ref)\n",
    "chrom_dict_m, i = prime.genome_loader(path_m_ref)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# For Human  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = df_all[df_all['Chromosome']!='M'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all['Reference_Allele'] = [str(x) for x in df_all['Reference_Allele']]\n",
    "df_all['Tumor_Seq_Allele2'] = [str(x) for x in df_all['Tumor_Seq_Allele2']]\n",
    "df_all['Start_Position'] = [int(x) for x in df_all['Start_Position']]\n",
    "df_all['End_Position'] = [int(x) for x in df_all['End_Position']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Efficiency predictions on non-NGG PAMs are un-tested\n",
      "Error in mutant # 4074; Dropped from pegRNA generation\n",
      "Variant Type = TNP| REF = CAG | ALT = AAA\n",
      "GAAGGCAGCCTGCAGCCAGCCCGCACCGCGGGACCCGCGCCCGTGCCCTGGCCACCCGCAGCCTCACCTCGCTCCAGCCGCCCGCCACCTGGAGCGGACCGGCTCCCCGCCGGCTCGGGGCAGCCCGGCAGCCAGCACACACGCACATGCGCGCCGCCGAGAGCGCGGCCCGCACTCACAAGCGCGCCCCCCCCGCGCGCGCGCCCTCCCTCGCTCGCGCACGGCGGCCCGGACGCTCTCGGG\n",
      "GACCCTTTTTGTCTCCTAGCCTGGATCTCCTTGGTGACCGCTCTGCATCTAGTGCTGGGCCTCAGCGCCGTCCTGGGCCTGCTGCTGCTGAGGTGGCAGTTTCCTGCACACTACAGGTACCAGGCCCCCGCCAGGCAGGAGACTGGCGGTGGACCAGGTGGAGCCGAAGGCCTGTAAACAGGCATTCTTGGTTCGCTCTGTGACCCCAGATCTCCGTCCACCGCCCGTGCGCACCTACGGCTT\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>...</th>\n",
       "      <th>indel_size</th>\n",
       "      <th>PEGG2_Score</th>\n",
       "      <th>RF_Score</th>\n",
       "      <th>pegRNA_rank</th>\n",
       "      <th>sensor_wt</th>\n",
       "      <th>sensor_alt</th>\n",
       "      <th>sensor_orientation</th>\n",
       "      <th>sensor_error</th>\n",
       "      <th>contains_polyT_terminator</th>\n",
       "      <th>pegRNA_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>23.027879</td>\n",
       "      <td>16.035646</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTAGGTGTGATC...</td>\n",
       "      <td>TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTGGGTGTGATC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>19.778981</td>\n",
       "      <td>13.473911</td>\n",
       "      <td>2.0</td>\n",
       "      <td>TGACAGCAAGGCTGGCATGGAGGAAGATCACACCTACGAGGTAAGG...</td>\n",
       "      <td>TGACAGCAAGGCTGGCATGGAGGAAGATCACACCCACGAGGTAAGG...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>19.567352</td>\n",
       "      <td>13.473911</td>\n",
       "      <td>3.0</td>\n",
       "      <td>TGACAGCAAGGCTGGCATGGAGGAAGATCACACCTACGAGGTAAGG...</td>\n",
       "      <td>TGACAGCAAGGCTGGCATGGAGGAAGATCACACCCACGAGGTAAGG...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>16.589368</td>\n",
       "      <td>13.266724</td>\n",
       "      <td>4.0</td>\n",
       "      <td>CTCCTTACCTCGTAGGTGTGATCTTCCTCCATGCCAGCCTTGCTGT...</td>\n",
       "      <td>CTCCTTACCTCGTGGGTGTGATCTTCCTCCATGCCAGCCTTGCTGT...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>24.061402</td>\n",
       "      <td>12.680708</td>\n",
       "      <td>5.0</td>\n",
       "      <td>TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTAGGTGTGATC...</td>\n",
       "      <td>TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTGGGTGTGATC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25880</th>\n",
       "      <td>5177</td>\n",
       "      <td>SETD1B</td>\n",
       "      <td>ENSMUST00000163030.9</td>\n",
       "      <td>122242657</td>\n",
       "      <td>122242658</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>AACR-H0001887-M01</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>19.917952</td>\n",
       "      <td>27.489744</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_25880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25881</th>\n",
       "      <td>5177</td>\n",
       "      <td>SETD1B</td>\n",
       "      <td>ENSMUST00000163030.9</td>\n",
       "      <td>122242657</td>\n",
       "      <td>122242658</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>AACR-H0001887-M01</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>19.436025</td>\n",
       "      <td>24.974358</td>\n",
       "      <td>2.0</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_25881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25882</th>\n",
       "      <td>5177</td>\n",
       "      <td>SETD1B</td>\n",
       "      <td>ENSMUST00000163030.9</td>\n",
       "      <td>122242657</td>\n",
       "      <td>122242658</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>AACR-H0001887-M01</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>19.836103</td>\n",
       "      <td>24.506077</td>\n",
       "      <td>3.0</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_25882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25883</th>\n",
       "      <td>5177</td>\n",
       "      <td>SETD1B</td>\n",
       "      <td>ENSMUST00000163030.9</td>\n",
       "      <td>122242657</td>\n",
       "      <td>122242658</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>AACR-H0001887-M01</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>19.388880</td>\n",
       "      <td>24.506077</td>\n",
       "      <td>4.0</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_25883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25884</th>\n",
       "      <td>5177</td>\n",
       "      <td>SETD1B</td>\n",
       "      <td>ENSMUST00000163030.9</td>\n",
       "      <td>122242657</td>\n",
       "      <td>122242658</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>AACR-H0001887-M01</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>20.005823</td>\n",
       "      <td>16.147070</td>\n",
       "      <td>5.0</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_25884</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25885 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       mutation_idx Hugo_Symbol                tx_id_h  Start_Position  \\\n",
       "0                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "1                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "2                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "3                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "4                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "...             ...         ...                    ...             ...   \n",
       "25880          5177      SETD1B   ENSMUST00000163030.9       122242657   \n",
       "25881          5177      SETD1B   ENSMUST00000163030.9       122242657   \n",
       "25882          5177      SETD1B   ENSMUST00000163030.9       122242657   \n",
       "25883          5177      SETD1B   ENSMUST00000163030.9       122242657   \n",
       "25884          5177      SETD1B   ENSMUST00000163030.9       122242657   \n",
       "\n",
       "       End_Position Reference_Allele Tumor_Seq_Allele2 Variant_Type  \\\n",
       "0          62006799                A                 G          SNP   \n",
       "1          62006799                A                 G          SNP   \n",
       "2          62006799                A                 G          SNP   \n",
       "3          62006799                A                 G          SNP   \n",
       "4          62006799                A                 G          SNP   \n",
       "...             ...              ...               ...          ...   \n",
       "25880     122242658                -                 C          INS   \n",
       "25881     122242658                -                 C          INS   \n",
       "25882     122242658                -                 C          INS   \n",
       "25883     122242658                -                 C          INS   \n",
       "25884     122242658                -                 C          INS   \n",
       "\n",
       "                      ID Chromosome  ...  indel_size  PEGG2_Score   RF_Score  \\\n",
       "0      AACR-H0002100-M01         17  ...           0    23.027879  16.035646   \n",
       "1      AACR-H0002100-M01         17  ...           0    19.778981  13.473911   \n",
       "2      AACR-H0002100-M01         17  ...           0    19.567352  13.473911   \n",
       "3      AACR-H0002100-M01         17  ...           0    16.589368  13.266724   \n",
       "4      AACR-H0002100-M01         17  ...           0    24.061402  12.680708   \n",
       "...                  ...        ...  ...         ...          ...        ...   \n",
       "25880  AACR-H0001887-M01         12  ...           1    19.917952  27.489744   \n",
       "25881  AACR-H0001887-M01         12  ...           1    19.436025  24.974358   \n",
       "25882  AACR-H0001887-M01         12  ...           1    19.836103  24.506077   \n",
       "25883  AACR-H0001887-M01         12  ...           1    19.388880  24.506077   \n",
       "25884  AACR-H0001887-M01         12  ...           1    20.005823  16.147070   \n",
       "\n",
       "      pegRNA_rank                                          sensor_wt  \\\n",
       "0             1.0  TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTAGGTGTGATC...   \n",
       "1             2.0  TGACAGCAAGGCTGGCATGGAGGAAGATCACACCTACGAGGTAAGG...   \n",
       "2             3.0  TGACAGCAAGGCTGGCATGGAGGAAGATCACACCTACGAGGTAAGG...   \n",
       "3             4.0  CTCCTTACCTCGTAGGTGTGATCTTCCTCCATGCCAGCCTTGCTGT...   \n",
       "4             5.0  TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTAGGTGTGATC...   \n",
       "...           ...                                                ...   \n",
       "25880         1.0  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...   \n",
       "25881         2.0  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...   \n",
       "25882         3.0  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...   \n",
       "25883         4.0  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...   \n",
       "25884         5.0  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...   \n",
       "\n",
       "                                              sensor_alt  sensor_orientation  \\\n",
       "0      TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTGGGTGTGATC...  reverse-complement   \n",
       "1      TGACAGCAAGGCTGGCATGGAGGAAGATCACACCCACGAGGTAAGG...  reverse-complement   \n",
       "2      TGACAGCAAGGCTGGCATGGAGGAAGATCACACCCACGAGGTAAGG...  reverse-complement   \n",
       "3      CTCCTTACCTCGTGGGTGTGATCTTCCTCCATGCCAGCCTTGCTGT...  reverse-complement   \n",
       "4      TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTGGGTGTGATC...  reverse-complement   \n",
       "...                                                  ...                 ...   \n",
       "25880  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...  reverse-complement   \n",
       "25881  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...  reverse-complement   \n",
       "25882  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...  reverse-complement   \n",
       "25883  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...  reverse-complement   \n",
       "25884  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...  reverse-complement   \n",
       "\n",
       "      sensor_error  contains_polyT_terminator     pegRNA_id  \n",
       "0         No Error                      False      pegRNA_0  \n",
       "1         No Error                      False      pegRNA_1  \n",
       "2         No Error                      False      pegRNA_2  \n",
       "3         No Error                      False      pegRNA_3  \n",
       "4         No Error                      False      pegRNA_4  \n",
       "...            ...                        ...           ...  \n",
       "25880     No Error                      False  pegRNA_25880  \n",
       "25881     No Error                      False  pegRNA_25881  \n",
       "25882     No Error                      False  pegRNA_25882  \n",
       "25883     No Error                      False  pegRNA_25883  \n",
       "25884     No Error                      False  pegRNA_25884  \n",
       "\n",
       "[25885 rows x 46 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pe_h = prime.run(df_all, 'cBioPortal', chrom_dict=chrom_dict_h, PAM='NGN',rankby='RF_Score', pegRNAs_per_mut=5, RTT_lengths=[5, 10, 15, 25, 30],PBS_lengths=[8, 10, 13, 15], min_RHA_size=1, RE_sites=None, polyT_threshold=4,proto_size=19, context_size=120, before_proto_context=5, sensor_length=60,sensor_orientation='reverse-complement', sensor=True)\n",
    "df_pe_h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>...</th>\n",
       "      <th>indel_size</th>\n",
       "      <th>PEGG2_Score</th>\n",
       "      <th>RF_Score</th>\n",
       "      <th>pegRNA_rank</th>\n",
       "      <th>sensor_wt</th>\n",
       "      <th>sensor_alt</th>\n",
       "      <th>sensor_orientation</th>\n",
       "      <th>sensor_error</th>\n",
       "      <th>contains_polyT_terminator</th>\n",
       "      <th>pegRNA_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>23.027879</td>\n",
       "      <td>16.035646</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTAGGTGTGATC...</td>\n",
       "      <td>TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTGGGTGTGATC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>19.778981</td>\n",
       "      <td>13.473911</td>\n",
       "      <td>2.0</td>\n",
       "      <td>TGACAGCAAGGCTGGCATGGAGGAAGATCACACCTACGAGGTAAGG...</td>\n",
       "      <td>TGACAGCAAGGCTGGCATGGAGGAAGATCACACCCACGAGGTAAGG...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>19.567352</td>\n",
       "      <td>13.473911</td>\n",
       "      <td>3.0</td>\n",
       "      <td>TGACAGCAAGGCTGGCATGGAGGAAGATCACACCTACGAGGTAAGG...</td>\n",
       "      <td>TGACAGCAAGGCTGGCATGGAGGAAGATCACACCCACGAGGTAAGG...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>16.589368</td>\n",
       "      <td>13.266724</td>\n",
       "      <td>4.0</td>\n",
       "      <td>CTCCTTACCTCGTAGGTGTGATCTTCCTCCATGCCAGCCTTGCTGT...</td>\n",
       "      <td>CTCCTTACCTCGTGGGTGTGATCTTCCTCCATGCCAGCCTTGCTGT...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>24.061402</td>\n",
       "      <td>12.680708</td>\n",
       "      <td>5.0</td>\n",
       "      <td>TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTAGGTGTGATC...</td>\n",
       "      <td>TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTGGGTGTGATC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25880</th>\n",
       "      <td>5177</td>\n",
       "      <td>SETD1B</td>\n",
       "      <td>ENSMUST00000163030.9</td>\n",
       "      <td>122242657</td>\n",
       "      <td>122242658</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>AACR-H0001887-M01</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>19.917952</td>\n",
       "      <td>27.489744</td>\n",
       "      <td>1.0</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_25880</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25881</th>\n",
       "      <td>5177</td>\n",
       "      <td>SETD1B</td>\n",
       "      <td>ENSMUST00000163030.9</td>\n",
       "      <td>122242657</td>\n",
       "      <td>122242658</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>AACR-H0001887-M01</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>19.436025</td>\n",
       "      <td>24.974358</td>\n",
       "      <td>2.0</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_25881</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25882</th>\n",
       "      <td>5177</td>\n",
       "      <td>SETD1B</td>\n",
       "      <td>ENSMUST00000163030.9</td>\n",
       "      <td>122242657</td>\n",
       "      <td>122242658</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>AACR-H0001887-M01</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>19.836103</td>\n",
       "      <td>24.506077</td>\n",
       "      <td>3.0</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_25882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25883</th>\n",
       "      <td>5177</td>\n",
       "      <td>SETD1B</td>\n",
       "      <td>ENSMUST00000163030.9</td>\n",
       "      <td>122242657</td>\n",
       "      <td>122242658</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>AACR-H0001887-M01</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>19.388880</td>\n",
       "      <td>24.506077</td>\n",
       "      <td>4.0</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_25883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25884</th>\n",
       "      <td>5177</td>\n",
       "      <td>SETD1B</td>\n",
       "      <td>ENSMUST00000163030.9</td>\n",
       "      <td>122242657</td>\n",
       "      <td>122242658</td>\n",
       "      <td>-</td>\n",
       "      <td>C</td>\n",
       "      <td>INS</td>\n",
       "      <td>AACR-H0001887-M01</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>20.005823</td>\n",
       "      <td>16.147070</td>\n",
       "      <td>5.0</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_25884</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25885 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       mutation_idx Hugo_Symbol                tx_id_h  Start_Position  \\\n",
       "0                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "1                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "2                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "3                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "4                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "...             ...         ...                    ...             ...   \n",
       "25880          5177      SETD1B   ENSMUST00000163030.9       122242657   \n",
       "25881          5177      SETD1B   ENSMUST00000163030.9       122242657   \n",
       "25882          5177      SETD1B   ENSMUST00000163030.9       122242657   \n",
       "25883          5177      SETD1B   ENSMUST00000163030.9       122242657   \n",
       "25884          5177      SETD1B   ENSMUST00000163030.9       122242657   \n",
       "\n",
       "       End_Position Reference_Allele Tumor_Seq_Allele2 Variant_Type  \\\n",
       "0          62006799                A                 G          SNP   \n",
       "1          62006799                A                 G          SNP   \n",
       "2          62006799                A                 G          SNP   \n",
       "3          62006799                A                 G          SNP   \n",
       "4          62006799                A                 G          SNP   \n",
       "...             ...              ...               ...          ...   \n",
       "25880     122242658                -                 C          INS   \n",
       "25881     122242658                -                 C          INS   \n",
       "25882     122242658                -                 C          INS   \n",
       "25883     122242658                -                 C          INS   \n",
       "25884     122242658                -                 C          INS   \n",
       "\n",
       "                      ID Chromosome  ...  indel_size  PEGG2_Score   RF_Score  \\\n",
       "0      AACR-H0002100-M01         17  ...           0    23.027879  16.035646   \n",
       "1      AACR-H0002100-M01         17  ...           0    19.778981  13.473911   \n",
       "2      AACR-H0002100-M01         17  ...           0    19.567352  13.473911   \n",
       "3      AACR-H0002100-M01         17  ...           0    16.589368  13.266724   \n",
       "4      AACR-H0002100-M01         17  ...           0    24.061402  12.680708   \n",
       "...                  ...        ...  ...         ...          ...        ...   \n",
       "25880  AACR-H0001887-M01         12  ...           1    19.917952  27.489744   \n",
       "25881  AACR-H0001887-M01         12  ...           1    19.436025  24.974358   \n",
       "25882  AACR-H0001887-M01         12  ...           1    19.836103  24.506077   \n",
       "25883  AACR-H0001887-M01         12  ...           1    19.388880  24.506077   \n",
       "25884  AACR-H0001887-M01         12  ...           1    20.005823  16.147070   \n",
       "\n",
       "      pegRNA_rank                                          sensor_wt  \\\n",
       "0             1.0  TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTAGGTGTGATC...   \n",
       "1             2.0  TGACAGCAAGGCTGGCATGGAGGAAGATCACACCTACGAGGTAAGG...   \n",
       "2             3.0  TGACAGCAAGGCTGGCATGGAGGAAGATCACACCTACGAGGTAAGG...   \n",
       "3             4.0  CTCCTTACCTCGTAGGTGTGATCTTCCTCCATGCCAGCCTTGCTGT...   \n",
       "4             5.0  TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTAGGTGTGATC...   \n",
       "...           ...                                                ...   \n",
       "25880         1.0  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...   \n",
       "25881         2.0  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...   \n",
       "25882         3.0  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...   \n",
       "25883         4.0  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...   \n",
       "25884         5.0  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...   \n",
       "\n",
       "                                              sensor_alt  sensor_orientation  \\\n",
       "0      TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTGGGTGTGATC...  reverse-complement   \n",
       "1      TGACAGCAAGGCTGGCATGGAGGAAGATCACACCCACGAGGTAAGG...  reverse-complement   \n",
       "2      TGACAGCAAGGCTGGCATGGAGGAAGATCACACCCACGAGGTAAGG...  reverse-complement   \n",
       "3      CTCCTTACCTCGTGGGTGTGATCTTCCTCCATGCCAGCCTTGCTGT...  reverse-complement   \n",
       "4      TCAGAGCTGCTGGGCCTGCCCCTCTCCTTACCTCGTGGGTGTGATC...  reverse-complement   \n",
       "...                                                  ...                 ...   \n",
       "25880  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...  reverse-complement   \n",
       "25881  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...  reverse-complement   \n",
       "25882  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...  reverse-complement   \n",
       "25883  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...  reverse-complement   \n",
       "25884  CTGGTTTTCAGGTTGGGTTAACGGCATGGAGAACAGTCACCCCCCC...  reverse-complement   \n",
       "\n",
       "      sensor_error  contains_polyT_terminator     pegRNA_id  \n",
       "0         No Error                      False      pegRNA_0  \n",
       "1         No Error                      False      pegRNA_1  \n",
       "2         No Error                      False      pegRNA_2  \n",
       "3         No Error                      False      pegRNA_3  \n",
       "4         No Error                      False      pegRNA_4  \n",
       "...            ...                        ...           ...  \n",
       "25880     No Error                      False  pegRNA_25880  \n",
       "25881     No Error                      False  pegRNA_25881  \n",
       "25882     No Error                      False  pegRNA_25882  \n",
       "25883     No Error                      False  pegRNA_25883  \n",
       "25884     No Error                      False  pegRNA_25884  \n",
       "\n",
       "[25885 rows x 46 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_pe_h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5177"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_pe_h['ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hugo_Symbol\n",
       "TP53      588\n",
       "APC       118\n",
       "PIK3CA    100\n",
       "PTEN       95\n",
       "ARID1A     85\n",
       "         ... \n",
       "FGF14       1\n",
       "FGA         1\n",
       "ERCC1       1\n",
       "ERRFI1      1\n",
       "SETD1B      1\n",
       "Name: count, Length: 614, dtype: int64"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all['Hugo_Symbol'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pe_h.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_human_result.csv',index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
