{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This script runs prime editing guide RNA design for H2M database.  \n",
    "\n",
    "Kexin Dong  \n",
    "\n",
    "May 1, 2024  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kexindong/Documents/GitHub/h2m-public/.conda/lib/python3.10/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import bioh2m as h2m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kexindong/Documents/GitHub/h2m-public/.conda/lib/python3.10/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from pegg import prime\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_human.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4798"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_all['ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_h_ref, path_m_ref = '/Users/kexindong/Documents/GitHub/Database/RefGenome/ncbi-2023-09-12/GCF_000001405.25_GRCh37.p13_genomic.fna.gz', '/Users/kexindong/Documents/GitHub/Database/RefGenome/mouse-2023-09-13/GCF_000001635.27_GRCm39_genomic.fna.gz'\n",
    "chrom_dict_h, i = prime.genome_loader(path_h_ref)\n",
    "chrom_dict_m, i = prime.genome_loader(path_m_ref)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# For Human  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = df_all[df_all['Chromosome']!='M'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all['Reference_Allele'] = [str(x) for x in df_all['Reference_Allele']]\n",
    "df_all['Tumor_Seq_Allele2'] = [str(x) for x in df_all['Tumor_Seq_Allele2']]\n",
    "df_all['Start_Position'] = [int(x) for x in df_all['Start_Position']]\n",
    "df_all['End_Position'] = [int(x) for x in df_all['End_Position']]\n",
    "df_pe_h = prime.run(df_all, 'cBioPortal', chrom_dict=chrom_dict_h, PAM='NGN',rankby='RF_Score', pegRNAs_per_mut=5, RTT_lengths=[5, 10, 15, 25, 30],PBS_lengths=[8, 10, 13, 15], min_RHA_size=1, RE_sites=None, polyT_threshold=4,proto_size=19, context_size=120, before_proto_context=5, sensor_length=60,sensor_orientation='reverse-complement', sensor=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pe = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_result_all.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_be = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/BEGG/be_result_all.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pe_h = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_human_result.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['mutation_idx', 'Hugo_Symbol', 'Start_Position', 'End_Position',\n",
       "       'Reference_Allele', 'Tumor_Seq_Allele2', 'Variant_Type', 'ID',\n",
       "       'Chromosome', 'seq_start', 'seq_end', 'wt_w_context', 'alt_w_context',\n",
       "       'left_context', 'right_context', 'REF', 'ALT', 'PAM_start', 'PAM_end',\n",
       "       'PAM', 'PAM_strand', 'Protospacer_30', 'Protospacer',\n",
       "       'Protospacer_Location', 'Ideal_Edit_Window', 'Editor',\n",
       "       'OnTarget_Azimuth_Score', 'sensor_wt', 'sensor_alt',\n",
       "       'sensor_orientation', 'sensor_error', 'contains_polyT_terminator'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_be.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_be = df_be.drop_duplicates(subset=list(df_be.columns)[1:]).reset_index(drop=True)\n",
    "df_be.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/BEGG/be_result_all.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pe = df_pe.drop_duplicates(subset=list(df_pe.columns)[1:]).reset_index(drop=True)\n",
    "df_pe.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_result_all.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4944"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_pe[[len(x.split('-'))==2 for x in df_pe['ID']]]['ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>seq_start</th>\n",
       "      <th>...</th>\n",
       "      <th>Protospacer</th>\n",
       "      <th>Protospacer_Location</th>\n",
       "      <th>Ideal_Edit_Window</th>\n",
       "      <th>Editor</th>\n",
       "      <th>OnTarget_Azimuth_Score</th>\n",
       "      <th>sensor_wt</th>\n",
       "      <th>sensor_alt</th>\n",
       "      <th>sensor_orientation</th>\n",
       "      <th>sensor_error</th>\n",
       "      <th>contains_polyT_terminator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100</td>\n",
       "      <td>17</td>\n",
       "      <td>62006679</td>\n",
       "      <td>...</td>\n",
       "      <td>GCCCCTCTCCTTACCTCGTA</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>44.0</td>\n",
       "      <td>GAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCAGGCC</td>\n",
       "      <td>GAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCAGGCC</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100</td>\n",
       "      <td>17</td>\n",
       "      <td>62006679</td>\n",
       "      <td>...</td>\n",
       "      <td>GCCTCTCCTTACCTCGTAGG</td>\n",
       "      <td>18</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>54.0</td>\n",
       "      <td>TGGAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCAGG</td>\n",
       "      <td>TGGAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCAGG</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100</td>\n",
       "      <td>17</td>\n",
       "      <td>62006679</td>\n",
       "      <td>...</td>\n",
       "      <td>GTCTCCTTACCTCGTAGGTG</td>\n",
       "      <td>16</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>59.0</td>\n",
       "      <td>CATGGAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCA</td>\n",
       "      <td>CATGGAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCA</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100</td>\n",
       "      <td>17</td>\n",
       "      <td>62006679</td>\n",
       "      <td>...</td>\n",
       "      <td>GTAGGTGTGATCTTCCTCCA</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>59.0</td>\n",
       "      <td>ACAGCAAGGCTGGCATGGAGGAAGATCACACCTACGAGGT</td>\n",
       "      <td>ACAGCAAGGCTGGCATGGAGGAAGATCACACCCACGAGGT</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>62006798</td>\n",
       "      <td>62006798</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002993</td>\n",
       "      <td>17</td>\n",
       "      <td>62006678</td>\n",
       "      <td>...</td>\n",
       "      <td>GACACCTACGAGGTAAGGAG</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>ABE</td>\n",
       "      <td>66.0</td>\n",
       "      <td>GCTGGGCCTGCCCCTCTCCTTACCTCGTAGGTGTGATCTT</td>\n",
       "      <td>GCTGGGCCTGCCCCTCTCCTTACCTCGCAGGTGTGATCTT</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24675</th>\n",
       "      <td>1896</td>\n",
       "      <td>Xpo1</td>\n",
       "      <td>23235114</td>\n",
       "      <td>23235114</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005028-M01-V01</td>\n",
       "      <td>11</td>\n",
       "      <td>23234994</td>\n",
       "      <td>...</td>\n",
       "      <td>GAATAGCTCAAAAATGCCGT</td>\n",
       "      <td>14</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>59.0</td>\n",
       "      <td>CTGAACAAAATGCCTACGGCATTTTTGAGCTATTTTAATA</td>\n",
       "      <td>CTGAACAAAATGCCTACGGCACTTTTGAGCTATTTTAATA</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24676</th>\n",
       "      <td>1896</td>\n",
       "      <td>Xpo1</td>\n",
       "      <td>23235114</td>\n",
       "      <td>23235114</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005028-M01-V01</td>\n",
       "      <td>11</td>\n",
       "      <td>23234994</td>\n",
       "      <td>...</td>\n",
       "      <td>GATAGCTCAAAAATGCCGTA</td>\n",
       "      <td>13</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>60.0</td>\n",
       "      <td>CCTGAACAAAATGCCTACGGCATTTTTGAGCTATTTTAAT</td>\n",
       "      <td>CCTGAACAAAATGCCTACGGCACTTTTGAGCTATTTTAAT</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24677</th>\n",
       "      <td>1896</td>\n",
       "      <td>Xpo1</td>\n",
       "      <td>23235114</td>\n",
       "      <td>23235114</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005028-M01-V01</td>\n",
       "      <td>11</td>\n",
       "      <td>23234994</td>\n",
       "      <td>...</td>\n",
       "      <td>GAAAAATGCCGTAGGCATTT</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>ABE</td>\n",
       "      <td>24.0</td>\n",
       "      <td>ACTTGAACCTGAACAAAATGCCTACGGCATTTTTGAGCTA</td>\n",
       "      <td>ACTTGAACCTGAACAAAATGCCTACGGCACTTTTGAGCTA</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24678</th>\n",
       "      <td>1896</td>\n",
       "      <td>Xpo1</td>\n",
       "      <td>23235114</td>\n",
       "      <td>23235114</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005028-M01-V01</td>\n",
       "      <td>11</td>\n",
       "      <td>23234994</td>\n",
       "      <td>...</td>\n",
       "      <td>GTGCCGTAGGCATTTTGTTC</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>33.0</td>\n",
       "      <td>CTCCAACTTGAACCTGAACAAAATGCCTACGGCATTTTTG</td>\n",
       "      <td>CTCCAACTTGAACCTGAACAAAATGCCTACGGCACTTTTG</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24679</th>\n",
       "      <td>1897</td>\n",
       "      <td>Bcl9</td>\n",
       "      <td>97117238</td>\n",
       "      <td>97117238</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005030-M01-V01</td>\n",
       "      <td>3</td>\n",
       "      <td>97117118</td>\n",
       "      <td>...</td>\n",
       "      <td>GGACCACCTGCTCCTGTTTC</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>CBE</td>\n",
       "      <td>26.0</td>\n",
       "      <td>TGAAGAGAAGAGGCGGAAACAGGAGCAGGTGGTCGTCCAG</td>\n",
       "      <td>TGAAGAGAAGAGGCGAAAACAGGAGCAGGTGGTCGTCCAG</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24680 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       mutation_idx Hugo_Symbol  Start_Position  End_Position  \\\n",
       "0                 0       CD79B        62006799      62006799   \n",
       "1                 0       CD79B        62006799      62006799   \n",
       "2                 0       CD79B        62006799      62006799   \n",
       "3                 0       CD79B        62006799      62006799   \n",
       "4                 1       CD79B        62006798      62006798   \n",
       "...             ...         ...             ...           ...   \n",
       "24675          1896        Xpo1        23235114      23235114   \n",
       "24676          1896        Xpo1        23235114      23235114   \n",
       "24677          1896        Xpo1        23235114      23235114   \n",
       "24678          1896        Xpo1        23235114      23235114   \n",
       "24679          1897        Bcl9        97117238      97117238   \n",
       "\n",
       "      Reference_Allele Tumor_Seq_Allele2 Variant_Type                     ID  \\\n",
       "0                    A                 G          SNP          AACR-H0002100   \n",
       "1                    A                 G          SNP          AACR-H0002100   \n",
       "2                    A                 G          SNP          AACR-H0002100   \n",
       "3                    A                 G          SNP          AACR-H0002100   \n",
       "4                    T                 C          SNP          AACR-H0002993   \n",
       "...                ...               ...          ...                    ...   \n",
       "24675                A                 G          SNP  AACR-H0005028-M01-V01   \n",
       "24676                A                 G          SNP  AACR-H0005028-M01-V01   \n",
       "24677                A                 G          SNP  AACR-H0005028-M01-V01   \n",
       "24678                A                 G          SNP  AACR-H0005028-M01-V01   \n",
       "24679                C                 T          SNP  AACR-H0005030-M01-V01   \n",
       "\n",
       "      Chromosome  seq_start  ...           Protospacer Protospacer_Location  \\\n",
       "0             17   62006679  ...  GCCCCTCTCCTTACCTCGTA                   20   \n",
       "1             17   62006679  ...  GCCTCTCCTTACCTCGTAGG                   18   \n",
       "2             17   62006679  ...  GTCTCCTTACCTCGTAGGTG                   16   \n",
       "3             17   62006679  ...  GTAGGTGTGATCTTCCTCCA                    3   \n",
       "4             17   62006678  ...  GACACCTACGAGGTAAGGAG                    8   \n",
       "...          ...        ...  ...                   ...                  ...   \n",
       "24675         11   23234994  ...  GAATAGCTCAAAAATGCCGT                   14   \n",
       "24676         11   23234994  ...  GATAGCTCAAAAATGCCGTA                   13   \n",
       "24677         11   23234994  ...  GAAAAATGCCGTAGGCATTT                    6   \n",
       "24678         11   23234994  ...  GTGCCGTAGGCATTTTGTTC                    1   \n",
       "24679          3   97117118  ...  GGACCACCTGCTCCTGTTTC                   20   \n",
       "\n",
       "      Ideal_Edit_Window Editor OnTarget_Azimuth_Score  \\\n",
       "0                 False    ABE                   44.0   \n",
       "1                 False    ABE                   54.0   \n",
       "2                 False    ABE                   59.0   \n",
       "3                 False    ABE                   59.0   \n",
       "4                  True    ABE                   66.0   \n",
       "...                 ...    ...                    ...   \n",
       "24675             False    ABE                   59.0   \n",
       "24676             False    ABE                   60.0   \n",
       "24677              True    ABE                   24.0   \n",
       "24678             False    ABE                   33.0   \n",
       "24679             False    CBE                   26.0   \n",
       "\n",
       "                                      sensor_wt  \\\n",
       "0      GAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCAGGCC   \n",
       "1      TGGAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCAGG   \n",
       "2      CATGGAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCA   \n",
       "3      ACAGCAAGGCTGGCATGGAGGAAGATCACACCTACGAGGT   \n",
       "4      GCTGGGCCTGCCCCTCTCCTTACCTCGTAGGTGTGATCTT   \n",
       "...                                         ...   \n",
       "24675  CTGAACAAAATGCCTACGGCATTTTTGAGCTATTTTAATA   \n",
       "24676  CCTGAACAAAATGCCTACGGCATTTTTGAGCTATTTTAAT   \n",
       "24677  ACTTGAACCTGAACAAAATGCCTACGGCATTTTTGAGCTA   \n",
       "24678  CTCCAACTTGAACCTGAACAAAATGCCTACGGCATTTTTG   \n",
       "24679  TGAAGAGAAGAGGCGGAAACAGGAGCAGGTGGTCGTCCAG   \n",
       "\n",
       "                                     sensor_alt  sensor_orientation  \\\n",
       "0      GAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCAGGCC  reverse-complement   \n",
       "1      TGGAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCAGG  reverse-complement   \n",
       "2      CATGGAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCA  reverse-complement   \n",
       "3      ACAGCAAGGCTGGCATGGAGGAAGATCACACCCACGAGGT  reverse-complement   \n",
       "4      GCTGGGCCTGCCCCTCTCCTTACCTCGCAGGTGTGATCTT  reverse-complement   \n",
       "...                                         ...                 ...   \n",
       "24675  CTGAACAAAATGCCTACGGCACTTTTGAGCTATTTTAATA  reverse-complement   \n",
       "24676  CCTGAACAAAATGCCTACGGCACTTTTGAGCTATTTTAAT  reverse-complement   \n",
       "24677  ACTTGAACCTGAACAAAATGCCTACGGCACTTTTGAGCTA  reverse-complement   \n",
       "24678  CTCCAACTTGAACCTGAACAAAATGCCTACGGCACTTTTG  reverse-complement   \n",
       "24679  TGAAGAGAAGAGGCGAAAACAGGAGCAGGTGGTCGTCCAG  reverse-complement   \n",
       "\n",
       "       sensor_error contains_polyT_terminator  \n",
       "0          No Error                     False  \n",
       "1          No Error                     False  \n",
       "2          No Error                     False  \n",
       "3          No Error                     False  \n",
       "4          No Error                     False  \n",
       "...             ...                       ...  \n",
       "24675      No Error                     False  \n",
       "24676      No Error                     False  \n",
       "24677      No Error                     False  \n",
       "24678      No Error                      True  \n",
       "24679      No Error                     False  \n",
       "\n",
       "[24680 rows x 32 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_be"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pe_h.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_human_result.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# For Mouse  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_mouse.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "dict_of_type = {'TNP':'ONP'}\n",
    "df_all['Variant_Type'] = [dict_of_type.get(x,x) for x in df_all['Variant_Type']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = df_all[df_all['Chromosome']!='M'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Warning: Efficiency predictions on non-NGG PAMs are un-tested\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>...</th>\n",
       "      <th>indel_size</th>\n",
       "      <th>PEGG2_Score</th>\n",
       "      <th>RF_Score</th>\n",
       "      <th>pegRNA_rank</th>\n",
       "      <th>sensor_wt</th>\n",
       "      <th>sensor_alt</th>\n",
       "      <th>sensor_orientation</th>\n",
       "      <th>sensor_error</th>\n",
       "      <th>contains_polyT_terminator</th>\n",
       "      <th>pegRNA_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "      <td>145192497</td>\n",
       "      <td>145192497</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000001-M01-V01</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>20.660669</td>\n",
       "      <td>30.350048</td>\n",
       "      <td>1.0</td>\n",
       "      <td>AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGGTGGCGTAGG...</td>\n",
       "      <td>AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGATGGCGTAGG...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "      <td>145192497</td>\n",
       "      <td>145192497</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000001-M01-V01</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>20.310904</td>\n",
       "      <td>27.444430</td>\n",
       "      <td>2.0</td>\n",
       "      <td>AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGGTGGCGTAGG...</td>\n",
       "      <td>AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGATGGCGTAGG...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "      <td>145192497</td>\n",
       "      <td>145192497</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000001-M01-V01</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>21.117170</td>\n",
       "      <td>24.007503</td>\n",
       "      <td>3.0</td>\n",
       "      <td>AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGGTGGCGTAGG...</td>\n",
       "      <td>AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGATGGCGTAGG...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "      <td>145192497</td>\n",
       "      <td>145192497</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000001-M01-V01</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>20.669312</td>\n",
       "      <td>23.256523</td>\n",
       "      <td>4.0</td>\n",
       "      <td>AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGGTGGCGTAGG...</td>\n",
       "      <td>AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGATGGCGTAGG...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "      <td>145192497</td>\n",
       "      <td>145192497</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000001-M01-V01</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>20.302260</td>\n",
       "      <td>22.063635</td>\n",
       "      <td>5.0</td>\n",
       "      <td>AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGGTGGCGTAGG...</td>\n",
       "      <td>AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGATGGCGTAGG...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23530</th>\n",
       "      <td>4708</td>\n",
       "      <td>Bcl9</td>\n",
       "      <td>ENSMUST00000046521.14</td>\n",
       "      <td>97117238</td>\n",
       "      <td>97117238</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005030-M01-V01</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>17.825962</td>\n",
       "      <td>27.045849</td>\n",
       "      <td>1.0</td>\n",
       "      <td>GAGCACTGCTGGACGACCACCTGCTCCTGTTTCCGCCTCTTCTCTT...</td>\n",
       "      <td>GAGCACTGCTGGACGACCACCTGCTCCTGTTTTCGCCTCTTCTCTT...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>True</td>\n",
       "      <td>pegRNA_23530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23531</th>\n",
       "      <td>4708</td>\n",
       "      <td>Bcl9</td>\n",
       "      <td>ENSMUST00000046521.14</td>\n",
       "      <td>97117238</td>\n",
       "      <td>97117238</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005030-M01-V01</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>17.151807</td>\n",
       "      <td>21.992069</td>\n",
       "      <td>2.0</td>\n",
       "      <td>GAGTTTTATGAAGAGAAGAGGCGGAAACAGGAGCAGGTGGTCGTCC...</td>\n",
       "      <td>GAGTTTTATGAAGAGAAGAGGCGAAAACAGGAGCAGGTGGTCGTCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_23531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23532</th>\n",
       "      <td>4708</td>\n",
       "      <td>Bcl9</td>\n",
       "      <td>ENSMUST00000046521.14</td>\n",
       "      <td>97117238</td>\n",
       "      <td>97117238</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005030-M01-V01</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>17.394716</td>\n",
       "      <td>21.161375</td>\n",
       "      <td>3.0</td>\n",
       "      <td>GAGCACTGCTGGACGACCACCTGCTCCTGTTTCCGCCTCTTCTCTT...</td>\n",
       "      <td>GAGCACTGCTGGACGACCACCTGCTCCTGTTTTCGCCTCTTCTCTT...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>True</td>\n",
       "      <td>pegRNA_23532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23533</th>\n",
       "      <td>4708</td>\n",
       "      <td>Bcl9</td>\n",
       "      <td>ENSMUST00000046521.14</td>\n",
       "      <td>97117238</td>\n",
       "      <td>97117238</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005030-M01-V01</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>17.173775</td>\n",
       "      <td>19.967213</td>\n",
       "      <td>4.0</td>\n",
       "      <td>GAGTTTTATGAAGAGAAGAGGCGGAAACAGGAGCAGGTGGTCGTCC...</td>\n",
       "      <td>GAGTTTTATGAAGAGAAGAGGCGAAAACAGGAGCAGGTGGTCGTCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_23533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23534</th>\n",
       "      <td>4708</td>\n",
       "      <td>Bcl9</td>\n",
       "      <td>ENSMUST00000046521.14</td>\n",
       "      <td>97117238</td>\n",
       "      <td>97117238</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005030-M01-V01</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>17.160451</td>\n",
       "      <td>19.123854</td>\n",
       "      <td>5.0</td>\n",
       "      <td>GAGTTTTATGAAGAGAAGAGGCGGAAACAGGAGCAGGTGGTCGTCC...</td>\n",
       "      <td>GAGTTTTATGAAGAGAAGAGGCGAAAACAGGAGCAGGTGGTCGTCC...</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "      <td>pegRNA_23534</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23535 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       mutation_idx Hugo_Symbol                tx_id_m  Start_Position  \\\n",
       "0                 0        Kras   ENSMUST00000111710.8       145192497   \n",
       "1                 0        Kras   ENSMUST00000111710.8       145192497   \n",
       "2                 0        Kras   ENSMUST00000111710.8       145192497   \n",
       "3                 0        Kras   ENSMUST00000111710.8       145192497   \n",
       "4                 0        Kras   ENSMUST00000111710.8       145192497   \n",
       "...             ...         ...                    ...             ...   \n",
       "23530          4708        Bcl9  ENSMUST00000046521.14        97117238   \n",
       "23531          4708        Bcl9  ENSMUST00000046521.14        97117238   \n",
       "23532          4708        Bcl9  ENSMUST00000046521.14        97117238   \n",
       "23533          4708        Bcl9  ENSMUST00000046521.14        97117238   \n",
       "23534          4708        Bcl9  ENSMUST00000046521.14        97117238   \n",
       "\n",
       "       End_Position Reference_Allele Tumor_Seq_Allele2 Variant_Type  \\\n",
       "0         145192497                C                 T          SNP   \n",
       "1         145192497                C                 T          SNP   \n",
       "2         145192497                C                 T          SNP   \n",
       "3         145192497                C                 T          SNP   \n",
       "4         145192497                C                 T          SNP   \n",
       "...             ...              ...               ...          ...   \n",
       "23530      97117238                C                 T          SNP   \n",
       "23531      97117238                C                 T          SNP   \n",
       "23532      97117238                C                 T          SNP   \n",
       "23533      97117238                C                 T          SNP   \n",
       "23534      97117238                C                 T          SNP   \n",
       "\n",
       "                          ID Chromosome  ...  indel_size  PEGG2_Score  \\\n",
       "0      AACR-H0000001-M01-V01          6  ...           0    20.660669   \n",
       "1      AACR-H0000001-M01-V01          6  ...           0    20.310904   \n",
       "2      AACR-H0000001-M01-V01          6  ...           0    21.117170   \n",
       "3      AACR-H0000001-M01-V01          6  ...           0    20.669312   \n",
       "4      AACR-H0000001-M01-V01          6  ...           0    20.302260   \n",
       "...                      ...        ...  ...         ...          ...   \n",
       "23530  AACR-H0005030-M01-V01          3  ...           0    17.825962   \n",
       "23531  AACR-H0005030-M01-V01          3  ...           0    17.151807   \n",
       "23532  AACR-H0005030-M01-V01          3  ...           0    17.394716   \n",
       "23533  AACR-H0005030-M01-V01          3  ...           0    17.173775   \n",
       "23534  AACR-H0005030-M01-V01          3  ...           0    17.160451   \n",
       "\n",
       "        RF_Score pegRNA_rank  \\\n",
       "0      30.350048         1.0   \n",
       "1      27.444430         2.0   \n",
       "2      24.007503         3.0   \n",
       "3      23.256523         4.0   \n",
       "4      22.063635         5.0   \n",
       "...          ...         ...   \n",
       "23530  27.045849         1.0   \n",
       "23531  21.992069         2.0   \n",
       "23532  21.161375         3.0   \n",
       "23533  19.967213         4.0   \n",
       "23534  19.123854         5.0   \n",
       "\n",
       "                                               sensor_wt  \\\n",
       "0      AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGGTGGCGTAGG...   \n",
       "1      AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGGTGGCGTAGG...   \n",
       "2      AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGGTGGCGTAGG...   \n",
       "3      AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGGTGGCGTAGG...   \n",
       "4      AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGGTGGCGTAGG...   \n",
       "...                                                  ...   \n",
       "23530  GAGCACTGCTGGACGACCACCTGCTCCTGTTTCCGCCTCTTCTCTT...   \n",
       "23531  GAGTTTTATGAAGAGAAGAGGCGGAAACAGGAGCAGGTGGTCGTCC...   \n",
       "23532  GAGCACTGCTGGACGACCACCTGCTCCTGTTTCCGCCTCTTCTCTT...   \n",
       "23533  GAGTTTTATGAAGAGAAGAGGCGGAAACAGGAGCAGGTGGTCGTCC...   \n",
       "23534  GAGTTTTATGAAGAGAAGAGGCGGAAACAGGAGCAGGTGGTCGTCC...   \n",
       "\n",
       "                                              sensor_alt  sensor_orientation  \\\n",
       "0      AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGATGGCGTAGG...  reverse-complement   \n",
       "1      AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGATGGCGTAGG...  reverse-complement   \n",
       "2      AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGATGGCGTAGG...  reverse-complement   \n",
       "3      AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGATGGCGTAGG...  reverse-complement   \n",
       "4      AAATGACTGAGTATAAACTTGTGGTGGTTGGAGCTGATGGCGTAGG...  reverse-complement   \n",
       "...                                                  ...                 ...   \n",
       "23530  GAGCACTGCTGGACGACCACCTGCTCCTGTTTTCGCCTCTTCTCTT...  reverse-complement   \n",
       "23531  GAGTTTTATGAAGAGAAGAGGCGAAAACAGGAGCAGGTGGTCGTCC...  reverse-complement   \n",
       "23532  GAGCACTGCTGGACGACCACCTGCTCCTGTTTTCGCCTCTTCTCTT...  reverse-complement   \n",
       "23533  GAGTTTTATGAAGAGAAGAGGCGAAAACAGGAGCAGGTGGTCGTCC...  reverse-complement   \n",
       "23534  GAGTTTTATGAAGAGAAGAGGCGAAAACAGGAGCAGGTGGTCGTCC...  reverse-complement   \n",
       "\n",
       "      sensor_error  contains_polyT_terminator     pegRNA_id  \n",
       "0         No Error                      False      pegRNA_0  \n",
       "1         No Error                      False      pegRNA_1  \n",
       "2         No Error                      False      pegRNA_2  \n",
       "3         No Error                      False      pegRNA_3  \n",
       "4         No Error                      False      pegRNA_4  \n",
       "...            ...                        ...           ...  \n",
       "23530     No Error                       True  pegRNA_23530  \n",
       "23531     No Error                      False  pegRNA_23531  \n",
       "23532     No Error                       True  pegRNA_23532  \n",
       "23533     No Error                      False  pegRNA_23533  \n",
       "23534     No Error                      False  pegRNA_23534  \n",
       "\n",
       "[23535 rows x 46 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all['Reference_Allele'] = [str(x) for x in df_all['Reference_Allele']]\n",
    "df_all['Tumor_Seq_Allele2'] = [str(x) for x in df_all['Tumor_Seq_Allele2']]\n",
    "df_all['Start_Position'] = [int(x) for x in df_all['Start_Position']]\n",
    "df_all['End_Position'] = [int(x) for x in df_all['End_Position']]\n",
    "df_pe_m = prime.run(df_all, 'cBioPortal', chrom_dict=chrom_dict_m, PAM='NGN',rankby='RF_Score', pegRNAs_per_mut=5, RTT_lengths=[5, 10, 15, 25, 30],PBS_lengths=[8, 10, 13, 15], min_RHA_size=1, RE_sites=None, polyT_threshold=4,proto_size=19, context_size=120, before_proto_context=5, sensor_length=60,sensor_orientation='reverse-complement', sensor=True)\n",
    "df_pe_m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4707"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_pe_m['ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Hugo_Symbol\n",
       "TP53      588\n",
       "APC       118\n",
       "PIK3CA    100\n",
       "PTEN       95\n",
       "ARID1A     85\n",
       "         ... \n",
       "FGF14       1\n",
       "FGA         1\n",
       "ERCC1       1\n",
       "ERRFI1      1\n",
       "SETD1B      1\n",
       "Name: count, Length: 614, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_all['Hugo_Symbol'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pe_m.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_mouse_result.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bind human and mouse data  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pe_h = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_human_result.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pe_h = df_pe_h.rename(columns = {'tx_id_h':'tx_id'})\n",
    "df_pe_m = df_pe_m.rename(columns = {'tx_id_m':'tx_id'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.concat([df_pe_h,df_pe_m]).drop(['pegRNA_id','tx_id'], axis=1).to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_result_all.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Statistics  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4944"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_pe[[len(x.split('-'))==2 for x in df_pe['ID']]]['ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4707"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_pe[[len(x.split('-'))==4 for x in df_pe['ID']]]['ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2720"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_be[[len(x.split('-'))==2 for x in df_be['ID']]]['ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1892"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_be[[len(x.split('-'))==4 for x in df_be['ID']]]['ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
