Search.setIndex({"alltitles": {"1.0.0": [[0, "id1"]], "1.0.1": [[0, "id2"]], "1.0.2": [[0, "id3"]], "1.0.3": [[0, "id4"]], "About H2M": [[0, "about-h2m"]], "Additional Usage Hint": [[5, "additional-usage-hint"]], "Additional function 1: modeling M2H": [[5, "additional-function-1-modeling-m2h"]], "Additional function 2: modeling H2H/M2M paralogs": [[5, "additional-function-2-modeling-h2h-m2m-paralogs"]], "Additional function 3: modeling for base editing": [[5, "additional-function-3-modeling-for-base-editing"]], "Additional function 4: modeling by amino acid change input": [[5, "additional-function-4-modeling-by-amino-acid-change-input"]], "Alternative modeling": [[5, "alternative-modeling"]], "Basic usage": [[5, "basic-usage"]], "Batch Processing": [[5, "batch-processing"]], "Complete Documentation": [[1, "complete-documentation"]], "Compute the murine variant equivalents": [[5, "compute-the-murine-variant-equivalents"]], "Contents:": [[3, null]], "Flank Size": [[5, "flank-size"]], "Functions": [[1, "functions"]], "Get canonical transcript IDs for human": [[5, "get-canonical-transcript-ids-for-human"]], "Get canonical transcript IDs for mouse": [[5, "get-canonical-transcript-ids-for-mouse"]], "Get transcript ID": [[5, "get-transcript-id"]], "H2M Database Online Portal": [[2, "h2m-database-online-portal"]], "H2M Documentation": [[3, "h2m-documentation"]], "H2M Modeling Class Description": [[1, "h2m-modeling-class-description"]], "H2M Output Data Description": [[1, "h2m-output-data-description"]], "Importing packages": [[5, "importing-packages"]], "Input format": [[5, "input-format"]], "Jupyter Notebook Tutorial": [[4, "jupyter-notebook-tutorial"]], "Loading data": [[5, "loading-data"]], "Modeling human variants in the mouse genome": [[5, "modeling-human-variants-in-the-mouse-genome"]], "Original modeling with uncertain effects": [[5, "original-modeling-with-uncertain-effects"]], "PEGG: Prime Editing Guide Generator": [[2, "pegg-prime-editing-guide-generator"]], "Package Installation": [[3, "package-installation"], [5, "package-installation"]], "Query orthologous genes": [[5, "query-orthologous-genes"], [5, "id1"]], "Quick Start": [[5, "quick-start"]], "Read from ClinVar": [[5, "read-from-clinvar"]], "Read from GenomAD  - VCF": [[5, "read-from-genomad-vcf"]], "Read from cBioPortal - MAF": [[5, "read-from-cbioportal-maf"]], "Result visualization": [[5, "result-visualization"]], "Single variant input": [[5, "single-variant-input"]], "Version History": [[0, "version-history"]]}, "docnames": ["about", "documentation", "downstream", "index", "jupyter", "quickstart"], "envversion": {"sphinx": 61, "sphinx.domains.c": 3, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 9, "sphinx.domains.index": 1, "sphinx.domains.javascript": 3, "sphinx.domains.math": 2, "sphinx.domains.python": 4, "sphinx.domains.rst": 2, "sphinx.domains.std": 2}, "filenames": ["about.md", "documentation.md", "downstream.md", "index.md", "jupyter.md", "quickstart.md"], "indexentries": {"anno_loader() (in module bioh2m)": [[1, "bioh2m.anno_loader", false]], "cbio_reader() (in module bioh2m)": [[1, "bioh2m.cbio_reader", false]], "clinvar_reader() (in module bioh2m)": [[1, "bioh2m.clinvar_reader", false]], "clinvar_to_maf() (in module bioh2m)": [[1, "bioh2m.clinvar_to_maf", false]], "genome_loader() (in module bioh2m)": [[1, "bioh2m.genome_loader", false]], "get_tx_batch() (in module bioh2m)": [[1, "bioh2m.get_tx_batch", false]], "get_tx_id() (in module bioh2m)": [[1, "bioh2m.get_tx_id", false]], "get_variant_type() (in module bioh2m)": [[1, "bioh2m.get_variant_type", false]], "model() (in module bioh2m)": [[1, "bioh2m.model", false]], "model_batch() (in module bioh2m)": [[1, "bioh2m.model_batch", false]], "query() (in module bioh2m)": [[1, "bioh2m.query", false]], "query_batch() (in module bioh2m)": [[1, "bioh2m.query_batch", false]], "vcf_reader() (in module bioh2m)": [[1, "bioh2m.vcf_reader", false]], "vcf_to_maf() (in module bioh2m)": [[1, "bioh2m.vcf_to_maf", false]], "visualization() (in module bioh2m)": [[1, "bioh2m.visualization", false]]}, "objects": {"bioh2m": [[1, 0, 1, "", "anno_loader"], [1, 0, 1, "", "cbio_reader"], [1, 0, 1, "", "clinvar_reader"], [1, 0, 1, "", "clinvar_to_maf"], [1, 0, 1, "", "genome_loader"], [1, 0, 1, "", "get_tx_batch"], [1, 0, 1, "", "get_tx_id"], [1, 0, 1, "", "get_variant_type"], [1, 0, 1, "", "model"], [1, 0, 1, "", "model_batch"], [1, 0, 1, "", "query"], [1, 0, 1, "", "query_batch"], [1, 0, 1, "", "vcf_reader"], [1, 0, 1, "", "vcf_to_maf"], [1, 0, 1, "", "visualization"]]}, "objnames": {"0": ["py", "function", "Python function"]}, "objtypes": {"0": "py:function"}, "terms": {"": [3, 5], "0": [1, 3, 5], "000": 2, "0705": 5, "0_ensg00000141510": 5, "0_ensg00000141510_2024_02_07_11_36_03": 1, "1": [1, 2, 3], "10": [1, 5], "10000": 1, "10602915": 5, "11": [1, 5], "11143999": 5, "1191613": 1, "12": [1, 3, 5], "1205375": 1, "13": [1, 5], "14": [1, 5], "140953": 5, "15": [1, 5], "15aa": 5, "16": [1, 5], "17": [1, 5], "175": 5, "17578": 5, "1796995": 5, "18": [1, 5], "19": [1, 5], "1g": 5, "2": [1, 2, 3], "20": [1, 5], "208043": 1, "21": 1, "2115855": 5, "22": 1, "23": 1, "233866": 5, "24": [1, 2, 5], "25": 1, "255": 2, "25_grch37": 5, "26": 1, "27": 1, "27_grcm39_genom": 5, "28": 1, "29": 1, "3": [1, 2, 3], "30": 1, "300": 1, "308061": 1, "31": [1, 5], "325626": [1, 5], "32798013": 5, "3457": 5, "3537": 5, "361149": 1, "367": 5, "37": [0, 1, 5], "375926": 5, "376": 5, "38": [1, 5], "3a": 5, "3b": 5, "4": [1, 2, 3], "48": 2, "4aa": 5, "5": [0, 1, 5], "514_516cgc": 5, "515_516gc": 5, "515g": 5, "573320": 5, "6": [1, 5], "62": 5, "661t": 5, "663c": 5, "68": 5, "680": 2, "69479338": 5, "69479339": 5, "69480434": 5, "7": [1, 5], "707": 2, "720": 2, "75": 5, "7577021": 5, "7577115": 5, "7577116": 5, "7577120": [1, 5], "7577523": 5, "7577534": 5, "7578405": 5, "7578407": 5, "7578555": 5, "77": 5, "8": [1, 5], "809g": 5, "813_814": 5, "818g": 5, "822_823": 5, "892": 2, "9": [1, 3, 5], "907_908ag": 5, "907_909aga": 5, "908g": 5, "917g": 5, "925434": 1, "925574": 1, "925707": 1, "926695": 1, "94": 5, "944": 2, "A": [1, 4, 5], "AT": 5, "BE": [1, 5], "By": 5, "For": 5, "INS": [1, 5], "If": [1, 3, 5], "In": 5, "It": 1, "One": [1, 5], "The": [1, 3, 5], "Then": 5, "These": 5, "To": [3, 5], "aa": [1, 5], "aacr": [2, 5], "about": [3, 5], "accept": [1, 5], "access": 4, "achiev": 5, "acid": [1, 3], "ad": [0, 1], "addit": [1, 3], "adgra2": 5, "adgrb3": 5, "adgrl3": 5, "afdn": 5, "after": [1, 5], "again": 5, "alias": 5, "align": 1, "align_dict": 1, "align_input": 1, "all": [1, 5], "also": [3, 5], "alt": 1, "alt_col": 1, "alt_seq": 1, "alt_seq_h": [1, 5], "alt_seq_h_1": 5, "alt_seq_h_2": 5, "alt_seq_m": [1, 5], "alt_seq_m_ori": [1, 5], "alter": 5, "altern": [0, 1, 3], "alwai": 1, "amino": [1, 3], "an": [1, 5], "analysi": 5, "anno_load": [1, 3, 5], "annot": [1, 5], "ar": [1, 3, 5], "assembli": [0, 5], "attach": 1, "author": 0, "avail": [2, 3, 5], "babam2": 5, "bai3": 5, "base": [1, 2, 3], "basic": 3, "batch": [1, 3], "been": [3, 5], "being": 5, "below": 5, "between": [1, 3, 5], "bind": 1, "bioh2m": [1, 3, 5], "biologi": 0, "bool": 1, "borcs8": 5, "both": [1, 5], "bre": 5, "break": 1, "browser": 5, "build": 5, "built": 5, "c": [1, 5], "c272lf": 5, "c275lf": 5, "ca": 5, "cac": 5, "cag": 5, "call": 5, "can": [1, 3, 4, 5], "cancer": 0, "cannot": [1, 5], "canon": [1, 3], "casc5": 5, "case": 1, "cbio": 1, "cbio_read": [1, 3, 5], "cbioport": [1, 3], "cersa": 3, "cgc": 5, "chang": [1, 3], "check": 5, "choos": 5, "chr11": 5, "chr17": 5, "chr_h": [1, 5], "chr_m": [1, 5], "chromosom": 1, "circumsat": 5, "class": [0, 3, 5], "classif": 1, "classification_h": [1, 5], "classification_m": [1, 5], "clinvar": [1, 2, 3], "clinvar_read": [1, 3, 5], "clinvar_to_maf": [1, 3, 5], "code": [1, 5], "codon": [1, 5], "col_nam": 1, "column": [1, 5], "com": [0, 2], "common": 5, "compat": [1, 2, 3, 5], "complet": 3, "comput": [1, 2, 3], "confid": 5, "connect": 5, "consid": 1, "consider": 1, "consist": 1, "contain": 1, "convert": [1, 5], "coor": [1, 5], "coordin": [1, 5], "cop1": 5, "correspond": 5, "cosmic": 2, "could": [1, 5], "csv": [1, 5], "data": 3, "data_mut": 1, "data_mutations_extend": 5, "databas": [1, 3, 5], "datafram": [1, 5], "dataset": 5, "db": [1, 5], "db_h": [1, 5], "db_m": [1, 5], "de": 1, "death": 1, "default": [1, 5], "definitli": 5, "del": [1, 5], "depart": 0, "depend": [1, 3, 5], "deriv": 2, "descript": 3, "design": 2, "desir": 5, "df": [1, 5], "df_fail": 5, "df_queri": 5, "df_result": 5, "dictionari": [1, 5], "differ": [3, 5], "dirc2": 5, "direct": [1, 5], "directli": 5, "dna": 1, "dnp": [1, 5], "do": 5, "document": 5, "done": 5, "dong": 0, "download": [3, 5], "drop": 5, "dropbox": 5, "due": [3, 5], "e_7": [1, 5], "each": [1, 5], "ecords_h": 1, "edit": [1, 3], "effect": [1, 3], "element": 5, "eloc": 5, "email": 0, "end": [1, 5], "end_h": [1, 5], "end_h_1": 5, "end_h_2": 5, "end_m": [1, 5], "end_m_ori": [1, 5], "engin": 2, "ensembl": [1, 5], "ensg00000141510": 5, "ensmusg00000059552": 5, "ensmust00000005371": 5, "ensmust00000108657": 5, "ensmust00000108658": 5, "ensmust00000130540": 5, "ensmust00000147512": 5, "ensmust00000164812": 5, "ensmust00000171247": 5, "enst00000171111": 5, "enst00000269305": 5, "enst00000302401": 5, "enst00000324954": 5, "enst00000344626": 5, "enst00000349721": 5, "enst00000357248": 5, "enst00000358026": 5, "enst00000359597": 5, "enst00000382183": 5, "enst00000382185": 5, "enst00000382186": 5, "enst00000382194": 5, "enst00000382203": 5, "enst00000413465": 5, "enst00000413806": 5, "enst00000416751": 5, "enst00000417599": 5, "enst00000420246": 5, "enst00000423555": 5, "enst00000429416": 5, "enst00000439732": 5, "enst00000444061": 5, "enst00000445888": 5, "enst00000450198": 5, "enst00000450717": 5, "enst00000452193": 5, "enst00000455263": 5, "enst00000457226": 5, "enst00000491574": 5, "enst00000503591": 5, "enst00000504290": 5, "enst00000504937": 5, "enst00000505014": 5, "enst00000508793": 5, "enst00000509690": 5, "enst00000510385": 5, "enst00000514944": 5, "enst00000538456": 5, "enst00000541122": 5, "enst00000574684": 5, "enst00000576024": 5, "enst00000585799": 5, "enst00000586122": 5, "enst00000586892": 5, "enst00000586921": 5, "enst00000586985": 5, "enst00000587988": 5, "enst00000589677": 5, "enst00000590574": 5, "enst00000591545": 5, "enst00000591595": 5, "enst00000592158": 5, "enst00000592604": 5, "enst00000604348": 5, "entri": 1, "equal": 1, "equival": [2, 3], "error": 1, "etc": [1, 5], "exactli": 1, "exampl": [1, 5], "except": 5, "exist": 5, "exon": 1, "exon_h": [1, 5], "exon_m": [1, 5], "exon_num_h": [1, 5], "exon_num_m": [1, 5], "experienc": [3, 5], "express": 1, "exsist": 5, "f221l": 5, "failur": 5, "fall": 1, "fals": [1, 5], "fam46c": 5, "featuredb": 1, "few": 5, "fidel": 5, "file": [0, 1, 3, 5], "filepath": [1, 5], "find": 5, "first": [0, 1], "flank": [1, 3], "flank_siz": [1, 5], "flank_size_left": [1, 5], "flank_size_right": [1, 5], "fna": 5, "folder": 5, "follow": 5, "form": 1, "format": [1, 3], "found": [1, 5], "fram": 1, "frame": [1, 5], "from": [1, 2, 3], "function": 3, "further": 5, "g": 5, "g1164r": 5, "g1194r": 5, "gba": 5, "gba1": 5, "gc": 5, "gcf_000001405": 5, "gcf_000001635": 5, "gckr": 5, "gdc": 5, "gencod": [1, 5], "gencode_v19_grch37": 5, "gencode_vm33_grcm39": 5, "gene": [0, 1, 3], "gene_id": [1, 5], "gene_id_h": [1, 5], "gene_id_m": [1, 5], "gene_name_h": [1, 5], "gene_name_h_1": 5, "gene_name_h_2": 5, "gene_name_m": [1, 5], "gener": [1, 3], "genet": 3, "geni": [2, 5], "genom": [0, 1, 2, 3], "genomad": [1, 3], "genome_load": [1, 3, 5], "get": [1, 3], "get_tx_batch": [1, 3, 5], "get_tx_id": [1, 3, 5], "get_variant_typ": [1, 3], "gg": 5, "github": [3, 5], "give": 5, "given": [1, 3, 5], "gm22866": 5, "gm25938": 5, "gmail": 0, "gnb2l1": 5, "gnomad_v4": [1, 5], "go": 5, "gorkordkx": 0, "gpr124": 5, "grch37": 5, "grch37_clinvar_20230923": 1, "grch37_clinvar_20240206": 5, "grch38": 5, "grcm39": 5, "grna": 2, "gtgaggatgggc": 5, "guid": 3, "gz": [1, 5], "h": [1, 5], "h1": 5, "h2ac1": 5, "h2bc13": 5, "h2bc5": 5, "h2bc8": 5, "h2h": [1, 3], "h2m": [4, 5], "h3": 5, "h3c1": 5, "h3c12": 5, "h3c2": 5, "h3c3": 5, "h3c4": 5, "h3f3a": 5, "h3f3b": 5, "h4c5": 5, "ha": [1, 3, 5], "hand": 5, "have": 5, "here": 4, "hgvsc": 1, "hgvsc_h": [1, 5], "hgvsc_m": [1, 5], "hgvsc_m_ori": [1, 5], "hgvsp": 1, "hgvsp_h": [1, 5], "hgvsp_h_1": 5, "hgvsp_h_2": 5, "hgvsp_m": [1, 5], "hgvsp_m_ori": [1, 5], "high": 3, "hint": 3, "hist1h1": 5, "hist1h1b": 5, "hist1h1c": 5, "hist1h1d": 5, "hist1h2ac": 5, "hist1h2bd": 5, "hist1h2bj": 5, "hist1h2bk": 5, "hist1h3": 5, "hist1h3b": 5, "hist1h3d": 5, "hist1h3f": 5, "hist1h3j": 5, "hist1h4": 5, "hist3h3": 5, "histori": 3, "homolog": [1, 5], "homologi": 5, "how": 5, "http": 2, "huaman": 5, "human": [0, 1, 2, 3], "human2mous": 2, "i": [1, 2, 3, 5], "i_5": 1, "ick": 5, "id": [0, 1, 3], "ident": [1, 5], "identif": 1, "identifi": 5, "identificaiton": 5, "idh1": 5, "import": 3, "includ": [1, 5], "indel": 1, "index": [1, 3, 5], "index_list_h": [1, 5], "index_list_m": [1, 5], "infom": 5, "inform": [1, 5], "input": [0, 1, 3], "institut": 0, "int": 1, "intact": 1, "integr": 0, "interest": [2, 5], "internet": [1, 5], "introduc": 5, "intron": 1, "issu": [3, 5], "its": [1, 5], "jupyt": 3, "keap1": 5, "keep": [1, 5], "kei": 5, "kernel": 1, "kexin": 0, "kexindong": 5, "kmt5a": 5, "knl1": 5, "koch": 0, "lab": 0, "larg": 5, "large1": 5, "lead": 1, "least": 5, "left": [1, 5], "len": 5, "length": [1, 5], "lenth": 1, "list": [1, 5], "list_of_id": 1, "list_tx_id_h": 5, "list_tx_id_m": 5, "load": [1, 3], "local": 5, "locat": 1, "long": 1, "lphn3": 5, "m": [1, 5], "m2h": [1, 3], "m2m": [1, 3], "maf": [1, 3], "mai": [1, 5], "main": 3, "mani": 5, "manual": [0, 5], "map": 1, "massachusett": 0, "match": [1, 5], "max_altern": [1, 5], "maximum": [1, 5], "maxlength": 1, "mef2b": 5, "mef2bnb": 5, "memory_protect": 1, "memory_s": 1, "mgea5": 5, "middl": 1, "mini": [3, 5], "miss": 0, "missens": [1, 5], "mkl1": 5, "mllt4": 5, "model": 3, "model_batch": [1, 3, 5], "model_result": [1, 5], "model_result_long": 5, "modifi": 5, "more": [3, 5], "most": [3, 5], "mous": [1, 2, 3], "mre11": 5, "mre11a": 5, "mrtfa": 5, "multipl": [1, 5], "murin": [1, 2, 3], "must": 1, "mutat": [1, 2, 3, 5], "mutaton": 5, "name": [1, 5], "napb": 5, "nc": 1, "necessari": 1, "necesssari": 1, "nectin4": 5, "need": [1, 5], "neg": 1, "neucleotid": 1, "non": [1, 5], "none": 1, "nonsens": 1, "nonstop": 1, "nonstop_s": 1, "note": 5, "notebook": 3, "now": 5, "nsd1": 5, "nsd3": 5, "nucleotid": 5, "number": [0, 1, 5], "off": [3, 5], "offer": 5, "oga": 5, "one": [1, 5], "one2mani": 5, "one2on": 5, "ones": 5, "onli": [1, 5], "onlin": 3, "onp": [1, 5], "opiton": 1, "option": [1, 5], "order": 5, "origi": 1, "origin": [1, 3], "ortholog": [1, 2, 3], "other": 1, "output": [2, 3, 5], "over": 2, "p13_genom": 5, "packag": 2, "pak5": 5, "pak7": 5, "panda": [1, 5], "paralog": 3, "param": [1, 5], "paramet": [1, 5], "park2": 5, "path": [1, 5], "path_aacr": 5, "path_h_anno": [1, 5], "path_h_ref": [1, 5], "path_m_anno": 5, "path_m_ref": 5, "pd": 5, "pegg": 3, "pegrna": 2, "peptid": 1, "percentag": 5, "pip": [3, 5], "place": 1, "pleas": 5, "plot": 1, "portal": 3, "posit": [1, 5], "postion": 5, "pre": 2, "precis": [2, 3], "prepar": 1, "presenc": 5, "prime": 3, "print": 1, "print_siz": [1, 5], "prkn": 5, "problem": [3, 5], "process": [1, 3], "protein": 5, "provid": 5, "public": 5, "publicdatabas": 5, "pulic": 3, "pvrl4": 5, "pypi": [3, 5], "pysam": [3, 5], "python": [2, 3, 5], "queri": [1, 3], "query_batch": [1, 3, 5], "query_result": 5, "quick": 3, "r": 5, "r172g": 5, "r172h": 5, "r175h": 5, "r249_t253delinss": 5, "r270h": 5, "r273h": 5, "r303k": 5, "r303q": 5, "r306q": 5, "rack1": 5, "read": 3, "recommend": [1, 3, 5], "record": [1, 5], "records_h": [1, 5], "records_m": [1, 5], "ref": [1, 5], "ref_col": 1, "ref_seq": 1, "ref_seq_h": [1, 5], "ref_seq_h_1": 5, "ref_seq_h_2": 5, "ref_seq_m": [1, 5], "ref_seq_m_ori": [1, 5], "refer": [0, 1, 5], "refernc": 1, "region": [1, 5], "releas": 0, "rememb": 5, "renam": 5, "renfer": 1, "replac": 5, "repositori": [3, 5], "research": 0, "reset_index": 5, "result": [1, 3], "return": [1, 5], "rfwd2": 5, "right": [1, 5], "rivera": 0, "rna": 2, "round": [3, 5], "row": 1, "rule": 1, "same": [1, 5], "save": 1, "search": 5, "search_altern": [1, 5], "search_alternatv": 5, "second": 1, "see": [1, 3, 5], "sens": 5, "sept5": 5, "sept9": 5, "septin5": 5, "seq": 1, "sequenc": [0, 1, 5], "set": [1, 5], "setd8": 5, "shift": [1, 5], "should": [1, 5], "show": [1, 5], "show_sequ": [1, 5], "side": [1, 5], "simalar": 5, "similar": [1, 5], "simpli": 5, "sinc": [3, 5], "singl": 3, "site": 5, "size": 3, "slc49a4": 5, "slightli": 5, "small": 0, "smaller": 1, "smarca2": 5, "smarca4": 5, "snp": [1, 5], "so": 5, "solv": [3, 5], "some": 5, "sometim": 5, "sourc": 3, "speci": [1, 5], "specif": [1, 5], "specifi": [1, 5], "splice": 5, "splicing_s": 1, "stabl": 1, "start": [1, 3], "start_h": [1, 5], "start_h_1": 5, "start_h_2": 5, "start_m": [1, 5], "start_m_ori": [1, 5], "statement": [1, 5], "statu": [1, 5], "sting1": 5, "stop": 1, "str": 1, "strand": 1, "strand_h": [1, 5], "strand_m": [1, 5], "strategi": 5, "string": 1, "sty": 1, "style": [1, 5], "successfulli": [1, 5], "summari": 1, "symbol": 5, "s\u00e1nchez": 0, "t": [1, 5], "take": 5, "target": [1, 2, 5], "tbodi": 1, "tceb1": 5, "tcga": 5, "technologi": 0, "teh": 5, "tent5c": 5, "test": [3, 5], "text": 1, "th": 1, "than": [1, 5], "thead": 1, "them": [3, 5], "thi": [1, 2, 3, 5], "three": 5, "through": [3, 5], "throughput": 3, "time": 1, "tmem173b": 5, "tno": 1, "tnp": [1, 5], "too": 5, "top": 1, "total": [1, 5], "tp53": [1, 5], "tpt1p6": 5, "tr": 1, "transcirpt": 1, "transcript": [0, 1, 3], "translat": 1, "trp53": 5, "true": [1, 5], "tutori": [0, 3, 5], "two": [1, 5], "tx_id": [1, 5], "tx_id_1_h": 5, "tx_id_2_h": 5, "tx_id_h": [1, 5], "tx_id_h_2": 5, "tx_id_m": [1, 5], "tx_id_m_2": 5, "txt": [1, 5], "ty": [1, 5], "ty_h": [1, 5], "type": [1, 5], "type_h": [1, 5], "type_m": [1, 5], "u1": 5, "uncertain": 3, "under": 5, "updat": 0, "upload": 5, "us": [1, 2, 3, 5], "usag": 3, "user": 5, "vairant": 1, "valu": [1, 5], "variant": [1, 2, 3], "variat": [1, 5], "variation_id": [1, 5], "vcf": [1, 3], "vcf_reader": [1, 3, 5], "vcf_to_maf": [1, 3, 5], "ver": [1, 5], "versa": [1, 3, 5], "version": [1, 3, 4, 5], "vertic": 1, "vice": [3, 5], "vise": 1, "visual": [1, 3], "visul": 5, "visula": 5, "wai": 5, "want": 5, "we": [2, 5], "well": 5, "wheel": [3, 5], "when": [1, 5], "which": [3, 5], "while": 1, "whole": 1, "whsc1": 5, "whsc1l1": 5, "wild": 5, "wish": 5, "without": [1, 3, 5], "would": 1, "x122_splice": 5, "x125_splice": 5, "you": [3, 5], "your": [1, 5]}, "titles": ["About H2M", "Complete Documentation", "H2M Database Online Portal", "H2M Documentation", "Jupyter Notebook Tutorial", "Quick Start"], "titleterms": {"0": 0, "1": [0, 5], "2": [0, 5], "3": [0, 5], "4": 5, "about": 0, "acid": 5, "addit": 5, "altern": 5, "amino": 5, "base": 5, "basic": 5, "batch": 5, "canon": 5, "cbioport": 5, "chang": 5, "class": 1, "clinvar": 5, "complet": 1, "comput": 5, "content": 3, "data": [1, 5], "databas": 2, "descript": 1, "document": [1, 3], "edit": [2, 5], "effect": 5, "equival": 5, "flank": 5, "format": 5, "from": 5, "function": [1, 5], "gene": 5, "gener": 2, "genom": 5, "genomad": 5, "get": 5, "guid": 2, "h2h": 5, "h2m": [0, 1, 2, 3], "hint": 5, "histori": 0, "human": 5, "id": 5, "import": 5, "input": 5, "instal": [3, 5], "jupyt": 4, "load": 5, "m2h": 5, "m2m": 5, "maf": 5, "model": [1, 5], "mous": 5, "murin": 5, "notebook": 4, "onlin": 2, "origin": 5, "ortholog": 5, "output": 1, "packag": [3, 5], "paralog": 5, "pegg": 2, "portal": 2, "prime": 2, "process": 5, "queri": 5, "quick": 5, "read": 5, "result": 5, "singl": 5, "size": 5, "start": 5, "transcript": 5, "tutori": 4, "uncertain": 5, "usag": 5, "variant": 5, "vcf": 5, "version": 0, "visual": 5}})