{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script generates base editing guide RNA library for a subset of frequently mutated mutations (for both human and mouse).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bioh2m as h2m\n",
    "from pegg import prime\n",
    "from pegg import base\n",
    "import pandas as pd\n",
    "df_all = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/BEGG/be_human.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# For Human  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000001</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>ENST00000263967.3</td>\n",
       "      <td>178936091</td>\n",
       "      <td>178936091</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000004</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>ENST00000263967.3</td>\n",
       "      <td>178952085</td>\n",
       "      <td>178952085</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000005</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7578406</td>\n",
       "      <td>7578406</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000007</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>ENST00000263967.3</td>\n",
       "      <td>178936082</td>\n",
       "      <td>178936082</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000008</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3138</th>\n",
       "      <td>FGFR3</td>\n",
       "      <td>ENST00000260795.2</td>\n",
       "      <td>1806162</td>\n",
       "      <td>1806162</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005025</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3139</th>\n",
       "      <td>NTRK2</td>\n",
       "      <td>ENST00000323115.4</td>\n",
       "      <td>87359980</td>\n",
       "      <td>87359980</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005026</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3140</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>55211111</td>\n",
       "      <td>55211111</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005027</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3141</th>\n",
       "      <td>XPO1</td>\n",
       "      <td>ENST00000401558.2</td>\n",
       "      <td>61719275</td>\n",
       "      <td>61719275</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005028</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3142</th>\n",
       "      <td>BCL9</td>\n",
       "      <td>ENST00000234739.3</td>\n",
       "      <td>147091416</td>\n",
       "      <td>147091416</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005030</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3143 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Hugo_Symbol            tx_id_h  Start_Position  End_Position  \\\n",
       "0           KRAS  ENST00000256078.4        25398284      25398284   \n",
       "1         PIK3CA  ENST00000263967.3       178936091     178936091   \n",
       "2         PIK3CA  ENST00000263967.3       178952085     178952085   \n",
       "3           TP53  ENST00000269305.4         7578406       7578406   \n",
       "4         PIK3CA  ENST00000263967.3       178936082     178936082   \n",
       "...          ...                ...             ...           ...   \n",
       "3138       FGFR3  ENST00000260795.2         1806162       1806162   \n",
       "3139       NTRK2  ENST00000323115.4        87359980      87359980   \n",
       "3140        EGFR  ENST00000275493.2        55211111      55211111   \n",
       "3141        XPO1  ENST00000401558.2        61719275      61719275   \n",
       "3142        BCL9  ENST00000234739.3       147091416     147091416   \n",
       "\n",
       "     Reference_Allele Tumor_Seq_Allele2 Variant_Type             ID Chromosome  \n",
       "0                   C                 T          SNP  AACR-H0000001         12  \n",
       "1                   G                 A          SNP  AACR-H0000004          3  \n",
       "2                   A                 G          SNP  AACR-H0000005          3  \n",
       "3                   C                 T          SNP  AACR-H0000007         17  \n",
       "4                   G                 A          SNP  AACR-H0000008          3  \n",
       "...               ...               ...          ...            ...        ...  \n",
       "3138                C                 T          SNP  AACR-H0005025          4  \n",
       "3139                C                 T          SNP  AACR-H0005026          9  \n",
       "3140                C                 T          SNP  AACR-H0005027          7  \n",
       "3141                T                 C          SNP  AACR-H0005028          2  \n",
       "3142                G                 A          SNP  AACR-H0005030          1  \n",
       "\n",
       "[3143 rows x 9 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_h_ref, path_m_ref = '/Users/kexindong/Documents/GitHub/Database/RefGenome/ncbi-2023-09-12/GCF_000001405.25_GRCh37.p13_genomic.fna.gz', '/Users/kexindong/Documents/GitHub/Database/RefGenome/mouse-2023-09-13/GCF_000001635.27_GRCm39_genomic.fna.gz'\n",
    "chrom_dict_h, i = prime.genome_loader(path_h_ref)\n",
    "chrom_dict_m, i = prime.genome_loader(path_m_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>...</th>\n",
       "      <th>Protospacer_Location</th>\n",
       "      <th>Ideal_Edit_Window</th>\n",
       "      <th>Editor</th>\n",
       "      <th>gRNA_id</th>\n",
       "      <th>OnTarget_Azimuth_Score</th>\n",
       "      <th>sensor_wt</th>\n",
       "      <th>sensor_alt</th>\n",
       "      <th>sensor_orientation</th>\n",
       "      <th>sensor_error</th>\n",
       "      <th>contains_polyT_terminator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>GAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCAGGCC</td>\n",
       "      <td>GAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCAGGCC</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_1</td>\n",
       "      <td>54.0</td>\n",
       "      <td>TGGAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCAGG</td>\n",
       "      <td>TGGAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCAGG</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_2</td>\n",
       "      <td>59.0</td>\n",
       "      <td>CATGGAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCA</td>\n",
       "      <td>CATGGAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCA</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_3</td>\n",
       "      <td>59.0</td>\n",
       "      <td>ACAGCAAGGCTGGCATGGAGGAAGATCACACCTACGAGGT</td>\n",
       "      <td>ACAGCAAGGCTGGCATGGAGGAAGATCACACCCACGAGGT</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006798</td>\n",
       "      <td>62006798</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002993-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_4</td>\n",
       "      <td>66.0</td>\n",
       "      <td>GCTGGGCCTGCCCCTCTCCTTACCTCGTAGGTGTGATCTT</td>\n",
       "      <td>GCTGGGCCTGCCCCTCTCCTTACCTCGCAGGTGTGATCTT</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15376</th>\n",
       "      <td>2850</td>\n",
       "      <td>RNF213</td>\n",
       "      <td>ENSMUST00000093902.12</td>\n",
       "      <td>78357600</td>\n",
       "      <td>78357600</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0003804-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_15376</td>\n",
       "      <td>38.0</td>\n",
       "      <td>GCAATGGACCACACTTTTCCGGGGCAGGTGGGGCAGGAAG</td>\n",
       "      <td>GCAATGGACCACACTTTCCCGGGGCAGGTGGGGCAGGAAG</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15377</th>\n",
       "      <td>2850</td>\n",
       "      <td>RNF213</td>\n",
       "      <td>ENSMUST00000093902.12</td>\n",
       "      <td>78357600</td>\n",
       "      <td>78357600</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0003804-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_15377</td>\n",
       "      <td>46.0</td>\n",
       "      <td>GAGCAATGGACCACACTTTTCCGGGGCAGGTGGGGCAGGA</td>\n",
       "      <td>GAGCAATGGACCACACTTTCCCGGGGCAGGTGGGGCAGGA</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15378</th>\n",
       "      <td>2850</td>\n",
       "      <td>RNF213</td>\n",
       "      <td>ENSMUST00000093902.12</td>\n",
       "      <td>78357600</td>\n",
       "      <td>78357600</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0003804-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_15378</td>\n",
       "      <td>60.0</td>\n",
       "      <td>TAGAGCAATGGACCACACTTTTCCGGGGCAGGTGGGGCAG</td>\n",
       "      <td>TAGAGCAATGGACCACACTTTCCCGGGGCAGGTGGGGCAG</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15379</th>\n",
       "      <td>2850</td>\n",
       "      <td>RNF213</td>\n",
       "      <td>ENSMUST00000093902.12</td>\n",
       "      <td>78357600</td>\n",
       "      <td>78357600</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0003804-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_15379</td>\n",
       "      <td>41.0</td>\n",
       "      <td>CCAAATCTTAGAGCAATGGACCACACTTTTCCGGGGCAGG</td>\n",
       "      <td>CCAAATCTTAGAGCAATGGACCACACTTTCCCGGGGCAGG</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15380</th>\n",
       "      <td>2850</td>\n",
       "      <td>RNF213</td>\n",
       "      <td>ENSMUST00000093902.12</td>\n",
       "      <td>78357600</td>\n",
       "      <td>78357600</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0003804-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_15380</td>\n",
       "      <td>59.0</td>\n",
       "      <td>TTAGAGCAATGGACCACACTTTTCCGGGGCAGGTGGGGCA</td>\n",
       "      <td>TTAGAGCAATGGACCACACTTTCCCGGGGCAGGTGGGGCA</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15381 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       mutation_idx Hugo_Symbol                tx_id_h  Start_Position  \\\n",
       "0                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "1                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "2                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "3                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "4                 1       CD79B  ENSMUST00000044228.11        62006798   \n",
       "...             ...         ...                    ...             ...   \n",
       "15376          2850      RNF213  ENSMUST00000093902.12        78357600   \n",
       "15377          2850      RNF213  ENSMUST00000093902.12        78357600   \n",
       "15378          2850      RNF213  ENSMUST00000093902.12        78357600   \n",
       "15379          2850      RNF213  ENSMUST00000093902.12        78357600   \n",
       "15380          2850      RNF213  ENSMUST00000093902.12        78357600   \n",
       "\n",
       "       End_Position Reference_Allele Tumor_Seq_Allele2 Variant_Type  \\\n",
       "0          62006799                A                 G          SNP   \n",
       "1          62006799                A                 G          SNP   \n",
       "2          62006799                A                 G          SNP   \n",
       "3          62006799                A                 G          SNP   \n",
       "4          62006798                T                 C          SNP   \n",
       "...             ...              ...               ...          ...   \n",
       "15376      78357600                A                 G          SNP   \n",
       "15377      78357600                A                 G          SNP   \n",
       "15378      78357600                A                 G          SNP   \n",
       "15379      78357600                A                 G          SNP   \n",
       "15380      78357600                A                 G          SNP   \n",
       "\n",
       "                      ID Chromosome  ...  Protospacer_Location  \\\n",
       "0      AACR-H0002100-M01         17  ...                    20   \n",
       "1      AACR-H0002100-M01         17  ...                    18   \n",
       "2      AACR-H0002100-M01         17  ...                    16   \n",
       "3      AACR-H0002100-M01         17  ...                     3   \n",
       "4      AACR-H0002993-M01         17  ...                     8   \n",
       "...                  ...        ...  ...                   ...   \n",
       "15376  AACR-H0003804-M01         17  ...                    18   \n",
       "15377  AACR-H0003804-M01         17  ...                    16   \n",
       "15378  AACR-H0003804-M01         17  ...                    14   \n",
       "15379  AACR-H0003804-M01         17  ...                     6   \n",
       "15380  AACR-H0003804-M01         17  ...                    13   \n",
       "\n",
       "       Ideal_Edit_Window Editor     gRNA_id OnTarget_Azimuth_Score  \\\n",
       "0                  False    ABE      gRNA_0                   44.0   \n",
       "1                  False    ABE      gRNA_1                   54.0   \n",
       "2                  False    ABE      gRNA_2                   59.0   \n",
       "3                  False    ABE      gRNA_3                   59.0   \n",
       "4                   True    ABE      gRNA_4                   66.0   \n",
       "...                  ...    ...         ...                    ...   \n",
       "15376              False    ABE  gRNA_15376                   38.0   \n",
       "15377              False    ABE  gRNA_15377                   46.0   \n",
       "15378              False    ABE  gRNA_15378                   60.0   \n",
       "15379               True    ABE  gRNA_15379                   41.0   \n",
       "15380              False    ABE  gRNA_15380                   59.0   \n",
       "\n",
       "                                      sensor_wt  \\\n",
       "0      GAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCAGGCC   \n",
       "1      TGGAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCAGG   \n",
       "2      CATGGAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCA   \n",
       "3      ACAGCAAGGCTGGCATGGAGGAAGATCACACCTACGAGGT   \n",
       "4      GCTGGGCCTGCCCCTCTCCTTACCTCGTAGGTGTGATCTT   \n",
       "...                                         ...   \n",
       "15376  GCAATGGACCACACTTTTCCGGGGCAGGTGGGGCAGGAAG   \n",
       "15377  GAGCAATGGACCACACTTTTCCGGGGCAGGTGGGGCAGGA   \n",
       "15378  TAGAGCAATGGACCACACTTTTCCGGGGCAGGTGGGGCAG   \n",
       "15379  CCAAATCTTAGAGCAATGGACCACACTTTTCCGGGGCAGG   \n",
       "15380  TTAGAGCAATGGACCACACTTTTCCGGGGCAGGTGGGGCA   \n",
       "\n",
       "                                     sensor_alt  sensor_orientation  \\\n",
       "0      GAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCAGGCC  reverse-complement   \n",
       "1      TGGAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCAGG  reverse-complement   \n",
       "2      CATGGAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCA  reverse-complement   \n",
       "3      ACAGCAAGGCTGGCATGGAGGAAGATCACACCCACGAGGT  reverse-complement   \n",
       "4      GCTGGGCCTGCCCCTCTCCTTACCTCGCAGGTGTGATCTT  reverse-complement   \n",
       "...                                         ...                 ...   \n",
       "15376  GCAATGGACCACACTTTCCCGGGGCAGGTGGGGCAGGAAG  reverse-complement   \n",
       "15377  GAGCAATGGACCACACTTTCCCGGGGCAGGTGGGGCAGGA  reverse-complement   \n",
       "15378  TAGAGCAATGGACCACACTTTCCCGGGGCAGGTGGGGCAG  reverse-complement   \n",
       "15379  CCAAATCTTAGAGCAATGGACCACACTTTCCCGGGGCAGG  reverse-complement   \n",
       "15380  TTAGAGCAATGGACCACACTTTCCCGGGGCAGGTGGGGCA  reverse-complement   \n",
       "\n",
       "       sensor_error  contains_polyT_terminator  \n",
       "0          No Error                      False  \n",
       "1          No Error                      False  \n",
       "2          No Error                      False  \n",
       "3          No Error                      False  \n",
       "4          No Error                      False  \n",
       "...             ...                        ...  \n",
       "15376      No Error                      False  \n",
       "15377      No Error                      False  \n",
       "15378      No Error                      False  \n",
       "15379      No Error                      False  \n",
       "15380      No Error                      False  \n",
       "\n",
       "[15381 rows x 34 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all = df_all[df_all['Chromosome']!='M'].reset_index(drop=True)\n",
    "df_all['Reference_Allele'] = [str(x) for x in df_all['Reference_Allele']]\n",
    "df_all['Tumor_Seq_Allele2'] = [str(x) for x in df_all['Tumor_Seq_Allele2']]\n",
    "df_all['Start_Position'] = [int(x) for x in df_all['Start_Position']]\n",
    "df_all['End_Position'] = [int(x) for x in df_all['End_Position']]\n",
    "df_be_h = base.run_base(df_all, 'cBioPortal', chrom_dict_h, PAM='NGN',filtration='ABE+CBE', ideal_edit_window=[4, 8], auto_SNP_filter=True, proto_size=19,context_size=120, RE_sites=None, polyT_threshold=4, before_proto_context=5,sensor_length=40, sensor_orientation='reverse-complement', sensor=True)\n",
    "df_be_h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AACR-H0002363-M01',\n",
       " 'AACR-H0002999-M01',\n",
       " 'AACR-H0002588-M01',\n",
       " 'AACR-H0003738-M01',\n",
       " 'AACR-H0002965-M01',\n",
       " 'AACR-H0005030-M01',\n",
       " 'AACR-H0000928-M01',\n",
       " 'AACR-H0000291-M01',\n",
       " 'AACR-H0001653-M01',\n",
       " 'AACR-H0004831-M01',\n",
       " 'AACR-H0004152-M01',\n",
       " 'AACR-H0003347-M01',\n",
       " 'AACR-H0000973-M01',\n",
       " 'AACR-H0000313-M01',\n",
       " 'AACR-H0001629-M01']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(set(df_all['ID'])-set(df_be_h['ID']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_be_h['ID'] = [x[:-4] for x in df_be_h['ID']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_be_h.to_csv('be_human_result',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# For Mouse  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/BEGG/be_mouse.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_h_ref, path_m_ref = '/Users/kexindong/Documents/GitHub/Database/RefGenome/ncbi-2023-09-12/GCF_000001405.25_GRCh37.p13_genomic.fna.gz', '/Users/kexindong/Documents/GitHub/Database/RefGenome/mouse-2023-09-13/GCF_000001635.27_GRCm39_genomic.fna.gz'\n",
    "# remember to replace the paths with yours\n",
    "records_h, index_list_h = h2m.genome_loader(path_h_ref)\n",
    "records_m, index_list_m  = h2m.genome_loader(path_m_ref)\n",
    "path_h_anno, path_m_anno = '/Users/kexindong/Documents/GitHub/Database/Genecode/gencode_v19_GRCh37.db', '/Users/kexindong/Documents/GitHub/Database/Genecode/gencode_vm33_GRCm39.db'\n",
    "# remember to replace the paths with yours\n",
    "db_h, db_m = h2m.anno_loader(path_h_anno), h2m.anno_loader(path_m_anno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No error occurs.\n",
      "No error occurs.\n",
      "There were rows that could not be processed.\n"
     ]
    }
   ],
   "source": [
    "df_all.columns = ['gene_name_h','tx_id_h','start_h','end_h','ref_seq_h','alt_seq_h','type_h','index','chr_h']\n",
    "df_all = h2m.query_batch(df_all)[0]\n",
    "df_all = h2m.get_tx_batch(df_all, 'm')[0]\n",
    "df_result, df_fail = h2m.model_batch(df_all, records_h, index_list_h, records_m, index_list_m, db_h, db_m, 37, param='BE', memory_size=2000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>...</th>\n",
       "      <th>Protospacer_Location</th>\n",
       "      <th>Ideal_Edit_Window</th>\n",
       "      <th>Editor</th>\n",
       "      <th>gRNA_id</th>\n",
       "      <th>OnTarget_Azimuth_Score</th>\n",
       "      <th>sensor_wt</th>\n",
       "      <th>sensor_alt</th>\n",
       "      <th>sensor_orientation</th>\n",
       "      <th>sensor_error</th>\n",
       "      <th>contains_polyT_terminator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "      <td>145192497</td>\n",
       "      <td>145192497</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000001-M01-V01</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>19</td>\n",
       "      <td>False</td>\n",
       "      <td>CBE</td>\n",
       "      <td>gRNA_0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>GTGGTGGTTGGAGCTGGTGGCGTAGGCAAGAGCGCCTTGA</td>\n",
       "      <td>GTGGTGGTTGGAGCTGATGGCGTAGGCAAGAGCGCCTTGA</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "      <td>145192497</td>\n",
       "      <td>145192497</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000001-M01-V01</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>CBE</td>\n",
       "      <td>gRNA_1</td>\n",
       "      <td>68.0</td>\n",
       "      <td>GACTGAGTATAAACTTGTGGTGGTTGGAGCTGGTGGCGTA</td>\n",
       "      <td>GACTGAGTATAAACTTGTGGTGGTTGGAGCTGATGGCGTA</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>Pik3ca</td>\n",
       "      <td>ENSMUST00000029201.14</td>\n",
       "      <td>32502197</td>\n",
       "      <td>32502197</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000004-M01-V01</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>False</td>\n",
       "      <td>CBE</td>\n",
       "      <td>gRNA_2</td>\n",
       "      <td>15.0</td>\n",
       "      <td>CGGGACCCACTATCTGAAATCACTGAACAAGAGAAAGACT</td>\n",
       "      <td>CGGGACCCACTATCTGAAATCACTAAACAAGAGAAAGACT</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>Pik3ca</td>\n",
       "      <td>ENSMUST00000029201.14</td>\n",
       "      <td>32502197</td>\n",
       "      <td>32502197</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000004-M01-V01</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "      <td>CBE</td>\n",
       "      <td>gRNA_3</td>\n",
       "      <td>47.0</td>\n",
       "      <td>CACCCGGGACCCACTATCTGAAATCACTGAACAAGAGAAA</td>\n",
       "      <td>CACCCGGGACCCACTATCTGAAATCACTAAACAAGAGAAA</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>Pik3ca</td>\n",
       "      <td>ENSMUST00000029201.14</td>\n",
       "      <td>32502197</td>\n",
       "      <td>32502197</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000004-M01-V01</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>CBE</td>\n",
       "      <td>gRNA_4</td>\n",
       "      <td>47.0</td>\n",
       "      <td>TGCACCCGGGACCCACTATCTGAAATCACTGAACAAGAGA</td>\n",
       "      <td>TGCACCCGGGACCCACTATCTGAAATCACTAAACAAGAGA</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10102</th>\n",
       "      <td>1896</td>\n",
       "      <td>Xpo1</td>\n",
       "      <td>ENSMUST00000102869.8</td>\n",
       "      <td>23235114</td>\n",
       "      <td>23235114</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005028-M01-V01</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_10102</td>\n",
       "      <td>59.0</td>\n",
       "      <td>CTGAACAAAATGCCTACGGCATTTTTGAGCTATTTTAATA</td>\n",
       "      <td>CTGAACAAAATGCCTACGGCACTTTTGAGCTATTTTAATA</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10103</th>\n",
       "      <td>1896</td>\n",
       "      <td>Xpo1</td>\n",
       "      <td>ENSMUST00000102869.8</td>\n",
       "      <td>23235114</td>\n",
       "      <td>23235114</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005028-M01-V01</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_10103</td>\n",
       "      <td>60.0</td>\n",
       "      <td>CCTGAACAAAATGCCTACGGCATTTTTGAGCTATTTTAAT</td>\n",
       "      <td>CCTGAACAAAATGCCTACGGCACTTTTGAGCTATTTTAAT</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10104</th>\n",
       "      <td>1896</td>\n",
       "      <td>Xpo1</td>\n",
       "      <td>ENSMUST00000102869.8</td>\n",
       "      <td>23235114</td>\n",
       "      <td>23235114</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005028-M01-V01</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_10104</td>\n",
       "      <td>24.0</td>\n",
       "      <td>ACTTGAACCTGAACAAAATGCCTACGGCATTTTTGAGCTA</td>\n",
       "      <td>ACTTGAACCTGAACAAAATGCCTACGGCACTTTTGAGCTA</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10105</th>\n",
       "      <td>1896</td>\n",
       "      <td>Xpo1</td>\n",
       "      <td>ENSMUST00000102869.8</td>\n",
       "      <td>23235114</td>\n",
       "      <td>23235114</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005028-M01-V01</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_10105</td>\n",
       "      <td>33.0</td>\n",
       "      <td>CTCCAACTTGAACCTGAACAAAATGCCTACGGCATTTTTG</td>\n",
       "      <td>CTCCAACTTGAACCTGAACAAAATGCCTACGGCACTTTTG</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10106</th>\n",
       "      <td>1897</td>\n",
       "      <td>Bcl9</td>\n",
       "      <td>ENSMUST00000046521.14</td>\n",
       "      <td>97117238</td>\n",
       "      <td>97117238</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005030-M01-V01</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>CBE</td>\n",
       "      <td>gRNA_10106</td>\n",
       "      <td>26.0</td>\n",
       "      <td>TGAAGAGAAGAGGCGGAAACAGGAGCAGGTGGTCGTCCAG</td>\n",
       "      <td>TGAAGAGAAGAGGCGAAAACAGGAGCAGGTGGTCGTCCAG</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10107 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       mutation_idx Hugo_Symbol                tx_id_m  Start_Position  \\\n",
       "0                 0        Kras   ENSMUST00000111710.8       145192497   \n",
       "1                 0        Kras   ENSMUST00000111710.8       145192497   \n",
       "2                 1      Pik3ca  ENSMUST00000029201.14        32502197   \n",
       "3                 1      Pik3ca  ENSMUST00000029201.14        32502197   \n",
       "4                 1      Pik3ca  ENSMUST00000029201.14        32502197   \n",
       "...             ...         ...                    ...             ...   \n",
       "10102          1896        Xpo1   ENSMUST00000102869.8        23235114   \n",
       "10103          1896        Xpo1   ENSMUST00000102869.8        23235114   \n",
       "10104          1896        Xpo1   ENSMUST00000102869.8        23235114   \n",
       "10105          1896        Xpo1   ENSMUST00000102869.8        23235114   \n",
       "10106          1897        Bcl9  ENSMUST00000046521.14        97117238   \n",
       "\n",
       "       End_Position Reference_Allele Tumor_Seq_Allele2 Variant_Type  \\\n",
       "0         145192497                C                 T          SNP   \n",
       "1         145192497                C                 T          SNP   \n",
       "2          32502197                G                 A          SNP   \n",
       "3          32502197                G                 A          SNP   \n",
       "4          32502197                G                 A          SNP   \n",
       "...             ...              ...               ...          ...   \n",
       "10102      23235114                A                 G          SNP   \n",
       "10103      23235114                A                 G          SNP   \n",
       "10104      23235114                A                 G          SNP   \n",
       "10105      23235114                A                 G          SNP   \n",
       "10106      97117238                C                 T          SNP   \n",
       "\n",
       "                          ID Chromosome  ...  Protospacer_Location  \\\n",
       "0      AACR-H0000001-M01-V01          6  ...                    19   \n",
       "1      AACR-H0000001-M01-V01          6  ...                     3   \n",
       "2      AACR-H0000004-M01-V01          3  ...                    11   \n",
       "3      AACR-H0000004-M01-V01          3  ...                     7   \n",
       "4      AACR-H0000004-M01-V01          3  ...                     5   \n",
       "...                      ...        ...  ...                   ...   \n",
       "10102  AACR-H0005028-M01-V01         11  ...                    14   \n",
       "10103  AACR-H0005028-M01-V01         11  ...                    13   \n",
       "10104  AACR-H0005028-M01-V01         11  ...                     6   \n",
       "10105  AACR-H0005028-M01-V01         11  ...                     1   \n",
       "10106  AACR-H0005030-M01-V01          3  ...                    20   \n",
       "\n",
       "       Ideal_Edit_Window Editor     gRNA_id OnTarget_Azimuth_Score  \\\n",
       "0                  False    CBE      gRNA_0                   48.0   \n",
       "1                  False    CBE      gRNA_1                   68.0   \n",
       "2                  False    CBE      gRNA_2                   15.0   \n",
       "3                   True    CBE      gRNA_3                   47.0   \n",
       "4                   True    CBE      gRNA_4                   47.0   \n",
       "...                  ...    ...         ...                    ...   \n",
       "10102              False    ABE  gRNA_10102                   59.0   \n",
       "10103              False    ABE  gRNA_10103                   60.0   \n",
       "10104               True    ABE  gRNA_10104                   24.0   \n",
       "10105              False    ABE  gRNA_10105                   33.0   \n",
       "10106              False    CBE  gRNA_10106                   26.0   \n",
       "\n",
       "                                      sensor_wt  \\\n",
       "0      GTGGTGGTTGGAGCTGGTGGCGTAGGCAAGAGCGCCTTGA   \n",
       "1      GACTGAGTATAAACTTGTGGTGGTTGGAGCTGGTGGCGTA   \n",
       "2      CGGGACCCACTATCTGAAATCACTGAACAAGAGAAAGACT   \n",
       "3      CACCCGGGACCCACTATCTGAAATCACTGAACAAGAGAAA   \n",
       "4      TGCACCCGGGACCCACTATCTGAAATCACTGAACAAGAGA   \n",
       "...                                         ...   \n",
       "10102  CTGAACAAAATGCCTACGGCATTTTTGAGCTATTTTAATA   \n",
       "10103  CCTGAACAAAATGCCTACGGCATTTTTGAGCTATTTTAAT   \n",
       "10104  ACTTGAACCTGAACAAAATGCCTACGGCATTTTTGAGCTA   \n",
       "10105  CTCCAACTTGAACCTGAACAAAATGCCTACGGCATTTTTG   \n",
       "10106  TGAAGAGAAGAGGCGGAAACAGGAGCAGGTGGTCGTCCAG   \n",
       "\n",
       "                                     sensor_alt  sensor_orientation  \\\n",
       "0      GTGGTGGTTGGAGCTGATGGCGTAGGCAAGAGCGCCTTGA  reverse-complement   \n",
       "1      GACTGAGTATAAACTTGTGGTGGTTGGAGCTGATGGCGTA  reverse-complement   \n",
       "2      CGGGACCCACTATCTGAAATCACTAAACAAGAGAAAGACT  reverse-complement   \n",
       "3      CACCCGGGACCCACTATCTGAAATCACTAAACAAGAGAAA  reverse-complement   \n",
       "4      TGCACCCGGGACCCACTATCTGAAATCACTAAACAAGAGA  reverse-complement   \n",
       "...                                         ...                 ...   \n",
       "10102  CTGAACAAAATGCCTACGGCACTTTTGAGCTATTTTAATA  reverse-complement   \n",
       "10103  CCTGAACAAAATGCCTACGGCACTTTTGAGCTATTTTAAT  reverse-complement   \n",
       "10104  ACTTGAACCTGAACAAAATGCCTACGGCACTTTTGAGCTA  reverse-complement   \n",
       "10105  CTCCAACTTGAACCTGAACAAAATGCCTACGGCACTTTTG  reverse-complement   \n",
       "10106  TGAAGAGAAGAGGCGAAAACAGGAGCAGGTGGTCGTCCAG  reverse-complement   \n",
       "\n",
       "       sensor_error  contains_polyT_terminator  \n",
       "0          No Error                      False  \n",
       "1          No Error                      False  \n",
       "2          No Error                      False  \n",
       "3          No Error                      False  \n",
       "4          No Error                      False  \n",
       "...             ...                        ...  \n",
       "10102      No Error                      False  \n",
       "10103      No Error                      False  \n",
       "10104      No Error                      False  \n",
       "10105      No Error                       True  \n",
       "10106      No Error                      False  \n",
       "\n",
       "[10107 rows x 34 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all = df_all[df_all['Chromosome']!='M'].reset_index(drop=True)\n",
    "df_all['Reference_Allele'] = [str(x) for x in df_all['Reference_Allele']]\n",
    "df_all['Tumor_Seq_Allele2'] = [str(x) for x in df_all['Tumor_Seq_Allele2']]\n",
    "df_all['Start_Position'] = [int(x) for x in df_all['Start_Position']]\n",
    "df_all['End_Position'] = [int(x) for x in df_all['End_Position']]\n",
    "df_be_m = base.run_base(df_all, 'cBioPortal', chrom_dict_m, PAM='NGN',filtration='ABE+CBE', ideal_edit_window=[4, 8], auto_SNP_filter=True, proto_size=19,context_size=120, RE_sites=None, polyT_threshold=4, before_proto_context=5,sensor_length=40, sensor_orientation='reverse-complement', sensor=True)\n",
    "df_be_m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_be_m.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/BEGG/be_mouse_rusult.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bind mouse and human "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_be_h = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/BEGG/be_human_result')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.concat([df_be_h.drop(['gRNA_id','tx_id_h'], axis=1), df_be_m.drop(['gRNA_id','tx_id_m'], axis=1)]).to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/BEGG/be_result_all.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/kexindong/Documents/GitHub/h2m-public/.conda/lib/python3.10/site-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "import bioh2m as h2m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sift_h = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_human.csv')\n",
    "df_sift_m = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/PEGG/pe_mouse.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sift_h = df_sift_h[df_sift_h['Variant_Type']=='SNP'].reset_index(drop=True)\n",
    "df_sift_m = df_sift_m[df_sift_m['Variant_Type']=='SNP'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000001</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000002</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>ENST00000288602.6</td>\n",
       "      <td>140453136</td>\n",
       "      <td>140453136</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000003</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>ENST00000263967.3</td>\n",
       "      <td>178936091</td>\n",
       "      <td>178936091</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000004</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>ENST00000263967.3</td>\n",
       "      <td>178952085</td>\n",
       "      <td>178952085</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000005</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4471</th>\n",
       "      <td>FGFR3</td>\n",
       "      <td>ENST00000260795.2</td>\n",
       "      <td>1806162</td>\n",
       "      <td>1806162</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005025</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4472</th>\n",
       "      <td>NTRK2</td>\n",
       "      <td>ENST00000323115.4</td>\n",
       "      <td>87359980</td>\n",
       "      <td>87359980</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005026</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4473</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>55211111</td>\n",
       "      <td>55211111</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005027</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4474</th>\n",
       "      <td>XPO1</td>\n",
       "      <td>ENST00000401558.2</td>\n",
       "      <td>61719275</td>\n",
       "      <td>61719275</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005028</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4475</th>\n",
       "      <td>BCL9</td>\n",
       "      <td>ENST00000234739.3</td>\n",
       "      <td>147091416</td>\n",
       "      <td>147091416</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005030</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4476 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Hugo_Symbol            tx_id_h  Start_Position  End_Position  \\\n",
       "0           KRAS  ENST00000256078.4        25398284      25398284   \n",
       "1           KRAS  ENST00000256078.4        25398284      25398284   \n",
       "2           BRAF  ENST00000288602.6       140453136     140453136   \n",
       "3         PIK3CA  ENST00000263967.3       178936091     178936091   \n",
       "4         PIK3CA  ENST00000263967.3       178952085     178952085   \n",
       "...          ...                ...             ...           ...   \n",
       "4471       FGFR3  ENST00000260795.2         1806162       1806162   \n",
       "4472       NTRK2  ENST00000323115.4        87359980      87359980   \n",
       "4473        EGFR  ENST00000275493.2        55211111      55211111   \n",
       "4474        XPO1  ENST00000401558.2        61719275      61719275   \n",
       "4475        BCL9  ENST00000234739.3       147091416     147091416   \n",
       "\n",
       "     Reference_Allele Tumor_Seq_Allele2 Variant_Type             ID Chromosome  \n",
       "0                   C                 T          SNP  AACR-H0000001         12  \n",
       "1                   C                 A          SNP  AACR-H0000002         12  \n",
       "2                   A                 T          SNP  AACR-H0000003          7  \n",
       "3                   G                 A          SNP  AACR-H0000004          3  \n",
       "4                   A                 G          SNP  AACR-H0000005          3  \n",
       "...               ...               ...          ...            ...        ...  \n",
       "4471                C                 T          SNP  AACR-H0005025          4  \n",
       "4472                C                 T          SNP  AACR-H0005026          9  \n",
       "4473                C                 T          SNP  AACR-H0005027          7  \n",
       "4474                T                 C          SNP  AACR-H0005028          2  \n",
       "4475                G                 A          SNP  AACR-H0005030          1  \n",
       "\n",
       "[4476 rows x 9 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sift_h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query human gene: KMT2D;\n",
      "Mouse ortholog(s): Kmt2d;\n",
      "Homology type: one2one;\n",
      "Sequence Simalarity(%):88.9471.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'gene_name_m': 'Kmt2d',\n",
       "  'gene_id_m': 'ENSMUSG00000048154',\n",
       "  'homology_type': 'ortholog_one2one',\n",
       "  'similarity': 88.9471}]"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h2m.query('KMT2D')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query human gene: KIT;\n",
      "Mouse ortholog(s): Kit;\n",
      "Homology type: one2one;\n",
      "Sequence Simalarity(%):82.7869.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'gene_name_m': 'Kit',\n",
       "  'gene_id_m': 'ENSMUSG00000005672',\n",
       "  'homology_type': 'ortholog_one2one',\n",
       "  'similarity': 82.7869}]"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h2m.query('KIT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>tx_id_m</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000001-M01-V01</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Kras</td>\n",
       "      <td>ENSMUST00000111710.8</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>145192497.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000002-M01-V01</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Braf</td>\n",
       "      <td>ENSMUST00000002487.15</td>\n",
       "      <td>39604717.0</td>\n",
       "      <td>39604717.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000003-M01-V01</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Pik3ca</td>\n",
       "      <td>ENSMUST00000029201.14</td>\n",
       "      <td>32502197.0</td>\n",
       "      <td>32502197.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000004-M01-V01</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Pik3ca</td>\n",
       "      <td>ENSMUST00000029201.14</td>\n",
       "      <td>32516913.0</td>\n",
       "      <td>32516913.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000005-M01-V01</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3173</th>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUST00000108658.10</td>\n",
       "      <td>69480068.0</td>\n",
       "      <td>69480068.0</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005024-M01-V01</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3174</th>\n",
       "      <td>Ntrk2</td>\n",
       "      <td>ENSMUST00000079828.7</td>\n",
       "      <td>59022231.0</td>\n",
       "      <td>59022231.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005026-M01-V01</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3175</th>\n",
       "      <td>Egfr</td>\n",
       "      <td>ENSMUST00000020329.13</td>\n",
       "      <td>16810181.0</td>\n",
       "      <td>16810181.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005027-M01-V01</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3176</th>\n",
       "      <td>Xpo1</td>\n",
       "      <td>ENSMUST00000102869.8</td>\n",
       "      <td>23235114.0</td>\n",
       "      <td>23235114.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005028-M01-V01</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3177</th>\n",
       "      <td>Bcl9</td>\n",
       "      <td>ENSMUST00000046521.14</td>\n",
       "      <td>97117238.0</td>\n",
       "      <td>97117238.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005030-M01-V01</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3178 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Hugo_Symbol                tx_id_m  Start_Position  End_Position  \\\n",
       "0           Kras   ENSMUST00000111710.8     145192497.0   145192497.0   \n",
       "1           Kras   ENSMUST00000111710.8     145192497.0   145192497.0   \n",
       "2           Braf  ENSMUST00000002487.15      39604717.0    39604717.0   \n",
       "3         Pik3ca  ENSMUST00000029201.14      32502197.0    32502197.0   \n",
       "4         Pik3ca  ENSMUST00000029201.14      32516913.0    32516913.0   \n",
       "...          ...                    ...             ...           ...   \n",
       "3173       Trp53  ENSMUST00000108658.10      69480068.0    69480068.0   \n",
       "3174       Ntrk2   ENSMUST00000079828.7      59022231.0    59022231.0   \n",
       "3175        Egfr  ENSMUST00000020329.13      16810181.0    16810181.0   \n",
       "3176        Xpo1   ENSMUST00000102869.8      23235114.0    23235114.0   \n",
       "3177        Bcl9  ENSMUST00000046521.14      97117238.0    97117238.0   \n",
       "\n",
       "     Reference_Allele Tumor_Seq_Allele2 Variant_Type                     ID  \\\n",
       "0                   C                 T          SNP  AACR-H0000001-M01-V01   \n",
       "1                   C                 A          SNP  AACR-H0000002-M01-V01   \n",
       "2                   A                 T          SNP  AACR-H0000003-M01-V01   \n",
       "3                   G                 A          SNP  AACR-H0000004-M01-V01   \n",
       "4                   A                 G          SNP  AACR-H0000005-M01-V01   \n",
       "...               ...               ...          ...                    ...   \n",
       "3173                A                 C          SNP  AACR-H0005024-M01-V01   \n",
       "3174                C                 T          SNP  AACR-H0005026-M01-V01   \n",
       "3175                C                 T          SNP  AACR-H0005027-M01-V01   \n",
       "3176                A                 G          SNP  AACR-H0005028-M01-V01   \n",
       "3177                C                 T          SNP  AACR-H0005030-M01-V01   \n",
       "\n",
       "     Chromosome  \n",
       "0             6  \n",
       "1             6  \n",
       "2             6  \n",
       "3             3  \n",
       "4             3  \n",
       "...         ...  \n",
       "3173         11  \n",
       "3174         13  \n",
       "3175         11  \n",
       "3176         11  \n",
       "3177          3  \n",
       "\n",
       "[3178 rows x 9 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sift_m"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Build all SNP into a VCF file "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sift_h['Reference_Allele'] = df_sift_h['Reference_Allele'].fillna('')\n",
    "df_sift_m['Reference_Allele'] = df_sift_m['Reference_Allele'].fillna('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "sift_h = df_sift_h[['Chromosome',\n",
    "                       'Start_Position',\n",
    "                       'ID',\n",
    "                       'Reference_Allele',\n",
    "                       'Tumor_Seq_Allele2']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [],
   "source": [
    "sift_h = sift_h.sort_values('Chromosome').reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "sift_h['QUAL'] = None\n",
    "sift_h['FILTER'] = 'PASS'\n",
    "sift_h['INFO'] = '-'\n",
    "sift_h['FORMAT'] = '-'\n",
    "sift_h['MP96-Ripoll-MRC-Harwell'] = '-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "sift_h = sift_h.sort_values(by=['Chromosome','Start_Position']).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "sift_h.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/SIFT/sift_h.txt',sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##fileformat=VCFv4.1\t\t\t\t\t\t\t\t\t\n",
    "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tMP96-Ripoll-MRC-Harwell"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cut KIT AACR #>=5 and run SIFT  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/df_result_cleaned.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_binded = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/database-output-final/df_result_cleaned.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_input = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/aacr_input_original.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_kit = df_binded[(df_binded['gene_name_h']=='KIT') & (df_binded['status']==True) & (df_binded['type_h']=='SNP') & (df_binded['Database']=='AACR')]\n",
    "del df_binded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_input_sub = df_input[(df_input['gene_name_h']=='KIT') & (df_input['count']>=5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_kit = df_kit[df_kit['ID_human'].isin(df_input_sub['ID'])]\n",
    "df_kit = df_kit.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_name_h</th>\n",
       "      <th>gene_id_h</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>chr_h</th>\n",
       "      <th>exon_num_h</th>\n",
       "      <th>strand_h</th>\n",
       "      <th>match</th>\n",
       "      <th>start_h</th>\n",
       "      <th>end_h</th>\n",
       "      <th>ref_seq_h</th>\n",
       "      <th>...</th>\n",
       "      <th>start_m</th>\n",
       "      <th>end_m</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>alt_seq_m</th>\n",
       "      <th>HGVSc_m</th>\n",
       "      <th>HGVSp_m</th>\n",
       "      <th>ID_human</th>\n",
       "      <th>ID_mouse</th>\n",
       "      <th>Database</th>\n",
       "      <th>ID_db</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KIT</td>\n",
       "      <td>ENSG00000157404.11</td>\n",
       "      <td>ENST00000288135.5</td>\n",
       "      <td>chr4</td>\n",
       "      <td>21</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55593610</td>\n",
       "      <td>55593610</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>75801387.0</td>\n",
       "      <td>75801387.0</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSMUST00000005815.7:c.1685T&gt;G</td>\n",
       "      <td>V562G</td>\n",
       "      <td>AACR-H0000156</td>\n",
       "      <td>AACR-H0000156-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KIT</td>\n",
       "      <td>ENSG00000157404.11</td>\n",
       "      <td>ENST00000288135.5</td>\n",
       "      <td>chr4</td>\n",
       "      <td>21</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55599159</td>\n",
       "      <td>55599159</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>75810135.0</td>\n",
       "      <td>75810135.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000005815.7:c.2368-72G&gt;A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AACR-H0000368</td>\n",
       "      <td>AACR-H0000368-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KIT</td>\n",
       "      <td>ENSG00000157404.11</td>\n",
       "      <td>ENST00000288135.5</td>\n",
       "      <td>chr4</td>\n",
       "      <td>21</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55599321</td>\n",
       "      <td>55599321</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>75810292.0</td>\n",
       "      <td>75810292.0</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000005815.7:c.2453A&gt;T</td>\n",
       "      <td>D818V</td>\n",
       "      <td>AACR-H0000390</td>\n",
       "      <td>AACR-H0000390-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KIT</td>\n",
       "      <td>ENSG00000157404.11</td>\n",
       "      <td>ENST00000288135.5</td>\n",
       "      <td>chr4</td>\n",
       "      <td>21</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55593661</td>\n",
       "      <td>55593661</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>75801438.0</td>\n",
       "      <td>75801438.0</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSMUST00000005815.7:c.1736T&gt;C</td>\n",
       "      <td>L579P</td>\n",
       "      <td>AACR-H0000420</td>\n",
       "      <td>AACR-H0000420-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KIT</td>\n",
       "      <td>ENSG00000157404.11</td>\n",
       "      <td>ENST00000288135.5</td>\n",
       "      <td>chr4</td>\n",
       "      <td>21</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55593613</td>\n",
       "      <td>55593613</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>75801390.0</td>\n",
       "      <td>75801390.0</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSMUST00000005815.7:c.1688T&gt;G</td>\n",
       "      <td>V563G</td>\n",
       "      <td>AACR-H0000423</td>\n",
       "      <td>AACR-H0000423-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>KIT</td>\n",
       "      <td>ENSG00000157404.11</td>\n",
       "      <td>ENST00000288135.5</td>\n",
       "      <td>chr4</td>\n",
       "      <td>21</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55604668</td>\n",
       "      <td>55604668</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>75815150.0</td>\n",
       "      <td>75815151.0</td>\n",
       "      <td>CG</td>\n",
       "      <td>AT</td>\n",
       "      <td>ENSMUST00000005815.7:c.2885_2886CG&gt;AT</td>\n",
       "      <td>S962Y</td>\n",
       "      <td>AACR-H0048399</td>\n",
       "      <td>AACR-H0048399-M01-V02</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>KIT</td>\n",
       "      <td>ENSG00000157404.11</td>\n",
       "      <td>ENST00000288135.5</td>\n",
       "      <td>chr4</td>\n",
       "      <td>21</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55604638</td>\n",
       "      <td>55604638</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>75815108.0</td>\n",
       "      <td>75815108.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000005815.7:c.2843C&gt;T</td>\n",
       "      <td>P948L</td>\n",
       "      <td>AACR-H0048772</td>\n",
       "      <td>AACR-H0048772-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>KIT</td>\n",
       "      <td>ENSG00000157404.11</td>\n",
       "      <td>ENST00000288135.5</td>\n",
       "      <td>chr4</td>\n",
       "      <td>21</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55603445</td>\n",
       "      <td>55603445</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>75813984.0</td>\n",
       "      <td>75813984.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSMUST00000005815.7:c.2807A&gt;G</td>\n",
       "      <td>H936R</td>\n",
       "      <td>AACR-H0048819</td>\n",
       "      <td>AACR-H0048819-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>KIT</td>\n",
       "      <td>ENSG00000157404.11</td>\n",
       "      <td>ENST00000288135.5</td>\n",
       "      <td>chr4</td>\n",
       "      <td>21</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55593586</td>\n",
       "      <td>55593586</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>75801363.0</td>\n",
       "      <td>75801363.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSMUST00000005815.7:c.1661C&gt;T</td>\n",
       "      <td>P554L</td>\n",
       "      <td>AACR-H0049226</td>\n",
       "      <td>AACR-H0049226-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>KIT</td>\n",
       "      <td>ENSG00000157404.11</td>\n",
       "      <td>ENST00000288135.5</td>\n",
       "      <td>chr4</td>\n",
       "      <td>21</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>55561914</td>\n",
       "      <td>55561914</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>75767925.0</td>\n",
       "      <td>75767925.0</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSMUST00000005815.7:c.307G&gt;A</td>\n",
       "      <td>G103S</td>\n",
       "      <td>AACR-H0049395</td>\n",
       "      <td>AACR-H0049395-M01-V01</td>\n",
       "      <td>AACR</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129 rows × 46 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_name_h           gene_id_h            tx_id_h chr_h  exon_num_h  \\\n",
       "0           KIT  ENSG00000157404.11  ENST00000288135.5  chr4          21   \n",
       "1           KIT  ENSG00000157404.11  ENST00000288135.5  chr4          21   \n",
       "2           KIT  ENSG00000157404.11  ENST00000288135.5  chr4          21   \n",
       "3           KIT  ENSG00000157404.11  ENST00000288135.5  chr4          21   \n",
       "4           KIT  ENSG00000157404.11  ENST00000288135.5  chr4          21   \n",
       "..          ...                 ...                ...   ...         ...   \n",
       "124         KIT  ENSG00000157404.11  ENST00000288135.5  chr4          21   \n",
       "125         KIT  ENSG00000157404.11  ENST00000288135.5  chr4          21   \n",
       "126         KIT  ENSG00000157404.11  ENST00000288135.5  chr4          21   \n",
       "127         KIT  ENSG00000157404.11  ENST00000288135.5  chr4          21   \n",
       "128         KIT  ENSG00000157404.11  ENST00000288135.5  chr4          21   \n",
       "\n",
       "    strand_h  match   start_h     end_h ref_seq_h  ...     start_m  \\\n",
       "0          +   True  55593610  55593610         T  ...  75801387.0   \n",
       "1          +   True  55599159  55599159         G  ...  75810135.0   \n",
       "2          +   True  55599321  55599321         A  ...  75810292.0   \n",
       "3          +   True  55593661  55593661         T  ...  75801438.0   \n",
       "4          +   True  55593613  55593613         T  ...  75801390.0   \n",
       "..       ...    ...       ...       ...       ...  ...         ...   \n",
       "124        +   True  55604668  55604668         C  ...  75815150.0   \n",
       "125        +   True  55604638  55604638         C  ...  75815108.0   \n",
       "126        +   True  55603445  55603445         A  ...  75813984.0   \n",
       "127        +   True  55593586  55593586         C  ...  75801363.0   \n",
       "128        +   True  55561914  55561914         G  ...  75767925.0   \n",
       "\n",
       "          end_m ref_seq_m alt_seq_m                                HGVSc_m  \\\n",
       "0    75801387.0         T         G         ENSMUST00000005815.7:c.1685T>G   \n",
       "1    75810135.0         G         A      ENSMUST00000005815.7:c.2368-72G>A   \n",
       "2    75810292.0         A         T         ENSMUST00000005815.7:c.2453A>T   \n",
       "3    75801438.0         T         C         ENSMUST00000005815.7:c.1736T>C   \n",
       "4    75801390.0         T         G         ENSMUST00000005815.7:c.1688T>G   \n",
       "..          ...       ...       ...                                    ...   \n",
       "124  75815151.0        CG        AT  ENSMUST00000005815.7:c.2885_2886CG>AT   \n",
       "125  75815108.0         C         T         ENSMUST00000005815.7:c.2843C>T   \n",
       "126  75813984.0         A         G         ENSMUST00000005815.7:c.2807A>G   \n",
       "127  75801363.0         C         T         ENSMUST00000005815.7:c.1661C>T   \n",
       "128  75767925.0         G         A          ENSMUST00000005815.7:c.307G>A   \n",
       "\n",
       "    HGVSp_m       ID_human               ID_mouse Database ID_db  \n",
       "0     V562G  AACR-H0000156  AACR-H0000156-M01-V01     AACR   NaN  \n",
       "1       NaN  AACR-H0000368  AACR-H0000368-M01-V01     AACR   NaN  \n",
       "2     D818V  AACR-H0000390  AACR-H0000390-M01-V01     AACR   NaN  \n",
       "3     L579P  AACR-H0000420  AACR-H0000420-M01-V01     AACR   NaN  \n",
       "4     V563G  AACR-H0000423  AACR-H0000423-M01-V01     AACR   NaN  \n",
       "..      ...            ...                    ...      ...   ...  \n",
       "124   S962Y  AACR-H0048399  AACR-H0048399-M01-V02     AACR   NaN  \n",
       "125   P948L  AACR-H0048772  AACR-H0048772-M01-V01     AACR   NaN  \n",
       "126   H936R  AACR-H0048819  AACR-H0048819-M01-V01     AACR   NaN  \n",
       "127   P554L  AACR-H0049226  AACR-H0049226-M01-V01     AACR   NaN  \n",
       "128   G103S  AACR-H0049395  AACR-H0049395-M01-V01     AACR   NaN  \n",
       "\n",
       "[129 rows x 46 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_kit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "sift_h = df_kit[['chr_h',\n",
    "                       'start_h',\n",
    "                       'ID_human',\n",
    "                       'ref_seq_h',\n",
    "                       'alt_seq_h']].drop_duplicates().reset_index(drop=True)\n",
    "sift_h['QUAL'] = None\n",
    "sift_h['FILTER'] = 'PASS'\n",
    "sift_h['INFO'] = '-'\n",
    "sift_h['FORMAT'] = '-'\n",
    "sift_h['MP96-Ripoll-MRC-Harwell'] = '-'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sift_h = df_kit[['chr_h',\n",
    "                       'start_h',\n",
    "                       'ID_human',\n",
    "                       'ref_seq_h',\n",
    "                       'alt_seq_h']].drop_duplicates().reset_index(drop=True)\n",
    "sift_h['QUAL'] = None\n",
    "sift_h['FILTER'] = 'PASS'\n",
    "sift_h['INFO'] = '-'\n",
    "sift_h['FORMAT'] = '-'\n",
    "sift_h['MP96-Ripoll-MRC-Harwell'] = '-'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [],
   "source": [
    "sift_h = sift_h.sort_values('start_h').reset_index(drop=True)\n",
    "sift_h.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/SIFT/sift_h.txt',sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "sift_m = df_kit[['chr_m',\n",
    "                       'start_m',\n",
    "                       'ID_human',\n",
    "                       'ref_seq_m',\n",
    "                       'alt_seq_m']].dropna().reset_index(drop=True).drop_duplicates().reset_index(drop=True)\n",
    "sift_m['QUAL'] = None\n",
    "sift_m['FILTER'] = 'PASS'\n",
    "sift_m['INFO'] = '-'\n",
    "sift_m['FORMAT'] = '-'\n",
    "sift_m['start_m'] = [int(x) for x in sift_m['start_m']] \n",
    "sift_m['MP96-Ripoll-MRC-Harwell'] = '-'\n",
    "sift_m = sift_m.sort_values('start_m').reset_index(drop=True)\n",
    "sift_m.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/SIFT/sift_m.txt',sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr_m</th>\n",
       "      <th>start_m</th>\n",
       "      <th>ID_human</th>\n",
       "      <th>ref_seq_m</th>\n",
       "      <th>alt_seq_m</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>MP96-Ripoll-MRC-Harwell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75735766</td>\n",
       "      <td>AACR-H0010107</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>None</td>\n",
       "      <td>PASS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75735782</td>\n",
       "      <td>AACR-H0003061</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>None</td>\n",
       "      <td>PASS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75767719</td>\n",
       "      <td>AACR-H0004152</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>None</td>\n",
       "      <td>PASS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75767760</td>\n",
       "      <td>AACR-H0012039</td>\n",
       "      <td>GTT</td>\n",
       "      <td>ATG</td>\n",
       "      <td>None</td>\n",
       "      <td>PASS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75767760</td>\n",
       "      <td>AACR-H0010103</td>\n",
       "      <td>GTT</td>\n",
       "      <td>CTC</td>\n",
       "      <td>None</td>\n",
       "      <td>PASS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75815140</td>\n",
       "      <td>AACR-H0010923</td>\n",
       "      <td>AGG</td>\n",
       "      <td>CAA</td>\n",
       "      <td>None</td>\n",
       "      <td>PASS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75815150</td>\n",
       "      <td>AACR-H0048399</td>\n",
       "      <td>CG</td>\n",
       "      <td>AC</td>\n",
       "      <td>None</td>\n",
       "      <td>PASS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75815150</td>\n",
       "      <td>AACR-H0048399</td>\n",
       "      <td>CG</td>\n",
       "      <td>AT</td>\n",
       "      <td>None</td>\n",
       "      <td>PASS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75815155</td>\n",
       "      <td>AACR-H0014362</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>None</td>\n",
       "      <td>PASS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75815197</td>\n",
       "      <td>AACR-H0012190</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>None</td>\n",
       "      <td>PASS</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>129 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr_m   start_m       ID_human ref_seq_m alt_seq_m  QUAL FILTER INFO  \\\n",
       "0    chr5  75735766  AACR-H0010107         C         T  None   PASS    -   \n",
       "1    chr5  75735782  AACR-H0003061         G         A  None   PASS    -   \n",
       "2    chr5  75767719  AACR-H0004152         C         T  None   PASS    -   \n",
       "3    chr5  75767760  AACR-H0012039       GTT       ATG  None   PASS    -   \n",
       "4    chr5  75767760  AACR-H0010103       GTT       CTC  None   PASS    -   \n",
       "..    ...       ...            ...       ...       ...   ...    ...  ...   \n",
       "124  chr5  75815140  AACR-H0010923       AGG       CAA  None   PASS    -   \n",
       "125  chr5  75815150  AACR-H0048399        CG        AC  None   PASS    -   \n",
       "126  chr5  75815150  AACR-H0048399        CG        AT  None   PASS    -   \n",
       "127  chr5  75815155  AACR-H0014362         G         A  None   PASS    -   \n",
       "128  chr5  75815197  AACR-H0012190         G         A  None   PASS    -   \n",
       "\n",
       "    FORMAT MP96-Ripoll-MRC-Harwell  \n",
       "0        -                       -  \n",
       "1        -                       -  \n",
       "2        -                       -  \n",
       "3        -                       -  \n",
       "4        -                       -  \n",
       "..     ...                     ...  \n",
       "124      -                       -  \n",
       "125      -                       -  \n",
       "126      -                       -  \n",
       "127      -                       -  \n",
       "128      -                       -  \n",
       "\n",
       "[129 rows x 10 columns]"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sift_m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "list_of_print = []\n",
    "for i,x in sift_m.iterrows():\n",
    "    xx = int(x['start_m'])\n",
    "    list_of_print.append(f'chr5 {xx} {xx}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "129"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(list_of_print)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr5 75896426 75896426\n",
      "chr5 75896442 75896442\n",
      "chr5 75928379 75928379\n",
      "chr5 75928420 75928420\n",
      "chr5 75928420 75928420\n",
      "chr5 75928420 75928420\n",
      "chr5 75928420 75928420\n",
      "chr5 75928420 75928420\n",
      "chr5 75928420 75928420\n",
      "chr5 75928432 75928432\n",
      "chr5 75928460 75928460\n",
      "chr5 75928568 75928568\n",
      "chr5 75928585 75928585\n",
      "chr5 75930601 75930601\n",
      "chr5 75930643 75930643\n",
      "chr5 75930655 75930655\n",
      "chr5 75930656 75930656\n",
      "chr5 75930692 75930692\n",
      "chr5 75930721 75930721\n",
      "chr5 75930727 75930727\n",
      "chr5 75930730 75930730\n",
      "chr5 75930742 75930742\n",
      "chr5 75930756 75930756\n",
      "chr5 75930781 75930781\n",
      "chr5 75930782 75930782\n",
      "chr5 75930793 75930793\n",
      "chr5 75932044 75932044\n",
      "chr5 75932216 75932216\n",
      "chr5 75936681 75936681\n",
      "chr5 75936699 75936699\n",
      "chr5 75936727 75936727\n",
      "chr5 75942277 75942277\n",
      "chr5 75944332 75944332\n",
      "chr5 75944349 75944349\n",
      "chr5 75958609 75958609\n",
      "chr5 75958635 75958635\n",
      "chr5 75958637 75958637\n",
      "chr5 75960341 75960341\n",
      "chr5 75960424 75960424\n",
      "chr5 75960435 75960435\n",
      "chr5 75960464 75960464\n",
      "chr5 75961863 75961863\n",
      "chr5 75961893 75961893\n",
      "chr5 75961910 75961910\n",
      "chr5 75962023 75962023\n",
      "chr5 75962034 75962034\n",
      "chr5 75962038 75962038\n",
      "chr5 75962040 75962040\n",
      "chr5 75962040 75962040\n",
      "chr5 75962040 75962040\n",
      "chr5 75962043 75962043\n",
      "chr5 75962045 75962045\n",
      "chr5 75962047 75962047\n",
      "chr5 75962047 75962047\n",
      "chr5 75962047 75962047\n",
      "chr5 75962050 75962050\n",
      "chr5 75962050 75962050\n",
      "chr5 75962059 75962059\n",
      "chr5 75962065 75962065\n",
      "chr5 75962074 75962074\n",
      "chr5 75962098 75962098\n",
      "chr5 75962100 75962100\n",
      "chr5 75962502 75962502\n",
      "chr5 75962516 75962516\n",
      "chr5 75962534 75962534\n",
      "chr5 75962667 75962667\n",
      "chr5 75962667 75962667\n",
      "chr5 75962691 75962691\n",
      "chr5 75962728 75962728\n",
      "chr5 75967166 75967166\n",
      "chr5 75967174 75967174\n",
      "chr5 75967221 75967221\n",
      "chr5 75967222 75967222\n",
      "chr5 75969121 75969121\n",
      "chr5 75969132 75969132\n",
      "chr5 75969145 75969145\n",
      "chr5 75969716 75969716\n",
      "chr5 75969736 75969736\n",
      "chr5 75969754 75969754\n",
      "chr5 75969766 75969766\n",
      "chr5 75969820 75969820\n",
      "chr5 75969822 75969822\n",
      "chr5 75969823 75969823\n",
      "chr5 75970764 75970764\n",
      "chr5 75970795 75970795\n",
      "chr5 75970899 75970899\n",
      "chr5 75970916 75970916\n",
      "chr5 75970951 75970951\n",
      "chr5 75970951 75970951\n",
      "chr5 75970952 75970952\n",
      "chr5 75970953 75970953\n",
      "chr5 75970963 75970963\n",
      "chr5 75970964 75970964\n",
      "chr5 75970964 75970964\n",
      "chr5 75970965 75970965\n",
      "chr5 75970969 75970969\n",
      "chr5 75970970 75970970\n",
      "chr5 75970971 75970971\n",
      "chr5 75970971 75970971\n",
      "chr5 75970972 75970972\n",
      "chr5 75970989 75970989\n",
      "chr5 75970990 75970990\n",
      "chr5 75973870 75973870\n",
      "chr5 75973873 75973873\n",
      "chr5 75973889 75973889\n",
      "chr5 75973925 75973925\n",
      "chr5 75973939 75973939\n",
      "chr5 75973971 75973971\n",
      "chr5 75974102 75974102\n",
      "chr5 75974106 75974106\n",
      "chr5 75974126 75974126\n",
      "chr5 75974142 75974142\n",
      "chr5 75974143 75974143\n",
      "chr5 75974163 75974163\n",
      "chr5 75974574 75974574\n",
      "chr5 75974575 75974575\n",
      "chr5 75974589 75974589\n",
      "chr5 75974617 75974617\n",
      "chr5 75974644 75974644\n",
      "chr5 75975736 75975736\n",
      "chr5 75975768 75975768\n",
      "chr5 75975779 75975779\n",
      "chr5 75975800 75975800\n",
      "chr5 75975800 75975800\n",
      "chr5 75975800 75975800\n",
      "chr5 75975810 75975810\n",
      "chr5 75975810 75975810\n",
      "chr5 75975815 75975815\n",
      "chr5 75975857 75975857\n"
     ]
    }
   ],
   "source": [
    "print('\\n'.join(list_of_print))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# 读取 BED 文件\n",
    "bed_file_path = 'path/to/your/file.bed'\n",
    "bed_df = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/SIFT/tgWLU4k4xoKksrlj.bed', sep='\\t', header=None)\n",
    "\n",
    "# 设置列名（根据 BED 文件格式）\n",
    "#bed_df.columns = ['chrom', 'chromStart', 'chromEnd', 'name', 'score', 'strand']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75609407</td>\n",
       "      <td>75609407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75609410</td>\n",
       "      <td>75609410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75609422</td>\n",
       "      <td>75609422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75609436</td>\n",
       "      <td>75609436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75609461</td>\n",
       "      <td>75609461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75654480</td>\n",
       "      <td>75654480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75654490</td>\n",
       "      <td>75654490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75654490</td>\n",
       "      <td>75654490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75654495</td>\n",
       "      <td>75654495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>chr5</td>\n",
       "      <td>75654537</td>\n",
       "      <td>75654537</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        0         1         2\n",
       "0    chr5  75609407  75609407\n",
       "1    chr5  75609410  75609410\n",
       "2    chr5  75609422  75609422\n",
       "3    chr5  75609436  75609436\n",
       "4    chr5  75609461  75609461\n",
       "..    ...       ...       ...\n",
       "105  chr5  75654480  75654480\n",
       "106  chr5  75654490  75654490\n",
       "107  chr5  75654490  75654490\n",
       "108  chr5  75654495  75654495\n",
       "109  chr5  75654537  75654537\n",
       "\n",
       "[110 rows x 3 columns]"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "129"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(list(bed_df.iloc[:,1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "sift_m['start_m'] = list(bed_df.iloc[:,1])\n",
    "sift_m = sift_m.sort_values('start_m').reset_index(drop=True)\n",
    "sift_m.to_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/SIFT/sift_m.txt',sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
