{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script generates base editing guide RNA library for a subset of frequently mutated mutations (for both human and mouse).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import bioh2m as h2m\n",
    "from pegg import prime\n",
    "from pegg import base\n",
    "import pandas as pd\n",
    "df_all = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/BEGG/be_human.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# For Human  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Chromosome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>ENST00000256078.4</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000001</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>ENST00000263967.3</td>\n",
       "      <td>178936091</td>\n",
       "      <td>178936091</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000004</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>ENST00000263967.3</td>\n",
       "      <td>178952085</td>\n",
       "      <td>178952085</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000005</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TP53</td>\n",
       "      <td>ENST00000269305.4</td>\n",
       "      <td>7578406</td>\n",
       "      <td>7578406</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000007</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PIK3CA</td>\n",
       "      <td>ENST00000263967.3</td>\n",
       "      <td>178936082</td>\n",
       "      <td>178936082</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0000008</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3138</th>\n",
       "      <td>FGFR3</td>\n",
       "      <td>ENST00000260795.2</td>\n",
       "      <td>1806162</td>\n",
       "      <td>1806162</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005025</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3139</th>\n",
       "      <td>NTRK2</td>\n",
       "      <td>ENST00000323115.4</td>\n",
       "      <td>87359980</td>\n",
       "      <td>87359980</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005026</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3140</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>ENST00000275493.2</td>\n",
       "      <td>55211111</td>\n",
       "      <td>55211111</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005027</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3141</th>\n",
       "      <td>XPO1</td>\n",
       "      <td>ENST00000401558.2</td>\n",
       "      <td>61719275</td>\n",
       "      <td>61719275</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005028</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3142</th>\n",
       "      <td>BCL9</td>\n",
       "      <td>ENST00000234739.3</td>\n",
       "      <td>147091416</td>\n",
       "      <td>147091416</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0005030</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3143 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Hugo_Symbol            tx_id_h  Start_Position  End_Position  \\\n",
       "0           KRAS  ENST00000256078.4        25398284      25398284   \n",
       "1         PIK3CA  ENST00000263967.3       178936091     178936091   \n",
       "2         PIK3CA  ENST00000263967.3       178952085     178952085   \n",
       "3           TP53  ENST00000269305.4         7578406       7578406   \n",
       "4         PIK3CA  ENST00000263967.3       178936082     178936082   \n",
       "...          ...                ...             ...           ...   \n",
       "3138       FGFR3  ENST00000260795.2         1806162       1806162   \n",
       "3139       NTRK2  ENST00000323115.4        87359980      87359980   \n",
       "3140        EGFR  ENST00000275493.2        55211111      55211111   \n",
       "3141        XPO1  ENST00000401558.2        61719275      61719275   \n",
       "3142        BCL9  ENST00000234739.3       147091416     147091416   \n",
       "\n",
       "     Reference_Allele Tumor_Seq_Allele2 Variant_Type             ID Chromosome  \n",
       "0                   C                 T          SNP  AACR-H0000001         12  \n",
       "1                   G                 A          SNP  AACR-H0000004          3  \n",
       "2                   A                 G          SNP  AACR-H0000005          3  \n",
       "3                   C                 T          SNP  AACR-H0000007         17  \n",
       "4                   G                 A          SNP  AACR-H0000008          3  \n",
       "...               ...               ...          ...            ...        ...  \n",
       "3138                C                 T          SNP  AACR-H0005025          4  \n",
       "3139                C                 T          SNP  AACR-H0005026          9  \n",
       "3140                C                 T          SNP  AACR-H0005027          7  \n",
       "3141                T                 C          SNP  AACR-H0005028          2  \n",
       "3142                G                 A          SNP  AACR-H0005030          1  \n",
       "\n",
       "[3143 rows x 9 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_h_ref, path_m_ref = '/Users/kexindong/Documents/GitHub/Database/RefGenome/ncbi-2023-09-12/GCF_000001405.25_GRCh37.p13_genomic.fna.gz', '/Users/kexindong/Documents/GitHub/Database/RefGenome/mouse-2023-09-13/GCF_000001635.27_GRCm39_genomic.fna.gz'\n",
    "chrom_dict_h, i = prime.genome_loader(path_h_ref)\n",
    "chrom_dict_m, i = prime.genome_loader(path_m_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mutation_idx</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>tx_id_h</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>...</th>\n",
       "      <th>Protospacer_Location</th>\n",
       "      <th>Ideal_Edit_Window</th>\n",
       "      <th>Editor</th>\n",
       "      <th>gRNA_id</th>\n",
       "      <th>OnTarget_Azimuth_Score</th>\n",
       "      <th>sensor_wt</th>\n",
       "      <th>sensor_alt</th>\n",
       "      <th>sensor_orientation</th>\n",
       "      <th>sensor_error</th>\n",
       "      <th>contains_polyT_terminator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>GAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCAGGCC</td>\n",
       "      <td>GAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCAGGCC</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_1</td>\n",
       "      <td>54.0</td>\n",
       "      <td>TGGAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCAGG</td>\n",
       "      <td>TGGAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCAGG</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_2</td>\n",
       "      <td>59.0</td>\n",
       "      <td>CATGGAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCA</td>\n",
       "      <td>CATGGAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCA</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006799</td>\n",
       "      <td>62006799</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002100-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_3</td>\n",
       "      <td>59.0</td>\n",
       "      <td>ACAGCAAGGCTGGCATGGAGGAAGATCACACCTACGAGGT</td>\n",
       "      <td>ACAGCAAGGCTGGCATGGAGGAAGATCACACCCACGAGGT</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>CD79B</td>\n",
       "      <td>ENSMUST00000044228.11</td>\n",
       "      <td>62006798</td>\n",
       "      <td>62006798</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0002993-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_4</td>\n",
       "      <td>66.0</td>\n",
       "      <td>GCTGGGCCTGCCCCTCTCCTTACCTCGTAGGTGTGATCTT</td>\n",
       "      <td>GCTGGGCCTGCCCCTCTCCTTACCTCGCAGGTGTGATCTT</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15376</th>\n",
       "      <td>2850</td>\n",
       "      <td>RNF213</td>\n",
       "      <td>ENSMUST00000093902.12</td>\n",
       "      <td>78357600</td>\n",
       "      <td>78357600</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0003804-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>18</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_15376</td>\n",
       "      <td>38.0</td>\n",
       "      <td>GCAATGGACCACACTTTTCCGGGGCAGGTGGGGCAGGAAG</td>\n",
       "      <td>GCAATGGACCACACTTTCCCGGGGCAGGTGGGGCAGGAAG</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15377</th>\n",
       "      <td>2850</td>\n",
       "      <td>RNF213</td>\n",
       "      <td>ENSMUST00000093902.12</td>\n",
       "      <td>78357600</td>\n",
       "      <td>78357600</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0003804-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>16</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_15377</td>\n",
       "      <td>46.0</td>\n",
       "      <td>GAGCAATGGACCACACTTTTCCGGGGCAGGTGGGGCAGGA</td>\n",
       "      <td>GAGCAATGGACCACACTTTCCCGGGGCAGGTGGGGCAGGA</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15378</th>\n",
       "      <td>2850</td>\n",
       "      <td>RNF213</td>\n",
       "      <td>ENSMUST00000093902.12</td>\n",
       "      <td>78357600</td>\n",
       "      <td>78357600</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0003804-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>14</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_15378</td>\n",
       "      <td>60.0</td>\n",
       "      <td>TAGAGCAATGGACCACACTTTTCCGGGGCAGGTGGGGCAG</td>\n",
       "      <td>TAGAGCAATGGACCACACTTTCCCGGGGCAGGTGGGGCAG</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15379</th>\n",
       "      <td>2850</td>\n",
       "      <td>RNF213</td>\n",
       "      <td>ENSMUST00000093902.12</td>\n",
       "      <td>78357600</td>\n",
       "      <td>78357600</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0003804-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_15379</td>\n",
       "      <td>41.0</td>\n",
       "      <td>CCAAATCTTAGAGCAATGGACCACACTTTTCCGGGGCAGG</td>\n",
       "      <td>CCAAATCTTAGAGCAATGGACCACACTTTCCCGGGGCAGG</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15380</th>\n",
       "      <td>2850</td>\n",
       "      <td>RNF213</td>\n",
       "      <td>ENSMUST00000093902.12</td>\n",
       "      <td>78357600</td>\n",
       "      <td>78357600</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNP</td>\n",
       "      <td>AACR-H0003804-M01</td>\n",
       "      <td>17</td>\n",
       "      <td>...</td>\n",
       "      <td>13</td>\n",
       "      <td>False</td>\n",
       "      <td>ABE</td>\n",
       "      <td>gRNA_15380</td>\n",
       "      <td>59.0</td>\n",
       "      <td>TTAGAGCAATGGACCACACTTTTCCGGGGCAGGTGGGGCA</td>\n",
       "      <td>TTAGAGCAATGGACCACACTTTCCCGGGGCAGGTGGGGCA</td>\n",
       "      <td>reverse-complement</td>\n",
       "      <td>No Error</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15381 rows Ã— 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       mutation_idx Hugo_Symbol                tx_id_h  Start_Position  \\\n",
       "0                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "1                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "2                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "3                 0       CD79B  ENSMUST00000044228.11        62006799   \n",
       "4                 1       CD79B  ENSMUST00000044228.11        62006798   \n",
       "...             ...         ...                    ...             ...   \n",
       "15376          2850      RNF213  ENSMUST00000093902.12        78357600   \n",
       "15377          2850      RNF213  ENSMUST00000093902.12        78357600   \n",
       "15378          2850      RNF213  ENSMUST00000093902.12        78357600   \n",
       "15379          2850      RNF213  ENSMUST00000093902.12        78357600   \n",
       "15380          2850      RNF213  ENSMUST00000093902.12        78357600   \n",
       "\n",
       "       End_Position Reference_Allele Tumor_Seq_Allele2 Variant_Type  \\\n",
       "0          62006799                A                 G          SNP   \n",
       "1          62006799                A                 G          SNP   \n",
       "2          62006799                A                 G          SNP   \n",
       "3          62006799                A                 G          SNP   \n",
       "4          62006798                T                 C          SNP   \n",
       "...             ...              ...               ...          ...   \n",
       "15376      78357600                A                 G          SNP   \n",
       "15377      78357600                A                 G          SNP   \n",
       "15378      78357600                A                 G          SNP   \n",
       "15379      78357600                A                 G          SNP   \n",
       "15380      78357600                A                 G          SNP   \n",
       "\n",
       "                      ID Chromosome  ...  Protospacer_Location  \\\n",
       "0      AACR-H0002100-M01         17  ...                    20   \n",
       "1      AACR-H0002100-M01         17  ...                    18   \n",
       "2      AACR-H0002100-M01         17  ...                    16   \n",
       "3      AACR-H0002100-M01         17  ...                     3   \n",
       "4      AACR-H0002993-M01         17  ...                     8   \n",
       "...                  ...        ...  ...                   ...   \n",
       "15376  AACR-H0003804-M01         17  ...                    18   \n",
       "15377  AACR-H0003804-M01         17  ...                    16   \n",
       "15378  AACR-H0003804-M01         17  ...                    14   \n",
       "15379  AACR-H0003804-M01         17  ...                     6   \n",
       "15380  AACR-H0003804-M01         17  ...                    13   \n",
       "\n",
       "       Ideal_Edit_Window Editor     gRNA_id OnTarget_Azimuth_Score  \\\n",
       "0                  False    ABE      gRNA_0                   44.0   \n",
       "1                  False    ABE      gRNA_1                   54.0   \n",
       "2                  False    ABE      gRNA_2                   59.0   \n",
       "3                  False    ABE      gRNA_3                   59.0   \n",
       "4                   True    ABE      gRNA_4                   66.0   \n",
       "...                  ...    ...         ...                    ...   \n",
       "15376              False    ABE  gRNA_15376                   38.0   \n",
       "15377              False    ABE  gRNA_15377                   46.0   \n",
       "15378              False    ABE  gRNA_15378                   60.0   \n",
       "15379               True    ABE  gRNA_15379                   41.0   \n",
       "15380              False    ABE  gRNA_15380                   59.0   \n",
       "\n",
       "                                      sensor_wt  \\\n",
       "0      GAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCAGGCC   \n",
       "1      TGGAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCAGG   \n",
       "2      CATGGAGGAAGATCACACCTACGAGGTAAGGAGAGGGGCA   \n",
       "3      ACAGCAAGGCTGGCATGGAGGAAGATCACACCTACGAGGT   \n",
       "4      GCTGGGCCTGCCCCTCTCCTTACCTCGTAGGTGTGATCTT   \n",
       "...                                         ...   \n",
       "15376  GCAATGGACCACACTTTTCCGGGGCAGGTGGGGCAGGAAG   \n",
       "15377  GAGCAATGGACCACACTTTTCCGGGGCAGGTGGGGCAGGA   \n",
       "15378  TAGAGCAATGGACCACACTTTTCCGGGGCAGGTGGGGCAG   \n",
       "15379  CCAAATCTTAGAGCAATGGACCACACTTTTCCGGGGCAGG   \n",
       "15380  TTAGAGCAATGGACCACACTTTTCCGGGGCAGGTGGGGCA   \n",
       "\n",
       "                                     sensor_alt  sensor_orientation  \\\n",
       "0      GAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCAGGCC  reverse-complement   \n",
       "1      TGGAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCAGG  reverse-complement   \n",
       "2      CATGGAGGAAGATCACACCCACGAGGTAAGGAGAGGGGCA  reverse-complement   \n",
       "3      ACAGCAAGGCTGGCATGGAGGAAGATCACACCCACGAGGT  reverse-complement   \n",
       "4      GCTGGGCCTGCCCCTCTCCTTACCTCGCAGGTGTGATCTT  reverse-complement   \n",
       "...                                         ...                 ...   \n",
       "15376  GCAATGGACCACACTTTCCCGGGGCAGGTGGGGCAGGAAG  reverse-complement   \n",
       "15377  GAGCAATGGACCACACTTTCCCGGGGCAGGTGGGGCAGGA  reverse-complement   \n",
       "15378  TAGAGCAATGGACCACACTTTCCCGGGGCAGGTGGGGCAG  reverse-complement   \n",
       "15379  CCAAATCTTAGAGCAATGGACCACACTTTCCCGGGGCAGG  reverse-complement   \n",
       "15380  TTAGAGCAATGGACCACACTTTCCCGGGGCAGGTGGGGCA  reverse-complement   \n",
       "\n",
       "       sensor_error  contains_polyT_terminator  \n",
       "0          No Error                      False  \n",
       "1          No Error                      False  \n",
       "2          No Error                      False  \n",
       "3          No Error                      False  \n",
       "4          No Error                      False  \n",
       "...             ...                        ...  \n",
       "15376      No Error                      False  \n",
       "15377      No Error                      False  \n",
       "15378      No Error                      False  \n",
       "15379      No Error                      False  \n",
       "15380      No Error                      False  \n",
       "\n",
       "[15381 rows x 34 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all = df_all[df_all['Chromosome']!='M'].reset_index(drop=True)\n",
    "df_all['Reference_Allele'] = [str(x) for x in df_all['Reference_Allele']]\n",
    "df_all['Tumor_Seq_Allele2'] = [str(x) for x in df_all['Tumor_Seq_Allele2']]\n",
    "df_all['Start_Position'] = [int(x) for x in df_all['Start_Position']]\n",
    "df_all['End_Position'] = [int(x) for x in df_all['End_Position']]\n",
    "df_be_h = base.run_base(df_all, 'cBioPortal', chrom_dict_h, PAM='NGN',filtration='ABE+CBE', ideal_edit_window=[4, 8], auto_SNP_filter=True, proto_size=19,context_size=120, RE_sites=None, polyT_threshold=4, before_proto_context=5,sensor_length=40, sensor_orientation='reverse-complement', sensor=True)\n",
    "df_be_h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AACR-H0002363-M01',\n",
       " 'AACR-H0002999-M01',\n",
       " 'AACR-H0002588-M01',\n",
       " 'AACR-H0003738-M01',\n",
       " 'AACR-H0002965-M01',\n",
       " 'AACR-H0005030-M01',\n",
       " 'AACR-H0000928-M01',\n",
       " 'AACR-H0000291-M01',\n",
       " 'AACR-H0001653-M01',\n",
       " 'AACR-H0004831-M01',\n",
       " 'AACR-H0004152-M01',\n",
       " 'AACR-H0003347-M01',\n",
       " 'AACR-H0000973-M01',\n",
       " 'AACR-H0000313-M01',\n",
       " 'AACR-H0001629-M01']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(set(df_all['ID'])-set(df_be_h['ID']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_be_h['ID'] = [x[:-4] for x in df_be_h['ID']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_be_h.to_csv('be_human_result',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# For Mouse  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all = pd.read_csv('/Users/kexindong/Documents/GitHub/Output/h2m_database/BEGG/be_mouse.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_h_ref, path_m_ref = '/Users/kexindong/Documents/GitHub/Database/RefGenome/ncbi-2023-09-12/GCF_000001405.25_GRCh37.p13_genomic.fna.gz', '/Users/kexindong/Documents/GitHub/Database/RefGenome/mouse-2023-09-13/GCF_000001635.27_GRCm39_genomic.fna.gz'\n",
    "# remember to replace the paths with yours\n",
    "records_h, index_list_h = h2m.genome_loader(path_h_ref)\n",
    "records_m, index_list_m  = h2m.genome_loader(path_m_ref)\n",
    "path_h_anno, path_m_anno = '/Users/kexindong/Documents/GitHub/Database/Genecode/gencode_v19_GRCh37.db', '/Users/kexindong/Documents/GitHub/Database/Genecode/gencode_vm33_GRCm39.db'\n",
    "# remember to replace the paths with yours\n",
    "db_h, db_m = h2m.anno_loader(path_h_anno), h2m.anno_loader(path_m_anno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No error occurs.\n",
      "No error occurs.\n",
      "There were rows that could not be processed.\n"
     ]
    }
   ],
   "source": [
    "df_all.columns = ['gene_name_h','tx_id_h','start_h','end_h','ref_seq_h','alt_seq_h','type_h','index','chr_h']\n",
    "df_all = h2m.query_batch(df_all)[0]\n",
    "df_all = h2m.get_tx_batch(df_all, 'm')[0]\n",
    "df_result, df_fail = h2m.model_batch(df_all, records_h, index_list_h, records_m, index_list_m, db_h, db_m, 37, param='BE', memory_size=2000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'chrom_dict_m' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[40], line 6\u001b[0m\n\u001b[1;32m      4\u001b[0m df_all[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mStart_Position\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m [\u001b[38;5;28mint\u001b[39m(x) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m df_all[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mStart_Position\u001b[39m\u001b[38;5;124m'\u001b[39m]]\n\u001b[1;32m      5\u001b[0m df_all[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mEnd_Position\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m [\u001b[38;5;28mint\u001b[39m(x) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m df_all[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mEnd_Position\u001b[39m\u001b[38;5;124m'\u001b[39m]]\n\u001b[0;32m----> 6\u001b[0m df_be_h \u001b[38;5;241m=\u001b[39m base\u001b[38;5;241m.\u001b[39mrun_base(df_all, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mcBioPortal\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[43mchrom_dict_m\u001b[49m, PAM\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNGN\u001b[39m\u001b[38;5;124m'\u001b[39m,filtration\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mABE+CBE\u001b[39m\u001b[38;5;124m'\u001b[39m, ideal_edit_window\u001b[38;5;241m=\u001b[39m[\u001b[38;5;241m4\u001b[39m, \u001b[38;5;241m8\u001b[39m], auto_SNP_filter\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, proto_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m19\u001b[39m,context_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m120\u001b[39m, RE_sites\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, polyT_threshold\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m4\u001b[39m, before_proto_context\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m5\u001b[39m,sensor_length\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m40\u001b[39m, sensor_orientation\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mreverse-complement\u001b[39m\u001b[38;5;124m'\u001b[39m, sensor\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m      7\u001b[0m df_be_h\n",
      "\u001b[0;31mNameError\u001b[0m: name 'chrom_dict_m' is not defined"
     ]
    }
   ],
   "source": [
    "df_all = df_all[df_all['Chromosome']!='M'].reset_index(drop=True)\n",
    "df_all['Reference_Allele'] = [str(x) for x in df_all['Reference_Allele']]\n",
    "df_all['Tumor_Seq_Allele2'] = [str(x) for x in df_all['Tumor_Seq_Allele2']]\n",
    "df_all['Start_Position'] = [int(x) for x in df_all['Start_Position']]\n",
    "df_all['End_Position'] = [int(x) for x in df_all['End_Position']]\n",
    "df_be_h = base.run_base(df_all, 'cBioPortal', chrom_dict_m, PAM='NGN',filtration='ABE+CBE', ideal_edit_window=[4, 8], auto_SNP_filter=True, proto_size=19,context_size=120, RE_sites=None, polyT_threshold=4, before_proto_context=5,sensor_length=40, sensor_orientation='reverse-complement', sensor=True)\n",
    "df_be_h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
